<?xml version="1.0" ?>
<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.0 Transitional//EN">
<html>
<head>
<!-- saved from url=(0017)http://localhost/ -->
<script language="JavaScript" src="../displayToc.js"></script>
<script language="JavaScript" src="../tocParas.js"></script>
<script language="JavaScript" src="../tocTab.js"></script>
<link rel="stylesheet" type="text/css" href="../scineplex.css">
<title></title>
<link rel="stylesheet" href="../Active.css" type="text/css" />
<meta http-equiv="content-type" content="text/html; charset=utf-8" />
<link rev="made" href="mailto:support@ActiveState.com" />
</head>

<body>



<ul id="index">
  <li><a href="#NAME">NAME</a></li>
  <li><a href="#SYNOPSIS">SYNOPSIS</a></li>
  <li><a href="#DESCRIPTION">DESCRIPTION</a>
    <ul>
      <li><a href="#NOTES">NOTES</a></li>
    </ul>
  </li>
  <li><a href="#COMMAND-LINE-OPTIONS">COMMAND-LINE OPTIONS</a></li>
  <li><a href="#SEE-ALSO">SEE ALSO</a></li>
  <li><a href="#AUTHOR">AUTHOR</a></li>
</ul>

<h1 id="NAME">NAME</h1>

<p>bp_fast_load_gff.pl - Fast-load a Bio::DB::GFF database from GFF files.</p>

<h1 id="SYNOPSIS">SYNOPSIS</h1>

<pre><code><code>  % bp_fast_load_gff.pl -d testdb dna1.fa dna2.fa features1.gff features2.gff ...</code></code></pre>

<h1 id="DESCRIPTION">DESCRIPTION</h1>

<p>This script loads a Bio::DB::GFF database with the features contained in a list of GFF files and/or FASTA sequence files. You must use the exact variant of GFF described in <a href="../../.cpan/build/BioPerl-1.6.901-ghKY8U/blib/lib/Bio/DB/GFF.html">Bio::DB::GFF</a>. Various command-line options allow you to control which database to load and whether to allow an existing database to be overwritten.</p>

<p>This script is similar to load_gff.pl, but is much faster. However, it is hard-coded to use MySQL and probably only works on Unix platforms due to its reliance on pipes. See <a>bp_load_gff.pl</a> for an incremental loader that works with all databases supported by Bio::DB::GFF, and <a>bp_bulk_load_gff.pl</a> for a fast MySQL loader that supports all platforms.</p>

<h2 id="NOTES">NOTES</h2>

<p>If the filename is given as &quot;-&quot; then the input is taken from standard input. Compressed files (.gz, .Z, .bz2) are automatically uncompressed.</p>

<p>FASTA format files are distinguished from GFF files by their filename extensions. Files ending in .fa, .fasta, .fast, .seq, .dna and their uppercase variants are treated as FASTA files. Everything else is treated as a GFF file. If you wish to load -fasta files from STDIN, then use the -f command-line swith with an argument of &#39;-&#39;, as in</p>

<pre><code><code>    gunzip my_data.fa.gz | bp_fast_load_gff.pl -d test -f -</code></code></pre>

<p>The nature of the load requires that the database be on the local machine and that the indicated user have the &quot;file&quot; privilege to load the tables and have enough room in /usr/tmp (or whatever is specified by the \$TMPDIR environment variable), to hold the tables transiently. If your MySQL is version 3.22.6 and was compiled using the &quot;load local file&quot; option, then you may be able to load remote databases with local data using the --local option.</p>

<p>About maxfeature: the default value is 100,000,000 bases. If you have features that are close to or greater that 100Mb in length, then the value of maxfeature should be increased to 1,000,000,000. This value must be a power of 10.</p>

<p>If the list of GFF or fasta files exceeds the kernel limit for the maximum number of command-line arguments, use the --long_list /path/to/files option.</p>

<p>The adaptor used is dbi::mysqlopt. There is currently no way to change this.</p>

<h1 id="COMMAND-LINE-OPTIONS">COMMAND-LINE OPTIONS</h1>

<p>Command-line options can be abbreviated to single-letter options. e.g. -d instead of --database.</p>

<pre><code><code>   --database &lt;dsn&gt;      Mysql database name
   --create              Reinitialize/create data tables without asking
   --local               Try to load a remote database using local data.
   --user                Username to log in as
   --fasta               File or directory containing fasta files to load
   --password            Password to use for authentication
   --long_list           Directory containing a very large number of
                         GFF and/or FASTA files
   --maxfeature          Set the value of the maximum feature size (default 100Mb; must be a power of 10)
   --group               A list of one or more tag names (comma or space separated)
                         to be used for grouping in the 9th column.
   --gff3_munge          Activate GFF3 name munging (see Bio::DB::GFF)
   --summary             Generate summary statistics for drawing coverage histograms.
                           This can be run on a previously loaded database or during
                           the load.
   --Temporary           Location of a writable scratch directory</code></code></pre>

<h1 id="SEE-ALSO">SEE ALSO</h1>

<p><a href="../../.cpan/build/BioPerl-1.6.901-ghKY8U/blib/lib/Bio/DB/GFF.html">Bio::DB::GFF</a>, <a>bulk_load_gff.pl</a>, <a>load_gff.pl</a></p>

<h1 id="AUTHOR">AUTHOR</h1>

<p>Lincoln Stein, lstein@cshl.org</p>

<p>Copyright (c) 2002 Cold Spring Harbor Laboratory</p>

<p>This library is free software; you can redistribute it and/or modify it under the same terms as Perl itself. See DISCLAIMER.txt for disclaimers of warranty.</p>


</body>

</html>


