<?xml version="1.0" ?>
<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.0 Transitional//EN">
<html>
<head>
<!-- saved from url=(0017)http://localhost/ -->
<script language="JavaScript" src="../displayToc.js"></script>
<script language="JavaScript" src="../tocParas.js"></script>
<script language="JavaScript" src="../tocTab.js"></script>
<link rel="stylesheet" type="text/css" href="../scineplex.css">
<title></title>
<link rel="stylesheet" href="../Active.css" type="text/css" />
<meta http-equiv="content-type" content="text/html; charset=utf-8" />
<link rev="made" href="mailto:support@ActiveState.com" />
</head>

<body>



<ul id="index">
  <li><a href="#NAME">NAME</a></li>
  <li><a href="#SYNOPSIS">SYNOPSIS</a></li>
  <li><a href="#DESCRIPTION">DESCRIPTION</a></li>
  <li><a href="#OPTIONS">OPTIONS</a></li>
  <li><a href="#OUTPUT-FORMAT">OUTPUT FORMAT</a></li>
  <li><a href="#EXAMPLE">EXAMPLE</a></li>
  <li><a href="#TODO">TODO</a>
    <ul>
      <li><a href="#Mailing-Lists">Mailing Lists</a></li>
      <li><a href="#Reporting-Bugs">Reporting Bugs</a></li>
    </ul>
  </li>
  <li><a href="#AUTHOR---Heikki-Lehvaslaiho">AUTHOR - Heikki Lehvaslaiho</a></li>
</ul>

<h1 id="NAME">NAME</h1>

<p>flanks - finding flanking sequences for a variant in a sequence position</p>

<h1 id="SYNOPSIS">SYNOPSIS</h1>

<pre><code><code>  flanks --position POS [-p POS ...]  [--flanklen INT]
         accession | filename</code></code></pre>

<h1 id="DESCRIPTION">DESCRIPTION</h1>

<p>This script allows you to extract a subsequence around a region of interest from an existing sequence. The output if fasta formatted sequence entry where the header line contains additional information about the location.</p>

<h1 id="OPTIONS">OPTIONS</h1>

<p>The script takes one unnamed argument which be either a file name in the local file system or a nucleotide sequence accession number.</p>

<pre><code><code>  -p         Position uses simple nucleotide sequence feature table
  --position notation to define the region of interest, typically a
             SNP or microsatellite repeat around which the flanks are
             defined.

             There can be more than one position option or you can
             give a comma separated list to one position option.

             The format of a position is:

                 [id:] int | range | in-between [-]

             The optional id is the name you want to call the new
             sequence. If it not given in joins running number to the
             entry name with an underscore.

             The position is either a point (e.g. 234), a range (e.g
             250..300) or insertion point between nucleotides
             (e.g. 234^235)

             If the position is not completely within the source
             sequence the output sequence will be truncated and it
             will print a warning.

             The optional hyphen [-] at the end of the position
             indicates that that you want the retrieved sequence to be
             in the opposite strand.


  -f         Defaults to 100. This is the length of the nucleotides
  --flanklen sequence retrieved on both sides of the given position.

             If the source file does not contain </code></code></pre>

<h1 id="OUTPUT-FORMAT">OUTPUT FORMAT</h1>

<p>The output is a fasta formatted entry where the description file contains tag=value pairs for information about where in the original sequence the subsequence was taken.</p>

<p>The ID of the fasta entry is the name given at the command line joined by hyphen to the filename or accesion of the source sequence. If no id is given a series of consequtive integers is used.</p>

<p>The tag=value pairs are:</p>

<dl>

<dt id="oripos-int">oripos=int</dt>
<dd>

<p>position in the source file</p>

</dd>
<dt id="strand-1--1">strand=1|-1</dt>
<dd>

<p>strand of the sequence compared to the source sequence</p>

</dd>
<dt id="allelepos-int">allelepos=int</dt>
<dd>

<p>position of the region of interest in the current entry. The tag is the same as used by dbSNP@NCBI</p>

</dd>
</dl>

<p>The sequence highlights the allele variant position by showing it in upper case and rest of the sequence in lower case characters.</p>

<h1 id="EXAMPLE">EXAMPLE</h1>

<pre><code><code>  % flanks ~/seq/ar.embl

  &gt;1_/HOME/HEIKKI/SEQ/AR.EMBL oripos=100 strand=1 allelepos=100
  taataactcagttcttatttgcacctacttcagtggacactgaatttggaaggtggagga
  ttttgtttttttcttttaagatctgggcatcttttgaatCtacccttcaagtattaagag
  acagactgtgagcctagcagggcagatcttgtccaccgtgtgtcttcttctgcacgagac
  tttgaggctgtcagagcgct</code></code></pre>

<h1 id="TODO">TODO</h1>

<p>The input files are assumed to be in EMBL format and the sequences are retrieved only from the EMB database. Make this more generic and use the registry.</p>

<p>head1 FEEDBACK</p>

<h2 id="Mailing-Lists">Mailing Lists</h2>

<p>User feedback is an integral part of the evolution of this and other Bioperl modules. Send your comments and suggestions preferably to the Bioperl mailing lists Your participation is much appreciated.</p>

<pre><code><code>  bioperl-l@bioperl.org                  - General discussion
  http://bioperl.org/wiki/Mailing_lists  - About the mailing lists</code></code></pre>

<h2 id="Reporting-Bugs">Reporting Bugs</h2>

<p>Report bugs to the Bioperl bug tracking system to help us keep track the bugs and their resolution. Bug reports can be submitted via the web:</p>

<pre><code><code>  https://redmine.open-bio.org/projects/bioperl/</code></code></pre>

<h1 id="AUTHOR---Heikki-Lehvaslaiho">AUTHOR - Heikki Lehvaslaiho</h1>

<p>Email: &lt;heikki-at-bioperl-dot-org&gt;</p>


</body>

</html>


