<?xml version="1.0" ?>
<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.0 Transitional//EN">
<html>
<head>
<!-- saved from url=(0017)http://localhost/ -->
<script language="JavaScript" src="../displayToc.js"></script>
<script language="JavaScript" src="../tocParas.js"></script>
<script language="JavaScript" src="../tocTab.js"></script>
<link rel="stylesheet" type="text/css" href="../scineplex.css">
<title></title>
<link rel="stylesheet" href="../Active.css" type="text/css" />
<meta http-equiv="content-type" content="text/html; charset=utf-8" />
<link rev="made" href="mailto:support@ActiveState.com" />
</head>

<body>



<ul id="index">
  <li><a href="#NAME">NAME</a></li>
  <li><a href="#SYNOPSIS">SYNOPSIS</a></li>
  <li><a href="#DESCRIPTION">DESCRIPTION</a>
    <ul>
      <li><a href="#Notes">Notes</a>
        <ul>
          <li><a href="#split-and-nolump-produce-many-files">&#39;split&#39; and &#39;nolump&#39; produce many files</a></li>
          <li><a href="#Designed-for-RefSeq">Designed for RefSeq</a></li>
          <li><a href="#G-R-P-E-Gene-Model">G-R-P-E Gene Model</a></li>
        </ul>
      </li>
    </ul>
  </li>
  <li><a href="#TODO">TODO</a>
    <ul>
      <li><a href="#Are-these-additions-desired-">Are these additions desired?</a></li>
      <li><a href="#Related-bugfixes-tests">Related bugfixes/tests</a></li>
    </ul>
  </li>
  <li><a href="#AUTHOR">AUTHOR</a>
    <ul>
      <li><a href="#AUTHOR-of-hacks-for-GFF2Chado-loading">AUTHOR of hacks for GFF2Chado loading</a></li>
    </ul>
  </li>
</ul>

<h1 id="NAME">NAME</h1>

<p>genbank2gff3.pl -- Genbank-&gt;gbrowse-friendly GFF3</p>

<h1 id="SYNOPSIS">SYNOPSIS</h1>

<pre><code><code>  <span class="variable">genbank2gff3</span><span class="operator">.</span><span class="variable">pl</span> <span class="operator">[</span><span class="variable">options</span><span class="operator">]</span> <span class="variable">filename</span><span class="operator">(</span><span class="regex">s)
  
  # process a directory containing GenBank flatfiles
  perl genbank2gff3.pl --dir path_to_files --zip
  
  # process a single file, ignore explicit exons and introns
  perl genbank2gff3.pl --filter exon --filter intron file.gbk.gz
  
  # process a list of files 
  perl genbank2gff3.pl *gbk.gz
  
  # process data from URL, with Chado GFF model (-noCDS), and pipe to database loader
  curl ftp://ftp.ncbi.nih.gov/genomes/Saccharomyces_cerevisiae/CHR_X/NC_001142.gbk \
  | perl genbank2gff3.pl -noCDS -in stdin -out stdout \
  | perl gmod_bulk_load_gff3.pl -dbname mychado -organism fromdata
  
    Options:
        --noinfer  -r  don't infer exon/mRNA subfeatures
        --conf     -i  path to the curation configuration file that contains user preferences
                       for Genbank entries (must be YAML format)</span>
                       <span class="operator">(</span><span class="keyword">if</span> <span class="operator">--</span><span class="variable">manual</span> <span class="variable">is</span> <span class="variable">passed</span> <span class="variable">without</span> <span class="operator">--</span><span class="variable">ini</span><span class="operator">,</span> <span class="variable">user</span> <span class="variable">will</span> <span class="variable">be</span> <span class="variable">prompted</span> <span class="variable">to</span> 
                        <span class="variable">create</span> <span class="variable">the</span> <span class="variable">file</span> <span class="keyword">if</span> <span class="variable">any</span> <span class="variable">manual</span> <span class="variable">input</span> <span class="variable">is</span> <span class="variable">saved</span><span class="operator">)</span>
        <span class="operator">--</span><span class="variable">sofile</span>  <span class="keyword">-l</span>  <span class="variable">path</span> <span class="variable">to</span> <span class="variable">to</span> <span class="variable">the</span> <span class="variable">so</span><span class="operator">.</span><span class="variable">obo</span> <span class="variable">file</span> <span class="variable">to</span> <span class="keyword">use</span> <span class="keyword">for</span> <span class="variable">feature</span> <span class="variable">type</span> <span class="variable">mapping</span>
                       <span class="operator">(--</span><span class="variable">sofile</span> <span class="variable">live</span> <span class="variable">will</span> <span class="variable">download</span> <span class="variable">the</span> <span class="variable">latest</span> <span class="variable">online</span> <span class="variable">revision</span><span class="operator">)</span>
        <span class="operator">--</span><span class="variable">manual</span>   <span class="operator">-</span><span class="regex">m  when trying to guess the proper SO term, if more than
                       one option matches the primary tag, the converter will</span> 
                       <span class="keyword">wait</span> <span class="keyword">for</span> <span class="variable">user</span> <span class="variable">input</span> <span class="variable">to</span> <span class="variable">choose</span> <span class="variable">the</span> <span class="variable">correct</span> <span class="variable">one</span>
                       <span class="operator">(</span><span class="variable">only</span> <span class="variable">works</span> <span class="variable">with</span> <span class="operator">--</span><span class="variable">sofile</span><span class="operator">)</span>
        <span class="operator">--</span><span class="variable">dir</span>      <span class="keyword">-d</span>  <span class="variable">path</span> <span class="variable">to</span> <span class="variable">a</span> <span class="variable">list</span> <span class="variable">of</span> <span class="variable">genbank</span> <span class="variable">flatfiles</span>
        <span class="operator">--</span><span class="variable">outdir</span>   <span class="keyword">-o</span>  <span class="variable">location</span> <span class="variable">to</span> <span class="keyword">write</span> <span class="variable">GFF</span> <span class="variable">files</span> <span class="operator">(</span><span class="variable">can</span> <span class="variable">be</span> <span class="string">'stdout'</span> <span class="keyword">or</span> <span class="string">'-'</span> <span class="keyword">for</span> <span class="keyword">pipe</span><span class="operator">)</span>
        <span class="operator">--</span><span class="variable">zip</span>      <span class="keyword">-z</span>  <span class="variable">compress</span> <span class="variable">GFF3</span> <span class="variable">output</span> <span class="variable">files</span> <span class="variable">with</span> <span class="variable">gzip</span>
        <span class="operator">--</span><span class="variable">summary</span>  <span class="keyword">-s</span>  <span class="keyword">print</span> <span class="variable">a</span> <span class="variable">summary</span> <span class="variable">of</span> <span class="variable">the</span> <span class="variable">features</span> <span class="variable">in</span> <span class="keyword">each</span> <span class="variable">contig</span>
        <span class="operator">--</span><span class="variable">filter</span>   <span class="keyword">-x</span>  <span class="variable">genbank</span> <span class="variable">feature</span> <span class="variable">type</span><span class="operator">(</span><span class="regex">s) to ignore
        --split    -y  split output to separate GFF and fasta files for
                       each genbank record
        --nolump   -n  separate file for each reference sequence
                       (default is to lump all records together into one 
                       output file for each input file)
        --ethresh  -e  error threshold for unflattener
                       set this high (&gt;2)</span> <span class="variable">to</span> <span class="variable">ignore</span> <span class="variable">all</span> <span class="variable">unflattener</span> <span class="variable">errors</span>
        <span class="operator">--</span><span class="operator">[</span><span class="keyword">no</span><span class="operator">]</span><span class="variable">CDS</span>  <span class="keyword">-c</span>  <span class="variable">Keep</span> <span class="variable">CDS</span><span class="operator">-</span><span class="variable">exons</span><span class="operator">,</span> <span class="keyword">or</span> <span class="variable">convert</span> <span class="variable">to</span> <span class="variable">alternate</span> <span class="variable">gene</span><span class="operator">-</span><span class="variable">RNA</span><span class="operator">-</span><span class="variable">protein</span><span class="operator">-</span><span class="variable">exon</span> 
                       <span class="variable">model</span><span class="operator">.</span> <span class="operator">--</span><span class="variable">CDS</span> <span class="variable">is</span> <span class="keyword">default</span><span class="operator">.</span> <span class="variable">Use</span> <span class="operator">--</span><span class="variable">CDS</span> <span class="variable">to</span> <span class="variable">keep</span> <span class="keyword">default</span> <span class="variable">GFF</span> <span class="variable">gene</span> <span class="variable">model</span><span class="operator">,</span> 
                       <span class="keyword">use</span> <span class="operator">--</span><span class="variable">noCDS</span> <span class="variable">to</span> <span class="variable">convert</span> <span class="variable">to</span> <span class="variable">g</span><span class="keyword">-r-p-e</span><span class="operator">.</span>
        <span class="operator">--</span><span class="keyword">format</span>   <span class="keyword">-f</span>  <span class="variable">Input</span> <span class="keyword">format</span> <span class="operator">(</span><span class="variable">SeqIO</span> <span class="variable">types</span><span class="operator">):</span> <span class="variable">GenBank</span><span class="operator">,</span> <span class="variable">Swiss</span> <span class="keyword">or</span> <span class="variable">Uniprot</span><span class="operator">,</span> <span class="variable">EMBL</span> <span class="variable">work</span>
                       <span class="operator">(</span><span class="variable">GenBank</span> <span class="variable">is</span> <span class="keyword">default</span><span class="operator">)</span>
        <span class="operator">--</span><span class="variable">GFF_VERSION</span>  <span class="number">3</span> <span class="variable">is</span> <span class="keyword">default</span><span class="operator">,</span> <span class="number">2</span> <span class="keyword">and</span> <span class="number">2.5</span> <span class="keyword">and</span> <span class="variable">other</span> <span class="variable">Bio::Tools::GFF</span> <span class="variable">versions</span> <span class="variable">available</span>
        <span class="operator">--</span><span class="variable">quiet</span>        <span class="variable">don't</span> <span class="variable">talk</span> <span class="variable">about</span> <span class="variable">what</span> <span class="variable">is</span> <span class="variable">being</span> <span class="variable">processed</span> 
        <span class="operator">--</span><span class="variable">typesource</span>   <span class="variable">SO</span> <span class="variable">sequence</span> <span class="variable">type</span> <span class="keyword">for</span> <span class="variable">source</span> <span class="operator">(</span><span class="variable">e</span><span class="operator">.</span><span class="variable">g</span><span class="operator">.</span> <span class="variable">chromosome</span><span class="operator">;</span> <span class="variable">region</span><span class="operator">;</span> <span class="variable">contig</span><span class="operator">)</span>
        <span class="operator">--</span><span class="variable">help</span>     <span class="operator">-</span><span class="variable">h</span>  <span class="variable">display</span> <span class="variable">this</span> <span class="variable">message</span>
</code></code></pre>

<h1 id="DESCRIPTION">DESCRIPTION</h1>

<p>This script uses Bio::SeqFeature::Tools::Unflattener and Bio::Tools::GFF to convert GenBank flatfiles to GFF3 with gene containment hierarchies mapped for optimal display in gbrowse.</p>

<p>The input files are assumed to be gzipped GenBank flatfiles for refseq contigs. The files may contain multiple GenBank records. Either a single file or an entire directory can be processed. By default, the DNA sequence is embedded in the GFF but it can be saved into separate fasta file with the --split(-y) option.</p>

<p>If an input file contains multiple records, the default behaviour is to dump all GFF and sequence to a file of the same name (with .gff appended). Using the &#39;nolump&#39; option will create a separate file for each genbank record. Using the &#39;split&#39; option will create separate GFF and Fasta files for each genbank record.</p>

<h2 id="Notes">Notes</h2>

<h3 id="split-and-nolump-produce-many-files">&#39;split&#39; and &#39;nolump&#39; produce many files</h3>

<p>In cases where the input files contain many GenBank records (for example, the chromosome files for the mouse genome build), a very large number of output files will be produced if the &#39;split&#39; or &#39;nolump&#39; options are selected. If you do have lists of files &gt; 6000, use the --long_list option in bp_bulk_load_gff.pl or bp_fast_load_gff.pl to load the gff and/ or fasta files.</p>

<h3 id="Designed-for-RefSeq">Designed for RefSeq</h3>

<p>This script is designed for RefSeq genomic sequence entries. It may work for third party annotations but this has not been tested. But see below, Uniprot/Swissprot works, EMBL and possibly EMBL/Ensembl if you don&#39;t mind some gene model unflattener errors (dgg).</p>

<h3 id="G-R-P-E-Gene-Model">G-R-P-E Gene Model</h3>

<p>Don Gilbert worked this over with needs to produce GFF3 suited to loading to GMOD Chado databases. Most of the changes I believe are suited for general use. One main chado-specific addition is the --[no]cds2protein flag</p>

<p>My favorite GFF is to set the above as ON by default (disable with --nocds2prot) For general use it probably should be OFF, enabled with --cds2prot.</p>

<p>This writes GFF with an alternate, but useful Gene model, instead of the consensus model for GFF3</p>

<pre><code><code>  [ gene &gt; mRNA&gt; (exon,CDS,UTR) ]</code></code></pre>

<p>This alternate is</p>

<pre><code><code>  gene &gt; mRNA &gt; polypeptide &gt; exon </code></code></pre>

<p>means the only feature with dna bases is the exon. The others specify only location ranges on a genome. Exon of course is a child of mRNA and protein/peptide.</p>

<p>The protein/polypeptide feature is an important one, having all the annotations of the GenBank CDS feature, protein ID, translation, GO terms, Dbxrefs to other proteins.</p>

<p>UTRs, introns, CDS-exons are all inferred from the primary exon bases inside/outside appropriate higher feature ranges. Other special gene model features remain the same.</p>

<p>Several other improvements and bugfixes, minor but useful are included</p>

<pre><code><code>  * IO pipes now work:
    curl ftp://ncbigenomes/... | genbank2gff3 --in stdin --out stdout | gff2chado ...

  * GenBank main record fields are added to source feature, e.g. organism, date,
    and the sourcetype, commonly chromosome for  genomes, is used.

  * Gene Model handling for ncRNA, pseudogenes are added.

  * GFF header is cleaner, more informative.
    --GFF_VERSION flag allows choice of v2 as well as default v3

  * GFF ##FASTA inclusion is improved, and
    CDS translation sequence is moved to FASTA records.

  * FT -&gt; GFF attribute mapping is improved.

  * --format choice of SeqIO input formats (GenBank default). 
    Uniprot/Swissprot and EMBL work and produce useful GFF.

  * SeqFeature::Tools::TypeMapper has a few FT -&gt; SOFA additions
      and more flexible usage.</code></code></pre>

<h1 id="TODO">TODO</h1>

<h2 id="Are-these-additions-desired-">Are these additions desired?</h2>

<pre><code><code> * filter input records by taxon (e.g. keep only organism=xxx or taxa level = classYYY
 * handle Entrezgene, other non-sequence SeqIO structures (really should change
    those parsers to produce consistent annotation tags).</code></code></pre>

<h2 id="Related-bugfixes-tests">Related bugfixes/tests</h2>

<p>These items from Bioperl mail were tested (sample data generating errors), and found corrected:</p>

<pre><code><code> From: Ed Green &lt;green &lt;at&gt; eva.mpg.de&gt;
 Subject: genbank2gff3.pl on new human RefSeq
 Date: 2006-03-13 21:22:26 GMT 
   -- unspecified errors (sample data works now).

 From: Eric Just &lt;e-just &lt;at&gt; northwestern.edu&gt;
 Subject: bp_genbank2gff3.pl
 Date: 2007-01-26 17:08:49 GMT
   -- bug fixed in genbank2gff3 for multi-record handling</code></code></pre>

<p>This error is for a /trans_splice gene that is hard to handle, and unflattner/genbank2 doesn&#39;t</p>

<pre><code><code> From: Chad Matsalla &lt;chad &lt;at&gt; dieselwurks.com&gt; 
 Subject: genbank2gff3.PLS and the unflatenner - Inconsistent   order?
 Date: 2005-07-15 19:51:48 GMT </code></code></pre>

<h1 id="AUTHOR">AUTHOR</h1>

<p>Sheldon McKay (mckays@cshl.edu)</p>

<p>Copyright (c) 2004 Cold Spring Harbor Laboratory.</p>

<h2 id="AUTHOR-of-hacks-for-GFF2Chado-loading">AUTHOR of hacks for GFF2Chado loading</h2>

<p>Don Gilbert (gilbertd@indiana.edu)</p>


</body>

</html>


