<?xml version="1.0" ?>
<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.0 Transitional//EN">
<html>
<head>
<!-- saved from url=(0017)http://localhost/ -->
<script language="JavaScript" src="../displayToc.js"></script>
<script language="JavaScript" src="../tocParas.js"></script>
<script language="JavaScript" src="../tocTab.js"></script>
<link rel="stylesheet" type="text/css" href="../scineplex.css">
<title></title>
<link rel="stylesheet" href="../Active.css" type="text/css" />
<meta http-equiv="content-type" content="text/html; charset=utf-8" />
<link rev="made" href="mailto:support@ActiveState.com" />
</head>

<body>



<ul id="index">
  <li><a href="#NAME">NAME</a></li>
  <li><a href="#SYNOPSIS">SYNOPSIS</a></li>
  <li><a href="#DESCRIPTION">DESCRIPTION</a></li>
  <li><a href="#OPTIONS">OPTIONS</a></li>
  <li><a href="#FEEDBACK">FEEDBACK</a>
    <ul>
      <li><a href="#Mailing-Lists">Mailing Lists</a></li>
      <li><a href="#Reporting-Bugs">Reporting Bugs</a></li>
    </ul>
  </li>
  <li><a href="#AUTHOR---Charles-C.-Kim">AUTHOR - Charles C. Kim</a></li>
  <li><a href="#HISTORY">HISTORY</a></li>
</ul>

<h1 id="NAME">NAME</h1>

<p>oligo_count - oligo count and frequency</p>

<h1 id="SYNOPSIS">SYNOPSIS</h1>

<pre><code><code>  Usage:  oligo_count [-h/--help] [-l/--length OLIGOLENGTH]
          [-f/--format SEQFORMAT] [-i/--in/-s/--sequence SEQFILE]
          [-o/--out OUTFILE]</code></code></pre>

<h1 id="DESCRIPTION">DESCRIPTION</h1>

<p>This scripts counts occurrence and frequency for all oligonucleotides of given length.</p>

<p>It can be used to determine what primers are useful for frequent priming of nucleic acid for random labeling.</p>

<p>Note that this script could be run by utilizing the compseq program which is part of EMBOSS.</p>

<h1 id="OPTIONS">OPTIONS</h1>

<p>The default sequence format is fasta. If no outfile is given, the results will be printed to standard out. All other options can entered interactively.</p>

<h1 id="FEEDBACK">FEEDBACK</h1>

<h2 id="Mailing-Lists">Mailing Lists</h2>

<p>User feedback is an integral part of the evolution of this and other Bioperl modules. Send your comments and suggestions preferably to the Bioperl mailing list. Your participation is much appreciated.</p>

<pre><code><code>  bioperl-l@bioperl.org                  - General discussion
  http://bioperl.org/wiki/Mailing_lists  - About the mailing lists</code></code></pre>

<h2 id="Reporting-Bugs">Reporting Bugs</h2>

<p>Report bugs to the Bioperl bug tracking system to help us keep track of the bugs and their resolution. Bug reports can be submitted via the web:</p>

<pre><code><code>  https://redmine.open-bio.org/projects/bioperl/</code></code></pre>

<h1 id="AUTHOR---Charles-C.-Kim">AUTHOR - Charles C. Kim</h1>

<p>Email cckim@stanford.edu</p>

<h1 id="HISTORY">HISTORY</h1>

<p>Written July 2, 2001</p>

<p>Submitted to bioperl scripts project 2001/08/06</p>

<p>&gt;&gt; 100 x speed optimization by Heikki Lehvaslaiho</p>


</body>

</html>


