<?xml version="1.0" ?>
<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.0 Transitional//EN">
<html>
<head>
<!-- saved from url=(0017)http://localhost/ -->
<script language="JavaScript" src="../displayToc.js"></script>
<script language="JavaScript" src="../tocParas.js"></script>
<script language="JavaScript" src="../tocTab.js"></script>
<link rel="stylesheet" type="text/css" href="../scineplex.css">
<title></title>
<link rel="stylesheet" href="../Active.css" type="text/css" />
<meta http-equiv="content-type" content="text/html; charset=utf-8" />
<link rev="made" href="mailto:support@ActiveState.com" />
</head>

<body>



<ul id="index">
  <li><a href="#NAME">NAME</a></li>
  <li><a href="#SYNOPSIS">SYNOPSIS</a></li>
  <li><a href="#DESCRIPTION">DESCRIPTION</a></li>
  <li><a href="#SEE-ALSO">SEE ALSO</a></li>
  <li><a href="#AUTHOR">AUTHOR</a></li>
</ul>

<h1 id="NAME">NAME</h1>

<p>process_gadfly.pl - Massage Gadfly/FlyBase GFF files into a version suitable for the Generic Genome Browser</p>

<h1 id="SYNOPSIS">SYNOPSIS</h1>

<pre><code><code>  % process_gadfly.pl ./RELEASE2 &gt; gadfly.gff</code></code></pre>

<h1 id="DESCRIPTION">DESCRIPTION</h1>

<p>This script massages the RELEASE 3 Flybase/Gadfly GFF files located at http://www.fruitfly.org/sequence/release3download.shtml into the &quot;correct&quot; version of the GFF format.</p>

<p>To use this script, download the whole genome FASTA file and save it to disk. (The downloaded file will be called something like &quot;na_whole-genome_genomic_dmel_RELEASE3.FASTA&quot;, but the link on the HTML page doesn&#39;t give the filename.) Do the same for the whole genome GFF annotation file (the saved file will be called something like &quot;whole-genome_annotation-feature-region_dmel_RELEASE3.GFF&quot;.) If you wish you can download the ZIP compressed versions of these files.</p>

<p>Next run this script on the two files, indicating the name of the downloaded FASTA file first, followed by the gff file:</p>

<pre><code><code> % process_gadfly.pl na_whole-genome_genomic_dmel_RELEASE3.FASTA whole-genome_annotation-feature-region_dmel_RELEASE3.GFF &gt; fly.gff</code></code></pre>

<p>The gadfly.gff file and the fasta file can now be loaded into a Bio::DB::GFF database using the following command:</p>

<pre><code><code>  % bulk_load_gff.pl -d fly -fasta na_whole-genome_genomic_dmel_RELEASE3.FASTA fly.gff </code></code></pre>

<p>(Where &quot;fly&quot; is the name of the database. Change it as appropriate. The database must already exist and be writable by you!)</p>

<p>The resulting database will have the following feature types (represented as &quot;method:source&quot;):</p>

<pre><code><code>  Component:arm              A chromosome arm
  Component:scaffold         A chromosome scaffold (accession #)
  Component:gap              A gap in the assembly
  clone:clonelocator         A BAC clone
  gene:gadfly                A gene accession number
  transcript:gadfly          A transcript accession number
  translation:gadfly         A translation
  codon:gadfly               Significance unknown
  exon:gadfly                An exon
  symbol:gadfly              A classical gene symbol
  similarity:blastn          A BLASTN hit
  similarity:blastx          A BLASTX hit
  similarity:sim4            EST-&gt;genome using SIM4
  similarity:groupest        EST-&gt;genome using GROUPEST
  similarity:repeatmasker    A repeat</code></code></pre>

<p>IMPORTANT NOTE: This script will *only* work with the RELEASE3 gadfly files and will not work with earlier releases.</p>

<h1 id="SEE-ALSO">SEE ALSO</h1>

<p><a href="../../.cpan/build/BioPerl-1.6.901-ghKY8U/blib/lib/Bio/DB/GFF.html">Bio::DB::GFF</a>, <a>bulk_load_gff.pl</a>, <a>load_gff.pl</a></p>

<h1 id="AUTHOR">AUTHOR</h1>

<p>Lincoln Stein, lstein@cshl.org</p>

<p>Copyright (c) 2002 Cold Spring Harbor Laboratory</p>

<p>This library is free software; you can redistribute it and/or modify it under the same terms as Perl itself. See DISCLAIMER.txt for disclaimers of warranty.</p>


</body>

</html>


