<?xml version="1.0" ?>
<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.0 Transitional//EN">
<html>
<head>
<!-- saved from url=(0017)http://localhost/ -->
<script language="JavaScript" src="../displayToc.js"></script>
<script language="JavaScript" src="../tocParas.js"></script>
<script language="JavaScript" src="../tocTab.js"></script>
<link rel="stylesheet" type="text/css" href="../scineplex.css">
<title></title>
<link rel="stylesheet" href="../Active.css" type="text/css" />
<meta http-equiv="content-type" content="text/html; charset=utf-8" />
<link rev="made" href="mailto:support@ActiveState.com" />
</head>

<body>



<ul id="index">
  <li><a href="#NAME">NAME</a></li>
  <li><a href="#SYNOPSIS">SYNOPSIS</a></li>
  <li><a href="#DESCRIPTION">DESCRIPTION</a></li>
  <li><a href="#SEE-ALSO">SEE ALSO</a></li>
  <li><a href="#AUTHOR">AUTHOR</a></li>
</ul>

<h1 id="NAME">NAME</h1>

<p>process_wormbase.pl - Massage WormBase GFF files into a version suitable for the Generic Genome Browser</p>

<h1 id="SYNOPSIS">SYNOPSIS</h1>

<pre><code><code>  % process_wormbase.pl ./WS61 &gt; wormbase.gff</code></code></pre>

<h1 id="DESCRIPTION">DESCRIPTION</h1>

<p>This script massages the Wormbase GFF files located at ftp://www.wormbase.org/pub/wormbase/GENE_DUMPS into a version of the GFF format suitable for display by the generic genome browser. It mainly adds comments to the annotations and designates certain well-spaced genetic loci as framework landmarks.</p>

<p>This script requires the AcePerl distribution, which is available on CPAN (look for the &quot;Ace&quot; module).</p>

<p>To use this script, get the WormBase GFF files from the FTP site listed above and place them in a directory. It might be a good idea to name the directory after the current release, such as WS61. You do not need to uncompress the files.</p>

<p>Then give that directory as the argument to this script and capture the script&#39;s output to a file:</p>

<pre><code><code>  % process_wormbase.pl ./WS61 &gt; wormbase.gff</code></code></pre>

<p>It may take a while before you see output from this script, since it must first fetch gene and protein database from the remote AceDB running at www.wormbase.org. The wormbase.gff file can then be loaded into a Bio::DB::GFF database using the following command:</p>

<pre><code><code>  % bulk_load_gff.pl -d &lt;databasename&gt; wormbase.gff</code></code></pre>

<h1 id="SEE-ALSO">SEE ALSO</h1>

<p><a href="../../.cpan/build/BioPerl-1.6.901-ghKY8U/blib/lib/Bio/DB/GFF.html">Bio::DB::GFF</a>, <a>bulk_load_gff.pl</a>, <a>load_gff.pl</a></p>

<h1 id="AUTHOR">AUTHOR</h1>

<p>Lincoln Stein &lt;lstein@cshl.org&gt;</p>

<p>Copyright (c) 2002 Cold Spring Harbor Laboratory</p>

<p>This library is free software; you can redistribute it and/or modify it under the same terms as Perl itself. See DISCLAIMER.txt for disclaimers of warranty.</p>


</body>

</html>


