<?xml version="1.0" ?>
<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.0 Transitional//EN">
<html>
<head>
<!-- saved from url=(0017)http://localhost/ -->
<script language="JavaScript" src="../displayToc.js"></script>
<script language="JavaScript" src="../tocParas.js"></script>
<script language="JavaScript" src="../tocTab.js"></script>
<link rel="stylesheet" type="text/css" href="../scineplex.css">
<title></title>
<link rel="stylesheet" href="../Active.css" type="text/css" />
<meta http-equiv="content-type" content="text/html; charset=utf-8" />
<link rev="made" href="mailto:support@ActiveState.com" />
</head>

<body>



<ul id="index">
  <li><a href="#NAME">NAME</a></li>
  <li><a href="#DESCRIPTION">DESCRIPTION</a></li>
  <li><a href="#SYNOPSIS">SYNOPSIS</a></li>
  <li><a href="#OPTIONS">OPTIONS</a></li>
</ul>

<h1 id="NAME">NAME</h1>

<p>bp_seqfeature_load.pl - Load GFF into a SeqFeature database</p>

<h1 id="DESCRIPTION">DESCRIPTION</h1>

<p>Pass any number of GFF or fasta format files (or GFF with embedded fasta) to load the features and sequences into a SeqFeature database. The database (and adaptor) to use is specified on the command line. Use the --create flag to create a new SeqFeature database.</p>

<h1 id="SYNOPSIS">SYNOPSIS</h1>

<pre><code><code> bp_seqfeature_load.pl [options] gff_or_fasta_file1 [gff_or_fasta_file2 [...]]</code></code></pre>

<p>Try &#39;bp_seqfeature_load.pl --help&#39; or &#39;--man&#39; for more information.</p>

<h1 id="OPTIONS">OPTIONS</h1>

<dl>

<dt id="d---dsn">-d, --dsn</dt>
<dd>

<p>DBI data source (default dbi:mysql:test)</p>

</dd>
<dt id="n---namespace">-n, --namespace</dt>
<dd>

<p>The table prefix to use (default undef) Allows several independent sequence feature databases to be stored in a single database</p>

</dd>
<dt id="s---seqfeature">-s, --seqfeature</dt>
<dd>

<p>The type of SeqFeature to create... RTSC (default Bio::DB::SeqFeature)</p>

</dd>
<dt id="a---adaptor">-a, --adaptor</dt>
<dd>

<p>The storage adaptor (class) to use (default DBI::mysql)</p>

</dd>
<dt id="v---verbose">-v, --verbose</dt>
<dd>

<p>Turn on verbose progress reporting (default true) Use --noverbose to switch this off.</p>

</dd>
<dt id="f---fast">-f, --fast</dt>
<dd>

<p>Activate fast loading. (default 0) Only available for some adaptors.</p>

</dd>
<dt id="T---temporary-directory">-T, --temporary-directory</dt>
<dd>

<p>Specify temporary directory for fast loading (default File::Spec-&gt;tmpdir())</p>

</dd>
<dt id="i---ignore-seqregion">-i, --ignore-seqregion</dt>
<dd>

<p>If true, then ignore ##sequence-region directives in the GFF3 file (default, create a feature for each region)</p>

</dd>
<dt id="c---create">-c, --create</dt>
<dd>

<p>Create the database and reinitialize it (default false) Note, this will erase previous database contents, if any.</p>

</dd>
<dt id="u---user">-u, --user</dt>
<dd>

<p>User to connect to database as</p>

</dd>
<dt id="p---password">-p, --password</dt>
<dd>

<p>Password to use to connect to database</p>

</dd>
<dt id="z---zip">-z, --zip</dt>
<dd>

<p>Compress database tables to save space (default false)</p>

</dd>
<dt id="S---subfeatures">-S, --subfeatures</dt>
<dd>

<p>Turn on indexing of subfeatures (default true) Use --nosubfeatures to switch this off.</p>

</dd>
<dt id="summary">--summary</dt>
<dd>

<p>Generate summary statistics for coverage graphs (default false) This can be run on a previously loaded database or during the load. It will default to true if --create is used.</p>

</dd>
<dt id="noalias-target">--noalias-target</dt>
<dd>

<p>Don&#39;t create an Alias attribute whose value is the target_id in a Target attribute (if the feature contains a Target attribute, the default is to create an Alias attribute whose value is the target_id in the Target attribute)</p>

</dd>
</dl>

<p>Please see http://www.sequenceontology.org/gff3.shtml for information about the GFF3 format. BioPerl extends the format slightly by adding a ##index-subfeatures directive. Set this to a true value if you wish the database to be able to retrieve a feature&#39;s individual parts (such as the exons of a transcript) independently of the top level feature:</p>

<pre><code><code>  <span class="comment">##index-subfeatures 1</span>
</code></code></pre>

<p>It is also possible to control the indexing of subfeatures on a case-by-case basis by adding &quot;index=1&quot; or &quot;index=0&quot; to the feature&#39;s attribute list. This should only be used for subfeatures.</p>

<p>Subfeature indexing is true by default. Set to false (0) to save lots of database space and speed performance. You may use --nosubfeatures to force this.</p>


</body>

</html>


