<?xml version="1.0" ?>
<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.0 Transitional//EN">
<html>
<head>
<!-- saved from url=(0017)http://localhost/ -->
<script language="JavaScript" src="../../../../displayToc.js"></script>
<script language="JavaScript" src="../../../../tocParas.js"></script>
<script language="JavaScript" src="../../../../tocTab.js"></script>
<link rel="stylesheet" type="text/css" href="../../../../scineplex.css">
<title></title>
<link rel="stylesheet" href="../../../../Active.css" type="text/css" />
<meta http-equiv="content-type" content="text/html; charset=utf-8" />
<link rev="made" href="mailto:support@ActiveState.com" />
</head>

<body>



<ul id="index">
  <li><a href="#NAME">NAME</a></li>
  <li><a href="#SYNOPSIS">SYNOPSIS</a></li>
  <li><a href="#DESCRIPTION">DESCRIPTION</a></li>
  <li><a href="#FEEDBACK">FEEDBACK</a>
    <ul>
      <li><a href="#Support">Support</a></li>
      <li><a href="#Reporting-Bugs">Reporting Bugs</a></li>
    </ul>
  </li>
  <li><a href="#AUTHORS">AUTHORS</a></li>
  <li><a href="#APPENDIX">APPENDIX</a>
    <ul>
      <li><a href="#next_aln">next_aln</a></li>
      <li><a href="#write_aln">write_aln</a></li>
    </ul>
  </li>
</ul>

<h1 id="NAME">NAME</h1>

<p>Bio::AlignIO::arp - ARP MSA Sequence input/output stream</p>

<h1 id="SYNOPSIS">SYNOPSIS</h1>

<p>Do not use this module directly. Use it via the <a href="../../../../../.cpan/build/BioPerl-1.6.901-ghKY8U/blib/lib/Bio/AlignIO.html">Bio::AlignIO</a> class.</p>

<h1 id="DESCRIPTION">DESCRIPTION</h1>

<p>This object can create <a href="../../../../../.cpan/build/BioPerl-1.6.901-ghKY8U/blib/lib/Bio/SimpleAlign.html">Bio::SimpleAlign</a> objects from ARP flat files. These are typically configuration-like data files for the program Arlequin. For more information, see:</p>

<pre><code><code>  http://lgb.unige.ch/arlequin/</code></code></pre>

<p>For the moment, this retains the allele sequence data in the DATA section and inserts them into SimpleAlign objects. ARP files that contain other data (RFLP, etc.) are not expected to parse properly. Also, if the DNA data is actually SNP data, then the LocatableSeq object instantiation will throw an error.</p>

<p>This is now set up as a generic parser (i.e. it parses everything) and collects as much data as possible into the SimpleAlign object. The following in a general mapping of where data can be found:</p>

<pre><code><code>    Tag        SimpleAlign
               Method  
    ----------------------------------------------------------------------
    Title      description
    SampleName id  
    ----------------------------------------------------------------------

    Tag        Bio::Annotation   TagName                    Bio::Annotation
               Class                                        Parameters
    ----------------------------------------------------------------------
     NE        SimpleValue       pfam_family_accession      value
     NL        SimpleValue       sequence_start_stop        value
     SS        SimpleValue       sec_structure_source       value
     BM        SimpleValue       build_model                value
     RN        Reference         reference                  *
    ----------------------------------------------------------------------
  * RN is generated based on the number of Bio::Annotation::Reference objects</code></code></pre>

<p>In addition, the number of samples found in the alignment is retained in a Bio::Annotation::TagTree object in the annotation collection and is accessible via:</p>

<pre><code><code>  <span class="operator">(</span><span class="variable">$samples</span><span class="operator">)</span> <span class="operator">=</span> <span class="variable">$aln</span><span class="operator">-&gt;</span><span class="variable">annotation</span><span class="operator">-&gt;</span><span class="variable">get_Annotations</span><span class="operator">(</span><span class="string">'Samples'</span><span class="operator">);</span>
  <span class="keyword">say</span> <span class="variable">$samples</span><span class="operator">-&gt;</span><span class="variable">display_text</span><span class="operator">;</span>
  <span class="comment"># or use other relevant TagTree methods to retrieve data</span>
</code></code></pre>

<h1 id="FEEDBACK">FEEDBACK</h1>

<h2 id="Support">Support</h2>

<p>Please direct usage questions or support issues to the mailing list:</p>

<p><i>bioperl-l@bioperl.org</i></p>

<p>rather than to the module maintainer directly. Many experienced and reponsive experts will be able look at the problem and quickly address it. Please include a thorough description of the problem with code and data examples if at all possible.</p>

<h2 id="Reporting-Bugs">Reporting Bugs</h2>

<p>Report bugs to the Bioperl bug tracking system to help us keep track the bugs and their resolution. Bug reports can be submitted via the web:</p>

<pre><code><code>  https://redmine.open-bio.org/projects/bioperl/</code></code></pre>

<h1 id="AUTHORS">AUTHORS</h1>

<p>Chris Fields (cjfields)</p>

<h1 id="APPENDIX">APPENDIX</h1>

<p>The rest of the documentation details each of the object methods. Internal methods are usually preceded with a _</p>

<h2 id="next_aln">next_aln</h2>

<pre><code><code> Title   : next_aln
 Usage   : $aln = $stream-&gt;next_aln
 Function: returns the next alignment in the stream.
 Returns : Bio::Align::AlignI object - returns 0 on end of file
            or on error
 Args    : -width =&gt; optional argument to specify the width sequence
           will be written (60 chars by default)</code></code></pre>

<p>See <a href="../../../../../.cpan/build/BioPerl-1.6.901-ghKY8U/blib/lib/Bio/Align/AlignI.html">Bio::Align::AlignI</a></p>

<h2 id="write_aln">write_aln</h2>

<pre><code><code> Title   : write_aln
 Usage   : $stream-&gt;write_aln(@aln)
 Function: writes the $aln object into the stream in xmfa format
 Returns : 1 for success and 0 for error
 Args    : L&lt;Bio::Align::AlignI&gt; object</code></code></pre>

<p>See <a href="../../../../../.cpan/build/BioPerl-1.6.901-ghKY8U/blib/lib/Bio/Align/AlignI.html">Bio::Align::AlignI</a></p>


</body>

</html>


