<?xml version="1.0" ?>
<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.0 Transitional//EN">
<html>
<head>
<!-- saved from url=(0017)http://localhost/ -->
<script language="JavaScript" src="../../../../displayToc.js"></script>
<script language="JavaScript" src="../../../../tocParas.js"></script>
<script language="JavaScript" src="../../../../tocTab.js"></script>
<link rel="stylesheet" type="text/css" href="../../../../scineplex.css">
<title></title>
<link rel="stylesheet" href="../../../../Active.css" type="text/css" />
<meta http-equiv="content-type" content="text/html; charset=utf-8" />
<link rev="made" href="mailto:support@ActiveState.com" />
</head>

<body>



<ul id="index">
  <li><a href="#NAME">NAME</a></li>
  <li><a href="#SYNOPSIS">SYNOPSIS</a></li>
  <li><a href="#DESCRIPTION">DESCRIPTION</a></li>
  <li><a href="#FEEDBACK">FEEDBACK</a>
    <ul>
      <li><a href="#Mailing-Lists">Mailing Lists</a></li>
      <li><a href="#Support">Support</a></li>
      <li><a href="#Reporting-Bugs">Reporting Bugs</a></li>
    </ul>
  </li>
  <li><a href="#AUTHOR---Heikki-Lehvaslaiho">AUTHOR - Heikki Lehvaslaiho</a></li>
  <li><a href="#APPENDIX">APPENDIX</a>
    <ul>
      <li><a href="#id">id</a></li>
      <li><a href="#name">name</a></li>
      <li><a href="#tables">tables</a></li>
      <li><a href="#translate">translate</a></li>
      <li><a href="#translate_strict">translate_strict</a></li>
      <li><a href="#revtranslate">revtranslate</a></li>
      <li><a href="#reverse_translate_all">reverse_translate_all</a></li>
      <li><a href="#reverse_translate_best">reverse_translate_best</a></li>
      <li><a href="#is_start_codon">is_start_codon</a></li>
      <li><a href="#is_ter_codon">is_ter_codon</a></li>
      <li><a href="#is_unknown_codon">is_unknown_codon</a></li>
      <li><a href="#unambiguous_codons">unambiguous_codons</a></li>
      <li><a href="#unambiquous_codons">_unambiquous_codons</a></li>
      <li><a href="#add_table">add_table</a></li>
    </ul>
  </li>
</ul>

<h1 id="NAME">NAME</h1>

<p>Bio::Tools::CodonTable - Codon table object</p>

<h1 id="SYNOPSIS">SYNOPSIS</h1>

<pre><code><code>  <span class="comment"># This is a read-only class for all known codon tables.  The IDs are</span>
  <span class="comment"># the ones used by nucleotide sequence databases.  All common IUPAC</span>
  <span class="comment"># ambiguity codes for DNA, RNA and amino acids are recognized.</span>
  
  <span class="keyword">use</span> <span class="variable">Bio::Tools::CodonTable</span><span class="operator">;</span>
  
  <span class="comment"># defaults to ID 1 "Standard"</span>
  <span class="variable">$myCodonTable</span>   <span class="operator">=</span> <span class="variable">Bio::Tools::CodonTable</span><span class="operator">-&gt;</span><span class="variable">new</span><span class="operator">();</span>
  <span class="variable">$myCodonTable2</span>  <span class="operator">=</span> <span class="variable">Bio::Tools::CodonTable</span><span class="operator">-&gt;</span><span class="variable">new</span><span class="operator">(</span> <span class="string">-id</span> <span class="operator">=&gt;</span> <span class="number">3</span> <span class="operator">);</span>
  
  <span class="comment"># change codon table</span>
  <span class="variable">$myCodonTable</span><span class="operator">-&gt;</span><span class="variable">id</span><span class="operator">(</span><span class="number">5</span><span class="operator">);</span>
  
  <span class="comment"># examine codon table</span>
  <span class="keyword">print</span>  <span class="keyword">join</span> <span class="operator">(</span><span class="string">' '</span><span class="operator">,</span> <span class="string">"The name of the codon table no."</span><span class="operator">,</span> <span class="variable">$myCodonTable</span><span class="operator">-&gt;</span><span class="variable">id</span><span class="operator">(</span><span class="number">4</span><span class="operator">),</span>
           <span class="string">"is:"</span><span class="operator">,</span> <span class="variable">$myCodonTable</span><span class="operator">-&gt;</span><span class="variable">name</span><span class="operator">(),</span> <span class="string">"\n"</span><span class="operator">);</span>
  
  <span class="comment"># print possible codon tables</span>
  <span class="variable">$tables</span> <span class="operator">=</span> <span class="variable">Bio::Tools::CodonTable</span><span class="operator">-&gt;</span><span class="variable">tables</span><span class="operator">;</span>
  <span class="keyword">while</span> <span class="operator">(</span> <span class="operator">(</span><span class="variable">$id</span><span class="operator">,</span><span class="variable">$name</span><span class="operator">)</span> <span class="operator">=</span> <span class="keyword">each</span> <span class="variable">%</span><span class="operator">{</span><span class="variable">$tables</span><span class="operator">}</span> <span class="operator">)</span> <span class="operator">{</span>
    <span class="keyword">print</span> <span class="string">"</span><span class="variable">$id</span><span class="string"> = </span><span class="variable">$name</span><span class="string">\n"</span><span class="operator">;</span>
  <span class="operator">}</span>
  
  <span class="comment"># translate a codon</span>
  <span class="variable">$aa</span> <span class="operator">=</span> <span class="variable">$myCodonTable</span><span class="operator">-&gt;</span><span class="variable">translate</span><span class="operator">(</span><span class="string">'ACU'</span><span class="operator">);</span>
  <span class="variable">$aa</span> <span class="operator">=</span> <span class="variable">$myCodonTable</span><span class="operator">-&gt;</span><span class="variable">translate</span><span class="operator">(</span><span class="string">'act'</span><span class="operator">);</span>
  <span class="variable">$aa</span> <span class="operator">=</span> <span class="variable">$myCodonTable</span><span class="operator">-&gt;</span><span class="variable">translate</span><span class="operator">(</span><span class="string">'ytr'</span><span class="operator">);</span>
  
  <span class="comment"># reverse translate an amino acid</span>
  <span class="variable">@codons</span> <span class="operator">=</span> <span class="variable">$myCodonTable</span><span class="operator">-&gt;</span><span class="variable">revtranslate</span><span class="operator">(</span><span class="string">'A'</span><span class="operator">);</span>
  <span class="variable">@codons</span> <span class="operator">=</span> <span class="variable">$myCodonTable</span><span class="operator">-&gt;</span><span class="variable">revtranslate</span><span class="operator">(</span><span class="string">'Ser'</span><span class="operator">);</span>
  <span class="variable">@codons</span> <span class="operator">=</span> <span class="variable">$myCodonTable</span><span class="operator">-&gt;</span><span class="variable">revtranslate</span><span class="operator">(</span><span class="string">'Glx'</span><span class="operator">);</span>
  <span class="variable">@codons</span> <span class="operator">=</span> <span class="variable">$myCodonTable</span><span class="operator">-&gt;</span><span class="variable">revtranslate</span><span class="operator">(</span><span class="string">'cYS'</span><span class="operator">,</span> <span class="string">'rna'</span><span class="operator">);</span>
  
  <span class="comment"># reverse translate an entire amino acid sequence into a IUPAC</span>
  <span class="comment"># nucleotide string</span>
  
  <span class="keyword">my</span> <span class="variable">$seqobj</span>    <span class="operator">=</span> <span class="variable">Bio::PrimarySeq</span><span class="operator">-&gt;</span><span class="variable">new</span><span class="operator">(</span><span class="string">-seq</span> <span class="operator">=&gt;</span> <span class="string">'FHGERHEL'</span><span class="operator">);</span>
  <span class="keyword">my</span> <span class="variable">$iupac_str</span> <span class="operator">=</span> <span class="variable">$myCodonTable</span><span class="operator">-&gt;</span><span class="variable">reverse_translate_all</span><span class="operator">(</span><span class="variable">$seqobj</span><span class="operator">);</span>
  
  <span class="comment"># boolean tests</span>
  <span class="keyword">print</span> <span class="string">"Is a start\n"</span>       <span class="keyword">if</span> <span class="variable">$myCodonTable</span><span class="operator">-&gt;</span><span class="variable">is_start_codon</span><span class="operator">(</span><span class="string">'ATG'</span><span class="operator">);</span>
  <span class="keyword">print</span> <span class="string">"Is a terminator\n"</span> <span class="keyword">if</span> <span class="variable">$myCodonTable</span><span class="operator">-&gt;</span><span class="variable">is_ter_codon</span><span class="operator">(</span><span class="string">'tar'</span><span class="operator">);</span>
  <span class="keyword">print</span> <span class="string">"Is a unknown\n"</span>     <span class="keyword">if</span> <span class="variable">$myCodonTable</span><span class="operator">-&gt;</span><span class="variable">is_unknown_codon</span><span class="operator">(</span><span class="string">'JTG'</span><span class="operator">);</span>
</code></code></pre>

<h1 id="DESCRIPTION">DESCRIPTION</h1>

<p>Codon tables are also called translation tables or genetic codes since that is what they represent. A bit more complete picture of the full complexity of codon usage in various taxonomic groups is presented at the NCBI Genetic Codes Home page.</p>

<p>CodonTable is a BioPerl class that knows all current translation tables that are used by primary nucleotide sequence databases (GenBank, EMBL and DDBJ). It provides methods to output information about tables and relationships between codons and amino acids.</p>

<p>This class and its methods recognized all common IUPAC ambiguity codes for DNA, RNA and animo acids. The translation method follows the conventions in EMBL and TREMBL databases.</p>

<p>It is a nuisance to separate RNA and cDNA representations of nucleic acid transcripts. The CodonTable object accepts codons of both type as input and allows the user to set the mode for output when reverse translating. Its default for output is DNA.</p>

<p>Note:</p>

<p>This class deals primarily with individual codons and amino acids. However in the interest of speed you can <a>translate</a> longer sequence, too. The full complexity of protein translation is tackled by <a>Bio::PrimarySeqI::translate</a>.</p>

<p>The amino acid codes are IUPAC recommendations for common amino acids:</p>

<pre><code><code>          A           Ala            Alanine
          R           Arg            Arginine
          N           Asn            Asparagine
          D           Asp            Aspartic acid
          C           Cys            Cysteine
          Q           Gln            Glutamine
          E           Glu            Glutamic acid
          G           Gly            Glycine
          H           His            Histidine
          I           Ile            Isoleucine
          L           Leu            Leucine
          K           Lys            Lysine
          M           Met            Methionine
          F           Phe            Phenylalanine
          P           Pro            Proline
          O           Pyl            Pyrrolysine (22nd amino acid)
          U           Sec            Selenocysteine (21st amino acid)
          S           Ser            Serine
          T           Thr            Threonine
          W           Trp            Tryptophan
          Y           Tyr            Tyrosine
          V           Val            Valine
          B           Asx            Aspartic acid or Asparagine
          Z           Glx            Glutamine or Glutamic acid
          J           Xle            Isoleucine or Valine (mass spec ambiguity)
          X           Xaa            Any or unknown amino acid</code></code></pre>

<p>It is worth noting that, &quot;Bacterial&quot; codon table no. 11 produces an polypeptide that is, confusingly, identical to the standard one. The only differences are in available initiator codons.</p>

<p>NCBI Genetic Codes home page: http://www.ncbi.nlm.nih.gov/Taxonomy/Utils/wprintgc.cgi?mode=c</p>

<p>EBI Translation Table Viewer: http://www.ebi.ac.uk/cgi-bin/mutations/trtables.cgi</p>

<p>Amended ASN.1 version with ids 16 and 21 is at: ftp://ftp.ebi.ac.uk/pub/databases/geneticcode/</p>

<p>Thanks to Matteo diTomasso for the original Perl implementation of these tables.</p>

<h1 id="FEEDBACK">FEEDBACK</h1>

<h2 id="Mailing-Lists">Mailing Lists</h2>

<p>User feedback is an integral part of the evolution of this and other Bioperl modules. Send your comments and suggestions preferably to the Bioperl mailing lists Your participation is much appreciated.</p>

<pre><code><code>  bioperl-l@bioperl.org                  - General discussion
  http://bioperl.org/wiki/Mailing_lists  - About the mailing lists</code></code></pre>

<h2 id="Support">Support</h2>

<p>Please direct usage questions or support issues to the mailing list:</p>

<p><i>bioperl-l@bioperl.org</i></p>

<p>rather than to the module maintainer directly. Many experienced and reponsive experts will be able look at the problem and quickly address it. Please include a thorough description of the problem with code and data examples if at all possible.</p>

<h2 id="Reporting-Bugs">Reporting Bugs</h2>

<p>Report bugs to the Bioperl bug tracking system to help us keep track the bugs and their resolution. Bug reports can be submitted via the web:</p>

<pre><code><code>  https://redmine.open-bio.org/projects/bioperl/</code></code></pre>

<h1 id="AUTHOR---Heikki-Lehvaslaiho">AUTHOR - Heikki Lehvaslaiho</h1>

<p>Email: heikki-at-bioperl-dot-org</p>

<h1 id="APPENDIX">APPENDIX</h1>

<p>The rest of the documentation details each of the object methods. Internal methods are usually preceded with a _</p>

<h2 id="id">id</h2>

<pre><code><code> <span class="variable">Title</span>   <span class="operator">:</span> <span class="variable">id</span>
 <span class="variable">Usage</span>   <span class="operator">:</span> <span class="variable">$obj</span><span class="operator">-&gt;</span><span class="variable">id</span><span class="operator">(</span><span class="number">3</span><span class="operator">);</span> <span class="variable">$id_integer</span> <span class="operator">=</span> <span class="variable">$obj</span><span class="operator">-&gt;</span><span class="variable">id</span><span class="operator">();</span>
 <span class="variable">Function</span><span class="operator">:</span> <span class="variable">Sets</span> <span class="keyword">or</span> <span class="variable">returns</span> <span class="variable">the</span> <span class="variable">id</span> <span class="variable">of</span> <span class="variable">the</span> <span class="variable">translation</span> <span class="variable">table</span><span class="operator">.</span>  <span class="variable">IDs</span> <span class="variable">are</span>
           <span class="variable">integers</span> <span class="variable">from</span> <span class="number">1</span> <span class="variable">to</span> <span class="number">15</span><span class="operator">,</span> <span class="variable">excluding</span> <span class="number">7</span> <span class="keyword">and</span> <span class="number">8</span> <span class="variable">which</span> <span class="variable">have</span> <span class="variable">been</span>
           <span class="variable">removed</span> <span class="variable">as</span> <span class="variable">redundant</span><span class="operator">.</span> <span class="variable">If</span> <span class="variable">an</span> <span class="variable">invalid</span> <span class="variable">ID</span> <span class="variable">is</span> <span class="keyword">given</span> <span class="variable">the</span> <span class="variable">method</span>
           <span class="variable">returns</span> <span class="number">0</span><span class="operator">,</span> <span class="variable">false</span><span class="operator">.</span>
 <span class="variable">Example</span> <span class="operator">:</span>
 <span class="variable">Returns</span> <span class="operator">:</span> <span class="variable">value</span> <span class="variable">of</span> <span class="variable">id</span><span class="operator">,</span> <span class="variable">a</span> <span class="keyword">scalar</span><span class="operator">,</span> <span class="number">0</span> <span class="keyword">if</span> <span class="keyword">not</span> <span class="variable">a</span> <span class="variable">valid</span>
 <span class="variable">Args</span>    <span class="operator">:</span> <span class="variable">newvalue</span> <span class="operator">(</span><span class="variable">optional</span><span class="operator">)</span>
</code></code></pre>

<h2 id="name">name</h2>

<pre><code><code> Title   : name
 Usage   : $obj-&gt;name()
 Function: returns the descriptive name of the translation table
 Example :
 Returns : A string
 Args    : None</code></code></pre>

<h2 id="tables">tables</h2>

<pre><code><code> Title   : tables
 Usage   : $obj-&gt;tables()  or  Bio::Tools::CodonTable-&gt;tables()
 Function: returns a hash reference where each key is a valid codon
           table id() number, and each value is the corresponding
           codon table name() string
 Example :
 Returns : A hashref
 Args    : None</code></code></pre>

<h2 id="translate">translate</h2>

<pre><code><code> Title   : translate
 Usage   : $obj-&gt;translate(&#39;YTR&#39;)
 Function: Returns a string of one letter amino acid codes from 
           nucleotide sequence input. The imput can be of any length.

           Returns &#39;X&#39; for unknown codons and codons that code for
           more than one amino acid. Returns an empty string if input
           is not three characters long. Exceptions for these are:

             - IUPAC amino acid code B for Aspartic Acid and
               Asparagine, is used.
             - IUPAC amino acid code Z for Glutamic Acid, Glutamine is
               used.
             - if the codon is two nucleotides long and if by adding
               an a third character &#39;N&#39;, it codes for a single amino
               acid (with exceptions above), return that, otherwise
               return empty string.

           Returns empty string for other input strings that are not
           three characters long.

 Example :
 Returns : a string of one letter ambiguous IUPAC amino acid codes
 Args    : ambiguous IUPAC nucleotide string</code></code></pre>

<h2 id="translate_strict">translate_strict</h2>

<pre><code><code> Title   : translate_strict
 Usage   : $obj-&gt;translate_strict(&#39;ACT&#39;)
 Function: returns one letter amino acid code for a codon input

           Fast and simple translation. User is responsible to resolve
           ambiguous nucleotide codes before calling this
           method. Returns &#39;X&#39; for unknown codons and an empty string
           for input strings that are not three characters long.

           It is not recommended to use this method in a production
           environment. Use method translate, instead.

 Example :
 Returns : A string
 Args    : a codon = a three nucleotide character string</code></code></pre>

<h2 id="revtranslate">revtranslate</h2>

<pre><code><code> <span class="variable">Title</span>   <span class="operator">:</span> <span class="variable">revtranslate</span>
 <span class="variable">Usage</span>   <span class="operator">:</span> <span class="variable">$obj</span><span class="operator">-&gt;</span><span class="variable">revtranslate</span><span class="operator">(</span><span class="string">'G'</span><span class="operator">)</span>
 <span class="variable">Function</span><span class="operator">:</span> <span class="variable">returns</span> <span class="variable">codons</span> <span class="keyword">for</span> <span class="variable">an</span> <span class="variable">amino</span> <span class="variable">acid</span>
 
           <span class="variable">Returns</span> <span class="variable">an</span> <span class="variable">empty</span> <span class="variable">string</span> <span class="keyword">for</span> <span class="variable">unknown</span> <span class="variable">amino</span> <span class="variable">acid</span>
           <span class="variable">codes</span><span class="operator">.</span> <span class="variable">Ambiguous</span> <span class="variable">IUPAC</span> <span class="variable">codes</span> <span class="variable">Asx</span><span class="operator">,</span><span class="variable">B</span><span class="operator">,</span> <span class="operator">(</span><span class="variable">Asp</span><span class="operator">,</span><span class="variable">D</span><span class="operator">;</span> <span class="variable">Asn</span><span class="operator">,</span><span class="variable">N</span><span class="operator">)</span> <span class="keyword">and</span>
           <span class="variable">Glx</span><span class="operator">,</span><span class="variable">Z</span> <span class="operator">(</span><span class="variable">Glu</span><span class="operator">,</span><span class="variable">E</span><span class="operator">;</span> <span class="variable">Gln</span><span class="operator">,</span><span class="variable">Q</span><span class="operator">)</span> <span class="variable">are</span> <span class="variable">resolved</span><span class="operator">.</span> <span class="variable">Both</span> <span class="variable">single</span> <span class="keyword">and</span> <span class="variable">three</span>
           <span class="variable">letter</span> <span class="variable">amino</span> <span class="variable">acid</span> <span class="variable">codes</span> <span class="variable">are</span> <span class="variable">accepted</span><span class="operator">.</span> <span class="string">'*'</span> <span class="keyword">and</span> <span class="string">'Ter'</span> <span class="variable">are</span>
           <span class="variable">used</span> <span class="keyword">for</span> <span class="variable">terminator</span><span class="operator">.</span>
 
           <span class="variable">By</span> <span class="keyword">default</span><span class="operator">,</span> <span class="variable">the</span> <span class="variable">output</span> <span class="variable">codons</span> <span class="variable">are</span> <span class="variable">shown</span> <span class="variable">in</span> <span class="variable">DNA</span><span class="operator">.</span>  <span class="variable">If</span> <span class="variable">the</span>
           <span class="variable">output</span> <span class="variable">is</span> <span class="variable">needed</span> <span class="variable">in</span> <span class="variable">RNA</span> <span class="operator">(</span><span class="regex">tr/t/u/</span><span class="operator">),</span> <span class="variable">add</span> <span class="variable">a</span> <span class="variable">second</span> <span class="variable">argument</span>
           <span class="string">'RNA'</span><span class="operator">.</span>
 
 <span class="variable">Example</span> <span class="operator">:</span> <span class="variable">$obj</span><span class="operator">-&gt;</span><span class="variable">revtranslate</span><span class="operator">(</span><span class="string">'Gly'</span><span class="operator">,</span> <span class="string">'RNA'</span><span class="operator">)</span>
 <span class="variable">Returns</span> <span class="operator">:</span> <span class="variable">An</span> <span class="variable">array</span> <span class="variable">of</span> <span class="variable">three</span> <span class="variable">lower</span> <span class="variable">case</span> <span class="variable">letter</span> <span class="variable">strings</span> <span class="variable">i</span><span class="operator">.</span><span class="variable">e</span><span class="operator">.</span> <span class="variable">codons</span>
 <span class="variable">Args</span>    <span class="operator">:</span> <span class="variable">amino</span> <span class="variable">acid</span><span class="operator">,</span> <span class="string">'RNA'</span>
</code></code></pre>

<h2 id="reverse_translate_all">reverse_translate_all</h2>

<pre><code><code> <span class="variable">Title</span>   <span class="operator">:</span> <span class="variable">reverse_translate_all</span>
 <span class="variable">Usage</span>   <span class="operator">:</span> <span class="keyword">my</span> <span class="variable">$iup_str</span> <span class="operator">=</span> <span class="variable">$cttable</span><span class="operator">-&gt;</span><span class="variable">reverse_translate_all</span><span class="operator">(</span><span class="variable">$seq_object</span><span class="operator">)</span>
           <span class="keyword">my</span> <span class="variable">$iup_str</span> <span class="operator">=</span> <span class="variable">$cttable</span><span class="operator">-&gt;</span><span class="variable">reverse_translate_all</span><span class="operator">(</span><span class="variable">$seq_object</span><span class="operator">,</span>
                                                         <span class="variable">$cutable</span><span class="operator">,</span>
                                                         <span class="number">15</span><span class="operator">);</span>
 <span class="variable">Function</span><span class="operator">:</span> <span class="keyword">reverse</span> <span class="variable">translates</span> <span class="variable">a</span> <span class="variable">protein</span> <span class="variable">sequence</span> <span class="variable">into</span> <span class="variable">IUPAC</span> <span class="variable">nucleotide</span>
           <span class="variable">sequence</span><span class="operator">.</span> <span class="variable">An</span> <span class="string">'X'</span> <span class="variable">in</span> <span class="variable">the</span> <span class="variable">protein</span> <span class="variable">sequence</span> <span class="variable">is</span> <span class="variable">converted</span> <span class="variable">to</span> <span class="string">'NNN'</span>
           <span class="variable">in</span> <span class="variable">the</span> <span class="variable">nucleotide</span> <span class="variable">sequence</span><span class="operator">.</span>
 <span class="variable">Returns</span> <span class="operator">:</span> <span class="variable">a</span> <span class="variable">string</span>
 <span class="variable">Args</span>    <span class="operator">:</span> <span class="variable">a</span> <span class="variable">Bio::PrimarySeqI</span> <span class="variable">compatible</span> <span class="variable">object</span> <span class="operator">(</span><span class="variable">mandatory</span><span class="operator">)</span>
           <span class="variable">a</span> <span class="variable">Bio::CodonUsage::Table</span> <span class="variable">object</span> <span class="keyword">and</span> <span class="variable">a</span> <span class="variable">threshold</span> <span class="keyword">if</span> <span class="variable">only</span>
             <span class="variable">codons</span> <span class="variable">with</span> <span class="variable">a</span> <span class="variable">relative</span> <span class="variable">frequency</span> <span class="variable">above</span> <span class="variable">the</span> <span class="variable">threshold</span> <span class="variable">are</span>
             <span class="variable">to</span> <span class="variable">be</span> <span class="variable">considered</span><span class="operator">.</span>
</code></code></pre>

<h2 id="reverse_translate_best">reverse_translate_best</h2>

<pre><code><code> <span class="variable">Title</span>   <span class="operator">:</span> <span class="variable">reverse_translate_best</span>
 <span class="variable">Usage</span>   <span class="operator">:</span> <span class="keyword">my</span> <span class="variable">$str</span> <span class="operator">=</span> <span class="variable">$cttable</span><span class="operator">-&gt;</span><span class="variable">reverse_translate_best</span><span class="operator">(</span><span class="variable">$seq_object</span><span class="operator">,</span><span class="variable">$cutable</span><span class="operator">);</span>
 <span class="variable">Function</span><span class="operator">:</span> <span class="variable">Reverse</span> <span class="variable">translates</span> <span class="variable">a</span> <span class="variable">protein</span> <span class="variable">sequence</span> <span class="variable">into</span> <span class="variable">plain</span> <span class="variable">nucleotide</span>
           <span class="variable">sequence</span> <span class="operator">(</span><span class="variable">GATC</span><span class="operator">),</span> <span class="variable">uses</span> <span class="variable">the</span> <span class="variable">most</span> <span class="variable">common</span> <span class="variable">codon</span> <span class="keyword">for</span> <span class="keyword">each</span> <span class="variable">amino</span> <span class="variable">acid</span>
 <span class="variable">Returns</span> <span class="operator">:</span> <span class="variable">A</span> <span class="variable">string</span>
 <span class="variable">Args</span>    <span class="operator">:</span> <span class="variable">A</span> <span class="variable">Bio::PrimarySeqI</span> <span class="variable">compatible</span> <span class="variable">object</span> <span class="keyword">and</span> <span class="variable">a</span> <span class="variable">Bio::CodonUsage::Table</span> <span class="variable">object</span>
</code></code></pre>

<h2 id="is_start_codon">is_start_codon</h2>

<pre><code><code> Title   : is_start_codon
 Usage   : $obj-&gt;is_start_codon(&#39;ATG&#39;)
 Function: returns true (1) for all codons that can be used as a
           translation start, false (0) for others.
 Example : $myCodonTable-&gt;is_start_codon(&#39;ATG&#39;)
 Returns : boolean
 Args    : codon</code></code></pre>

<h2 id="is_ter_codon">is_ter_codon</h2>

<pre><code><code> Title   : is_ter_codon
 Usage   : $obj-&gt;is_ter_codon(&#39;GAA&#39;)
 Function: returns true (1) for all codons that can be used as a
           translation tarminator, false (0) for others.
 Example : $myCodonTable-&gt;is_ter_codon(&#39;ATG&#39;)
 Returns : boolean
 Args    : codon</code></code></pre>

<h2 id="is_unknown_codon">is_unknown_codon</h2>

<pre><code><code> Title   : is_unknown_codon
 Usage   : $obj-&gt;is_unknown_codon(&#39;GAJ&#39;)
 Function: returns false (0) for all codons that are valid,
        true (1) for others.
 Example : $myCodonTable-&gt;is_unknown_codon(&#39;NTG&#39;)
 Returns : boolean
 Args    : codon</code></code></pre>

<h2 id="unambiguous_codons">unambiguous_codons</h2>

<pre><code><code> Title   : unambiguous_codons
 Usage   : @codons = $self-&gt;unambiguous_codons(&#39;ACN&#39;)
 Returns : array of strings (one-letter unambiguous amino acid codes)
 Args    : a codon = a three IUPAC nucleotide character string</code></code></pre>

<h2 id="unambiquous_codons">_unambiquous_codons</h2>

<p>deprecated, now an alias for unambiguous_codons</p>

<h2 id="add_table">add_table</h2>

<pre><code><code> Title   : add_table
 Usage   : $newid = $ct-&gt;add_table($name, $table, $starts)
 Function: Add a custom Codon Table into the object.
           Know what you are doing, only the length of
           the argument strings is checked!
 Returns : the id of the new codon table
 Args    : name, a string, optional (can be empty)
           table, a string of 64 characters
           startcodons, a string of 64 characters, defaults to standard</code></code></pre>


</body>

</html>


