<?xml version="1.0" ?>
<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.0 Transitional//EN">
<html>
<head>
<!-- saved from url=(0017)http://localhost/ -->
<script language="JavaScript" src="../../../../displayToc.js"></script>
<script language="JavaScript" src="../../../../tocParas.js"></script>
<script language="JavaScript" src="../../../../tocTab.js"></script>
<link rel="stylesheet" type="text/css" href="../../../../scineplex.css">
<title></title>
<link rel="stylesheet" href="../../../../Active.css" type="text/css" />
<meta http-equiv="content-type" content="text/html; charset=utf-8" />
<link rev="made" href="mailto:support@ActiveState.com" />
</head>

<body>



<ul id="index">
  <li><a href="#NAME">NAME</a></li>
  <li><a href="#SYNOPSIS">SYNOPSIS</a></li>
  <li><a href="#DESCRIPTION">DESCRIPTION</a></li>
  <li><a href="#FEEDBACK">FEEDBACK</a>
    <ul>
      <li><a href="#Mailing-Lists">Mailing Lists</a></li>
      <li><a href="#Support">Support</a></li>
      <li><a href="#Reporting-Bugs">Reporting Bugs</a></li>
    </ul>
  </li>
  <li><a href="#AUTHOR---Keith-James">AUTHOR - Keith James</a></li>
  <li><a href="#APPENDIX">APPENDIX</a>
    <ul>
      <li><a href="#new">new</a></li>
      <li><a href="#next_prediction">next_prediction</a></li>
      <li><a href="#add_exon">_add_exon</a></li>
      <li><a href="#set_strand">_set_strand</a></li>
      <li><a href="#target_id">_target_id</a></li>
    </ul>
  </li>
  <li><a href="#POD-ERRORS">POD ERRORS</a></li>
</ul>

<h1 id="NAME">NAME</h1>

<p>Bio::Tools::Geneid - Results of one geneid run</p>

<h1 id="SYNOPSIS">SYNOPSIS</h1>

<pre><code><code>  <span class="keyword">use</span> <span class="variable">Bio::Tools::Geneid</span><span class="operator">;</span>
  <span class="keyword">my</span> <span class="variable">$gid</span> <span class="operator">=</span> <span class="variable">Bio::Tools::Geneid</span><span class="operator">(</span><span class="string">-file</span> <span class="operator">=&gt;</span> <span class="string">"geneid.out"</span><span class="operator">);</span>
  
  <span class="keyword">while</span> <span class="operator">(</span><span class="keyword">my</span> <span class="variable">$gene</span> <span class="operator">=</span> <span class="variable">$gid</span><span class="operator">-&gt;</span><span class="variable">next_prediction</span><span class="operator">)</span>
  <span class="operator">{</span>
    <span class="keyword">my</span> <span class="variable">@transcripts</span> <span class="operator">=</span> <span class="variable">$gene</span><span class="operator">-&gt;</span><span class="variable">transcripts</span><span class="operator">;</span>
      <span class="keyword">foreach</span> <span class="keyword">my</span> <span class="variable">$t</span> <span class="operator">(</span><span class="variable">@transcripts</span><span class="operator">)</span>
      <span class="operator">{</span>
        <span class="keyword">my</span> <span class="variable">@exons</span> <span class="operator">=</span> <span class="variable">$t</span><span class="operator">-&gt;</span><span class="variable">exons</span><span class="operator">;</span>
        <span class="keyword">foreach</span> <span class="keyword">my</span> <span class="variable">$e</span> <span class="operator">(</span><span class="variable">@exons</span><span class="operator">)</span>
        <span class="operator">{</span>
          <span class="keyword">printf</span><span class="operator">(</span><span class="string">"Exon %d..%d\n"</span><span class="operator">,</span> <span class="variable">$e</span><span class="operator">-&gt;</span><span class="variable">start</span><span class="operator">,</span> <span class="variable">$e</span><span class="operator">-&gt;</span><span class="variable">end</span><span class="operator">);</span>
        <span class="operator">}</span>
      <span class="operator">}</span>
  <span class="operator">}</span>
</code></code></pre>

<h1 id="DESCRIPTION">DESCRIPTION</h1>

<p>This is the parser for the output of geneid by Enrique Blanco and Roderic Guig&oacute; (IMIM-UPF). See http://www1.imim.es/software/geneid. It relies on native geneid output format internally and will work with geneid versions 1.0 and 1.1. Currently this module supports only the default mode of operation which is to predict exons and assemble an optimal gene prediction.</p>

<p>It takes either a file handle or a file name and returns a Bio::SeqFeature::Gene::GeneStructure object.</p>

<h1 id="FEEDBACK">FEEDBACK</h1>

<h2 id="Mailing-Lists">Mailing Lists</h2>

<p>User feedback is an integral part of the evolution of this and other Bioperl modules. Send your comments and suggestions preferably to one of the Bioperl mailing lists. Your participation is much appreciated.</p>

<pre><code><code>  bioperl-l@bioperl.org                  - General discussion
  http://bioperl.org/wiki/Mailing_lists  - About the mailing lists</code></code></pre>

<h2 id="Support">Support</h2>

<p>Please direct usage questions or support issues to the mailing list:</p>

<p><i>bioperl-l@bioperl.org</i></p>

<p>rather than to the module maintainer directly. Many experienced and reponsive experts will be able look at the problem and quickly address it. Please include a thorough description of the problem with code and data examples if at all possible.</p>

<h2 id="Reporting-Bugs">Reporting Bugs</h2>

<p>Report bugs to the Bioperl bug tracking system to help us keep track the bugs and their resolution. Bug reports can be submitted via the web:</p>

<pre><code><code>  https://redmine.open-bio.org/projects/bioperl/</code></code></pre>

<h1 id="AUTHOR---Keith-James">AUTHOR - Keith James</h1>

<pre><code><code> Email: kdj@sanger.ac.uk</code></code></pre>

<h1 id="APPENDIX">APPENDIX</h1>

<p>The rest of the documentation details each of the object methods. Internal methods are usually preceded with a _</p>

<h2 id="new">new</h2>

<pre><code><code> <span class="variable">Title</span>   <span class="operator">:</span> <span class="variable">new</span>
 <span class="variable">Usage</span>   <span class="operator">:</span> <span class="variable">$obj</span><span class="operator">-&gt;</span><span class="variable">new</span><span class="operator">(-</span><span class="variable">file</span> <span class="operator">=</span> <span class="string">"&lt;geneid.out"</span><span class="operator">);</span>
           <span class="variable">$obj</span><span class="operator">-&gt;</span><span class="variable">new</span><span class="operator">(</span><span class="string">-fh</span> <span class="operator">=&gt;</span> <span class="operator">\</span><span class="variable">*GI</span><span class="operator">);</span>
 <span class="variable">Function</span><span class="operator">:</span> <span class="variable">Constructor</span> <span class="keyword">for</span> <span class="variable">geneid</span> <span class="variable">wrapper</span><span class="operator">.</span> <span class="variable">Takes</span> <span class="variable">either</span> <span class="variable">a</span> <span class="variable">file</span>
         <span class="operator">:</span> <span class="keyword">or</span> <span class="variable">filehandle</span>
 <span class="variable">Returns</span> <span class="operator">:</span> <span class="variable">L</span><span class="operator">&lt;</span><span class="variable">Bio::Tools::Geneid</span><span class="operator">&gt;</span>
</code></code></pre>

<h2 id="next_prediction">next_prediction</h2>

<pre><code><code> <span class="variable">Title</span>   <span class="operator">:</span> <span class="variable">next_prediction</span>
 <span class="variable">Usage</span>   <span class="operator">:</span> <span class="keyword">while</span><span class="operator">(</span><span class="variable">$gene</span> <span class="operator">=</span> <span class="variable">$geneid</span><span class="operator">-&gt;</span><span class="variable">next_prediction</span><span class="operator">)</span>
           <span class="operator">{</span>
               <span class="comment"># do something</span>
           <span class="operator">}</span>
 <span class="variable">Function</span><span class="operator">:</span> <span class="variable">Returns</span> <span class="variable">the</span> <span class="variable">gene</span> <span class="variable">structure</span> <span class="variable">prediction</span> <span class="variable">of</span> <span class="variable">the</span> <span class="variable">geneid</span> <span class="variable">result</span>
           <span class="variable">file</span><span class="operator">.</span> <span class="variable">Call</span> <span class="variable">this</span> <span class="variable">method</span> <span class="variable">repeatedly</span> <span class="keyword">until</span> <span class="variable">FALSE</span> <span class="variable">is</span> <span class="variable">returned</span><span class="operator">.</span>
 <span class="variable">Returns</span> <span class="operator">:</span> <span class="variable">A</span> <span class="variable">Bio::SeqFeature::Gene::GeneStructure</span> <span class="variable">object</span>
 <span class="variable">Args</span>    <span class="operator">:</span> <span class="variable">None</span>
</code></code></pre>

<h2 id="add_exon">_add_exon</h2>

<pre><code><code> Title   : _add_exon
 Usage   : $obj-&gt;_add_exon($gene, $transcript, ... exon data ...)
 Function: Adds a new exon to both gene and transcript from the data
         : supplied as args
 Example :
 Returns : Nothing</code></code></pre>

<h2 id="set_strand">_set_strand</h2>

<pre><code><code> Title   : _set_strand
 Usage   : $obj-&gt;_set_strand($gene)
 Function: Sets the overall gene strand to the same strand as all
         : the exons if they are all on the same strand, or to strand 0
         : if the exons are on different strands.
 Example :
 Returns : Nothing</code></code></pre>

<h2 id="target_id">_target_id</h2>

<pre><code><code> Title   : _target_id
 Usage   : $obj-&gt;_target_id
 Function: get/set for genomic sequence id
 Example :
 Returns : A target ID</code></code></pre>

<h1 id="POD-ERRORS">POD ERRORS</h1>

<p>Hey! <b>The above document had some coding errors, which are explained below:</b></p>

<dl>

<dt id="Around-line-37:">Around line 37:</dt>
<dd>

<p>Non-ASCII character seen before =encoding in &#39;Guig&oacute;&#39;. Assuming ISO8859-1</p>

</dd>
</dl>


</body>

</html>


