<?xml version="1.0" ?>
<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.0 Transitional//EN">
<html>
<head>
<!-- saved from url=(0017)http://localhost/ -->
<script language="JavaScript" src="../../../../displayToc.js"></script>
<script language="JavaScript" src="../../../../tocParas.js"></script>
<script language="JavaScript" src="../../../../tocTab.js"></script>
<link rel="stylesheet" type="text/css" href="../../../../scineplex.css">
<title></title>
<link rel="stylesheet" href="../../../../Active.css" type="text/css" />
<meta http-equiv="content-type" content="text/html; charset=utf-8" />
<link rev="made" href="mailto:support@ActiveState.com" />
</head>

<body>



<ul id="index">
  <li><a href="#NAME">NAME</a></li>
  <li><a href="#SYNOPSIS">SYNOPSIS</a></li>
  <li><a href="#INSTALLATION">INSTALLATION</a></li>
  <li><a href="#DESCRIPTION">DESCRIPTION</a></li>
  <li><a href="#FEEDBACK">FEEDBACK</a>
    <ul>
      <li><a href="#Mailing-Lists">Mailing Lists</a></li>
      <li><a href="#Support">Support</a></li>
      <li><a href="#Reporting-Bugs">Reporting Bugs</a></li>
    </ul>
  </li>
  <li><a href="#AUTHOR">AUTHOR</a></li>
  <li><a href="#CONTRIBUTORS">CONTRIBUTORS</a></li>
  <li><a href="#APPENDIX">APPENDIX</a>
    <ul>
      <li><a href="#pairwise_alignment">pairwise_alignment</a></li>
      <li><a href="#align_and_show">align_and_show</a></li>
      <li><a href="#matrix">matrix</a></li>
      <li><a href="#gap">gap</a></li>
      <li><a href="#ext">ext</a></li>
    </ul>
  </li>
</ul>

<h1 id="NAME">NAME</h1>

<p>Bio::Tools::pSW - pairwise Smith Waterman object</p>

<h1 id="SYNOPSIS">SYNOPSIS</h1>

<pre><code><code>    <span class="keyword">use</span> <span class="variable">Bio::Tools::pSW</span><span class="operator">;</span>
    <span class="keyword">use</span> <span class="variable">Bio::AlignIO</span><span class="operator">;</span>
    <span class="keyword">my</span> <span class="variable">$factory</span> <span class="operator">=</span> <span class="variable">Bio::Tools::pSW</span><span class="operator">-&gt;</span><span class="variable">new</span><span class="operator">(</span> <span class="string">'-matrix'</span> <span class="operator">=&gt;</span> <span class="string">'blosum62.bla'</span><span class="operator">,</span>
                                       <span class="string">'-gap'</span> <span class="operator">=&gt;</span> <span class="number">12</span><span class="operator">,</span>
                                       <span class="string">'-ext'</span> <span class="operator">=&gt;</span> <span class="number">2</span><span class="operator">,</span>
                                       <span class="operator">);</span>
    
    <span class="comment">#use the factory to make some output</span>
    
    <span class="variable">$factory</span><span class="operator">-&gt;</span><span class="variable">align_and_show</span><span class="operator">(</span><span class="variable">$seq1</span><span class="operator">,</span><span class="variable">$seq2</span><span class="operator">,</span><span class="variable">STDOUT</span><span class="operator">);</span>
    
    <span class="comment"># make a Bio::SimpleAlign and do something with it</span>
    
    <span class="keyword">my</span> <span class="variable">$aln</span> <span class="operator">=</span> <span class="variable">$factory</span><span class="operator">-&gt;</span><span class="variable">pairwise_alignment</span><span class="operator">(</span><span class="variable">$seq1</span><span class="operator">,</span><span class="variable">$seq2</span><span class="operator">);</span>
    <span class="keyword">my</span> <span class="variable">$alnout</span> <span class="operator">=</span> <span class="variable">Bio::AlignIO</span><span class="operator">-&gt;</span><span class="variable">new</span><span class="operator">(</span><span class="string">-format</span> <span class="operator">=&gt;</span> <span class="string">'msf'</span><span class="operator">,</span>
                                  <span class="string">-fh</span>     <span class="operator">=&gt;</span> <span class="operator">\</span><span class="variable">*STDOUT</span><span class="operator">);</span>
    
    <span class="variable">$alnout</span><span class="operator">-&gt;</span><span class="variable">write_aln</span><span class="operator">(</span><span class="variable">$aln</span><span class="operator">);</span>
</code></code></pre>

<h1 id="INSTALLATION">INSTALLATION</h1>

<p>This module is included with the central Bioperl distribution:</p>

<pre><code><code>   http://bio.perl.org/Core/Latest
   ftp://bio.perl.org/pub/DIST</code></code></pre>

<p>Follow the installation instructions included in the INSTALL file.</p>

<h1 id="DESCRIPTION">DESCRIPTION</h1>

<p>pSW is an Alignment Factory for protein sequences. It builds pairwise alignments using the Smith-Waterman algorithm. The alignment algorithm is implemented in C and added in using an XS extension. The XS extension basically comes from the Wise2 package, but has been slimmed down to only be the alignment part of that (this is a good thing!). The XS extension comes from the bioperl-ext package which is distributed along with bioperl. <i>Warning:</i> This package will not work if you have not compiled the bioperl-ext package.</p>

<p>The mixture of C and Perl is ideal for this sort of problem. Here are some plus points for this strategy:</p>

<dl>

<dt id="Speed-and-Memory">Speed and Memory</dt>
<dd>

<p>The algorithm is actually implemented in C, which means it is faster than a pure perl implementation (I have never done one, so I have no idea how faster) and will use considerably less memory, as it efficiently assigns memory for the calculation.</p>

</dd>
<dt id="Algorithm-efficiency">Algorithm efficiency</dt>
<dd>

<p>The algorithm was written using Dynamite, and so contains an automatic switch to the linear space divide-and-conquer method. This means you could effectively align very large sequences without killing your machine (it could take a while though!).</p>

</dd>
</dl>

<h1 id="FEEDBACK">FEEDBACK</h1>

<h2 id="Mailing-Lists">Mailing Lists</h2>

<p>User feedback is an integral part of the evolution of this and other Bioperl modules. Send your comments and suggestions preferably to one of the Bioperl mailing lists. Your participation is much appreciated.</p>

<pre><code><code>  bioperl-l@bioperl.org                  - General discussion
  http://bioperl.org/wiki/Mailing_lists  - About the mailing lists</code></code></pre>

<h2 id="Support">Support</h2>

<p>Please direct usage questions or support issues to the mailing list:</p>

<p><i>bioperl-l@bioperl.org</i></p>

<p>rather than to the module maintainer directly. Many experienced and reponsive experts will be able look at the problem and quickly address it. Please include a thorough description of the problem with code and data examples if at all possible.</p>

<h2 id="Reporting-Bugs">Reporting Bugs</h2>

<p>Report bugs to the Bioperl bug tracking system to help us keep track the bugs and their resolution. Bug reports can be submitted via the web:</p>

<pre><code><code>  https://redmine.open-bio.org/projects/bioperl/           </code></code></pre>

<h1 id="AUTHOR">AUTHOR</h1>

<p>Ewan Birney, birney-at-sanger.ac.uk or birney-at-ebi.ac.uk</p>

<h1 id="CONTRIBUTORS">CONTRIBUTORS</h1>

<p>Jason Stajich, jason-at-bioperl.org</p>

<h1 id="APPENDIX">APPENDIX</h1>

<p>The rest of the documentation details each of the object methods. Internal methods are usually preceded with an underscore &quot;_&quot;.</p>

<h2 id="pairwise_alignment">pairwise_alignment</h2>

<pre><code><code> Title   : pairwise_alignment
 Usage   : $aln = $factory-&gt;pairwise_alignment($seq1,$seq2)
 Function: Makes a SimpleAlign object from two sequences
 Returns : A SimpleAlign object
 Args    :</code></code></pre>

<h2 id="align_and_show">align_and_show</h2>

<pre><code><code> Title   : align_and_show
 Usage   : $factory-&gt;align_and_show($seq1,$seq2,STDOUT)</code></code></pre>

<h2 id="matrix">matrix</h2>

<pre><code><code> <span class="variable">Title</span>     <span class="operator">:</span> <span class="variable">matrix</span><span class="operator">()</span>
 <span class="variable">Usage</span>     <span class="operator">:</span> <span class="variable">$factory</span><span class="operator">-&gt;</span><span class="variable">matrix</span><span class="operator">(</span><span class="string">'blosum62.bla'</span><span class="operator">);</span>
 <span class="variable">Function</span>  <span class="operator">:</span> <span class="variable">Reads</span> <span class="variable">in</span> <span class="variable">comparison</span> <span class="variable">matrix</span> <span class="variable">based</span> <span class="variable">on</span> <span class="variable">name</span>
           <span class="operator">:</span>
 <span class="variable">Returns</span>   <span class="operator">:</span> 
 <span class="variable">Argument</span>  <span class="operator">:</span> <span class="variable">comparison</span> <span class="variable">matrix</span>
</code></code></pre>

<h2 id="gap">gap</h2>

<pre><code><code> Title     : gap
 Usage     : $gap = $factory-&gt;gap() #get
           : $factory-&gt;gap($value) #set
 Function  : the set get for the gap penalty
 Example   :
 Returns   : gap value 
 Arguments : new value</code></code></pre>

<h2 id="ext">ext</h2>

<pre><code><code> Title     : ext
 Usage     : $ext = $factory-&gt;ext() #get
           : $factory-&gt;ext($value) #set
 Function  : the set get for the ext penalty
 Example   :
 Returns   : ext value 
 Arguments : new value</code></code></pre>


</body>

</html>


