<?xml version="1.0" ?>
<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.0 Transitional//EN">
<html>
<head>
<!-- saved from url=(0017)http://localhost/ -->
<script language="JavaScript" src="../../../../displayToc.js"></script>
<script language="JavaScript" src="../../../../tocParas.js"></script>
<script language="JavaScript" src="../../../../tocTab.js"></script>
<link rel="stylesheet" type="text/css" href="../../../../scineplex.css">
<title></title>
<link rel="stylesheet" href="../../../../Active.css" type="text/css" />
<meta http-equiv="content-type" content="text/html; charset=utf-8" />
<link rev="made" href="mailto:support@ActiveState.com" />
</head>

<body>



<ul id="index">
  <li><a href="#NAME">NAME</a></li>
  <li><a href="#SYNOPSIS">SYNOPSIS</a>
    <ul>
      <li><a href="#Object-Creation">Object Creation</a></li>
      <li><a href="#Object-Methods-Accessors">Object Methods &amp; Accessors</a></li>
    </ul>
  </li>
  <li><a href="#DESCRIPTION">DESCRIPTION</a></li>
  <li><a href="#FEEDBACK">FEEDBACK</a>
    <ul>
      <li><a href="#Support">Support</a></li>
      <li><a href="#Reporting-Bugs">Reporting Bugs</a></li>
    </ul>
  </li>
  <li><a href="#AUTHOR">AUTHOR</a></li>
  <li><a href="#CONTRIBUTORS">CONTRIBUTORS</a></li>
  <li><a href="#VERSION">VERSION</a></li>
  <li><a href="#COPYRIGHT">COPYRIGHT</a></li>
  <li><a href="#REFERENCES-SEE-ALSO">REFERENCES / SEE ALSO</a></li>
  <li><a href="#APPENDIX">APPENDIX</a></li>
  <li><a href="#threshold">threshold</a></li>
  <li><a href="#matrix">matrix</a></li>
  <li><a href="#seq">seq</a></li>
  <li><a href="#Analyze">_Analyze</a></li>
  <li><a href="#signals">signals</a></li>
  <li><a href="#result_count">result_count</a></li>
  <li><a href="#pretty_print">pretty_print</a></li>
</ul>

<h1 id="NAME">NAME</h1>

<p>Bio::Tools::Sigcleave - Bioperl object for sigcleave analysis</p>

<h1 id="SYNOPSIS">SYNOPSIS</h1>

<h2 id="Object-Creation">Object Creation</h2>

<pre><code><code>  <span class="keyword">use</span> <span class="variable">Bio::Tools::Sigcleave</span> <span class="operator">();</span>
  
  <span class="comment"># to keep the module backwar compatible, you can pass it a sequence string, but</span>
  <span class="comment"># there recommended say is to pass it a Seq object</span>
  
  <span class="comment"># this works</span>
  <span class="variable">$seq</span> <span class="operator">=</span> <span class="string">"MVLLLILSVLLLKEDVRGSAQSSERRVVAHMPGDIIIGALFSVHHQPTVDKVHERKCGAVREQYGI"</span><span class="operator">;</span>
  <span class="variable">$sig</span> <span class="operator">=</span> <span class="variable">Bio::Tools::Sigcleave</span><span class="operator">-&gt;</span><span class="variable">new</span><span class="operator">(</span><span class="string">-seq</span>  <span class="operator">=&gt;</span> <span class="variable">$seq</span><span class="operator">,</span>
                                                <span class="string">-type</span> <span class="operator">=&gt;</span> <span class="string">'protein'</span><span class="operator">,</span>
                                                <span class="string">-threshold</span><span class="operator">=&gt;</span><span class="string">'3.5'</span><span class="operator">,</span>
                                                <span class="operator">);</span>
  <span class="comment"># but you do:</span>
  <span class="variable">$seqobj</span> <span class="operator">=</span> <span class="variable">Bio::PrimarySeq</span><span class="operator">-&gt;</span><span class="variable">new</span><span class="operator">(</span><span class="string">-seq</span> <span class="operator">=&gt;</span> <span class="variable">$seq</span><span class="operator">);</span>
  
  <span class="variable">$sig</span> <span class="operator">=</span> <span class="variable">Bio::Tools::Sigcleave</span><span class="operator">-&gt;</span><span class="variable">new</span><span class="operator">(</span><span class="string">-seq</span>  <span class="operator">=&gt;</span> <span class="variable">$seqobj</span><span class="operator">,</span>
                                                <span class="string">-threshold</span><span class="operator">=&gt;</span><span class="string">'3.5'</span><span class="operator">,</span>
                                                <span class="operator">);</span>
  
  <span class="comment"># now you can detect procaryotic signal sequences as well as eucaryotic</span>
  <span class="variable">$sig</span><span class="operator">-&gt;</span><span class="variable">matrix</span><span class="operator">(</span><span class="string">'eucaryotic'</span><span class="operator">);</span> <span class="comment"># or 'procaryotic'</span>
</code></code></pre>

<h2 id="Object-Methods-Accessors">Object Methods &amp; Accessors</h2>

<pre><code><code>  <span class="comment"># you can use this method to fine tune the threshod before printing out the results</span>
  <span class="variable">$sig</span><span class="operator">-&gt;</span><span class="variable">result_count</span><span class="operator">:</span>
  
  <span class="variable">%raw_results</span>      <span class="operator">=</span> <span class="variable">$sig</span><span class="operator">-&gt;</span><span class="variable">signals</span><span class="operator">;</span>
  <span class="variable">$formatted_output</span> <span class="operator">=</span> <span class="variable">$sig</span><span class="operator">-&gt;</span><span class="variable">pretty_print</span><span class="operator">;</span>
</code></code></pre>

<h1 id="DESCRIPTION">DESCRIPTION</h1>

<p>&quot;Sigcleave&quot; was a program distributed as part of the free EGCG add-on to earlier versions of the GCG Sequence Analysis package. A new implementation of the algorithm is now part of EMBOSS package.</p>

<p>From the EGCG documentation:</p>

<pre><code><code>  SigCleave uses the von Heijne method to locate signal sequences, and
  to identify the cleavage site. The method is 95% accurate in
  resolving signal sequences from non-signal sequences with a cutoff
  score of 3.5, and 75-80% accurate in identifying the cleavage
  site. The program reports all hits above a minimum value.</code></code></pre>

<p>The EGCG Sigcleave program was written by Peter Rice (E-mail: pmr@sanger.ac.uk Post: Informatics Division, The Sanger Centre, Wellcome Trust Genome Campus, Hinxton, Cambs, CB10 1SA, UK).</p>

<p>Since EGCG is no longer distributed for the latest versions of GCG, this code was developed to emulate the output of the original program as much as possible for those who lost access to sigcleave when upgrading to newer versions of GCG.</p>

<p>There are 2 accessor methods for this object. &quot;signals&quot; will return a perl associative array containing the sigcleave scores keyed by amino acid position. &quot;pretty_print&quot; returns a formatted string similar to the output of the original sigcleave utility.</p>

<p>In both cases, the &quot;threshold&quot; setting controls the score reporting level. If no value for threshold is passed in by the user, the code defaults to a reporting value of 3.5.</p>

<p>In this implemntation the accessor will never return any score/position pair which does not meet the threshold limit. This is the slightly different from the behaviour of the 8.1 EGCG sigcleave program which will report the highest of the under-threshold results if nothing else is found.</p>

<p>Example of pretty_print output:</p>

<pre><code><code>        SIGCLEAVE of sigtest from: 1 to 146

        Report scores over 3.5
        Maximum score 4.9 at residue 131

         Sequence:  FVILAAMSIQGSA-NLQTQWKSTASLALET
                    | (signal)    | (mature peptide)
                118            131

         Other entries above 3.5

        Maximum score 3.7 at residue 112

         Sequence:  CSRQLFGWLFCKV-HPGAIVFVILAAMSIQGSANLQTQWKSTASLALET
                   | (signal)    | (mature peptide)
                99            112</code></code></pre>

<h1 id="FEEDBACK">FEEDBACK</h1>

<p>When updating and maintaining a module, it helps to know that people are actually using it. Let us know if you find a bug, think this code is useful or have any improvements/features to suggest.</p>

<h2 id="Support">Support</h2>

<p>Please direct usage questions or support issues to the mailing list:</p>

<p><i>bioperl-l@bioperl.org</i></p>

<p>rather than to the module maintainer directly. Many experienced and reponsive experts will be able look at the problem and quickly address it. Please include a thorough description of the problem with code and data examples if at all possible.</p>

<h2 id="Reporting-Bugs">Reporting Bugs</h2>

<p>Report bugs to the Bioperl bug tracking system to help us keep track the bugs and their resolution. Bug reports can be submitted via the web:</p>

<pre><code><code>  https://redmine.open-bio.org/projects/bioperl/</code></code></pre>

<h1 id="AUTHOR">AUTHOR</h1>

<p>Chris Dagdigian, dag-at-sonsorol.org &amp; others</p>

<h1 id="CONTRIBUTORS">CONTRIBUTORS</h1>

<p>Heikki Lehvaslaiho, heikki-at-bioperl-dot-org</p>

<h1 id="VERSION">VERSION</h1>

<p>Bio::Tools::Sigcleave, $Id$</p>

<h1 id="COPYRIGHT">COPYRIGHT</h1>

<p>Copyright (c) 1999 Chris Dagdigian &amp; others. All Rights Reserved. This module is free software; you can redistribute it and/or modify it under the same terms as Perl itself.</p>

<h1 id="REFERENCES-SEE-ALSO">REFERENCES / SEE ALSO</h1>

<p>von Heijne G. (1986) &quot;A new method for predicting signal sequences cleavage sites.&quot; Nucleic Acids Res. 14, 4683-4690.</p>

<p>von Heijne G. (1987) in &quot;Sequence Analysis in Molecular Biology: Treasure Trove or Trivial Pursuit&quot; (Acad. Press, (1987), 113-117).</p>

<h1 id="APPENDIX">APPENDIX</h1>

<p>The following documentation describes the various functions contained in this module. Some functions are for internal use and are not meant to be called by the user; they are preceded by an underscore (&quot;_&quot;).</p>

<h1 id="threshold">threshold</h1>

<pre><code><code> Title     : threshold
 Usage     : $value = $self-&gt;threshold
 Purpose   : Read/write method sigcleave score reporting threshold.
 Returns   : float.
 Argument  : new value, float
 Throws    : on non-number argument
 Comments  : defaults to 3.5
 See Also   : n/a</code></code></pre>

<h1 id="matrix">matrix</h1>

<pre><code><code> Title     : matrix
 Usage     : $value = $self-&gt;matrix(&#39;procaryotic&#39;)
 Purpose   : Read/write method sigcleave matrix.
 Returns   : float.
 Argument  : new value: &#39;eucaryotic&#39; or &#39;procaryotic&#39;
 Throws    : on non-number argument
 Comments  : defaults to 3.5
 See Also   : n/a</code></code></pre>

<h1 id="seq">seq</h1>

<pre><code><code> Title     : seq
 Usage     : $value = $self-&gt;seq($seq_object)
 Purpose   : set the Seq object to be used
 Returns   : Seq object
 Argument  : protein sequence or Seq object
 See Also   : n/a</code></code></pre>

<h1 id="Analyze">_Analyze</h1>

<pre><code><code> Title     : _Analyze
 Usage     : N/A This is an internal method. Not meant to be called from outside
           : the package
           :
 Purpose   : calculates sigcleave score and amino acid position for the
           : given protein sequence. The score reporting threshold can
           : be adjusted by passing in the &quot;threshold&quot; parameter during
           : object construction. If no threshold is passed in, the code
           : defaults to reporting any scores equal to or above 3.5
           :
 Returns   : nothing. results are added to the object
 Argument  : none.
 Throws    : nothing.
 Comments  : nothing.
See Also   : n/a</code></code></pre>

<h1 id="signals">signals</h1>

<pre><code><code> <span class="variable">Title</span>     <span class="operator">:</span> <span class="variable">signals</span>
 <span class="variable">Usage</span>     <span class="operator">:</span> <span class="variable">%sigcleave_results</span> <span class="operator">=</span> <span class="variable">$sig</span><span class="operator">-&gt;</span><span class="variable">signals</span><span class="operator">;</span>
           <span class="operator">:</span>
 <span class="variable">Purpose</span>   <span class="operator">:</span> <span class="variable">Accessor</span> <span class="variable">method</span> <span class="keyword">for</span> <span class="variable">sigcleave</span> <span class="variable">results</span>
           <span class="operator">:</span> 
 <span class="variable">Returns</span>   <span class="operator">:</span> <span class="variable">Associative</span> <span class="variable">array</span><span class="operator">.</span> <span class="variable">The</span> <span class="variable">key</span> <span class="variable">value</span> <span class="variable">represents</span> <span class="variable">the</span> <span class="variable">amino</span> <span class="variable">acid</span> <span class="variable">position</span>
           <span class="operator">:</span> <span class="keyword">and</span> <span class="variable">the</span> <span class="variable">value</span> <span class="variable">represents</span> <span class="variable">the</span> <span class="variable">score</span><span class="operator">.</span> <span class="variable">Only</span> <span class="variable">scores</span> <span class="variable">that</span>
           <span class="operator">:</span> <span class="variable">are</span> <span class="variable">greater</span> <span class="variable">than</span> <span class="keyword">or</span> <span class="variable">equal</span> <span class="variable">to</span> <span class="variable">the</span> <span class="variable">THRESHOLD</span> <span class="variable">value</span> <span class="variable">are</span> <span class="variable">reported</span><span class="operator">.</span>
           <span class="operator">:</span> 
 <span class="variable">Argument</span>  <span class="operator">:</span> <span class="variable">none</span><span class="operator">.</span>
 <span class="variable">Throws</span>    <span class="operator">:</span> <span class="variable">none</span><span class="operator">.</span>
 <span class="variable">Comments</span>  <span class="operator">:</span> <span class="variable">none</span><span class="operator">.</span>
 <span class="variable">See</span> <span class="variable">Also</span>   <span class="operator">:</span> <span class="variable">THRESHOLD</span>
</code></code></pre>

<h1 id="result_count">result_count</h1>

<pre><code><code> <span class="variable">Title</span>     <span class="operator">:</span> <span class="variable">result_count</span>
 <span class="variable">Usage</span>     <span class="operator">:</span> <span class="variable">$count</span> <span class="operator">=</span> <span class="variable">$sig</span><span class="operator">-&gt;</span><span class="variable">result_count</span><span class="operator">;</span>
           <span class="operator">:</span>
 <span class="variable">Purpose</span>   <span class="operator">:</span> <span class="variable">Accessor</span> <span class="variable">method</span> <span class="keyword">for</span> <span class="variable">sigcleave</span> <span class="variable">results</span>
           <span class="operator">:</span> 
 <span class="variable">Returns</span>   <span class="operator">:</span> <span class="variable">Integer</span><span class="operator">,</span> <span class="variable">number</span> <span class="variable">of</span> <span class="variable">results</span> <span class="variable">above</span> <span class="variable">the</span> <span class="variable">threshold</span>
           <span class="operator">:</span> 
 <span class="variable">Argument</span>  <span class="operator">:</span> <span class="variable">none</span><span class="operator">.</span>
 <span class="variable">Throws</span>    <span class="operator">:</span> <span class="variable">none</span><span class="operator">.</span>
 <span class="variable">Comments</span>  <span class="operator">:</span> <span class="variable">none</span><span class="operator">.</span>
</code></code></pre>

<p>See Also : THRESHOLD</p>

<h1 id="pretty_print">pretty_print</h1>

<pre><code><code> <span class="variable">Title</span>     <span class="operator">:</span> <span class="variable">pretty_print</span>
 <span class="variable">Usage</span>     <span class="operator">:</span> <span class="variable">$output</span> <span class="operator">=</span> <span class="variable">$sig</span><span class="operator">-&gt;</span><span class="variable">pretty_print</span><span class="operator">;</span>
           <span class="operator">:</span> <span class="keyword">print</span> <span class="variable">$sig</span><span class="operator">-&gt;</span><span class="variable">pretty_print</span><span class="operator">;</span>
           <span class="operator">:</span>
 <span class="variable">Purpose</span>   <span class="operator">:</span> <span class="variable">Emulates</span> <span class="variable">the</span> <span class="variable">output</span> <span class="variable">of</span> <span class="variable">the</span> <span class="variable">EGCG</span> <span class="variable">Sigcleave</span>
           <span class="operator">:</span> <span class="variable">utility</span><span class="operator">.</span>
           <span class="operator">:</span> 
 <span class="variable">Returns</span>   <span class="operator">:</span> <span class="variable">A</span> <span class="variable">formatted</span> <span class="variable">string</span><span class="operator">.</span>
 <span class="variable">Argument</span>  <span class="operator">:</span> <span class="variable">none</span><span class="operator">.</span>
 <span class="variable">Throws</span>    <span class="operator">:</span> <span class="variable">none</span><span class="operator">.</span>
 <span class="variable">Comments</span>  <span class="operator">:</span> <span class="variable">none</span><span class="operator">.</span>
 <span class="variable">See</span> <span class="variable">Also</span>   <span class="operator">:</span> <span class="variable">n</span><span class="operator">/</span><span class="variable">a</span>
</code></code></pre>


</body>

</html>


