<?xml version="1.0" ?>
<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.0 Transitional//EN">
<html>
<head>
<!-- saved from url=(0017)http://localhost/ -->
<script language="JavaScript" src="../../../../displayToc.js"></script>
<script language="JavaScript" src="../../../../tocParas.js"></script>
<script language="JavaScript" src="../../../../tocTab.js"></script>
<link rel="stylesheet" type="text/css" href="../../../../scineplex.css">
<title></title>
<link rel="stylesheet" href="../../../../Active.css" type="text/css" />
<meta http-equiv="content-type" content="text/html; charset=utf-8" />
<link rev="made" href="mailto:support@ActiveState.com" />
</head>

<body>



<ul id="index">
  <li><a href="#NAME">NAME</a></li>
  <li><a href="#SYNOPSIS">SYNOPSIS</a></li>
  <li><a href="#DESCRIPTION">DESCRIPTION</a></li>
  <li><a href="#FEEDBACK">FEEDBACK</a>
    <ul>
      <li><a href="#Mailing-Lists">Mailing Lists</a></li>
      <li><a href="#Support">Support</a></li>
      <li><a href="#Reporting-Bugs">Reporting Bugs</a></li>
    </ul>
  </li>
  <li><a href="#AUTHOR---Lincoln-Stein">AUTHOR - Lincoln Stein</a></li>
  <li><a href="#APPENDIX">APPENDIX</a>
    <ul>
      <li><a href="#seq_id">seq_id</a></li>
      <li><a href="#display_name">display_name</a></li>
      <li><a href="#start">start</a></li>
      <li><a href="#end">end</a></li>
      <li><a href="#length">length</a></li>
      <li><a href="#seq">seq</a></li>
      <li><a href="#ref">ref</a></li>
      <li><a href="#absolute">absolute</a></li>
      <li><a href="#features">features</a></li>
      <li><a href="#overlapping_features">overlapping_features</a></li>
      <li><a href="#contained_features">contained_features</a></li>
      <li><a href="#contained_in">contained_in</a></li>
      <li><a href="#get_feature_stream">get_feature_stream</a></li>
      <li><a href="#factory">factory</a></li>
      <li><a href="#primary_tag">primary_tag</a></li>
      <li><a href="#strand">strand</a></li>
    </ul>
  </li>
</ul>

<h1 id="NAME">NAME</h1>

<p>Bio::Das::SegmentI - DAS-style access to a feature database</p>

<h1 id="SYNOPSIS">SYNOPSIS</h1>

<pre><code><code>  <span class="comment"># Get a Bio::Das::SegmentI object from a Bio::DasI database...</span>
  
  <span class="variable">$segment</span> <span class="operator">=</span> <span class="variable">$das</span><span class="operator">-&gt;</span><span class="variable">segment</span><span class="operator">(</span><span class="string">-name</span><span class="operator">=&gt;</span><span class="string">'Landmark'</span><span class="operator">,</span>
                           <span class="string">-start</span><span class="operator">=&gt;</span><span class="variable">$start</span><span class="operator">,</span>
                           <span class="string">-end</span> <span class="operator">=&gt;</span> <span class="variable">$end</span><span class="operator">);</span>
  
  <span class="variable">@features</span> <span class="operator">=</span> <span class="variable">$segment</span><span class="operator">-&gt;</span><span class="variable">overlapping_features</span><span class="operator">(</span><span class="string">-type</span><span class="operator">=&gt;</span><span class="operator">[</span><span class="string">'type1'</span><span class="operator">,</span><span class="string">'type2'</span><span class="operator">]</span><span class="operator">);</span>
  <span class="comment"># each feature is a Bio::SeqFeatureI-compliant object</span>
  
  <span class="variable">@features</span> <span class="operator">=</span> <span class="variable">$segment</span><span class="operator">-&gt;</span><span class="variable">contained_features</span><span class="operator">(</span><span class="string">-type</span><span class="operator">=&gt;</span><span class="operator">[</span><span class="string">'type1'</span><span class="operator">,</span><span class="string">'type2'</span><span class="operator">]</span><span class="operator">);</span>
  
  <span class="variable">@features</span> <span class="operator">=</span> <span class="variable">$segment</span><span class="operator">-&gt;</span><span class="variable">contained_in</span><span class="operator">(</span><span class="string">-type</span><span class="operator">=&gt;</span><span class="operator">[</span><span class="string">'type1'</span><span class="operator">,</span><span class="string">'type2'</span><span class="operator">]</span><span class="operator">);</span>
  
  <span class="variable">$stream</span> <span class="operator">=</span> <span class="variable">$segment</span><span class="operator">-&gt;</span><span class="variable">get_feature_stream</span><span class="operator">(</span><span class="string">-type</span><span class="operator">=&gt;</span><span class="operator">[</span><span class="string">'type1'</span><span class="operator">,</span><span class="string">'type2'</span><span class="operator">,</span><span class="string">'type3'</span><span class="operator">]</span><span class="operator">;</span>
  <span class="keyword">while</span> <span class="operator">(</span><span class="keyword">my</span> <span class="variable">$feature</span> <span class="operator">=</span> <span class="variable">$stream</span><span class="operator">-&gt;</span><span class="variable">next_seq</span><span class="operator">)</span> <span class="operator">{</span>
     <span class="comment"># do something with feature</span>
  <span class="operator">}</span>
  
  <span class="variable">$count</span> <span class="operator">=</span> <span class="variable">$segment</span><span class="operator">-&gt;</span><span class="variable">features_callback</span><span class="operator">(</span><span class="string">-type</span><span class="operator">=&gt;</span><span class="operator">[</span><span class="string">'type1'</span><span class="operator">,</span><span class="string">'type2'</span><span class="operator">,</span><span class="string">'type3'</span><span class="operator">]</span><span class="operator">,</span>
                                       <span class="string">-callback</span> <span class="operator">=&gt;</span> <span class="keyword">sub</span><span class="variable"> </span><span class="operator">{</span> <span class="operator">...</span> <span class="operator">{</span> <span class="operator">}</span>
                                       <span class="operator">);</span>
</code></code></pre>

<h1 id="DESCRIPTION">DESCRIPTION</h1>

<p>Bio::Das::SegmentI is a simplified alternative interface to sequence annotation databases used by the distributed annotation system. In this scheme, the genome is represented as a series of landmarks. Each Bio::Das::SegmentI object (&quot;segment&quot;) corresponds to a genomic region defined by a landmark and a start and end position relative to that landmark. A segment is created using the Bio::DasI segment() method.</p>

<p>Features can be filtered by the following attributes:</p>

<pre><code><code>  1) their location relative to the segment (whether overlapping,
          contained within, or completely containing)

  2) their type

  3) other attributes using tag/value semantics</code></code></pre>

<p>Access to the feature list uses three distinct APIs:</p>

<pre><code><code>  1) fetching entire list of features at a time

  2) fetching an iterator across features

  3) a callback</code></code></pre>

<h1 id="FEEDBACK">FEEDBACK</h1>

<h2 id="Mailing-Lists">Mailing Lists</h2>

<p>User feedback is an integral part of the evolution of this and other Bioperl modules. Send your comments and suggestions preferably to one of the Bioperl mailing lists. Your participation is much appreciated.</p>

<pre><code><code>  bioperl-l@bio.perl.org</code></code></pre>

<h2 id="Support">Support</h2>

<p>Please direct usage questions or support issues to the mailing list:</p>

<p><i>bioperl-l@bioperl.org</i></p>

<p>rather than to the module maintainer directly. Many experienced and reponsive experts will be able look at the problem and quickly address it. Please include a thorough description of the problem with code and data examples if at all possible.</p>

<h2 id="Reporting-Bugs">Reporting Bugs</h2>

<p>Report bugs to the Bioperl bug tracking system to help us keep track the bugs and their resolution. Bug reports can be submitted via the web:</p>

<pre><code><code>  https://redmine.open-bio.org/projects/bioperl/</code></code></pre>

<h1 id="AUTHOR---Lincoln-Stein">AUTHOR - Lincoln Stein</h1>

<p>Email lstein@cshl.org</p>

<h1 id="APPENDIX">APPENDIX</h1>

<p>The rest of the documentation details each of the object methods. Internal methods are usually preceded with a _</p>

<h2 id="seq_id">seq_id</h2>

<pre><code><code> Title   : seq_id
 Usage   : $ref = $s-&gt;seq_id
 Function: return the ID of the landmark
 Returns : a string
 Args    : none
 Status  : Public</code></code></pre>

<h2 id="display_name">display_name</h2>

<pre><code><code> Title   : seq_name
 Usage   : $ref = $s-&gt;seq_name
 Function: return the human-readable name for the landmark
 Returns : a string
 Args    : none
 Status  : Public</code></code></pre>

<p>This defaults to the same as seq_id.</p>

<h2 id="start">start</h2>

<pre><code><code> Title   : start
 Usage   : $s-&gt;start
 Function: start of segment
 Returns : integer
 Args    : none
 Status  : Public</code></code></pre>

<p>This is a read-only accessor for the start of the segment. Alias to low() for Gadfly compatibility.</p>

<h2 id="end">end</h2>

<pre><code><code> Title   : end
 Usage   : $s-&gt;end
 Function: end of segment
 Returns : integer
 Args    : none
 Status  : Public</code></code></pre>

<p>This is a read-only accessor for the end of the segment. Alias to high() for Gadfly compatibility.</p>

<h2 id="length">length</h2>

<pre><code><code> Title   : length
 Usage   : $s-&gt;length
 Function: length of segment
 Returns : integer
 Args    : none
 Status  : Public</code></code></pre>

<p>Returns the length of the segment. Always a positive number.</p>

<h2 id="seq">seq</h2>

<pre><code><code> Title   : seq
 Usage   : $s-&gt;seq
 Function: get the sequence string for this segment
 Returns : a string
 Args    : none
 Status  : Public</code></code></pre>

<p>Returns the sequence for this segment as a simple string.</p>

<h2 id="ref">ref</h2>

<pre><code><code> Title   : ref
 Usage   : $ref = $s-&gt;ref([$newlandmark])
 Function: get/set the reference landmark for addressing
 Returns : a string
 Args    : none
 Status  : Public</code></code></pre>

<p>This method is used to examine/change the reference landmark used to establish the coordinate system. By default, the landmark cannot be changed and therefore this has the same effect as seq_id(). The new landmark might be an ID, or another Das::SegmentI object.</p>

<h2 id="absolute">absolute</h2>

<pre><code><code> Title   : absolute
 Usage   : $s-&gt;absolute([$new_value])
 Function: get/set absolute addressing mode
 Returns : flag
 Args    : new flag (optional)
 Status  : Public</code></code></pre>

<p>Turn on and off absolute-addressing mode. In absolute addressing mode, coordinates are relative to some underlying &quot;top level&quot; coordinate system (such as a chromosome). ref() returns the identity of the top level landmark, and start() and end() return locations relative to that landmark. In relative addressing mode, coordinates are relative to the landmark sequence specified at the time of segment creation or later modified by the ref() method.</p>

<p>The default is to return false and to do nothing in response to attempts to set absolute addressing mode.</p>

<h2 id="features">features</h2>

<pre><code><code> Title   : features
 Usage   : @features = $s-&gt;features(@args)
 Function: get features that overlap this segment
 Returns : a list of Bio::SeqFeatureI objects
 Args    : see below
 Status  : Public</code></code></pre>

<p>This method will find all features that intersect the segment in a variety of ways and return a list of Bio::SeqFeatureI objects. The feature locations will use coordinates relative to the reference sequence in effect at the time that features() was called.</p>

<p>The returned list can be limited to certain types, attributes or range intersection modes. Types of range intersection are one of:</p>

<pre><code><code>   &quot;overlaps&quot;      the default
   &quot;contains&quot;      return features completely contained within the segment
   &quot;contained_in&quot;  return features that completely contain the segment</code></code></pre>

<p>Two types of argument lists are accepted. In the positional argument form, the arguments are treated as a list of feature types. In the named parameter form, the arguments are a series of -name=&gt;value pairs.</p>

<pre><code><code>  Argument    Description
  --------   ------------

  -types      An array reference to type names in the format
              &quot;method:source&quot;

  -attributes A hashref containing a set of attributes to match

  -rangetype  One of &quot;overlaps&quot;, &quot;contains&quot;, or &quot;contained_in&quot;.

  -iterator   Return an iterator across the features.

  -callback   A callback to invoke on each feature</code></code></pre>

<p>The -attributes argument is a hashref containing one or more attributes to match against:</p>

<pre><code><code>  <span class="string">-attributes</span> <span class="operator">=&gt;</span> <span class="operator">{</span> <span class="string">Gene</span> <span class="operator">=&gt;</span> <span class="string">'abc-1'</span><span class="operator">,</span>
                   <span class="string">Note</span> <span class="operator">=&gt;</span> <span class="string">'confirmed'</span> <span class="operator">}</span>
</code></code></pre>

<p>Attribute matching is simple string matching, and multiple attributes are ANDed together. More complex filtering can be performed using the -callback option (see below).</p>

<p>If -iterator is true, then the method returns an object reference that implements the next_seq() method. Each call to next_seq() returns a new Bio::SeqFeatureI object.</p>

<p>If -callback is passed a code reference, the code reference will be invoked on each feature returned. The code will be passed two arguments consisting of the current feature and the segment object itself, and must return a true value. If the code returns a false value, feature retrieval will be aborted.</p>

<p>-callback and -iterator are mutually exclusive options. If -iterator is defined, then -callback is ignored.</p>

<p>NOTE: the following methods all build on top of features(), and do not need to be explicitly implemented.</p>

<pre><code><code>    overlapping_features()
    contained_features()
    contained_in()
    get_feature_stream()</code></code></pre>

<h2 id="overlapping_features">overlapping_features</h2>

<pre><code><code> Title   : overlapping_features
 Usage   : @features = $s-&gt;overlapping_features(@args)
 Function: get features that overlap this segment
 Returns : a list of Bio::SeqFeatureI objects
 Args    : see below
 Status  : Public</code></code></pre>

<p>This method is identical to features() except that it defaults to finding overlapping features.</p>

<h2 id="contained_features">contained_features</h2>

<pre><code><code> Title   : contained_features
 Usage   : @features = $s-&gt;contained_features(@args)
 Function: get features that are contained in this segment
 Returns : a list of Bio::SeqFeatureI objects
 Args    : see below
 Status  : Public</code></code></pre>

<p>This method is identical to features() except that it defaults to a range type of &#39;contained&#39;.</p>

<h2 id="contained_in">contained_in</h2>

<pre><code><code> Title   : contained_in
 Usage   : @features = $s-&gt;contained_in(@args)
 Function: get features that contain this segment
 Returns : a list of Bio::SeqFeatureI objects
 Args    : see below
 Status  : Public</code></code></pre>

<p>This method is identical to features() except that it defaults to a range type of &#39;contained_in&#39;.</p>

<h2 id="get_feature_stream">get_feature_stream</h2>

<pre><code><code> Title   : get_feature_stream
 Usage   : $iterator = $s-&gt;get_feature_stream(@args)
 Function: get an iterator across the segment
 Returns : an object that implements next_seq()
 Args    : see below
 Status  : Public</code></code></pre>

<p>This method is identical to features() except that it always generates an iterator.</p>

<p>NOTE: This is defined in the interface in terms of features(). You do not have to implement it.</p>

<h2 id="factory">factory</h2>

<pre><code><code> Title   : factory
 Usage   : $factory = $s-&gt;factory
 Function: return the segment factory
 Returns : a Bio::DasI object
 Args    : see below
 Status  : Public</code></code></pre>

<p>This method returns a Bio::DasI object that can be used to fetch more segments. This is typically the Bio::DasI object from which the segment was originally generated.</p>

<h2 id="primary_tag">primary_tag</h2>

<pre><code><code> Title   : primary_tag
 Usage   : $tag = $s-&gt;primary_tag
 Function: identifies the segment as type &quot;DasSegment&quot;
 Returns : a string named &quot;DasSegment&quot;
 Args    : none
 Status  : Public, but see below</code></code></pre>

<p>This method provides Bio::Das::Segment objects with a primary_tag() field that identifies them as being of type &quot;DasSegment&quot;. This allows the Bio::Graphics engine to render segments just like a feature in order nis way useful.</p>

<p>This does not need to be implemented. It is defined by the interface.</p>

<h2 id="strand">strand</h2>

<pre><code><code> Title   : strand
 Usage   : $strand = $s-&gt;strand
 Function: identifies the segment strand as 0
 Returns : the number 0
 Args    : none
 Status  : Public, but see below</code></code></pre>

<p>This method provides Bio::Das::Segment objects with a strand() field that identifies it as being strandless. This allows the Bio::Graphics engine to render segments just like a feature in order nis way useful.</p>

<p>This does not need to be implemented. It is defined by the interface.</p>


</body>

</html>


