<?xml version="1.0" ?>
<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.0 Transitional//EN">
<html>
<head>
<!-- saved from url=(0017)http://localhost/ -->
<script language="JavaScript" src="../../../../../../../displayToc.js"></script>
<script language="JavaScript" src="../../../../../../../tocParas.js"></script>
<script language="JavaScript" src="../../../../../../../tocTab.js"></script>
<link rel="stylesheet" type="text/css" href="../../../../../../../scineplex.css">
<title></title>
<link rel="stylesheet" href="../../../../../../../Active.css" type="text/css" />
<meta http-equiv="content-type" content="text/html; charset=utf-8" />
<link rev="made" href="mailto:support@ActiveState.com" />
</head>

<body>



<ul id="index">
  <li><a href="#NAME">NAME</a></li>
  <li><a href="#SYNOPSIS">SYNOPSIS</a></li>
  <li><a href="#DESCRIPTION">DESCRIPTION</a></li>
  <li><a href="#FEEDBACK">FEEDBACK</a>
    <ul>
      <li><a href="#Mailing-Lists">Mailing Lists</a></li>
      <li><a href="#Support">Support</a></li>
      <li><a href="#Reporting-Bugs">Reporting Bugs</a></li>
    </ul>
  </li>
  <li><a href="#AUTHOR---Jason-Stajich">AUTHOR - Jason Stajich</a></li>
  <li><a href="#CONTRIBUTORS">CONTRIBUTORS</a></li>
  <li><a href="#APPENDIX">APPENDIX</a>
    <ul>
      <li><a href="#Default-Values">Default Values</a></li>
      <li><a href="#program_name">program_name</a></li>
      <li><a href="#program_dir">program_dir</a></li>
      <li><a href="#new">new</a></li>
      <li><a href="#prepare">prepare</a></li>
      <li><a href="#run">run</a></li>
      <li><a href="#error_string">error_string</a></li>
      <li><a href="#alignment">alignment</a></li>
      <li><a href="#tree">tree</a></li>
      <li><a href="#get_parameters">get_parameters</a></li>
      <li><a href="#set_parameter">set_parameter</a></li>
      <li><a href="#set_default_parameters">set_default_parameters</a></li>
    </ul>
  </li>
  <li><a href="#Bio::Tools::Run::WrapperBase-methods">Bio::Tools::Run::WrapperBase methods</a>
    <ul>
      <li><a href="#no_param_checks">no_param_checks</a></li>
      <li><a href="#save_tempfiles">save_tempfiles</a></li>
      <li><a href="#outfile_name">outfile_name</a></li>
      <li><a href="#tempdir">tempdir</a></li>
      <li><a href="#cleanup">cleanup</a></li>
      <li><a href="#io">io</a></li>
    </ul>
  </li>
</ul>

<h1 id="NAME">NAME</h1>

<p>Bio::Tools::Run::Phylo::PAML::Codeml - Wrapper aroud the PAML program codeml</p>

<h1 id="SYNOPSIS">SYNOPSIS</h1>

<pre><code><code>  <span class="keyword">use</span> <span class="variable">Bio::Tools::Run::Phylo::PAML::Codeml</span><span class="operator">;</span>
  <span class="keyword">use</span> <span class="variable">Bio::AlignIO</span><span class="operator">;</span>
  
  <span class="keyword">my</span> <span class="variable">$alignio</span> <span class="operator">=</span> <span class="variable">Bio::AlignIO</span><span class="operator">-&gt;</span><span class="variable">new</span><span class="operator">(</span><span class="string">-format</span> <span class="operator">=&gt;</span> <span class="string">'phylip'</span><span class="operator">,</span>
                                 <span class="string">-file</span>   <span class="operator">=&gt;</span> <span class="string">'t/data/gf-s85.phylip'</span><span class="operator">);</span>
  
  <span class="keyword">my</span> <span class="variable">$aln</span> <span class="operator">=</span> <span class="variable">$alignio</span><span class="operator">-&gt;</span><span class="variable">next_aln</span><span class="operator">;</span>
  
  <span class="keyword">my</span> <span class="variable">$codeml</span> <span class="operator">=</span> <span class="variable">Bio::Tools::Run::Phylo::PAML::Codeml</span><span class="operator">-&gt;</span><span class="variable">new</span><span class="operator">();</span>
  <span class="variable">$codeml</span><span class="operator">-&gt;</span><span class="variable">alignment</span><span class="operator">(</span><span class="variable">$aln</span><span class="operator">);</span>
  <span class="keyword">my</span> <span class="operator">(</span><span class="variable">$rc</span><span class="operator">,</span><span class="variable">$parser</span><span class="operator">)</span> <span class="operator">=</span> <span class="variable">$codeml</span><span class="operator">-&gt;</span><span class="variable">run</span><span class="operator">();</span>
  <span class="keyword">my</span> <span class="variable">$result</span> <span class="operator">=</span> <span class="variable">$parser</span><span class="operator">-&gt;</span><span class="variable">next_result</span><span class="operator">;</span>
  <span class="keyword">my</span> <span class="variable">$MLmatrix</span> <span class="operator">=</span> <span class="variable">$result</span><span class="operator">-&gt;</span><span class="variable">get_MLmatrix</span><span class="operator">();</span>
  <span class="keyword">print</span> <span class="string">"Ka = "</span><span class="operator">,</span> <span class="variable">$MLmatrix</span><span class="operator">-&gt;</span><span class="operator">[</span><span class="number">0</span><span class="operator">]</span><span class="operator">-&gt;</span><span class="operator">[</span><span class="number">1</span><span class="operator">]</span><span class="operator">-&gt;</span><span class="operator">{</span><span class="string">'dN'</span><span class="operator">}</span><span class="operator">,</span><span class="string">"\n"</span><span class="operator">;</span>
  <span class="keyword">print</span> <span class="string">"Ks = "</span><span class="operator">,</span> <span class="variable">$MLmatrix</span><span class="operator">-&gt;</span><span class="operator">[</span><span class="number">0</span><span class="operator">]</span><span class="operator">-&gt;</span><span class="operator">[</span><span class="number">1</span><span class="operator">]</span><span class="operator">-&gt;</span><span class="operator">{</span><span class="string">'dS'</span><span class="operator">}</span><span class="operator">,</span><span class="string">"\n"</span><span class="operator">;</span>
  <span class="keyword">print</span> <span class="string">"Ka/Ks = "</span><span class="operator">,</span> <span class="variable">$MLmatrix</span><span class="operator">-&gt;</span><span class="operator">[</span><span class="number">0</span><span class="operator">]</span><span class="operator">-&gt;</span><span class="operator">[</span><span class="number">1</span><span class="operator">]</span><span class="operator">-&gt;</span><span class="operator">{</span><span class="string">'omega'</span><span class="operator">}</span><span class="operator">,</span><span class="string">"\n"</span><span class="operator">;</span>
</code></code></pre>

<h1 id="DESCRIPTION">DESCRIPTION</h1>

<p>This is a wrapper around the codeml program of PAML (Phylogenetic Analysis by Maximum Likelihood) package of Ziheng Yang. See http://abacus.gene.ucl.ac.uk/software/paml.html for more information.</p>

<p>This module is more about generating the properl codeml.ctl file and will run the program in a separate temporary directory to avoid creating temp files all over the place.</p>

<h1 id="FEEDBACK">FEEDBACK</h1>

<h2 id="Mailing-Lists">Mailing Lists</h2>

<p>User feedback is an integral part of the evolution of this and other Bioperl modules. Send your comments and suggestions preferably to the Bioperl mailing list. Your participation is much appreciated.</p>

<pre><code><code>  bioperl-l@bioperl.org                  - General discussion
  http://bioperl.org/wiki/Mailing_lists  - About the mailing lists</code></code></pre>

<h2 id="Support">Support</h2>

<p>Please direct usage questions or support issues to the mailing list:</p>

<p><i>bioperl-l@bioperl.org</i></p>

<p>rather than to the module maintainer directly. Many experienced and reponsive experts will be able look at the problem and quickly address it. Please include a thorough description of the problem with code and data examples if at all possible.</p>

<h2 id="Reporting-Bugs">Reporting Bugs</h2>

<p>Report bugs to the Bioperl bug tracking system to help us keep track of the bugs and their resolution. Bug reports can be submitted via the web:</p>

<pre><code><code>  http://redmine.open-bio.org/projects/bioperl/</code></code></pre>

<h1 id="AUTHOR---Jason-Stajich">AUTHOR - Jason Stajich</h1>

<p>Email jason-at-bioperl-dot-org</p>

<h1 id="CONTRIBUTORS">CONTRIBUTORS</h1>

<p>Additional contributors names and emails here</p>

<h1 id="APPENDIX">APPENDIX</h1>

<p>The rest of the documentation details each of the object methods. Internal methods are usually preceded with a _</p>

<h2 id="Default-Values">Default Values</h2>

<p>Valid and default values for codeml programs are listed below. The default values are always the first one listed. These descriptions are essentially lifted from the example codeml.ctl file and pamlDOC documentation provided by the author.</p>

<p><b>CodonFreq</b> specifies the equilibrium codon frequencies in codon substitution model. These frequencies can be assumed to be equal (1/61 each for the standard genetic code, <b>CodonFreq</b> = 0), calculated from the average nucleotide frequencies (<b>CodonFreq</b> = 1), from the average nucleotide frequencies at the three codon positions (<b>CodonFreq</b> = 2), or used as free parameters (<b>CodonFreq</b> = 3). The number of parameters involved in those models of codon frequencies is 0, 3, 9, and 60 (under the universal code), for <b>CodonFreq</b> = 0, 1, 2, and 3 respectively.</p>

<p><b>aaDist</b> specifies whether equal amino acid distances are assumed (= 0) or Grantham&#39;s matrix is used (= 1) (Yang et al. 1998).</p>

<p><b>runmode</b> = -2 performs ML estimation of dS and dN in pairwise comparisons. The program will collect estimates of dS and dN into the files 2ML.dS and 2ML.dN. Since many users seem interested in looking at dN /dS ratios among lineages, examination of the tree shapes indicated by branch lengths calculated from the two rates may be interesting although the analysis is ad hoc. If your species names have no more than 10 characters, you can use the output distance matrices as input to Phylip programs such as neighbor without change. Otherwise you need to edit the files to cut the names short.</p>

<p><b>model</b> concerns assumptions about the dN/dS rate ratios among branches (Yang 1998; Yang and Nielsen 1998). <b>model</b> =0 means a single dN/dS ratio for all lineages (branches), 1 means one ratio for each branch (free ratio model), and 2 means arbitrary number of rations (such as the 2-ratios or 3-ratios models. with <b>model</b> =2, you may specify the omega ratios for the branches using branch labels (read about the tree structure file in the document). This option seems rather easy to use. Otherwise, the program will ask the user to input a branch mark for the dN/dS ratio assumed for each branch. This should be an integral number between 0 to k - 1 if k different dN/dS ratios (omega_0 - omega_k - 1) are assumed for the branches of the tree. <b>Bioperl</b> note basically, doing this interactively is not going to work very well, so this module is really focused around using the 0 or 1 parameters. Read the program documentation if you&#39;d like some more detailed instructions.</p>

<p><b>NSsites</b> specifies models that allow the dN/dS ratio (omega) to vary among sites (Nielsen and Yang 1998, Yang et al. 2000) <b>Nssites</b> = m corresponds to model Mm in Yang et al (2000). The variable <b>ncatG</b> is used to specify the number of categories in the omega distribution under some models. The values of ncatG() used to perform our analyses are 3 for M3 (discrete), 5 for M4 (freq), 10 for the continuous distributions (M5: gamma, M6: 2gamma, M7: beta, M8:beta&amp;amp;w, M9:beta&amp;amp;gamma, M10: beta&amp;gamma+1, M11:beta&amp;amp;normal&amp;gt;1, and M12:0&amp;amp;2normal&amp;gt;1, M13:3normal&amp;gt;0). This means M8 will have 11 site classes (10 from the beta distribution plus 1 additional class). The posterior probabilities for site classes as well as the expected omega values for sites are listed in the file rst, which may be useful to pinpoint sites under positive selection, if they exist.</p>

<p>To make it easy to run several <b>Nssites</b> models in one go, the executable <a>Bio::Tools::Run::Phylo::PAML::Codemlsites</a> can be used, which asks you how many and which models to run at the start of the program. The number of categories used will then match those used in Yang et al(2000).</p>

<p>As noted in that paper, some of the models are hard to use, in particular, M12 and M13. Recommended models are 0 (one-ratio), 1 (neutral), 2 (selection), 3 (discrete), 7 (beta), and 8 (beta&amp;amp;omega ). Some of the models like M2 and M8 are noted to be prone to the problem of multiple local optima. You are advised to run the program at least twice, once with a starting omega value &lt;1 and a second time with a value &gt;1, and use the results corresponding to the highest likelihood. The continuous neutral and selection models of Nielsen and Yang (1998) are not implemented in the program.</p>

<p><b>icode</b> for genetic code and these correspond to 1-11 in the genbank transl table. 0:universal code 1:mamalian mt 2:yeast mt 3:mold mt, 4:invertebrate mt 5:ciliate nuclear 6:echinoderm mt 7:euplotid mt 8:alternative yeast nu. 9:ascidian mt 10:blepharisma nu</p>

<p><b>RateAncestor</b> For codon sequences, ancestral reconstruction is not implemented for the models of variable dN/dS ratios among sites. The output under codon-based models usually shows the encoded amino acid for each codon. The output under &quot;Prob of best character at each node, listed by site&quot; has two posterior probabilities for each node at each codon (amino acid) site. The first is for the best codon. The second, in parentheses, is for the most likely amino acid under the codon substitution model. This is a sum of posterior probabilities across synonymous codons. In theory it is possible although rare for the most likely amino acid not to match the most likely codon.</p>

<p><b>Output</b> for codon sequences (seqtype = 1): The codon frequencies in each sequence are counted and listed in a genetic code table, together with their sums across species. Each table contains six or fewer species. For data of multiple genes (option G in the sequence file), codon frequencies in each gene (summed over species) are also listed. The nucleotide distributions at the three codon positions are also listed. The method of Nei and Gojobori (1986) is used to calculate the number of synonymous substitutions per synonymous site (dS ) and the number of nonsynonymous substitutions per nonsynonymous site (dN ) and their ratio (dN /dS ). These are used to construct initial estimates of branch lengths for the likelihood analysis but are not MLEs themselves. Note that the estimates of these quantities for the a- and b-globin genes shown in Table 2 of Goldman and Yang (1994), calculated using the MEGA package (Kumar et al., 1993), are not accurate.</p>

<p>Results of ancestral reconstructions (<b>RateAncestor</b> = 1) are collected in the file rst. Under models of variable dN/dS ratios among sites (NSsites models), the posterior probabilities for site classes as well as positively selected sites are listed in rst.</p>

<p>INCOMPLETE DOCUMENTATION OF ALL METHODS</p>

<h2 id="program_name">program_name</h2>

<pre><code><code> Title   : program_name
 Usage   : $factory-&gt;program_name()
 Function: holds the program name
 Returns:  string
 Args    : None</code></code></pre>

<h2 id="program_dir">program_dir</h2>

<pre><code><code> Title   : program_dir
 Usage   : -&gt;program_dir()
 Function: returns the program directory, obtained from ENV variable.
 Returns:  string
 Args    :</code></code></pre>

<h2 id="new">new</h2>

<pre><code><code> <span class="variable">Title</span>   <span class="operator">:</span> <span class="variable">new</span>
 <span class="variable">Usage</span>   <span class="operator">:</span> <span class="keyword">my</span> <span class="variable">$obj</span> <span class="operator">=</span> <span class="variable">Bio::Tools::Run::Phylo::PAML::Codeml</span><span class="operator">-&gt;</span><span class="variable">new</span><span class="operator">();</span>
 <span class="variable">Function</span><span class="operator">:</span> <span class="variable">Builds</span> <span class="variable">a</span> <span class="variable">new</span> <span class="variable">Bio::Tools::Run::Phylo::PAML::Codeml</span> <span class="variable">object</span> 
 <span class="variable">Returns</span> <span class="operator">:</span> <span class="variable">Bio::Tools::Run::Phylo::PAML::Codeml</span>
 <span class="variable">Args</span>    <span class="operator">:</span> <span class="string">-alignment</span> <span class="operator">=&gt;</span> <span class="variable">the</span> <span class="variable">Bio::Align::AlignI</span> <span class="variable">object</span>
           <span class="operator">-</span><span class="string">save_tempfiles</span> <span class="operator">=&gt;</span> <span class="variable">boolean</span> <span class="variable">to</span> <span class="variable">save</span> <span class="variable">the</span> <span class="variable">generated</span> <span class="variable">tempfiles</span> <span class="keyword">and</span>
                              <span class="variable">NOT</span> <span class="variable">cleanup</span> <span class="variable">after</span> <span class="variable">onesself</span> <span class="operator">(</span><span class="keyword">default</span> <span class="variable">FALSE</span><span class="operator">)</span>
           <span class="string">-tree</span> <span class="operator">=&gt;</span> <span class="variable">the</span> <span class="variable">Bio::Tree::TreeI</span> <span class="variable">object</span>
           <span class="operator">-</span><span class="string">branchlengths</span> <span class="operator">=&gt;</span> <span class="number">0</span><span class="operator">:</span> <span class="variable">ignore</span> <span class="variable">any</span> <span class="variable">branch</span> <span class="variable">lengths</span> <span class="variable">found</span> <span class="variable">on</span> <span class="variable">the</span> <span class="variable">tree</span>
                             <span class="number">1</span><span class="operator">:</span> <span class="keyword">use</span> <span class="variable">as</span> <span class="variable">initial</span> <span class="keyword">values</span>
                             <span class="number">2</span><span class="operator">:</span> <span class="variable">fix</span> <span class="variable">branch</span> <span class="variable">lengths</span>
           <span class="operator">-</span><span class="string">params</span> <span class="operator">=&gt;</span> <span class="variable">a</span> <span class="variable">hashref</span> <span class="variable">of</span> <span class="variable">PAML</span> <span class="variable">parameters</span> <span class="operator">(</span><span class="variable">all</span> <span class="variable">passed</span> <span class="variable">to</span> <span class="variable">set_parameter</span><span class="operator">)</span>
           <span class="string">-executable</span> <span class="operator">=&gt;</span> <span class="variable">where</span> <span class="variable">the</span> <span class="variable">codeml</span> <span class="variable">executable</span> <span class="variable">resides</span>
</code></code></pre>

<p>See also: <a href="../../../../../../../lib/perl5/Bio/Tree/TreeI.html">Bio::Tree::TreeI</a>, <a href="../../../../../../../lib/perl5/Bio/Align/AlignI.html">Bio::Align::AlignI</a></p>

<h2 id="prepare">prepare</h2>

<pre><code><code> <span class="variable">Title</span>   <span class="operator">:</span> <span class="variable">prepare</span>
 <span class="variable">Usage</span>   <span class="operator">:</span> <span class="keyword">my</span> <span class="variable">$rundir</span> <span class="operator">=</span> <span class="variable">$codeml</span><span class="operator">-&gt;</span><span class="variable">prepare</span><span class="operator">(</span><span class="variable">$aln</span><span class="operator">);</span>
 <span class="variable">Function</span><span class="operator">:</span> <span class="variable">prepare</span> <span class="variable">the</span> <span class="variable">codeml</span> <span class="variable">analysis</span> <span class="variable">using</span> <span class="variable">the</span> <span class="keyword">default</span> <span class="keyword">or</span> <span class="variable">updated</span> <span class="variable">parameters</span>
           <span class="variable">the</span> <span class="variable">alignment</span> <span class="variable">parameter</span> <span class="variable">must</span> <span class="variable">have</span> <span class="variable">been</span> <span class="variable">set</span>
 <span class="variable">Returns</span> <span class="operator">:</span> <span class="variable">value</span> <span class="variable">of</span> <span class="variable">rundir</span>
 <span class="variable">Args</span>    <span class="operator">:</span> <span class="variable">L</span><span class="operator">&lt;</span><span class="variable">Bio::Align::AlignI</span><span class="operator">&gt;</span> <span class="variable">object</span><span class="operator">,</span>
           <span class="variable">L</span><span class="operator">&lt;</span><span class="variable">Bio::Tree::TreeI</span><span class="operator">&gt;</span> <span class="variable">object</span> <span class="operator">[</span><span class="variable">optional</span><span class="operator">]</span>
</code></code></pre>

<h2 id="run">run</h2>

<pre><code><code> <span class="variable">Title</span>   <span class="operator">:</span> <span class="variable">run</span>
 <span class="variable">Usage</span>   <span class="operator">:</span> <span class="keyword">my</span> <span class="operator">(</span><span class="variable">$rc</span><span class="operator">,</span><span class="variable">$parser</span><span class="operator">)</span> <span class="operator">=</span> <span class="variable">$codeml</span><span class="operator">-&gt;</span><span class="variable">run</span><span class="operator">(</span><span class="variable">$aln</span><span class="operator">,</span><span class="variable">$tree</span><span class="operator">);</span>
 <span class="variable">Function</span><span class="operator">:</span> <span class="variable">run</span> <span class="variable">the</span> <span class="variable">codeml</span> <span class="variable">analysis</span> <span class="variable">using</span> <span class="variable">the</span> <span class="keyword">default</span> <span class="keyword">or</span> <span class="variable">updated</span> <span class="variable">parameters</span>
           <span class="variable">the</span> <span class="variable">alignment</span> <span class="variable">parameter</span> <span class="variable">must</span> <span class="variable">have</span> <span class="variable">been</span> <span class="variable">set</span>
 <span class="variable">Returns</span> <span class="operator">:</span> <span class="variable">Return</span> <span class="variable">code</span><span class="operator">,</span> <span class="variable">L</span><span class="operator">&lt;</span><span class="variable">Bio::Tools::Phylo::PAML</span><span class="operator">&gt;</span>
 <span class="variable">Args</span>    <span class="operator">:</span> <span class="variable">L</span><span class="operator">&lt;</span><span class="variable">Bio::Align::AlignI</span><span class="operator">&gt;</span> <span class="variable">object</span><span class="operator">,</span>
           <span class="variable">L</span><span class="operator">&lt;</span><span class="variable">Bio::Tree::TreeI</span><span class="operator">&gt;</span> <span class="variable">object</span> <span class="operator">[</span><span class="variable">optional</span><span class="operator">]</span>
</code></code></pre>

<h2 id="error_string">error_string</h2>

<pre><code><code> Title   : error_string
 Usage   : $obj-&gt;error_string($newval)
 Function: Where the output from the last analysus run is stored.
 Returns : value of error_string
 Args    : newvalue (optional)</code></code></pre>

<h2 id="alignment">alignment</h2>

<pre><code><code> <span class="variable">Title</span>   <span class="operator">:</span> <span class="variable">alignment</span>
 <span class="variable">Usage</span>   <span class="operator">:</span> <span class="variable">$codeml</span><span class="operator">-&gt;</span><span class="variable">align</span><span class="operator">(</span><span class="variable">$aln</span><span class="operator">);</span>
 <span class="variable">Function</span><span class="operator">:</span> <span class="variable">Get</span><span class="operator">/</span><span class="variable">Set</span> <span class="variable">the</span> <span class="variable">L</span><span class="operator">&lt;</span><span class="variable">Bio::Align::AlignI</span><span class="operator">&gt;</span> <span class="variable">object</span>
 <span class="variable">Returns</span> <span class="operator">:</span> <span class="variable">L</span><span class="operator">&lt;</span><span class="variable">Bio::Align::AlignI</span><span class="operator">&gt;</span> <span class="variable">object</span>
 <span class="variable">Args</span>    <span class="operator">:</span> <span class="operator">[</span><span class="variable">optional</span><span class="operator">]</span> <span class="variable">L</span><span class="operator">&lt;</span><span class="variable">Bio::Align::AlignI</span><span class="operator">&gt;</span>
 <span class="variable">Comment</span> <span class="operator">:</span> <span class="variable">We</span> <span class="variable">could</span> <span class="variable">potentially</span> <span class="variable">add</span> <span class="variable">support</span> <span class="keyword">for</span> <span class="variable">running</span> <span class="variable">directly</span> <span class="variable">on</span> <span class="variable">a</span> <span class="variable">file</span>
           <span class="variable">but</span> <span class="variable">we</span> <span class="variable">shall</span> <span class="variable">keep</span> <span class="variable">it</span> <span class="variable">simple</span>
 <span class="variable">See</span> <span class="variable">also</span><span class="operator">:</span> <span class="variable">L</span><span class="operator">&lt;</span><span class="variable">Bio::SimpleAlign</span><span class="operator">&gt;</span>
</code></code></pre>

<h2 id="tree">tree</h2>

<pre><code><code> <span class="variable">Title</span>   <span class="operator">:</span> <span class="variable">tree</span>
 <span class="variable">Usage</span>   <span class="operator">:</span> <span class="variable">$codeml</span><span class="operator">-&gt;</span><span class="variable">tree</span><span class="operator">(</span><span class="variable">$tree</span><span class="operator">,</span> <span class="variable">%params</span><span class="operator">);</span>
 <span class="variable">Function</span><span class="operator">:</span> <span class="variable">Get</span><span class="operator">/</span><span class="variable">Set</span> <span class="variable">the</span> <span class="variable">L</span><span class="operator">&lt;</span><span class="variable">Bio::Tree::TreeI</span><span class="operator">&gt;</span> <span class="variable">object</span>
 <span class="variable">Returns</span> <span class="operator">:</span> <span class="variable">L</span><span class="operator">&lt;</span><span class="variable">Bio::Tree::TreeI</span><span class="operator">&gt;</span> 
 <span class="variable">Args</span>    <span class="operator">:</span> <span class="operator">[</span><span class="variable">optional</span><span class="operator">]</span> <span class="variable">$tree</span> <span class="operator">=&gt;</span> <span class="variable">L</span><span class="operator">&lt;</span><span class="variable">Bio::Tree::TreeI</span><span class="operator">&gt;,</span>
           <span class="operator">[</span><span class="variable">optional</span><span class="operator">]</span> <span class="variable">%parameters</span> <span class="operator">=&gt;</span> <span class="variable">hash</span> <span class="variable">of</span> <span class="variable">tree</span><span class="operator">-</span><span class="variable">specific</span> <span class="variable">parameters</span><span class="operator">:</span>
                  <span class="variable">branchLengths</span><span class="operator">:</span> <span class="number">0</span><span class="operator">,</span> <span class="number">1</span> <span class="keyword">or</span> <span class="number">2</span>
                  <span class="variable">out</span>
 
 <span class="variable">Comment</span> <span class="operator">:</span> <span class="variable">We</span> <span class="variable">could</span> <span class="variable">potentially</span> <span class="variable">add</span> <span class="variable">support</span> <span class="keyword">for</span> <span class="variable">running</span> <span class="variable">directly</span> <span class="variable">on</span> <span class="variable">a</span> <span class="variable">file</span>
           <span class="variable">but</span> <span class="variable">we</span> <span class="variable">shall</span> <span class="variable">keep</span> <span class="variable">it</span> <span class="variable">simple</span>
 <span class="variable">See</span> <span class="variable">also</span><span class="operator">:</span> <span class="variable">L</span><span class="operator">&lt;</span><span class="variable">Bio::Tree::Tree</span><span class="operator">&gt;</span>
</code></code></pre>

<h2 id="get_parameters">get_parameters</h2>

<pre><code><code> <span class="variable">Title</span>   <span class="operator">:</span> <span class="variable">get_parameters</span>
 <span class="variable">Usage</span>   <span class="operator">:</span> <span class="keyword">my</span> <span class="variable">%params</span> <span class="operator">=</span> <span class="variable">$self</span><span class="operator">-&gt;</span><span class="variable">get_parameters</span><span class="operator">();</span>
 <span class="variable">Function</span><span class="operator">:</span> <span class="variable">returns</span> <span class="variable">the</span> <span class="variable">list</span> <span class="variable">of</span> <span class="variable">parameters</span> <span class="variable">as</span> <span class="variable">a</span> <span class="variable">hash</span>
 <span class="variable">Returns</span> <span class="operator">:</span> <span class="variable">associative</span> <span class="variable">array</span> <span class="variable">keyed</span> <span class="variable">on</span> <span class="variable">parameter</span> <span class="variable">names</span>
 <span class="variable">Args</span>    <span class="operator">:</span> <span class="variable">none</span>
</code></code></pre>

<h2 id="set_parameter">set_parameter</h2>

<pre><code><code> <span class="variable">Title</span>   <span class="operator">:</span> <span class="variable">set_parameter</span>
 <span class="variable">Usage</span>   <span class="operator">:</span> <span class="variable">$codeml</span><span class="operator">-&gt;</span><span class="variable">set_parameter</span><span class="operator">(</span><span class="variable">$param</span><span class="operator">,</span><span class="variable">$val</span><span class="operator">);</span>
 <span class="variable">Function</span><span class="operator">:</span> <span class="variable">Sets</span> <span class="variable">a</span> <span class="variable">codeml</span> <span class="variable">parameter</span><span class="operator">,</span> <span class="variable">will</span> <span class="variable">be</span> <span class="variable">validated</span> <span class="variable">against</span>
           <span class="variable">the</span> <span class="variable">valid</span> <span class="keyword">values</span> <span class="variable">as</span> <span class="variable">set</span> <span class="variable">in</span> <span class="variable">the</span> <span class="variable">%VALIDVALUES</span> <span class="variable">class</span> <span class="variable">variable</span><span class="operator">.</span>  
           <span class="variable">The</span> <span class="variable">checks</span> <span class="variable">can</span> <span class="variable">be</span> <span class="variable">ignored</span> <span class="keyword">if</span> <span class="variable">one</span> <span class="variable">turns</span> <span class="variable">off</span> <span class="variable">param</span> <span class="variable">checks</span> <span class="variable">like</span> <span class="variable">this</span><span class="operator">:</span>
             <span class="variable">$codeml</span><span class="operator">-&gt;</span><span class="variable">no_param_checks</span><span class="operator">(</span><span class="number">1</span><span class="operator">)</span>
 <span class="variable">Returns</span> <span class="operator">:</span> <span class="variable">boolean</span> <span class="keyword">if</span> <span class="variable">set</span> <span class="variable">was</span> <span class="variable">success</span><span class="operator">,</span> <span class="keyword">if</span> <span class="variable">verbose</span> <span class="variable">is</span> <span class="variable">set</span> <span class="variable">to</span> <span class="operator">-</span><span class="number">1</span>
           <span class="variable">then</span> <span class="keyword">no</span> <span class="variable">warning</span> <span class="variable">will</span> <span class="variable">be</span> <span class="variable">reported</span>
 <span class="variable">Args</span>    <span class="operator">:</span> <span class="variable">$param</span> <span class="operator">=&gt;</span> <span class="variable">name</span> <span class="variable">of</span> <span class="variable">the</span> <span class="variable">parameter</span>
           <span class="variable">$value</span> <span class="operator">=&gt;</span> <span class="variable">value</span> <span class="variable">to</span> <span class="variable">set</span> <span class="variable">the</span> <span class="variable">parameter</span> <span class="variable">to</span>
 <span class="variable">See</span> <span class="variable">also</span><span class="operator">:</span> <span class="variable">L</span><span class="operator">&lt;</span><span class="variable">no_param_checks</span><span class="operator">()&gt;</span>
</code></code></pre>

<h2 id="set_default_parameters">set_default_parameters</h2>

<pre><code><code> <span class="variable">Title</span>   <span class="operator">:</span> <span class="variable">set_default_parameters</span>
 <span class="variable">Usage</span>   <span class="operator">:</span> <span class="variable">$codeml</span><span class="operator">-&gt;</span><span class="variable">set_default_parameters</span><span class="operator">(</span><span class="number">0</span><span class="operator">);</span>
 <span class="variable">Function</span><span class="operator">:</span> <span class="operator">(</span><span class="variable">Re</span><span class="operator">)</span><span class="variable">set</span> <span class="variable">the</span> <span class="keyword">default</span> <span class="variable">parameters</span> <span class="variable">from</span> <span class="variable">the</span> <span class="variable">defaults</span>
           <span class="operator">(</span><span class="variable">the</span> <span class="variable">first</span> <span class="variable">value</span> <span class="variable">in</span> <span class="keyword">each</span> <span class="variable">array</span> <span class="variable">in</span> <span class="variable">the</span> 
            <span class="variable">%VALIDVALUES</span> <span class="variable">class</span> <span class="variable">variable</span><span class="operator">)</span>
 <span class="variable">Returns</span> <span class="operator">:</span> <span class="variable">none</span>
 <span class="variable">Args</span>    <span class="operator">:</span> <span class="variable">boolean</span><span class="operator">:</span> <span class="variable">keep</span> <span class="variable">existing</span> <span class="variable">parameter</span> <span class="keyword">values</span>
</code></code></pre>

<h1 id="Bio::Tools::Run::WrapperBase-methods">Bio::Tools::Run::WrapperBase methods</h1>

<h2 id="no_param_checks">no_param_checks</h2>

<pre><code><code> Title   : no_param_checks
 Usage   : $obj-&gt;no_param_checks($newval)
 Function: Boolean flag as to whether or not we should
           trust the sanity checks for parameter values  
 Returns : value of no_param_checks
 Args    : newvalue (optional)</code></code></pre>

<h2 id="save_tempfiles">save_tempfiles</h2>

<pre><code><code> Title   : save_tempfiles
 Usage   : $obj-&gt;save_tempfiles($newval)
 Function: 
 Returns : value of save_tempfiles
 Args    : newvalue (optional)</code></code></pre>

<h2 id="outfile_name">outfile_name</h2>

<pre><code><code> <span class="variable">Title</span>   <span class="operator">:</span> <span class="variable">outfile_name</span>
 <span class="variable">Usage</span>   <span class="operator">:</span> <span class="keyword">my</span> <span class="variable">$outfile</span> <span class="operator">=</span> <span class="variable">$codeml</span><span class="operator">-&gt;</span><span class="variable">outfile_name</span><span class="operator">();</span>
 <span class="variable">Function</span><span class="operator">:</span> <span class="variable">Get</span><span class="operator">/</span><span class="variable">Set</span> <span class="variable">the</span> <span class="variable">name</span> <span class="variable">of</span> <span class="variable">the</span> <span class="variable">output</span> <span class="variable">file</span> <span class="keyword">for</span> <span class="variable">this</span> <span class="variable">run</span>
           <span class="operator">(</span><span class="keyword">if</span> <span class="variable">you</span> <span class="variable">wanted</span> <span class="variable">to</span> <span class="keyword">do</span> <span class="variable">something</span> <span class="variable">special</span><span class="operator">)</span>
 <span class="variable">Returns</span> <span class="operator">:</span> <span class="variable">string</span>
 <span class="variable">Args</span>    <span class="operator">:</span> <span class="operator">[</span><span class="variable">optional</span><span class="operator">]</span> <span class="variable">string</span> <span class="variable">to</span> <span class="variable">set</span> <span class="variable">value</span> <span class="variable">to</span>
</code></code></pre>

<h2 id="tempdir">tempdir</h2>

<pre><code><code> <span class="variable">Title</span>   <span class="operator">:</span> <span class="variable">tempdir</span>
 <span class="variable">Usage</span>   <span class="operator">:</span> <span class="keyword">my</span> <span class="variable">$tmpdir</span> <span class="operator">=</span> <span class="variable">$self</span><span class="operator">-&gt;</span><span class="variable">tempdir</span><span class="operator">();</span>
 <span class="variable">Function</span><span class="operator">:</span> <span class="variable">Retrieve</span> <span class="variable">a</span> <span class="variable">temporary</span> <span class="variable">directory</span> <span class="variable">name</span> <span class="operator">(</span><span class="variable">which</span> <span class="variable">is</span> <span class="variable">created</span><span class="operator">)</span>
 <span class="variable">Returns</span> <span class="operator">:</span> <span class="variable">string</span> <span class="variable">which</span> <span class="variable">is</span> <span class="variable">the</span> <span class="variable">name</span> <span class="variable">of</span> <span class="variable">the</span> <span class="variable">temporary</span> <span class="variable">directory</span>
 <span class="variable">Args</span>    <span class="operator">:</span> <span class="variable">none</span>
</code></code></pre>

<h2 id="cleanup">cleanup</h2>

<pre><code><code> <span class="variable">Title</span>   <span class="operator">:</span> <span class="variable">cleanup</span>
 <span class="variable">Usage</span>   <span class="operator">:</span> <span class="variable">$codeml</span><span class="operator">-&gt;</span><span class="variable">cleanup</span><span class="operator">();</span>
 <span class="variable">Function</span><span class="operator">:</span> <span class="variable">Will</span> <span class="variable">cleanup</span> <span class="variable">the</span> <span class="variable">tempdir</span> <span class="variable">directory</span> <span class="variable">after</span> <span class="variable">a</span> <span class="variable">PAML</span> <span class="variable">run</span>
 <span class="variable">Returns</span> <span class="operator">:</span> <span class="variable">none</span>
 <span class="variable">Args</span>    <span class="operator">:</span> <span class="variable">none</span>
</code></code></pre>

<h2 id="io">io</h2>

<pre><code><code> Title   : io
 Usage   : $obj-&gt;io($newval)
 Function:  Gets a L&lt;Bio::Root::IO&gt; object
 Returns : L&lt;Bio::Root::IO&gt;
 Args    : none</code></code></pre>


</body>

</html>


