<?xml version="1.0" ?>
<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.0 Transitional//EN">
<html>
<head>
<!-- saved from url=(0017)http://localhost/ -->
<script language="JavaScript" src="../../../../displayToc.js"></script>
<script language="JavaScript" src="../../../../tocParas.js"></script>
<script language="JavaScript" src="../../../../tocTab.js"></script>
<link rel="stylesheet" type="text/css" href="../../../../scineplex.css">
<title></title>
<link rel="stylesheet" href="../../../../Active.css" type="text/css" />
<meta http-equiv="content-type" content="text/html; charset=utf-8" />
<link rev="made" href="mailto:support@ActiveState.com" />
</head>

<body>



<ul id="index">
  <li><a href="#NAME">NAME</a></li>
  <li><a href="#SYNOPSIS">SYNOPSIS</a></li>
  <li><a href="#DESCRIPTION">DESCRIPTION</a>
    <ul>
      <li><a href="#RATIONALE">RATIONALE</a></li>
      <li><a href="#CONSTRUCTORS">CONSTRUCTORS</a></li>
      <li><a href="#OBJECT-METHODS">OBJECT METHODS</a></li>
    </ul>
  </li>
  <li><a href="#DETAILS">DETAILS</a>
    <ul>
      <li><a href="#How-the-process-is-working-with-one-example">How the process is working with one example</a></li>
      <li><a href="#Working-with-fuzzy-data.">Working with fuzzy data.</a></li>
    </ul>
  </li>
  <li><a href="#FEEDBACK">FEEDBACK</a>
    <ul>
      <li><a href="#Mailing-Lists">Mailing Lists</a></li>
      <li><a href="#Support">Support</a></li>
      <li><a href="#Reporting-Bugs">Reporting Bugs</a></li>
    </ul>
  </li>
  <li><a href="#AUTHOR---Pedro-M.-Gomez-Fabre">AUTHOR - Pedro M. Gomez-Fabre</a></li>
  <li><a href="#APPENDIX">APPENDIX</a>
    <ul>
      <li><a href="#new">new</a></li>
      <li><a href="#haplotype_block">haplotype_block</a></li>
      <li><a href="#snp_ids">snp_ids</a></li>
      <li><a href="#pattern_freq">pattern_freq</a></li>
      <li><a href="#check_input">_check_input</a></li>
      <li><a href="#haplotype_length_error">_haplotype_length_error</a></li>
      <li><a href="#population_error">_population_error</a></li>
      <li><a href="#do_it">_do_it</a></li>
      <li><a href="#input_block">input_block</a></li>
      <li><a href="#hap_length">hap_length</a></li>
      <li><a href="#pop_freq">pop_freq</a></li>
      <li><a href="#deg_snp">deg_snp</a></li>
      <li><a href="#snp_type">snp_type</a></li>
      <li><a href="#silent_snp">silent_snp</a></li>
      <li><a href="#useful_snp">useful_snp</a></li>
      <li><a href="#ht_type">ht_type</a></li>
      <li><a href="#ht_set">ht_set</a></li>
      <li><a href="#snp_type_code">snp_type_code</a></li>
      <li><a href="#snp_and_code">snp_and_code</a></li>
      <li><a href="#deg_pattern">deg_pattern</a></li>
      <li><a href="#split_hap">split_hap</a></li>
      <li><a href="#split_haplo">_split_haplo</a></li>
      <li><a href="#to_upper_case">_to_upper_case</a></li>
      <li><a href="#remove_deg">_remove_deg</a></li>
      <li><a href="#rem_silent_snp">_rem_silent_snp</a></li>
      <li><a href="#find_silent_snps">_find_silent_snps</a></li>
      <li><a href="#find_indet">_find_indet</a></li>
      <li><a href="#remove_col">_remove_col</a></li>
      <li><a href="#remove_snp_id">_remove_snp_id</a></li>
      <li><a href="#find_deg_pattern">_find_deg_pattern</a></li>
      <li><a href="#keep_these_patterns">_keep_these_patterns</a></li>
      <li><a href="#compare_arrays">compare_arrays</a></li>
      <li><a href="#convert_to_numbers">_convert_to_numbers</a></li>
      <li><a href="#snp_type_code1">_snp_type_code</a></li>
      <li><a href="#alleles_number">_alleles_number</a></li>
      <li><a href="#htSNP">_htSNP</a></li>
      <li><a href="#snp_and_code_summary">_snp_and_code_summary</a></li>
    </ul>
  </li>
</ul>

<h1 id="NAME">NAME</h1>

<p>Bio::PopGen::HtSNP.pm- Select htSNP from a haplotype set</p>

<h1 id="SYNOPSIS">SYNOPSIS</h1>

<pre><code><code>    <span class="keyword">use</span> <span class="variable">Bio::PopGen::HtSNP</span><span class="operator">;</span>
    
    <span class="keyword">my</span> <span class="variable">$obj</span> <span class="operator">=</span> <span class="variable">Bio::PopGen::HtSNP</span><span class="operator">-&gt;</span><span class="variable">new</span><span class="operator">(</span><span class="variable">$hap</span><span class="operator">,</span><span class="variable">$snp</span><span class="operator">,</span><span class="variable">$pop</span><span class="operator">);</span>
</code></code></pre>

<h1 id="DESCRIPTION">DESCRIPTION</h1>

<p>Select the minimal set of SNP that contains the full information about the haplotype without redundancies.</p>

<p>Take as input the followin values:</p>

<dl>

<dt id="the-haplotype-block-array-of-array-.">- the haplotype block (array of array).</dt>
<dd>

</dd>
<dt id="the-snp-id-array-.">- the snp id (array).</dt>
<dd>

</dd>
<dt id="family-information-and-frequency-array-of-array-.">- family information and frequency (array of array).</dt>
<dd>

</dd>
</dl>

<p>The final haplotype is generated in a numerical format and the SNP&#39;s sets can be retrieve from the module.</p>

<p><b>considerations:</b></p>

<p>- If you force to include a family with indetermination, the SNP&#39;s with indetermination will be removed from the analysis, so consider before to place your data set what do you really want to do.</p>

<p>- If two families have the same information (identical haplotype), one of them will be removed and the removed files will be stored classify as removed.</p>

<p>- Only are accepted for calculation A, C, G, T and - (as deletion) and their combinations. Any other value as n or ? will be considered as degenerations due to lack of information.</p>

<h2 id="RATIONALE">RATIONALE</h2>

<p>On a haplotype set is expected that some of the SNP and their variations contribute in the same way to the haplotype. Eliminating redundancies will produce a minimal set of SNP&#39;s that can be used as input for a taging selection process. On the process SNP&#39;s with the same variation are clustered on the same group.</p>

<p>The idea is that because the tagging haplotype process is exponential. All redundant information we could eliminate on the tagging process will help to find a quick result.</p>

<h2 id="CONSTRUCTORS">CONSTRUCTORS</h2>

<pre><code><code>  <span class="keyword">my</span> <span class="variable">$obj</span> <span class="operator">=</span> <span class="variable">Bio::PopGen::HtSNP</span><span class="operator">-&gt;</span><span class="variable">new</span>
    <span class="operator">(</span><span class="string">-haplotype_block</span> <span class="operator">=&gt;</span> <span class="operator">\</span><span class="variable">@haplotype_patterns</span><span class="operator">,</span>
     <span class="string">-snp_ids</span>         <span class="operator">=&gt;</span> <span class="operator">\</span><span class="variable">@snp_ids</span><span class="operator">,</span>
     <span class="string">-pattern_freq</span>    <span class="operator">=&gt;</span> <span class="operator">\</span><span class="variable">@pattern_name_and_freq</span><span class="operator">);</span>
</code></code></pre>

<p>where $hap, $snp and $pop are in the format:</p>

<pre><code><code>  <span class="keyword">my</span> <span class="variable">$hap</span> <span class="operator">=</span> <span class="operator">[</span>
             <span class="string">'acgt'</span><span class="operator">,</span>
             <span class="string">'agtc'</span><span class="operator">,</span>
             <span class="string">'cgtc'</span>
            <span class="operator">]</span><span class="operator">;</span>                     <span class="comment"># haplotype patterns' id</span>
  
  <span class="keyword">my</span> <span class="variable">$snp</span> <span class="operator">=</span> <span class="operator">[</span><span class="string">qw/s1 s2 s3 s4/</span><span class="operator">]</span><span class="operator">;</span>     <span class="comment"># snps' Id's</span>
  
  <span class="keyword">my</span> <span class="variable">$pop</span> <span class="operator">=</span> <span class="operator">[</span>
             <span class="operator">[</span><span class="string">qw/ uno    0.20/</span><span class="operator">]</span><span class="operator">,</span>
             <span class="operator">[</span><span class="string">qw/ dos    0.20/</span><span class="operator">]</span><span class="operator">,</span>
             <span class="operator">[</span><span class="string">qw/ tres   0.15/</span><span class="operator">]</span><span class="operator">,</span>
            <span class="operator">]</span><span class="operator">;</span>                     <span class="comment"># haplotype_pattern_id    Frequency</span>
</code></code></pre>

<h2 id="OBJECT-METHODS">OBJECT METHODS</h2>

<pre><code><code>    See Below for more detailed summaries.</code></code></pre>

<h1 id="DETAILS">DETAILS</h1>

<h2 id="How-the-process-is-working-with-one-example">How the process is working with one example</h2>

<p>Let&#39;s begin with one general example of the code.</p>

<p>Input haplotype:</p>

<pre><code><code>  acgtcca-t
  cggtagtgc
  cccccgtgc
  cgctcgtgc</code></code></pre>

<p>The first thing to to is to <b>split the haplotype</b> into characters.</p>

<pre><code><code>  a       c       g       t       c       c       a       -       t
  c       g       g       t       a       g       t       g       c
  c       c       c       c       c       g       t       g       c
  c       g       c       t       c       g       t       g       c</code></code></pre>

<p>Now we have to <b>convert</b> the haplotype to <b>Upercase</b>. This will produce the same SNP if we have input a or A.</p>

<pre><code><code>  A       C       G       T       C       C       A       -       T
  C       G       G       T       A       G       T       G       C
  C       C       C       C       C       G       T       G       C
  C       G       C       T       C       G       T       G       C</code></code></pre>

<p>The program admit as values any combination of ACTG and - (deletions). The haplotype is <b>converted to number</b>, considering the first variation as zero and the alternate value as 1 (see expanded description below).</p>

<pre><code><code>  0       0       0       0       0       0       0       0       0
  1       1       0       0       1       1       1       1       1
  1       0       1       1       0       1       1       1       1
  1       1       1       0       0       1       1       1       1</code></code></pre>

<p>Once we have the haplotype converted to numbers we have to generate the snp type information for the haplotype.</p>

<p><b>SNP code = SUM ( value * multiplicity ^ position );</b></p>

<pre><code><code>    where:
      SUM is the sum of the values for the SNP
      value is the SNP number code (0 [generally for the mayor allele],
                                    1 [for the minor allele].
      position is the position on the block.</code></code></pre>

<p>For this example the code is:</p>

<pre><code><code>  0       0       0       0       0       0       0       0       0
  1       1       0       0       1       1       1       1       1
  1       0       1       1       0       1       1       1       1
  1       1       1       0       0       1       1       1       1
 ------------------------------------------------------------------
  14      10      12      4       2       14      14      14      14

  14 = 0*2^0 + 1*2^1 + 1*2^2 + 1*2^3
  12 = 0*2^0 + 1*2^1 + 0*2^2 + 1*2^3
  ....</code></code></pre>

<p>Once we have the families classify. We will <b>take</b> just the SNP&#39;s <b>not redundant</b>.</p>

<pre><code><code>  14      10      12      4       2</code></code></pre>

<p>This information will be <b>passed to the tag module</b> is you want to tag the htSNP.</p>

<p>Whatever it happens to one SNPs of a class will happen to a SNP of the same class. Therefore you don&#39;t need to scan redundancies</p>

<h2 id="Working-with-fuzzy-data.">Working with fuzzy data.</h2>

<p>This module is designed to work with fuzzy data. As the source of the haplotype is diverse. The program assume that some haplotypes can be generated using different values. If there is any indetermination (? or n) or any other degenerated value or invalid. The program will take away This SNP and will leave that for a further analysis.</p>

<p>On a complex situation:</p>

<pre><code><code>  a       c       g       t       ?       c       a       c       t
  a       c       g       t       ?       c       a       -       t
  c       g       ?       t       a       g       ?       g       c
  c       a       c       t       c       g       t       g       c
  c       g       c       t       c       g       t       g       c
  c       g       g       t       a       g       ?       g       c
  a       c       ?       t       ?       c       a       c       t</code></code></pre>

<p>On this haplotype everything is happening. We have a multialelic variance. We have indeterminations. We have deletions and we have even one SNP which is not a real SNP.</p>

<p>The buiding process will be the same on this situation.</p>

<p>Convert the haplotype to uppercase.</p>

<pre><code><code>  A       C       G       T       ?       C       A       C       T
  A       C       G       T       ?       C       A       -       T
  C       G       ?       T       A       G       ?       G       C
  C       A       C       T       C       G       T       G       C
  C       G       C       T       C       G       T       G       C
  C       G       G       T       A       G       ?       G       C
  A       C       ?       T       ?       C       A       C       T</code></code></pre>

<p>All columns that present indeterminations will be removed from the analysis on this Step.</p>

<p>hapotype after remove columns:</p>

<pre><code><code>  A       C       T       C       C       T
  A       C       T       C       -       T
  C       G       T       G       G       C
  C       A       T       G       G       C
  C       G       T       G       G       C
  C       G       T       G       G       C
  A       C       T       C       C       T</code></code></pre>

<p>All changes made on the haplotype matrix, will be also made on the SNP list.</p>

<pre><code><code>  snp_id_1 snp_id_2 snp_id_4 snp_id_6 snp_id_8 snp_id_9</code></code></pre>

<p>now the SNP that is not one SNP will be removed from the analysis. SNP with Id snp_id_4 (the one with all T&#39;s).</p>

<p>because of the removing. Some of the families will become the same and will be clustered. A posteriori analysis will diference these families. but because of the indetermination can not be distinguish.</p>

<pre><code><code>  A       C       C       C       T
  A       C       C       -       T
  C       G       G       G       C
  C       A       G       G       C
  C       G       G       G       C
  C       G       G       G       C
  A       C       C       C       T</code></code></pre>

<p>The result of the mergering will go like:</p>

<pre><code><code>  A       C       C       C       T
  A       C       C       -       T
  C       G       G       G       C
  C       A       G       G       C</code></code></pre>

<p>Once again the changes made on the families and we merge the frequency (<i>to be implemented</i>)</p>

<p>Before to convert the haplotype into numbers we consider how many variations we have on the set. On this case the variations are 3.</p>

<p>The control code will use on this situation base three as mutiplicity</p>

<pre><code><code>  0       0       0       0       0
  0       0       0       1       0
  1       1       1       2       1
  1       2       1       2       1
 -----------------------------------
  36      63      36      75      36</code></code></pre>

<p>And the minimal set for this combination is</p>

<pre><code><code>  0       0       0
  0       0       1
  1       1       2
  1       2       2</code></code></pre>

<p><b>NOTE:</b> this second example is a remote example an on normal conditions. This conditions makes no sense, but as the haplotypes, can come from many sources we have to be ready for all kind of combinations.</p>

<h1 id="FEEDBACK">FEEDBACK</h1>

<h2 id="Mailing-Lists">Mailing Lists</h2>

<p>User feedback is an integral part of the evolution of this and other Bioperl modules. Send your comments and suggestions preferably to the Bioperl mailing list. Your participation is much appreciated.</p>

<pre><code><code>  bioperl-l@bioperl.org                  - General discussion
  http://bioperl.org/wiki/Mailing_lists  - About the mailing lists</code></code></pre>

<h2 id="Support">Support</h2>

<p>Please direct usage questions or support issues to the mailing list:</p>

<p><i>bioperl-l@bioperl.org</i></p>

<p>rather than to the module maintainer directly. Many experienced and reponsive experts will be able look at the problem and quickly address it. Please include a thorough description of the problem with code and data examples if at all possible.</p>

<h2 id="Reporting-Bugs">Reporting Bugs</h2>

<p>Report bugs to the Bioperl bug tracking system to help us keep track of the bugs and their resolution. Bug reports can be submitted via the web:</p>

<pre><code><code>  https://redmine.open-bio.org/projects/bioperl/</code></code></pre>

<h1 id="AUTHOR---Pedro-M.-Gomez-Fabre">AUTHOR - Pedro M. Gomez-Fabre</h1>

<p>Email pgf18872-at-gsk-dot-com</p>

<h1 id="APPENDIX">APPENDIX</h1>

<p>The rest of the documentation details each of the object methods. Internal methods are usually preceded with a _</p>

<h2 id="new">new</h2>

<pre><code><code> Title   : new
 Function: constructor of the class.
 Usage   : $obj-&gt; Bio::PopGen::HtSNP-&gt;new(-haplotype_block
                                          -snp_ids
                                          -pattern_freq)
 Returns : self hash
 Args    : input haplotype (array of array)
           snp_ids         (array)
           pop_freq        (array of array)
 Status  : public</code></code></pre>

<h2 id="haplotype_block">haplotype_block</h2>

<pre><code><code> <span class="variable">Title</span>   <span class="operator">:</span> <span class="variable">haplotype_block</span> 
 <span class="variable">Usage</span>   <span class="operator">:</span> <span class="keyword">my</span> <span class="variable">$haplotype_block</span> <span class="operator">=</span> <span class="variable">$HtSNP</span><span class="operator">-&gt;</span><span class="variable">haplotype_block</span><span class="operator">();</span>
 <span class="variable">Function</span><span class="operator">:</span> <span class="variable">Get</span> <span class="variable">the</span> <span class="variable">haplotype</span> <span class="variable">block</span> <span class="keyword">for</span> <span class="variable">a</span> <span class="variable">haplotype</span> <span class="variable">tagging</span> <span class="variable">selection</span>
 <span class="variable">Returns</span> <span class="operator">:</span> <span class="variable">reference</span> <span class="variable">of</span> <span class="variable">array</span> 
 <span class="variable">Args</span>    <span class="operator">:</span> <span class="variable">reference</span> <span class="variable">of</span> <span class="variable">array</span> <span class="variable">with</span> <span class="variable">haplotype</span> <span class="variable">pattern</span> 
</code></code></pre>

<h2 id="snp_ids">snp_ids</h2>

<pre><code><code> <span class="variable">Title</span>   <span class="operator">:</span> <span class="variable">snp_ids</span> 
 <span class="variable">Usage</span>   <span class="operator">:</span> <span class="keyword">my</span> <span class="variable">$snp_ids</span> <span class="operator">=</span> <span class="variable">$HtSNP</span><span class="operator">-&gt;</span><span class="variable">$snp_ids</span><span class="operator">();</span>
 <span class="variable">Function</span><span class="operator">:</span> <span class="variable">Get</span> <span class="variable">the</span> <span class="variable">ids</span> <span class="keyword">for</span> <span class="variable">a</span> <span class="variable">haplotype</span> <span class="variable">tagging</span> <span class="variable">selection</span>
 <span class="variable">Returns</span> <span class="operator">:</span> <span class="variable">reference</span> <span class="variable">of</span> <span class="variable">array</span>
 <span class="variable">Args</span>    <span class="operator">:</span> <span class="variable">reference</span> <span class="variable">of</span> <span class="variable">array</span> <span class="variable">with</span> <span class="variable">SNP</span> <span class="variable">ids</span>
</code></code></pre>

<h2 id="pattern_freq">pattern_freq</h2>

<pre><code><code> <span class="variable">Title</span>   <span class="operator">:</span> <span class="variable">pattern_freq</span>
 <span class="variable">Usage</span>   <span class="operator">:</span> <span class="keyword">my</span> <span class="variable">$pattern_freq</span> <span class="operator">=</span> <span class="variable">$HtSNP</span><span class="operator">-&gt;</span><span class="variable">pattern_freq</span><span class="operator">();</span>
 <span class="variable">Function</span><span class="operator">:</span> <span class="variable">Get</span> <span class="variable">the</span> <span class="variable">pattern</span> <span class="variable">id</span> <span class="keyword">and</span> <span class="variable">frequency</span>  <span class="keyword">for</span> <span class="variable">a</span> <span class="variable">haplotype</span>
           <span class="variable">tagging</span> <span class="variable">selection</span>
 <span class="variable">Returns</span> <span class="operator">:</span> <span class="variable">reference</span> <span class="variable">of</span> <span class="variable">array</span>
 <span class="variable">Args</span>    <span class="operator">:</span> <span class="variable">reference</span> <span class="variable">of</span> <span class="variable">array</span> <span class="variable">with</span> <span class="variable">SNP</span> <span class="variable">ids</span>
</code></code></pre>

<h2 id="check_input">_check_input</h2>

<pre><code><code> Title   : _check_input
 Usage   : _check_input($self)
 Function: check for errors on the input
 Returns : self hash
 Args    : self
 Status  : internal</code></code></pre>

<h2 id="haplotype_length_error">_haplotype_length_error</h2>

<pre><code><code> Title   : _haplotype_length_error
 Usage   : _haplotype_length_error($self)
 Function: check if the haplotype length is the same that the one on the
           SNP id list. If not break and exit
 Returns : self hash
 Args    : self
 Status  : internal</code></code></pre>

<h2 id="population_error">_population_error</h2>

<pre><code><code> Title   : _population_error
 Usage   : _population_error($self)
 Function: use input_block and pop_freq test if the number of elements
           match. If doesn&#39;t break and quit.
 Returns : self hash
 Args    : self
 Status  : internal</code></code></pre>

<h2 id="do_it">_do_it</h2>

<pre><code><code> Title   : _do_it
 Usage   : _do_it($self)
 Function: Process the input generating the results.
 Returns : self hash
 Args    : self
 Status  : internal</code></code></pre>

<h2 id="input_block">input_block</h2>

<pre><code><code> Title   : input_block
 Usage   : $obj-&gt;input_block()
 Function: returns input block
 Returns : reference to array of array
 Args    : none
 Status  : public</code></code></pre>

<h2 id="hap_length">hap_length</h2>

<pre><code><code> Title   : hap_length
 Usage   : $obj-&gt;hap_length()
 Function: get numbers of SNP on the haplotype
 Returns : scalar
 Args    : none
 Status  : public</code></code></pre>

<h2 id="pop_freq">pop_freq</h2>

<pre><code><code> Title   : pop_freq
 Usage   : $obj-&gt;pop_freq()
 Function: returns population frequency
 Returns : reference to array
 Args    : none
 Status  : public</code></code></pre>

<h2 id="deg_snp">deg_snp</h2>

<pre><code><code> Title   : deg_snp
 Usage   : $obj-&gt;deg_snp()
 Function: returns snp_removes due to indetermination on their values
 Returns : reference to array
 Args    : none
 Status  : public</code></code></pre>

<h2 id="snp_type">snp_type</h2>

<pre><code><code> Title   : snp_type
 Usage   : $obj-&gt;snp_type()
 Function: returns hash with SNP type
 Returns : reference to hash
 Args    : none
 Status  : public</code></code></pre>

<h2 id="silent_snp">silent_snp</h2>

<pre><code><code> Title   : silent_snp
 Usage   : $obj-&gt;silent_snp()
 Function: some SNP&#39;s are silent (not contibuting to the haplotype)
           and are not considering for this analysis
 Returns : reference to a array
 Args    : none
 Status  : public</code></code></pre>

<h2 id="useful_snp">useful_snp</h2>

<pre><code><code> Title   : useful_snp
 Usage   : $obj-&gt;useful_snp()
 Function: returns list of SNP&#39;s that are can be used as htSNP. Some
           of them can produce the same information. But this is
           not considered here.
 Returns : reference to a array
 Args    : none
 Status  : public</code></code></pre>

<h2 id="ht_type">ht_type</h2>

<pre><code><code> Title   : ht_type
 Usage   : $obj-&gt;ht_type()
 Function: every useful SNP has a numeric code dependending of its
           value and position. For a better description see
           description of the module.
 Returns : reference to a array
 Args    : none
 Status  : public</code></code></pre>

<h2 id="ht_set">ht_set</h2>

<pre><code><code> Title   : ht_set
 Usage   : $obj-&gt;ht_set()
 Function: returns the minimal haplotype in numerical format. This
           haplotype contains the maximal information about the
           haplotype variations but with no redundancies. It&#39;s the
           minimal set that describes the haplotype.
 Returns : reference to an array of arrays
 Args    : none
 Status  : public</code></code></pre>

<h2 id="snp_type_code">snp_type_code</h2>

<pre><code><code> Title   : snp_type_code
 Usage   : $obj-&gt;snp_type_code()
 Function: returns the numeric code of the SNPs that need to be
           tagged that correspond to the SNP&#39;s considered in ht_set.
 Returns : reference to an array
 Args    : none
 Status  : public</code></code></pre>

<h2 id="snp_and_code">snp_and_code</h2>

<pre><code><code> Title   : snp_and_code
 Usage   : $obj-&gt;snp_and_code()
 Function: Returns the full list of SNP&#39;s and the code associate to
           them. If the SNP belongs to the group useful_snp it keep
           this code. If the SNP is silent the code is 0. And if the
           SNP is degenerated the code is -1.
 Returns : reference to an array of array
 Args    : none
 Status  : public</code></code></pre>

<h2 id="deg_pattern">deg_pattern</h2>

<pre><code><code> Title   : deg_pattern
 Usage   : $obj-&gt;deg_pattern()
 Function: Returns the a list with the degenerated haplotype.
           Sometimes due to degeneration some haplotypes looks
           the same and if we don&#39;t remove them it won&#39;t find
           any tag.
 Returns : reference to a hash of array
 Args    : none
 Status  : public</code></code></pre>

<h2 id="split_hap">split_hap</h2>

<pre><code><code> Title   : split_hap
 Usage   : $obj-&gt;split_hap()
 Function: simple representation of the haplotype base by base
           Same information that input haplotype but base based.
 Returns : reference to an array of array
 Args    : none
 Status  : public</code></code></pre>

<h2 id="split_haplo">_split_haplo</h2>

<pre><code><code> Title   : _split_haplo
 Usage   : _split_haplo($self)
 Function: Take a haplotype and split it into bases
 Returns : self
 Args    : none
 Status  : internal</code></code></pre>

<h2 id="to_upper_case">_to_upper_case</h2>

<pre><code><code> Title   : _to_upper_case
 Usage   : _to_upper_case()
 Function: make SNP or in-dels Upper case
 Returns : self
 Args    : an AoA ref
 Status  : private</code></code></pre>

<h2 id="remove_deg">_remove_deg</h2>

<pre><code><code> Title   : _remove_deg
 Usage   : _remove_deg()
 Function: when have a indetermination or strange value this SNP
           is removed
 Returns : haplotype family set and degeneration list
 Args    : ref to an AoA and a ref to an array
 Status  : internal</code></code></pre>

<h2 id="rem_silent_snp">_rem_silent_snp</h2>

<pre><code><code> Title   : _rem_silent_snp
 Usage   : _rem_silent_snp()
 Function: there is the remote possibilty that one SNP won&#39;t be a
           real SNP on this situation we have to remove this SNP,
           otherwise the program won&#39;t find any tag
 Returns : nonthing
 Args    : ref to an AoA and a ref to an array
 Status  : internal</code></code></pre>

<h2 id="find_silent_snps">_find_silent_snps</h2>

<pre><code><code> Title   : _find_silent_snps
 Usage   :
 Function: list of snps that are not SNPs. All values for that
           SNPs on the set is the same one. Look stupid but can
           happend and if this happend you will not find any tag
 Returns : nothing
 Args    :
 Status  :</code></code></pre>

<h2 id="find_indet">_find_indet</h2>

<pre><code><code> Title   : _find_indet
 Usage   :
 Function: find column (SNP) with invalid or degenerated values
           and store this values into the second parameter supplied.
 Returns : nothing
 Args    : ref to AoA and ref to an array
 Status  : internal</code></code></pre>

<h2 id="remove_col">_remove_col</h2>

<pre><code><code> Title   : _remove_col
 Usage   :
 Function: remove columns contained on the second array from
           the first arr
 Returns : nothing
 Args    : array of array reference and array reference
 Status  : internal</code></code></pre>

<h2 id="remove_snp_id">_remove_snp_id</h2>

<pre><code><code> Title   : _remove_snp_id
 Usage   :
 Function: remove columns contained on the second array from
           the first arr
 Returns : nothing
 Args    : array of array reference and array reference
 Status  : internal</code></code></pre>

<h2 id="find_deg_pattern">_find_deg_pattern</h2>

<pre><code><code> Title   : _find_deg_pattern
 Usage   :
 Function: create a list with the degenerated patterns
 Returns : @array
 Args    : a ref to AoA
 Status  : public</code></code></pre>

<h2 id="keep_these_patterns">_keep_these_patterns</h2>

<pre><code><code> Title   : _keep_these_patterns
 Usage   :
 Function: this is a basic approach, take a LoL and a list,
           keep just the columns included on the list
 Returns : nothing
 Args    : an AoA and an array
 Status  : public</code></code></pre>

<h2 id="compare_arrays">compare_arrays</h2>

<pre><code><code> Title   : compare_arrays
 Usage   :
 Function: take two arrays and compare their values
 Returns : 1 if the two values are the same
           0 if the values are different
 Args    : an AoA and an array
 Status  : public</code></code></pre>

<h2 id="convert_to_numbers">_convert_to_numbers</h2>

<pre><code><code> Title   : _convert_to_numbers
 Usage   : _convert_to_numbers()
 Function: tranform the haplotype into numbers. before to do that
           we have to consider the variation on the set.
 Returns : nonthing
 Args    : ref to an AoA and a ref to an array
 Status  : internal</code></code></pre>

<h2 id="snp_type_code1">_snp_type_code</h2>

<pre><code><code> Title   : _snp_type_code
 Usage   :
 Function:
           we have to create the snp type code for each version.
           The way the snp type is created is the following:

           we take the number value for every SNP and do the
           following calculation

           let be a SNP set as follow:

           0    0
           1    1
           1    2

           and multiplicity 3
           on this case the situation is:

           sum (value * multiplicity ^ position) for each SNP

           0 * 3 ^ 0 + 1 * 3 ^ 1 + 1 * 3 ^ 2 = 12
           0 * 3 ^ 0 + 1 * 3 ^ 1 + 2 * 3 ^ 2 = 21
 Returns : nothing
 Args    : $self
 Status  : private</code></code></pre>

<h2 id="alleles_number">_alleles_number</h2>

<pre><code><code> Title   : _alleles_number
 Usage   :
 Function: calculate the max number of alleles for a haplotype and
           if the number. For each SNP the number is stored and the
           max number of alleles for a SNP on the set is returned
 Returns : max number of alleles (a scalar storing a number)
 Args    : ref to AoA
 Status  : public</code></code></pre>

<h2 id="htSNP">_htSNP</h2>

<pre><code><code> Title   : _htSNP
 Usage   : _htSNP()
 Function: calculate the minimal set that contains all information of the
           haplotype.
 Returns : nonthing
 Args    : ref to an AoA and a ref to an array
 Status  : internal</code></code></pre>

<h2 id="snp_and_code_summary">_snp_and_code_summary</h2>

<pre><code><code> Title   : _snp_and_code_summary
 Usage   : _snp_and_code_summary()
 Function: compile on a list all SNP and the code for each. This
           information can be also obtained combining snp_type and
           snp_type_code but on these results the information about
           the rest of SNP&#39;s are not compiled as table.

           0 will be silent SNPs
           -1 are degenerated SNPs
           and the rest of positive values are the code for useful SNP

 Returns : nonthing
 Args    : ref to an AoA and a ref to an array
 Status  : internal</code></code></pre>


</body>

</html>


