<?xml version="1.0" ?>
<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.0 Transitional//EN">
<html>
<head>
<!-- saved from url=(0017)http://localhost/ -->
<script language="JavaScript" src="../../../../../displayToc.js"></script>
<script language="JavaScript" src="../../../../../tocParas.js"></script>
<script language="JavaScript" src="../../../../../tocTab.js"></script>
<link rel="stylesheet" type="text/css" href="../../../../../scineplex.css">
<title></title>
<link rel="stylesheet" href="../../../../../Active.css" type="text/css" />
<meta http-equiv="content-type" content="text/html; charset=utf-8" />
<link rev="made" href="mailto:support@ActiveState.com" />
</head>

<body>



<ul id="index">
  <li><a href="#NAME">NAME</a></li>
  <li><a href="#SYNOPSIS">SYNOPSIS</a></li>
  <li><a href="#DESCRIPTION">DESCRIPTION</a>
    <ul>
      <li><a href="#new">new</a></li>
      <li><a href="#Bio::SeqFeatureI-methods">Bio::SeqFeatureI methods</a></li>
      <li><a href="#add_SeqFeature">add_SeqFeature</a></li>
      <li><a href="#update">update</a></li>
      <li><a href="#get_SeqFeatures">get_SeqFeatures</a></li>
      <li><a href="#object_store">object_store</a></li>
      <li><a href="#overloaded_names">overloaded_names</a></li>
      <li><a href="#segment">segment</a></li>
      <li><a href="#AUTOLOADED-methods">AUTOLOADED methods</a></li>
      <li><a href="#load_id">load_id</a></li>
      <li><a href="#notes">notes</a></li>
      <li><a href="#primary_id">primary_id</a></li>
      <li><a href="#target">target</a></li>
      <li><a href="#Internal-methods">Internal methods</a></li>
    </ul>
  </li>
  <li><a href="#BUGS">BUGS</a></li>
  <li><a href="#SEE-ALSO">SEE ALSO</a></li>
  <li><a href="#AUTHOR">AUTHOR</a></li>
</ul>

<h1 id="NAME">NAME</h1>

<p>Bio::DB::SeqFeature::NormalizedFeature -- Normalized feature for use with Bio::DB::SeqFeature::Store</p>

<h1 id="SYNOPSIS">SYNOPSIS</h1>

<pre><code><code> <span class="keyword">use</span> <span class="variable">Bio::DB::SeqFeature::Store</span><span class="operator">;</span>
 <span class="comment"># Open the sequence database</span>
 <span class="keyword">my</span> <span class="variable">$db</span>      <span class="operator">=</span> <span class="variable">Bio::DB::SeqFeature::Store</span><span class="operator">-&gt;</span><span class="variable">new</span><span class="operator">(</span> <span class="string">-adaptor</span> <span class="operator">=&gt;</span> <span class="string">'DBI::mysql'</span><span class="operator">,</span>
                                                <span class="string">-dsn</span>     <span class="operator">=&gt;</span> <span class="string">'dbi:mysql:test'</span><span class="operator">);</span>
 <span class="keyword">my</span> <span class="operator">(</span><span class="variable">$feature</span><span class="operator">)</span>   <span class="operator">=</span> <span class="variable">$db</span><span class="operator">-&gt;</span><span class="variable">get_features_by_name</span><span class="operator">(</span><span class="string">'ZK909'</span><span class="operator">);</span>
 <span class="keyword">my</span> <span class="variable">@subfeatures</span> <span class="operator">=</span> <span class="variable">$feature</span><span class="operator">-&gt;</span><span class="variable">get_SeqFeatures</span><span class="operator">();</span>
 <span class="keyword">my</span> <span class="variable">@exons_only</span>  <span class="operator">=</span> <span class="variable">$feature</span><span class="operator">-&gt;</span><span class="variable">get_SeqFeatures</span><span class="operator">(</span><span class="string">'exon'</span><span class="operator">);</span>
 
 <span class="comment"># create a new object</span>
 <span class="variable">$db</span><span class="operator">-&gt;</span><span class="variable">seqfeature_class</span><span class="operator">(</span><span class="string">'Bio::DB::SeqFeature::NormalizedFeature'</span><span class="operator">);</span>
 <span class="keyword">my</span> <span class="variable">$new</span> <span class="operator">=</span> <span class="variable">$db</span><span class="operator">-&gt;</span><span class="variable">new_feature</span><span class="operator">(</span><span class="string">-primary_tag</span><span class="operator">=&gt;</span><span class="string">'gene'</span><span class="operator">,</span>
                            <span class="string">-seq_id</span>     <span class="operator">=&gt;</span> <span class="string">'chr3'</span><span class="operator">,</span>
                            <span class="string">-start</span>      <span class="operator">=&gt;</span> <span class="number">10000</span><span class="operator">,</span>
                            <span class="string">-end</span>        <span class="operator">=&gt;</span> <span class="number">11000</span><span class="operator">);</span>
 
 <span class="comment"># add a new exon</span>
 <span class="variable">$feature</span><span class="operator">-&gt;</span><span class="variable">add_SeqFeature</span><span class="operator">(</span><span class="variable">$db</span><span class="operator">-&gt;</span><span class="variable">new_feature</span><span class="operator">(</span><span class="string">-primary_tag</span><span class="operator">=&gt;</span><span class="string">'exon'</span><span class="operator">,</span>
                                           <span class="string">-seq_id</span>     <span class="operator">=&gt;</span> <span class="string">'chr3'</span><span class="operator">,</span>
                                           <span class="string">-start</span>      <span class="operator">=&gt;</span> <span class="number">5000</span><span class="operator">,</span>
                                           <span class="string">-end</span>        <span class="operator">=&gt;</span> <span class="number">5551</span><span class="operator">));</span>
</code></code></pre>

<h1 id="DESCRIPTION">DESCRIPTION</h1>

<p>The Bio::DB::SeqFeature::NormalizedFeature object is an alternative representation of SeqFeatures for use with Bio::DB::SeqFeature::Store database system. It is identical to Bio::DB::SeqFeature, except that instead of storing feature/subfeature relationships in a database table, the information is stored in the object itself. This actually makes the objects somewhat inconvenient to work with from SQL, but does speed up access somewhat.</p>

<p>To use this class, pass the name of the class to the Bio::DB::SeqFeature::Store object&#39;s seqfeature_class() method. After this, $db-&gt;new_feature() will create objects of type Bio::DB::SeqFeature::NormalizedFeature. If you are using the GFF3 loader, pass Bio::DB::SeqFeature::Store::GFF3Loader-&gt;new() the -seqfeature_class argument:</p>

<pre><code><code>  <span class="keyword">use</span> <span class="variable">Bio::DB::SeqFeature::Store::GFF3Loader</span><span class="operator">;</span>
  
  <span class="keyword">my</span> <span class="variable">$store</span>  <span class="operator">=</span> <span class="variable">connect_to_db_somehow</span><span class="operator">();</span>
  <span class="keyword">my</span> <span class="variable">$loader</span> <span class="operator">=</span> <span class="variable">Bio::DB::SeqFeature::Store::GFF3Loader</span><span class="operator">-&gt;</span><span class="variable">new</span><span class="operator">(</span>
                <span class="string">-store</span><span class="operator">=&gt;</span><span class="variable">$db</span><span class="operator">,</span>
                <span class="string">-seqfeature_class</span> <span class="operator">=&gt;</span> <span class="string">'Bio::DB::SeqFeature::NormalizedFeature'</span>
               <span class="operator">);</span>
</code></code></pre>

<h2 id="new">new</h2>

<pre><code><code> Title   : new
 Usage   : $feature = Bio::DB::SeqFeature::NormalizedFeature-&gt;new(@args)
 Function: create a new feature
 Returns : the new seqfeature
 Args    : see below
 Status  : public</code></code></pre>

<p>This method creates and, if possible stores into a database, a new Bio::DB::SeqFeature::NormalizedFeature object using the specialized Bio::DB::SeqFeature class.</p>

<p>The arguments are the same to Bio::SeqFeature::Generic-&gt;new() and Bio::Graphics::Feature-&gt;new(). The most important difference is the <b>-store</b> option, which if present creates the object in a Bio::DB::SeqFeature::Store database, and he <b>-index</b> option, which controls whether the feature will be indexed for retrieval (default is true). Ordinarily, you would only want to turn indexing on when creating top level features, and off only when storing subfeatures. The default is on.</p>

<p>Arguments are as follows:</p>

<pre><code><code>  -seq_id       the reference sequence
  -start        the start position of the feature
  -end          the stop position of the feature
  -display_name the feature name (returned by seqname)
  -primary_tag  the feature type (returned by primary_tag)
  -source       the source tag
  -score        the feature score (for GFF compatibility)
  -desc         a description of the feature
  -segments     a list of subfeatures (see Bio::Graphics::Feature)
  -subtype      the type to use when creating subfeatures
  -strand       the strand of the feature (one of -1, 0 or +1)
  -phase        the phase of the feature (0..2)
  -url          a URL to link to when rendered with Bio::Graphics
  -attributes   a hashref of tag value attributes, in which the key is the tag
                  and the value is an array reference of values
  -store        a previously-opened Bio::DB::SeqFeature::Store object
  -index        index this feature if true</code></code></pre>

<p>Aliases:</p>

<pre><code><code>  -id           an alias for -display_name
  -seqname      an alias for -display_name
  -display_id   an alias for -display_name
  -name         an alias for -display_name
  -stop         an alias for end
  -type         an alias for primary_tag</code></code></pre>

<h2 id="Bio::SeqFeatureI-methods">Bio::SeqFeatureI methods</h2>

<p>The following Bio::SeqFeatureI methods are supported:</p>

<pre><code><code> seq_id(), start(), end(), strand(), get_SeqFeatures(),
 display_name(), primary_tag(), source_tag(), seq(),
 location(), primary_id(), overlaps(), contains(), equals(),
 intersection(), union(), has_tag(), remove_tag(),
 add_tag_value(), get_tag_values(), get_all_tags()</code></code></pre>

<p>Some methods that do not make sense in the context of a genome annotation database system, such as attach_seq(), are not supported.</p>

<p>Please see <a href="../../../../../../.cpan/build/BioPerl-1.6.901-ghKY8U/blib/lib/Bio/SeqFeatureI.html">Bio::SeqFeatureI</a> for more details.</p>

<h2 id="add_SeqFeature">add_SeqFeature</h2>

<pre><code><code> Title   : add_SeqFeature
 Usage   : $flag = $feature-&gt;add_SeqFeature(@features)
 Function: Add subfeatures to the feature
 Returns : true if successful
 Args    : list of Bio::SeqFeatureI objects
 Status  : public</code></code></pre>

<p>Add one or more subfeatures to the feature. For best results, subfeatures should be of the same class as the parent feature (i.e. don&#39;t try mixing Bio::DB::SeqFeature::NormalizedFeature with other feature types).</p>

<p>An alias for this method is add_segment().</p>

<h2 id="update">update</h2>

<pre><code><code> Title   : update
 Usage   : $flag = $feature-&gt;update()
 Function: Update feature in the database
 Returns : true if successful
 Args    : none
 Status  : public</code></code></pre>

<p>After changing any fields in the feature, call update() to write it to the database. This is not needed for add_SeqFeature() as update() is invoked automatically.</p>

<h2 id="get_SeqFeatures">get_SeqFeatures</h2>

<pre><code><code> Title   : get_SeqFeature
 Usage   : @subfeatures = $feature-&gt;get_SeqFeatures([@types])
 Function: return subfeatures of this feature
 Returns : list of subfeatures
 Args    : list of subfeature primary_tags (optional)
 Status  : public</code></code></pre>

<p>This method extends the Bio::SeqFeatureI get_SeqFeatures() slightly by allowing you to pass a list of primary_tags, in which case only subfeatures whose primary_tag is contained on the list will be returned. Without any types passed all subfeatures are returned.</p>

<h2 id="object_store">object_store</h2>

<pre><code><code> Title   : object_store
 Usage   : $store = $feature-&gt;object_store([$new_store])
 Function: get or set the database handle
 Returns : current database handle
 Args    : new database handle (optional)
 Status  : public</code></code></pre>

<p>This method will get or set the Bio::DB::SeqFeature::Store object that is associated with the feature. After changing the store, you should probably unset the feature&#39;s primary_id() and call update() to ensure that the object is written into the database as a new feature.</p>

<h2 id="overloaded_names">overloaded_names</h2>

<pre><code><code> Title   : overloaded_names
 Usage   : $overload = $feature-&gt;overloaded_names([$new_overload])
 Function: get or set overloading of object strings
 Returns : current flag
 Args    : new flag (optional)
 Status  : public</code></code></pre>

<p>For convenience, when objects of this class are stringified, they are represented in the form &quot;primary_tag(display_name)&quot;. To turn this feature off, call overloaded_names() with a false value. You can invoke this on an individual feature object or on the class:</p>

<pre><code><code>  <span class="variable">Bio::DB::SeqFeature::NormalizedFeature</span><span class="operator">-&gt;</span><span class="variable">overloaded_names</span><span class="operator">(</span><span class="number">0</span><span class="operator">);</span>
</code></code></pre>

<h2 id="segment">segment</h2>

<pre><code><code> Title   : segment
 Usage   : $segment = $feature-&gt;segment
 Function: return a Segment object corresponding to feature
 Returns : a Bio::DB::SeqFeature::Segment
 Args    : none
 Status  : public</code></code></pre>

<p>This turns the feature into a Bio::DB::SeqFeature::Segment object, which you can then use to query for overlapping features. See <a href="../../../../../../.cpan/build/BioPerl-1.6.901-ghKY8U/blib/lib/Bio/DB/SeqFeature/Segment.html">Bio::DB::SeqFeature::Segment</a>.</p>

<h2 id="AUTOLOADED-methods">AUTOLOADED methods</h2>

<pre><code><code> <span class="variable">@subfeatures</span> <span class="operator">=</span> <span class="variable">$feature</span><span class="operator">-&gt;</span><span class="variable">Exon</span><span class="operator">;</span>
</code></code></pre>

<p>If you use an unknown method that begins with a capital letter, then the feature autogenerates a call to get_SeqFeatures() using the lower-cased method name as the primary_tag. In other words $feature-&gt;Exon is equivalent to:</p>

<pre><code><code> @subfeature s= $feature-&gt;get_SeqFeatures(&#39;exon&#39;)</code></code></pre>

<p>If you use an unknown method that begins with Tag_(tagname), Att_(tagname) Is_(tagname), then it will be the same as calling the each_tag_value() method with the tagname. In a list context, these autogenerated procedures return the list of results. In scalar context, they return the first item in the list!!</p>

<h2 id="load_id">load_id</h2>

<pre><code><code> Title   : load_id
 Usage   : $id = $feature-&gt;load_id
 Function: get the GFF3 load ID
 Returns : the GFF3 load ID (string)
 Args    : none
 Status  : public</code></code></pre>

<p>For features that were originally loaded by the GFF3 loader, this method returns the GFF3 load ID. This method may not be supported in future versions of the module.</p>

<h2 id="notes">notes</h2>

<pre><code><code> Title   : notes
 Usage   : @notes = $feature-&gt;notes
 Function: get contents of the GFF3 Note tag
 Returns : List of GFF3 Note tags
 Args    : none
 Status  : public</code></code></pre>

<p>For features that were originally loaded by the GFF3 loader, this method returns the contents of the Note tag as a list. This is a convenience for Bio::Graphics, which looks for notes() when it constructs a default description line.</p>

<h2 id="primary_id">primary_id</h2>

<pre><code><code> Title   : primary_id
 Usage   : $id = $feature-&gt;primary_id([$new_id])
 Function: get/set the feature&#39;s database ID
 Returns : the current primary ID
 Args    : none
 Status  : public</code></code></pre>

<p>This method gets or sets the primary ID of the feature in the underlying Bio::DB::SeqFeature::Store database. If you change this field and then call update(), it will have the effect of making a copy of the feature in the database under a new ID.</p>

<h2 id="target">target</h2>

<pre><code><code> Title   : target
 Usage   : $segment = $feature-&gt;target
 Function: return the segment correspondent to the &quot;Target&quot; attribute
 Returns : a Bio::DB::SeqFeature::Segment object
 Args    : none
 Status  : public</code></code></pre>

<p>For features that are aligned with others via the GFF3 Target attribute, this returns a segment corresponding to the aligned region. The CIGAR gap string is not yet supported.</p>

<h2 id="Internal-methods">Internal methods</h2>

<dl>

<dt id="feature-as_string-">$feature-&gt;as_string()</dt>
<dd>

<p>Internal method used to implement overloaded stringification.</p>

</dd>
<dt id="boolean-feature-type_match-list_of_types-">$boolean = $feature-&gt;type_match(@list_of_types)</dt>
<dd>

<p>Internal method that will return true if the feature&#39;s primary_tag and source_tag match any of the list of types (in primary_tag:source_tag format) provided.</p>

</dd>
</dl>

<h1 id="BUGS">BUGS</h1>

<p>This is an early version, so there are certainly some bugs. Please use the BioPerl bug tracking system to report bugs.</p>

<h1 id="SEE-ALSO">SEE ALSO</h1>

<p><a>bioperl</a>, <a href="../../../../../../.cpan/build/BioPerl-1.6.901-ghKY8U/blib/lib/Bio/DB/SeqFeature.html">Bio::DB::SeqFeature</a>, <a href="../../../../../../.cpan/build/BioPerl-1.6.901-ghKY8U/blib/lib/Bio/DB/SeqFeature/Store.html">Bio::DB::SeqFeature::Store</a>, <a href="../../../../../../.cpan/build/BioPerl-1.6.901-ghKY8U/blib/lib/Bio/DB/SeqFeature/Segment.html">Bio::DB::SeqFeature::Segment</a>, <a>Bio::DB::SeqFeature::GFF3Loader</a>, <a href="../../../../../../.cpan/build/BioPerl-1.6.901-ghKY8U/blib/lib/Bio/DB/SeqFeature/Store/DBI/mysql.html">Bio::DB::SeqFeature::Store::DBI::mysql</a>, <a href="../../../../../../.cpan/build/BioPerl-1.6.901-ghKY8U/blib/lib/Bio/DB/SeqFeature/Store/bdb.html">Bio::DB::SeqFeature::Store::bdb</a></p>

<h1 id="AUTHOR">AUTHOR</h1>

<p>Lincoln Stein &lt;lstein@cshl.org&gt;.</p>

<p>Copyright (c) 2006 Cold Spring Harbor Laboratory.</p>

<p>This library is free software; you can redistribute it and/or modify it under the same terms as Perl itself.</p>


</body>

</html>


