<?xml version="1.0" ?>
<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.0 Transitional//EN">
<html>
<head>
<!-- saved from url=(0017)http://localhost/ -->
<script language="JavaScript" src="../../../../../displayToc.js"></script>
<script language="JavaScript" src="../../../../../tocParas.js"></script>
<script language="JavaScript" src="../../../../../tocTab.js"></script>
<link rel="stylesheet" type="text/css" href="../../../../../scineplex.css">
<title></title>
<link rel="stylesheet" href="../../../../../Active.css" type="text/css" />
<meta http-equiv="content-type" content="text/html; charset=utf-8" />
<link rev="made" href="mailto:support@ActiveState.com" />
</head>

<body>



<ul id="index">
  <li><a href="#NAME">NAME</a></li>
  <li><a href="#SYNOPSIS">SYNOPSIS</a></li>
  <li><a href="#DESCRIPTION">DESCRIPTION</a>
    <ul>
      <li><a href="#Important-note-about-start-vs-end-">Important note about start() vs end()</a></li>
    </ul>
  </li>
  <li><a href="#API">API</a>
    <ul>
      <li><a href="#new_from_parent">new_from_parent</a></li>
      <li><a href="#new">new</a></li>
      <li><a href="#type">type</a></li>
      <li><a href="#method">method</a></li>
      <li><a href="#source">source</a></li>
      <li><a href="#score">score</a></li>
      <li><a href="#phase">phase</a></li>
      <li><a href="#strand">strand</a></li>
      <li><a href="#group">group</a></li>
      <li><a href="#display_id">display_id</a></li>
      <li><a href="#info">info</a></li>
      <li><a href="#target">target</a></li>
      <li><a href="#flatten_target">flatten_target</a></li>
      <li><a href="#hit">hit</a></li>
      <li><a href="#id">id</a></li>
      <li><a href="#group_id">group_id</a></li>
      <li><a href="#clone">clone</a></li>
      <li><a href="#compound">compound</a></li>
      <li><a href="#sub_SeqFeature">sub_SeqFeature</a></li>
      <li><a href="#add_subfeature">add_subfeature</a></li>
      <li><a href="#attach_seq">attach_seq</a></li>
      <li><a href="#location">location</a></li>
      <li><a href="#entire_seq">entire_seq</a></li>
      <li><a href="#merged_segments">merged_segments</a></li>
      <li><a href="#sub_types">sub_types</a></li>
      <li><a href="#attributes">attributes</a></li>
      <li><a href="#notes">notes</a></li>
      <li><a href="#aliases">aliases</a></li>
      <li><a href="#Autogenerated-Methods">Autogenerated Methods</a></li>
      <li><a href="#SeqFeatureI-methods">SeqFeatureI methods</a></li>
      <li><a href="#adjust_bounds">adjust_bounds</a></li>
      <li><a href="#sort_features">sort_features</a></li>
      <li><a href="#asString">asString</a></li>
      <li><a href="#gff_string">gff_string</a></li>
      <li><a href="#gff3_string">gff3_string</a></li>
      <li><a href="#version">version</a></li>
      <li><a href="#cmap_link-">cmap_link()</a></li>
    </ul>
  </li>
  <li><a href="#A-Note-About-Similarities">A Note About Similarities</a></li>
  <li><a href="#BUGS">BUGS</a></li>
  <li><a href="#SEE-ALSO">SEE ALSO</a></li>
  <li><a href="#AUTHOR">AUTHOR</a></li>
</ul>

<h1 id="NAME">NAME</h1>

<p>Bio::DB::GFF::Feature -- A relative segment identified by a feature type</p>

<h1 id="SYNOPSIS">SYNOPSIS</h1>

<p>See <a href="../../../../../../.cpan/build/BioPerl-1.6.901-ghKY8U/blib/lib/Bio/DB/GFF.html">Bio::DB::GFF</a>.</p>

<h1 id="DESCRIPTION">DESCRIPTION</h1>

<p>Bio::DB::GFF::Feature is a stretch of sequence that corresponding to a single annotation in a GFF database. It inherits from Bio::DB::GFF::RelSegment, and so has all the support for relative addressing of this class and its ancestors. It also inherits from Bio::SeqFeatureI and so has the familiar start(), stop(), primary_tag() and location() methods (it implements Bio::LocationI too, if needed).</p>

<p>Bio::DB::GFF::Feature adds new methods to retrieve the annotation type, group, and other GFF attributes. Annotation types are represented by Bio::DB::GFF::Typename objects, a simple class that has two methods called method() and source(). These correspond to the method and source fields of a GFF file.</p>

<p>Annotation groups serve the dual purpose of giving the annotation a human-readable name, and providing higher-order groupings of subfeatures into features. The groups returned by this module are objects of the Bio::DB::GFF::Featname class.</p>

<p>Bio::DB::GFF::Feature inherits from and implements the abstract methods of Bio::SeqFeatureI, allowing it to interoperate with other Bioperl modules.</p>

<p>Generally, you will not create or manipulate Bio::DB::GFF::Feature objects directly, but use those that are returned by the Bio::DB::GFF::RelSegment-&gt;features() method.</p>

<h2 id="Important-note-about-start-vs-end-">Important note about start() vs end()</h2>

<p>If features are derived from segments that use relative addressing (which is the default), then start() will be less than end() if the feature is on the opposite strand from the reference sequence. This breaks Bio::SeqI compliance, but is necessary to avoid having the real genomic locations designated by start() and end() swap places when changing reference points.</p>

<p>To avoid this behavior, call $segment-&gt;absolute(1) before fetching features from it. This will force everything into absolute coordinates.</p>

<p>For example:</p>

<pre><code><code> <span class="keyword">my</span> <span class="variable">$segment</span> <span class="operator">=</span> <span class="variable">$db</span><span class="operator">-&gt;</span><span class="variable">segment</span><span class="operator">(</span><span class="string">'CHROMOSOME_I'</span><span class="operator">);</span>
 <span class="variable">$segment</span><span class="operator">-&gt;</span><span class="variable">absolute</span><span class="operator">(</span><span class="number">1</span><span class="operator">);</span>
 <span class="keyword">my</span> <span class="variable">@features</span> <span class="operator">=</span> <span class="variable">$segment</span><span class="operator">-&gt;</span><span class="variable">features</span><span class="operator">(</span><span class="string">'transcript'</span><span class="operator">);</span>
</code></code></pre>

<h1 id="API">API</h1>

<p>The remainder of this document describes the public and private methods implemented by this module.</p>

<h2 id="new_from_parent">new_from_parent</h2>

<pre><code><code> <span class="variable">Title</span>   <span class="operator">:</span> <span class="variable">new_from_parent</span>
 <span class="variable">Usage</span>   <span class="operator">:</span> <span class="variable">$f</span> <span class="operator">=</span> <span class="variable">Bio::DB::GFF::Feature</span><span class="operator">-&gt;</span><span class="variable">new_from_parent</span><span class="operator">(</span><span class="variable">@args</span><span class="operator">);</span>
 <span class="variable">Function</span><span class="operator">:</span> <span class="variable">create</span> <span class="variable">a</span> <span class="variable">new</span> <span class="variable">feature</span> <span class="variable">object</span>
 <span class="variable">Returns</span> <span class="operator">:</span> <span class="variable">new</span> <span class="variable">Bio::DB::GFF::Feature</span> <span class="variable">object</span>
 <span class="variable">Args</span>    <span class="operator">:</span> <span class="variable">see</span> <span class="variable">below</span>
 <span class="variable">Status</span>  <span class="operator">:</span> <span class="variable">Internal</span>
</code></code></pre>

<p>This method is called by Bio::DB::GFF to create a new feature using information obtained from the GFF database. It is one of two similar constructors. This one is called when the feature is generated from a RelSegment object, and should inherit the coordinate system of that object.</p>

<p>The 13 arguments are positional (sorry):</p>

<pre><code><code>  $parent       a Bio::DB::GFF::RelSegment object (or descendent)
  $start        start of this feature
  $stop         stop of this feature
  $method       this feature&#39;s GFF method
  $source       this feature&#39;s GFF source
  $score               this feature&#39;s score
  $fstrand      this feature&#39;s strand (relative to the source
                      sequence, which has its own strandedness!)
  $phase        this feature&#39;s phase
  $group        this feature&#39;s group (a Bio::DB::GFF::Featname object)
  $db_id        this feature&#39;s internal database ID
  $group_id     this feature&#39;s internal group database ID
  $tstart       this feature&#39;s target start
  $tstop        this feature&#39;s target stop</code></code></pre>

<p>tstart and tstop are not used for anything at the moment, since the information is embedded in the group object.</p>

<h2 id="new">new</h2>

<pre><code><code> <span class="variable">Title</span>   <span class="operator">:</span> <span class="variable">new</span>
 <span class="variable">Usage</span>   <span class="operator">:</span> <span class="variable">$f</span> <span class="operator">=</span> <span class="variable">Bio::DB::GFF::Feature</span><span class="operator">-&gt;</span><span class="variable">new</span><span class="operator">(</span><span class="variable">@args</span><span class="operator">);</span>
 <span class="variable">Function</span><span class="operator">:</span> <span class="variable">create</span> <span class="variable">a</span> <span class="variable">new</span> <span class="variable">feature</span> <span class="variable">object</span>
 <span class="variable">Returns</span> <span class="operator">:</span> <span class="variable">new</span> <span class="variable">Bio::DB::GFF::Feature</span> <span class="variable">object</span>
 <span class="variable">Args</span>    <span class="operator">:</span> <span class="variable">see</span> <span class="variable">below</span>
 <span class="variable">Status</span>  <span class="operator">:</span> <span class="variable">Internal</span>
</code></code></pre>

<p>This method is called by Bio::DB::GFF to create a new feature using information obtained from the GFF database. It is one of two similar constructors. This one is called when the feature is generated without reference to a RelSegment object, and should therefore use its default coordinate system (relative to itself).</p>

<p>The 11 arguments are positional:</p>

<pre><code><code>  $factory      a Bio::DB::GFF adaptor object (or descendent)
  $srcseq       the source sequence
  $start        start of this feature
  $stop         stop of this feature
  $method       this feature&#39;s GFF method
  $source       this feature&#39;s GFF source
  $score               this feature&#39;s score
  $fstrand      this feature&#39;s strand (relative to the source
                      sequence, which has its own strandedness!)
  $phase        this feature&#39;s phase
  $group        this feature&#39;s group
  $db_id        this feature&#39;s internal database ID</code></code></pre>

<h2 id="type">type</h2>

<pre><code><code> Title   : type
 Usage   : $type = $f-&gt;type([$newtype])
 Function: get or set the feature type
 Returns : a Bio::DB::GFF::Typename object
 Args    : a new Typename object (optional)
 Status  : Public</code></code></pre>

<p>This method gets or sets the type of the feature. The type is a Bio::DB::GFF::Typename object, which encapsulates the feature method and source.</p>

<p>The method() and source() methods described next provide shortcuts to the individual fields of the type.</p>

<h2 id="method">method</h2>

<pre><code><code> Title   : method
 Usage   : $method = $f-&gt;method([$newmethod])
 Function: get or set the feature method
 Returns : a string
 Args    : a new method (optional)
 Status  : Public</code></code></pre>

<p>This method gets or sets the feature method. It is a convenience feature that delegates the task to the feature&#39;s type object.</p>

<h2 id="source">source</h2>

<pre><code><code> Title   : source
 Usage   : $source = $f-&gt;source([$newsource])
 Function: get or set the feature source
 Returns : a string
 Args    : a new source (optional)
 Status  : Public</code></code></pre>

<p>This method gets or sets the feature source. It is a convenience feature that delegates the task to the feature&#39;s type object.</p>

<h2 id="score">score</h2>

<pre><code><code> Title   : score
 Usage   : $score = $f-&gt;score([$newscore])
 Function: get or set the feature score
 Returns : a string
 Args    : a new score (optional)
 Status  : Public</code></code></pre>

<p>This method gets or sets the feature score.</p>

<h2 id="phase">phase</h2>

<pre><code><code> Title   : phase
 Usage   : $phase = $f-&gt;phase([$phase])
 Function: get or set the feature phase
 Returns : a string
 Args    : a new phase (optional)
 Status  : Public</code></code></pre>

<p>This method gets or sets the feature phase.</p>

<h2 id="strand">strand</h2>

<pre><code><code> Title   : strand
 Usage   : $strand = $f-&gt;strand
 Function: get the feature strand
 Returns : +1, 0 -1
 Args    : none
 Status  : Public</code></code></pre>

<p>Returns the strand of the feature. Unlike the other methods, the strand cannot be changed once the object is created (due to coordinate considerations).</p>

<h2 id="group">group</h2>

<pre><code><code> Title   : group
 Usage   : $group = $f-&gt;group([$new_group])
 Function: get or set the feature group
 Returns : a Bio::DB::GFF::Featname object
 Args    : a new group (optional)
 Status  : Public</code></code></pre>

<p>This method gets or sets the feature group. The group is a Bio::DB::GFF::Featname object, which has an ID and a class.</p>

<h2 id="display_id">display_id</h2>

<pre><code><code> Title   : display_id
 Usage   : $display_id = $f-&gt;display_id([$display_id])
 Function: get or set the feature display id
 Returns : a Bio::DB::GFF::Featname object
 Args    : a new display_id (optional)
 Status  : Public</code></code></pre>

<p>This method is an alias for group(). It is provided for Bio::SeqFeatureI compatibility.</p>

<h2 id="info">info</h2>

<pre><code><code> Title   : info
 Usage   : $info = $f-&gt;info([$new_info])
 Function: get or set the feature group
 Returns : a Bio::DB::GFF::Featname object
 Args    : a new group (optional)
 Status  : Public</code></code></pre>

<p>This method is an alias for group(). It is provided for AcePerl compatibility.</p>

<h2 id="target">target</h2>

<pre><code><code> Title   : target
 Usage   : $target = $f-&gt;target([$new_target])
 Function: get or set the feature target
 Returns : a Bio::DB::GFF::Homol object
 Args    : a new group (optional)
 Status  : Public</code></code></pre>

<p>This method works like group(), but only returns the group if it implements the start() method. This is typical for similarity/assembly features, where the target encodes the start and stop location of the alignment.</p>

<p>The returned object is of type Bio::DB::GFF::Homol, which is a subclass of Bio::DB::GFF::Segment.</p>

<h2 id="flatten_target">flatten_target</h2>

<pre><code><code> Title   : flatten_target
 Usage   : $target = $f-&gt;flatten_target($f-&gt;target)
 Function: flatten a target object
 Returns : a string (GFF2), an array [GFF2.5] or an array ref [GFF3]
 Args    : a target object (required), GFF version (optional) 
 Status  : Public</code></code></pre>

<p>This method flattens a target object into text for GFF dumping. If a second argument is provided, version-specific vocabulary is used for the flattened target.</p>

<h2 id="hit">hit</h2>

<pre><code><code> Title   : hit
 Usage   : $hit = $f-&gt;hit([$new_hit])
 Function: get or set the feature hit
 Returns : a Bio::DB::GFF::Featname object
 Args    : a new group (optional)
 Status  : Public</code></code></pre>

<p>This is the same as target(), for compatibility with Bio::SeqFeature::SimilarityPair.</p>

<h2 id="id">id</h2>

<pre><code><code> Title   : id
 Usage   : $id = $f-&gt;id
 Function: get the feature ID
 Returns : a database identifier
 Args    : none
 Status  : Public</code></code></pre>

<p>This method retrieves the database identifier for the feature. It cannot be changed.</p>

<h2 id="group_id">group_id</h2>

<pre><code><code> Title   : group_id
 Usage   : $id = $f-&gt;group_id
 Function: get the feature ID
 Returns : a database identifier
 Args    : none
 Status  : Public</code></code></pre>

<p>This method retrieves the database group identifier for the feature. It cannot be changed. Often the group identifier is more useful than the feature identifier, since it is used to refer to a complex object containing subparts.</p>

<h2 id="clone">clone</h2>

<pre><code><code> Title   : clone
 Usage   : $feature = $f-&gt;clone
 Function: make a copy of the feature
 Returns : a new Bio::DB::GFF::Feature object
 Args    : none
 Status  : Public</code></code></pre>

<p>This method returns a copy of the feature.</p>

<h2 id="compound">compound</h2>

<pre><code><code> Title   : compound
 Usage   : $flag = $f-&gt;compound([$newflag])
 Function: get or set the compound flag
 Returns : a boolean
 Args    : a new flag (optional)
 Status  : Public</code></code></pre>

<p>This method gets or sets a flag indicated that the feature is not a primary one from the database, but the result of aggregation.</p>

<h2 id="sub_SeqFeature">sub_SeqFeature</h2>

<pre><code><code> Title   : sub_SeqFeature
 Usage   : @feat = $feature-&gt;sub_SeqFeature([$method])
 Function: get subfeatures
 Returns : a list of Bio::DB::GFF::Feature objects
 Args    : a feature method (optional)
 Status  : Public</code></code></pre>

<p>This method returns a list of any subfeatures that belong to the main feature. For those features that contain heterogeneous subfeatures, you can retrieve a subset of the subfeatures by providing a method name to filter on.</p>

<p>This method may also be called as segments() or get_SeqFeatures().</p>

<h2 id="add_subfeature">add_subfeature</h2>

<pre><code><code> Title   : add_subfeature
 Usage   : $feature-&gt;add_subfeature($feature)
 Function: add a subfeature to the feature
 Returns : nothing
 Args    : a Bio::DB::GFF::Feature object
 Status  : Public</code></code></pre>

<p>This method adds a new subfeature to the object. It is used internally by aggregators, but is available for public use as well.</p>

<h2 id="attach_seq">attach_seq</h2>

<pre><code><code> Title   : attach_seq
 Usage   : $sf-&gt;attach_seq($seq)
 Function: Attaches a Bio::Seq object to this feature. This
           Bio::Seq object is for the *entire* sequence: ie
           from 1 to 10000
 Example :
 Returns : TRUE on success
 Args    : a Bio::PrimarySeqI compliant object</code></code></pre>

<h2 id="location">location</h2>

<pre><code><code> Title   : location
 Usage   : my $location = $seqfeature-&gt;location()
 Function: returns a location object suitable for identifying location 
           of feature on sequence or parent feature  
 Returns : Bio::LocationI object
 Args    : none</code></code></pre>

<h2 id="entire_seq">entire_seq</h2>

<pre><code><code> Title   : entire_seq
 Usage   : $whole_seq = $sf-&gt;entire_seq()
 Function: gives the entire sequence that this seqfeature is attached to
 Example :
 Returns : a Bio::PrimarySeqI compliant object, or undef if there is no
           sequence attached
 Args    : none</code></code></pre>

<h2 id="merged_segments">merged_segments</h2>

<pre><code><code> Title   : merged_segments
 Usage   : @segs = $feature-&gt;merged_segments([$method])
 Function: get merged subfeatures
 Returns : a list of Bio::DB::GFF::Feature objects
 Args    : a feature method (optional)
 Status  : Public</code></code></pre>

<p>This method acts like sub_SeqFeature, except that it merges overlapping segments of the same time into contiguous features. For those features that contain heterogeneous subfeatures, you can retrieve a subset of the subfeatures by providing a method name to filter on.</p>

<p>A side-effect of this method is that the features are returned in sorted order by their start tposition.</p>

<h2 id="sub_types">sub_types</h2>

<pre><code><code> Title   : sub_types
 Usage   : @methods = $feature-&gt;sub_types
 Function: get methods of all sub-seqfeatures
 Returns : a list of method names
 Args    : none
 Status  : Public</code></code></pre>

<p>For those features that contain subfeatures, this method will return a unique list of method names of those subfeatures, suitable for use with sub_SeqFeature().</p>

<h2 id="attributes">attributes</h2>

<pre><code><code> Title   : attributes
 Usage   : @attributes = $feature-&gt;attributes($name)
 Function: get the &quot;attributes&quot; on a particular feature
 Returns : an array of string
 Args    : feature ID
 Status  : public</code></code></pre>

<p>Some GFF version 2 files use the groups column to store a series of attribute/value pairs. In this interpretation of GFF, the first such pair is treated as the primary group for the feature; subsequent pairs are treated as attributes. Two attributes have special meaning: &quot;Note&quot; is for backward compatibility and is used for unstructured text remarks. &quot;Alias&quot; is considered as a synonym for the feature name.</p>

<pre><code><code> <span class="variable">@gene_names</span> <span class="operator">=</span> <span class="variable">$feature</span><span class="operator">-&gt;</span><span class="variable">attributes</span><span class="operator">(</span><span class="string">'Gene'</span><span class="operator">);</span>
 <span class="variable">@aliases</span>    <span class="operator">=</span> <span class="variable">$feature</span><span class="operator">-&gt;</span><span class="variable">attributes</span><span class="operator">(</span><span class="string">'Alias'</span><span class="operator">);</span>
</code></code></pre>

<p>If no name is provided, then attributes() returns a flattened hash, of attribute=&gt;value pairs. This lets you do:</p>

<pre><code><code>  <span class="variable">%attributes</span> <span class="operator">=</span> <span class="variable">$db</span><span class="operator">-&gt;</span><span class="variable">attributes</span><span class="operator">;</span>
</code></code></pre>

<h2 id="notes">notes</h2>

<pre><code><code> Title   : notes
 Usage   : @notes = $feature-&gt;notes
 Function: get the &quot;notes&quot; on a particular feature
 Returns : an array of string
 Args    : feature ID
 Status  : public</code></code></pre>

<p>Some GFF version 2 files use the groups column to store various notes and remarks. Adaptors can elect to store the notes in the database, or just ignore them. For those adaptors that store the notes, the notes() method will return them as a list.</p>

<h2 id="aliases">aliases</h2>

<pre><code><code> Title   : aliases
 Usage   : @aliases = $feature-&gt;aliases
 Function: get the &quot;aliases&quot; on a particular feature
 Returns : an array of string
 Args    : feature ID
 Status  : public</code></code></pre>

<p>This method will return a list of attributes of type &#39;Alias&#39;.</p>

<h2 id="Autogenerated-Methods">Autogenerated Methods</h2>

<pre><code><code> Title   : AUTOLOAD
 Usage   : @subfeat = $feature-&gt;Method
 Function: Return subfeatures using autogenerated methods
 Returns : a list of Bio::DB::GFF::Feature objects
 Args    : none
 Status  : Public</code></code></pre>

<p>Any method that begins with an initial capital letter will be passed to AUTOLOAD and treated as a call to sub_SeqFeature with the method name used as the method argument. For instance, this call:</p>

<pre><code><code>  <span class="variable">@exons</span> <span class="operator">=</span> <span class="variable">$feature</span><span class="operator">-&gt;</span><span class="variable">Exon</span><span class="operator">;</span>
</code></code></pre>

<p>is equivalent to this call:</p>

<pre><code><code>  <span class="variable">@exons</span> <span class="operator">=</span> <span class="variable">$feature</span><span class="operator">-&gt;</span><span class="variable">sub_SeqFeature</span><span class="operator">(</span><span class="string">'exon'</span><span class="operator">);</span>
</code></code></pre>

<h2 id="SeqFeatureI-methods">SeqFeatureI methods</h2>

<p>The following Bio::SeqFeatureI methods are implemented:</p>

<p>primary_tag(), source_tag(), all_tags(), has_tag(), each_tag_value() [renamed get_tag_values()].</p>

<h2 id="adjust_bounds">adjust_bounds</h2>

<pre><code><code> Title   : adjust_bounds
 Usage   : $feature-&gt;adjust_bounds
 Function: adjust the bounds of a feature
 Returns : ($start,$stop,$strand)
 Args    : none
 Status  : Public</code></code></pre>

<p>This method adjusts the boundaries of the feature to enclose all its subfeatures. It returns the new start, stop and strand of the enclosing feature.</p>

<h2 id="sort_features">sort_features</h2>

<pre><code><code> Title   : sort_features
 Usage   : $feature-&gt;sort_features
 Function: sort features
 Returns : nothing
 Args    : none
 Status  : Public</code></code></pre>

<p>This method sorts subfeatures in ascending order by their start position. For reverse strand features, it sorts subfeatures in descending order. After this is called sub_SeqFeature will return the features in order.</p>

<p>This method is called internally by merged_segments().</p>

<h2 id="asString">asString</h2>

<pre><code><code> Title   : asString
 Usage   : $string = $feature-&gt;asString
 Function: return human-readabled representation of feature
 Returns : a string
 Args    : none
 Status  : Public</code></code></pre>

<p>This method returns a human-readable representation of the feature and is called by the overloaded &quot;&quot; operator.</p>

<h2 id="gff_string">gff_string</h2>

<pre><code><code> Title   : gff_string
 Usage   : $string = $feature-&gt;gff_string
 Function: return GFF2 of GFF2.5 representation of feature
 Returns : a string
 Args    : none
 Status  : Public</code></code></pre>

<h2 id="gff3_string">gff3_string</h2>

<pre><code><code> Title   : gff3_string
 Usage   : $string = $feature-&gt;gff3_string([$recurse])
 Function: return GFF3 representation of feature
 Returns : a string
 Args    : An optional flag, which if true, will cause the feature to recurse over
           subfeatures.
 Status  : Public</code></code></pre>

<h2 id="version">version</h2>

<pre><code><code> Title   : version
 Usage   : $feature-&gt;version()
 Function: get/set the GFF version to be returned by gff_string
 Returns : the GFF version (default is 2)
 Args    : the GFF version (2, 2.5 of 3)
 Status  : Public</code></code></pre>

<h2 id="cmap_link-">cmap_link()</h2>

<pre><code><code> Title   : cmap_link
 Usage   : $link = $feature-&gt;cmap_link
 Function: returns a URL link to the corresponding feature in cmap
 Returns : a string
 Args    : none
 Status  : Public</code></code></pre>

<p>If integrated cmap/gbrowse installation, it returns a link to the map otherwise it returns a link to a feature search on the feature name. See the cmap documentation for more information.</p>

<p>This function is intended primarily to be used in gbrowse conf files. For example:</p>

<pre><code><code>  <span class="keyword">link</span>       <span class="operator">=</span> <span class="keyword">sub</span><span class="variable"> </span><span class="operator">{</span><span class="keyword">my</span> <span class="variable">$self</span> <span class="operator">=</span> <span class="keyword">shift</span><span class="operator">;</span> <span class="keyword">return</span> <span class="variable">$self</span><span class="operator">-&gt;</span><span class="variable">cmap_viewer_link</span><span class="operator">(</span><span class="variable">data_source</span><span class="operator">);}</span>
</code></code></pre>

<h1 id="A-Note-About-Similarities">A Note About Similarities</h1>

<p>The current default aggregator for GFF &quot;similarity&quot; features creates a composite Bio::DB::GFF::Feature object of type &quot;gapped_alignment&quot;. The target() method for the feature as a whole will return a RelSegment object that is as long as the extremes of the similarity hit target, but will not necessarily be the same length as the query sequence. The length of each &quot;similarity&quot; subfeature will be exactly the same length as its target(). These subfeatures are essentially the HSPs of the match.</p>

<p>The following illustrates this:</p>

<pre><code><code>  <span class="variable">@similarities</span> <span class="operator">=</span> <span class="variable">$segment</span><span class="operator">-&gt;</span><span class="variable">feature</span><span class="operator">(</span><span class="string">'similarity:BLASTN'</span><span class="operator">);</span>
  <span class="variable">$sim</span>          <span class="operator">=</span> <span class="variable">$similarities</span><span class="operator">[</span><span class="number">0</span><span class="operator">]</span><span class="operator">;</span>
  
  <span class="keyword">print</span> <span class="variable">$sim</span><span class="operator">-&gt;</span><span class="variable">type</span><span class="operator">;</span>        <span class="comment"># yields "gapped_similarity:BLASTN"</span>
  
  <span class="variable">$query_length</span>  <span class="operator">=</span> <span class="variable">$sim</span><span class="operator">-&gt;</span><span class="variable">length</span><span class="operator">;</span>
  <span class="variable">$target_length</span> <span class="operator">=</span> <span class="variable">$sim</span><span class="operator">-&gt;</span><span class="variable">target</span><span class="operator">-&gt;</span><span class="variable">length</span><span class="operator">;</span>  <span class="comment"># $query_length != $target_length</span>
  
  <span class="variable">@matches</span> <span class="operator">=</span> <span class="variable">$sim</span><span class="operator">-&gt;</span><span class="variable">Similarity</span><span class="operator">;</span>   <span class="comment"># use autogenerated method</span>
  <span class="variable">$query1_length</span>  <span class="operator">=</span> <span class="variable">$matches</span><span class="operator">[</span><span class="number">0</span><span class="operator">]</span><span class="operator">-&gt;</span><span class="variable">length</span><span class="operator">;</span>
  <span class="variable">$target1_length</span> <span class="operator">=</span> <span class="variable">$matches</span><span class="operator">[</span><span class="number">0</span><span class="operator">]</span><span class="operator">-&gt;</span><span class="variable">target</span><span class="operator">-&gt;</span><span class="variable">length</span><span class="operator">;</span> <span class="comment"># $query1_length == $target1_length</span>
</code></code></pre>

<p>If you merge segments by calling merged_segments(), then the length of the query sequence segments will no longer necessarily equal the length of the targets, because the alignment information will have been lost. Nevertheless, the targets are adjusted so that the first and last base pairs of the query match the first and last base pairs of the target.</p>

<h1 id="BUGS">BUGS</h1>

<p>This module is still under development.</p>

<h1 id="SEE-ALSO">SEE ALSO</h1>

<p><a>bioperl</a>, <a href="../../../../../../.cpan/build/BioPerl-1.6.901-ghKY8U/blib/lib/Bio/DB/GFF.html">Bio::DB::GFF</a>, <a>Bio::DB::RelSegment</a></p>

<h1 id="AUTHOR">AUTHOR</h1>

<p>Lincoln Stein &lt;lstein@cshl.org&gt;.</p>

<p>Copyright (c) 2001 Cold Spring Harbor Laboratory.</p>

<p>This library is free software; you can redistribute it and/or modify it under the same terms as Perl itself.</p>


</body>

</html>


