<?xml version="1.0" ?>
<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.0 Transitional//EN">
<html>
<head>
<!-- saved from url=(0017)http://localhost/ -->
<script language="JavaScript" src="../../../../../displayToc.js"></script>
<script language="JavaScript" src="../../../../../tocParas.js"></script>
<script language="JavaScript" src="../../../../../tocTab.js"></script>
<link rel="stylesheet" type="text/css" href="../../../../../scineplex.css">
<title></title>
<link rel="stylesheet" href="../../../../../Active.css" type="text/css" />
<meta http-equiv="content-type" content="text/html; charset=utf-8" />
<link rev="made" href="mailto:support@ActiveState.com" />
</head>

<body>



<ul id="index">
  <li><a href="#NAME">NAME</a></li>
  <li><a href="#SYNOPSIS">SYNOPSIS</a></li>
  <li><a href="#DEPRECATION-WARNING">DEPRECATION WARNING</a></li>
  <li><a href="#DESCRIPTION">DESCRIPTION</a></li>
  <li><a href="#OPTIONS">OPTIONS</a></li>
  <li><a href="#FILES">FILES</a></li>
  <li><a href="#FEEDBACK">FEEDBACK</a>
    <ul>
      <li><a href="#Mailing-Lists">Mailing Lists</a></li>
      <li><a href="#Support">Support</a></li>
      <li><a href="#Reporting-Bugs">Reporting Bugs</a></li>
    </ul>
  </li>
  <li><a href="#AUTHOR---Dan-Kortschak">AUTHOR - Dan Kortschak</a></li>
  <li><a href="#CONTRIBUTORS">CONTRIBUTORS</a></li>
  <li><a href="#APPENDIX">APPENDIX</a>
    <ul>
      <li><a href="#new-">new()</a></li>
      <li><a href="#run-">run()</a></li>
      <li><a href="#want-">want()</a></li>
      <li><a href="#result-">result()</a></li>
      <li><a href="#determine_format-">_determine_format()</a></li>
      <li><a href="#read_bed-">_read_bed()</a></li>
      <li><a href="#read_bedpe-">_read_bedpe()</a></li>
      <li><a href="#validate_file_input-">_validate_file_input()</a></li>
      <li><a href="#version-">version()</a></li>
    </ul>
  </li>
</ul>

<h1 id="NAME">NAME</h1>

<p>Bio::Tools::Run::BEDTools - Run wrapper for the BEDTools suite of programs *BETA*</p>

<h1 id="SYNOPSIS">SYNOPSIS</h1>

<pre><code><code> <span class="comment"># use a BEDTools program</span>
 <span class="variable">$bedtools_fac</span> <span class="operator">=</span> <span class="variable">Bio::Tools::Run::BEDTools</span><span class="operator">-&gt;</span><span class="variable">new</span><span class="operator">(</span> <span class="string">-command</span> <span class="operator">=&gt;</span> <span class="string">'subtract'</span> <span class="operator">);</span>
 <span class="variable">$result_file</span> <span class="operator">=</span> <span class="variable">$bedtools_fac</span><span class="operator">-&gt;</span><span class="variable">run</span><span class="operator">(</span> <span class="string">-bed1</span> <span class="operator">=&gt;</span> <span class="string">'genes.bed'</span><span class="operator">,</span> <span class="string">-bed2</span> <span class="operator">=&gt;</span> <span class="string">'mask.bed'</span> <span class="operator">);</span>
 
 <span class="comment"># if IO::Uncompress::Gunzip is available...</span>
 <span class="variable">$result_file</span> <span class="operator">=</span> <span class="variable">$bedtools_fac</span><span class="operator">-&gt;</span><span class="variable">run</span><span class="operator">(</span> <span class="string">-bed1</span> <span class="operator">=&gt;</span> <span class="string">'genes.bed.gz'</span><span class="operator">,</span> <span class="string">-bed2</span> <span class="operator">=&gt;</span> <span class="string">'mask.bed.gz'</span> <span class="operator">);</span>
 
 <span class="comment"># be more strict</span>
 <span class="variable">$bedtools_fac</span><span class="operator">-&gt;</span><span class="variable">set_parameters</span><span class="operator">(</span> <span class="string">-strandedness</span> <span class="operator">=&gt;</span> <span class="number">1</span> <span class="operator">);</span>
 
 <span class="comment"># and even more...</span>
 <span class="variable">$bedtools_fac</span><span class="operator">-&gt;</span><span class="variable">set_parameters</span><span class="operator">(</span> <span class="string">-minimum_overlap</span> <span class="operator">=&gt;</span> <span class="number">1e-6</span> <span class="operator">);</span>
 
 <span class="comment"># create a Bio::SeqFeature::Collection object</span>
 <span class="variable">$features</span> <span class="operator">=</span> <span class="variable">$bedtools_fac</span><span class="operator">-&gt;</span><span class="variable">result</span><span class="operator">(</span> <span class="string">-want</span> <span class="operator">=&gt;</span> <span class="string">'Bio::SeqFeature::Collection'</span> <span class="operator">);</span>
</code></code></pre>

<h1 id="DEPRECATION-WARNING">DEPRECATION WARNING</h1>

<p>Most executables from BEDTools v&gt;=2.10.1 can read GFF and VCF formats in addition to BED format. This requires the use of a new input file param, shown in the following documentation, &#39;-bgv&#39;, in place of &#39;-bed&#39; for the executables that can do this.</p>

<p>This behaviour breaks existing scripts.</p>

<h1 id="DESCRIPTION">DESCRIPTION</h1>

<p>This module provides a wrapper interface for Aaron R. Quinlan and Ira M. Hall&#39;s utilities <code><code>BEDTools</code></code> that allow for (among other things):</p>

<ul>

<li><p>Intersecting two BED files in search of overlapping features.</p>

</li>
<li><p>Merging overlapping features.</p>

</li>
<li><p>Screening for paired-end (PE) overlaps between PE sequences and existing genomic features.</p>

</li>
<li><p>Calculating the depth and breadth of sequence coverage across defined &quot;windows&quot; in a genome.</p>

</li>
</ul>

<p>(see <a href="http://code.google.com/p/bedtools/">http://code.google.com/p/bedtools/</a> for manuals and downloads).</p>

<h1 id="OPTIONS">OPTIONS</h1>

<p><code><code>BEDTools</code></code> is a suite of 17 commandline executable. This module attempts to provide and options comprehensively. You can browse the choices like so:</p>

<pre><code><code> <span class="variable">$bedtools_fac</span> <span class="operator">=</span> <span class="variable">Bio::Tools::Run::BEDTools</span><span class="operator">-&gt;</span><span class="variable">new</span><span class="operator">;</span>
 
 <span class="comment"># all bowtie commands</span>
 <span class="variable">@all_commands</span> <span class="operator">=</span> <span class="variable">$bedtools_fac</span><span class="operator">-&gt;</span><span class="variable">available_parameters</span><span class="operator">(</span><span class="string">'commands'</span><span class="operator">);</span>
 <span class="variable">@all_commands</span> <span class="operator">=</span> <span class="variable">$bedtools_fac</span><span class="operator">-&gt;</span><span class="variable">available_commands</span><span class="operator">;</span> <span class="comment"># alias</span>
 
 <span class="comment"># just for default command ('bam_to_bed')</span>
 <span class="variable">@btb_params</span> <span class="operator">=</span> <span class="variable">$bedtools_fac</span><span class="operator">-&gt;</span><span class="variable">available_parameters</span><span class="operator">(</span><span class="string">'params'</span><span class="operator">);</span>
 <span class="variable">@btb_switches</span> <span class="operator">=</span> <span class="variable">$bedtools_fac</span><span class="operator">-&gt;</span><span class="variable">available_parameters</span><span class="operator">(</span><span class="string">'switches'</span><span class="operator">);</span>
 <span class="variable">@btb_all_options</span> <span class="operator">=</span> <span class="variable">$bedtools_fac</span><span class="operator">-&gt;</span><span class="variable">available_parameters</span><span class="operator">();</span>
</code></code></pre>

<p>Reasonably mnemonic names have been assigned to the single-letter command line options. These are the names returned by <code><code>available_parameters</code></code>, and can be used in the factory constructor like typical BioPerl named parameters.</p>

<p>As a number of options are mutually exclusive, and the interpretation of intent is based on last-pass option reaching bowtie with potentially unpredicted results. This module will prevent inconsistent switches and parameters from being passed.</p>

<p>See <a href="http://code.google.com/p/bedtools/">http://code.google.com/p/bedtools/</a> for details of BEDTools options.</p>

<h1 id="FILES">FILES</h1>

<p>When a command requires filenames, these are provided to the <code><code>run</code></code> method, not the constructor (<code><code>new()</code></code>). To see the set of files required by a command, use <code><code>available_parameters(&#39;filespec&#39;)</code></code> or the alias <code><code>filespec()</code></code>:</p>

<pre><code><code>  <span class="variable">$bedtools_fac</span> <span class="operator">=</span> <span class="variable">Bio::Tools::Run::BEDTools</span><span class="operator">-&gt;</span><span class="variable">new</span><span class="operator">(</span> <span class="string">-command</span> <span class="operator">=&gt;</span> <span class="string">'pair_to_bed'</span> <span class="operator">);</span>
  <span class="variable">@filespec</span> <span class="operator">=</span> <span class="variable">$bedtools_fac</span><span class="operator">-&gt;</span><span class="variable">filespec</span><span class="operator">;</span>
</code></code></pre>

<p>This example returns the following array:</p>

<pre><code><code> #bedpe
 #bam
 bed
 #out</code></code></pre>

<p>This indicates that the bed (<code><code>BEDTools</code></code> BED format) file MUST be specified, and that the out, bedpe (<code><code>BEDTools</code></code> BEDPE format) and bam (<code><code>SAM</code></code> binary format) file MAY be specified (Note that in this case you MUST provide ONE of bedpe OR bam, the module at this stage does not allow this information to be queried). Use these in the <code><code>run</code></code> call like so:</p>

<pre><code><code> <span class="variable">$bedtools_fac</span><span class="operator">-&gt;</span><span class="variable">run</span><span class="operator">(</span> <span class="string">-bedpe</span> <span class="operator">=&gt;</span> <span class="string">'paired.bedpe'</span><span class="operator">,</span>
                     <span class="string">-bgv</span> <span class="operator">=&gt;</span> <span class="string">'genes.bed'</span><span class="operator">,</span>
                     <span class="string">-out</span> <span class="operator">=&gt;</span> <span class="string">'overlap'</span> <span class="operator">);</span>
</code></code></pre>

<p>The object will store the programs STDERR output for you in the <code><code>stderr()</code></code> attribute:</p>

<pre><code><code> <span class="variable">handle_bed_warning</span><span class="operator">(</span><span class="variable">$bedtools_fac</span><span class="operator">)</span> <span class="keyword">if</span> <span class="operator">(</span><span class="variable">$bedtools_fac</span><span class="operator">-&gt;</span><span class="variable">stderr</span> <span class="operator">=~</span> <span class="regex">/Usage:/</span><span class="operator">);</span>
</code></code></pre>

<p>For the commands &#39;fasta_from_bed&#39; and &#39;mask_fasta_from_bed&#39; STDOUT will also be captured in the <code><code>stdout()</code></code> attribute by default and all other commands can be forced to capture program output in STDOUT by setting the -out filespec parameter to &#39;-&#39;.</p>

<h1 id="FEEDBACK">FEEDBACK</h1>

<h2 id="Mailing-Lists">Mailing Lists</h2>

<p>User feedback is an integral part of the evolution of this and other Bioperl modules. Send your comments and suggestions preferably to the Bioperl mailing list. Your participation is much appreciated.</p>

<pre><code><code>  bioperl-l@bioperl.org                  - General discussion
  http://bioperl.org/wiki/Mailing_lists  - About the mailing lists</code></code></pre>

<h2 id="Support">Support</h2>

<p>Please direct usage questions or support issues to the mailing list:</p>

<p><a>bioperl-l@bioperl.org</a></p>

<p>Rather than to the module maintainer directly. Many experienced and reponsive experts will be able look at the problem and quickly address it. Please include a thorough description of the problem with code and data examples if at all possible.</p>

<h2 id="Reporting-Bugs">Reporting Bugs</h2>

<p>Report bugs to the Bioperl bug tracking system to help us keep track of the bugs and their resolution. Bug reports can be submitted via the web:</p>

<pre><code><code>  http://redmine.open-bio.org/projects/bioperl/</code></code></pre>

<h1 id="AUTHOR---Dan-Kortschak">AUTHOR - Dan Kortschak</h1>

<pre><code><code> Email dan.kortschak adelaide.edu.au</code></code></pre>

<h1 id="CONTRIBUTORS">CONTRIBUTORS</h1>

<p>Additional contributors names and emails here</p>

<h1 id="APPENDIX">APPENDIX</h1>

<p>The rest of the documentation details each of the object methods. Internal methods are usually preceded with a _</p>

<h2 id="new-">new()</h2>

<pre><code><code> <span class="variable">Title</span>   <span class="operator">:</span> <span class="variable">new</span>
 <span class="variable">Usage</span>   <span class="operator">:</span> <span class="keyword">my</span> <span class="variable">$obj</span> <span class="operator">=</span> <span class="variable">new</span> <span class="variable">Bio::Tools::Run::BEDTools</span><span class="operator">();</span>
 <span class="variable">Function</span><span class="operator">:</span> <span class="variable">Builds</span> <span class="variable">a</span> <span class="variable">new</span> <span class="variable">Bio::Tools::Run::BEDTools</span> <span class="variable">object</span>
 <span class="variable">Returns</span> <span class="operator">:</span> <span class="variable">an</span> <span class="variable">instance</span> <span class="variable">of</span> <span class="variable">Bio::Tools::Run::BEDTools</span>
 <span class="variable">Args</span>    <span class="operator">:</span>
</code></code></pre>

<h2 id="run-">run()</h2>

<pre><code><code> <span class="variable">Title</span>   <span class="operator">:</span> <span class="variable">run</span>
 <span class="variable">Usage</span>   <span class="operator">:</span> <span class="variable">$result</span> <span class="operator">=</span> <span class="variable">$bedtools_fac</span><span class="operator">-&gt;</span><span class="variable">run</span><span class="operator">(</span><span class="variable">%params</span><span class="operator">);</span>
 <span class="variable">Function</span><span class="operator">:</span> <span class="variable">Run</span> <span class="variable">a</span> <span class="variable">BEDTools</span> <span class="variable">command</span><span class="operator">.</span>
 <span class="variable">Returns</span> <span class="operator">:</span> <span class="variable">Command</span> <span class="variable">results</span> <span class="operator">(</span><span class="variable">file</span><span class="operator">,</span> <span class="variable">IO</span> <span class="variable">object</span> <span class="keyword">or</span> <span class="variable">Bio</span> <span class="variable">object</span><span class="operator">)</span>
 <span class="variable">Args</span>    <span class="operator">:</span> <span class="variable">Dependent</span> <span class="variable">on</span> <span class="variable">filespec</span> <span class="keyword">for</span> <span class="variable">command</span><span class="operator">.</span> 
           <span class="variable">See</span> <span class="variable">$bedtools_fac</span><span class="operator">-&gt;</span><span class="variable">filespec</span> <span class="keyword">and</span> <span class="variable">BEDTools</span> <span class="variable">Manual</span><span class="operator">.</span>
           <span class="variable">Also</span> <span class="variable">accepts</span> <span class="operator">-</span><span class="string">want</span> <span class="operator">=&gt;</span> <span class="string">'(raw|format|&lt;object_class&gt;)'</span> <span class="operator">-</span> <span class="variable">see</span> <span class="variable">want</span><span class="operator">().</span>
 <span class="variable">Note</span>    <span class="operator">:</span> <span class="variable">gzipped</span> <span class="variable">inputs</span> <span class="variable">are</span> <span class="variable">allowed</span> <span class="keyword">if</span> <span class="variable">IO::Uncompress::Gunzip</span>
           <span class="variable">is</span> <span class="variable">available</span>
</code></code></pre>

<pre><code><code>           Command                      &lt;in&gt;                    &lt;out&gt;

           annotate             bgv             ann(s)          #out</code></code></pre>

<pre><code><code>           graph_union          bg_files                        #out</code></code></pre>

<pre><code><code>           fasta_from_bed       seq             bgv             #out

           mask_fasta_from_bed  seq             bgv             #out</code></code></pre>

<pre><code><code>           bam_to_bed           bam                             #out</code></code></pre>

<pre><code><code>           bed_to_IGV           bgv                             #out

           merge                bgv                             #out

           sort                 bgv                             #out

           links                bgv                             #out</code></code></pre>

<pre><code><code>           b12_to_b6            bed                             #out

           overlap              bed                             #out

           group_by             bed                             #out</code></code></pre>

<pre><code><code>           bed_to_bam           bgv                             #out

           shuffle              bgv             genome          #out

           slop                 bgv             genome          #out

           complement           bgv             genome          #out</code></code></pre>

<pre><code><code>           genome_coverage      bed             genome          #out</code></code></pre>

<pre><code><code>           window               bgv1            bgv2            #out

           closest              bgv1            bgv2            #out

           coverage             bgv1            bgv2            #out

           subtract             bgv1            bgv2            #out</code></code></pre>

<pre><code><code>           pair_to_pair         bedpe1          bedpe2          #out</code></code></pre>

<pre><code><code>           intersect            bgv1|bam        bgv2            #out</code></code></pre>

<pre><code><code>           pair_to_bed          bedpe|bam       bgv             #out

           bgv* signifies any of BED, GFF or VCF. ann is a bgv.
           
           NOTE: Replace &#39;bgv&#39; with &#39;bed&#39; unless $use_bgv is set.</code></code></pre>

<h2 id="want-">want()</h2>

<pre><code><code> Title   : want
 Usage   : $bowtiefac-&gt;want( $class )
 Function: make factory return $class, or &#39;raw&#39; results in file
           or &#39;format&#39; for result format
           All commands can return Bio::Root::IO
           commands returning:       can return object:
           - BED or BEDPE            - Bio::SeqFeature::Collection
           - sequence                - Bio::SeqIO
 Returns : return wanted type
 Args    : [optional] string indicating class or raw of wanted result</code></code></pre>

<h2 id="result-">result()</h2>

<pre><code><code> Title   : result
 Usage   : $bedtoolsfac-&gt;result( [-want =&gt; $type|$format] )
 Function: return result in wanted format
 Returns : results
 Args    : [optional] hashref of wanted type
 Note    : -want arg does not persist between result() call when
           specified in result(), for persistence, use want()</code></code></pre>

<h2 id="determine_format-">_determine_format()</h2>

<pre><code><code> Title   : _determine_format( $has_run )
 Usage   : $bedtools-fac-&gt;_determine_format
 Function: determine the format of output for current options
 Returns : format of bowtie output
 Args    : [optional] boolean to indicate result exists</code></code></pre>

<h2 id="read_bed-">_read_bed()</h2>

<pre><code><code> Title   : _read_bed()
 Usage   : $bedtools_fac-&gt;_read_bed
 Function: return a Bio::SeqFeature::Collection object from a BED file 
 Returns : Bio::SeqFeature::Collection
 Args    : </code></code></pre>

<h2 id="read_bedpe-">_read_bedpe()</h2>

<pre><code><code> Title   : _read_bedpe()
 Usage   : $bedtools_fac-&gt;_read_bedpe
 Function: return a Bio::SeqFeature::Collection object from a BEDPE file 
 Returns : Bio::SeqFeature::Collection
 Args    : </code></code></pre>

<h2 id="validate_file_input-">_validate_file_input()</h2>

<pre><code><code> Title   : _validate_file_input
 Usage   : $bedtools_fac-&gt;_validate_file_input( -type =&gt; $file )
 Function: validate file type for file spec
 Returns : file type if valid type for file spec
 Args    : hash of filespec =&gt; file_name</code></code></pre>

<h2 id="version-">version()</h2>

<pre><code><code> Title   : version
 Usage   : $version = $bedtools_fac-&gt;version()
 Function: Returns the program version (if available)
 Returns : string representing location and version of the program</code></code></pre>


</body>

</html>


