<?xml version="1.0" ?>
<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.0 Transitional//EN">
<html>
<head>
<!-- saved from url=(0017)http://localhost/ -->
<script language="JavaScript" src="../../../../../displayToc.js"></script>
<script language="JavaScript" src="../../../../../tocParas.js"></script>
<script language="JavaScript" src="../../../../../tocTab.js"></script>
<link rel="stylesheet" type="text/css" href="../../../../../scineplex.css">
<title></title>
<link rel="stylesheet" href="../../../../../Active.css" type="text/css" />
<meta http-equiv="content-type" content="text/html; charset=utf-8" />
<link rev="made" href="mailto:support@ActiveState.com" />
</head>

<body>



<ul id="index">
  <li><a href="#NAME">NAME</a></li>
  <li><a href="#SYNOPSIS">SYNOPSIS</a></li>
  <li><a href="#DESCRIPTION">DESCRIPTION</a></li>
  <li><a href="#FEEDBACK">FEEDBACK</a>
    <ul>
      <li><a href="#Mailing-Lists">Mailing Lists</a></li>
      <li><a href="#Support">Support</a></li>
      <li><a href="#Reporting-Bugs">Reporting Bugs</a></li>
    </ul>
  </li>
  <li><a href="#AUTHORS">AUTHORS</a></li>
  <li><a href="#APPENDIX">APPENDIX</a>
    <ul>
      <li><a href="#new">new</a></li>
      <li><a href="#out_type">out_type</a></li>
      <li><a href="#run">run</a></li>
      <li><a href="#run1">_run</a></li>
    </ul>
  </li>
</ul>

<h1 id="NAME">NAME</h1>

<p>Bio::Tools::Run::Cap3 - wrapper for CAP3</p>

<h1 id="SYNOPSIS">SYNOPSIS</h1>

<pre><code><code>  <span class="keyword">use</span> <span class="variable">Bio::Tools::Run::Cap3</span><span class="operator">;</span>
  <span class="comment"># Run Cap3 using an input FASTA file</span>
  <span class="keyword">my</span> <span class="variable">$factory</span> <span class="operator">=</span> <span class="variable">Bio::Tools::Run::Cap3</span><span class="operator">-&gt;</span><span class="variable">new</span><span class="operator">(</span> <span class="string">-clipping_range</span> <span class="operator">=&gt;</span> <span class="number">150</span> <span class="operator">);</span>
  <span class="keyword">my</span> <span class="variable">$asm_obj</span> <span class="operator">=</span> <span class="variable">$factory</span><span class="operator">-&gt;</span><span class="variable">run</span><span class="operator">(</span><span class="variable">$fasta_file</span><span class="operator">,</span> <span class="variable">$qual_file</span><span class="operator">);</span>
  <span class="comment"># An assembly object is returned by default</span>
  <span class="keyword">for</span> <span class="keyword">my</span> <span class="variable">$contig</span> <span class="operator">(</span><span class="variable">$assembly</span><span class="operator">-&gt;</span><span class="variable">all_contigs</span><span class="operator">)</span> <span class="operator">{</span>
    <span class="operator">...</span> <span class="keyword">do</span> <span class="variable">something</span> <span class="operator">...</span>
  <span class="operator">}</span>
  
  <span class="comment"># Read some sequences</span>
  <span class="keyword">use</span> <span class="variable">Bio::SeqIO</span><span class="operator">;</span>
  <span class="keyword">my</span> <span class="variable">$sio</span> <span class="operator">=</span> <span class="variable">Bio::SeqIO</span><span class="operator">-&gt;</span><span class="variable">new</span><span class="operator">(</span><span class="string">-file</span> <span class="operator">=&gt;</span> <span class="variable">$fasta_file</span><span class="operator">,</span> <span class="string">-format</span> <span class="operator">=&gt;</span> <span class="string">'fasta'</span><span class="operator">);</span>
  <span class="keyword">my</span> <span class="variable">@seqs</span><span class="operator">;</span>
  <span class="keyword">while</span> <span class="operator">(</span><span class="keyword">my</span> <span class="variable">$seq</span> <span class="operator">=</span> <span class="variable">$sio</span><span class="operator">-&gt;</span><span class="variable">next_seq</span><span class="operator">())</span> <span class="operator">{</span>
    <span class="keyword">push</span> <span class="variable">@seqs</span><span class="operator">,</span><span class="variable">$seq</span><span class="operator">;</span>
  <span class="operator">}</span>
  
  <span class="comment"># Run Cap3 using input sequence objects and returning an assembly file</span>
  <span class="keyword">my</span> <span class="variable">$asm_file</span> <span class="operator">=</span> <span class="string">'results.ace'</span><span class="operator">;</span>
  <span class="variable">$factory</span><span class="operator">-&gt;</span><span class="variable">out_type</span><span class="operator">(</span><span class="variable">$asm_file</span><span class="operator">);</span>
  <span class="variable">$factory</span><span class="operator">-&gt;</span><span class="variable">run</span><span class="operator">(\</span><span class="variable">@seqs</span><span class="operator">);</span>
</code></code></pre>

<h1 id="DESCRIPTION">DESCRIPTION</h1>

<pre><code><code>  Wrapper module for CAP3 program</code></code></pre>

<h1 id="FEEDBACK">FEEDBACK</h1>

<h2 id="Mailing-Lists">Mailing Lists</h2>

<p>User feedback is an integral part of the evolution of this and other Bioperl modules. Send your comments and suggestions preferably to one of the Bioperl mailing lists. Your participation is much appreciated.</p>

<pre><code><code>  bioperl-l@bioperl.org                  - General discussion
  http://bioperl.org/wiki/Mailing_lists  - About the mailing lists</code></code></pre>

<h2 id="Support">Support</h2>

<p>Please direct usage questions or support issues to the mailing list:</p>

<p><i>bioperl-l@bioperl.org</i></p>

<p>rather than to the module maintainer directly. Many experienced and reponsive experts will be able look at the problem and quickly address it. Please include a thorough description of the problem with code and data examples if at all possible.</p>

<h2 id="Reporting-Bugs">Reporting Bugs</h2>

<p>Report bugs to the Bioperl bug tracking system to help us keep track the bugs and their resolution. Bug reports can be submitted via the web:</p>

<pre><code><code>  http://redmine.open-bio.org/projects/bioperl/</code></code></pre>

<h1 id="AUTHORS">AUTHORS</h1>

<p>Marc Logghe</p>

<h1 id="APPENDIX">APPENDIX</h1>

<p>The rest of the documentation details each of the object methods. Internal methods are usually preceded with a _</p>

<h2 id="new">new</h2>

<pre><code><code> Title   : new
 Usage   : $factory-&gt;new(
             -overlap_length_cutoff   =&gt; 35,
             -overlap_identity_cutoff =&gt; 98 # %
           }
 Function: Create a new Cap3 factory
 Returns : A Bio::Tools::Run::Cap3 object
 Args    : Cap3 options available in this module:
  band_expansion_size        specify band expansion size N &gt; 10 (20)
  differences_quality_cutoff specify base quality cutoff for differences N &gt; 15 (20)
  clipping_quality_cutoff    specify base quality cutoff for clipping N &gt; 5 (12)
  max_qscore_sum             specify max qscore sum at differences N &gt; 20 (200)
  extra_nof_differences      specify clearance between no. of diff N &gt; 10 (30)
  max_gap_length             specify max gap length in any overlap N &gt; 1 (20)
  gap_penalty_factor         specify gap penalty factor N &gt; 0 (6)
  max_overhang_percent       specify max overhang percent length N &gt; 2 (20)
  match_score_factor         specify match score factor N &gt; 0 (2)
  mismatch_score_factor      specify mismatch score factor N &lt; 0 (-5)
  overlap_length_cutoff / minimum_overlap_length
                             specify overlap length cutoff &gt; 20 (40)
  overlap_identity_cutoff / minimum_overlap_similarity
                             specify overlap percent identity cutoff N &gt; 65 (80)
  reverse_orientation_value  specify reverse orientation value N &gt;= 0 (1)
  overlap_score_cutoff       specify overlap similarity score cutoff N &gt; 400 (900)
  max_word_occurrences       specify max number of word matches N &gt; 30 (300)
  min_correction_constraints specify min number of constraints for correction N &gt; 0 (3)
  min_linking_constraints    specify min number of constraints for linking N &gt; 0 (2)
  clipping_info_file         specify file name for clipping information (none)
  output_prefix_string       specify prefix string for output file names (cap)
  clipping_range             specify clipping range N &gt; 5 (250)
  min_clip_good_reads        specify min no. of good reads at clip pos N &gt; 0 (3)</code></code></pre>

<h2 id="out_type">out_type</h2>

<pre><code><code> Title   : out_type
 Usage   : $assembler-&gt;out_type(&#39;Bio::Assembly::ScaffoldI&#39;)
 Function: Get/set the desired type of output
 Returns : The type of results to return
 Args    : Desired type of results to return (optional):
                 &#39;Bio::Assembly::IO&#39; object
                 &#39;Bio::Assembly::ScaffoldI&#39; object (default)
                 The name of a file to save the results in</code></code></pre>

<h2 id="run">run</h2>

<pre><code><code> <span class="variable">Title</span>   <span class="operator">:</span>   <span class="variable">run</span>
 <span class="variable">Usage</span>   <span class="operator">:</span>   <span class="variable">$asm</span> <span class="operator">=</span> <span class="variable">$factory</span><span class="operator">-&gt;</span><span class="variable">run</span><span class="operator">(</span><span class="variable">$fasta_file</span><span class="operator">);</span>
 <span class="variable">Function</span><span class="operator">:</span>   <span class="variable">Run</span> <span class="variable">CAP3</span>
 <span class="variable">Returns</span> <span class="operator">:</span>   <span class="variable">Assembly</span> <span class="variable">results</span> <span class="operator">(</span><span class="variable">file</span><span class="operator">,</span> <span class="variable">IO</span> <span class="variable">object</span> <span class="keyword">or</span> <span class="variable">assembly</span> <span class="variable">object</span><span class="operator">)</span>
 <span class="variable">Args</span>    <span class="operator">:</span>   <span class="operator">-</span> <span class="variable">sequence</span> <span class="variable">input</span> <span class="operator">(</span><span class="variable">FASTA</span> <span class="variable">file</span> <span class="keyword">or</span> <span class="variable">sequence</span> <span class="variable">object</span> <span class="variable">arrayref</span><span class="operator">)</span>
             <span class="operator">-</span> <span class="variable">optional</span> <span class="variable">quality</span> <span class="variable">score</span> <span class="variable">input</span> <span class="operator">(</span><span class="variable">QUAL</span> <span class="variable">file</span> <span class="keyword">or</span> <span class="variable">quality</span> <span class="variable">score</span> <span class="variable">object</span>
               <span class="variable">arrayref</span><span class="operator">)</span>
</code></code></pre>

<h2 id="run1">_run</h2>

<pre><code><code> Title   :   _run
 Usage   :   $factory-&gt;_run()
 Function:   Make a system call and run Cap3
 Returns :   An assembly file
 Args    :   - FASTA file
             - optional QUAL file</code></code></pre>


</body>

</html>


