<?xml version="1.0" ?>
<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.0 Transitional//EN">
<html>
<head>
<!-- saved from url=(0017)http://localhost/ -->
<script language="JavaScript" src="../../../../../displayToc.js"></script>
<script language="JavaScript" src="../../../../../tocParas.js"></script>
<script language="JavaScript" src="../../../../../tocTab.js"></script>
<link rel="stylesheet" type="text/css" href="../../../../../scineplex.css">
<title></title>
<link rel="stylesheet" href="../../../../../Active.css" type="text/css" />
<meta http-equiv="content-type" content="text/html; charset=utf-8" />
<link rev="made" href="mailto:support@ActiveState.com" />
</head>

<body>



<ul id="index">
  <li><a href="#NAME">NAME</a></li>
  <li><a href="#SYNOPSIS">SYNOPSIS</a></li>
  <li><a href="#DESCRIPTION">DESCRIPTION</a></li>
  <li><a href="#FEEDBACK">FEEDBACK</a>
    <ul>
      <li><a href="#Mailing-Lists">Mailing Lists</a></li>
      <li><a href="#Support">Support</a></li>
      <li><a href="#Reporting-Bugs">Reporting Bugs</a></li>
    </ul>
  </li>
  <li><a href="#AUTHOR---Heikki-Lehvaslaiho">AUTHOR - Heikki Lehvaslaiho</a></li>
  <li><a href="#APPENDIX">APPENDIX</a>
    <ul>
      <li><a href="#new">new</a></li>
      <li><a href="#name">name</a></li>
      <li><a href="#print">print</a></li>
      <li><a href="#mandatory">mandatory</a></li>
      <li><a href="#Qualifier-queries">Qualifier queries</a></li>
      <li><a href="#qualifier">qualifier</a></li>
      <li><a href="#category">category</a></li>
      <li><a href="#values">values</a></li>
      <li><a href="#descr">descr</a></li>
      <li><a href="#unnamed">unnamed</a></li>
      <li><a href="#default">default</a></li>
    </ul>
  </li>
</ul>

<h1 id="NAME">NAME</h1>

<p>Bio::Tools::Run::EMBOSSacd - class for EMBOSS Application qualifiers</p>

<h1 id="SYNOPSIS">SYNOPSIS</h1>

<pre><code><code>  <span class="keyword">use</span> <span class="variable">Bio::Factory::EMBOSS</span><span class="operator">;</span>
  <span class="comment"># get an EMBOSS application object from the EMBOSS factory</span>
  <span class="variable">$factory</span> <span class="operator">=</span> <span class="variable">Bio::Factory::EMBOSS</span><span class="operator">-&gt;</span><span class="variable">new</span><span class="operator">();</span>
  <span class="variable">$application</span> <span class="operator">=</span> <span class="variable">$factory</span><span class="operator">-&gt;</span><span class="variable">program</span><span class="operator">(</span><span class="string">'embossversion'</span><span class="operator">);</span>
  <span class="comment"># run the application with an optional hash containing parameters</span>
  <span class="variable">$result</span> <span class="operator">=</span> <span class="variable">$application</span><span class="operator">-&gt;</span><span class="variable">run</span><span class="operator">();</span> <span class="comment"># returns a string or creates a file</span>
  <span class="keyword">print</span> <span class="variable">$result</span> <span class="operator">.</span> <span class="string">"\n"</span><span class="operator">;</span>
  
  <span class="variable">$water</span> <span class="operator">=</span> <span class="variable">$factory</span><span class="operator">-&gt;</span><span class="variable">program</span><span class="operator">(</span><span class="string">'water'</span><span class="operator">);</span>
  
  <span class="comment"># here is an example of running the application</span>
  <span class="comment"># water can compare 1 seq against 1-&gt;many sequences</span>
  <span class="comment"># in a database using Smith-Waterman</span>
  <span class="keyword">my</span> <span class="variable">$seq_to_test</span><span class="operator">;</span> <span class="comment"># this would have a seq here</span>
  <span class="keyword">my</span> <span class="variable">@seqs_to_check</span><span class="operator">;</span> <span class="comment"># this would be a list of seqs to compare </span>
                     <span class="comment"># (could be just 1)</span>
  <span class="keyword">my</span> <span class="variable">$wateroutfile</span> <span class="operator">=</span> <span class="string">'out.water'</span><span class="operator">;</span>
  <span class="variable">$water</span><span class="operator">-&gt;</span><span class="variable">run</span><span class="operator">(</span><span class="operator">{</span> <span class="string">-sequencea</span> <span class="operator">=&gt;</span> <span class="variable">$seq_to_test</span><span class="operator">,</span>
                <span class="string">-seqall</span>    <span class="operator">=&gt;</span> <span class="operator">\</span><span class="variable">@seqs_to_check</span><span class="operator">,</span>
                <span class="string">-gapopen</span>   <span class="operator">=&gt;</span> <span class="string">'10.0'</span><span class="operator">,</span>
                <span class="string">-gapextend</span> <span class="operator">=&gt;</span> <span class="string">'0.5'</span><span class="operator">,</span>
                <span class="string">-outfile</span>   <span class="operator">=&gt;</span> <span class="variable">$wateroutfile</span><span class="operator">}</span><span class="operator">);</span>
  <span class="comment"># now you might want to get the alignment</span>
  <span class="keyword">use</span> <span class="variable">Bio::AlignIO</span><span class="operator">;</span>
  <span class="keyword">my</span> <span class="variable">$alnin</span> <span class="operator">=</span> <span class="variable">Bio::AlignIO</span><span class="operator">-&gt;</span><span class="variable">new</span><span class="operator">(</span><span class="string">-format</span> <span class="operator">=&gt;</span> <span class="string">'emboss'</span><span class="operator">,</span>
                                              <span class="string">-file</span>   <span class="operator">=&gt;</span> <span class="variable">$wateroutfile</span><span class="operator">);</span>
  
  <span class="keyword">while</span><span class="operator">(</span> <span class="keyword">my</span> <span class="variable">$aln</span> <span class="operator">=</span> <span class="variable">$alnin</span><span class="operator">-&gt;</span><span class="variable">next_aln</span> <span class="operator">)</span> <span class="operator">{</span>
      <span class="comment"># process the alignment -- these will be Bio::SimpleAlign objects</span>
  <span class="operator">}</span>
</code></code></pre>

<h1 id="DESCRIPTION">DESCRIPTION</h1>

<p>The EMBOSSacd represents all the possible command line arguments that can be given to an EMBOSS application.</p>

<p>Do not create this object directly. It will be created and attached to its corresponding Bio::Tools::Run::EMBOSSApplication if you set</p>

<pre><code><code>  $application-&gt;verbose &gt; 0</code></code></pre>

<p>Call</p>

<pre><code><code>  $application-&gt;acd</code></code></pre>

<p>to retrive the Bio::Tools::Run::EMBOSSApplication::EMBOSSacd object.</p>

<p>See also <a href="../../../../../../../../public/home/fxing/.cpan/build/BioPerl-Run-1.006900-_NTLno/blib/lib/Bio/Tools/Run/EMBOSSApplication.html">Bio::Tools::Run::EMBOSSApplication</a> and <a href="../../../../../../../../public/home/fxing/.cpan/build/BioPerl-Run-1.006900-_NTLno/blib/lib/Bio/Factory/EMBOSS.html">Bio::Factory::EMBOSS</a>.</p>

<h1 id="FEEDBACK">FEEDBACK</h1>

<h2 id="Mailing-Lists">Mailing Lists</h2>

<p>User feedback is an integral part of the evolution of this and other Bioperl modules. Send your comments and suggestions preferably to the Bioperl mailing lists Your participation is much appreciated.</p>

<pre><code><code>  bioperl-l@bioperl.org                  - General discussion
  http://bioperl.org/wiki/Mailing_lists  - About the mailing lists</code></code></pre>

<h2 id="Support">Support</h2>

<p>Please direct usage questions or support issues to the mailing list:</p>

<p><i>bioperl-l@bioperl.org</i></p>

<p>rather than to the module maintainer directly. Many experienced and reponsive experts will be able look at the problem and quickly address it. Please include a thorough description of the problem with code and data examples if at all possible.</p>

<h2 id="Reporting-Bugs">Reporting Bugs</h2>

<p>report bugs to the Bioperl bug tracking system to help us keep track the bugs and their resolution. Bug reports can be submitted via the web:</p>

<pre><code><code>  http://redmine.open-bio.org/projects/bioperl/</code></code></pre>

<h1 id="AUTHOR---Heikki-Lehvaslaiho">AUTHOR - Heikki Lehvaslaiho</h1>

<p>Email: heikki-at-bioperl-dot-org Address:</p>

<pre><code><code>     EMBL Outstation, European Bioinformatics Institute
     Wellcome Trust Genome Campus, Hinxton
     Cambs. CB10 1SD, United Kingdom</code></code></pre>

<h1 id="APPENDIX">APPENDIX</h1>

<p>The rest of the documentation details each of the object methods. Internal methods are usually preceded with a _</p>

<h2 id="new">new</h2>

<pre><code><code> <span class="variable">Title</span>   <span class="operator">:</span> <span class="variable">new</span>
 <span class="variable">Usage</span>   <span class="operator">:</span> <span class="variable">$emboss_prog</span><span class="operator">-&gt;</span><span class="variable">acd</span><span class="operator">(</span><span class="variable">$prog_name</span><span class="operator">);</span>
 <span class="variable">Function</span><span class="operator">:</span> <span class="variable">Constructor</span> <span class="keyword">for</span> <span class="variable">the</span> <span class="variable">class</span><span class="operator">.</span>
           <span class="variable">Calls</span> <span class="variable">EMBOSS</span> <span class="variable">program</span> <span class="string">'acdc'</span><span class="operator">,</span> <span class="variable">converts</span> <span class="variable">the</span>
           <span class="variable">HTML</span> <span class="variable">output</span> <span class="variable">into</span> <span class="variable">XML</span> <span class="keyword">and</span> <span class="variable">uses</span> <span class="variable">XML::Twig</span> <span class="variable">XML</span> 
           <span class="variable">parser</span> <span class="variable">to</span> <span class="keyword">write</span> <span class="variable">out</span> <span class="variable">a</span> <span class="variable">hash</span> <span class="variable">of</span> <span class="variable">qualifiers</span> <span class="variable">which</span> <span class="variable">is</span>
           <span class="variable">then</span> <span class="variable">blessed</span><span class="operator">.</span>
 <span class="variable">Throws</span>  <span class="operator">:</span> <span class="variable">without</span> <span class="variable">program</span> <span class="variable">name</span>
 <span class="variable">Returns</span> <span class="operator">:</span> <span class="variable">new</span> <span class="variable">object</span>
 <span class="variable">Args</span>    <span class="operator">:</span> <span class="variable">EMBOSS</span> <span class="variable">program</span> <span class="variable">name</span>
</code></code></pre>

<h2 id="name">name</h2>

<pre><code><code> Title   : name
 Usage   : $embossacd-&gt;name
 Function: sets/gets the name of the EMBOSS program
           Setting is done by the EMBOSSApplication object,
           you should only get it.
 Throws  : 
 Returns : name string
 Args    : None</code></code></pre>

<h2 id="print">print</h2>

<pre><code><code> <span class="variable">Title</span>   <span class="operator">:</span> <span class="keyword">print</span>
 <span class="variable">Usage</span>   <span class="operator">:</span> <span class="variable">$embossacd</span><span class="operator">-&gt;</span><span class="variable">print</span><span class="operator">;</span> <span class="variable">$embossacd</span><span class="operator">-&gt;</span><span class="variable">print</span><span class="operator">(</span><span class="string">'-word'</span><span class="operator">);</span>
 <span class="variable">Function</span><span class="operator">:</span> <span class="variable">Print</span> <span class="variable">out</span> <span class="variable">the</span> <span class="variable">qualifiers</span><span class="operator">.</span>
           <span class="variable">Uses</span> <span class="variable">Data::Dumper</span> <span class="variable">to</span> <span class="keyword">print</span> <span class="variable">the</span> <span class="variable">qualifiers</span> <span class="variable">into</span> <span class="variable">STDOUT</span><span class="operator">.</span>
           <span class="variable">A</span> <span class="variable">valid</span> <span class="variable">qualifier</span> <span class="variable">name</span> <span class="keyword">given</span> <span class="variable">as</span> <span class="variable">an</span> <span class="variable">argment</span> <span class="variable">limits</span> <span class="variable">the</span> <span class="variable">output</span><span class="operator">.</span>
 <span class="variable">Throws</span>  <span class="operator">:</span> 
 <span class="variable">Returns</span> <span class="operator">:</span> <span class="keyword">print</span> <span class="variable">string</span>
 <span class="variable">Args</span>    <span class="operator">:</span> <span class="variable">optional</span> <span class="variable">qualifier</span> <span class="variable">name</span>
</code></code></pre>

<h2 id="mandatory">mandatory</h2>

<pre><code><code> Title   : mandatory
 Usage   : $acd-&gt;mandatory
 Function: gets a  mandatory subset of qualifiers
 Throws  : 
 Returns : Bio::Tools::Run::EMBOSSacd object
 Args    : none</code></code></pre>

<h2 id="Qualifier-queries">Qualifier queries</h2>

<p>These methods can be used test qualifier names and read values.</p>

<h2 id="qualifier">qualifier</h2>

<pre><code><code> Title   : qualifier
 Usage   : $acd-&gt;qualifier($string)
 Function: tests for the existence of the qualifier
 Throws  : 
 Returns : boolean
 Args    : string, name of the qualifier</code></code></pre>

<h2 id="category">category</h2>

<pre><code><code> Title   : category
 Usage   : $acd-&gt;category($qual_name)
 Function: Return the category of the qualifier
 Throws  : 
 Returns : &#39;mandatory&#39; or &#39;optional&#39; or &#39;advanced&#39; or 
            &#39;associated&#39; or &#39;general&#39;
 Args    : string, name of the qualifier</code></code></pre>

<h2 id="values">values</h2>

<pre><code><code> Title   : values
 Usage   : $acd-&gt;values($qual_name)
 Function: Return the possible values for the qualifier
 Throws  : 
 Returns : string
 Args    : string, name of the qualifier</code></code></pre>

<h2 id="descr">descr</h2>

<pre><code><code> Title   : descr
 Usage   : $acd-&gt;descr($qual_name)
 Function: Return the description of the qualifier
 Throws  : 
 Returns : boolean
 Args    : string, name of the qualifier</code></code></pre>

<h2 id="unnamed">unnamed</h2>

<pre><code><code> Title   : unnamed
 Usage   : $acd-&gt;unnamed($qual_name)
 Function: Find if the qualifier can be left unnamed
 Throws  : 
 Returns : 0 if needs to be named, order number otherwise
 Args    : string, name of the qualifier</code></code></pre>

<h2 id="default">default</h2>

<pre><code><code> Title   : default
 Usage   : $acd-&gt;default($qual_name)
 Function: Return the default value for the qualifier
 Throws  : 
 Returns : scalar
 Args    : string, name of the qualifier</code></code></pre>


</body>

</html>


