<?xml version="1.0" ?>
<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.0 Transitional//EN">
<html>
<head>
<!-- saved from url=(0017)http://localhost/ -->
<script language="JavaScript" src="../../../../../displayToc.js"></script>
<script language="JavaScript" src="../../../../../tocParas.js"></script>
<script language="JavaScript" src="../../../../../tocTab.js"></script>
<link rel="stylesheet" type="text/css" href="../../../../../scineplex.css">
<title></title>
<link rel="stylesheet" href="../../../../../Active.css" type="text/css" />
<meta http-equiv="content-type" content="text/html; charset=utf-8" />
<link rev="made" href="mailto:support@ActiveState.com" />
</head>

<body>



<ul id="index">
  <li><a href="#NAME">NAME</a></li>
  <li><a href="#SYNOPSIS">SYNOPSIS</a></li>
  <li><a href="#DESCRIPTION">DESCRIPTION</a>
    <ul>
      <li><a href="#Available-Params:">Available Params:</a></li>
    </ul>
  </li>
  <li><a href="#FEEDBACK">FEEDBACK</a>
    <ul>
      <li><a href="#Mailing-Lists">Mailing Lists</a></li>
      <li><a href="#Support">Support</a></li>
      <li><a href="#Reporting-Bugs">Reporting Bugs</a></li>
    </ul>
  </li>
  <li><a href="#AUTHOR---FUGU-Student-Intern">AUTHOR - FUGU Student Intern</a></li>
  <li><a href="#CONTRIBUTORS">CONTRIBUTORS</a></li>
  <li><a href="#APPENDIX">APPENDIX</a>
    <ul>
      <li><a href="#program_name">program_name</a></li>
      <li><a href="#program_dir">program_dir</a></li>
      <li><a href="#version">version</a></li>
      <li><a href="#predict_genes">predict_genes</a></li>
      <li><a href="#run">run</a></li>
      <li><a href="#run1">_run</a></li>
      <li><a href="#setparams">_setparams</a></li>
      <li><a href="#query_pep_seq">_query_pep_seq</a></li>
      <li><a href="#subject_dna_seq">_subject_dna_seq</a></li>
    </ul>
  </li>
</ul>

<h1 id="NAME">NAME</h1>

<p>Bio::Tools::Run::Genewise - Object for predicting genes in a given sequence given a protein</p>

<h1 id="SYNOPSIS">SYNOPSIS</h1>

<pre><code><code>  <span class="comment"># Build a Genewise alignment factory</span>
  <span class="keyword">my</span> <span class="variable">$factory</span> <span class="operator">=</span> <span class="variable">Bio::Tools::Run::Genewise</span><span class="operator">-&gt;</span><span class="variable">new</span><span class="operator">();</span>
  
  <span class="comment"># Pass the factory 2 Bio:SeqI objects (in the order of query peptide</span>
  <span class="comment"># and target_genomic).</span>
  
  <span class="comment"># @genes is an array of Bio::SeqFeature::Gene::GeneStructure objects</span>
  <span class="keyword">my</span> <span class="variable">@genes</span> <span class="operator">=</span> <span class="variable">$factory</span><span class="operator">-&gt;</span><span class="variable">run</span><span class="operator">(</span><span class="variable">$protein_seq</span><span class="operator">,</span> <span class="variable">$genomic_seq</span><span class="operator">);</span>
  
  <span class="comment"># Alternatively pass the factory a profile HMM filename and a</span>
  <span class="comment"># Bio:SeqI object (in the order of query HMM and target_genomic).</span>
  
  <span class="comment"># Set hmmer switch first to tell genewise to expect an HMM</span>
  <span class="variable">$factory</span><span class="operator">-&gt;</span><span class="variable">hmmer</span><span class="operator">(</span><span class="number">1</span><span class="operator">);</span>
  <span class="keyword">my</span> <span class="variable">@genes</span> <span class="operator">=</span> <span class="variable">$factory</span><span class="operator">-&gt;</span><span class="variable">run</span><span class="operator">(</span><span class="variable">$hmmfile</span><span class="operator">,</span> <span class="variable">$genomic_seq</span><span class="operator">);</span>
</code></code></pre>

<h1 id="DESCRIPTION">DESCRIPTION</h1>

<p>Genewise is a gene prediction program developed by Ewan Birney http://www.sanger.ac.uk/software/wise2.</p>

<h2 id="Available-Params:">Available Params:</h2>

<p>NB: These should be passed without the &#39;-&#39; or they will be ignored, except switches such as &#39;hmmer&#39; (which have no corresponding value) which should be set on the factory object using the AUTOLOADed methods of the same name.</p>

<pre><code><code>  Model    [-codon,-gene,-cfreq,-splice,-subs,-indel,-intron,-null]
  Alg      [-kbyte,-alg]
  HMM      [-hmmer]
  Output   [-gff,-gener,-alb,-pal,-block,-divide]
  Standard [-help,-version,-silent,-quiet,-errorlog]</code></code></pre>

<h1 id="FEEDBACK">FEEDBACK</h1>

<h2 id="Mailing-Lists">Mailing Lists</h2>

<p>User feedback is an integral part of the evolution of this and other Bioperl modules. Send your comments and suggestions preferably to one of the Bioperl mailing lists. Your participation is much appreciated.</p>

<pre><code><code>  bioperl-l@bioperl.org                  - General discussion
  http://bioperl.org/wiki/Mailing_lists  - About the mailing lists</code></code></pre>

<h2 id="Support">Support</h2>

<p>Please direct usage questions or support issues to the mailing list:</p>

<p><i>bioperl-l@bioperl.org</i></p>

<p>rather than to the module maintainer directly. Many experienced and reponsive experts will be able look at the problem and quickly address it. Please include a thorough description of the problem with code and data examples if at all possible.</p>

<h2 id="Reporting-Bugs">Reporting Bugs</h2>

<p>Report bugs to the Bioperl bug tracking system to help us keep track the bugs and their resolution. Bug reports can be submitted via the web:</p>

<pre><code><code>  http://redmine.open-bio.org/projects/bioperl/</code></code></pre>

<h1 id="AUTHOR---FUGU-Student-Intern">AUTHOR - FUGU Student Intern</h1>

<p>Email: fugui@worf.fugu-sg.org</p>

<h1 id="CONTRIBUTORS">CONTRIBUTORS</h1>

<p>Jason Stajich jason-AT-bioperl_DOT_org Keith James kdj@sanger.ac.uk</p>

<h1 id="APPENDIX">APPENDIX</h1>

<p>The rest of the documentation details each of the object methods. Internal methods are usually preceded with a _</p>

<h2 id="program_name">program_name</h2>

<pre><code><code> Title   : program_name
 Usage   : $factory&gt;program_name()
 Function: holds the program name
 Returns:  string
 Args    : None</code></code></pre>

<h2 id="program_dir">program_dir</h2>

<pre><code><code> Title   : program_dir
 Usage   : $factory-&gt;program_dir(@params)
 Function: returns the program directory, obtained from ENV variable.
 Returns:  string
 Args    :</code></code></pre>

<h2 id="version">version</h2>

<pre><code><code> Title   : version
 Usage   : exit if $prog-&gt;version() &lt; 1.8
 Function: Determine the version number of the program
 Example :
 Returns : float or undef
 Args    : none</code></code></pre>

<h2 id="predict_genes">predict_genes</h2>

<pre><code><code> Title   : predict_genes
 Usage   : DEPRECATED. Use $factory-&gt;run($seq1,$seq2)
 Function: Predict genes
 Returns : A Bio::Seqfeature::Gene:GeneStructure object
 Args    : Name of a file containing a set of 2 fasta sequences in the order of
           peptide and genomic sequences
           or else 2  Bio::Seq objects.</code></code></pre>

<p>Throws an exception if argument is not either a string (eg a filename) or 2 Bio::Seq objects. If arguments are strings, throws exception if file corresponding to string name can not be found.</p>

<h2 id="run">run</h2>

<pre><code><code> <span class="variable">Title</span>   <span class="operator">:</span> <span class="variable">run</span>
 <span class="variable">Usage</span>   <span class="operator">:</span> <span class="number">2</span> <span class="variable">sequence</span> <span class="variable">objects</span>
           <span class="variable">$genes</span> <span class="operator">=</span> <span class="variable">$factory</span><span class="operator">-&gt;</span><span class="variable">run</span><span class="operator">(</span><span class="variable">$seq1</span><span class="operator">,</span> <span class="variable">$seq2</span><span class="operator">);</span>
 <span class="variable">Function</span><span class="operator">:</span> <span class="variable">run</span>
 <span class="variable">Returns</span> <span class="operator">:</span> <span class="variable">A</span> <span class="variable">Bio::Seqfeature::Gene</span><span class="operator">:</span><span class="variable">GeneStructure</span> <span class="variable">object</span>
 <span class="variable">Args</span>    <span class="operator">:</span> <span class="variable">Names</span> <span class="variable">of</span> <span class="variable">a</span> <span class="variable">files</span> <span class="keyword">each</span> <span class="variable">containing</span> <span class="variable">a</span> <span class="variable">fasta</span> <span class="variable">sequence</span> <span class="variable">in</span> <span class="variable">the</span> <span class="variable">order</span>
           <span class="variable">of</span> <span class="variable">either</span> <span class="operator">(</span><span class="variable">peptide</span> <span class="variable">sequence</span><span class="operator">,</span> <span class="variable">genomic</span> <span class="variable">sequence</span><span class="operator">)</span> <span class="keyword">or</span> <span class="operator">(</span><span class="variable">profile</span> <span class="variable">HMM</span><span class="operator">,</span>
           <span class="variable">genomic</span> <span class="variable">sequence</span><span class="operator">).</span> <span class="variable">Alternatively</span> <span class="variable">any</span> <span class="variable">of</span> <span class="variable">the</span> <span class="variable">fasta</span> <span class="variable">sequence</span>
           <span class="variable">filenames</span> <span class="variable">may</span> <span class="variable">be</span> <span class="variable">substituted</span> <span class="variable">with</span> <span class="variable">a</span> <span class="variable">Bio::Seq</span> <span class="variable">object</span><span class="operator">.</span>
</code></code></pre>

<p>Throws an exception if argument is not either a string (eg a filename) or Bio::Seq objects. If arguments are strings, throws exception if file corresponding to string name can not be found. Also throws an exception if a profile HMM is expected (the -hmmer genewise switch has been set).</p>

<h2 id="run1">_run</h2>

<pre><code><code> Title   : _run
 Usage   : Internal function, not to be called directly
 Function: Makes actual system call to a genewise program
 Example :
 Returns : L&lt;Bio::SeqFeature::Gene::GeneStructure&gt;
 Args    : Name of a files containing 2 sequences in the order of peptide and genomic</code></code></pre>

<h2 id="setparams">_setparams</h2>

<pre><code><code> Title   :  _setparams
 Usage   :  Internal function, not to be called directly
 Function:  creates a string of params to be used in the command string
 Example :
 Returns :  string of params
 Args    :</code></code></pre>

<h2 id="query_pep_seq">_query_pep_seq</h2>

<pre><code><code> Title   :  _query_pep_seq
 Usage   :  Internal function, not to be called directly
 Function:  get/set for the query sequence
 Example :
 Returns :
 Args    :</code></code></pre>

<h2 id="subject_dna_seq">_subject_dna_seq</h2>

<pre><code><code> Title   :  _subject_dna_seq
 Usage   :  Internal function, not to be called directly
 Function:  get/set for the subject sequence
 Example :
 Returns :

 Args    :</code></code></pre>


</body>

</html>


