<?xml version="1.0" ?>
<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.0 Transitional//EN">
<html>
<head>
<!-- saved from url=(0017)http://localhost/ -->
<script language="JavaScript" src="../../../../../displayToc.js"></script>
<script language="JavaScript" src="../../../../../tocParas.js"></script>
<script language="JavaScript" src="../../../../../tocTab.js"></script>
<link rel="stylesheet" type="text/css" href="../../../../../scineplex.css">
<title></title>
<link rel="stylesheet" href="../../../../../Active.css" type="text/css" />
<meta http-equiv="content-type" content="text/html; charset=utf-8" />
<link rev="made" href="mailto:support@ActiveState.com" />
</head>

<body>



<ul id="index">
  <li><a href="#NAME">NAME</a></li>
  <li><a href="#SYNOPSIS">SYNOPSIS</a></li>
  <li><a href="#DESCRIPTION">DESCRIPTION</a></li>
  <li><a href="#FEEDBACK">FEEDBACK</a>
    <ul>
      <li><a href="#Mailing-Lists">Mailing Lists</a></li>
      <li><a href="#Support">Support</a></li>
      <li><a href="#Reporting-Bugs">Reporting Bugs</a></li>
    </ul>
  </li>
  <li><a href="#AUTHOR---Shawn-Hoon">AUTHOR - Shawn Hoon</a></li>
  <li><a href="#APPENDIX">APPENDIX</a>
    <ul>
      <li><a href="#program_name">program_name</a></li>
      <li><a href="#program_dir">program_dir</a></li>
      <li><a href="#version">version</a></li>
      <li><a href="#run">run</a></li>
      <li><a href="#run1">_run</a></li>
      <li><a href="#setparams">_setparams</a></li>
      <li><a href="#query_pep_seq">_query_pep_seq</a></li>
      <li><a href="#subject_dna_seq">_subject_dna_seq</a></li>
    </ul>
  </li>
</ul>

<h1 id="NAME">NAME</h1>

<p>Bio::Tools::Run::Promoterwise - Wrapper for aligning two sequences using promoterwise</p>

<h1 id="SYNOPSIS">SYNOPSIS</h1>

<pre><code><code>  <span class="comment"># Build a Promoterwise alignment factory</span>
  <span class="keyword">my</span> <span class="variable">@params</span> <span class="operator">=</span> <span class="operator">(</span><span class="string">'-s'</span><span class="operator">=&gt;</span><span class="number">1</span><span class="operator">,</span><span class="string">'-query_start'</span><span class="operator">=&gt;</span><span class="number">10</span><span class="operator">,</span><span class="string">'-dymem'</span><span class="operator">=&gt;</span><span class="number">1</span><span class="operator">);</span>
  <span class="keyword">my</span>  <span class="variable">$factory</span> <span class="operator">=</span> <span class="variable">Bio::Tools::Run::Promoterwise</span><span class="operator">-&gt;</span><span class="variable">new</span><span class="operator">(</span><span class="variable">@params</span><span class="operator">);</span>
  
  <span class="keyword">my</span> <span class="operator">(</span><span class="variable">@fp</span><span class="operator">)=</span> <span class="variable">$factory</span><span class="operator">-&gt;</span><span class="variable">run</span><span class="operator">(</span><span class="variable">$seq1</span><span class="operator">,</span><span class="variable">$seq2</span><span class="operator">);</span>
  
  <span class="comment"># each feature pair is a group of hsps</span>
  <span class="keyword">foreach</span> <span class="keyword">my</span> <span class="variable">$fp</span><span class="operator">(</span><span class="variable">@fp</span><span class="operator">){</span>
    <span class="keyword">print</span> <span class="string">"Hit Length: "</span><span class="operator">.</span><span class="variable">$fp</span><span class="operator">-&gt;</span><span class="variable">feature1</span><span class="operator">-&gt;</span><span class="variable">length</span><span class="operator">.</span><span class="string">"\n"</span><span class="operator">;</span>
    <span class="keyword">print</span> <span class="string">"Hit Start: "</span><span class="operator">.</span><span class="variable">$fp</span><span class="operator">-&gt;</span><span class="variable">feature1</span><span class="operator">-&gt;</span><span class="variable">start</span><span class="operator">.</span><span class="string">"\n"</span><span class="operator">;</span>
    <span class="keyword">print</span> <span class="string">"Hit End: "</span><span class="operator">.</span><span class="variable">$fp</span><span class="operator">-&gt;</span><span class="variable">feature1</span><span class="operator">-&gt;</span><span class="variable">end</span><span class="operator">.</span><span class="string">"\n"</span><span class="operator">;</span>
    <span class="keyword">print</span> <span class="string">"Hsps: \n"</span><span class="operator">;</span>
    <span class="keyword">my</span> <span class="variable">@first_hsp</span> <span class="operator">=</span> <span class="variable">$fp</span><span class="operator">-&gt;</span><span class="variable">feature1</span><span class="operator">-&gt;</span><span class="variable">sub_SeqFeature</span><span class="operator">;</span>
    <span class="keyword">my</span> <span class="variable">@second_hsp</span> <span class="operator">=</span> <span class="variable">$fp</span><span class="operator">-&gt;</span><span class="variable">feature2</span><span class="operator">-&gt;</span><span class="variable">sub_SeqFeature</span><span class="operator">;</span>
    <span class="keyword">for</span> <span class="operator">(</span><span class="variable">$i</span><span class="operator">..</span><span class="variable">$#first_hsp</span><span class="operator">){</span>
      <span class="keyword">print</span> <span class="variable">$first_hsp</span><span class="operator">[</span><span class="variable">$i</span><span class="operator">]</span><span class="operator">-&gt;</span><span class="variable">seq</span><span class="operator">.</span><span class="string">" "</span><span class="operator">.</span><span class="variable">$second_hsp</span><span class="operator">[</span><span class="variable">$i</span><span class="operator">]</span><span class="operator">-&gt;</span><span class="variable">seq</span><span class="operator">.</span><span class="string">"\n"</span><span class="operator">;</span>
    <span class="operator">}</span>
  <span class="operator">}</span>
  <span class="keyword">print</span> <span class="string">"end: "</span><span class="operator">.</span> <span class="variable">$fp</span><span class="operator">-&gt;</span><span class="variable">feature2</span><span class="operator">-&gt;</span><span class="variable">start</span><span class="operator">.</span><span class="string">"\t"</span><span class="operator">.</span><span class="variable">$fp</span><span class="operator">-&gt;</span><span class="variable">feature2</span><span class="operator">-&gt;</span><span class="variable">end</span><span class="operator">.</span><span class="string">"\n"</span><span class="operator">;</span>
  
  <span class="comment">#Available parameters include:</span>
  <span class="comment">#( S T U V QUERY_START QUERY_END TARGET_START</span>
  <span class="comment">#TARGET_END LHWINDOW LHSEED LHALN LHSCORE LHREJECT</span>
  <span class="comment">#LHREJECT LHMAX DYMEM KBYTE DYCACHE)</span>
  <span class="comment">#For an explanation of these parameters, please see documentation</span>
  <span class="comment">#from the Wise package.</span>
</code></code></pre>

<h1 id="DESCRIPTION">DESCRIPTION</h1>

<p>Promoterwise is an alignment algorithm that relaxes the constraint that local alignments have to be co-linear. Otherwise it provides a similar model to DBA, which is designed for promoter sequence alignments by Ewan Birney. It is part of the wise2 package available at: http://www.sanger.ac.uk/software/wise2.</p>

<h1 id="FEEDBACK">FEEDBACK</h1>

<h2 id="Mailing-Lists">Mailing Lists</h2>

<p>User feedback is an integral part of the evolution of this and other Bioperl modules. Send your comments and suggestions preferably to one of the Bioperl mailing lists. Your participation is much appreciated.</p>

<pre><code><code>  bioperl-l@bioperl.org                  - General discussion
  http://bioperl.org/wiki/Mailing_lists  - About the mailing lists</code></code></pre>

<h2 id="Support">Support</h2>

<p>Please direct usage questions or support issues to the mailing list:</p>

<p><i>bioperl-l@bioperl.org</i></p>

<p>rather than to the module maintainer directly. Many experienced and reponsive experts will be able look at the problem and quickly address it. Please include a thorough description of the problem with code and data examples if at all possible.</p>

<h2 id="Reporting-Bugs">Reporting Bugs</h2>

<p>Report bugs to the Bioperl bug tracking system to help us keep track the bugs and their resolution. Bug reports can be submitted via the web:</p>

<pre><code><code>  http://redmine.open-bio.org/projects/bioperl/</code></code></pre>

<h1 id="AUTHOR---Shawn-Hoon">AUTHOR - Shawn Hoon</h1>

<p>Email: shawnh@fugu-sg.org</p>

<h1 id="APPENDIX">APPENDIX</h1>

<p>The rest of the documentation details each of the object methods. Internal methods are usually preceded with a _</p>

<h2 id="program_name">program_name</h2>

<pre><code><code> Title   : program_name
 Usage   : $factory&gt;program_name()
 Function: holds the program name
 Returns:  string
 Args    : None</code></code></pre>

<h2 id="program_dir">program_dir</h2>

<pre><code><code> Title   : program_dir
 Usage   : $factory-&gt;program_dir(@params)
 Function: returns the program directory, obtained from ENV variable.
 Returns:  string
 Args    :</code></code></pre>

<h2 id="version">version</h2>

<pre><code><code> Title   : version
 Usage   : exit if $prog-&gt;version() &lt; 1.8
 Function: Determine the version number of the program
 Example :
 Returns : float or undef
 Args    : none</code></code></pre>

<h2 id="run">run</h2>

<pre><code><code> <span class="variable">Title</span>   <span class="operator">:</span> <span class="variable">run</span> 
 <span class="variable">Usage</span>   <span class="operator">:</span> <span class="number">2</span> <span class="variable">sequence</span> <span class="variable">objects</span>
           <span class="variable">@fp</span> <span class="operator">=</span> <span class="variable">$factory</span><span class="operator">-&gt;</span><span class="variable">run</span><span class="operator">(</span><span class="variable">$seq1</span><span class="operator">,</span> <span class="variable">$seq2</span><span class="operator">);</span>
 <span class="variable">Function</span><span class="operator">:</span> <span class="variable">run</span> 
 <span class="variable">Returns</span> <span class="operator">:</span> <span class="variable">An</span> <span class="variable">array</span> <span class="variable">of</span> <span class="operator">&lt;</span><span class="variable">Bio::SeqFeature::FeaturePair</span><span class="operator">&gt;</span> 
 <span class="variable">Args</span>    <span class="operator">:</span> <span class="variable">Name</span> <span class="variable">of</span> <span class="variable">a</span> <span class="variable">file</span> <span class="variable">containing</span> <span class="variable">a</span> <span class="variable">set</span> <span class="variable">of</span> <span class="number">2</span> <span class="variable">fasta</span> <span class="variable">sequences</span> 
           <span class="keyword">or</span> <span class="keyword">else</span> <span class="number">2</span>  <span class="variable">Bio::Seq</span> <span class="variable">objects</span><span class="operator">.</span>
</code></code></pre>

<p>Throws an exception if argument is not either a string (eg a filename) or 2 Bio::Seq objects. If arguments are strings, throws exception if file corresponding to string name can not be found.</p>

<h2 id="run1">_run</h2>

<pre><code><code> Title   :  _run
 Usage   :  Internal function, not to be called directly
 Function:   makes actual system call to a promoterwise program
 Example :
 Returns : L&lt;Bio::SeqFeature::FeaturePair&gt;
 Args    : Name of a files containing 2 sequences in the order of peptide and genomic</code></code></pre>

<h2 id="setparams">_setparams</h2>

<pre><code><code> Title   :  _setparams
 Usage   :  Internal function, not to be called directly
 Function:  creates a string of params to be used in the command string
 Example :
 Returns :  string of params
 Args    :  </code></code></pre>

<h2 id="query_pep_seq">_query_pep_seq</h2>

<pre><code><code> Title   :  _query_pep_seq
 Usage   :  Internal function, not to be called directly
 Function:  get/set for the query sequence
 Example :
 Returns :  
 Args    :</code></code></pre>

<h2 id="subject_dna_seq">_subject_dna_seq</h2>

<pre><code><code> Title   :  _subject_dna_seq
 Usage   :  Internal function, not to be called directly
 Function:  get/set for the subject sequence
 Example :
 Returns :

 Args    :</code></code></pre>


</body>

</html>


