<?xml version="1.0" ?>
<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.0 Transitional//EN">
<html>
<head>
<!-- saved from url=(0017)http://localhost/ -->
<script language="JavaScript" src="../../../../../displayToc.js"></script>
<script language="JavaScript" src="../../../../../tocParas.js"></script>
<script language="JavaScript" src="../../../../../tocTab.js"></script>
<link rel="stylesheet" type="text/css" href="../../../../../scineplex.css">
<title></title>
<link rel="stylesheet" href="../../../../../Active.css" type="text/css" />
<meta http-equiv="content-type" content="text/html; charset=utf-8" />
<link rev="made" href="mailto:support@ActiveState.com" />
</head>

<body>



<ul id="index">
  <li><a href="#NAME">NAME</a></li>
  <li><a href="#SYNOPSIS">SYNOPSIS</a></li>
  <li><a href="#DESCRIPTION">DESCRIPTION</a></li>
  <li><a href="#FEEDBACK">FEEDBACK</a>
    <ul>
      <li><a href="#Mailing-Lists">Mailing Lists</a></li>
      <li><a href="#Support">Support</a></li>
      <li><a href="#Reporting-Bugs">Reporting Bugs</a></li>
    </ul>
  </li>
  <li><a href="#AUTHOR---Kiran">AUTHOR - Kiran</a></li>
  <li><a href="#APPENDIX">APPENDIX</a>
    <ul>
      <li><a href="#program_name">program_name</a></li>
      <li><a href="#program_dir">program_dir</a></li>
      <li><a href="#version">version</a></li>
      <li><a href="#predict_genes">predict_genes</a></li>
      <li><a href="#run">run</a></li>
      <li><a href="#run1">_run</a></li>
      <li><a href="#parse_results">_parse_results</a></li>
      <li><a href="#setinput-">_setinput()</a></li>
      <li><a href="#query_pep_seq-">_query_pep_seq()</a></li>
      <li><a href="#query_cdna_seq-">_query_cdna_seq()</a></li>
      <li><a href="#subject_dna_seq-">_subject_dna_seq()</a></li>
    </ul>
  </li>
</ul>

<h1 id="NAME">NAME</h1>

<p>Bio::Tools::Run::Pseudowise - Object for prediting pseudogenes in a given sequence given a protein and a cdna sequence</p>

<h1 id="SYNOPSIS">SYNOPSIS</h1>

<pre><code><code>  <span class="comment"># Build a pseudowise alignment factory</span>
  <span class="keyword">my</span> <span class="variable">$factory</span> <span class="operator">=</span> <span class="variable">Bio::Tools::Run::Pseudowise</span><span class="operator">-&gt;</span><span class="variable">new</span><span class="operator">();</span>
  
  <span class="comment"># Pass the factory 3 Bio:SeqI objects (in the order of query</span>
  <span class="comment"># peptide and cdna and target_genomic)</span>
  <span class="comment"># @genes is an array of GenericSeqFeature objects</span>
  <span class="keyword">my</span> <span class="variable">@genes</span> <span class="operator">=</span> <span class="variable">$factory</span><span class="operator">-&gt;</span><span class="variable">run</span><span class="operator">(</span><span class="variable">$seq1</span><span class="operator">,</span> <span class="variable">$seq2</span><span class="operator">,</span> <span class="variable">$seq3</span><span class="operator">);</span>
</code></code></pre>

<h1 id="DESCRIPTION">DESCRIPTION</h1>

<p>Pseudowise is a pseudogene predition program developed by Ewan Birney http://www.sanger.ac.uk/software/wise2.</p>

<h1 id="FEEDBACK">FEEDBACK</h1>

<h2 id="Mailing-Lists">Mailing Lists</h2>

<p>User feedback is an integral part of the evolution of this and other Bioperl modules. Send your comments and suggestions preferably to one of the Bioperl mailing lists. Your participation is much appreciated.</p>

<pre><code><code>  bioperl-l@bioperl.org                  - General discussion
  http://bioperl.org/wiki/Mailing_lists  - About the mailing lists</code></code></pre>

<h2 id="Support">Support</h2>

<p>Please direct usage questions or support issues to the mailing list:</p>

<p><i>bioperl-l@bioperl.org</i></p>

<p>rather than to the module maintainer directly. Many experienced and reponsive experts will be able look at the problem and quickly address it. Please include a thorough description of the problem with code and data examples if at all possible.</p>

<h2 id="Reporting-Bugs">Reporting Bugs</h2>

<p>Report bugs to the Bioperl bug tracking system to help us keep track the bugs and their resolution. Bug reports can be submitted via the web:</p>

<pre><code><code>  http://redmine.open-bio.org/projects/bioperl/</code></code></pre>

<h1 id="AUTHOR---Kiran">AUTHOR - Kiran</h1>

<p>Email kiran@fugu-sg.org</p>

<h1 id="APPENDIX">APPENDIX</h1>

<p>The rest of the documentation details each of the object methods. Internal methods are usually preceded with a _</p>

<h2 id="program_name">program_name</h2>

<pre><code><code> Title   : program_name
 Usage   : $factory&gt;program_name()
 Function: holds the program name
 Returns:  string
 Args    : None</code></code></pre>

<h2 id="program_dir">program_dir</h2>

<pre><code><code> Title   : program_dir
 Usage   : $factory-&gt;program_dir(@params)
 Function: returns the program directory, obtained from ENV variable.
 Returns:  string
 Args    :</code></code></pre>

<h2 id="version">version</h2>

<pre><code><code> Title   : version
 Usage   : exit if $prog-&gt;version() &lt; 1.8
 Function: Determine the version number of the program
 Example :
 Returns : float or undef
 Args    : none</code></code></pre>

<h2 id="predict_genes">predict_genes</h2>

<pre><code><code> Title   : predict_genes 
 Usage   : DEPRECATED. Use $factory-&gt;run instead 
 Function: Predict pseudogenes
 Returns : An array of Bio::Seqfeature::Generic objects 
 Args    : Name of a file containing a set of 3 fasta sequences in the order of 
           peptide, cdna and genomic sequences
           or else 3  Bio::Seq objects.</code></code></pre>

<p>Throws an exception if argument is not either a string (eg a filename) or 3 Bio::Seq objects. If arguments are strings, throws exception if file corresponding to string name can not be found.</p>

<h2 id="run">run</h2>

<pre><code><code> <span class="variable">Title</span>   <span class="operator">:</span> <span class="variable">run</span>
 <span class="variable">Usage</span>   <span class="operator">:</span> <span class="keyword">my</span> <span class="variable">@feats</span> <span class="operator">=</span> <span class="variable">$factory</span><span class="operator">-&gt;</span><span class="variable">run</span><span class="operator">(</span><span class="variable">$seq1</span><span class="operator">,</span> <span class="variable">$seq2</span><span class="operator">,</span> <span class="variable">$seq3</span><span class="operator">);</span>
 <span class="variable">Function</span><span class="operator">:</span> <span class="variable">Executes</span> <span class="variable">pseudogene</span> <span class="variable">binary</span>
 <span class="variable">Returns</span> <span class="operator">:</span> <span class="variable">An</span> <span class="variable">array</span> <span class="variable">of</span> <span class="variable">Bio::Seqfeature::Generic</span> <span class="variable">objects</span> 
 <span class="variable">Args</span>    <span class="operator">:</span> <span class="variable">Name</span> <span class="variable">of</span> <span class="variable">a</span> <span class="variable">file</span> <span class="variable">containing</span> <span class="variable">a</span> <span class="variable">set</span> <span class="variable">of</span> <span class="number">3</span> <span class="variable">fasta</span> <span class="variable">sequences</span> <span class="variable">in</span> <span class="variable">the</span> <span class="variable">order</span> <span class="variable">of</span> 
           <span class="variable">peptide</span><span class="operator">,</span> <span class="variable">cdna</span> <span class="keyword">and</span> <span class="variable">genomic</span> <span class="variable">sequences</span>
           <span class="keyword">or</span> <span class="keyword">else</span> <span class="number">3</span>  <span class="variable">Bio::Seq</span> <span class="variable">objects</span><span class="operator">.</span>
</code></code></pre>

<p>Throws an exception if argument is not either a string (eg a filename) or 3 Bio::Seq objects. If arguments are strings, throws exception if file corresponding to string name can not be found.</p>

<h2 id="run1">_run</h2>

<pre><code><code> <span class="variable">Title</span>   <span class="operator">:</span>  <span class="variable">_run</span>
 <span class="variable">Usage</span>   <span class="operator">:</span>  <span class="variable">Internal</span> <span class="variable">function</span><span class="operator">,</span> <span class="keyword">not</span> <span class="variable">to</span> <span class="variable">be</span> <span class="variable">called</span> <span class="variable">directly</span>
 <span class="variable">Function</span><span class="operator">:</span>   <span class="variable">makes</span> <span class="variable">actual</span> <span class="keyword">system</span> <span class="variable">call</span> <span class="variable">to</span> <span class="variable">a</span> <span class="variable">pseudowise</span> <span class="variable">program</span>
 <span class="variable">Example</span> <span class="operator">:</span>
 <span class="variable">Returns</span> <span class="operator">:</span> <span class="variable">nothing</span><span class="operator">;</span> <span class="variable">pseudowise</span>  <span class="variable">output</span> <span class="variable">is</span> <span class="variable">written</span> <span class="variable">to</span> <span class="variable">a</span>
           <span class="variable">temporary</span> <span class="variable">file</span> <span class="variable">$TMPOUTFILE</span>
 <span class="variable">Args</span>    <span class="operator">:</span> <span class="variable">Name</span> <span class="variable">of</span> <span class="variable">a</span> <span class="variable">files</span> <span class="variable">containing</span> <span class="number">3</span> <span class="variable">sequences</span> <span class="variable">in</span> <span class="variable">the</span> <span class="variable">order</span> <span class="variable">of</span> <span class="variable">peptide</span><span class="operator">,</span> <span class="variable">cdna</span> <span class="keyword">and</span> <span class="variable">genomic</span> 
</code></code></pre>

<h2 id="parse_results">_parse_results</h2>

<pre><code><code> Title   :  __parse_results
 Usage   :  Internal function, not to be called directly
 Function:  Parses pseudowise output 
 Example :
 Returns : an reference to an array of Seqfeatures 
 Args    : the name of the output file </code></code></pre>

<h2 id="setinput-">_setinput()</h2>

<pre><code><code> Title   :  _setinput
 Usage   :  Internal function, not to be called directly
 Function:   Create input files for pseudowise program
 Example :
 Returns : name of file containing dba data input
 Args    : Seq objects in the order of query protein and cdna and target genomic sequence </code></code></pre>

<h2 id="query_pep_seq-">_query_pep_seq()</h2>

<pre><code><code> Title   :  _query_pep_seq
 Usage   :  Internal function, not to be called directly
 Function:  get/set for the query sequence 
 Example :
 Returns : 
 Args    : </code></code></pre>

<h2 id="query_cdna_seq-">_query_cdna_seq()</h2>

<pre><code><code> Title   :  _query_cdna_seq
 Usage   :  Internal function, not to be called directly
 Function:  get/set for the query sequence
 Example :
 Returns :
 Args    :</code></code></pre>

<h2 id="subject_dna_seq-">_subject_dna_seq()</h2>

<pre><code><code> Title   :  _subject_dna_seq
 Usage   :  Internal function, not to be called directly
 Function:  get/set for the subject sequence
 Example :
 Returns :

 Args    :</code></code></pre>


</body>

</html>


