<?xml version="1.0" ?>
<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.0 Transitional//EN">
<html>
<head>
<!-- saved from url=(0017)http://localhost/ -->
<script language="JavaScript" src="../../../../../displayToc.js"></script>
<script language="JavaScript" src="../../../../../tocParas.js"></script>
<script language="JavaScript" src="../../../../../tocTab.js"></script>
<link rel="stylesheet" type="text/css" href="../../../../../scineplex.css">
<title></title>
<link rel="stylesheet" href="../../../../../Active.css" type="text/css" />
<meta http-equiv="content-type" content="text/html; charset=utf-8" />
<link rev="made" href="mailto:support@ActiveState.com" />
</head>

<body>



<ul id="index">
  <li><a href="#NAME">NAME</a></li>
  <li><a href="#SYNOPSIS">SYNOPSIS</a></li>
  <li><a href="#DESCRIPTION">DESCRIPTION</a></li>
  <li><a href="#FEEDBACK">FEEDBACK</a>
    <ul>
      <li><a href="#Mailing-Lists">Mailing Lists</a></li>
      <li><a href="#Support">Support</a></li>
      <li><a href="#Reporting-Bugs">Reporting Bugs</a></li>
    </ul>
  </li>
  <li><a href="#AUTHOR">AUTHOR</a></li>
  <li><a href="#APPENDIX">APPENDIX</a>
    <ul>
      <li><a href="#new">new</a></li>
      <li><a href="#java">java</a></li>
      <li><a href="#run">run</a></li>
      <li><a href="#setinput">_setinput</a></li>
      <li><a href="#outfile">outfile</a></li>
      <li><a href="#min_perc_id">min_perc_id</a></li>
      <li><a href="#quiet">quiet</a></li>
      <li><a href="#verbose">verbose</a></li>
      <li><a href="#annotation_format">annotation_format</a></li>
      <li><a href="#region_file">region_file</a></li>
      <li><a href="#score_file">score_file</a></li>
      <li><a href="#alignment_file">alignment_file</a></li>
      <li><a href="#contigs_file">contigs_file</a></li>
      <li><a href="#diffs">diffs</a></li>
      <li><a href="#plotfile">plotfile</a></li>
      <li><a href="#min_length">min_length</a></li>
      <li><a href="#plotmin">plotmin</a></li>
      <li><a href="#annotation">annotation</a></li>
      <li><a href="#bases">bases</a></li>
      <li><a href="#tickdist">tickdist</a></li>
      <li><a href="#resolution">resolution</a></li>
      <li><a href="#title">title</a></li>
      <li><a href="#window">window</a></li>
      <li><a href="#numwindows">numwindows</a></li>
      <li><a href="#start">start</a></li>
      <li><a href="#end">end</a></li>
      <li><a href="#num_plot_lines">num_plot_lines</a></li>
      <li><a href="#legend">legend</a></li>
      <li><a href="#filename">filename</a></li>
      <li><a href="#axis_label">axis_label</a></li>
      <li><a href="#ticks_file">ticks_file</a></li>
      <li><a href="#color">color</a></li>
      <li><a href="#use_order">use_order</a></li>
      <li><a href="#gaps">gaps</a></li>
      <li><a href="#snps_file">snps_file</a></li>
      <li><a href="#repeats_file">repeats_file</a></li>
      <li><a href="#filter_repeats">filter_repeats</a></li>
    </ul>
  </li>
</ul>

<h1 id="NAME">NAME</h1>

<p>Bio::Tools::Run::Vista</p>

<p>Wrapper for Vista</p>

<h1 id="SYNOPSIS">SYNOPSIS</h1>

<pre><code><code>  <span class="keyword">use</span> <span class="variable">Bio::Tools::Run::Vista</span><span class="operator">;</span>
  <span class="keyword">use</span> <span class="variable">Bio::Tools::Run::Alignment::Lagan</span><span class="operator">;</span>
  <span class="keyword">use</span> <span class="variable">Bio::AlignIO</span><span class="operator">;</span>
  
  <span class="keyword">my</span> <span class="variable">$sio</span> <span class="operator">=</span> <span class="variable">Bio::SeqIO</span><span class="operator">-&gt;</span><span class="variable">new</span><span class="operator">(</span><span class="string">-file</span><span class="operator">=&gt;</span><span class="variable">$ARGV</span><span class="operator">[</span><span class="number">0</span><span class="operator">]</span><span class="operator">,</span><span class="string">-format</span><span class="operator">=&gt;</span><span class="string">'genbank'</span><span class="operator">);</span>
  <span class="keyword">my</span> <span class="variable">@seq</span><span class="operator">;</span>
  <span class="keyword">my</span> <span class="variable">$reference</span> <span class="operator">=</span> <span class="variable">$sio</span><span class="operator">-&gt;</span><span class="variable">next_seq</span><span class="operator">;</span>
  <span class="keyword">push</span> <span class="variable">@seq</span><span class="operator">,</span> <span class="variable">$reference</span><span class="operator">;</span>
  <span class="keyword">while</span><span class="operator">(</span><span class="keyword">my</span> <span class="variable">$seq</span> <span class="operator">=</span> <span class="variable">$sio</span><span class="operator">-&gt;</span><span class="variable">next_seq</span><span class="operator">){</span>
    <span class="keyword">push</span> <span class="variable">@seq</span><span class="operator">,</span><span class="variable">$seq</span><span class="operator">;</span>
  <span class="operator">}</span>
  <span class="keyword">my</span> <span class="variable">@features</span> <span class="operator">=</span> <span class="keyword">grep</span><span class="operator">{</span><span class="variable">$_</span><span class="operator">-&gt;</span><span class="variable">primary_tag</span> <span class="keyword">eq</span> <span class="string">'CDS'</span><span class="operator">}</span> <span class="variable">$reference</span><span class="operator">-&gt;</span><span class="variable">get_SeqFeatures</span><span class="operator">;</span>
  
  <span class="keyword">my</span> <span class="variable">$lagan</span> <span class="operator">=</span> <span class="variable">Bio::Tools::Run::Alignment::Lagan</span><span class="operator">-&gt;</span><span class="variable">new</span><span class="operator">;</span>
  
  <span class="keyword">my</span> <span class="variable">$aln</span> <span class="operator">=</span> <span class="variable">$lagan</span><span class="operator">-&gt;</span><span class="variable">mlagan</span><span class="operator">(\</span><span class="variable">@seq</span><span class="operator">,</span><span class="string">'(fugu (mouse human))'</span><span class="operator">);</span>
  
  
  <span class="keyword">my</span> <span class="variable">$vis</span> <span class="operator">=</span> <span class="variable">Bio::Tools::Run::Vista</span><span class="operator">-&gt;</span><span class="variable">new</span><span class="operator">(</span><span class="string">'outfile'</span><span class="operator">=&gt;</span><span class="string">"outfile.pdf"</span><span class="operator">,</span>
                                        <span class="string">'title'</span> <span class="operator">=&gt;</span> <span class="string">"My Vista Plot"</span><span class="operator">,</span>
                                        <span class="string">'annotation'</span><span class="operator">=&gt;\</span><span class="variable">@features</span><span class="operator">,</span>
                                        <span class="string">'annotation_format'</span><span class="operator">=&gt;</span><span class="string">'GFF'</span><span class="operator">,</span>
                                        <span class="string">'min_perc_id'</span><span class="operator">=&gt;</span><span class="number">75</span><span class="operator">,</span>
                                        <span class="string">'min_length'</span><span class="operator">=&gt;</span><span class="number">100</span><span class="operator">,</span>
                                        <span class="string">'plotmin'</span>   <span class="operator">=&gt;</span> <span class="number">50</span><span class="operator">,</span>
                                        <span class="string">'tickdist'</span> <span class="operator">=&gt;</span> <span class="number">2000</span><span class="operator">,</span>
                                        <span class="string">'window'</span><span class="operator">=&gt;</span><span class="number">40</span><span class="operator">,</span>
                                        <span class="string">'numwindows'</span><span class="operator">=&gt;</span><span class="number">4</span><span class="operator">,</span>
                                        <span class="string">'start'</span><span class="operator">=&gt;</span><span class="number">50</span><span class="operator">,</span>
                                        <span class="string">'end'</span><span class="operator">=&gt;</span><span class="number">1500</span><span class="operator">,</span>
                                        <span class="string">'tickdist'</span><span class="operator">=&gt;</span><span class="number">100</span><span class="operator">,</span>
                                        <span class="string">'bases'</span><span class="operator">=&gt;</span><span class="number">1000</span><span class="operator">,</span>
                                        <span class="string">'java_param'</span><span class="operator">=&gt;</span><span class="string">"-Xmx128m"</span><span class="operator">,</span>
                                        <span class="string">'num_pages'</span><span class="operator">=&gt;</span><span class="number">1</span><span class="operator">,</span>
                                        <span class="string">'color'</span><span class="operator">=&gt;</span> <span class="operator">{</span><span class="string">'EXON'</span><span class="operator">=&gt;</span><span class="string">'100 0 0'</span><span class="operator">,</span>
                                                   <span class="string">'CNS'</span><span class="operator">=&gt;</span><span class="string">'0 0 100'</span><span class="operator">}</span><span class="operator">,</span>
                                        <span class="string">'quiet'</span><span class="operator">=&gt;</span><span class="number">1</span><span class="operator">);</span>
  
  <span class="keyword">my</span> <span class="variable">$referenceid</span><span class="operator">=</span> <span class="string">'human'</span><span class="operator">;</span>
  <span class="variable">$vis</span><span class="operator">-&gt;</span><span class="variable">run</span><span class="operator">(</span><span class="variable">$aln</span><span class="operator">,</span><span class="variable">$referenceid</span><span class="operator">);</span> 
  
  <span class="comment">#alternative one can choose pairwise alignments to plot</span>
  <span class="comment">#where the second id in each pair is the reference sequence</span>
  <span class="variable">$vis</span><span class="operator">-&gt;</span><span class="variable">run</span><span class="operator">(</span><span class="variable">$aln</span><span class="operator">,(</span><span class="operator">[</span><span class="variable">mouse</span><span class="operator">,</span><span class="variable">human</span><span class="operator">]</span><span class="operator">,</span><span class="operator">[</span><span class="variable">fugu</span><span class="operator">,</span><span class="variable">human</span><span class="operator">]</span><span class="operator">,</span><span class="operator">[</span><span class="variable">mouse</span><span class="operator">,</span><span class="variable">fugu</span><span class="operator">]</span><span class="operator">));</span>
</code></code></pre>

<h1 id="DESCRIPTION">DESCRIPTION</h1>

<p>Pls see Vista documentation for plotfile options</p>

<p>Wrapper for Vista :</p>

<p>C. Mayor, M. Brudno, J. R. Schwartz, A. Poliakov, E. M. Rubin, K. A. Frazer, L. S. Pachter, I. Dubchak. VISTA: Visualizing global DNA sequence alignments of arbitrary length. Bioinformatics, 2000 Nov;16(11):1046-1047. Get it here: http://www-gsd.lbl.gov/vista/VISTAdownload2.html</p>

<p>On the command line, it is assumed that this can be executed:</p>

<p>java Vista plotfile</p>

<p>Some of the code was adapted from MLAGAN toolkit</p>

<p>M. Brudno, C.B. Do, G. Cooper, M.F. Kim, E. Davydov, NISC Sequencing Consortium, E.D. Green, A. Sidow and S. Batzoglou LAGAN and Multi-LAGAN: Efficient Tools for Large-Scale Multiple Alignment of Genomic DNA, Genome Research, in press</p>

<p>get lagan here:</p>

<p>http://lagan.stanford.edu/</p>

<h1 id="FEEDBACK">FEEDBACK</h1>

<h2 id="Mailing-Lists">Mailing Lists</h2>

<p>User feedback is an integral part of the evolution of this and other Bioperl modules. Send your comments and suggestions preferably to one of the Bioperl mailing lists. Your participation is much appreciated.</p>

<pre><code><code>  bioperl-l@bioperl.org                  - General discussion
  http://bioperl.org/wiki/Mailing_lists  - About the mailing lists</code></code></pre>

<h2 id="Support">Support</h2>

<p>Please direct usage questions or support issues to the mailing list:</p>

<p><i>bioperl-l@bioperl.org</i></p>

<p>rather than to the module maintainer directly. Many experienced and reponsive experts will be able look at the problem and quickly address it. Please include a thorough description of the problem with code and data examples if at all possible.</p>

<h2 id="Reporting-Bugs">Reporting Bugs</h2>

<p>Report bugs to the Bioperl bug tracking system to help us keep track the bugs and their resolution. Bug reports can be submitted via the web:</p>

<pre><code><code>  http://redmine.open-bio.org/projects/bioperl/</code></code></pre>

<h1 id="AUTHOR">AUTHOR</h1>

<p>Shawn Hoon Email shawnh@fugu-sg.org</p>

<h1 id="APPENDIX">APPENDIX</h1>

<p>The rest of the documentation details each of the object methods. Internal methods are usually preceded with a _</p>

<h2 id="new">new</h2>

<pre><code><code>    <span class="variable">Title</span>   <span class="operator">:</span>   <span class="variable">new</span>
    <span class="variable">Usage</span>   <span class="operator">:</span>   <span class="keyword">my</span> <span class="variable">$vis</span> <span class="operator">=</span> <span class="variable">Bio::Tools::Run::Vista</span><span class="operator">-&gt;</span><span class="variable">new</span><span class="operator">(</span><span class="string">'outfile'</span><span class="operator">=&gt;</span><span class="variable">$out</span><span class="operator">,</span>
                                        <span class="string">'title'</span> <span class="operator">=&gt;</span> <span class="string">"My Vista Plot"</span><span class="operator">,</span>
                                        <span class="string">'annotation'</span><span class="operator">=&gt;\</span><span class="variable">@features</span><span class="operator">,</span>
                                        <span class="string">'annotation_format'</span><span class="operator">=&gt;</span><span class="string">'GFF'</span><span class="operator">,</span>
                                        <span class="string">'min_perc_id'</span><span class="operator">=&gt;</span><span class="number">75</span><span class="operator">,</span>
                                        <span class="string">'min_length'</span><span class="operator">=&gt;</span><span class="number">100</span><span class="operator">,</span>
                                        <span class="string">'plotmin'</span>   <span class="operator">=&gt;</span> <span class="number">50</span><span class="operator">,</span>
                                        <span class="string">'tickdist'</span> <span class="operator">=&gt;</span> <span class="number">2000</span><span class="operator">,</span>
                                        <span class="string">'window'</span><span class="operator">=&gt;</span><span class="number">40</span><span class="operator">,</span>
                                        <span class="string">'numwindows'</span><span class="operator">=&gt;</span><span class="number">4</span><span class="operator">,</span>
                                        <span class="string">'start'</span><span class="operator">=&gt;</span><span class="number">50</span><span class="operator">,</span>
                                        <span class="string">'end'</span><span class="operator">=&gt;</span><span class="number">1500</span><span class="operator">,</span>
                                        <span class="string">'tickdist'</span><span class="operator">=&gt;</span><span class="number">100</span><span class="operator">,</span>
                                        <span class="string">'bases'</span><span class="operator">=&gt;</span><span class="number">1000</span><span class="operator">,</span>
                                        <span class="string">'color'</span><span class="operator">=&gt;</span> <span class="operator">{</span><span class="string">'EXON'</span><span class="operator">=&gt;</span><span class="string">'100 0 0'</span><span class="operator">,</span>
                                                   <span class="string">'CNS'</span><span class="operator">=&gt;</span><span class="string">'0 0 100'</span><span class="operator">}</span><span class="operator">,</span>
                                        <span class="string">'quiet'</span><span class="operator">=&gt;</span><span class="number">1</span><span class="operator">);</span>
    <span class="variable">Function</span><span class="operator">:</span>   <span class="variable">Construtor</span> <span class="keyword">for</span> <span class="variable">Vista</span> <span class="variable">wrapper</span>
    <span class="variable">Args</span>    <span class="operator">:</span>   <span class="variable">outfile</span> <span class="operator">-</span> <span class="variable">location</span> <span class="variable">of</span> <span class="variable">the</span> <span class="variable">pdf</span> <span class="variable">generated</span>
                <span class="variable">annotation</span> <span class="operator">-</span> <span class="variable">either</span> <span class="variable">a</span> <span class="variable">file</span> <span class="keyword">or</span> <span class="keyword">and</span> <span class="variable">array</span> <span class="keyword">ref</span> <span class="variable">of</span> <span class="variable">Bio::SeqFeatureI</span>
                             <span class="variable">indicating</span> <span class="variable">the</span> <span class="variable">exons</span> 
                <span class="variable">regmin</span>     <span class="operator">-</span><span class="variable">region</span> <span class="variable">min</span>
</code></code></pre>

<h2 id="java">java</h2>

<pre><code><code>    <span class="variable">Title</span>   <span class="operator">:</span>   <span class="variable">java</span>
    <span class="variable">Usage</span>   <span class="operator">:</span>   <span class="variable">$obj</span><span class="operator">-&gt;</span><span class="variable">java</span><span class="operator">(</span><span class="string">'/usr/opt/java130/bin/java'</span><span class="operator">);</span>
    <span class="variable">Function</span><span class="operator">:</span>   <span class="variable">Get</span><span class="operator">/</span><span class="variable">set</span> <span class="variable">method</span> <span class="keyword">for</span> <span class="variable">the</span> <span class="variable">location</span> <span class="variable">of</span> <span class="variable">java</span> <span class="variable">VM</span>
    <span class="variable">Args</span>    <span class="operator">:</span>   <span class="variable">File</span> <span class="variable">path</span> <span class="operator">(</span><span class="variable">optional</span><span class="operator">)</span>
</code></code></pre>

<h2 id="run">run</h2>

<pre><code><code> Title   : run
 Usage   : my @genes = $self-&gt;run($seq)
 Function: runs Vista 
 Returns : A boolean 1 if no errors 
 Args    : Argument 1: Bio::Align::Align required
           Argument 2: a string or number, which is the sequence id of the
                       reference sequence or the rank of the sequence
                       in the alignment</code></code></pre>

<h2 id="setinput">_setinput</h2>

<pre><code><code> Title   : _setinput
 Usage   : Internal function, not to be called directly
 Function: writes input sequence to file and return the file name
 Example :
 Returns : string
 Args    :</code></code></pre>

<h2 id="outfile">outfile</h2>

<pre><code><code>  Title    : outfile
  Usage    : $obj-&gt;outfile
  Function : Get/Set method outfile
  Args     : </code></code></pre>

<h2 id="min_perc_id">min_perc_id</h2>

<pre><code><code>  Title    : min_perc_id
  Usage    : $obj-&gt;min_perc_id
  Function : Get/Set method min_perc_id
  Args     : </code></code></pre>

<h2 id="quiet">quiet</h2>

<pre><code><code>  Title    : quiet
  Usage    : $obj-&gt;quiet
  Function : Get/Set method quiet
  Args     : </code></code></pre>

<h2 id="verbose">verbose</h2>

<pre><code><code>  Title    : verbose
  Usage    : $obj-&gt;verbose
  Function : Get/Set method verbose
  Args     : </code></code></pre>

<h2 id="annotation_format">annotation_format</h2>

<pre><code><code>  Title    : annotation_format
  Usage    : $obj-&gt;annotation_format
  Function : Get/Set method annotation_format
  Args     : </code></code></pre>

<h2 id="region_file">region_file</h2>

<pre><code><code>  Title    : region_file
  Usage    : $obj-&gt;region_file
  Function : Get/Set method region_file
  Args     : </code></code></pre>

<h2 id="score_file">score_file</h2>

<pre><code><code>  Title    : score_file
  Usage    : $obj-&gt;score_file
  Function : Get/Set method score_file
  Args     : </code></code></pre>

<h2 id="alignment_file">alignment_file</h2>

<pre><code><code>  Title    : alignment_file
  Usage    : $obj-&gt;alignment_file
  Function : Get/Set method alignment_file
  Args     : </code></code></pre>

<h2 id="contigs_file">contigs_file</h2>

<pre><code><code>  Title    : contigs_file
  Usage    : $obj-&gt;contigs_file
  Function : Get/Set method contigs_file
  Args     : </code></code></pre>

<h2 id="diffs">diffs</h2>

<pre><code><code>  Title    : diffs
  Usage    : $obj-&gt;diffs
  Function : Get/Set method diffs
  Args     : </code></code></pre>

<h2 id="plotfile">plotfile</h2>

<pre><code><code>  Title    : plotfile
  Usage    : $obj-&gt;plotfile
  Function : Get/Set method plotfile
  Args     : </code></code></pre>

<h2 id="min_length">min_length</h2>

<pre><code><code>  Title    : min_length
  Usage    : $obj-&gt;min_length
  Function : Get/Set method min_length
  Args     : </code></code></pre>

<h2 id="plotmin">plotmin</h2>

<pre><code><code>  Title    : plotmin
  Usage    : $obj-&gt;plotmin
  Function : Get/Set method plotmin
  Args     : </code></code></pre>

<h2 id="annotation">annotation</h2>

<pre><code><code>  Title    : annotation
  Usage    : $obj-&gt;annotation
  Function : Get/Set method annotation
  Args     : </code></code></pre>

<h2 id="bases">bases</h2>

<pre><code><code>  Title    : bases
  Usage    : $obj-&gt;bases
  Function : Get/Set method bases
  Args     : </code></code></pre>

<h2 id="tickdist">tickdist</h2>

<pre><code><code>  Title    : tickdist
  Usage    : $obj-&gt;tickdist
  Function : Get/Set method tickdist
  Args     : </code></code></pre>

<h2 id="resolution">resolution</h2>

<pre><code><code>  Title    : resolution
  Usage    : $obj-&gt;resolution
  Function : Get/Set method resolution
  Args     : </code></code></pre>

<h2 id="title">title</h2>

<pre><code><code>  Title    : title
  Usage    : $obj-&gt;title
  Function : Get/Set method title
  Args     : </code></code></pre>

<h2 id="window">window</h2>

<pre><code><code>  Title    : window
  Usage    : $obj-&gt;window
  Function : Get/Set method window
  Args     : </code></code></pre>

<h2 id="numwindows">numwindows</h2>

<pre><code><code>  Title    : numwindows
  Usage    : $obj-&gt;numwindows
  Function : Get/Set method numwindows
  Args     : </code></code></pre>

<h2 id="start">start</h2>

<pre><code><code>  Title    : start
  Usage    : $obj-&gt;start
  Function : Get/Set method start
  Args     : </code></code></pre>

<h2 id="end">end</h2>

<pre><code><code>  Title    : end
  Usage    : $obj-&gt;end
  Function : Get/Set method end
  Args     : </code></code></pre>

<h2 id="num_plot_lines">num_plot_lines</h2>

<pre><code><code>  Title    : num_plot_lines
  Usage    : $obj-&gt;num_plot_lines
  Function : Get/Set method num_plot_lines
  Args     : </code></code></pre>

<h2 id="legend">legend</h2>

<pre><code><code>  Title    : legend
  Usage    : $obj-&gt;legend
  Function : Get/Set method legend
  Args     : </code></code></pre>

<h2 id="filename">filename</h2>

<pre><code><code>  Title    : filename
  Usage    : $obj-&gt;filename
  Function : Get/Set method filename
  Args     : </code></code></pre>

<h2 id="axis_label">axis_label</h2>

<pre><code><code>  Title    : axis_label
  Usage    : $obj-&gt;axis_label
  Function : Get/Set method axis_label
  Args     : </code></code></pre>

<h2 id="ticks_file">ticks_file</h2>

<pre><code><code>  Title    : ticks_file
  Usage    : $obj-&gt;ticks_file
  Function : Get/Set method ticks_file
  Args     : </code></code></pre>

<h2 id="color">color</h2>

<pre><code><code>  Title    : color
  Usage    : $obj-&gt;color
  Function : Get/Set method color
  Args     : </code></code></pre>

<h2 id="use_order">use_order</h2>

<pre><code><code>  Title    : use_order
  Usage    : $obj-&gt;use_order
  Function : Get/Set method use_order
  Args     : </code></code></pre>

<h2 id="gaps">gaps</h2>

<pre><code><code>  Title    : gaps
  Usage    : $obj-&gt;gaps
  Function : Get/Set method gaps
  Args     : </code></code></pre>

<h2 id="snps_file">snps_file</h2>

<pre><code><code>  Title    : snps_file
  Usage    : $obj-&gt;snps_file
  Function : Get/Set method snps_file
  Args     : </code></code></pre>

<h2 id="repeats_file">repeats_file</h2>

<pre><code><code>  Title    : repeats_file
  Usage    : $obj-&gt;repeats_file
  Function : Get/Set method repeats_file
  Args     : </code></code></pre>

<h2 id="filter_repeats">filter_repeats</h2>

<pre><code><code>  Title    : filter_repeats
  Usage    : $obj-&gt;filter_repeats
  Function : Get/Set method filter_repeats
  Args     : </code></code></pre>


</body>

</html>


