<?xml version="1.0" ?>
<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.0 Transitional//EN">
<html>
<head>
<!-- saved from url=(0017)http://localhost/ -->
<script language="JavaScript" src="../../../../displayToc.js"></script>
<script language="JavaScript" src="../../../../tocParas.js"></script>
<script language="JavaScript" src="../../../../tocTab.js"></script>
<link rel="stylesheet" type="text/css" href="../../../../scineplex.css">
<title></title>
<link rel="stylesheet" href="../../../../Active.css" type="text/css" />
<meta http-equiv="content-type" content="text/html; charset=utf-8" />
<link rev="made" href="mailto:support@ActiveState.com" />
</head>

<body>



<ul id="index">
  <li><a href="#NAME">NAME</a></li>
  <li><a href="#SYNOPSIS">SYNOPSIS</a></li>
  <li><a href="#DESCRIPTION">DESCRIPTION</a></li>
  <li><a href="#FEEDBACK">FEEDBACK</a>
    <ul>
      <li><a href="#Mailing-Lists">Mailing Lists</a></li>
      <li><a href="#Support">Support</a></li>
      <li><a href="#Reporting-Bugs">Reporting Bugs</a></li>
    </ul>
  </li>
  <li><a href="#AUTHOR---Heikki-Lehvaslaiho">AUTHOR - Heikki Lehvaslaiho</a></li>
  <li><a href="#APPENDIX">APPENDIX</a>
    <ul>
      <li><a href="#RNAChange">RNAChange</a></li>
      <li><a href="#label">label</a></li>
      <li><a href="#similarity_score">similarity_score</a></li>
      <li><a href="#trivname">trivname</a></li>
    </ul>
  </li>
</ul>

<h1 id="NAME">NAME</h1>

<p>Bio::Variation::AAChange - Sequence change class for polypeptides</p>

<h1 id="SYNOPSIS">SYNOPSIS</h1>

<pre><code><code>   <span class="variable">$aamut</span> <span class="operator">=</span> <span class="variable">Bio::Variation::AAChange</span><span class="operator">-&gt;</span><span class="variable">new</span>
       <span class="operator">(</span><span class="string">'-start'</span>         <span class="operator">=&gt;</span> <span class="variable">$start</span><span class="operator">,</span>
        <span class="string">'-end'</span>           <span class="operator">=&gt;</span> <span class="variable">$end</span><span class="operator">,</span>
        <span class="string">'-length'</span>        <span class="operator">=&gt;</span> <span class="variable">$len</span><span class="operator">,</span>
        <span class="string">'-proof'</span>         <span class="operator">=&gt;</span> <span class="variable">$proof</span><span class="operator">,</span>
        <span class="string">'-isMutation'</span>    <span class="operator">=&gt;</span> <span class="number">1</span><span class="operator">,</span>
        <span class="string">'-mut_number'</span>    <span class="operator">=&gt;</span> <span class="variable">$mut_number</span>
        <span class="operator">);</span>
   
   <span class="keyword">my</span> <span class="variable">$a1</span> <span class="operator">=</span> <span class="variable">Bio::Variation::Allele</span><span class="operator">-&gt;</span><span class="variable">new</span><span class="operator">;</span>
   <span class="variable">$a1</span><span class="operator">-&gt;</span><span class="variable">seq</span><span class="operator">(</span><span class="variable">$ori</span><span class="operator">)</span> <span class="keyword">if</span> <span class="variable">$ori</span><span class="operator">;</span>
   <span class="variable">$aamut</span><span class="operator">-&gt;</span><span class="variable">allele_ori</span><span class="operator">(</span><span class="variable">$a1</span><span class="operator">);</span>
   <span class="keyword">my</span> <span class="variable">$a2</span> <span class="operator">=</span> <span class="variable">Bio::Variation::Allele</span><span class="operator">-&gt;</span><span class="variable">new</span><span class="operator">;</span>
   <span class="variable">$a2</span><span class="operator">-&gt;</span><span class="variable">seq</span><span class="operator">(</span><span class="variable">$mut</span><span class="operator">)</span> <span class="keyword">if</span> <span class="variable">$mut</span><span class="operator">;</span>
   <span class="variable">$aachange</span><span class="operator">-&gt;</span><span class="variable">add_Allele</span><span class="operator">(</span><span class="variable">$a2</span><span class="operator">);</span>
   <span class="variable">$aachange</span><span class="operator">-&gt;</span><span class="variable">allele_mut</span><span class="operator">(</span><span class="variable">$a2</span><span class="operator">);</span>
   
   <span class="keyword">print</span>  <span class="string">"\n"</span><span class="operator">;</span>
   
   <span class="comment"># add it to a SeqDiff container object</span>
   <span class="variable">$seqdiff</span><span class="operator">-&gt;</span><span class="variable">add_Variant</span><span class="operator">(</span><span class="variable">$rnachange</span><span class="operator">);</span>
   
   <span class="comment"># and create links to and from RNA level variant objects</span>
   <span class="variable">$aamut</span><span class="operator">-&gt;</span><span class="variable">RNAChange</span><span class="operator">(</span><span class="variable">$rnachange</span><span class="operator">);</span>
   <span class="variable">$rnachange</span><span class="operator">-&gt;</span><span class="variable">AAChange</span><span class="operator">(</span><span class="variable">$rnachange</span><span class="operator">);</span>
</code></code></pre>

<h1 id="DESCRIPTION">DESCRIPTION</h1>

<p>The instantiable class Bio::Variation::RNAChange describes basic sequence changes at polypeptide level. It uses methods defined in superclass Bio::Variation::VariantI, see <a href="../../../../../.cpan/build/BioPerl-1.6.901-ghKY8U/blib/lib/Bio/Variation/VariantI.html">Bio::Variation::VariantI</a> for details.</p>

<p>If the variation described by a AAChange object has a known Bio::Variation::RNAAChange object, create the link with method AAChange(). See <a href="../../../../../.cpan/build/BioPerl-1.6.901-ghKY8U/blib/lib/Bio/Variation/AAChange.html">Bio::Variation::AAChange</a> for more information.</p>

<h1 id="FEEDBACK">FEEDBACK</h1>

<h2 id="Mailing-Lists">Mailing Lists</h2>

<p>User feedback is an integral part of the evolution of this and other Bioperl modules. Send your comments and suggestions preferably to the Bioperl mailing lists Your participation is much appreciated.</p>

<pre><code><code>  bioperl-l@bioperl.org                  - General discussion
  http://bioperl.org/wiki/Mailing_lists  - About the mailing lists</code></code></pre>

<h2 id="Support">Support</h2>

<p>Please direct usage questions or support issues to the mailing list:</p>

<p><i>bioperl-l@bioperl.org</i></p>

<p>rather than to the module maintainer directly. Many experienced and reponsive experts will be able look at the problem and quickly address it. Please include a thorough description of the problem with code and data examples if at all possible.</p>

<h2 id="Reporting-Bugs">Reporting Bugs</h2>

<p>Report bugs to the Bioperl bug tracking system to help us keep track the bugs and their resolution. Bug reports can be submitted via the web:</p>

<pre><code><code>  https://redmine.open-bio.org/projects/bioperl/</code></code></pre>

<h1 id="AUTHOR---Heikki-Lehvaslaiho">AUTHOR - Heikki Lehvaslaiho</h1>

<p>Email: heikki-at-bioperl-dot-org</p>

<h1 id="APPENDIX">APPENDIX</h1>

<p>The rest of the documentation details each of the object methods. Internal methods are usually preceded with a _</p>

<h2 id="RNAChange">RNAChange</h2>

<pre><code><code> <span class="variable">Title</span>   <span class="operator">:</span> <span class="variable">RNAChange</span>
 <span class="variable">Usage</span>   <span class="operator">:</span> <span class="variable">$mutobj</span> <span class="operator">=</span> <span class="variable">$self</span><span class="operator">-&gt;</span><span class="variable">RNAChange</span><span class="operator">;</span>
         <span class="operator">:</span> <span class="variable">$mutobj</span> <span class="operator">=</span> <span class="variable">$self</span><span class="operator">-&gt;</span><span class="variable">RNAChange</span><span class="operator">(</span><span class="variable">$objref</span><span class="operator">);</span>
 <span class="variable">Function</span><span class="operator">:</span> <span class="variable">Returns</span> <span class="keyword">or</span> <span class="variable">sets</span> <span class="variable">the</span> <span class="keyword">link</span><span class="operator">-</span><span class="variable">reference</span> <span class="variable">to</span> <span class="variable">a</span> <span class="variable">mutation</span><span class="operator">/</span><span class="variable">change</span> <span class="variable">object</span><span class="operator">.</span>
           <span class="variable">If</span> <span class="variable">there</span> <span class="variable">is</span> <span class="keyword">no</span> <span class="keyword">link</span><span class="operator">,</span> <span class="variable">it</span> <span class="variable">will</span> <span class="keyword">return</span> <span class="keyword">undef</span>
 <span class="variable">Returns</span> <span class="operator">:</span> <span class="variable">an</span> <span class="variable">obj_ref</span> <span class="keyword">or</span> <span class="keyword">undef</span>
</code></code></pre>

<h2 id="label">label</h2>

<pre><code><code> <span class="variable">Title</span>   <span class="operator">:</span> <span class="variable">label</span>
 <span class="variable">Usage</span>   <span class="operator">:</span> <span class="variable">$obj</span><span class="operator">-&gt;</span><span class="variable">label</span><span class="operator">();</span>
 <span class="variable">Function</span><span class="operator">:</span>
 
            <span class="variable">Sets</span> <span class="keyword">and</span> <span class="variable">returns</span> <span class="variable">mutation</span> <span class="variable">event</span> <span class="variable">label</span><span class="operator">(</span><span class="regex">s).  If value is not
            set, or no argument is given returns false.  Each
            instantiable subclass of L&lt;Bio::Variation::VariantI&gt; needs
            to implement this method. Valid values are listed in
            'Mutation event controlled vocabulary' in
            http://www.ebi.ac.uk/mutations/recommendations/mutevent.html.
 
 Example : 
 Returns : string
 Args    : string
 </span>
</code></code></pre>

<h2 id="similarity_score">similarity_score</h2>

<pre><code><code> Title   : similarity_score
 Usage   : $self-&gt;similarity_score
 Function: Measure for evolutionary conservativeness
           of single amino substitutions. Uses BLOSUM62.
           Negative numbers are noncoservative changes.
 Returns : integer, undef if not single amino acid change</code></code></pre>

<h2 id="trivname">trivname</h2>

<pre><code><code> Title   : trivname
 Usage   : $self-&gt;trivname
 Function: 

           Given a Bio::Variation::AAChange object with linked
           Bio::Variation::RNAChange and Bio::Variation::DNAMutation
           objects, this subroutine creates a string corresponding to
           the &#39;trivial name&#39; of the mutation. Trivial name is
           specified in Antonorakis &amp; MDI Nomenclature Working Group:
           Human Mutation 11:1-3, 1998.

 Returns : string</code></code></pre>


</body>

</html>


