<?xml version="1.0" ?>
<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.0 Transitional//EN">
<html>
<head>
<!-- saved from url=(0017)http://localhost/ -->
<script language="JavaScript" src="../../../../displayToc.js"></script>
<script language="JavaScript" src="../../../../tocParas.js"></script>
<script language="JavaScript" src="../../../../tocTab.js"></script>
<link rel="stylesheet" type="text/css" href="../../../../scineplex.css">
<title></title>
<link rel="stylesheet" href="../../../../Active.css" type="text/css" />
<meta http-equiv="content-type" content="text/html; charset=utf-8" />
<link rev="made" href="mailto:support@ActiveState.com" />
</head>

<body>



<ul id="index">
  <li><a href="#NAME">NAME</a></li>
  <li><a href="#SYNOPSIS">SYNOPSIS</a></li>
  <li><a href="#DESCRIPTION">DESCRIPTION</a></li>
  <li><a href="#FEEDBACK">FEEDBACK</a>
    <ul>
      <li><a href="#Mailing-Lists">Mailing Lists</a></li>
      <li><a href="#Support">Support</a></li>
      <li><a href="#Reporting-Bugs">Reporting Bugs</a></li>
    </ul>
  </li>
  <li><a href="#AUTHOR">AUTHOR</a></li>
  <li><a href="#APPENDIX">APPENDIX</a>
    <ul>
      <li><a href="#get-set-able-methods">get/set-able methods</a></li>
      <li><a href="#is_subsnp">is_subsnp</a></li>
      <li><a href="#subsnp">subsnp</a></li>
      <li><a href="#add_subsnp">add_subsnp</a></li>
      <li><a href="#each_subsnp">each_subsnp</a></li>
    </ul>
  </li>
</ul>

<h1 id="NAME">NAME</h1>

<p>Bio::Variation::SNP - submitted SNP</p>

<h1 id="SYNOPSIS">SYNOPSIS</h1>

<pre><code><code>  <span class="variable">$SNP</span> <span class="operator">=</span> <span class="variable">Bio::Variation::SNP</span><span class="operator">-&gt;</span><span class="variable">new</span> <span class="operator">();</span>
</code></code></pre>

<h1 id="DESCRIPTION">DESCRIPTION</h1>

<p>Inherits from Bio::Variation::SeqDiff and Bio::Variation::Allele, with additional methods that are (db)SNP specific (ie, refSNP/subSNP IDs, batch IDs, validation methods).</p>

<h1 id="FEEDBACK">FEEDBACK</h1>

<h2 id="Mailing-Lists">Mailing Lists</h2>

<p>User feedback is an integral part of the evolution of this and other Bioperl modules. Send your comments and suggestions preferably to one of the Bioperl mailing lists. Your participation is much appreciated.</p>

<pre><code><code>  bioperl-l@bioperl.org                  - General discussion
  http://bioperl.org/wiki/Mailing_lists  - About the mailing lists</code></code></pre>

<h2 id="Support">Support</h2>

<p>Please direct usage questions or support issues to the mailing list:</p>

<p><i>bioperl-l@bioperl.org</i></p>

<p>rather than to the module maintainer directly. Many experienced and reponsive experts will be able look at the problem and quickly address it. Please include a thorough description of the problem with code and data examples if at all possible.</p>

<h2 id="Reporting-Bugs">Reporting Bugs</h2>

<p>Report bugs to the Bioperl bug tracking system to help us keep track the bugs and their resolution. Bug reports can be submitted via the web:</p>

<pre><code><code>  https://redmine.open-bio.org/projects/bioperl/</code></code></pre>

<h1 id="AUTHOR">AUTHOR</h1>

<p>Allen Day &lt;allenday@ucla.edu&gt;</p>

<h1 id="APPENDIX">APPENDIX</h1>

<p>The rest of the documentation details each of the object methods. Internal methods are usually preceded with a _</p>

<h2 id="get-set-able-methods">get/set-able methods</h2>

<pre><code><code> Usage   : $is = $snp-&gt;method()
 Function: for getting/setting attributes
 Returns : a value.  probably a scalar.
 Args    : if you&#39;re trying to set an attribute, pass in the new value.

 Methods:
 --------
 id
 type
 observed
 seq_5
 seq_3
 ncbi_build
 ncbi_chr_hits
 ncbi_ctg_hits
 ncbi_seq_loc
 ucsc_build 
 ucsc_chr_hits
 ucsc_ctg_hits
 heterozygous
 heterozygous_SE
 validated
 genotype
 handle
 batch_id
 method
 locus_id
 symbol
 mrna
 protein
 functional_class</code></code></pre>

<h2 id="is_subsnp">is_subsnp</h2>

<pre><code><code> Title   : is_subsnp
 Usage   : $is = $snp-&gt;is_subsnp()
 Function: returns 1 if $snp is a subSNP
 Returns : 1 or undef
 Args    : NONE</code></code></pre>

<h2 id="subsnp">subsnp</h2>

<pre><code><code> Title   : subsnp
 Usage   : $subsnp = $snp-&gt;subsnp()
 Function: returns the currently active subSNP of $snp
 Returns : Bio::Variation::SNP
 Args    : NONE</code></code></pre>

<h2 id="add_subsnp">add_subsnp</h2>

<pre><code><code> Title   : add_subsnp
 Usage   : $subsnp = $snp-&gt;add_subsnp()
 Function: pushes the previous value returned by subsnp() onto a stack,
           accessible with each_subsnp().
           Sets return value of subsnp() to a new Bio::Variation::SNP
           object, and returns that object.
 Returns : Bio::Varitiation::SNP
 Args    : NONE</code></code></pre>

<h2 id="each_subsnp">each_subsnp</h2>

<pre><code><code> Title   : each_subsnp
 Usage   : @subsnps = $snp-&gt;each_subsnp()
 Function: returns a list of the subSNPs of a refSNP
 Returns : list
 Args    : NONE</code></code></pre>


</body>

</html>


