<?xml version="1.0" ?>
<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.0 Transitional//EN">
<html>
<head>
<!-- saved from url=(0017)http://localhost/ -->
<script language="JavaScript" src="../../../../../../displayToc.js"></script>
<script language="JavaScript" src="../../../../../../tocParas.js"></script>
<script language="JavaScript" src="../../../../../../tocTab.js"></script>
<link rel="stylesheet" type="text/css" href="../../../../../../scineplex.css">
<title></title>
<link rel="stylesheet" href="../../../../../../Active.css" type="text/css" />
<meta http-equiv="content-type" content="text/html; charset=utf-8" />
<link rev="made" href="mailto:support@ActiveState.com" />
</head>

<body>



<ul id="index">
  <li><a href="#NAME">NAME</a></li>
  <li><a href="#SYNOPSIS">SYNOPSIS</a></li>
  <li><a href="#DESCRIPTION">DESCRIPTION</a>
    <ul>
      <li><a href="#aggregate">aggregate</a></li>
      <li><a href="#method">method</a></li>
      <li><a href="#part_names">part_names</a></li>
    </ul>
  </li>
  <li><a href="#BUGS">BUGS</a></li>
  <li><a href="#SEE-ALSO">SEE ALSO</a></li>
  <li><a href="#AUTHOR">AUTHOR</a></li>
</ul>

<h1 id="NAME">NAME</h1>

<p>Bio::DB::GFF::Aggregator::clone -- Clone aggregator</p>

<h1 id="SYNOPSIS">SYNOPSIS</h1>

<pre><code><code>  <span class="keyword">use</span> <span class="variable">Bio::DB::GFF</span><span class="operator">;</span>
  
  <span class="comment"># Open the sequence database</span>
  <span class="keyword">my</span> <span class="variable">$db</span>      <span class="operator">=</span> <span class="variable">Bio::DB::GFF</span><span class="operator">-&gt;</span><span class="variable">new</span><span class="operator">(</span> <span class="string">-adaptor</span> <span class="operator">=&gt;</span> <span class="string">'dbi:mysql'</span><span class="operator">,</span>
                                   <span class="string">-dsn</span>     <span class="operator">=&gt;</span> <span class="string">'dbi:mysql:elegans42'</span><span class="operator">,</span>
                                   <span class="string">-aggregator</span> <span class="operator">=&gt;</span> <span class="operator">[</span><span class="string">'transcript'</span><span class="operator">,</span><span class="string">'clone'</span><span class="operator">]</span><span class="operator">,</span>
                                 <span class="operator">);</span>
  
   <span class="operator">----------------------------------------------------------------------------</span>
   <span class="variable">Aggregator</span> <span class="variable">method</span><span class="operator">:</span> <span class="variable">clone</span>
   <span class="variable">Main</span> <span class="variable">method</span><span class="operator">:</span>       <span class="operator">-</span><span class="variable">none</span><span class="operator">-</span>
   <span class="variable">Sub</span> <span class="variable">methods</span><span class="operator">:</span>       <span class="variable">Clone_left_end</span> <span class="variable">Clone_right_end</span> <span class="variable">region</span><span class="operator">:</span><span class="variable">Genomic_canonical</span>
   <span class="operator">----------------------------------------------------------------------------</span>
</code></code></pre>

<h1 id="DESCRIPTION">DESCRIPTION</h1>

<p>Bio::DB::GFF::Aggregator::clone is one of the default aggregators, and was written to be compatible with the C elegans GFF files. It aggregates raw &quot;Clone_left_end&quot;, &quot;Clone_right_end&quot;, and &quot;region:Genomic_canonical&quot; features into composite features of type &quot;clone&quot;.</p>

<h2 id="aggregate">aggregate</h2>

<pre><code><code> Title   : aggregate
 Usage   : $features = $a-&gt;aggregate($features,$factory)
 Function: aggregate a feature list into composite features
 Returns : an array reference containing modified features
 Args    : see L&lt;Bio::DB::GFF::Aggregator&gt;
 Status  : Public</code></code></pre>

<p>The WormBase GFF model is unusual in that clones aren&#39;t identified as a single feature with start and stop positions, but as two features, a &quot;left end&quot; and a &quot;right end&quot;. One or both of these features may be absent. In order to accommodate this, the aggregator will return undef for the start and/or stop if one or both of the ends are missing.</p>

<h2 id="method">method</h2>

<pre><code><code> Title   : method
 Usage   : $aggregator-&gt;method
 Function: return the method for the composite object
 Returns : the string &quot;clone&quot;
 Args    : none
 Status  : Public</code></code></pre>

<h2 id="part_names">part_names</h2>

<pre><code><code> Title   : part_names
 Usage   : $aggregator-&gt;part_names
 Function: return the methods for the sub-parts
 Returns : the list (&quot;Clone_left_end&quot;, &quot;Clone_right_end&quot;, &quot;region:Genomic_canonical&quot;)
 Args    : none
 Status  : Public</code></code></pre>

<h1 id="BUGS">BUGS</h1>

<p>None reported.</p>

<h1 id="SEE-ALSO">SEE ALSO</h1>

<p><a href="../../../../../../../.cpan/build/BioPerl-1.6.901-ghKY8U/blib/lib/Bio/DB/GFF.html">Bio::DB::GFF</a>, <a href="../../../../../../../.cpan/build/BioPerl-1.6.901-ghKY8U/blib/lib/Bio/DB/GFF/Aggregator.html">Bio::DB::GFF::Aggregator</a></p>

<h1 id="AUTHOR">AUTHOR</h1>

<p>Lincoln Stein &lt;lstein@cshl.org&gt;.</p>

<p>Copyright (c) 2001 Cold Spring Harbor Laboratory.</p>

<p>This library is free software; you can redistribute it and/or modify it under the same terms as Perl itself.</p>


</body>

</html>


