<?xml version="1.0" ?>
<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.0 Transitional//EN">
<html>
<head>
<!-- saved from url=(0017)http://localhost/ -->
<script language="JavaScript" src="../../../../../../displayToc.js"></script>
<script language="JavaScript" src="../../../../../../tocParas.js"></script>
<script language="JavaScript" src="../../../../../../tocTab.js"></script>
<link rel="stylesheet" type="text/css" href="../../../../../../scineplex.css">
<title></title>
<link rel="stylesheet" href="../../../../../../Active.css" type="text/css" />
<meta http-equiv="content-type" content="text/html; charset=utf-8" />
<link rev="made" href="mailto:support@ActiveState.com" />
</head>

<body>



<ul id="index">
  <li><a href="#NAME">NAME</a></li>
  <li><a href="#SYNOPSIS">SYNOPSIS</a></li>
  <li><a href="#DESCRIPTION">DESCRIPTION</a></li>
  <li><a href="#FEEDBACK">FEEDBACK</a>
    <ul>
      <li><a href="#Mailing-Lists">Mailing Lists</a></li>
      <li><a href="#Support">Support</a></li>
      <li><a href="#Reporting-Bugs">Reporting Bugs</a></li>
    </ul>
  </li>
  <li><a href="#AUTHOR---Sendu-Bala">AUTHOR - Sendu Bala</a></li>
  <li><a href="#APPENDIX">APPENDIX</a>
    <ul>
      <li><a href="#program_name">program_name</a></li>
      <li><a href="#program_dir">program_dir</a></li>
      <li><a href="#new">new</a></li>
      <li><a href="#run">run</a></li>
      <li><a href="#setparams">_setparams</a></li>
    </ul>
  </li>
</ul>

<h1 id="NAME">NAME</h1>

<p>Bio::Tools::Run::Phylo::Semphy - Wrapper for Semphy</p>

<h1 id="SYNOPSIS">SYNOPSIS</h1>

<pre><code><code>  <span class="keyword">use</span> <span class="variable">Bio::Tools::Run::Phylo::Semphy</span><span class="operator">;</span>
  
  <span class="comment"># Make a Semphy factory</span>
  <span class="variable">$factory</span> <span class="operator">=</span> <span class="variable">Bio::Tools::Run::Phylo::Semphy</span><span class="operator">-&gt;</span><span class="variable">new</span><span class="operator">();</span>
  
  <span class="comment"># Run Semphy with an alignment</span>
  <span class="keyword">my</span> <span class="variable">$tree</span> <span class="operator">=</span> <span class="variable">$factory</span><span class="operator">-&gt;</span><span class="variable">run</span><span class="operator">(</span><span class="variable">$alignfilename</span><span class="operator">);</span>
  
  <span class="comment"># or with alignment object</span>
  <span class="variable">$tree</span> <span class="operator">=</span> <span class="variable">$factory</span><span class="operator">-&gt;</span><span class="variable">run</span><span class="operator">(</span><span class="variable">$bio_simplalign</span><span class="operator">);</span>
  
  <span class="comment"># you can supply an initial tree as well, which can be a newick tree file,</span>
  <span class="comment"># Bio::Tree::Tree object...</span>
  <span class="variable">$tree</span> <span class="operator">=</span> <span class="variable">$factory</span><span class="operator">-&gt;</span><span class="variable">run</span><span class="operator">(</span><span class="variable">$bio_simplalign</span><span class="operator">,</span> <span class="variable">$tree_obj</span><span class="operator">);</span>
  
  <span class="comment"># ... or Bio::DB::Taxonomy object</span>
  <span class="variable">$tree</span> <span class="operator">=</span> <span class="variable">$factory</span><span class="operator">-&gt;</span><span class="variable">run</span><span class="operator">(</span><span class="variable">$bio_simplalign</span><span class="operator">,</span> <span class="variable">$bio_db_taxonomy</span><span class="operator">);</span>
  
  <span class="comment"># (mixtures of all the above are possible)</span>
  
  <span class="comment"># $tree isa Bio::Tree::Tree</span>
</code></code></pre>

<h1 id="DESCRIPTION">DESCRIPTION</h1>

<p>This is a wrapper for running the Semphy application by N. Friedman et a.. You can get details here: http://compbio.cs.huji.ac.il/semphy/. Semphy is used for phylogenetic reconstruction (making a tree with branch lengths from an aligned set of input sequences).</p>

<p>You can try supplying normal Semphy command-line arguments to new(), eg. new(-hky =&gt; 1) or calling arg-named methods (excluding the initial hyphen(s), eg. $factory-&gt;hky(1) to set the --hky switch to true). Note that Semphy args are case-sensitive. To distinguish between Bioperl&#39;s -verbose and the Semphy&#39;s --verbose, you must set Semphy&#39;s verbosity with -semphy_verbose or the semphy_verbose() method.</p>

<p>You will need to enable this Semphy wrapper to find the Semphy program. This can be done in (at least) three ways:</p>

<pre><code><code> <span class="number">1.</span> <span class="variable">Make</span> <span class="variable">sure</span> <span class="variable">the</span> <span class="variable">Semphy</span> <span class="variable">executable</span> <span class="variable">is</span> <span class="variable">in</span> <span class="variable">your</span> <span class="variable">path</span><span class="operator">.</span>
 <span class="number">2.</span> <span class="variable">Define</span> <span class="variable">an</span> <span class="variable">environmental</span> <span class="variable">variable</span> <span class="variable">SEMPHYDIR</span> <span class="variable">which</span> <span class="variable">is</span> <span class="variable">a</span> 
    <span class="variable">directory</span> <span class="variable">which</span> <span class="variable">contains</span> <span class="variable">the</span> <span class="variable">Semphy</span> <span class="variable">application</span><span class="operator">:</span>
    <span class="variable">In</span> <span class="variable">bash</span><span class="operator">:</span>
 
    <span class="variable">export</span> <span class="variable">SEMPHYDIR</span><span class="operator">=</span><span class="regex">/home/username</span><span class="operator">/</span><span class="variable">semphy</span><span class="operator">/</span>
 
    <span class="variable">In</span> <span class="variable">csh</span><span class="operator">/</span><span class="variable">tcsh</span><span class="operator">:</span>
 
    <span class="variable">setenv</span> <span class="variable">SEMPHYDIR</span> <span class="operator">/</span><span class="variable">home</span><span class="operator">/</span><span class="variable">username</span><span class="operator">/</span><span class="variable">semphy</span>
 
 <span class="number">3.</span> <span class="variable">Include</span> <span class="variable">a</span> <span class="variable">definition</span> <span class="variable">of</span> <span class="variable">an</span> <span class="variable">environmental</span> <span class="variable">variable</span> <span class="variable">SEMPHYDIR</span> <span class="variable">in</span>
    <span class="variable">every</span> <span class="variable">script</span> <span class="variable">that</span> <span class="variable">will</span> <span class="keyword">use</span> <span class="variable">this</span> <span class="variable">Semphy</span> <span class="variable">wrapper</span> <span class="variable">module</span><span class="operator">,</span> <span class="variable">e</span><span class="operator">.</span><span class="variable">g</span><span class="operator">.:</span>
 
    <span class="keyword">BEGIN</span> <span class="operator">{</span> <span class="variable">$ENV</span><span class="operator">{</span><span class="string">SEMPHYDIR</span><span class="operator">}</span> <span class="operator">=</span> <span class="string">'/home/username/semphy/'</span> <span class="operator">}</span>
    <span class="keyword">use</span> <span class="variable">Bio::Tools::Run::Phylo::Semphy</span><span class="operator">;</span>
</code></code></pre>

<h1 id="FEEDBACK">FEEDBACK</h1>

<h2 id="Mailing-Lists">Mailing Lists</h2>

<p>User feedback is an integral part of the evolution of this and other Bioperl modules. Send your comments and suggestions preferably to the Bioperl mailing list. Your participation is much appreciated.</p>

<pre><code><code>  bioperl-l@bioperl.org                  - General discussion
  http://bioperl.org/wiki/Mailing_lists  - About the mailing lists</code></code></pre>

<h2 id="Support">Support</h2>

<p>Please direct usage questions or support issues to the mailing list:</p>

<p><i>bioperl-l@bioperl.org</i></p>

<p>rather than to the module maintainer directly. Many experienced and reponsive experts will be able look at the problem and quickly address it. Please include a thorough description of the problem with code and data examples if at all possible.</p>

<h2 id="Reporting-Bugs">Reporting Bugs</h2>

<p>Report bugs to the Bioperl bug tracking system to help us keep track of the bugs and their resolution. Bug reports can be submitted via the web:</p>

<pre><code><code>  http://redmine.open-bio.org/projects/bioperl/</code></code></pre>

<h1 id="AUTHOR---Sendu-Bala">AUTHOR - Sendu Bala</h1>

<p>Email bix@sendu.me.uk</p>

<h1 id="APPENDIX">APPENDIX</h1>

<p>The rest of the documentation details each of the object methods. Internal methods are usually preceded with a _</p>

<h2 id="program_name">program_name</h2>

<pre><code><code> Title   : program_name
 Usage   : $factory&gt;program_name()
 Function: holds the program name
 Returns : string
 Args    : None</code></code></pre>

<h2 id="program_dir">program_dir</h2>

<pre><code><code> Title   : program_dir
 Usage   : $factory-&gt;program_dir(@params)
 Function: returns the program directory, obtained from ENV variable.
 Returns : string
 Args    : None</code></code></pre>

<h2 id="new">new</h2>

<pre><code><code> Title   : new
 Usage   : $factory = Bio::Tools::Run::Phylo::Semphy-&gt;new()
 Function: creates a new Semphy factory
 Returns : Bio::Tools::Run::Phylo::Semphy
 Args    : Most options understood by Semphy can be supplied as key =&gt;
           value pairs, with a true value for switches.

           These options can NOT be used with this wrapper (they are handled
           internally or don&#39;t make sense in this context):
           -h | --help | --fill-help
           -s | --sequence
           -t | --tree

           To distinguish between Bioperl&#39;s -verbose and the Semphy&#39;s --verbose,
           you must set Semphy&#39;s verbosity with -semphy_verbose</code></code></pre>

<h2 id="run">run</h2>

<pre><code><code> <span class="variable">Title</span>   <span class="operator">:</span> <span class="variable">run</span>
 <span class="variable">Usage</span>   <span class="operator">:</span> <span class="variable">$result</span> <span class="operator">=</span> <span class="variable">$factory</span><span class="operator">-&gt;</span><span class="variable">run</span><span class="operator">(</span><span class="variable">$fasta_align_file</span><span class="operator">);</span>
           <span class="operator">-</span><span class="keyword">or</span><span class="operator">-</span>
           <span class="variable">$result</span> <span class="operator">=</span> <span class="variable">$factory</span><span class="operator">-&gt;</span><span class="variable">run</span><span class="operator">(</span><span class="variable">$align_object</span><span class="operator">);</span>
           <span class="operator">-</span><span class="keyword">or</span><span class="operator">-</span>
           <span class="variable">$result</span> <span class="operator">=</span> <span class="variable">$factory</span><span class="operator">-&gt;</span><span class="variable">run</span><span class="operator">(</span><span class="variable">$fasta_align_file</span><span class="operator">,</span> <span class="variable">$newick_tree_file</span><span class="operator">);</span>
           <span class="operator">-</span><span class="keyword">or</span><span class="operator">-</span>
           <span class="variable">$result</span> <span class="operator">=</span> <span class="variable">$factory</span><span class="operator">-&gt;</span><span class="variable">run</span><span class="operator">(</span><span class="variable">$align_object</span><span class="operator">,</span> <span class="variable">$tree_object</span><span class="operator">);</span>
           <span class="operator">-</span><span class="keyword">or</span><span class="operator">-</span>
           <span class="variable">$result</span> <span class="operator">=</span> <span class="variable">$factory</span><span class="operator">-&gt;</span><span class="variable">run</span><span class="operator">(</span><span class="variable">$align_object</span><span class="operator">,</span> <span class="variable">$db_taxonomy_object</span><span class="operator">);</span>
 <span class="variable">Function</span><span class="operator">:</span> <span class="variable">Runs</span> <span class="variable">Semphy</span> <span class="variable">on</span> <span class="variable">an</span> <span class="variable">alignment</span><span class="operator">.</span>
 <span class="variable">Returns</span> <span class="operator">:</span> <span class="variable">Bio::Tree::Tree</span>
 <span class="variable">Args</span>    <span class="operator">:</span> <span class="variable">The</span> <span class="variable">first</span> <span class="variable">argument</span> <span class="variable">represents</span> <span class="variable">an</span> <span class="variable">alignment</span><span class="operator">,</span> <span class="variable">the</span> <span class="variable">second</span> <span class="operator">(</span><span class="variable">optional</span><span class="operator">)</span>
           <span class="variable">argument</span> <span class="variable">a</span> <span class="variable">species</span> <span class="variable">tree</span> <span class="operator">(</span><span class="variable">to</span> <span class="variable">set</span> <span class="variable">an</span> <span class="variable">initial</span> <span class="variable">tree</span><span class="operator">:</span> <span class="variable">normally</span> <span class="variable">the</span> <span class="keyword">-t</span>
           <span class="variable">option</span> <span class="variable">to</span> <span class="variable">Semphy</span><span class="operator">).</span>
           <span class="variable">The</span> <span class="variable">alignment</span> <span class="variable">can</span> <span class="variable">be</span> <span class="variable">provided</span> <span class="variable">as</span> <span class="variable">a</span> <span class="variable">multi</span><span class="operator">-</span><span class="variable">fasta</span> <span class="keyword">format</span> <span class="variable">alignment</span>
           <span class="variable">filename</span><span class="operator">,</span> <span class="keyword">or</span> <span class="variable">a</span> <span class="variable">Bio::Align::AlignI</span> <span class="variable">complient</span> <span class="variable">object</span> <span class="operator">(</span><span class="variable">eg</span><span class="operator">.</span> <span class="variable">a</span>
           <span class="variable">Bio::SimpleAlign</span><span class="operator">).</span>
           <span class="variable">The</span> <span class="variable">species</span> <span class="variable">tree</span> <span class="variable">can</span> <span class="variable">be</span> <span class="variable">provided</span> <span class="variable">as</span> <span class="variable">a</span> <span class="variable">newick</span> <span class="keyword">format</span> <span class="variable">tree</span> <span class="variable">filename</span>
           <span class="keyword">or</span> <span class="variable">a</span> <span class="variable">Bio::Tree::TreeI</span> <span class="variable">complient</span> <span class="variable">object</span><span class="operator">.</span> <span class="variable">Alternatively</span> <span class="variable">a</span>
           <span class="variable">Bio::DB::Taxonomy</span> <span class="variable">object</span> <span class="variable">can</span> <span class="variable">be</span> <span class="variable">supplied</span><span class="operator">,</span> <span class="variable">in</span> <span class="variable">which</span> <span class="variable">case</span> <span class="variable">the</span> <span class="variable">species</span>
           <span class="variable">tree</span> <span class="variable">will</span> <span class="variable">be</span> <span class="variable">generated</span> <span class="variable">by</span> <span class="variable">using</span> <span class="variable">the</span> <span class="variable">alignment</span> <span class="variable">sequence</span> <span class="variable">names</span> <span class="variable">as</span>
           <span class="variable">species</span> <span class="variable">names</span> <span class="keyword">and</span> <span class="variable">looking</span> <span class="keyword">for</span> <span class="variable">those</span> <span class="variable">in</span> <span class="variable">the</span> <span class="variable">supplied</span> <span class="variable">database</span><span class="operator">.</span>
           
           <span class="variable">In</span> <span class="variable">all</span> <span class="variable">cases</span> <span class="variable">where</span> <span class="variable">an</span> <span class="variable">initial</span> <span class="variable">tree</span> <span class="variable">was</span> <span class="variable">supplied</span><span class="operator">,</span> <span class="variable">the</span> <span class="variable">alignment</span>
           <span class="variable">sequence</span> <span class="variable">names</span> <span class="variable">must</span> <span class="variable">correspond</span> <span class="variable">to</span> <span class="variable">node</span> <span class="variable">ids</span> <span class="variable">in</span> <span class="variable">the</span> <span class="variable">species</span> <span class="variable">tree</span><span class="operator">.</span>
</code></code></pre>

<h2 id="setparams">_setparams</h2>

<pre><code><code> Title   : _setparams
 Usage   : Internal function, not to be called directly
 Function: Creates a string of params to be used in the command string
 Returns : string of params
 Args    : alignment and tree file names</code></code></pre>


</body>

</html>


