<?xml version="1.0" ?>
<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.0 Transitional//EN">
<html>
<head>
<!-- saved from url=(0017)http://localhost/ -->
<script language="JavaScript" src="../../../../../../displayToc.js"></script>
<script language="JavaScript" src="../../../../../../tocParas.js"></script>
<script language="JavaScript" src="../../../../../../tocTab.js"></script>
<link rel="stylesheet" type="text/css" href="../../../../../../scineplex.css">
<title></title>
<link rel="stylesheet" href="../../../../../../Active.css" type="text/css" />
<meta http-equiv="content-type" content="text/html; charset=utf-8" />
<link rev="made" href="mailto:support@ActiveState.com" />
</head>

<body>



<ul id="index">
  <li><a href="#NAME">NAME</a></li>
  <li><a href="#SYNOPSIS">SYNOPSIS</a></li>
  <li><a href="#DESCRIPTION">DESCRIPTION</a></li>
  <li><a href="#FEEDBACK">FEEDBACK</a>
    <ul>
      <li><a href="#Mailing-Lists">Mailing Lists</a></li>
      <li><a href="#Support">Support</a></li>
      <li><a href="#Reporting-Bugs">Reporting Bugs</a></li>
    </ul>
  </li>
  <li><a href="#AUTHOR---Albert-Vilella">AUTHOR - Albert Vilella</a></li>
  <li><a href="#CONTRIBUTORS">CONTRIBUTORS</a></li>
  <li><a href="#APPENDIX">APPENDIX</a>
    <ul>
      <li><a href="#Default-Values">Default Values</a></li>
      <li><a href="#Results-Format">Results Format</a></li>
      <li><a href="#program_name">program_name</a></li>
      <li><a href="#program_dir">program_dir</a></li>
      <li><a href="#new">new</a></li>
      <li><a href="#prepare">prepare</a></li>
      <li><a href="#run">run</a></li>
      <li><a href="#error_string">error_string</a></li>
      <li><a href="#alignment">alignment</a></li>
      <li><a href="#tree">tree</a></li>
      <li><a href="#get_parameters">get_parameters</a></li>
      <li><a href="#set_parameter">set_parameter</a></li>
      <li><a href="#set_default_parameters">set_default_parameters</a></li>
    </ul>
  </li>
  <li><a href="#Bio::Tools::Run::WrapperBase-methods">Bio::Tools::Run::WrapperBase methods</a>
    <ul>
      <li><a href="#no_param_checks">no_param_checks</a></li>
      <li><a href="#save_tempfiles">save_tempfiles</a></li>
      <li><a href="#outfile_name">outfile_name</a></li>
      <li><a href="#tempdir">tempdir</a></li>
      <li><a href="#cleanup">cleanup</a></li>
      <li><a href="#io">io</a></li>
    </ul>
  </li>
</ul>

<h1 id="NAME">NAME</h1>

<p>Bio::Tools::Run::Phylo::SLR - Wrapper around the SLR program</p>

<h1 id="SYNOPSIS">SYNOPSIS</h1>

<pre><code><code>  <span class="keyword">use</span> <span class="variable">Bio::Tools::Run::Phylo::SLR</span><span class="operator">;</span>
  <span class="keyword">use</span> <span class="variable">Bio::AlignIO</span><span class="operator">;</span>
  <span class="keyword">use</span> <span class="variable">Bio::TreeIO</span><span class="operator">;</span>
  <span class="keyword">use</span> <span class="variable">Bio::SimpleAlign</span><span class="operator">;</span>
  
  <span class="keyword">my</span> <span class="variable">$alignio</span> <span class="operator">=</span> <span class="variable">Bio::AlignIO</span><span class="operator">-&gt;</span><span class="variable">new</span>
      <span class="operator">(</span><span class="string">-format</span> <span class="operator">=&gt;</span> <span class="string">'fasta'</span><span class="operator">,</span>
       <span class="string">-file</span>   <span class="operator">=&gt;</span> <span class="string">'t/data/219877.cdna.fasta'</span><span class="operator">);</span>
  
  <span class="keyword">my</span> <span class="variable">$aln</span> <span class="operator">=</span> <span class="variable">$alignio</span><span class="operator">-&gt;</span><span class="variable">next_aln</span><span class="operator">;</span>
  
  <span class="keyword">my</span> <span class="variable">$treeio</span> <span class="operator">=</span> <span class="variable">Bio::TreeIO</span><span class="operator">-&gt;</span><span class="variable">new</span>
      <span class="operator">(</span><span class="string">-format</span> <span class="operator">=&gt;</span> <span class="string">'newick'</span><span class="operator">,</span> <span class="string">-file</span> <span class="operator">=&gt;</span> <span class="string">'t/data/219877.tree'</span><span class="operator">);</span>
  
  <span class="keyword">my</span> <span class="variable">$tree</span> <span class="operator">=</span> <span class="variable">$treeio</span><span class="operator">-&gt;</span><span class="variable">next_tree</span><span class="operator">;</span>
  
  <span class="keyword">my</span> <span class="variable">$slr</span> <span class="operator">=</span> <span class="variable">Bio::Tools::Run::Phylo::SLR</span><span class="operator">-&gt;</span><span class="variable">new</span><span class="operator">();</span>
  <span class="variable">$slr</span><span class="operator">-&gt;</span><span class="variable">alignment</span><span class="operator">(</span><span class="variable">$aln</span><span class="operator">);</span>
  <span class="variable">$slr</span><span class="operator">-&gt;</span><span class="variable">tree</span><span class="operator">(</span><span class="variable">$tree</span><span class="operator">);</span>
  <span class="comment"># $rc = 1 for success, 0 for errors</span>
  <span class="keyword">my</span> <span class="operator">(</span><span class="variable">$rc</span><span class="operator">,</span><span class="variable">$results</span><span class="operator">)</span> <span class="operator">=</span> <span class="variable">$slr</span><span class="operator">-&gt;</span><span class="variable">run</span><span class="operator">();</span>
  
  <span class="keyword">my</span> <span class="variable">$positive_sites</span> <span class="operator">=</span> <span class="variable">$results</span><span class="operator">-&gt;</span><span class="operator">{</span><span class="string">'positive'</span><span class="operator">}</span><span class="operator">;</span>
  
  <span class="keyword">print</span> <span class="string">"# Site\tNeutral\tOptimal\tOmega\t"</span><span class="operator">,</span>
        <span class="string">"lower\tupper\tLRT_Stat\tPval\tAdj.Pval\tResult\tNote\n"</span><span class="operator">;</span>
  <span class="keyword">foreach</span> <span class="keyword">my</span> <span class="variable">$positive_site</span> <span class="operator">(</span><span class="variable">@$positive_sites</span><span class="operator">)</span> <span class="operator">{</span>
      <span class="keyword">print</span> 
          <span class="variable">$positive_site</span><span class="operator">-&gt;</span><span class="operator">[</span><span class="number">0</span><span class="operator">]</span><span class="operator">,</span> <span class="string">"\t"</span><span class="operator">,</span>
          <span class="variable">$positive_site</span><span class="operator">-&gt;</span><span class="operator">[</span><span class="number">1</span><span class="operator">]</span><span class="operator">,</span> <span class="string">"\t"</span><span class="operator">,</span>
          <span class="variable">$positive_site</span><span class="operator">-&gt;</span><span class="operator">[</span><span class="number">2</span><span class="operator">]</span><span class="operator">,</span> <span class="string">"\t"</span><span class="operator">,</span>
          <span class="variable">$positive_site</span><span class="operator">-&gt;</span><span class="operator">[</span><span class="number">3</span><span class="operator">]</span><span class="operator">,</span> <span class="string">"\t"</span><span class="operator">,</span>
          <span class="variable">$positive_site</span><span class="operator">-&gt;</span><span class="operator">[</span><span class="number">4</span><span class="operator">]</span><span class="operator">,</span> <span class="string">"\t"</span><span class="operator">,</span>
          <span class="variable">$positive_site</span><span class="operator">-&gt;</span><span class="operator">[</span><span class="number">5</span><span class="operator">]</span><span class="operator">,</span> <span class="string">"\t"</span><span class="operator">,</span>
          <span class="variable">$positive_site</span><span class="operator">-&gt;</span><span class="operator">[</span><span class="number">6</span><span class="operator">]</span><span class="operator">,</span> <span class="string">"\t"</span><span class="operator">,</span>
          <span class="variable">$positive_site</span><span class="operator">-&gt;</span><span class="operator">[</span><span class="number">7</span><span class="operator">]</span><span class="operator">,</span> <span class="string">"\t"</span><span class="operator">,</span>
          <span class="variable">$positive_site</span><span class="operator">-&gt;</span><span class="operator">[</span><span class="number">8</span><span class="operator">]</span><span class="operator">,</span> <span class="string">"\t"</span><span class="operator">,</span>
          <span class="string">"positive\n"</span><span class="operator">;</span>
  <span class="operator">}</span>
</code></code></pre>

<h1 id="DESCRIPTION">DESCRIPTION</h1>

<p>This is a wrapper around the SLR program. See http://www.ebi.ac.uk/goldman/SLR/ for more information.</p>

<p>This module is more about generating the proper ctl file and will run the program in a separate temporary directory to avoid creating temp files all over the place.</p>

<h1 id="FEEDBACK">FEEDBACK</h1>

<h2 id="Mailing-Lists">Mailing Lists</h2>

<p>User feedback is an integral part of the evolution of this and other Bioperl modules. Send your comments and suggestions preferably to the Bioperl mailing list. Your participation is much appreciated.</p>

<pre><code><code>  bioperl-l@bioperl.org                  - General discussion
  http://bioperl.org/wiki/Mailing_lists  - About the mailing lists</code></code></pre>

<h2 id="Support">Support</h2>

<p>Please direct usage questions or support issues to the mailing list:</p>

<p><i>bioperl-l@bioperl.org</i></p>

<p>rather than to the module maintainer directly. Many experienced and reponsive experts will be able look at the problem and quickly address it. Please include a thorough description of the problem with code and data examples if at all possible.</p>

<h2 id="Reporting-Bugs">Reporting Bugs</h2>

<p>Report bugs to the Bioperl bug tracking system to help us keep track of the bugs and their resolution. Bug reports can be submitted via the web:</p>

<pre><code><code>  http://redmine.open-bio.org/projects/bioperl/</code></code></pre>

<h1 id="AUTHOR---Albert-Vilella">AUTHOR - Albert Vilella</h1>

<p>Email avilella-at-gmail-dot-com</p>

<h1 id="CONTRIBUTORS">CONTRIBUTORS</h1>

<p>Additional contributors names and emails here</p>

<h1 id="APPENDIX">APPENDIX</h1>

<p>The rest of the documentation details each of the object methods. Internal methods are usually preceded with a _</p>

<h2 id="Default-Values">Default Values</h2>

<p>INCOMPLETE DOCUMENTATION OF ALL METHODS</p>

<p>seqfile [incodon] File from which to read alignment of codon sequences. The file should be in PAML format.</p>

<p>treefile [intree] File from which tree should be read. The tree should be in Nexus format</p>

<p>outfile [slr.res] File to which results are written. If the file already exists, it will be overwritten.</p>

<p>reoptimise [1] Should the branch lengths, omega and kappa be reoptimized? 0 - no 1 - yes.</p>

<p>kappa [2.0] Value for kappa. If &#39;reoptimise&#39; is specified, the value given will be used as am initial estimate,</p>

<p>omega [0.1] Value for omega (dN/dS). If &#39;reoptimise&#39; is specified, the value given will be used as an initial estimate.</p>

<p>codonf [0] How codon frequencies are estimated: 0: F61/F60 Estimates used are the empirical frequencies from the data. 1: F3x4 The frequencies of nucleotides at each codon position are estimated from the data and then multiplied together to get the frequency of observing a given codon. The frequency of stop codons is set to zero, and all other frequencies scaled appropriately. 2: F1x4 Nucleotide frequencies are estimated from the data (not taking into account at which position in the codon it occurs). The nucleotide frequencies are multiplied together to get the frequency of observing and then corrected for stop codons.</p>

<p>freqtype [0] How codon frequencies are incorporated into the substitution matrix. 0: q_{ij} = pi_{j} s_{ij} 1: q_{ij} = \sqrt(pi_j/pi_i) s_{ij} 2: q_{ij} = \pi_{n} s_{ij}, where n is the nucleotide that the subsitution is to. 3: q_{ij} = s_{ij} / pi_i Option 0 is the tradition method of incorporating equilibrium frequencies into subsitution matrices (Felsenstein 1981; Goldman and Yang, 1994) Option 1 is described by Goldman and Whelan (2002), in this case with the additional parameter set to 0.5. Option 2 was suggested by Muse and Gaut (1994). Option 3 is included as an experiment, originally suggested by Bret Larget. it does not appear to describe evolution very successfully and should not be used for analyses.</p>

<pre><code><code>  Kosakovsky-Pond has repeatedly stated that he finds incorporating codon
  frequencies in the manner of option 2 to be superior to option 0. We find
  that option 1 tends to perform better than either of these options.</code></code></pre>

<p>positive_only [0] If only positively selected sites are of interest, set this to &quot;1&quot;. Calculation will be slightly faster, but information about sites under purifying selection is lost.</p>

<p>gencode [universal] Which genetic code to use when determining whether a given mutation is synonymous or nonsynonymous. Currently only &quot;universal&quot; and &quot;mammalian&quot; mitochondrial are supported.</p>

<p>nucleof [0] Allow for empirical exchangabilities for nucleotide substitution. 0: No adjustment. All nucleotides treated the same, modulo transition / transversion. 1: The rate at which a substitution caused a mutation from nucleotide a to nucleotide b is adjust by a constant N_{ab}. This adjustment is in addition to other adjustments (e.g. transition / transversion or base frequencies).</p>

<p>aminof [0] Incorporate amino acid similarity parameters into substitution matrix, adjusting omega for a change between amino acid i and amino acid j. A_{ij} is a symmetric matrix of constants representing amino acid similarities. 0: Constant omega for all amino acid changes 1: omega_{ij} = omega^{A_{ij}} 2: omega_{ij} = a_{ij} log(omega) / [ 1 - exp(-a_{ij} log(omega)) ] Option 1 has the same form as the original codon subsitution model proposed by Goldman and Yang (but with potentially different constants). Option 2 has a more population genetic derivtion, with omega being interpreted as the ratio of fixation probabilities.</p>

<p>nucfile [nuc.dat] If nucleof is non-zero, read nucleotide substitution constants from nucfile. If this file does not exist, hard coded constants are used.</p>

<p>aminofile [amino.dat] If aminof is non-zero, read amino acid similarity constants from aminofile. If this file does not exist, hard coded constants are used.</p>

<p>timemem [0] Print summary of real time and CPU time used. Will eventually print summary of memory use as well.</p>

<p>ldiff [3.841459] Twice log-likelihood difference used as a threshold for calculating support (confidence) intervals for sitewise omega estimates. This value should be the quantile from a chi-square distribution with one degree of freedom corresponding to the support required. E.g. qchisq(0.95,1) = 3.841459 0.4549364 = 50% support 1.323304 = 75% support 2.705543 = 90% support 3.841459 = 95% support 6.634897 = 99% support 7.879439 = 99.5% support 10.82757 = 99.9% support</p>

<p>paramin [] If not blank, read in parameters from file given by the argument.</p>

<p>paramout [] If not blank, write out parameter estimates to file given.</p>

<p>skipsitewise [0] Skip sitewise estimation of omega. Depending on other options given, either calculate maximum likelihood or likelihood fixed at parameter values given.</p>

<p>seed [0] Seed for random number generator. If seed is 0, then previously produced seed file (~/.rng64) is used. If this does not exist, the random number generator is initialised using the clock.</p>

<p>saveseed [1] If non-zero, save finial seed in file (~/.rng64) to be used as initial seed in future runs of program.</p>

<h2 id="Results-Format">Results Format</h2>

<p>Results file (default: slr.res) ------------ Results are presented in nine columns</p>

<p>Site Number of sites in alignment</p>

<p>Neutral (minus) Log-probability of observing site given that it was evolving neutrally (omega=1)</p>

<p>Optimal (minus) Log-probability of observing site given that it was evolving at the optimal value of omega.</p>

<p>Omega The value of omega which maximizes the log-probability of observing</p>

<p>LRT_Stat Log-likelihood ratio statistic for non-neutral selection (or positive selection if the positive_only option is set to 1). LRT_Stat = 2 * (Neutral-Optimal)</p>

<p>Pval P-value for non-neutral (or positive) selection at a site, unadjusted for multiple comparisons.</p>

<p>Adj. Pval P-value for non-neutral (or positive) selection at a site, after adjusting for multiple comparisons using the Hochberg procedure (see the file &quot;MultipleComparisons.txt&quot; in the doc directory).</p>

<p>Result A simple visual guide to the result. Sites detected as having been under positive selection are marked with a &#39;+&#39;, sites under purifying selection are marked with &#39;-&#39;. The number of symbols Number symbols Threshold 1 95% 2 99% 3 95% after adjustment 4 99% after adjustment</p>

<pre><code><code>  Occasionally the result may also contain an exclamation mark. This
  indicates that the observation at a site is not significantly
  different from random (equivalent to infinitely strong positive
  selection). This may indicate that the alignment at that site is bad</code></code></pre>

<p>Note</p>

<pre><code><code>  The following events are flagged:
  Synonymous            All codons at a site code for the same amino 
                        acid.
  Single character      Only one sequence at the site is ungapped,
                        the result of a recent insertion for example.
  All gaps              All sequences at a site contain a gap
                        character.

  Sites marked &quot;Single character&quot; or &quot;All gaps&quot; are not counted
  towards the number of sites for the purposes of correcting for
  multiple comparisons since it is not possible to detect selection
  from none or one observation under the assumptions made by the
  sitewise likelihood ratio test.</code></code></pre>

<h2 id="program_name">program_name</h2>

<pre><code><code> Title   : program_name
 Usage   : $factory-&gt;program_name()
 Function: holds the program name
 Returns:  string
 Args    : None</code></code></pre>

<h2 id="program_dir">program_dir</h2>

<pre><code><code> Title   : program_dir
 Usage   : -&gt;program_dir()
 Function: returns the program directory, obtained from ENV variable.
 Returns:  string
 Args    :</code></code></pre>

<h2 id="new">new</h2>

<pre><code><code> <span class="variable">Title</span>   <span class="operator">:</span> <span class="variable">new</span>
 <span class="variable">Usage</span>   <span class="operator">:</span> <span class="keyword">my</span> <span class="variable">$obj</span> <span class="operator">=</span> <span class="variable">Bio::Tools::Run::Phylo::SLR</span><span class="operator">-&gt;</span><span class="variable">new</span><span class="operator">();</span>
 <span class="variable">Function</span><span class="operator">:</span> <span class="variable">Builds</span> <span class="variable">a</span> <span class="variable">new</span> <span class="variable">Bio::Tools::Run::Phylo::SLR</span> <span class="variable">object</span> 
 <span class="variable">Returns</span> <span class="operator">:</span> <span class="variable">Bio::Tools::Run::Phylo::SLR</span>
 <span class="variable">Args</span>    <span class="operator">:</span> <span class="string">-alignment</span> <span class="operator">=&gt;</span> <span class="variable">the</span> <span class="variable">Bio::Align::AlignI</span> <span class="variable">object</span>
           <span class="operator">-</span><span class="string">save_tempfiles</span> <span class="operator">=&gt;</span> <span class="variable">boolean</span> <span class="variable">to</span> <span class="variable">save</span> <span class="variable">the</span> <span class="variable">generated</span> <span class="variable">tempfiles</span> <span class="keyword">and</span>
                              <span class="variable">NOT</span> <span class="variable">cleanup</span> <span class="variable">after</span> <span class="variable">onesself</span> <span class="operator">(</span><span class="keyword">default</span> <span class="variable">FALSE</span><span class="operator">)</span>
           <span class="string">-tree</span> <span class="operator">=&gt;</span> <span class="variable">the</span> <span class="variable">Bio::Tree::TreeI</span> <span class="variable">object</span>
           <span class="operator">-</span><span class="string">params</span> <span class="operator">=&gt;</span> <span class="variable">a</span> <span class="variable">hashref</span> <span class="variable">of</span> <span class="variable">SLR</span> <span class="variable">parameters</span> <span class="operator">(</span><span class="variable">all</span> <span class="variable">passed</span> <span class="variable">to</span> <span class="variable">set_parameter</span><span class="operator">)</span>
           <span class="string">-executable</span> <span class="operator">=&gt;</span> <span class="variable">where</span> <span class="variable">the</span> <span class="variable">SLR</span> <span class="variable">executable</span> <span class="variable">resides</span>
</code></code></pre>

<p>See also: <a href="../../../../../../lib/perl5/Bio/Tree/TreeI.html">Bio::Tree::TreeI</a>, <a href="../../../../../../lib/perl5/Bio/Align/AlignI.html">Bio::Align::AlignI</a></p>

<h2 id="prepare">prepare</h2>

<pre><code><code> <span class="variable">Title</span>   <span class="operator">:</span> <span class="variable">prepare</span>
 <span class="variable">Usage</span>   <span class="operator">:</span> <span class="keyword">my</span> <span class="variable">$rundir</span> <span class="operator">=</span> <span class="variable">$slr</span><span class="operator">-&gt;</span><span class="variable">prepare</span><span class="operator">(</span><span class="variable">$aln</span><span class="operator">);</span>
 <span class="variable">Function</span><span class="operator">:</span> <span class="variable">prepare</span> <span class="variable">the</span> <span class="variable">SLR</span> <span class="variable">analysis</span> <span class="variable">using</span> <span class="variable">the</span> <span class="keyword">default</span> <span class="keyword">or</span> <span class="variable">updated</span> <span class="variable">parameters</span>
           <span class="variable">the</span> <span class="variable">alignment</span> <span class="variable">parameter</span> <span class="variable">must</span> <span class="variable">have</span> <span class="variable">been</span> <span class="variable">set</span>
 <span class="variable">Returns</span> <span class="operator">:</span> <span class="variable">value</span> <span class="variable">of</span> <span class="variable">rundir</span>
 <span class="variable">Args</span>    <span class="operator">:</span> <span class="variable">L</span><span class="operator">&lt;</span><span class="variable">Bio::Align::AlignI</span><span class="operator">&gt;</span> <span class="variable">object</span><span class="operator">,</span>
           <span class="variable">L</span><span class="operator">&lt;</span><span class="variable">Bio::Tree::TreeI</span><span class="operator">&gt;</span> <span class="variable">object</span>
</code></code></pre>

<h2 id="run">run</h2>

<pre><code><code> <span class="variable">Title</span>   <span class="operator">:</span> <span class="variable">run</span>
 <span class="variable">Usage</span>   <span class="operator">:</span> <span class="keyword">my</span> <span class="operator">(</span><span class="variable">$rc</span><span class="operator">,</span><span class="variable">$parser</span><span class="operator">)</span> <span class="operator">=</span> <span class="variable">$slr</span><span class="operator">-&gt;</span><span class="variable">run</span><span class="operator">(</span><span class="variable">$aln</span><span class="operator">,</span><span class="variable">$tree</span><span class="operator">);</span>
 <span class="variable">Function</span><span class="operator">:</span> <span class="variable">run</span> <span class="variable">the</span> <span class="variable">SLR</span> <span class="variable">analysis</span> <span class="variable">using</span> <span class="variable">the</span> <span class="keyword">default</span> <span class="keyword">or</span> <span class="variable">updated</span> <span class="variable">parameters</span>
           <span class="variable">the</span> <span class="variable">alignment</span> <span class="variable">parameter</span> <span class="variable">must</span> <span class="variable">have</span> <span class="variable">been</span> <span class="variable">set</span>
 <span class="variable">Returns</span> <span class="operator">:</span> <span class="variable">Return</span> <span class="variable">code</span><span class="operator">,</span> <span class="variable">L</span><span class="operator">&lt;</span><span class="variable">Bio::Tools::Phylo::SLR</span><span class="operator">&gt;</span>
 <span class="variable">Args</span>    <span class="operator">:</span> <span class="variable">L</span><span class="operator">&lt;</span><span class="variable">Bio::Align::AlignI</span><span class="operator">&gt;</span> <span class="variable">object</span><span class="operator">,</span>
           <span class="variable">L</span><span class="operator">&lt;</span><span class="variable">Bio::Tree::TreeI</span><span class="operator">&gt;</span> <span class="variable">object</span>
</code></code></pre>

<h2 id="error_string">error_string</h2>

<pre><code><code> Title   : error_string
 Usage   : $obj-&gt;error_string($newval)
 Function: Where the output from the last analysus run is stored.
 Returns : value of error_string
 Args    : newvalue (optional)</code></code></pre>

<h2 id="alignment">alignment</h2>

<pre><code><code> <span class="variable">Title</span>   <span class="operator">:</span> <span class="variable">alignment</span>
 <span class="variable">Usage</span>   <span class="operator">:</span> <span class="variable">$slr</span><span class="operator">-&gt;</span><span class="variable">align</span><span class="operator">(</span><span class="variable">$aln</span><span class="operator">);</span>
 <span class="variable">Function</span><span class="operator">:</span> <span class="variable">Get</span><span class="operator">/</span><span class="variable">Set</span> <span class="variable">the</span> <span class="variable">L</span><span class="operator">&lt;</span><span class="variable">Bio::Align::AlignI</span><span class="operator">&gt;</span> <span class="variable">object</span>
 <span class="variable">Returns</span> <span class="operator">:</span> <span class="variable">L</span><span class="operator">&lt;</span><span class="variable">Bio::Align::AlignI</span><span class="operator">&gt;</span> <span class="variable">object</span>
 <span class="variable">Args</span>    <span class="operator">:</span> <span class="operator">[</span><span class="variable">optional</span><span class="operator">]</span> <span class="variable">L</span><span class="operator">&lt;</span><span class="variable">Bio::Align::AlignI</span><span class="operator">&gt;</span>
 <span class="variable">Comment</span> <span class="operator">:</span> <span class="variable">We</span> <span class="variable">could</span> <span class="variable">potentially</span> <span class="variable">add</span> <span class="variable">support</span> <span class="keyword">for</span> <span class="variable">running</span> <span class="variable">directly</span> <span class="variable">on</span> <span class="variable">a</span> <span class="variable">file</span>
           <span class="variable">but</span> <span class="variable">we</span> <span class="variable">shall</span> <span class="variable">keep</span> <span class="variable">it</span> <span class="variable">simple</span>
 <span class="variable">See</span> <span class="variable">also</span><span class="operator">:</span> <span class="variable">L</span><span class="operator">&lt;</span><span class="variable">Bio::SimpleAlign</span><span class="operator">&gt;</span>
</code></code></pre>

<h2 id="tree">tree</h2>

<pre><code><code> <span class="variable">Title</span>   <span class="operator">:</span> <span class="variable">tree</span>
 <span class="variable">Usage</span>   <span class="operator">:</span> <span class="variable">$slr</span><span class="operator">-&gt;</span><span class="variable">tree</span><span class="operator">(</span><span class="variable">$tree</span><span class="operator">,</span> <span class="variable">%params</span><span class="operator">);</span>
 <span class="variable">Function</span><span class="operator">:</span> <span class="variable">Get</span><span class="operator">/</span><span class="variable">Set</span> <span class="variable">the</span> <span class="variable">L</span><span class="operator">&lt;</span><span class="variable">Bio::Tree::TreeI</span><span class="operator">&gt;</span> <span class="variable">object</span>
 <span class="variable">Returns</span> <span class="operator">:</span> <span class="variable">L</span><span class="operator">&lt;</span><span class="variable">Bio::Tree::TreeI</span><span class="operator">&gt;</span> 
 <span class="variable">Args</span>    <span class="operator">:</span> <span class="operator">[</span><span class="variable">optional</span><span class="operator">]</span> <span class="variable">$tree</span> <span class="operator">=&gt;</span> <span class="variable">L</span><span class="operator">&lt;</span><span class="variable">Bio::Tree::TreeI</span><span class="operator">&gt;,</span>
 
 <span class="variable">Comment</span> <span class="operator">:</span> <span class="variable">We</span> <span class="variable">could</span> <span class="variable">potentially</span> <span class="variable">add</span> <span class="variable">support</span> <span class="keyword">for</span> <span class="variable">running</span> <span class="variable">directly</span> <span class="variable">on</span> <span class="variable">a</span> <span class="variable">file</span>
           <span class="variable">but</span> <span class="variable">we</span> <span class="variable">shall</span> <span class="variable">keep</span> <span class="variable">it</span> <span class="variable">simple</span>
 <span class="variable">See</span> <span class="variable">also</span><span class="operator">:</span> <span class="variable">L</span><span class="operator">&lt;</span><span class="variable">Bio::Tree::Tree</span><span class="operator">&gt;</span>
</code></code></pre>

<h2 id="get_parameters">get_parameters</h2>

<pre><code><code> <span class="variable">Title</span>   <span class="operator">:</span> <span class="variable">get_parameters</span>
 <span class="variable">Usage</span>   <span class="operator">:</span> <span class="keyword">my</span> <span class="variable">%params</span> <span class="operator">=</span> <span class="variable">$self</span><span class="operator">-&gt;</span><span class="variable">get_parameters</span><span class="operator">();</span>
 <span class="variable">Function</span><span class="operator">:</span> <span class="variable">returns</span> <span class="variable">the</span> <span class="variable">list</span> <span class="variable">of</span> <span class="variable">parameters</span> <span class="variable">as</span> <span class="variable">a</span> <span class="variable">hash</span>
 <span class="variable">Returns</span> <span class="operator">:</span> <span class="variable">associative</span> <span class="variable">array</span> <span class="variable">keyed</span> <span class="variable">on</span> <span class="variable">parameter</span> <span class="variable">names</span>
 <span class="variable">Args</span>    <span class="operator">:</span> <span class="variable">none</span>
</code></code></pre>

<h2 id="set_parameter">set_parameter</h2>

<pre><code><code> <span class="variable">Title</span>   <span class="operator">:</span> <span class="variable">set_parameter</span>
 <span class="variable">Usage</span>   <span class="operator">:</span> <span class="variable">$slr</span><span class="operator">-&gt;</span><span class="variable">set_parameter</span><span class="operator">(</span><span class="variable">$param</span><span class="operator">,</span><span class="variable">$val</span><span class="operator">);</span>
 <span class="variable">Function</span><span class="operator">:</span> <span class="variable">Sets</span> <span class="variable">a</span> <span class="variable">SLR</span> <span class="variable">parameter</span><span class="operator">,</span> <span class="variable">will</span> <span class="variable">be</span> <span class="variable">validated</span> <span class="variable">against</span>
           <span class="variable">the</span> <span class="variable">valid</span> <span class="keyword">values</span> <span class="variable">as</span> <span class="variable">set</span> <span class="variable">in</span> <span class="variable">the</span> <span class="variable">%VALIDVALUES</span> <span class="variable">class</span> <span class="variable">variable</span><span class="operator">.</span>  
           <span class="variable">The</span> <span class="variable">checks</span> <span class="variable">can</span> <span class="variable">be</span> <span class="variable">ignored</span> <span class="keyword">if</span> <span class="variable">one</span> <span class="variable">turns</span> <span class="variable">off</span> <span class="variable">param</span> <span class="variable">checks</span> <span class="variable">like</span> <span class="variable">this</span><span class="operator">:</span>
             <span class="variable">$slr</span><span class="operator">-&gt;</span><span class="variable">no_param_checks</span><span class="operator">(</span><span class="number">1</span><span class="operator">)</span>
 <span class="variable">Returns</span> <span class="operator">:</span> <span class="variable">boolean</span> <span class="keyword">if</span> <span class="variable">set</span> <span class="variable">was</span> <span class="variable">success</span><span class="operator">,</span> <span class="keyword">if</span> <span class="variable">verbose</span> <span class="variable">is</span> <span class="variable">set</span> <span class="variable">to</span> <span class="operator">-</span><span class="number">1</span>
           <span class="variable">then</span> <span class="keyword">no</span> <span class="variable">warning</span> <span class="variable">will</span> <span class="variable">be</span> <span class="variable">reported</span>
 <span class="variable">Args</span>    <span class="operator">:</span> <span class="variable">$param</span> <span class="operator">=&gt;</span> <span class="variable">name</span> <span class="variable">of</span> <span class="variable">the</span> <span class="variable">parameter</span>
           <span class="variable">$value</span> <span class="operator">=&gt;</span> <span class="variable">value</span> <span class="variable">to</span> <span class="variable">set</span> <span class="variable">the</span> <span class="variable">parameter</span> <span class="variable">to</span>
 <span class="variable">See</span> <span class="variable">also</span><span class="operator">:</span> <span class="variable">L</span><span class="operator">&lt;</span><span class="variable">no_param_checks</span><span class="operator">()&gt;</span>
</code></code></pre>

<h2 id="set_default_parameters">set_default_parameters</h2>

<pre><code><code> <span class="variable">Title</span>   <span class="operator">:</span> <span class="variable">set_default_parameters</span>
 <span class="variable">Usage</span>   <span class="operator">:</span> <span class="variable">$slr</span><span class="operator">-&gt;</span><span class="variable">set_default_parameters</span><span class="operator">(</span><span class="number">0</span><span class="operator">);</span>
 <span class="variable">Function</span><span class="operator">:</span> <span class="operator">(</span><span class="variable">Re</span><span class="operator">)</span><span class="variable">set</span> <span class="variable">the</span> <span class="keyword">default</span> <span class="variable">parameters</span> <span class="variable">from</span> <span class="variable">the</span> <span class="variable">defaults</span>
           <span class="operator">(</span><span class="variable">the</span> <span class="variable">first</span> <span class="variable">value</span> <span class="variable">in</span> <span class="keyword">each</span> <span class="variable">array</span> <span class="variable">in</span> <span class="variable">the</span> 
            <span class="variable">%VALIDVALUES</span> <span class="variable">class</span> <span class="variable">variable</span><span class="operator">)</span>
 <span class="variable">Returns</span> <span class="operator">:</span> <span class="variable">none</span>
 <span class="variable">Args</span>    <span class="operator">:</span> <span class="variable">boolean</span><span class="operator">:</span> <span class="variable">keep</span> <span class="variable">existing</span> <span class="variable">parameter</span> <span class="keyword">values</span>
</code></code></pre>

<h1 id="Bio::Tools::Run::WrapperBase-methods">Bio::Tools::Run::WrapperBase methods</h1>

<h2 id="no_param_checks">no_param_checks</h2>

<pre><code><code> Title   : no_param_checks
 Usage   : $obj-&gt;no_param_checks($newval)
 Function: Boolean flag as to whether or not we should
           trust the sanity checks for parameter values  
 Returns : value of no_param_checks
 Args    : newvalue (optional)</code></code></pre>

<h2 id="save_tempfiles">save_tempfiles</h2>

<pre><code><code> Title   : save_tempfiles
 Usage   : $obj-&gt;save_tempfiles($newval)
 Function: 
 Returns : value of save_tempfiles
 Args    : newvalue (optional)</code></code></pre>

<h2 id="outfile_name">outfile_name</h2>

<pre><code><code> <span class="variable">Title</span>   <span class="operator">:</span> <span class="variable">outfile_name</span>
 <span class="variable">Usage</span>   <span class="operator">:</span> <span class="keyword">my</span> <span class="variable">$outfile</span> <span class="operator">=</span> <span class="variable">$slr</span><span class="operator">-&gt;</span><span class="variable">outfile_name</span><span class="operator">();</span>
 <span class="variable">Function</span><span class="operator">:</span> <span class="variable">Get</span><span class="operator">/</span><span class="variable">Set</span> <span class="variable">the</span> <span class="variable">name</span> <span class="variable">of</span> <span class="variable">the</span> <span class="variable">output</span> <span class="variable">file</span> <span class="keyword">for</span> <span class="variable">this</span> <span class="variable">run</span>
           <span class="operator">(</span><span class="keyword">if</span> <span class="variable">you</span> <span class="variable">wanted</span> <span class="variable">to</span> <span class="keyword">do</span> <span class="variable">something</span> <span class="variable">special</span><span class="operator">)</span>
 <span class="variable">Returns</span> <span class="operator">:</span> <span class="variable">string</span>
 <span class="variable">Args</span>    <span class="operator">:</span> <span class="operator">[</span><span class="variable">optional</span><span class="operator">]</span> <span class="variable">string</span> <span class="variable">to</span> <span class="variable">set</span> <span class="variable">value</span> <span class="variable">to</span>
</code></code></pre>

<h2 id="tempdir">tempdir</h2>

<pre><code><code> <span class="variable">Title</span>   <span class="operator">:</span> <span class="variable">tempdir</span>
 <span class="variable">Usage</span>   <span class="operator">:</span> <span class="keyword">my</span> <span class="variable">$tmpdir</span> <span class="operator">=</span> <span class="variable">$self</span><span class="operator">-&gt;</span><span class="variable">tempdir</span><span class="operator">();</span>
 <span class="variable">Function</span><span class="operator">:</span> <span class="variable">Retrieve</span> <span class="variable">a</span> <span class="variable">temporary</span> <span class="variable">directory</span> <span class="variable">name</span> <span class="operator">(</span><span class="variable">which</span> <span class="variable">is</span> <span class="variable">created</span><span class="operator">)</span>
 <span class="variable">Returns</span> <span class="operator">:</span> <span class="variable">string</span> <span class="variable">which</span> <span class="variable">is</span> <span class="variable">the</span> <span class="variable">name</span> <span class="variable">of</span> <span class="variable">the</span> <span class="variable">temporary</span> <span class="variable">directory</span>
 <span class="variable">Args</span>    <span class="operator">:</span> <span class="variable">none</span>
</code></code></pre>

<h2 id="cleanup">cleanup</h2>

<pre><code><code> <span class="variable">Title</span>   <span class="operator">:</span> <span class="variable">cleanup</span>
 <span class="variable">Usage</span>   <span class="operator">:</span> <span class="variable">$slr</span><span class="operator">-&gt;</span><span class="variable">cleanup</span><span class="operator">();</span>
 <span class="variable">Function</span><span class="operator">:</span> <span class="variable">Will</span> <span class="variable">cleanup</span> <span class="variable">the</span> <span class="variable">tempdir</span> <span class="variable">directory</span> <span class="variable">after</span> <span class="variable">an</span> <span class="variable">SLR</span> <span class="variable">run</span>
 <span class="variable">Returns</span> <span class="operator">:</span> <span class="variable">none</span>
 <span class="variable">Args</span>    <span class="operator">:</span> <span class="variable">none</span>
</code></code></pre>

<h2 id="io">io</h2>

<pre><code><code> Title   : io
 Usage   : $obj-&gt;io($newval)
 Function:  Gets a L&lt;Bio::Root::IO&gt; object
 Returns : L&lt;Bio::Root::IO&gt;
 Args    : none</code></code></pre>


</body>

</html>


