<?xml version="1.0" ?>
<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.0 Transitional//EN">
<html>
<head>
<!-- saved from url=(0017)http://localhost/ -->
<script language="JavaScript" src="../../../../../../displayToc.js"></script>
<script language="JavaScript" src="../../../../../../tocParas.js"></script>
<script language="JavaScript" src="../../../../../../tocTab.js"></script>
<link rel="stylesheet" type="text/css" href="../../../../../../scineplex.css">
<title></title>
<link rel="stylesheet" href="../../../../../../Active.css" type="text/css" />
<meta http-equiv="content-type" content="text/html; charset=utf-8" />
<link rev="made" href="mailto:support@ActiveState.com" />
</head>

<body>



<ul id="index">
  <li><a href="#NAME">NAME</a></li>
  <li><a href="#SYNOPSIS">SYNOPSIS</a></li>
  <li><a href="#DESCRIPTION">DESCRIPTION</a>
    <ul>
      <li><a href="#Using-the-berkeleydb-adaptor">Using the berkeleydb adaptor</a></li>
    </ul>
  </li>
  <li><a href="#BUGS">BUGS</a></li>
  <li><a href="#SEE-ALSO">SEE ALSO</a></li>
  <li><a href="#AUTHOR">AUTHOR</a></li>
</ul>

<h1 id="NAME">NAME</h1>

<p>Bio::DB::SeqFeature::Store::berkeleydb -- Storage and retrieval of sequence annotation data in Berkeleydb files</p>

<h1 id="SYNOPSIS">SYNOPSIS</h1>

<pre><code><code>  <span class="keyword">use</span> <span class="variable">Bio::DB::SeqFeature::Store</span><span class="operator">;</span>
  
  <span class="comment"># Create a database from the feature files located in /home/fly4.3 and store</span>
  <span class="comment"># the database index in the same directory:</span>
  <span class="variable">$db</span> <span class="operator">=</span>  <span class="variable">Bio::DB::SeqFeature::Store</span><span class="operator">-&gt;</span><span class="variable">new</span><span class="operator">(</span> <span class="string">-adaptor</span> <span class="operator">=&gt;</span> <span class="string">'berkeleydb'</span><span class="operator">,</span>
                                          <span class="string">-dir</span>     <span class="operator">=&gt;</span> <span class="string">'/home/fly4.3'</span><span class="operator">);</span>
  
  <span class="comment"># Create a database that will monitor the files in /home/fly4.3, but store</span>
  <span class="comment"># the indexes in /var/databases/fly4.3</span>
  <span class="variable">$db</span>      <span class="operator">=</span> <span class="variable">Bio::DB::SeqFeature::Store</span><span class="operator">-&gt;</span><span class="variable">new</span><span class="operator">(</span> <span class="string">-adaptor</span>    <span class="operator">=&gt;</span> <span class="string">'berkeleydb'</span><span class="operator">,</span>
                                              <span class="string">-dsn</span>        <span class="operator">=&gt;</span> <span class="string">'/var/databases/fly4.3'</span><span class="operator">,</span>
                                              <span class="string">-dir</span>        <span class="operator">=&gt;</span> <span class="string">'/home/fly4.3'</span><span class="operator">);</span>
  
  <span class="comment"># Create a feature database from scratch</span>
  <span class="variable">$db</span>     <span class="operator">=</span> <span class="variable">Bio::DB::SeqFeature::Store</span><span class="operator">-&gt;</span><span class="variable">new</span><span class="operator">(</span> <span class="string">-adaptor</span> <span class="operator">=&gt;</span> <span class="string">'berkeleydb'</span><span class="operator">,</span>
                                             <span class="string">-dsn</span>     <span class="operator">=&gt;</span> <span class="string">'/var/databases/fly4.3'</span><span class="operator">,</span>
                                             <span class="string">-create</span>  <span class="operator">=&gt;</span> <span class="number">1</span><span class="operator">);</span>
  
  <span class="comment"># get a feature from somewhere</span>
  <span class="keyword">my</span> <span class="variable">$feature</span> <span class="operator">=</span> <span class="variable">Bio::SeqFeature::Generic</span><span class="operator">-&gt;</span><span class="variable">new</span><span class="operator">(...);</span>
  
  <span class="comment"># store it</span>
  <span class="variable">$db</span><span class="operator">-&gt;</span><span class="variable">store</span><span class="operator">(</span><span class="variable">$feature</span><span class="operator">)</span> <span class="keyword">or</span> <span class="keyword">die</span> <span class="string">"Couldn't store!"</span><span class="operator">;</span>
  
  <span class="comment"># primary ID of the feature is changed to indicate its primary ID</span>
  <span class="comment"># in the database...</span>
  <span class="keyword">my</span> <span class="variable">$id</span> <span class="operator">=</span> <span class="variable">$feature</span><span class="operator">-&gt;</span><span class="variable">primary_id</span><span class="operator">;</span>
  
  <span class="comment"># get the feature back out</span>
  <span class="keyword">my</span> <span class="variable">$f</span>  <span class="operator">=</span> <span class="variable">$db</span><span class="operator">-&gt;</span><span class="variable">fetch</span><span class="operator">(</span><span class="variable">$id</span><span class="operator">);</span>
  
  <span class="comment"># change the feature and update it</span>
  <span class="variable">$f</span><span class="operator">-&gt;</span><span class="variable">start</span><span class="operator">(</span><span class="number">100</span><span class="operator">);</span>
  <span class="variable">$db</span><span class="operator">-&gt;</span><span class="variable">update</span><span class="operator">(</span><span class="variable">$f</span><span class="operator">)</span> <span class="keyword">or</span> <span class="variable">$self</span><span class="operator">-&gt;</span><span class="variable">throw</span><span class="operator">(</span><span class="string">"Couldn't update!"</span><span class="operator">);</span>
  
  <span class="comment"># use the GFF3 loader to do a bulk write:</span>
  <span class="keyword">my</span> <span class="variable">$loader</span> <span class="operator">=</span> <span class="variable">Bio::DB::SeqFeature::Store::GFF3Loader</span><span class="operator">-&gt;</span><span class="variable">new</span><span class="operator">(</span><span class="string">-store</span>   <span class="operator">=&gt;</span> <span class="variable">$db</span><span class="operator">,</span>
                                                           <span class="string">-verbose</span> <span class="operator">=&gt;</span> <span class="number">0</span><span class="operator">,</span>
                                                           <span class="string">-fast</span>    <span class="operator">=&gt;</span> <span class="number">1</span><span class="operator">);</span>
  <span class="variable">$loader</span><span class="operator">-&gt;</span><span class="variable">load</span><span class="operator">(</span><span class="string">'/home/fly4.3/dmel-all.gff'</span><span class="operator">);</span>
  
  
  <span class="comment"># searching...</span>
  <span class="comment"># ...by id</span>
  <span class="keyword">my</span> <span class="variable">@features</span> <span class="operator">=</span> <span class="variable">$db</span><span class="operator">-&gt;</span><span class="variable">fetch_many</span><span class="operator">(</span><span class="variable">@list_of_ids</span><span class="operator">);</span>
  
  <span class="comment"># ...by name</span>
  <span class="variable">@features</span> <span class="operator">=</span> <span class="variable">$db</span><span class="operator">-&gt;</span><span class="variable">get_features_by_name</span><span class="operator">(</span><span class="string">'ZK909'</span><span class="operator">);</span>
  
  <span class="comment"># ...by alias</span>
  <span class="variable">@features</span> <span class="operator">=</span> <span class="variable">$db</span><span class="operator">-&gt;</span><span class="variable">get_features_by_alias</span><span class="operator">(</span><span class="string">'sma-3'</span><span class="operator">);</span>
  
  <span class="comment"># ...by type</span>
  <span class="variable">@features</span> <span class="operator">=</span> <span class="variable">$db</span><span class="operator">-&gt;</span><span class="variable">get_features_by_type</span><span class="operator">(</span><span class="string">'gene'</span><span class="operator">);</span>
  
  <span class="comment"># ...by location</span>
  <span class="variable">@features</span> <span class="operator">=</span> <span class="variable">$db</span><span class="operator">-&gt;</span><span class="variable">get_features_by_location</span><span class="operator">(</span><span class="string">-seq_id</span><span class="operator">=&gt;</span><span class="string">'Chr1'</span><span class="operator">,</span><span class="string">-start</span><span class="operator">=&gt;</span><span class="number">4000</span><span class="operator">,</span><span class="string">-end</span><span class="operator">=&gt;</span><span class="number">600000</span><span class="operator">);</span>
  
  <span class="comment"># ...by attribute</span>
  <span class="variable">@features</span> <span class="operator">=</span> <span class="variable">$db</span><span class="operator">-&gt;</span><span class="variable">get_features_by_attribute</span><span class="operator">(</span><span class="operator">{</span><span class="string">description</span> <span class="operator">=&gt;</span> <span class="string">'protein kinase'</span><span class="operator">}</span><span class="operator">)</span>
  
  <span class="comment"># ...by the GFF "Note" field</span>
  <span class="variable">@result_list</span> <span class="operator">=</span> <span class="variable">$db</span><span class="operator">-&gt;</span><span class="variable">search_notes</span><span class="operator">(</span><span class="string">'kinase'</span><span class="operator">);</span>
  
  <span class="comment"># ...by arbitrary combinations of selectors</span>
  <span class="variable">@features</span> <span class="operator">=</span> <span class="variable">$db</span><span class="operator">-&gt;</span><span class="variable">features</span><span class="operator">(</span><span class="string">-name</span> <span class="operator">=&gt;</span> <span class="variable">$name</span><span class="operator">,</span>
                            <span class="string">-type</span> <span class="operator">=&gt;</span> <span class="variable">$types</span><span class="operator">,</span>
                            <span class="string">-seq_id</span> <span class="operator">=&gt;</span> <span class="variable">$seqid</span><span class="operator">,</span>
                            <span class="string">-start</span>  <span class="operator">=&gt;</span> <span class="variable">$start</span><span class="operator">,</span>
                            <span class="string">-end</span>    <span class="operator">=&gt;</span> <span class="variable">$end</span><span class="operator">,</span>
                            <span class="string">-attributes</span> <span class="operator">=&gt;</span> <span class="variable">$attributes</span><span class="operator">);</span>
  
  <span class="comment"># ...using an iterator</span>
  <span class="keyword">my</span> <span class="variable">$iterator</span> <span class="operator">=</span> <span class="variable">$db</span><span class="operator">-&gt;</span><span class="variable">get_seq_stream</span><span class="operator">(</span><span class="string">-name</span> <span class="operator">=&gt;</span> <span class="variable">$name</span><span class="operator">,</span>
                                     <span class="string">-type</span> <span class="operator">=&gt;</span> <span class="variable">$types</span><span class="operator">,</span>
                                     <span class="string">-seq_id</span> <span class="operator">=&gt;</span> <span class="variable">$seqid</span><span class="operator">,</span>
                                     <span class="string">-start</span>  <span class="operator">=&gt;</span> <span class="variable">$start</span><span class="operator">,</span>
                                     <span class="string">-end</span>    <span class="operator">=&gt;</span> <span class="variable">$end</span><span class="operator">,</span>
                                     <span class="string">-attributes</span> <span class="operator">=&gt;</span> <span class="variable">$attributes</span><span class="operator">);</span>
  
  <span class="keyword">while</span> <span class="operator">(</span><span class="keyword">my</span> <span class="variable">$feature</span> <span class="operator">=</span> <span class="variable">$iterator</span><span class="operator">-&gt;</span><span class="variable">next_seq</span><span class="operator">)</span> <span class="operator">{</span>
    <span class="comment"># do something with the feature</span>
  <span class="operator">}</span>
  
  <span class="comment"># ...limiting the search to a particular region</span>
  <span class="keyword">my</span> <span class="variable">$segment</span>  <span class="operator">=</span> <span class="variable">$db</span><span class="operator">-&gt;</span><span class="variable">segment</span><span class="operator">(</span><span class="string">'Chr1'</span><span class="operator">,</span><span class="number">5000</span><span class="operator">=&gt;</span><span class="number">6000</span><span class="operator">);</span>
  <span class="keyword">my</span> <span class="variable">@features</span> <span class="operator">=</span> <span class="variable">$segment</span><span class="operator">-&gt;</span><span class="variable">features</span><span class="operator">(</span><span class="string">-type</span><span class="operator">=&gt;</span><span class="operator">[</span><span class="string">'mRNA'</span><span class="operator">,</span><span class="string">'match'</span><span class="operator">]</span><span class="operator">);</span>
  
  <span class="comment"># what feature types are defined in the database?</span>
  <span class="keyword">my</span> <span class="variable">@types</span>    <span class="operator">=</span> <span class="variable">$db</span><span class="operator">-&gt;</span><span class="variable">types</span><span class="operator">;</span>
  
  <span class="comment"># getting &amp; storing sequence information</span>
  <span class="comment"># Warning: this returns a string, and not a PrimarySeq object</span>
  <span class="variable">$db</span><span class="operator">-&gt;</span><span class="variable">insert_sequence</span><span class="operator">(</span><span class="string">'Chr1'</span><span class="operator">,</span><span class="string">'GATCCCCCGGGATTCCAAAA...'</span><span class="operator">);</span>
  <span class="keyword">my</span> <span class="variable">$sequence</span> <span class="operator">=</span> <span class="variable">$db</span><span class="operator">-&gt;</span><span class="variable">fetch_sequence</span><span class="operator">(</span><span class="string">'Chr1'</span><span class="operator">,</span><span class="number">5000</span><span class="operator">=&gt;</span><span class="number">6000</span><span class="operator">);</span>
  
  <span class="comment"># create a new feature in the database</span>
  <span class="keyword">my</span> <span class="variable">$feature</span> <span class="operator">=</span> <span class="variable">$db</span><span class="operator">-&gt;</span><span class="variable">new_feature</span><span class="operator">(</span><span class="string">-primary_tag</span> <span class="operator">=&gt;</span> <span class="string">'mRNA'</span><span class="operator">,</span>
                                 <span class="string">-seq_id</span>      <span class="operator">=&gt;</span> <span class="string">'chr3'</span><span class="operator">,</span>
                                 <span class="string">-start</span>      <span class="operator">=&gt;</span> <span class="number">10000</span><span class="operator">,</span>
                                 <span class="string">-end</span>        <span class="operator">=&gt;</span> <span class="number">11000</span><span class="operator">);</span>
</code></code></pre>

<h1 id="DESCRIPTION">DESCRIPTION</h1>

<p>Bio::DB::SeqFeature::Store::berkeleydb is the Berkeleydb adaptor for Bio::DB::SeqFeature::Store. You will not create it directly, but instead use Bio::DB::SeqFeature::Store-&gt;new() to do so.</p>

<p>See <a href="../../../../../../../.cpan/build/BioPerl-1.6.901-ghKY8U/blib/lib/Bio/DB/SeqFeature/Store.html">Bio::DB::SeqFeature::Store</a> for complete usage instructions.</p>

<h2 id="Using-the-berkeleydb-adaptor">Using the berkeleydb adaptor</h2>

<p>The Berkeley database consists of a series of Berkeleydb index files, and a couple of special purpose indexes. You can create the index files from scratch by creating a new database and calling new_feature() repeatedly, you can create the database and then bulk populate it using the GFF3 loader, or you can monitor a directory of preexisting GFF3 and FASTA files and rebuild the indexes whenever one or more of the fields changes. The last mode is probably the most convenient. Note that the indexer will only pay attention to files that end with .gff3, .wig and .fa.</p>

<dl>

<dt id="The-new-constructor">The new() constructor</dt>
<dd>

<p>The new() constructor method all the arguments recognized by Bio::DB::SeqFeature::Store, and a few additional ones.</p>

<p>Standard arguments:</p>

<pre><code><code> Name               Value
 ----               -----

 -adaptor           The name of the Adaptor class (default DBI::mysql)

 -serializer        The name of the serializer class (default Storable)

 -index_subfeatures Whether or not to make subfeatures searchable
                    (default true)

 -cache             Activate LRU caching feature -- size of cache

 -compress          Compresses features before storing them in database
                    using Compress::Zlib</code></code></pre>

<p>Adaptor-specific arguments</p>

<pre><code><code> Name               Value
 ----               -----

 -dsn               Where the index files are stored

 -dir               Where the source (GFF3, FASTA) files are stored

 -autoindex         An alias for -dir.

 -write             Pass true to open the index files for writing.

 -create            Pass true to create the index files if they don&#39;t exist
                    (implies -write=&gt;1)

 -locking           Use advisory locking to avoid one process trying to read
                    from the database while another is updating it (may not
                    work properly over NFS).

 -temp              Pass true to create temporary index files that will
                    be deleted once the script exits.

 -verbose           Pass true to report autoindexing operations on STDERR.
                    (default is true).</code></code></pre>

<p>Examples:</p>

<p>To create an empty database which will be populated using calls to store() or new_feature(), or which will be bulk-loaded using the GFF3 loader:</p>

<pre><code><code>  <span class="variable">$db</span>     <span class="operator">=</span> <span class="variable">Bio::DB::SeqFeature::Store</span><span class="operator">-&gt;</span><span class="variable">new</span><span class="operator">(</span> <span class="string">-adaptor</span> <span class="operator">=&gt;</span> <span class="string">'berkeleydb'</span><span class="operator">,</span>
                                             <span class="string">-dsn</span>     <span class="operator">=&gt;</span> <span class="string">'/var/databases/fly4.3'</span><span class="operator">,</span>
                                             <span class="string">-create</span>  <span class="operator">=&gt;</span> <span class="number">1</span><span class="operator">);</span>
</code></code></pre>

<p>To open a preexisting database in read-only mode:</p>

<pre><code><code>  <span class="variable">$db</span>     <span class="operator">=</span> <span class="variable">Bio::DB::SeqFeature::Store</span><span class="operator">-&gt;</span><span class="variable">new</span><span class="operator">(</span> <span class="string">-adaptor</span> <span class="operator">=&gt;</span> <span class="string">'berkeleydb'</span><span class="operator">,</span>
                                             <span class="string">-dsn</span>     <span class="operator">=&gt;</span> <span class="string">'/var/databases/fly4.3'</span><span class="operator">);</span>
</code></code></pre>

<p>To open a preexisting database in read/write (update) mode:</p>

<pre><code><code>  <span class="variable">$db</span>     <span class="operator">=</span> <span class="variable">Bio::DB::SeqFeature::Store</span><span class="operator">-&gt;</span><span class="variable">new</span><span class="operator">(</span> <span class="string">-adaptor</span> <span class="operator">=&gt;</span> <span class="string">'berkeleydb'</span><span class="operator">,</span>
                                             <span class="string">-dsn</span>     <span class="operator">=&gt;</span> <span class="string">'/var/databases/fly4.3'</span><span class="operator">,</span>
                                             <span class="string">-write</span>   <span class="operator">=&gt;</span> <span class="number">1</span><span class="operator">);</span>
</code></code></pre>

<p>To monitor a set of GFF3 and FASTA files located in a directory and create/update the database indexes as needed. The indexes will be stored in a new subdirectory named &quot;indexes&quot;:</p>

<pre><code><code>  <span class="variable">$db</span>     <span class="operator">=</span> <span class="variable">Bio::DB::SeqFeature::Store</span><span class="operator">-&gt;</span><span class="variable">new</span><span class="operator">(</span> <span class="string">-adaptor</span> <span class="operator">=&gt;</span> <span class="string">'berkeleydb'</span><span class="operator">,</span>
                                             <span class="string">-dir</span>     <span class="operator">=&gt;</span> <span class="string">'/var/databases/fly4.3'</span><span class="operator">);</span>
</code></code></pre>

<p>As above, but store the source files and index files in separate directories:</p>

<pre><code><code>  <span class="variable">$db</span>     <span class="operator">=</span> <span class="variable">Bio::DB::SeqFeature::Store</span><span class="operator">-&gt;</span><span class="variable">new</span><span class="operator">(</span> <span class="string">-adaptor</span> <span class="operator">=&gt;</span> <span class="string">'berkeleydb'</span><span class="operator">,</span>
                                             <span class="string">-dsn</span>     <span class="operator">=&gt;</span> <span class="string">'/var/databases/fly4.3'</span><span class="operator">,</span>
                                             <span class="string">-dir</span>     <span class="operator">=&gt;</span> <span class="string">'/home/gff3_files/fly4.3'</span><span class="operator">);</span>
</code></code></pre>

<p>To be indexed, files must end with one of .gff3 (GFF3 format), .fa (FASTA format) or .wig (WIG format).</p>

<p><b>-autoindex</b> is an alias for <b>-dir</b>.</p>

<p>You should specify <b>-locking</b> in a multiuser environment, including the case in which the database is being used by a web server at the same time another user might be updating it.</p>

</dd>
</dl>

<p>See <a href="../../../../../../../.cpan/build/BioPerl-1.6.901-ghKY8U/blib/lib/Bio/DB/SeqFeature/Store.html">Bio::DB::SeqFeature::Store</a> for all the access methods supported by this adaptor. The various methods for storing and updating features and sequences into the database are supported, but there is no locking. If two processes try to update the same database simultaneously, the database will likely become corrupted.</p>

<h1 id="BUGS">BUGS</h1>

<p>This is an early version, so there are certainly some bugs. Please use the BioPerl bug tracking system to report bugs.</p>

<h1 id="SEE-ALSO">SEE ALSO</h1>

<p><a>bioperl</a>, <a href="../../../../../../../.cpan/build/BioPerl-1.6.901-ghKY8U/blib/lib/Bio/DB/SeqFeature.html">Bio::DB::SeqFeature</a>, <a href="../../../../../../../.cpan/build/BioPerl-1.6.901-ghKY8U/blib/lib/Bio/DB/SeqFeature/Store.html">Bio::DB::SeqFeature::Store</a>, <a>Bio::DB::SeqFeature::GFF3Loader</a>, <a href="../../../../../../../.cpan/build/BioPerl-1.6.901-ghKY8U/blib/lib/Bio/DB/SeqFeature/Segment.html">Bio::DB::SeqFeature::Segment</a>, <a href="../../../../../../../.cpan/build/BioPerl-1.6.901-ghKY8U/blib/lib/Bio/DB/SeqFeature/Store/memory.html">Bio::DB::SeqFeature::Store::memory</a>, <a href="../../../../../../../.cpan/build/BioPerl-1.6.901-ghKY8U/blib/lib/Bio/DB/SeqFeature/Store/DBI/mysql.html">Bio::DB::SeqFeature::Store::DBI::mysql</a>,</p>

<h1 id="AUTHOR">AUTHOR</h1>

<p>Lincoln Stein &lt;lstein@cshl.org&gt;.</p>

<p>Copyright (c) 2006 Cold Spring Harbor Laboratory.</p>

<p>This library is free software; you can redistribute it and/or modify it under the same terms as Perl itself.</p>


</body>

</html>


