<?xml version="1.0" ?>
<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.0 Transitional//EN">
<html>
<head>
<!-- saved from url=(0017)http://localhost/ -->
<script language="JavaScript" src="../../../../../../displayToc.js"></script>
<script language="JavaScript" src="../../../../../../tocParas.js"></script>
<script language="JavaScript" src="../../../../../../tocTab.js"></script>
<link rel="stylesheet" type="text/css" href="../../../../../../scineplex.css">
<title></title>
<link rel="stylesheet" href="../../../../../../Active.css" type="text/css" />
<meta http-equiv="content-type" content="text/html; charset=utf-8" />
<link rev="made" href="mailto:support@ActiveState.com" />
</head>

<body>



<ul id="index">
  <li><a href="#NAME">NAME</a></li>
  <li><a href="#SYNOPSIS">SYNOPSIS</a></li>
  <li><a href="#DESCRIPTION">DESCRIPTION</a>
    <ul>
      <li><a href="#Using-the-memory-adaptor">Using the memory adaptor</a></li>
      <li><a href="#types">types</a></li>
    </ul>
  </li>
  <li><a href="#BUGS">BUGS</a></li>
  <li><a href="#SEE-ALSO">SEE ALSO</a></li>
  <li><a href="#AUTHOR">AUTHOR</a></li>
</ul>

<h1 id="NAME">NAME</h1>

<p>Bio::DB::SeqFeature::Store::memory -- In-memory implementation of Bio::DB::SeqFeature::Store</p>

<h1 id="SYNOPSIS">SYNOPSIS</h1>

<pre><code><code>  <span class="keyword">use</span> <span class="variable">Bio::DB::SeqFeature::Store</span><span class="operator">;</span>
  
  <span class="comment"># Open the sequence database</span>
  <span class="keyword">my</span> <span class="variable">$db</span>      <span class="operator">=</span> <span class="variable">Bio::DB::SeqFeature::Store</span><span class="operator">-&gt;</span><span class="variable">new</span><span class="operator">(</span> <span class="string">-adaptor</span> <span class="operator">=&gt;</span> <span class="string">'memory'</span><span class="operator">,</span>
                                                 <span class="string">-dsn</span>     <span class="operator">=&gt;</span> <span class="string">'/var/databases/test'</span><span class="operator">);</span>
  <span class="comment"># search... by id</span>
  <span class="keyword">my</span> <span class="variable">@features</span> <span class="operator">=</span> <span class="variable">$db</span><span class="operator">-&gt;</span><span class="variable">fetch_many</span><span class="operator">(</span><span class="variable">@list_of_ids</span><span class="operator">);</span>
  
  <span class="comment"># ...by name</span>
  <span class="variable">@features</span> <span class="operator">=</span> <span class="variable">$db</span><span class="operator">-&gt;</span><span class="variable">get_features_by_name</span><span class="operator">(</span><span class="string">'ZK909'</span><span class="operator">);</span>
  
  <span class="comment"># ...by alias</span>
  <span class="variable">@features</span> <span class="operator">=</span> <span class="variable">$db</span><span class="operator">-&gt;</span><span class="variable">get_features_by_alias</span><span class="operator">(</span><span class="string">'sma-3'</span><span class="operator">);</span>
  
  <span class="comment"># ...by type</span>
  <span class="variable">@features</span> <span class="operator">=</span> <span class="variable">$db</span><span class="operator">-&gt;</span><span class="variable">get_features_by_type</span><span class="operator">(</span><span class="string">'gene'</span><span class="operator">);</span>
  
  <span class="comment"># ...by location</span>
  <span class="variable">@features</span> <span class="operator">=</span> <span class="variable">$db</span><span class="operator">-&gt;</span><span class="variable">get_features_by_location</span><span class="operator">(</span><span class="string">-seq_id</span><span class="operator">=&gt;</span><span class="string">'Chr1'</span><span class="operator">,</span><span class="string">-start</span><span class="operator">=&gt;</span><span class="number">4000</span><span class="operator">,</span><span class="string">-end</span><span class="operator">=&gt;</span><span class="number">600000</span><span class="operator">);</span>
  
  <span class="comment"># ...by attribute</span>
  <span class="variable">@features</span> <span class="operator">=</span> <span class="variable">$db</span><span class="operator">-&gt;</span><span class="variable">get_features_by_attribute</span><span class="operator">(</span><span class="operator">{</span><span class="string">description</span> <span class="operator">=&gt;</span> <span class="string">'protein kinase'</span><span class="operator">}</span><span class="operator">)</span>
  
  <span class="comment"># ...by the GFF "Note" field</span>
  <span class="variable">@result_list</span> <span class="operator">=</span> <span class="variable">$db</span><span class="operator">-&gt;</span><span class="variable">search_notes</span><span class="operator">(</span><span class="string">'kinase'</span><span class="operator">);</span>
  
  <span class="comment"># ...by arbitrary combinations of selectors</span>
  <span class="variable">@features</span> <span class="operator">=</span> <span class="variable">$db</span><span class="operator">-&gt;</span><span class="variable">features</span><span class="operator">(</span><span class="string">-name</span> <span class="operator">=&gt;</span> <span class="variable">$name</span><span class="operator">,</span>
                            <span class="string">-type</span> <span class="operator">=&gt;</span> <span class="variable">$types</span><span class="operator">,</span>
                            <span class="string">-seq_id</span> <span class="operator">=&gt;</span> <span class="variable">$seqid</span><span class="operator">,</span>
                            <span class="string">-start</span>  <span class="operator">=&gt;</span> <span class="variable">$start</span><span class="operator">,</span>
                            <span class="string">-end</span>    <span class="operator">=&gt;</span> <span class="variable">$end</span><span class="operator">,</span>
                            <span class="string">-attributes</span> <span class="operator">=&gt;</span> <span class="variable">$attributes</span><span class="operator">);</span>
  
  <span class="comment"># ...using an iterator</span>
  <span class="keyword">my</span> <span class="variable">$iterator</span> <span class="operator">=</span> <span class="variable">$db</span><span class="operator">-&gt;</span><span class="variable">get_seq_stream</span><span class="operator">(</span><span class="string">-name</span> <span class="operator">=&gt;</span> <span class="variable">$name</span><span class="operator">,</span>
                                     <span class="string">-type</span> <span class="operator">=&gt;</span> <span class="variable">$types</span><span class="operator">,</span>
                                     <span class="string">-seq_id</span> <span class="operator">=&gt;</span> <span class="variable">$seqid</span><span class="operator">,</span>
                                     <span class="string">-start</span>  <span class="operator">=&gt;</span> <span class="variable">$start</span><span class="operator">,</span>
                                     <span class="string">-end</span>    <span class="operator">=&gt;</span> <span class="variable">$end</span><span class="operator">,</span>
                                     <span class="string">-attributes</span> <span class="operator">=&gt;</span> <span class="variable">$attributes</span><span class="operator">);</span>
  
  <span class="keyword">while</span> <span class="operator">(</span><span class="keyword">my</span> <span class="variable">$feature</span> <span class="operator">=</span> <span class="variable">$iterator</span><span class="operator">-&gt;</span><span class="variable">next_seq</span><span class="operator">)</span> <span class="operator">{</span>
    <span class="comment"># do something with the feature</span>
  <span class="operator">}</span>
  
  <span class="comment"># ...limiting the search to a particular region</span>
  <span class="keyword">my</span> <span class="variable">$segment</span>  <span class="operator">=</span> <span class="variable">$db</span><span class="operator">-&gt;</span><span class="variable">segment</span><span class="operator">(</span><span class="string">'Chr1'</span><span class="operator">,</span><span class="number">5000</span><span class="operator">=&gt;</span><span class="number">6000</span><span class="operator">);</span>
  <span class="keyword">my</span> <span class="variable">@features</span> <span class="operator">=</span> <span class="variable">$segment</span><span class="operator">-&gt;</span><span class="variable">features</span><span class="operator">(</span><span class="string">-type</span><span class="operator">=&gt;</span><span class="operator">[</span><span class="string">'mRNA'</span><span class="operator">,</span><span class="string">'match'</span><span class="operator">]</span><span class="operator">);</span>
  
  <span class="comment"># getting &amp; storing sequence information</span>
  <span class="comment"># Warning: this returns a string, and not a PrimarySeq object</span>
  <span class="variable">$db</span><span class="operator">-&gt;</span><span class="variable">insert_sequence</span><span class="operator">(</span><span class="string">'Chr1'</span><span class="operator">,</span><span class="string">'GATCCCCCGGGATTCCAAAA...'</span><span class="operator">);</span>
  <span class="keyword">my</span> <span class="variable">$sequence</span> <span class="operator">=</span> <span class="variable">$db</span><span class="operator">-&gt;</span><span class="variable">fetch_sequence</span><span class="operator">(</span><span class="string">'Chr1'</span><span class="operator">,</span><span class="number">5000</span><span class="operator">=&gt;</span><span class="number">6000</span><span class="operator">);</span>
  
  <span class="comment"># what feature types are defined in the database?</span>
  <span class="keyword">my</span> <span class="variable">@types</span>    <span class="operator">=</span> <span class="variable">$db</span><span class="operator">-&gt;</span><span class="variable">types</span><span class="operator">;</span>
  
  <span class="comment"># create a new feature in the database</span>
  <span class="keyword">my</span> <span class="variable">$feature</span> <span class="operator">=</span> <span class="variable">$db</span><span class="operator">-&gt;</span><span class="variable">new_feature</span><span class="operator">(</span><span class="string">-primary_tag</span> <span class="operator">=&gt;</span> <span class="string">'mRNA'</span><span class="operator">,</span>
                                 <span class="string">-seq_id</span>      <span class="operator">=&gt;</span> <span class="string">'chr3'</span><span class="operator">,</span>
                                 <span class="string">-start</span>      <span class="operator">=&gt;</span> <span class="number">10000</span><span class="operator">,</span>
                                 <span class="string">-end</span>        <span class="operator">=&gt;</span> <span class="number">11000</span><span class="operator">);</span>
</code></code></pre>

<h1 id="DESCRIPTION">DESCRIPTION</h1>

<p>Bio::DB::SeqFeature::Store::memory is the in-memory adaptor for Bio::DB::SeqFeature::Store. You will not create it directly, but instead use Bio::DB::SeqFeature::Store-&gt;new() to do so.</p>

<p>See <a href="../../../../../../../.cpan/build/BioPerl-1.6.901-ghKY8U/blib/lib/Bio/DB/SeqFeature/Store.html">Bio::DB::SeqFeature::Store</a> for complete usage instructions.</p>

<h2 id="Using-the-memory-adaptor">Using the memory adaptor</h2>

<p>Before using the memory adaptor, populate a readable-directory on the file system with annotation and/or sequence files. The annotation files must be in GFF3 format, and sholud end in the extension .gff or .gff3. They may be compressed with &quot;compress&quot;, &quot;gzip&quot; or &quot;bzip2&quot; (in which case the appropriate compression extension must be present as well.)</p>

<p>You may include sequence data inline in the GFF3 files, or put the sequence data in one or more separate FASTA-format files. These files must end with .fa or .fasta and may be compressed. Because of the way the adaptor works, you will get much better performance if you keep the sequence data in separate FASTA files.</p>

<p>Initialize the database using the -dsn option. This should point to the directory creating the annotation and sequence files, or to a single GFF3 file. Examples:</p>

<pre><code><code>  <span class="comment"># load all GFF3 and FASTA files located in /var/databases/test directory</span>
  <span class="variable">$db</span>  <span class="operator">=</span> <span class="variable">Bio::DB::SeqFeature::Store</span><span class="operator">-&gt;</span><span class="variable">new</span><span class="operator">(</span> <span class="string">-adaptor</span> <span class="operator">=&gt;</span> <span class="string">'memory'</span><span class="operator">,</span>
                                          <span class="string">-dsn</span>     <span class="operator">=&gt;</span> <span class="string">'/var/databases/test'</span><span class="operator">);</span>
  
  
  <span class="comment"># load the data in a single compressed GFF3 file located at</span>
  <span class="comment"># /usr/annotations/worm.gf33.gz</span>
  <span class="variable">$db</span>  <span class="operator">=</span> <span class="variable">Bio::DB::SeqFeature::Store</span><span class="operator">-&gt;</span><span class="variable">new</span><span class="operator">(</span> <span class="string">-adaptor</span> <span class="operator">=&gt;</span> <span class="string">'memory'</span><span class="operator">,</span>
                                          <span class="string">-dsn</span>     <span class="operator">=&gt;</span> <span class="string">'/usr/annotations/worm.gff3.gz'</span><span class="operator">);</span>
</code></code></pre>

<p>For compatibility with the Bio::DB::GFF memory adaptor, -gff is recognized as an alias for -dsn.</p>

<p>See <a href="../../../../../../../.cpan/build/BioPerl-1.6.901-ghKY8U/blib/lib/Bio/DB/SeqFeature/Store.html">Bio::DB::SeqFeature::Store</a> for all the access methods supported by this adaptor. The various methods for storing and updating features and sequences into the database are supported, including GFF3 loading support, but since this is an in-memory adaptor all changes you make will be lost when the script exits.</p>

<h2 id="types">types</h2>

<pre><code><code> Title   : types
 Usage   : @type_list = $db-&gt;types
 Function: Get all the types in the database
 Returns : array of Bio::DB::GFF::Typename objects (arrayref in scalar context)
 Args    : none
 Status  : public</code></code></pre>

<h1 id="BUGS">BUGS</h1>

<p>This is an early version, so there are certainly some bugs. Please use the BioPerl bug tracking system to report bugs.</p>

<h1 id="SEE-ALSO">SEE ALSO</h1>

<p><a>bioperl</a>, <a href="../../../../../../../.cpan/build/BioPerl-1.6.901-ghKY8U/blib/lib/Bio/DB/SeqFeature.html">Bio::DB::SeqFeature</a>, <a href="../../../../../../../.cpan/build/BioPerl-1.6.901-ghKY8U/blib/lib/Bio/DB/SeqFeature/Store.html">Bio::DB::SeqFeature::Store</a>, <a>Bio::DB::SeqFeature::GFF3Loader</a>, <a href="../../../../../../../.cpan/build/BioPerl-1.6.901-ghKY8U/blib/lib/Bio/DB/SeqFeature/Segment.html">Bio::DB::SeqFeature::Segment</a>, <a href="../../../../../../../.cpan/build/BioPerl-1.6.901-ghKY8U/blib/lib/Bio/DB/SeqFeature/Store/berkeleydb.html">Bio::DB::SeqFeature::Store::berkeleydb</a>, <a href="../../../../../../../.cpan/build/BioPerl-1.6.901-ghKY8U/blib/lib/Bio/DB/SeqFeature/Store/DBI/mysql.html">Bio::DB::SeqFeature::Store::DBI::mysql</a></p>

<h1 id="AUTHOR">AUTHOR</h1>

<p>Lincoln Stein &lt;lstein@cshl.org&gt;.</p>

<p>Copyright (c) 2006 Cold Spring Harbor Laboratory.</p>

<p>This library is free software; you can redistribute it and/or modify it under the same terms as Perl itself.</p>


</body>

</html>


