<?xml version="1.0" ?>
<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.0 Transitional//EN">
<html>
<head>
<!-- saved from url=(0017)http://localhost/ -->
<script language="JavaScript" src="../../../../../../displayToc.js"></script>
<script language="JavaScript" src="../../../../../../tocParas.js"></script>
<script language="JavaScript" src="../../../../../../tocTab.js"></script>
<link rel="stylesheet" type="text/css" href="../../../../../../scineplex.css">
<title></title>
<link rel="stylesheet" href="../../../../../../Active.css" type="text/css" />
<meta http-equiv="content-type" content="text/html; charset=utf-8" />
<link rev="made" href="mailto:support@ActiveState.com" />
</head>

<body>



<ul id="index">
  <li><a href="#NAME">NAME</a></li>
  <li><a href="#SYNOPSIS">SYNOPSIS</a></li>
  <li><a href="#DESCRIPTION">DESCRIPTION</a></li>
  <li><a href="#CONSTRUCTOR">CONSTRUCTOR</a></li>
  <li><a href="#METHODS">METHODS</a></li>
  <li><a href="#BUGS">BUGS</a></li>
  <li><a href="#SEE-ALSO">SEE ALSO</a></li>
  <li><a href="#AUTHORS">AUTHORS</a>
    <ul>
      <li><a href="#feature_by_name">_feature_by_name</a></li>
    </ul>
  </li>
</ul>

<h1 id="NAME">NAME</h1>

<p>Bio::DB::GFF::Adaptor::berkeleydb -- Bio::DB::GFF database adaptor for in-memory databases</p>

<h1 id="SYNOPSIS">SYNOPSIS</h1>

<pre><code><code>  <span class="keyword">use</span> <span class="variable">Bio::DB::GFF</span><span class="operator">;</span>
  <span class="keyword">my</span> <span class="variable">$db</span> <span class="operator">=</span> <span class="variable">Bio::DB::GFF</span><span class="operator">-&gt;</span><span class="variable">new</span><span class="operator">(</span><span class="string">-adaptor</span><span class="operator">=&gt;</span> <span class="string">'berkeleydb'</span><span class="operator">,</span>
                             <span class="string">-create</span> <span class="operator">=&gt;</span> <span class="number">1</span><span class="operator">,</span> <span class="comment"># on initial build you need this</span>
                             <span class="string">-dsn</span>    <span class="operator">=&gt;</span> <span class="string">'/usr/local/share/gff/dmel'</span><span class="operator">);</span>
  
  <span class="comment"># initialize an empty database, then load GFF and FASTA files</span>
  <span class="variable">$db</span><span class="operator">-&gt;</span><span class="variable">initialize</span><span class="operator">(</span><span class="number">1</span><span class="operator">);</span>
  <span class="variable">$db</span><span class="operator">-&gt;</span><span class="variable">load_gff</span><span class="operator">(</span><span class="string">'/home/drosophila_R3.2.gff'</span><span class="operator">);</span>
  <span class="variable">$db</span><span class="operator">-&gt;</span><span class="variable">load_fasta</span><span class="operator">(</span><span class="string">'/home/drosophila_R3.2.fa'</span><span class="operator">);</span>
  
  <span class="comment"># do queries</span>
  <span class="keyword">my</span> <span class="variable">$segment</span>  <span class="operator">=</span> <span class="variable">$db</span><span class="operator">-&gt;</span><span class="variable">segment</span><span class="operator">(</span><span class="string">Chromosome</span> <span class="operator">=&gt;</span> <span class="string">'1R'</span><span class="operator">);</span>
  <span class="keyword">my</span> <span class="variable">$subseg</span>   <span class="operator">=</span> <span class="variable">$segment</span><span class="operator">-&gt;</span><span class="variable">subseq</span><span class="operator">(</span><span class="number">5000</span><span class="operator">,</span><span class="number">6000</span><span class="operator">);</span>
  <span class="keyword">my</span> <span class="variable">@features</span> <span class="operator">=</span> <span class="variable">$subseg</span><span class="operator">-&gt;</span><span class="variable">features</span><span class="operator">(</span><span class="string">'gene'</span><span class="operator">);</span>
</code></code></pre>

<p>See <a href="../../../../../../../.cpan/build/BioPerl-1.6.901-ghKY8U/blib/lib/Bio/DB/GFF.html">Bio::DB::GFF</a> for other methods.</p>

<h1 id="DESCRIPTION">DESCRIPTION</h1>

<p>This adaptor implements a berkeleydb-indexed version of Bio::DB::GFF. It requires the DB_File and Storable modules. It can be used to store and retrieve short to medium-length GFF files of several million features in length.</p>

<h1 id="CONSTRUCTOR">CONSTRUCTOR</h1>

<p>Use Bio::DB::GFF-&gt;new() to construct new instances of this class. Three named arguments are recommended:</p>

<pre><code><code> Argument    Description
 --------    -----------

 -adaptor    Set to &quot;berkeleydb&quot; to create an instance of this class.

 -dsn        Path to directory where the database index files will be stored (alias -db)

 -autoindex  Monitor the indicated directory path for FASTA and GFF files, and update the
               indexes automatically if they change (alias -dir)

 -write      Set to a true value in order to update the database.

 -create     Set to a true value to create the database the first time
               (implies -write)

 -tmp        Location of temporary directory for storing intermediate files
               during certain queries.

 -preferred_groups  Specify the grouping tag. See L&lt;Bio::DB::GFF&gt;</code></code></pre>

<p>The -dsn argument selects the directory in which to store the database index files. If the directory does not exist it will be created automatically, provided that the current process has sufficient privileges. If no -dsn argument is specified, a database named &quot;test&quot; will be created in your system&#39;s temporary files directory.</p>

<p>The -tmp argument specifies the temporary directory to use for storing intermediate search results. If not specified, your system&#39;s temporary files directory will be used. On Unix systems, the TMPDIR environment variable is honored. Note that some queries can require a lot of space.</p>

<p>The -autoindex argument, if present, selects a directory to be monitored for GFF and FASTA files (which can be compressed with the gzip program if desired). Whenever any file in this directory is changed, the index files will be updated. Note that the indexing can take a long time to run: anywhere from 5 to 10 minutes for a million features. An alias for this argument is -dir, which gives this adaptor a similar flavor to the &quot;memory&quot; adaptor.</p>

<p>-dsn and -dir can point to the same directory. If -dir is given but -dsn is absent the index files will be stored into the directory containing the source files. For autoindexing to work, you must specify the same -dir path each time you open the database.</p>

<p>If you do not choose autoindexing, then you will want to load the database using the bp_load_gff.pl command-line tool. For example:</p>

<pre><code><code> bp_load_gff.pl -a berkeleydb -c -d /usr/local/share/gff/dmel dna1.fa dna2.fa features.gff</code></code></pre>

<h1 id="METHODS">METHODS</h1>

<p>See <a href="../../../../../../../.cpan/build/BioPerl-1.6.901-ghKY8U/blib/lib/Bio/DB/GFF.html">Bio::DB::GFF</a> for inherited methods</p>

<h1 id="BUGS">BUGS</h1>

<p>The various get_Stream_* methods and the features() method with the -iterator argument only return an iterator after the query runs completely and the module has been able to generate a temporary results file on disk. This means that iteration is not as big a win as it is for the relational-database adaptors.</p>

<p>Like the dbi::mysqlopt adaptor, this module uses a binning scheme to speed up range-based searches. The binning scheme used here imposes a hard-coded 1 gigabase (1000 Mbase) limit on the size of the largest chromosome or other reference sequence.</p>

<h1 id="SEE-ALSO">SEE ALSO</h1>

<p><a href="../../../../../../../.cpan/build/BioPerl-1.6.901-ghKY8U/blib/lib/Bio/DB/GFF.html">Bio::DB::GFF</a>, <a>bioperl</a></p>

<h1 id="AUTHORS">AUTHORS</h1>

<p>Vsevolod (Simon) Ilyushchenko &gt;simonf@cshl.edu&lt; Lincoln Stein &gt;lstein@cshl.edu&lt;</p>

<p>Copyright (c) 2005 Cold Spring Harbor Laboratory.</p>

<p>This library is free software; you can redistribute it and/or modify it under the same terms as Perl itself.</p>

<h2 id="feature_by_name">_feature_by_name</h2>

<pre><code><code> Title   : _feature_by_name
 Usage   : $db-&gt;get_features_by_name($class,$name,$callback)
 Function: get a list of features by name and class
 Returns : count of number of features retrieved
 Args    : name of feature, class of feature, and a callback
 Status  : protected</code></code></pre>

<p>This method is used internally. The callback arguments are those used by make_feature().</p>


</body>

</html>


