<?xml version="1.0" ?>
<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.0 Transitional//EN">
<html>
<head>
<!-- saved from url=(0017)http://localhost/ -->
<script language="JavaScript" src="../../../../../../displayToc.js"></script>
<script language="JavaScript" src="../../../../../../tocParas.js"></script>
<script language="JavaScript" src="../../../../../../tocTab.js"></script>
<link rel="stylesheet" type="text/css" href="../../../../../../scineplex.css">
<title></title>
<link rel="stylesheet" href="../../../../../../Active.css" type="text/css" />
<meta http-equiv="content-type" content="text/html; charset=utf-8" />
<link rev="made" href="mailto:support@ActiveState.com" />
</head>

<body>



<ul id="index">
  <li><a href="#NAME">NAME</a></li>
  <li><a href="#SYNOPSIS">SYNOPSIS</a></li>
  <li><a href="#DESCRIPTION">DESCRIPTION</a></li>
  <li><a href="#AUTHOR">AUTHOR</a>
    <ul>
      <li><a href="#new">new</a></li>
    </ul>
  </li>
</ul>

<h1 id="NAME">NAME</h1>

<p>Bio::DB::GFF::Adaptor::biofetch -- Cache BioFetch objects in a Bio::DB::GFF database</p>

<h1 id="SYNOPSIS">SYNOPSIS</h1>

<p>Proof of principle. Not for production use.</p>

<h1 id="DESCRIPTION">DESCRIPTION</h1>

<p>This adaptor is a proof-of-principle. It is used to fetch BioFetch sequences into a Bio::DB::GFF database (currently uses a hard-coded EMBL database) as needed. This allows the Generic Genome Browser to be used as a Genbank/EMBL browser.</p>

<h1 id="AUTHOR">AUTHOR</h1>

<p>Lincoln Stein &lt;lstein@cshl.org&gt;.</p>

<p>Copyright 2002 Cold Spring Harbor Laboratory.</p>

<p>This library is free software; you can redistribute it and/or modify it under the same terms as Perl itself.</p>

<h2 id="new">new</h2>

<pre><code><code> <span class="variable">Title</span>   <span class="operator">:</span> <span class="variable">new</span>
 <span class="variable">Usage</span>   <span class="operator">:</span> <span class="variable">$db</span> <span class="operator">=</span> <span class="variable">Bio::DB::GFF</span><span class="operator">-&gt;</span><span class="variable">new</span><span class="operator">(</span><span class="string">-adaptor</span><span class="operator">=&gt;</span><span class="string">'biofetch'</span><span class="operator">,</span><span class="variable">@args</span><span class="operator">)</span>
 <span class="variable">Function</span><span class="operator">:</span> <span class="variable">create</span> <span class="variable">a</span> <span class="variable">new</span> <span class="variable">adaptor</span>
 <span class="variable">Returns</span> <span class="operator">:</span> <span class="variable">a</span> <span class="variable">Bio::DB::GFF</span> <span class="variable">object</span>
 <span class="variable">Args</span>    <span class="operator">:</span>   <span class="operator">-</span><span class="variable">adaptor</span> <span class="operator">:</span> <span class="variable">required</span><span class="operator">.</span>  <span class="variable">Which</span> <span class="variable">adaptor</span> <span class="variable">to</span> <span class="keyword">use</span><span class="operator">;</span> <span class="variable">biofetch</span> <span class="keyword">for</span> <span class="variable">mysql</span><span class="operator">,</span> <span class="variable">biofetch_oracle</span> <span class="keyword">for</span> <span class="variable">Oracle</span>
             <span class="operator">-</span><span class="variable">preferred_tags</span> <span class="operator">:</span> <span class="variable">optional</span><span class="operator">.</span>  <span class="variable">A</span> <span class="variable">hash</span> <span class="variable">of</span> <span class="operator">{</span><span class="string">classname</span> <span class="operator">=&gt;</span> <span class="variable">weight</span><span class="operator">,...}</span>
                               <span class="variable">used</span> <span class="variable">to</span> <span class="variable">determine</span> <span class="variable">the</span> <span class="variable">class</span> <span class="keyword">and</span> <span class="variable">name</span> <span class="variable">of</span> <span class="variable">the</span> <span class="variable">feature</span>
                               <span class="keyword">when</span> <span class="variable">a</span> <span class="variable">choice</span> <span class="variable">of</span> <span class="variable">possible</span> <span class="variable">feature</span> <span class="variable">classes</span> <span class="variable">is</span> <span class="variable">available</span>
                               <span class="operator">(</span><span class="variable">e</span><span class="operator">.</span><span class="variable">g</span><span class="operator">.</span> <span class="variable">a</span> <span class="variable">feature</span> <span class="variable">has</span> <span class="variable">both</span> <span class="variable">a</span> <span class="string">'gene'</span> <span class="keyword">and</span> <span class="variable">a</span> <span class="string">'locus'</span> <span class="variable">tag</span><span class="operator">).</span>
                               <span class="variable">Common</span> <span class="variable">defaults</span> <span class="variable">are</span> <span class="variable">provided</span> <span class="variable">that</span> <span class="variable">work</span> <span class="variable">well</span> <span class="keyword">for</span> <span class="variable">eukaryotic</span>
                               <span class="variable">features</span> <span class="operator">(</span><span class="variable">but</span> <span class="keyword">not</span> <span class="variable">well</span> <span class="keyword">for</span> <span class="variable">viral</span><span class="operator">/</span><span class="variable">prokaryotic</span><span class="operator">)</span>
              <span class="variable">see</span> <span class="variable">below</span> <span class="keyword">for</span> <span class="variable">additional</span> <span class="variable">arguments</span><span class="operator">.</span>
 <span class="variable">Status</span>  <span class="operator">:</span> <span class="variable">Public</span>
</code></code></pre>

<p>This is the constructor for the adaptor. It is called automatically by Bio::DB::GFF-&gt;new. In addition to arguments that are common among all adaptors, the following class-specific arguments are recgonized:</p>

<pre><code><code>  Argument       Description
  --------       -----------

  -dsn           the DBI data source, e.g. &#39;dbi:mysql:ens0040&#39;

  -user          username for authentication

  -pass          the password for authentication

  -proxy         [[&#39;http&#39;,&#39;ftp&#39;],&#39;http://proxy:8080&#39;]

  -source        source to use for loaded features (&#39;EMBL&#39;)</code></code></pre>

<p>-dsn,-user and -pass indicate the local database to cache results in, and as are per Bio::DB::GFF::Adaptor::dbi. The -proxy argument allows you to set the biofetch web proxy, and uses the same syntax described for the proxy() method of <a href="../../../../../../../.cpan/build/BioPerl-1.6.901-ghKY8U/blib/lib/Bio/DB/WebDBSeqI.html">Bio::DB::WebDBSeqI</a>, except that the argument must be passed as an array reference.</p>


</body>

</html>


