<?xml version="1.0" ?>
<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.0 Transitional//EN">
<html>
<head>
<!-- saved from url=(0017)http://localhost/ -->
<script language="JavaScript" src="../../../displayToc.js"></script>
<script language="JavaScript" src="../../../tocParas.js"></script>
<script language="JavaScript" src="../../../tocTab.js"></script>
<link rel="stylesheet" type="text/css" href="../../../scineplex.css">
<title></title>
<link rel="stylesheet" href="../../../Active.css" type="text/css" />
<meta http-equiv="content-type" content="text/html; charset=utf-8" />
<link rev="made" href="mailto:support@ActiveState.com" />
</head>

<body>



<ul id="index">
  <li><a href="#NAME">NAME</a></li>
  <li><a href="#SYNOPSIS">SYNOPSIS</a></li>
  <li><a href="#DESCRIPTION">DESCRIPTION</a>
    <ul>
      <li><a href="#Boulder::LocusLink-methods">Boulder::LocusLink methods</a></li>
    </ul>
  </li>
  <li><a href="#OUTPUT-TAGS">OUTPUT TAGS</a>
    <ul>
      <li><a href="#Top-Level-Tags">Top-Level Tags</a></li>
    </ul>
  </li>
  <li><a href="#SEE-ALSO">SEE ALSO</a></li>
  <li><a href="#AUTHOR">AUTHOR</a></li>
</ul>

<h1 id="NAME">NAME</h1>

<p>Boulder::LocusLink - Fetch LocusLink data records as parsed Boulder Stones</p>

<h1 id="SYNOPSIS">SYNOPSIS</h1>

<pre><code><code>  <span class="comment"># parse a file of LocusLink records</span>
  <span class="variable">$ll</span> <span class="operator">=</span> <span class="variable">new</span> <span class="variable">Boulder::LocusLink</span><span class="operator">(</span><span class="string">-accessor</span><span class="operator">=&gt;</span><span class="string">'File'</span><span class="operator">,</span>
                             <span class="string">-param</span> <span class="operator">=&gt;</span> <span class="string">'/home/data/LocusLink/LL_tmpl'</span><span class="operator">);</span>
  <span class="keyword">while</span> <span class="operator">(</span><span class="keyword">my</span> <span class="variable">$s</span> <span class="operator">=</span> <span class="variable">$ll</span><span class="operator">-&gt;</span><span class="variable">get</span><span class="operator">)</span> <span class="operator">{</span>
    <span class="keyword">print</span> <span class="variable">$s</span><span class="operator">-&gt;</span><span class="variable">Identifier</span><span class="operator">;</span>
    <span class="keyword">print</span> <span class="variable">$s</span><span class="operator">-&gt;</span><span class="variable">Gene</span><span class="operator">;</span>
  <span class="operator">}</span>
  
  <span class="comment"># parse flatfile records yourself</span>
  <span class="keyword">open</span> <span class="operator">(</span><span class="variable">LL</span><span class="operator">,</span><span class="string">"/home/data/LocusLink/LL_tmpl"</span><span class="operator">);</span>
  <span class="keyword">local</span> <span class="variable">$/</span> <span class="operator">=</span> <span class="string">"*RECORD*"</span><span class="operator">;</span>
  <span class="keyword">while</span> <span class="operator">(&lt;</span><span class="variable">LL</span><span class="operator">&gt;)</span> <span class="operator">{</span>
     <span class="keyword">my</span> <span class="variable">$s</span> <span class="operator">=</span> <span class="variable">Boulder::LocusLink</span><span class="operator">-&gt;</span><span class="variable">parse</span><span class="operator">(</span><span class="variable">$_</span><span class="operator">);</span>
     <span class="comment"># etc.</span>
  <span class="operator">}</span>
</code></code></pre>

<h1 id="DESCRIPTION">DESCRIPTION</h1>

<p>Boulder::LocusLink provides retrieval and parsing services for LocusLink records</p>

<p>Boulder::LocusLink provides retrieval and parsing services for NCBI LocusLink records. It returns Unigene entries in <a href="../../../../.cpan/build/Boulder-1.30-1cMFjf/blib/lib/Stone.html">Stone</a> format, allowing easy access to the various fields and values. Boulder::LocusLink is a descendent of Boulder::Stream, and provides a stream-like interface to a series of Stone objects.</p>

<p>Access to LocusLink is provided by one <i>accessors</i>, which give access to local LocusLink database. When you create a new Boulder::LocusLink stream, you provide the accessors, along with accessor-specific parameters that control what entries to fetch. The accessors is:</p>

<dl>

<dt id="File">File</dt>
<dd>

<p>This provides access to local LocusLink entries by reading from a flat file (typically Hs.dat file downloadable from NCBI&#39;s Ftp site). The stream will return a Stone corresponding to each of the entries in the file, starting from the top of the file and working downward. The parameter is the path to the local file.</p>

</dd>
</dl>

<p>It is also possible to parse a single LocusLink entry from a text string stored in a scalar variable, returning a Stone object.</p>

<h2 id="Boulder::LocusLink-methods">Boulder::LocusLink methods</h2>

<p>This section lists the public methods that the <i>Boulder::LocusLink</i> class makes available.</p>

<dl>

<dt id="new-">new()</dt>
<dd>

<pre><code><code>   <span class="comment"># Local fetch via File</span>
   <span class="variable">$ug</span><span class="operator">=</span><span class="variable">new</span> <span class="variable">Boulder::LocusLink</span><span class="operator">(</span><span class="string">-accessor</span>  <span class="operator">=&gt;</span>  <span class="string">'File'</span><span class="operator">,</span>
                            <span class="string">-param</span>     <span class="operator">=&gt;</span>  <span class="string">'/data/LocusLink/Hs.dat'</span><span class="operator">);</span>
</code></code></pre>

<p>The new() method creates a new <i>Boulder::LocusLink</i> stream on the accessor provided. The only possible accessors is <b>File</b>. If successful, the method returns the stream object. Otherwise it returns undef.</p>

<p>new() takes the following arguments:</p>

<pre><code><code>        -accessor       Name of the accessor to use
        -param          Parameters to pass to the accessor</code></code></pre>

<p>Specify the accessor to use with the <b>-accessor</b> argument. If not specified, it defaults to <b>File</b>.</p>

<p><b>-param</b> is an accessor-specific argument. The possibilities is:</p>

<p>For <b>File</b>, the <b>-param</b> argument must point to a string-valued scalar, which will be interpreted as the path to the file to read LocusLink entries from.</p>

</dd>
<dt id="get-">get()</dt>
<dd>

<p>The get() method is inherited from <i>Boulder::Stream</i>, and simply returns the next parsed LocusLink Stone, or undef if there is nothing more to fetch. It has the same semantics as the parent class, including the ability to restrict access to certain top-level tags.</p>

</dd>
<dt id="put-">put()</dt>
<dd>

<p>The put() method is inherited from the parent Boulder::Stream class, and will write the passed Stone to standard output in Boulder format. This means that it is currently not possible to write a Boulder::LocusLink object back into LocusLink flatfile form.</p>

</dd>
</dl>

<h1 id="OUTPUT-TAGS">OUTPUT TAGS</h1>

<p>The tags returned by the parsing operation are taken from the names shown in the Flat file Hs.dat since no better description of them is provided yet by the database source producer.</p>

<h2 id="Top-Level-Tags">Top-Level Tags</h2>

<p>These are tags that appear at the top level of the parsed LocusLink entry.</p>

<dl>

<dt id="Identifier">Identifier</dt>
<dd>

<p>The LocusLink identifier of this entry. Identifier is a single-value tag.</p>

<p>Example:</p>

<pre><code><code>      <span class="keyword">my</span> <span class="variable">$identifierNo</span> <span class="operator">=</span> <span class="variable">$s</span><span class="operator">-&gt;</span><span class="variable">Identifier</span><span class="operator">;</span>
</code></code></pre>

</dd>
<dt id="Current_locusid">Current_locusid</dt>
<dd>

<p>If a locus has been merged with another, the Current_locusid contains the previous LOCUSID line (A bit confusing, shall be called &quot;previous_locusid&quot;, but this is defined in NCBI README File ... ).</p>

<p>Example: my $prevlocusid=$s-&gt;Current_locusid;</p>

</dd>
<dt id="Organism-Source-species-ased-on-NCBIs-Taxonomy">Organism Source species ased on NCBI&#39;s Taxonomy</dt>
<dd>

<p>Example: my $theorganism=$s-&gt;Organism;</p>

</dd>
<dt id="Status-Type-of-reference-sequence-record.-If-PROVISIONAL-then-means-that-is-generated-automatically-from-existing-Genbank-record-and-information-stored-in-the-LocusLink-database-no-curation.-If-REVIEWED-than-it-means-that-is-generated-from-the-most-representative-complete-GenBank-sequence-or-merge-of-GenBank-sequenes-and-from-information-stored-in-the-LocusLink-database">Status Type of reference sequence record. If &quot;PROVISIONAL&quot; then means that is generated automatically from existing Genbank record and information stored in the LocusLink database, no curation. If &quot;REVIEWED&quot; than it means that is generated from the most representative complete GenBank sequence or merge of GenBank sequenes and from information stored in the LocusLink database</dt>
<dd>

<p>Example: my $thestatus=$s-&gt;Status;</p>

</dd>
<dt id="LocAss-Here-comes-a-complex-record-...-made-up-of-LOCUS_STRING-NM-The-value-in-the-LOCUS-field-of-the-RefSeq-record-NP-The-RefSeq-accession-number-for-an-mRNA-record-PRODUCT-The-name-of-the-produc-tof-this-transcript-TRANSVAR-a-variant-specific-description-ASSEMBLY-The-Genbank-accession-used-to-assemble-the-refseq-record">LocAss Here comes a complex record ... made up of LOCUS_STRING, NM The value in the LOCUS field of the RefSeq record , NP The RefSeq accession number for an mRNA record, PRODUCT The name of the produc tof this transcript, TRANSVAR a variant-specific description, ASSEMBLY The Genbank accession used to assemble the refseq record</dt>
<dd>

<p>Example: my $theprod=$s-&gt;LocAss-&gt;Product;</p>

</dd>
<dt id="AccProt-Here-comes-a-complex-record-...-made-up-of-ACCNUM-Nucleotide-sequence-accessio-number-TYPE-e-EST-m-mRNA-g-Genomic-PROT-set-of-PID-values-for-the-coding-region-or-regions-annotated-on-the-nucleotide-record.-The-first-value-is-the-PID-an-integer-or-null-then-either-MMDB-or-na-separated-from-the-PID-by-a-.-If-MMDB-is-present-it-indicates-there-are-structur-edata-available-for-a-protein-related-to-the-protein-referenced-by-the-PID-Example:-my-theprot-s-AccProt-Prot-">AccProt Here comes a complex record ... made up of ACCNUM Nucleotide sequence accessio number TYPE e=EST, m=mRNA, g=Genomic PROT set of PID values for the coding region or regions annotated on the nucleotide record. The first value is the PID (an integer or null), then either MMDB or na, separated from the PID by a |. If MMDB is present, it indicates there are structur edata available for a protein related to the protein referenced by the PID Example: my $theprot=$s-&gt;AccProt-&gt;Prot;</dt>
<dd>

</dd>
<dt id="OFFICIAL_SYMBOL-The-symbol-used-for-gene-reports-validated-by-the-appropriate-nomenclature-committee">OFFICIAL_SYMBOL The symbol used for gene reports, validated by the appropriate nomenclature committee</dt>
<dd>

</dd>
<dt id="PREFERRED_SYMBOL-Interim-symbol-used-for-display">PREFERRED_SYMBOL Interim symbol used for display</dt>
<dd>

</dd>
<dt id="OFFICIAL_GENE_NAME-The-gene-description-used-for-gene-reports-validate-by-the-appropriate-nomenclatur-eommittee.-If-the-symbol-is-official-the-gene-name-will-be-official.-No-records-will-have-both-official-and-interim-nomenclature.">OFFICIAL_GENE_NAME The gene description used for gene reports validate by the appropriate nomenclatur eommittee. If the symbol is official, the gene name will be official. No records will have both official and interim nomenclature.</dt>
<dd>

</dd>
<dt id="PREFERRED_GENE_NAME-Interim-used-for-display">PREFERRED_GENE_NAME Interim used for display</dt>
<dd>

</dd>
<dt id="PREFERRED_PRODUCT-The-name-of-the-product-used-in-the-RefSeq-record">PREFERRED_PRODUCT The name of the product used in the RefSeq record</dt>
<dd>

</dd>
<dt id="ALIAS_SYMBOL-Other-symbols-associated-with-this-gene">ALIAS_SYMBOL Other symbols associated with this gene</dt>
<dd>

</dd>
<dt id="ALIAS_PROT-Other-protein-names-associated-with-this-gene">ALIAS_PROT Other protein names associated with this gene</dt>
<dd>

</dd>
<dt id="PhenoTable-A-complex-record-made-up-of-Phenotype-Phenotype_ID">PhenoTable A complex record made up of Phenotype Phenotype_ID</dt>
<dd>

</dd>
<dt id="SUmmary">SUmmary</dt>
<dd>

</dd>
<dt id="Unigene">Unigene</dt>
<dd>

</dd>
<dt id="Omim">Omim</dt>
<dd>

</dd>
<dt id="Chr">Chr</dt>
<dd>

</dd>
<dt id="Map">Map</dt>
<dd>

</dd>
<dt id="STS">STS</dt>
<dd>

</dd>
<dt id="ECNUM">ECNUM</dt>
<dd>

</dd>
<dt id="ButTable-BUTTON-LINK">ButTable BUTTON LINK</dt>
<dd>

</dd>
<dt id="DBTable-DB_DESCR-DB_LINK">DBTable DB_DESCR DB_LINK</dt>
<dd>

</dd>
<dt id="PMID-a-subset-of-publications-associated-with-this-locus-with-the-link-being-the-PubMed-unique-identifier-comma-separated">PMID a subset of publications associated with this locus with the link being the PubMed unique identifier comma separated</dt>
<dd>

</dd>
</dl>

<h1 id="SEE-ALSO">SEE ALSO</h1>

<p><a href="../../../../.cpan/build/Boulder-1.30-1cMFjf/blib/lib/Boulder.html">Boulder</a>, <a href="../../../../.cpan/build/Boulder-1.30-1cMFjf/blib/lib/Boulder/Blast.html">Boulder::Blast</a>, <a href="../../../../.cpan/build/Boulder-1.30-1cMFjf/blib/lib/Boulder/Genbank.html">Boulder::Genbank</a></p>

<h1 id="AUTHOR">AUTHOR</h1>

<p>Lincoln Stein &lt;lstein@cshl.org&gt;. Luca I.G. Toldo &lt;luca.toldo@merck.de&gt;</p>

<p>Copyright (c) 1997 Lincoln D. Stein Copyright (c) 1999 Luca I.G. Toldo</p>

<p>This library is free software; you can redistribute it and/or modify it under the same terms as Perl itself. See DISCLAIMER.txt for disclaimers of warranty.</p>


</body>

</html>


