<?xml version="1.0" ?>
<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.0 Transitional//EN">
<html>
<head>
<!-- saved from url=(0017)http://localhost/ -->
<script language="JavaScript" src="../../../../../displayToc.js"></script>
<script language="JavaScript" src="../../../../../tocParas.js"></script>
<script language="JavaScript" src="../../../../../tocTab.js"></script>
<link rel="stylesheet" type="text/css" href="../../../../../scineplex.css">
<title></title>
<link rel="stylesheet" href="../../../../../Active.css" type="text/css" />
<meta http-equiv="content-type" content="text/html; charset=utf-8" />
<link rev="made" href="mailto:support@ActiveState.com" />
</head>

<body>



<ul id="index">
  <li><a href="#NAME">NAME</a></li>
  <li><a href="#SYNOPSIS">SYNOPSIS</a></li>
  <li><a href="#DESCRIPTION">DESCRIPTION</a></li>
  <li><a href="#FEEDBACK">FEEDBACK</a>
    <ul>
      <li><a href="#Mailing-Lists">Mailing Lists</a></li>
      <li><a href="#Support">Support</a></li>
      <li><a href="#Reporting-Bugs">Reporting Bugs</a></li>
    </ul>
  </li>
  <li><a href="#AUTHOR---Sheldon-McKay">AUTHOR - Sheldon McKay</a></li>
  <li><a href="#APPENDIX">APPENDIX</a>
    <ul>
      <li><a href="#new">new</a></li>
      <li><a href="#add_source">add_source</a></li>
      <li><a href="#has_gene">has_gene</a></li>
      <li><a href="#has_CDS">_has_CDS</a></li>
      <li><a href="#add_annotation">add_annotation</a></li>
      <li><a href="#add_generic_annotation">_add_generic_annotation</a></li>
      <li><a href="#feature_set">feature_set</a></li>
      <li><a href="#build_feature_set">_build_feature_set</a></li>
      <li><a href="#add_feature_span">_add_feature_span</a></li>
      <li><a href="#add_CDS">_add_CDS</a></li>
    </ul>
  </li>
</ul>

<h1 id="NAME">NAME</h1>

<p>Bio::SeqIO::game::featHandler -- a class for handling feature elements</p>

<h1 id="SYNOPSIS">SYNOPSIS</h1>

<p>This module is not used directly</p>

<h1 id="DESCRIPTION">DESCRIPTION</h1>

<p>Bio::SeqIO::game::featHandler converts game XML &lt;annotation&gt; elements into flattened Bio::SeqFeature::Generic objects to be added to the sequence</p>

<h1 id="FEEDBACK">FEEDBACK</h1>

<h2 id="Mailing-Lists">Mailing Lists</h2>

<p>User feedback is an integral part of the evolution of this and other Bioperl modules. Send your comments and suggestions preferably to one of the Bioperl mailing lists.</p>

<p>Your participation is much appreciated.</p>

<pre><code><code>  bioperl-l@bioperl.org                  - General discussion
  http://bioperl.org/wiki/Mailing_lists  - About the mailing lists</code></code></pre>

<h2 id="Support">Support</h2>

<p>Please direct usage questions or support issues to the mailing list:</p>

<p><i>bioperl-l@bioperl.org</i></p>

<p>rather than to the module maintainer directly. Many experienced and reponsive experts will be able look at the problem and quickly address it. Please include a thorough description of the problem with code and data examples if at all possible.</p>

<h2 id="Reporting-Bugs">Reporting Bugs</h2>

<p>Report bugs to the Bioperl bug tracking system to help us keep track of the bugs and their resolution. Bug reports can be submitted via the web:</p>

<pre><code><code>  https://redmine.open-bio.org/projects/bioperl/</code></code></pre>

<h1 id="AUTHOR---Sheldon-McKay">AUTHOR - Sheldon McKay</h1>

<p>Email mckays@cshl.edu</p>

<h1 id="APPENDIX">APPENDIX</h1>

<p>The rest of the documentation details each of the object methods. Internal methods are usually preceded with a _</p>

<h2 id="new">new</h2>

<pre><code><code> Title   : new
 Usage   : my $featHandler = Bio::SeqIO::game::featHandler-&gt;new($seq, $seq_h, $ann_l)
 Function: creates an object to deal with sequence features 
 Returns : a handler object
 Args    : $seq   -- a Bio::SeqI compliant object
           $seq_h -- ref. to a hash of other sequences associated 
                     with the main sequence (proteins, etc)
           $ann_l -- ref. to a list of annotations</code></code></pre>

<h2 id="add_source">add_source</h2>

<pre><code><code> <span class="variable">Title</span>   <span class="operator">:</span> <span class="variable">add_source</span>
 <span class="variable">Usage</span>   <span class="operator">:</span> <span class="variable">$featHandler</span><span class="operator">-&gt;</span><span class="variable">add_source</span><span class="operator">(</span><span class="variable">$seq</span><span class="operator">-&gt;</span><span class="variable">length</span><span class="operator">,</span> <span class="operator">\</span><span class="variable">%tags</span><span class="operator">);</span>
 <span class="variable">Function</span><span class="operator">:</span> <span class="variable">creates</span> <span class="variable">a</span> <span class="variable">source</span> <span class="variable">feature</span>
 <span class="variable">Returns</span> <span class="operator">:</span> <span class="variable">a</span> <span class="variable">Bio::SeqFeature::Generic</span> <span class="variable">object</span> 
 <span class="variable">Args</span>    <span class="operator">:</span> <span class="variable">sequence</span> <span class="keyword">length</span> <span class="keyword">and</span> <span class="variable">a</span> <span class="keyword">ref</span><span class="operator">.</span> <span class="variable">to</span> <span class="variable">a</span> <span class="variable">hash</span> <span class="variable">of</span> <span class="variable">tag</span><span class="operator">/</span><span class="variable">value</span> <span class="variable">attributes</span>
</code></code></pre>

<h2 id="has_gene">has_gene</h2>

<pre><code><code> Title   : has_gene
 Usage   : my $gene = $self-&gt;_has_gene($gene, $gname, $id)
 Function: method to get/set the current gene feature
 Returns : a Bio::SeqFeature::Generic object (if there is a gene)
 Args    : (optional)
           $gene  -- an XML element for the annotation
           $gname -- gene name
           $id    -- gene ID (not always the same as the name)</code></code></pre>

<h2 id="has_CDS">_has_CDS</h2>

<pre><code><code> Title   : _has_CDS
 Usage   : my $cds = $self-&gt;_has_CDS
 Function: internal getter/setter for CDS features
 Returns : a Bio::SeqFeature::Generic transcript object (or nothing)
 Args    : a Bio::SeqFeature::Generic transcript feature</code></code></pre>

<h2 id="add_annotation">add_annotation</h2>

<pre><code><code> Title   : add_annotation
 Usage   : $featHandler-&gt;add_annotation($seq, $type, $id, $tags, $feats)
 Function: converts a containment hierarchy into an ordered list of flat features
 Returns : nothing
 Args    : $seq   -- a Bio::SeqI compliant object
           $type  -- the annotation type
           $id    -- the anotation ID
           $tags  -- ref. to a hash of tag/value attributes
           $feats -- ref to an array of Bio::SeqFeature::Generic objects</code></code></pre>

<h2 id="add_generic_annotation">_add_generic_annotation</h2>

<pre><code><code> Title   : _add_generic_annotation
 Usage   : $self-&gt;_add_generic_annotation($seq, $type, $id, $tags, $feats)
 Function: an internal method to handle non-gene annotations
 Returns : nothing
 Args    : $seq   -- a Bio::SeqI compliant object
           $type  -- the annotation type
           $id    -- the anotation ID
           $tags  -- ref. to a hash of tag/value attributes
           $feats -- ref to an array of Bio::SeqFeature::Generic objects</code></code></pre>

<h2 id="feature_set">feature_set</h2>

<pre><code><code> <span class="variable">Title</span>   <span class="operator">:</span> <span class="variable">feature_set</span>
 <span class="variable">Usage</span>   <span class="operator">:</span> <span class="keyword">push</span> <span class="variable">@feats</span><span class="operator">,</span> <span class="variable">$featHandler</span><span class="operator">-&gt;</span><span class="variable">feature_set</span><span class="operator">(</span><span class="variable">$id</span><span class="operator">,</span> <span class="variable">$gname</span><span class="operator">,</span> <span class="variable">$set</span><span class="operator">,</span> <span class="variable">$anntype</span><span class="operator">);</span>
 <span class="variable">Function</span><span class="operator">:</span> <span class="variable">handles</span> <span class="operator">&lt;</span><span class="variable">feature_span</span><span class="operator">&gt;</span> <span class="variable">hierarchies</span> <span class="operator">(</span><span class="variable">usually</span> <span class="variable">a</span> <span class="variable">transcript</span><span class="operator">)</span>
 <span class="variable">Returns</span> <span class="operator">:</span> <span class="variable">a</span> <span class="variable">list</span> <span class="variable">of</span> <span class="variable">Bio::SeqFeature::Generic</span> <span class="variable">objects</span>
 <span class="variable">Args</span>    <span class="operator">:</span> <span class="variable">$id</span>      <span class="operator">--</span> <span class="variable">ID</span> <span class="variable">of</span> <span class="variable">the</span> <span class="variable">feature</span> <span class="variable">set</span>
           <span class="variable">$gname</span>   <span class="operator">--</span> <span class="variable">name</span> <span class="variable">of</span> <span class="variable">the</span> <span class="variable">gene</span>
           <span class="variable">$set</span>     <span class="operator">--</span> <span class="variable">the</span> <span class="operator">&lt;</span><span class="variable">feature_set</span><span class="operator">&gt;</span> <span class="variable">object</span>
           <span class="variable">$anntype</span> <span class="operator">--</span> <span class="variable">type</span> <span class="variable">of</span> <span class="variable">the</span> <span class="variable">parent</span> <span class="variable">annotation</span>
</code></code></pre>

<h2 id="build_feature_set">_build_feature_set</h2>

<pre><code><code> Title   : _build_feature_set
 Usage   : $self-&gt;_build_feature_set($set, 1) # 1 flag means retain the exon as a subfeat
 Function: an internal method to process attributes and subfeats of a feature set
 Returns : nothing
 Args    : $set -- a &lt;feature_set&gt; element
           1    -- optional flag to retain exons as subfeats.  Otherwise, they will
                   be converted to sublocations of a parent CDS feature</code></code></pre>

<h2 id="add_feature_span">_add_feature_span</h2>

<pre><code><code> Title   : _add_feature_span
 Usage   : $self-&gt;_add_feature_span($el, 1)
 Function: an internal method to process &lt;feature_span&gt; elements
 Returns : nothing
 Args    : $el -- a &lt;feature_span&gt; element
           1   -- an optional flag to retain exons as subfeatures</code></code></pre>

<h2 id="add_CDS">_add_CDS</h2>

<pre><code><code> Title   : _add_CDS
 Usage   : my $cds = $self-&gt;_add_CDS($transcript, $tags)
 Function: an internal method to create a CDS feature from a transcript feature
 Returns : a Bio::SeqFeature::Generic object
 Args    : $transcript -- a Bio::SeqFeature::Generic object for a transcript
           $tags       -- ref. to a hash of tag/value attributes</code></code></pre>


</body>

</html>


