<?xml version="1.0" ?>
<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.0 Transitional//EN">
<html>
<head>
<!-- saved from url=(0017)http://localhost/ -->
<script language="JavaScript" src="../../../../../displayToc.js"></script>
<script language="JavaScript" src="../../../../../tocParas.js"></script>
<script language="JavaScript" src="../../../../../tocTab.js"></script>
<link rel="stylesheet" type="text/css" href="../../../../../scineplex.css">
<title></title>
<link rel="stylesheet" href="../../../../../Active.css" type="text/css" />
<meta http-equiv="content-type" content="text/html; charset=utf-8" />
<link rev="made" href="mailto:support@ActiveState.com" />
</head>

<body>



<ul id="index">
  <li><a href="#NAME">NAME</a></li>
  <li><a href="#SYNOPSIS">SYNOPSIS</a></li>
  <li><a href="#DESCRIPTION">DESCRIPTION</a></li>
  <li><a href="#AUTHOR---Joseph-A.L.-Insana">AUTHOR - Joseph A.L. Insana</a></li>
  <li><a href="#APPENDIX">APPENDIX</a>
    <ul>
      <li><a href="#load">load</a></li>
      <li><a href="#embl2hash">embl2hash</a></li>
      <li><a href="#novelaasequence2gene">novelaasequence2gene</a></li>
    </ul>
  </li>
</ul>

<h1 id="NAME">NAME</h1>

<p>Bio::LiveSeq::IO::BioPerl - Loader for LiveSeq from EMBL entries with BioPerl</p>

<h1 id="SYNOPSIS">SYNOPSIS</h1>

<pre><code><code>  <span class="keyword">my</span> <span class="variable">$db</span><span class="operator">=</span><span class="string">"EMBL"</span><span class="operator">;</span>
  <span class="keyword">my</span> <span class="variable">$file</span><span class="operator">=</span><span class="string">"../data/M20132"</span><span class="operator">;</span>
  <span class="keyword">my</span> <span class="variable">$id</span><span class="operator">=</span><span class="string">"HSANDREC"</span><span class="operator">;</span>
  
  <span class="keyword">my</span> <span class="variable">$loader</span><span class="operator">=</span><span class="variable">Bio::LiveSeq::IO::BioPerl</span><span class="operator">-&gt;</span><span class="variable">load</span><span class="operator">(</span><span class="string">-db</span><span class="operator">=&gt;</span><span class="string">"</span><span class="variable">$db</span><span class="string">"</span><span class="operator">,</span> <span class="string">-file</span><span class="operator">=&gt;</span><span class="string">"</span><span class="variable">$file</span><span class="string">"</span><span class="operator">);</span>
  <span class="comment">#                      or</span>
  <span class="keyword">my</span> <span class="variable">$loader</span><span class="operator">=</span><span class="variable">Bio::LiveSeq::IO::BioPerl</span><span class="operator">-&gt;</span><span class="variable">load</span><span class="operator">(</span><span class="string">-db</span><span class="operator">=&gt;</span><span class="string">"</span><span class="variable">$db</span><span class="string">"</span><span class="operator">,</span> <span class="string">-id</span><span class="operator">=&gt;</span><span class="string">"</span><span class="variable">$id</span><span class="string">"</span><span class="operator">);</span>
  
  <span class="keyword">my</span> <span class="variable">@translationobjects</span><span class="operator">=</span><span class="variable">$loader</span><span class="operator">-&gt;</span><span class="variable">entry2liveseq</span><span class="operator">();</span>
  
  <span class="keyword">my</span> <span class="variable">$genename</span><span class="operator">=</span><span class="string">"AR"</span><span class="operator">;</span>
  <span class="keyword">my</span> <span class="variable">$gene</span><span class="operator">=</span><span class="variable">$loader</span><span class="operator">-&gt;</span><span class="variable">gene2liveseq</span><span class="operator">(</span><span class="string">-gene_name</span> <span class="operator">=&gt;</span> <span class="string">"</span><span class="variable">$genename</span><span class="string">"</span><span class="operator">,</span>
                                    <span class="string">-getswissprotinfo</span> <span class="operator">=&gt;</span> <span class="number">0</span><span class="operator">);</span>
  
  <span class="comment">#NOTE1: The only -db now supported is EMBL. Hence it defaults to EMBL.</span>
  <span class="comment">#NOTE2: -file requires a filename (and path if necessary) containing an</span>
  <span class="comment">#             EMBL entry</span>
  <span class="comment">#       -id will use Bio::DB::EMBL.pm to fetch the sequence from the web,</span>
  <span class="comment">#             (bioperl wraparound to [w]getz from SRS)</span>
  <span class="comment">#NOTE3: To retrieve the swissprot (if possible) attached to the embl entry</span>
  <span class="comment">#             (to get protein domains at dna level), only Bio::DB::EMBL.pm</span>
  <span class="comment">#             is supported under BioPerl. Refer to Bio::LiveSeq::IO::SRS</span>
  <span class="comment">#             otherwise.</span>
  <span class="comment">#NOTE4: NOTE3 is not implemented yet for bioperl, working on it</span>
</code></code></pre>

<h1 id="DESCRIPTION">DESCRIPTION</h1>

<p>This package uses BioPerl (SeqIO) to fetch a sequence database entry, analyse it and create LiveSeq objects out of it.</p>

<p>A filename (or an ID that will fetch entry through the web) has to be passed to this package which will return references to all translation objects created from the EMBL entry. References to Transcription, DNA and Exon objects can all be retrieved departing from these.</p>

<p>Alternatively, a specific &quot;gene&quot; name can be specified, together with the embl-acc ID. This will create a LiveSeq::Gene object with all relevant gene features attached/created.</p>

<p>ATTENTION: if web fetching is requested, the package HTTP::Request needs to be installed.</p>

<h1 id="AUTHOR---Joseph-A.L.-Insana">AUTHOR - Joseph A.L. Insana</h1>

<p>Email: Insana@ebi.ac.uk, jinsana@gmx.net</p>

<h1 id="APPENDIX">APPENDIX</h1>

<p>The rest of the documentation details each of the object methods. Internal methods are usually preceded with a _</p>

<h2 id="load">load</h2>

<pre><code><code>  <span class="variable">Title</span>   <span class="operator">:</span> <span class="variable">load</span>
  <span class="variable">Usage</span>   <span class="operator">:</span> <span class="keyword">my</span> <span class="variable">$filename</span><span class="operator">=</span><span class="string">"../data/M20132"</span><span class="operator">;</span>
            <span class="variable">$loader</span><span class="operator">=</span><span class="variable">Bio::LiveSeq::IO::BioPerl</span><span class="operator">-&gt;</span><span class="variable">load</span><span class="operator">(</span><span class="string">-db</span><span class="operator">=&gt;</span><span class="string">"EMBL"</span><span class="operator">,</span> <span class="string">-file</span><span class="operator">=&gt;</span><span class="string">"</span><span class="variable">$filename</span><span class="string">"</span><span class="operator">);</span>
                                   <span class="keyword">or</span>
            <span class="variable">$loader</span><span class="operator">=</span><span class="variable">Bio::LiveSeq::IO::BioPerl</span><span class="operator">-&gt;</span><span class="variable">load</span><span class="operator">(</span><span class="string">-db</span><span class="operator">=&gt;</span><span class="string">"EMBL"</span><span class="operator">,</span> <span class="string">-id</span><span class="operator">=&gt;</span><span class="string">"HSANDREC"</span><span class="operator">);</span>
  
  <span class="variable">Function</span><span class="operator">:</span> <span class="variable">loads</span> <span class="variable">an</span> <span class="variable">entry</span> <span class="variable">with</span> <span class="variable">BioPerl</span> <span class="variable">from</span> <span class="variable">a</span> <span class="variable">database</span> <span class="variable">into</span> <span class="variable">a</span> <span class="variable">hash</span>
  <span class="variable">Returns</span> <span class="operator">:</span> <span class="variable">reference</span> <span class="variable">to</span> <span class="variable">a</span> <span class="variable">new</span> <span class="variable">object</span> <span class="variable">of</span> <span class="variable">class</span> <span class="variable">IO::BioPerl</span> <span class="variable">holding</span> <span class="variable">an</span> <span class="variable">entry</span>
  <span class="variable">Errorcode</span> <span class="number">0</span>
  <span class="variable">Args</span>    <span class="operator">:</span> <span class="variable">an</span> <span class="variable">filename</span> <span class="variable">containing</span> <span class="variable">an</span> <span class="variable">EMBL</span> <span class="variable">entry</span> <span class="variable">OR</span> <span class="variable">an</span> <span class="variable">ID</span> <span class="keyword">or</span> <span class="variable">ACCESSION</span> <span class="variable">code</span>
</code></code></pre>

<h2 id="embl2hash">embl2hash</h2>

<pre><code><code>  <span class="variable">Title</span>   <span class="operator">:</span> <span class="variable">embl2hash</span>
  <span class="variable">Function</span><span class="operator">:</span> <span class="variable">retrieves</span> <span class="variable">with</span> <span class="variable">BioPerl</span> <span class="variable">an</span> <span class="variable">EMBL</span> <span class="variable">entry</span><span class="operator">,</span> <span class="variable">parses</span> <span class="variable">it</span> <span class="keyword">and</span> <span class="variable">creates</span>
            <span class="variable">a</span> <span class="variable">hash</span> <span class="variable">that</span> <span class="variable">contains</span> <span class="variable">all</span> <span class="variable">the</span> <span class="variable">information</span><span class="operator">.</span>
  <span class="variable">Returns</span> <span class="operator">:</span> <span class="variable">a</span> <span class="variable">reference</span> <span class="variable">to</span> <span class="variable">a</span> <span class="variable">hash</span>
  <span class="variable">Errorcode</span><span class="operator">:</span> <span class="number">0</span>
  <span class="variable">Args</span>    <span class="operator">:</span> <span class="variable">a</span> <span class="variable">BioPerl</span> <span class="variable">Sequence</span> <span class="variable">Object</span> <span class="operator">(</span><span class="variable">from</span> <span class="variable">file</span> <span class="keyword">or</span> <span class="variable">web</span> <span class="variable">fetching</span><span class="operator">)</span>
            <span class="variable">two</span> <span class="variable">array</span> <span class="variable">references</span> <span class="variable">to</span> <span class="variable">skip</span> <span class="variable">features</span> <span class="keyword">and</span> <span class="variable">qualifiers</span> <span class="operator">(</span><span class="keyword">for</span>
            <span class="variable">performance</span><span class="operator">)</span>
  <span class="variable">Example</span><span class="operator">:</span> <span class="variable">@valid_features</span><span class="operator">=</span><span class="string">qw(CDS exon prim_transcript mRNA)</span><span class="operator">;</span>
           <span class="variable">@valid_qualifiers</span><span class="operator">=</span><span class="string">qw(gene codon_start db_xref product rpt_family)</span><span class="operator">;</span>
           <span class="variable">$hashref</span><span class="operator">=&amp;</span><span class="variable">embl2hash</span><span class="operator">(</span><span class="variable">$seqobj</span><span class="operator">,\</span><span class="variable">@valid_features</span><span class="operator">,\</span><span class="variable">@valid_qualifiers</span><span class="operator">);</span>
</code></code></pre>

<h2 id="novelaasequence2gene">novelaasequence2gene</h2>

<pre><code><code>  <span class="variable">Title</span>   <span class="operator">:</span> <span class="variable">novelaasequence2gene</span>
  <span class="variable">Usage</span>   <span class="operator">:</span> <span class="variable">$gene</span><span class="operator">=</span><span class="variable">Bio::LiveSeq::IO::BioPerl</span><span class="operator">-&gt;</span><span class="variable">novelaasequence2gene</span><span class="operator">(</span><span class="string">-aasequence</span> <span class="operator">=&gt;</span> <span class="string">"MGLAAPTRS*"</span><span class="operator">);</span>
          <span class="operator">:</span> <span class="variable">$gene</span><span class="operator">=</span><span class="variable">Bio::LiveSeq::IO::BioPerl</span><span class="operator">-&gt;</span><span class="variable">novelaasequence2gene</span><span class="operator">(</span><span class="string">-aasequence</span> <span class="operator">=&gt;</span> <span class="string">"MGLAAPTRS*"</span><span class="operator">,</span>
                                             <span class="string">-cusg_data</span> <span class="operator">=&gt;</span> <span class="string">"58 44 7 29 3 3 480 267 105 143 122 39 144 162 14 59 53 25 233 292 19 113 88 246 28 68 161 231 27 102 128 151 67 60 138 131 48 61 153 19 233 73 150 31 129 38 147 71 138 43 181 81 44 15 255 118 312 392 236 82 20 10 14 141"</span><span class="operator">);</span>
          <span class="operator">:</span> <span class="variable">$gene</span><span class="operator">=</span><span class="variable">Bio::LiveSeq::IO::BioPerl</span><span class="operator">-&gt;</span><span class="variable">novelaasequence2gene</span><span class="operator">(</span><span class="string">-aasequence</span> <span class="operator">=&gt;</span> <span class="string">"MGLAAPTRS*"</span><span class="operator">,</span>
                                             <span class="string">-cusg_data</span> <span class="operator">=&gt;</span> <span class="string">"58 44 7 29 3 3 480 267 105 143 122 39 144 162 14 59 53 25 233 292 19 113 88 246 28 68 161 231 27 102 128 151 67 60 138 131 48 61 153 19 233 73 150 31 129 38 147 71 138 43 181 81 44 15 255 118 312 392 236 82 20 10 14 141"</span><span class="operator">,</span>
                                             <span class="string">-translation_table</span> <span class="operator">=&gt;</span> <span class="string">"2"</span><span class="operator">,</span>
                                             <span class="string">-gene_name</span> <span class="operator">=&gt;</span> <span class="string">"tyr-kinase"</span><span class="operator">);</span>
  
  <span class="variable">Function</span><span class="operator">:</span> <span class="variable">creates</span> <span class="variable">LiveSeq</span> <span class="variable">objects</span> <span class="variable">from</span> <span class="variable">a</span> <span class="variable">novel</span> <span class="variable">amino</span> <span class="variable">acid</span> <span class="variable">sequence</span><span class="operator">,</span>
            <span class="variable">using</span> <span class="variable">codon</span> <span class="variable">usage</span> <span class="variable">information</span> <span class="operator">(</span><span class="variable">loaded</span> <span class="variable">from</span> <span class="variable">a</span> <span class="variable">file</span><span class="operator">)</span> <span class="variable">to</span> <span class="variable">choose</span>
            <span class="variable">codons</span> <span class="variable">according</span> <span class="variable">to</span> <span class="variable">relative</span> <span class="variable">frequencies</span><span class="operator">.</span>
            <span class="variable">If</span> <span class="variable">a</span> <span class="variable">codon_usage</span> <span class="variable">information</span> <span class="variable">is</span> <span class="keyword">not</span> <span class="variable">specified</span><span class="operator">,</span>
            <span class="variable">the</span> <span class="keyword">default</span> <span class="variable">is</span> <span class="variable">to</span> <span class="keyword">use</span> <span class="variable">Homo</span> <span class="variable">sapiens</span> <span class="variable">data</span> <span class="operator">(</span><span class="variable">taxonomy</span> <span class="variable">ID</span> <span class="number">9606</span><span class="operator">).</span>
            <span class="variable">If</span> <span class="variable">a</span> <span class="variable">translation_table</span> <span class="variable">ID</span> <span class="variable">is</span> <span class="keyword">not</span> <span class="variable">specified</span><span class="operator">,</span> <span class="variable">it</span> <span class="variable">will</span> <span class="keyword">default</span> <span class="variable">to</span> <span class="number">1</span>
            <span class="operator">(</span><span class="variable">standard</span> <span class="variable">code</span><span class="operator">).</span>
  <span class="variable">Returns</span> <span class="operator">:</span> <span class="variable">reference</span> <span class="variable">to</span> <span class="variable">a</span> <span class="variable">Gene</span> <span class="variable">object</span> <span class="variable">containing</span> <span class="variable">references</span> <span class="variable">to</span> <span class="variable">LiveSeq</span> <span class="variable">objects</span>
  <span class="variable">Errorcode</span> <span class="number">0</span>
  <span class="variable">Args</span>    <span class="operator">:</span> <span class="variable">string</span> <span class="variable">containing</span> <span class="variable">an</span> <span class="variable">amino</span> <span class="variable">acid</span> <span class="variable">sequence</span>
            <span class="variable">string</span> <span class="operator">(</span><span class="variable">optional</span><span class="operator">)</span> <span class="variable">with</span> <span class="variable">codon</span> <span class="variable">usage</span> <span class="variable">data</span> <span class="operator">(</span><span class="number">64</span> <span class="variable">integer</span> <span class="variable">numbers</span><span class="operator">)</span>
            <span class="variable">string</span> <span class="operator">(</span><span class="variable">optional</span><span class="operator">)</span> <span class="variable">specifying</span> <span class="variable">a</span> <span class="variable">gene_name</span>
            <span class="variable">integer</span> <span class="operator">(</span><span class="variable">optional</span><span class="operator">)</span> <span class="variable">specifying</span> <span class="variable">a</span> <span class="variable">translation_table</span> <span class="variable">ID</span>
</code></code></pre>


</body>

</html>


