<?xml version="1.0" ?>
<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.0 Transitional//EN">
<html>
<head>
<!-- saved from url=(0017)http://localhost/ -->
<script language="JavaScript" src="../../../displayToc.js"></script>
<script language="JavaScript" src="../../../tocParas.js"></script>
<script language="JavaScript" src="../../../tocTab.js"></script>
<link rel="stylesheet" type="text/css" href="../../../scineplex.css">
<title></title>
<link rel="stylesheet" href="../../../Active.css" type="text/css" />
<meta http-equiv="content-type" content="text/html; charset=utf-8" />
<link rev="made" href="mailto:support@ActiveState.com" />
</head>

<body>



<ul id="index">
  <li><a href="#NAME">NAME</a></li>
  <li><a href="#SYNOPSIS">SYNOPSIS</a></li>
  <li><a href="#DESCRIPTION">DESCRIPTION</a></li>
  <li><a href="#CONSTRUCTORS">CONSTRUCTORS</a>
    <ul>
      <li><a href="#Bio::AlignIO-new-">Bio::AlignIO-&gt;new()</a></li>
      <li><a href="#Bio::AlignIO-newFh-">Bio::AlignIO-&gt;newFh()</a></li>
    </ul>
  </li>
  <li><a href="#OBJECT-METHODS">OBJECT METHODS</a>
    <ul>
      <li><a href="#alignment-AlignIO-next_aln-">$alignment = $AlignIO-&gt;next_aln()</a></li>
      <li><a href="#AlignIO-write_aln-aln-">$AlignIO-&gt;write_aln($aln)</a></li>
      <li><a href="#TIEHANDLE-READLINE-PRINT-">TIEHANDLE(), READLINE(), PRINT()</a></li>
    </ul>
  </li>
  <li><a href="#FEEDBACK">FEEDBACK</a>
    <ul>
      <li><a href="#Mailing-Lists">Mailing Lists</a></li>
      <li><a href="#Support">Support</a></li>
      <li><a href="#Reporting-Bugs">Reporting Bugs</a></li>
    </ul>
  </li>
  <li><a href="#AUTHOR---Peter-Schattner">AUTHOR - Peter Schattner</a></li>
  <li><a href="#CONTRIBUTORS">CONTRIBUTORS</a></li>
  <li><a href="#APPENDIX">APPENDIX</a>
    <ul>
      <li><a href="#new">new</a></li>
      <li><a href="#newFh">newFh</a></li>
      <li><a href="#fh1">fh</a></li>
      <li><a href="#load_format_module">_load_format_module</a></li>
      <li><a href="#next_aln">next_aln</a></li>
      <li><a href="#write_aln">write_aln</a></li>
      <li><a href="#guess_format">_guess_format</a></li>
      <li><a href="#force_displayname_flat">force_displayname_flat</a></li>
      <li><a href="#alphabet">alphabet</a></li>
    </ul>
  </li>
</ul>

<h1 id="NAME">NAME</h1>

<p>Bio::AlignIO - Handler for AlignIO Formats</p>

<h1 id="SYNOPSIS">SYNOPSIS</h1>

<pre><code><code>    <span class="keyword">use</span> <span class="variable">Bio::AlignIO</span><span class="operator">;</span>
    
    <span class="variable">$inputfilename</span> <span class="operator">=</span> <span class="string">"testaln.fasta"</span><span class="operator">;</span>
    <span class="variable">$in</span>  <span class="operator">=</span> <span class="variable">Bio::AlignIO</span><span class="operator">-&gt;</span><span class="variable">new</span><span class="operator">(</span><span class="string">-file</span>   <span class="operator">=&gt;</span> <span class="variable">$inputfilename</span> <span class="operator">,</span>
                             <span class="string">-format</span> <span class="operator">=&gt;</span> <span class="string">'fasta'</span><span class="operator">);</span>
    <span class="variable">$out</span> <span class="operator">=</span> <span class="variable">Bio::AlignIO</span><span class="operator">-&gt;</span><span class="variable">new</span><span class="operator">(</span><span class="string">-file</span>   <span class="operator">=&gt;</span> <span class="string">"&gt;out.aln.pfam"</span> <span class="operator">,</span>
                             <span class="string">-format</span> <span class="operator">=&gt;</span> <span class="string">'pfam'</span><span class="operator">);</span>
    
    <span class="keyword">while</span> <span class="operator">(</span> <span class="keyword">my</span> <span class="variable">$aln</span> <span class="operator">=</span> <span class="variable">$in</span><span class="operator">-&gt;</span><span class="variable">next_aln</span><span class="operator">()</span> <span class="operator">)</span> <span class="operator">{</span>
        <span class="variable">$out</span><span class="operator">-&gt;</span><span class="variable">write_aln</span><span class="operator">(</span><span class="variable">$aln</span><span class="operator">);</span>
    <span class="operator">}</span>
    
    <span class="comment"># OR</span>
    
    <span class="keyword">use</span> <span class="variable">Bio::AlignIO</span><span class="operator">;</span>
    
    <span class="keyword">open</span> <span class="variable">MYIN</span><span class="operator">,</span><span class="string">"testaln.fasta"</span><span class="operator">;</span>
    <span class="variable">$in</span>  <span class="operator">=</span> <span class="variable">Bio::AlignIO</span><span class="operator">-&gt;</span><span class="variable">newFh</span><span class="operator">(</span><span class="string">-fh</span>     <span class="operator">=&gt;</span> <span class="operator">\</span><span class="variable">*MYIN</span><span class="operator">,</span>
                               <span class="string">-format</span> <span class="operator">=&gt;</span> <span class="string">'fasta'</span><span class="operator">);</span>
    <span class="keyword">open</span> <span class="keyword">my</span> <span class="variable">$MYOUT</span><span class="operator">,</span> <span class="string">'&gt;'</span><span class="operator">,</span> <span class="string">'testaln.pfam'</span><span class="operator">;</span>
    <span class="variable">$out</span> <span class="operator">=</span> <span class="variable">Bio::AlignIO</span><span class="operator">-&gt;</span><span class="variable">newFh</span><span class="operator">(</span><span class="string">-fh</span>     <span class="operator">=&gt;</span>  <span class="variable">$MYOUT</span><span class="operator">,</span>
                               <span class="string">-format</span> <span class="operator">=&gt;</span> <span class="string">'pfam'</span><span class="operator">);</span>
    
    <span class="comment"># World's smallest Fasta&lt;-&gt;pfam format converter:</span>
    <span class="keyword">print</span> <span class="variable">$out</span> <span class="variable">$_</span> <span class="keyword">while</span> <span class="operator">&lt;</span><span class="variable">$in</span><span class="operator">&gt;;</span>
</code></code></pre>

<h1 id="DESCRIPTION">DESCRIPTION</h1>

<p><a href="../../../../.cpan/build/BioPerl-1.6.901-ghKY8U/blib/lib/Bio/AlignIO.html">Bio::AlignIO</a> is a handler module for the formats in the AlignIO set, for example, <a href="../../../../.cpan/build/BioPerl-1.6.901-ghKY8U/blib/lib/Bio/AlignIO/fasta.html">Bio::AlignIO::fasta</a>. It is the officially sanctioned way of getting at the alignment objects. The resulting alignment is a <a href="../../../../.cpan/build/BioPerl-1.6.901-ghKY8U/blib/lib/Bio/Align/AlignI.html">Bio::Align::AlignI</a>-compliant object.</p>

<p>The idea is that you request an object for a particular format. All the objects have a notion of an internal file that is read from or written to. A particular AlignIO object instance is configured for either input or output, you can think of it as a stream object.</p>

<p>Each object has functions:</p>

<pre><code><code>   <span class="variable">$stream</span><span class="operator">-&gt;</span><span class="variable">next_aln</span><span class="operator">();</span>
</code></code></pre>

<p>And:</p>

<pre><code><code>   <span class="variable">$stream</span><span class="operator">-&gt;</span><span class="variable">write_aln</span><span class="operator">(</span><span class="variable">$aln</span><span class="operator">);</span>
</code></code></pre>

<p>Also:</p>

<pre><code><code>   $stream-&gt;type() # returns &#39;INPUT&#39; or &#39;OUTPUT&#39;</code></code></pre>

<p>As an added bonus, you can recover a filehandle that is tied to the AlignIO object, allowing you to use the standard &lt;&gt; and print operations to read and write alignment objects:</p>

<pre><code><code>    <span class="keyword">use</span> <span class="variable">Bio::AlignIO</span><span class="operator">;</span>
    
    <span class="comment"># read from standard input</span>
    <span class="variable">$stream</span> <span class="operator">=</span> <span class="variable">Bio::AlignIO</span><span class="operator">-&gt;</span><span class="variable">newFh</span><span class="operator">(</span><span class="string">-format</span> <span class="operator">=&gt;</span> <span class="string">'Fasta'</span><span class="operator">);</span>
    
    <span class="keyword">while</span> <span class="operator">(</span> <span class="variable">$aln</span> <span class="operator">=</span> <span class="operator">&lt;</span><span class="variable">$stream</span><span class="operator">&gt;</span> <span class="operator">)</span> <span class="operator">{</span>
             <span class="comment"># do something with $aln</span>
    <span class="operator">}</span>
</code></code></pre>

<p>And:</p>

<pre><code><code>    <span class="keyword">print</span> <span class="variable">$stream</span> <span class="variable">$aln</span><span class="operator">;</span> <span class="comment"># when stream is in output mode</span>
</code></code></pre>

<p><a href="../../../../.cpan/build/BioPerl-1.6.901-ghKY8U/blib/lib/Bio/AlignIO.html">Bio::AlignIO</a> is patterned on the <a href="../../../../.cpan/build/BioPerl-1.6.901-ghKY8U/blib/lib/Bio/SeqIO.html">Bio::SeqIO</a> module and shares most of its features. One significant difference is that <a href="../../../../.cpan/build/BioPerl-1.6.901-ghKY8U/blib/lib/Bio/AlignIO.html">Bio::AlignIO</a> usually handles IO for only a single alignment at a time, whereas <a href="../../../../.cpan/build/BioPerl-1.6.901-ghKY8U/blib/lib/Bio/SeqIO.html">Bio::SeqIO</a> handles IO for multiple sequences in a single stream. The principal reason for this is that whereas simultaneously handling multiple sequences is a common requirement, simultaneous handling of multiple alignments is not. The only current exception is format <code><code>bl2seq</code></code> which parses results of the BLAST <code><code>bl2seq</code></code> program and which may produce several alignment pairs. This set of alignment pairs can be read using multiple calls to <a>next_aln</a>.</p>

<h1 id="CONSTRUCTORS">CONSTRUCTORS</h1>

<h2 id="Bio::AlignIO-new-">Bio::AlignIO-&gt;new()</h2>

<pre><code><code>   <span class="variable">$seqIO</span> <span class="operator">=</span> <span class="variable">Bio::AlignIO</span><span class="operator">-&gt;</span><span class="variable">new</span><span class="operator">(</span><span class="string">-file</span> <span class="operator">=&gt;</span> <span class="string">'filename'</span><span class="operator">,</span>   <span class="string">-format</span><span class="operator">=&gt;</span><span class="variable">$format</span><span class="operator">);</span>
   <span class="variable">$seqIO</span> <span class="operator">=</span> <span class="variable">Bio::AlignIO</span><span class="operator">-&gt;</span><span class="variable">new</span><span class="operator">(</span><span class="string">-fh</span>   <span class="operator">=&gt;</span> <span class="operator">\</span><span class="variable">*FILEHANDLE</span><span class="operator">,</span> <span class="string">-format</span><span class="operator">=&gt;</span><span class="variable">$format</span><span class="operator">);</span>
   <span class="variable">$seqIO</span> <span class="operator">=</span> <span class="variable">Bio::AlignIO</span><span class="operator">-&gt;</span><span class="variable">new</span><span class="operator">(</span><span class="string">-format</span> <span class="operator">=&gt;</span> <span class="variable">$format</span><span class="operator">);</span>
   <span class="variable">$seqIO</span> <span class="operator">=</span> <span class="variable">Bio::AlignIO</span><span class="operator">-&gt;</span><span class="variable">new</span><span class="operator">(</span><span class="string">-fh</span> <span class="operator">=&gt;</span> <span class="operator">\</span><span class="variable">*STDOUT</span><span class="operator">,</span> <span class="string">-format</span> <span class="operator">=&gt;</span> <span class="variable">$format</span><span class="operator">);</span>
</code></code></pre>

<p>The <a>new</a> class method constructs a new <a href="../../../../.cpan/build/BioPerl-1.6.901-ghKY8U/blib/lib/Bio/AlignIO.html">Bio::AlignIO</a> object. The returned object can be used to retrieve or print alignment objects. <a>new</a> accepts the following parameters:</p>

<dl>

<dt id="file">-file</dt>
<dd>

<p>A file path to be opened for reading or writing. The usual Perl conventions apply:</p>

<pre><code><code>   &#39;file&#39;       # open file for reading
   &#39;&gt;file&#39;      # open file for writing
   &#39;&gt;&gt;file&#39;     # open file for appending
   &#39;+&lt;file&#39;     # open file read/write
   &#39;command |&#39;  # open a pipe from the command
   &#39;| command&#39;  # open a pipe to the command</code></code></pre>

</dd>
<dt id="fh">-fh</dt>
<dd>

<p>You may provide new() with a previously-opened filehandle. For example, to read from STDIN:</p>

<pre><code><code>   <span class="variable">$seqIO</span> <span class="operator">=</span> <span class="variable">Bio::AlignIO</span><span class="operator">-&gt;</span><span class="variable">new</span><span class="operator">(</span><span class="string">-fh</span> <span class="operator">=&gt;</span> <span class="operator">\</span><span class="variable">*STDIN</span><span class="operator">);</span>
</code></code></pre>

<p>Note that you must pass filehandles as references to globs.</p>

<p>If neither a filehandle nor a filename is specified, then the module will read from the @ARGV array or STDIN, using the familiar &lt;&gt; semantics.</p>

</dd>
<dt id="format">-format</dt>
<dd>

<p>Specify the format of the file. Supported formats include:</p>

<pre><code><code>   bl2seq      Bl2seq Blast output
   clustalw    clustalw (.aln) format
   emboss      EMBOSS water and needle format
   fasta       FASTA format
   maf         Multiple Alignment Format
   mase        mase (seaview) format
   mega        MEGA format
   meme        MEME format
   msf         msf (GCG) format
   nexus       Swofford et al NEXUS format
   pfam        Pfam sequence alignment format
   phylip      Felsenstein PHYLIP format
   prodom      prodom (protein domain) format
   psi         PSI-BLAST format
   selex       selex (hmmer) format
   stockholm   stockholm format</code></code></pre>

<p>Currently only those formats which were implemented in <a href="../../../../.cpan/build/BioPerl-1.6.901-ghKY8U/blib/lib/Bio/SimpleAlign.html">Bio::SimpleAlign</a> have been incorporated into <a href="../../../../.cpan/build/BioPerl-1.6.901-ghKY8U/blib/lib/Bio/AlignIO.html">Bio::AlignIO</a>. Specifically, <code><code>mase</code></code>, <code><code>stockholm</code></code> and <code><code>prodom</code></code> have only been implemented for input. See the specific module (e.g. <a href="../../../../.cpan/build/BioPerl-1.6.901-ghKY8U/blib/lib/Bio/AlignIO/prodom.html">Bio::AlignIO::prodom</a>) for notes on supported versions.</p>

<p>If no format is specified and a filename is given, then the module will attempt to deduce it from the filename suffix. If this is unsuccessful, <code><code>fasta</code></code> format is assumed.</p>

<p>The format name is case insensitive; <code><code>FASTA</code></code>, <code><code>Fasta</code></code> and <code><code>fasta</code></code> are all treated equivalently.</p>

</dd>
</dl>

<h2 id="Bio::AlignIO-newFh-">Bio::AlignIO-&gt;newFh()</h2>

<pre><code><code>   <span class="variable">$fh</span> <span class="operator">=</span> <span class="variable">Bio::AlignIO</span><span class="operator">-&gt;</span><span class="variable">newFh</span><span class="operator">(</span><span class="string">-fh</span>   <span class="operator">=&gt;</span> <span class="operator">\</span><span class="variable">*FILEHANDLE</span><span class="operator">,</span> <span class="string">-format</span><span class="operator">=&gt;</span><span class="variable">$format</span><span class="operator">);</span>
   <span class="comment"># read from STDIN or use @ARGV:</span>
   <span class="variable">$fh</span> <span class="operator">=</span> <span class="variable">Bio::AlignIO</span><span class="operator">-&gt;</span><span class="variable">newFh</span><span class="operator">(</span><span class="string">-format</span> <span class="operator">=&gt;</span> <span class="variable">$format</span><span class="operator">);</span>
</code></code></pre>

<p>This constructor behaves like <a>new</a>, but returns a tied filehandle rather than a <a href="../../../../.cpan/build/BioPerl-1.6.901-ghKY8U/blib/lib/Bio/AlignIO.html">Bio::AlignIO</a> object. You can read sequences from this object using the familiar &lt;&gt; operator, and write to it using <a>print</a>. The usual array and $_ semantics work. For example, you can read all sequence objects into an array like this:</p>

<pre><code><code>  <span class="variable">@sequences</span> <span class="operator">=</span> <span class="operator">&lt;</span><span class="variable">$fh</span><span class="operator">&gt;;</span>
</code></code></pre>

<p>Other operations, such as read(), sysread(), write(), close(), and printf() are not supported.</p>

<dl>

<dt id="flush">-flush</dt>
<dd>

<p>By default, all files (or filehandles) opened for writing alignments will be flushed after each write_aln() making the file immediately usable. If you do not need this facility and would like to marginally improve the efficiency of writing multiple sequences to the same file (or filehandle), pass the -flush option &#39;0&#39; or any other value that evaluates as defined but false:</p>

<pre><code><code>  <span class="keyword">my</span> <span class="variable">$clustal</span> <span class="operator">=</span> <span class="variable">Bio::AlignIO</span><span class="operator">-&gt;</span><span class="variable">new</span><span class="operator">(</span> <span class="string">-file</span>   <span class="operator">=&gt;</span> <span class="string">"&lt;prot.aln"</span><span class="operator">,</span>
                                   <span class="string">-format</span> <span class="operator">=&gt;</span> <span class="string">"clustalw"</span> <span class="operator">);</span>
  <span class="keyword">my</span> <span class="variable">$msf</span> <span class="operator">=</span> <span class="variable">Bio::AlignIO</span><span class="operator">-&gt;</span><span class="variable">new</span><span class="operator">(</span><span class="string">-file</span>   <span class="operator">=&gt;</span> <span class="string">"&gt;prot.msf"</span><span class="operator">,</span>
                              <span class="string">-format</span> <span class="operator">=&gt;</span> <span class="string">"msf"</span><span class="operator">,</span>
                              <span class="string">-flush</span>  <span class="operator">=&gt;</span> <span class="number">0</span> <span class="operator">);</span> <span class="comment"># go as fast as we can!</span>
  <span class="keyword">while</span><span class="operator">(</span><span class="variable">$seq</span> <span class="operator">=</span> <span class="variable">$clustal</span><span class="operator">-&gt;</span><span class="variable">next_aln</span><span class="operator">)</span> <span class="operator">{</span> <span class="variable">$msf</span><span class="operator">-&gt;</span><span class="variable">write_aln</span><span class="operator">(</span><span class="variable">$seq</span><span class="operator">)</span> <span class="operator">}</span>
</code></code></pre>

</dd>
</dl>

<h1 id="OBJECT-METHODS">OBJECT METHODS</h1>

<p>See below for more detailed summaries. The main methods are:</p>

<h2 id="alignment-AlignIO-next_aln-">$alignment = $AlignIO-&gt;next_aln()</h2>

<p>Fetch an alignment from a formatted file.</p>

<h2 id="AlignIO-write_aln-aln-">$AlignIO-&gt;write_aln($aln)</h2>

<p>Write the specified alignment to a file..</p>

<h2 id="TIEHANDLE-READLINE-PRINT-">TIEHANDLE(), READLINE(), PRINT()</h2>

<p>These provide the tie interface. See <a href="../../../lib/pod/perltie.html">perltie</a> for more details.</p>

<h1 id="FEEDBACK">FEEDBACK</h1>

<h2 id="Mailing-Lists">Mailing Lists</h2>

<p>User feedback is an integral part of the evolution of this and other Bioperl modules. Send your comments and suggestions preferably to one of the Bioperl mailing lists. Your participation is much appreciated.</p>

<pre><code><code>  bioperl-l@bioperl.org                  - General discussion
  http://bioperl.org/wiki/Mailing_lists  - About the mailing lists</code></code></pre>

<h2 id="Support">Support</h2>

<p>Please direct usage questions or support issues to the mailing list:</p>

<p><i>bioperl-l@bioperl.org</i></p>

<p>rather than to the module maintainer directly. Many experienced and reponsive experts will be able look at the problem and quickly address it. Please include a thorough description of the problem with code and data examples if at all possible.</p>

<h2 id="Reporting-Bugs">Reporting Bugs</h2>

<p>Report bugs to the Bioperl bug tracking system to help us keep track the bugs and their resolution. Bug reports can be submitted via the web:</p>

<pre><code><code>  https://redmine.open-bio.org/projects/bioperl/</code></code></pre>

<h1 id="AUTHOR---Peter-Schattner">AUTHOR - Peter Schattner</h1>

<p>Email: schattner@alum.mit.edu</p>

<h1 id="CONTRIBUTORS">CONTRIBUTORS</h1>

<p>Jason Stajich, jason@bioperl.org</p>

<h1 id="APPENDIX">APPENDIX</h1>

<p>The rest of the documentation details each of the object methods. Internal methods are usually preceded with a _</p>

<h2 id="new">new</h2>

<pre><code><code> Title   : new
 Usage   : $stream = Bio::AlignIO-&gt;new(-file =&gt; $filename,
                                       -format =&gt; &#39;Format&#39;)
 Function: Returns a new seqstream
 Returns : A Bio::AlignIO::Handler initialised with
           the appropriate format
 Args    : -file =&gt; $filename
           -format =&gt; format
           -fh =&gt; filehandle to attach to
           -displayname_flat =&gt; 1 [optional]
                                to force the displayname to not show start/end
                                information</code></code></pre>

<h2 id="newFh">newFh</h2>

<pre><code><code> <span class="variable">Title</span>   <span class="operator">:</span> <span class="variable">newFh</span>
 <span class="variable">Usage</span>   <span class="operator">:</span> <span class="variable">$fh</span> <span class="operator">=</span> <span class="variable">Bio::AlignIO</span><span class="operator">-&gt;</span><span class="variable">newFh</span><span class="operator">(</span><span class="string">-file</span><span class="operator">=&gt;</span><span class="variable">$filename</span><span class="operator">,</span><span class="string">-format</span><span class="operator">=&gt;</span><span class="string">'Format'</span><span class="operator">)</span>
 <span class="variable">Function</span><span class="operator">:</span> <span class="variable">does</span> <span class="variable">a</span> <span class="variable">new</span><span class="operator">()</span> <span class="variable">followed</span> <span class="variable">by</span> <span class="variable">an</span> <span class="variable">fh</span><span class="operator">()</span>
 <span class="variable">Example</span> <span class="operator">:</span> <span class="variable">$fh</span> <span class="operator">=</span> <span class="variable">Bio::AlignIO</span><span class="operator">-&gt;</span><span class="variable">newFh</span><span class="operator">(</span><span class="string">-file</span><span class="operator">=&gt;</span><span class="variable">$filename</span><span class="operator">,</span><span class="string">-format</span><span class="operator">=&gt;</span><span class="string">'Format'</span><span class="operator">)</span>
           <span class="variable">$sequence</span> <span class="operator">=</span> <span class="operator">&lt;</span><span class="variable">$fh</span><span class="operator">&gt;;</span>   <span class="comment"># read a sequence object</span>
           <span class="keyword">print</span> <span class="variable">$fh</span> <span class="variable">$sequence</span><span class="operator">;</span> <span class="comment"># write a sequence object</span>
 <span class="variable">Returns</span> <span class="operator">:</span> <span class="variable">filehandle</span> <span class="keyword">tied</span> <span class="variable">to</span> <span class="variable">the</span> <span class="variable">Bio::AlignIO::Fh</span> <span class="variable">class</span>
 <span class="variable">Args</span>    <span class="operator">:</span>
</code></code></pre>

<h2 id="fh1">fh</h2>

<pre><code><code> <span class="variable">Title</span>   <span class="operator">:</span> <span class="variable">fh</span>
 <span class="variable">Usage</span>   <span class="operator">:</span> <span class="variable">$obj</span><span class="operator">-&gt;</span><span class="variable">fh</span>
 <span class="variable">Function</span><span class="operator">:</span>
 <span class="variable">Example</span> <span class="operator">:</span> <span class="variable">$fh</span> <span class="operator">=</span> <span class="variable">$obj</span><span class="operator">-&gt;</span><span class="variable">fh</span><span class="operator">;</span>      <span class="comment"># make a tied filehandle</span>
           <span class="variable">$sequence</span> <span class="operator">=</span> <span class="operator">&lt;</span><span class="variable">$fh</span><span class="operator">&gt;;</span>   <span class="comment"># read a sequence object</span>
           <span class="keyword">print</span> <span class="variable">$fh</span> <span class="variable">$sequence</span><span class="operator">;</span> <span class="comment"># write a sequence object</span>
 <span class="variable">Returns</span> <span class="operator">:</span> <span class="variable">filehandle</span> <span class="keyword">tied</span> <span class="variable">to</span> <span class="variable">the</span> <span class="variable">Bio::AlignIO::Fh</span> <span class="variable">class</span>
 <span class="variable">Args</span>    <span class="operator">:</span>
</code></code></pre>

<h2 id="load_format_module">_load_format_module</h2>

<pre><code><code> Title   : _load_format_module
 Usage   : *INTERNAL AlignIO stuff*
 Function: Loads up (like use) a module at run time on demand
 Example :
 Returns :
 Args    :</code></code></pre>

<h2 id="next_aln">next_aln</h2>

<pre><code><code> Title   : next_aln
 Usage   : $aln = stream-&gt;next_aln
 Function: reads the next $aln object from the stream
 Returns : a Bio::Align::AlignI compliant object
 Args    :</code></code></pre>

<h2 id="write_aln">write_aln</h2>

<pre><code><code> Title   : write_aln
 Usage   : $stream-&gt;write_aln($aln)
 Function: writes the $aln object into the stream
 Returns : 1 for success and 0 for error
 Args    : Bio::Seq object</code></code></pre>

<h2 id="guess_format">_guess_format</h2>

<pre><code><code> Title   : _guess_format
 Usage   : $obj-&gt;_guess_format($filename)
 Function:
 Example :
 Returns : guessed format of filename (lower case)
 Args    :</code></code></pre>

<h2 id="force_displayname_flat">force_displayname_flat</h2>

<pre><code><code> Title   : force_displayname_flat
 Usage   : $obj-&gt;force_displayname_flat($newval)
 Function:
 Example :
 Returns : value of force_displayname_flat (a scalar)
 Args    : on set, new value (a scalar or undef, optional)</code></code></pre>

<h2 id="alphabet">alphabet</h2>

<pre><code><code> <span class="variable">Title</span>   <span class="operator">:</span> <span class="variable">alphabet</span>
 <span class="variable">Usage</span>   <span class="operator">:</span> <span class="variable">$obj</span><span class="operator">-&gt;</span><span class="variable">alphabet</span><span class="operator">(</span><span class="variable">$newval</span><span class="operator">)</span>
 <span class="variable">Function</span><span class="operator">:</span> <span class="variable">Get</span><span class="operator">/</span><span class="variable">Set</span> <span class="variable">alphabet</span> <span class="keyword">for</span> <span class="variable">purpose</span> <span class="variable">of</span> <span class="variable">passing</span> <span class="variable">to</span> <span class="variable">Bio::LocatableSeq</span> <span class="variable">creation</span>
 <span class="variable">Example</span> <span class="operator">:</span> <span class="variable">$obj</span><span class="operator">-&gt;</span><span class="variable">alphabet</span><span class="operator">(</span><span class="string">'dna'</span><span class="operator">);</span>
 <span class="variable">Returns</span> <span class="operator">:</span> <span class="variable">value</span> <span class="variable">of</span> <span class="variable">alphabet</span> <span class="operator">(</span><span class="variable">a</span> <span class="keyword">scalar</span><span class="operator">)</span>
 <span class="variable">Args</span>    <span class="operator">:</span> <span class="variable">on</span> <span class="variable">set</span><span class="operator">,</span> <span class="variable">new</span> <span class="variable">value</span> <span class="operator">(</span><span class="variable">a</span> <span class="keyword">scalar</span> <span class="keyword">or</span> <span class="keyword">undef</span><span class="operator">,</span> <span class="variable">optional</span><span class="operator">)</span>
</code></code></pre>


</body>

</html>


