<?xml version="1.0" ?>
<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.0 Transitional//EN">
<html>
<head>
<!-- saved from url=(0017)http://localhost/ -->
<script language="JavaScript" src="../../../displayToc.js"></script>
<script language="JavaScript" src="../../../tocParas.js"></script>
<script language="JavaScript" src="../../../tocTab.js"></script>
<link rel="stylesheet" type="text/css" href="../../../scineplex.css">
<title></title>
<link rel="stylesheet" href="../../../Active.css" type="text/css" />
<meta http-equiv="content-type" content="text/html; charset=utf-8" />
<link rev="made" href="mailto:support@ActiveState.com" />
</head>

<body>



<ul id="index">
  <li><a href="#NAME">NAME</a></li>
  <li><a href="#SYNOPSIS">SYNOPSIS</a></li>
  <li><a href="#DESCRIPTION">DESCRIPTION</a>
    <ul>
      <li><a href="#Phylogenetic-Networks">Phylogenetic Networks</a></li>
      <li><a href="#eNewick-description">eNewick description</a></li>
      <li><a href="#Tree-child-phylogenetic-networks">Tree-child phylogenetic networks</a></li>
      <li><a href="#Tripartitions">Tripartitions</a></li>
      <li><a href="#Time-consistency">Time-consistency</a></li>
    </ul>
  </li>
  <li><a href="#AUTHOR">AUTHOR</a></li>
  <li><a href="#SEE-ALSO">SEE ALSO</a></li>
  <li><a href="#APPENDIX">APPENDIX</a>
    <ul>
      <li><a href="#new">new</a></li>
      <li><a href="#is_leaf">is_leaf</a></li>
      <li><a href="#is_root">is_root</a></li>
      <li><a href="#is_tree_node">is_tree_node</a></li>
      <li><a href="#is_hybrid_node">is_hybrid_node</a></li>
      <li><a href="#is_tree_child">is_tree_child</a></li>
      <li><a href="#nodes">nodes</a></li>
      <li><a href="#leaves">leaves</a></li>
      <li><a href="#roots">roots</a></li>
      <li><a href="#internal_nodes">internal_nodes</a></li>
      <li><a href="#tree_nodes">tree_nodes</a></li>
      <li><a href="#hybrid_nodes">hybrid_nodes</a></li>
      <li><a href="#graph">graph</a></li>
      <li><a href="#edges">edges</a></li>
      <li><a href="#tree_edges">tree_edges</a></li>
      <li><a href="#hybrid_edges">hybrid_edges</a></li>
      <li><a href="#explode">explode</a></li>
      <li><a href="#mudata">mudata</a></li>
      <li><a href="#heights">heights</a></li>
      <li><a href="#mu_distance">mu_distance</a></li>
      <li><a href="#mu_distance_generalized">mu_distance_generalized</a></li>
      <li><a href="#tripartitions">tripartitions</a></li>
      <li><a href="#is_time_consistent">is_time_consistent</a></li>
      <li><a href="#temporal_representation">temporal_representation</a></li>
      <li><a href="#contract_elementary">contract_elementary</a></li>
      <li><a href="#optimal_alignment">optimal_alignment</a></li>
      <li><a href="#optimal_alignment_generalized">optimal_alignment_generalized</a></li>
      <li><a href="#topological_restriction">topological_restriction</a></li>
      <li><a href="#eNewick">eNewick</a></li>
      <li><a href="#eNewick_full">eNewick_full</a></li>
      <li><a href="#display">display</a></li>
    </ul>
  </li>
</ul>

<h1 id="NAME">NAME</h1>

<p>Bio::PhyloNetwork - Module to compute with Phylogenetic Networks</p>

<h1 id="SYNOPSIS">SYNOPSIS</h1>

<pre><code><code> <span class="keyword">use</span> <span class="variable">Bio::PhyloNetwork</span><span class="operator">;</span>
 
 <span class="comment"># Create a PhyloNetwork object from a eNewick string</span>
 <span class="keyword">my</span> <span class="variable">$net1</span><span class="operator">=</span><span class="variable">Bio::PhyloNetwork</span><span class="operator">-&gt;</span><span class="variable">new</span><span class="operator">(</span>
   <span class="string">-eNewick</span><span class="operator">=&gt;</span><span class="string">'t0:((H1,(H2,l2)),H2); H1:((H3,l1)); H2:((H3,(l3,H1))); H3:(l4);'</span>
 <span class="operator">);</span>
 
 <span class="comment"># Print all available data</span>
 <span class="keyword">print</span> <span class="variable">$net1</span><span class="operator">;</span>
 
 <span class="comment"># Rebuild $net1 from its mu_data</span>
 <span class="keyword">my</span> <span class="variable">%mudata</span><span class="operator">=</span><span class="variable">$net1</span><span class="operator">-&gt;</span><span class="variable">mudata</span><span class="operator">();</span>
 <span class="keyword">my</span> <span class="variable">$net2</span><span class="operator">=</span><span class="variable">Bio::PhyloNetwork</span><span class="operator">-&gt;</span><span class="variable">new</span><span class="operator">(</span><span class="string">-mudata</span><span class="operator">=&gt;\</span><span class="variable">%mudata</span><span class="operator">,</span><span class="string">-numleaves</span><span class="operator">=&gt;</span><span class="number">4</span><span class="operator">);</span>
 <span class="keyword">print</span> <span class="variable">$net2</span><span class="operator">;</span>
 <span class="keyword">print</span> <span class="string">"d="</span><span class="operator">.</span><span class="variable">$net1</span><span class="operator">-&gt;</span><span class="variable">mu_distance</span><span class="operator">(</span><span class="variable">$net2</span><span class="operator">).</span><span class="string">"\n"</span><span class="operator">;</span>
 
 <span class="comment"># Get another one and compute distance</span>
 <span class="keyword">my</span> <span class="variable">$net3</span><span class="operator">=</span><span class="variable">Bio::PhyloNetwork</span><span class="operator">-&gt;</span><span class="variable">new</span><span class="operator">(</span>
   <span class="string">-eNewick</span><span class="operator">=&gt;</span><span class="string">'(l2,((l1,(H1,l4)),H1))r; (l3)H1;'</span>
 <span class="operator">);</span>
 <span class="keyword">print</span> <span class="string">"d="</span><span class="operator">.</span><span class="variable">$net1</span><span class="operator">-&gt;</span><span class="variable">mu_distance</span><span class="operator">(</span><span class="variable">$net3</span><span class="operator">).</span><span class="string">"\n"</span><span class="operator">;</span>
 
 <span class="comment"># ...and find an optimal alignment w.r.t. the Manhattan distance (default)</span>
 <span class="keyword">my</span> <span class="operator">(</span><span class="variable">$weight</span><span class="operator">,</span><span class="variable">%alignment</span><span class="operator">)=</span><span class="variable">$net1</span><span class="operator">-&gt;</span><span class="variable">optimal_alignment</span><span class="operator">(</span><span class="variable">$net3</span><span class="operator">);</span>
 <span class="keyword">print</span> <span class="string">"weight:</span><span class="variable">$weight</span><span class="string">\n"</span><span class="operator">;</span>
 <span class="keyword">foreach</span> <span class="keyword">my</span> <span class="variable">$node1</span> <span class="operator">(</span><span class="keyword">keys</span> <span class="variable">%alignment</span><span class="operator">)</span> <span class="operator">{</span>
   <span class="keyword">print</span> <span class="string">"</span><span class="variable">$node1</span><span class="string"> =&gt; "</span><span class="operator">.</span><span class="variable">$alignment</span><span class="operator">{</span><span class="variable">$node1</span><span class="operator">}</span><span class="operator">.</span><span class="string">"\n"</span><span class="operator">;</span>
 <span class="operator">}</span>
 <span class="comment"># ...or the Hamming distance</span>
 
 <span class="keyword">my</span> <span class="operator">(</span><span class="variable">$weightH</span><span class="operator">,</span><span class="variable">%alignmentH</span><span class="operator">)=</span><span class="variable">$net1</span><span class="operator">-&gt;</span><span class="variable">optimal_alignment</span><span class="operator">(</span><span class="variable">$net3</span><span class="operator">,</span><span class="string">-metric</span><span class="operator">=&gt;</span><span class="string">'Hamming'</span><span class="operator">);</span>
 <span class="keyword">print</span> <span class="string">"weight:</span><span class="variable">$weightH</span><span class="string">\n"</span><span class="operator">;</span>
 <span class="keyword">foreach</span> <span class="keyword">my</span> <span class="variable">$node1</span> <span class="operator">(</span><span class="keyword">keys</span> <span class="variable">%alignmentH</span><span class="operator">)</span> <span class="operator">{</span>
   <span class="keyword">print</span> <span class="string">"</span><span class="variable">$node1</span><span class="string"> =&gt; "</span><span class="operator">.</span><span class="variable">$alignmentH</span><span class="operator">{</span><span class="variable">$node1</span><span class="operator">}</span><span class="operator">.</span><span class="string">"\n"</span><span class="operator">;</span>
 <span class="operator">}</span>
 
 <span class="comment"># Test for time consistency of $net1</span>
 <span class="keyword">if</span> <span class="operator">(</span><span class="variable">$net1</span><span class="operator">-&gt;</span><span class="variable">is_time_consistent</span><span class="operator">)</span> <span class="operator">{</span>
   <span class="keyword">print</span> <span class="string">"net1 is time consistent\n"</span>
 <span class="operator">}</span>
 <span class="keyword">else</span> <span class="operator">{</span>
   <span class="keyword">print</span> <span class="string">"net1 is not time consistent\n"</span>
 <span class="operator">}</span>
 
 <span class="comment"># create a network from the list of edges</span>
 <span class="keyword">my</span> <span class="variable">$net4</span><span class="operator">=</span><span class="variable">Bio::PhyloNetwork</span><span class="operator">-&gt;</span><span class="variable">new</span><span class="operator">(</span><span class="string">-edges</span><span class="operator">=&gt;</span>
   <span class="operator">[</span><span class="string">qw(r s r t s u s c t c t v u b u l3 u b v b v l4 b l2 c l1)</span><span class="operator">]</span><span class="operator">);</span>
 
 <span class="comment"># Test for time consistency of $net3</span>
 <span class="keyword">if</span> <span class="operator">(</span><span class="variable">$net4</span><span class="operator">-&gt;</span><span class="variable">is_time_consistent</span><span class="operator">)</span> <span class="operator">{</span>
   <span class="keyword">print</span> <span class="string">"net4 is time consistent\n"</span>
 <span class="operator">}</span>
 <span class="keyword">else</span> <span class="operator">{</span>
   <span class="keyword">print</span> <span class="string">"net4 is not time consistent\n"</span>
 <span class="operator">}</span>
 
 <span class="comment"># And print all information on net4</span>
 <span class="keyword">print</span> <span class="variable">$net4</span><span class="operator">;</span>
 
 <span class="comment"># Compute some tripartitions</span>
 <span class="keyword">my</span> <span class="variable">%triparts</span><span class="operator">=</span><span class="variable">$net1</span><span class="operator">-&gt;</span><span class="variable">tripartitions</span><span class="operator">();</span>
 
 <span class="comment"># Now these are stored</span>
 <span class="keyword">print</span> <span class="variable">$net1</span><span class="operator">;</span>
 
 <span class="comment"># And can compute the tripartition error</span>
 <span class="keyword">print</span> <span class="string">"dtr="</span><span class="operator">.</span><span class="variable">$net1</span><span class="operator">-&gt;</span><span class="variable">tripartition_error</span><span class="operator">(</span><span class="variable">$net3</span><span class="operator">).</span><span class="string">"\n"</span><span class="operator">;</span>
</code></code></pre>

<h1 id="DESCRIPTION">DESCRIPTION</h1>

<h2 id="Phylogenetic-Networks">Phylogenetic Networks</h2>

<p>This is a module to work with phylogenetic networks. Phylogenetic networks have been studied over the last years as a richer model of the evolutionary history of sets of organisms than phylogenetic trees, because they take not only mutation events but also recombination and horizontal gene transfer events into account.</p>

<p>The natural model for describing the evolutionary history of a set of sequences under recombination events is a DAG, hence this package relies on the package Graph::Directed to represent the underlying graph of a phylogenetic network. We refer the reader to [CRV1,CRV2] for formal definitions related to phylogenetic networks.</p>

<h2 id="eNewick-description">eNewick description</h2>

<p>With this package, phylogenetic networks can be given by its eNewick string. This description appeared in other packages related to phylogenetic networks (see [PhyloNet] and [NetGen]); in fact, these two packages use different descriptions. The Bio::PhyloNetwork package allows both of them, but uses the second one in its output.</p>

<p>The first approach [PhyloNet] goes as follows: For each hybrid node H, say with parents u_1,u_2,...,u_k and children v_1,v_2,...v_l: split H in k+1 different nodes; let each of the first k copies be a child of one of the u_1,...,u_k (one for each) and have no children (hence we will have k extra leaves); as for the last copy, let it have no parents and have v_1,...,v_l be its children. This way we get a forest; each of the trees will be rooted at either one root of the phylogenetic network or a hybrid node of it; the set of leaves (of the whole forest) will be the set of leaves of the original network together with the set of hybrid nodes (each of them repeated as many times as its in-degree). Then, the eNewick representation of the phylogenetic network will be the Newick representation of all the trees in the obtained forest, each of them with its root labeled.</p>

<p>The second approach [NetGen] goes as follows: For each hybrid node H, say with parents u_1,u_2,...,u_k and children v_1,v_2,...v_l: split H in k different nodes; let the first copy be a child of u_1 and have all v_1,v_2,...v_l as its children; let the other copies be child of u_2,...,u_k (one for each) and have no children. This way, we get a tree whose set of leaves is the set of leaves of the original network together with the set of hybrid nodes (possibly repeated). Then the Newick string of the obtained tree (note that some internal nodes will be labeled and some leaves will be repeated) is the eNewick string of the phylogenetic network.</p>

<p>For example, consider the network depicted below:</p>

<pre><code><code>       r
      / \
     /   \
    U     V
   / \   / \
  1   \ /   3
       H
       |
       2</code></code></pre>

<p>If the first approach is taken, we get the forest:</p>

<pre><code><code>       r
      / \
     /   \
    U     V
   / \   / \
  1   H H   3
       |
       H
       |
       2</code></code></pre>

<p>Hence, the eNewick string is &#39;((1,H),(H,3))r; (2)H;&#39;.</p>

<p>As for the second one, one gets the tree:</p>

<pre><code><code>       r
      / \
     /   \
    U     V
   / \   / \
  1   H |   3
        H
        |
        2</code></code></pre>

<p>Hence, the eNewick string is &#39;((1,H),((2)H,3))r;&#39;.</p>

<p>Note: when rooting a tree, this package allows the notations &#39;(subtree,subtree,...)root&#39; as well as &#39;root:(subtree,subtree,...)&#39;, but the first one is used when writing eNewick strings.</p>

<h2 id="Tree-child-phylogenetic-networks">Tree-child phylogenetic networks</h2>

<p>Tree-child (TC) phylogenetic networks are a special class of phylogenetic networks for which a distance, called mu-distance, is defined [CRV2] based on certain data (mu-data) associated to every node. Moreover, this distance extends the Robinson-Foulds on phylogenetic trees. This package allows testing for a phylogenetic network if it is TC and computes mu-distances between networks over the same set of leaves.</p>

<p>Moreover, the mu-data allows one to define the optimal (in some precise sense) alignment between networks over the same set of leaves. This package also computes this optimal alignment.</p>

<h2 id="Tripartitions">Tripartitions</h2>

<p>Although tripartitions (see [CRV1] and the references therein) do not allow to define distances, this package outputs tripartitions and computes a weak form of the tripartition error.</p>

<h2 id="Time-consistency">Time-consistency</h2>

<p>Another useful property of Phylogenetic Networks that appears in the literature is that of time-consistency or real-time hybrids [BSS]. Roughly speaking, a network admits a temporal representation if it can be drawn in such a way that tree arcs (those whose end is a tree node) are inclined downwards, while hybridization arcs (those whose end is a hybrid node) are horizontal. This package checks for time-consistency and, if so, a temporal representation is provided.</p>

<h1 id="AUTHOR">AUTHOR</h1>

<pre><code><code> Gabriel Cardona, gabriel(dot)cardona(at)uib(dot)es
 Gabriel Valiente, valiente(at)lsi(dot)upc(dot)edu</code></code></pre>

<h1 id="SEE-ALSO">SEE ALSO</h1>

<dl>

<dt id="CRV1-">[CRV1]</dt>
<dd>

<p>G. Cardona, F. Rossello, G. Valiente. Tripartitions do not always discriminate phylogenetic networks. arXiv:0707.2376v1 [q-bio.PE]</p>

</dd>
<dt id="CRV2-">[CRV2]</dt>
<dd>

<p>G. Cardona, F. Rossello, G. Valiente. A Distance Measure for Tree-Child Phylogenetic Networks. Preprint.</p>

</dd>
<dt id="NetGen-">[NetGen]</dt>
<dd>

<p>M.M. Morin, and B.M.E. Moret. NetGen: generating phylogenetic networks with diploid hybrids. Bioinformatics 22 (2006), 1921-1923</p>

</dd>
<dt id="PhyloNet-">[PhyloNet]</dt>
<dd>

<p>PhyloNet: &quot;Phylogenetic Networks Toolkit&quot;. http://bioinfo.cs.rice.edu/phylonet</p>

</dd>
<dt id="BSS-">[BSS]</dt>
<dd>

<p>M. Baroni, C. Semple, and M. Steel. Hybrids in Real Time. Syst. Biol. 55(1):46-56, 2006</p>

</dd>
</dl>

<h1 id="APPENDIX">APPENDIX</h1>

<p>The rest of the documentation details each of the object methods.</p>

<h2 id="new">new</h2>

<pre><code><code> <span class="variable">Title</span>   <span class="operator">:</span> <span class="variable">new</span>
 <span class="variable">Usage</span>   <span class="operator">:</span> <span class="keyword">my</span> <span class="variable">$obj</span> <span class="operator">=</span> <span class="variable">new</span> <span class="variable">Bio::PhyloNetwork</span><span class="operator">();</span>
 <span class="variable">Function</span><span class="operator">:</span> <span class="variable">Creates</span> <span class="variable">a</span> <span class="variable">new</span> <span class="variable">Bio::PhyloNetwork</span> <span class="variable">object</span>
 <span class="variable">Returns</span> <span class="operator">:</span> <span class="variable">Bio::PhyloNetwork</span>
 <span class="variable">Args</span>    <span class="operator">:</span> <span class="variable">none</span>
            <span class="variable">OR</span>
           <span class="operator">-</span><span class="string">eNewick</span> <span class="operator">=&gt;</span> <span class="variable">string</span>
            <span class="variable">OR</span>
           <span class="operator">-</span><span class="string">graph</span> <span class="operator">=&gt;</span> <span class="variable">Graph::Directed</span> <span class="variable">object</span>
            <span class="variable">OR</span>
           <span class="operator">-</span><span class="string">edges</span> <span class="operator">=&gt;</span> <span class="variable">reference</span> <span class="variable">to</span> <span class="variable">an</span> <span class="variable">array</span>
            <span class="variable">OR</span>
           <span class="operator">-</span><span class="string">tree</span> <span class="operator">=&gt;</span> <span class="variable">Bio::Tree::Tree</span> <span class="variable">object</span>
            <span class="variable">OR</span>
           <span class="operator">-</span><span class="string">mudata</span> <span class="operator">=&gt;</span> <span class="variable">reference</span> <span class="variable">to</span> <span class="variable">a</span> <span class="variable">hash</span><span class="operator">,</span>
           <span class="string">-leaves</span> <span class="operator">=&gt;</span> <span class="variable">reference</span> <span class="variable">to</span> <span class="variable">an</span> <span class="variable">array</span>
            <span class="variable">OR</span>
           <span class="operator">-</span><span class="string">mudata</span> <span class="operator">=&gt;</span> <span class="variable">reference</span> <span class="variable">to</span> <span class="variable">a</span> <span class="variable">hash</span><span class="operator">,</span>
           <span class="string">-numleaves</span> <span class="operator">=&gt;</span> <span class="variable">integer</span>
</code></code></pre>

<p>Returns a Bio::PhyloNetwork object, created according to the data given:</p>

<dl>

<dt id="new-">new()</dt>
<dd>

<p>creates an empty network.</p>

</dd>
<dt id="new--eNewick-str-">new(-eNewick =&gt; $str)</dt>
<dd>

<p>creates the network whose Extended Newick representation (see description above) is the string $str.</p>

</dd>
<dt id="new--graph-graph-">new(-graph =&gt; $graph)</dt>
<dd>

<p>creates the network with underlying graph given by the Graph::Directed object $graph</p>

</dd>
<dt id="new--tree-tree-">new(-tree =&gt; $tree)</dt>
<dd>

<p>creates a network as a copy of the Bio::Tree::Tree object in $tree</p>

</dd>
<dt id="new--mudata-mudata--leaves-leaves-">new(-mudata =&gt; \%mudata, -leaves =&gt; \@leaves)</dt>
<dd>

<p>creates the network by reconstructing it from its mu-data stored in \%mudata and with set of leaves in \@leaves.</p>

</dd>
<dt id="new--mudata-mudata--numleaves-numleaves-">new(-mudata =&gt; \%mudata, -numleaves =&gt; $numleaves)</dt>
<dd>

<p>creates the network by reconstructing it from its mu-data stored in \%mudata and with set of leaves in (&quot;l1&quot;..&quot;l$numleaves&quot;).</p>

</dd>
</dl>

<h2 id="is_leaf">is_leaf</h2>

<pre><code><code> Title   : is_leaf
 Usage   : my $b=$net-&gt;is_leaf($u)
 Function: tests if $u is a leaf in $net
 Returns : boolean
 Args    : scalar</code></code></pre>

<h2 id="is_root">is_root</h2>

<pre><code><code> Title   : is_root
 Usage   : my $b=$net-&gt;is_root($u)
 Function: tests if $u is the root of $net
 Returns : boolean
 Args    : scalar</code></code></pre>

<h2 id="is_tree_node">is_tree_node</h2>

<pre><code><code> Title   : is_tree_node
 Usage   : my $b=$net-&gt;is_tree_node($u)
 Function: tests if $u is a tree node in $net
 Returns : boolean
 Args    : scalar</code></code></pre>

<h2 id="is_hybrid_node">is_hybrid_node</h2>

<pre><code><code> Title   : is_hybrid_node
 Usage   : my $b=$net-&gt;is_hybrid_node($u)
 Function: tests if $u is a hybrid node in $net
 Returns : boolean
 Args    : scalar</code></code></pre>

<h2 id="is_tree_child">is_tree_child</h2>

<pre><code><code> Title   : is_tree_child
 Usage   : my $b=$net-&gt;is_tree_child()
 Function: tests if $net is a Tree-Child phylogenetic network
 Returns : boolean
 Args    : Bio::PhyloNetwork</code></code></pre>

<h2 id="nodes">nodes</h2>

<pre><code><code> Title   : nodes
 Usage   : my @nodes=$net-&gt;nodes()
 Function: returns the set of nodes of $net
 Returns : array
 Args    : none</code></code></pre>

<h2 id="leaves">leaves</h2>

<pre><code><code> Title   : leaves
 Usage   : my @leaves=$net-&gt;leaves()
 Function: returns the set of leaves of $net
 Returns : array
 Args    : none</code></code></pre>

<h2 id="roots">roots</h2>

<pre><code><code> Title   : roots
 Usage   : my @roots=$net-&gt;roots()
 Function: returns the set of roots of $net
 Returns : array
 Args    : none</code></code></pre>

<h2 id="internal_nodes">internal_nodes</h2>

<pre><code><code> Title   : internal_nodes
 Usage   : my @internal_nodes=$net-&gt;internal_nodes()
 Function: returns the set of internal nodes of $net
 Returns : array
 Args    : none</code></code></pre>

<h2 id="tree_nodes">tree_nodes</h2>

<pre><code><code> Title   : tree_nodes
 Usage   : my @tree_nodes=$net-&gt;tree_nodes()
 Function: returns the set of tree nodes of $net
 Returns : array
 Args    : none</code></code></pre>

<h2 id="hybrid_nodes">hybrid_nodes</h2>

<pre><code><code> Title   : hybrid_nodes
 Usage   : my @hybrid_nodes=$net-&gt;hybrid_nodes()
 Function: returns the set of hybrid nodes of $net
 Returns : array
 Args    : none</code></code></pre>

<h2 id="graph">graph</h2>

<pre><code><code> Title   : graph
 Usage   : my $graph=$net-&gt;graph()
 Function: returns the underlying graph of $net
 Returns : Graph::Directed
 Args    : none</code></code></pre>

<h2 id="edges">edges</h2>

<pre><code><code> Title   : edges
 Usage   : my @edges=$net-&gt;edges()
 Function: returns the set of edges of $net
 Returns : array
 Args    : none</code></code></pre>

<p>Each element in the array is an anonimous array whose first element is the head of the edge and the second one is the tail.</p>

<h2 id="tree_edges">tree_edges</h2>

<pre><code><code> Title   : tree_edges
 Usage   : my @tree_edges=$net-&gt;tree_edges()
 Function: returns the set of tree edges of $net
           (those whose tail is a tree node)
 Returns : array
 Args    : none</code></code></pre>

<h2 id="hybrid_edges">hybrid_edges</h2>

<pre><code><code> Title   : hybrid_edges
 Usage   : my @hybrid_edges=$net-&gt;hybrid_edges()
 Function: returns the set of hybrid edges of $net
           (those whose tail is a hybrid node)
 Returns : array
 Args    : none</code></code></pre>

<h2 id="explode">explode</h2>

<pre><code><code> Title   : explode
 Usage   : my @trees=$net-&gt;explode()
 Function: returns the representation of $net by a set of
           Bio::Tree:Tree objects
 Returns : array
 Args    : none</code></code></pre>

<h2 id="mudata">mudata</h2>

<pre><code><code> Title   : mudata
 Usage   : my %mudata=$net-&gt;mudata()
 Function: returns the representation of $net by its mu-data
 Returns : hash
 Args    : none</code></code></pre>

<p>$net-&gt;mudata() returns a hash with keys the nodes of $net and each value is a muVector object holding its mu-vector.</p>

<h2 id="heights">heights</h2>

<pre><code><code> Title   : heights
 Usage   : my %heights=$net-&gt;heights()
 Function: returns the heights of the nodes of $net
 Returns : hash
 Args    : none</code></code></pre>

<p>$net-&gt;heights() returns a hash with keys the nodes of $net and each value is its height.</p>

<h2 id="mu_distance">mu_distance</h2>

<pre><code><code> Title   : mu_distance
 Usage   : my $dist=$net1-&gt;mu_distance($net2)
 Function: Computes the mu-distance between the networks $net1 and $net2 on
           the same set of leaves
 Returns : scalar
 Args    : Bio::PhyloNetwork</code></code></pre>

<h2 id="mu_distance_generalized">mu_distance_generalized</h2>

<pre><code><code> Title   : mu_distance_generalized
 Usage   : my $dist=$net1-&gt;mu_distance($net2)
 Function: Computes the mu-distance between the topological restrictions of
           networks $net1 and $net2 on its common set of leaves
 Returns : scalar
 Args    : Bio::PhyloNetwork</code></code></pre>

<h2 id="tripartitions">tripartitions</h2>

<pre><code><code> Title   : tripartitions
 Usage   : my %tripartitions=$net-&gt;tripartitions()
 Function: returns the set of tripartitions of $net
 Returns : hash
 Args    : none</code></code></pre>

<p>$net-&gt;tripartitions() returns a hash with keys the nodes of $net and each value is a string representing the tripartition of the leaves induced by the node. A string &quot;BCA...&quot; associated with a node u (e.g.) means, the first leaf is in the set B(u), the second one in C(u), the third one in A(u), and so on.</p>

<h2 id="is_time_consistent">is_time_consistent</h2>

<pre><code><code> Title   : is_time_consistent
 Usage   : my $b=$net-&gt;is_time_consistent()
 Function: tests if $net is (strong) time-consistent
 Returns : boolean
 Args    : none</code></code></pre>

<h2 id="temporal_representation">temporal_representation</h2>

<pre><code><code> Title   : temporal_representation
 Usage   : my %time=$net-&gt;temporal_representation()
 Function: returns a hash containing a temporal representation of $net, or 0
           if $net is not time-consistent
 Returns : hash
 Args    : none</code></code></pre>

<h2 id="contract_elementary">contract_elementary</h2>

<pre><code><code> <span class="variable">Title</span>   <span class="operator">:</span> <span class="variable">contract_elementary</span>
 <span class="variable">Usage</span>   <span class="operator">:</span> <span class="keyword">my</span> <span class="operator">(</span><span class="variable">$contracted</span><span class="operator">,</span><span class="variable">$blocks</span><span class="operator">)=</span><span class="variable">$net</span><span class="operator">-&gt;</span><span class="variable">contract_elementary</span><span class="operator">();</span>
 <span class="variable">Function</span><span class="operator">:</span> <span class="variable">Returns</span> <span class="variable">the</span> <span class="variable">network</span> <span class="variable">$contracted</span><span class="operator">,</span> <span class="variable">obtained</span> <span class="variable">by</span> <span class="variable">contracting</span> <span class="variable">elementary</span>
           <span class="variable">paths</span> <span class="variable">of</span> <span class="variable">$net</span> <span class="variable">into</span> <span class="variable">edges</span><span class="operator">.</span> <span class="variable">The</span> <span class="variable">reference</span> <span class="variable">$blocks</span> <span class="variable">points</span> <span class="variable">to</span> <span class="variable">a</span> <span class="variable">hash</span>
           <span class="variable">where</span><span class="operator">,</span> <span class="keyword">for</span> <span class="keyword">each</span> <span class="variable">node</span> <span class="variable">of</span> <span class="variable">$contracted</span><span class="operator">,</span> <span class="variable">gives</span> <span class="variable">the</span> <span class="variable">corresponding</span> <span class="variable">nodes</span>
           <span class="variable">of</span> <span class="variable">$net</span> <span class="variable">that</span> <span class="variable">have</span> <span class="variable">been</span> <span class="variable">deleted</span><span class="operator">.</span>
 <span class="variable">Returns</span> <span class="operator">:</span> <span class="variable">Bio::PhyloNetwork</span><span class="operator">,</span><span class="variable">reference</span> <span class="variable">to</span> <span class="variable">hash</span>
 <span class="variable">Args</span>    <span class="operator">:</span> <span class="variable">none</span>
</code></code></pre>

<h2 id="optimal_alignment">optimal_alignment</h2>

<pre><code><code> Title   : optimal_alignment
 Usage   : my ($weight,$alignment,$wgts)=$net-&gt;optimal_alignment($net2)
 Function: returns the total weight of an optimal alignment,
           the alignment itself, and partial weights
           between the networks $net1 and $net2 on the same set of leaves.
           An optional argument allows one to use the Manhattan (default) or the
           Hamming distance between mu-vectors.
 Returns : scalar,reference to hash,reference to hash
 Args    : Bio::PhyloNetwork,
           -metric =&gt; string (optional)</code></code></pre>

<p>Supported strings for the -metric parameter are &#39;Manhattan&#39; or &#39;Hamming&#39;.</p>

<h2 id="optimal_alignment_generalized">optimal_alignment_generalized</h2>

<pre><code><code> Title   : optimal_alignment_generalized
 Usage   : my ($weight,%alignment)=$net-&gt;optimal_alignment_generalized($net2)
 Function: returns the wieght of an optimal alignment, and the alignment itself,
           between the topological restriction of the networks $net1 and $net2
           on the set of common leaves.
           An optional argument allows one to use the Manhattan (default) or the
           Hamming distance between mu-vectors.
 Returns : scalar,hash
 Args    : Bio::PhyloNetwork,
           -metric =&gt; string (optional)</code></code></pre>

<p>Supported strings for the -metric parameter are &#39;Manhattan&#39; or &#39;Hamming&#39;.</p>

<h2 id="topological_restriction">topological_restriction</h2>

<pre><code><code> Title   : topological_restriction
 Usage   : my ($netr1,$netr2)=$net1-&gt;topological_restriction($net2)
 Function: returns the topological restriction of $net1 and $net2 on its
           common set of leaves
 Returns : Bio::PhyloNetwork, Bio::PhyloNetwork
 Args    : Bio::PhyloNetwork</code></code></pre>

<h2 id="eNewick">eNewick</h2>

<pre><code><code> Title   : eNewick
 Usage   : my $str=$net-&gt;eNewick()
 Function: returns the eNewick representation of $net without labeling
           internal tree nodes
 Returns : string
 Args    : none</code></code></pre>

<h2 id="eNewick_full">eNewick_full</h2>

<pre><code><code> Title   : eNewick_full
 Usage   : my $str=$net-&gt;eNewick_full()
 Function: returns the eNewick representation of $net labeling
           internal tree nodes
 Returns : string
 Args    : none</code></code></pre>

<h2 id="display">display</h2>

<pre><code><code> Title   : display
 Usage   : my $str=$net-&gt;display()
 Function: returns a string containing all the available information on $net
 Returns : string
 Args    : none</code></code></pre>


</body>

</html>


