<?xml version="1.0" ?>
<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.0 Transitional//EN">
<html>
<head>
<!-- saved from url=(0017)http://localhost/ -->
<script language="JavaScript" src="../../../displayToc.js"></script>
<script language="JavaScript" src="../../../tocParas.js"></script>
<script language="JavaScript" src="../../../tocTab.js"></script>
<link rel="stylesheet" type="text/css" href="../../../scineplex.css">
<title></title>
<link rel="stylesheet" href="../../../Active.css" type="text/css" />
<meta http-equiv="content-type" content="text/html; charset=utf-8" />
<link rev="made" href="mailto:support@ActiveState.com" />
</head>

<body>



<ul id="index">
  <li><a href="#NAME">NAME</a></li>
  <li><a href="#SYNOPSIS">SYNOPSIS</a></li>
  <li><a href="#DESCRIPTION">DESCRIPTION</a></li>
  <li><a href="#FEEDBACK">FEEDBACK</a>
    <ul>
      <li><a href="#Support">Support</a></li>
      <li><a href="#Reporting-Bugs">Reporting Bugs</a></li>
    </ul>
  </li>
  <li><a href="#APPENDIX">APPENDIX</a></li>
  <li><a href="#Bio::SeqFeatureI-specific-methods">Bio::SeqFeatureI specific methods</a>
    <ul>
      <li><a href="#get_SeqFeatures">get_SeqFeatures</a></li>
      <li><a href="#display_name">display_name</a></li>
      <li><a href="#primary_tag">primary_tag</a></li>
      <li><a href="#source_tag">source_tag</a></li>
      <li><a href="#has_tag">has_tag</a></li>
      <li><a href="#get_tag_values">get_tag_values</a></li>
      <li><a href="#get_tagset_values">get_tagset_values</a></li>
      <li><a href="#get_all_tags">get_all_tags</a></li>
      <li><a href="#attach_seq">attach_seq</a></li>
      <li><a href="#seq">seq</a></li>
      <li><a href="#entire_seq">entire_seq</a></li>
      <li><a href="#seq_id">seq_id</a></li>
      <li><a href="#gff_string">gff_string</a></li>
      <li><a href="#static_gff_formatter">_static_gff_formatter</a></li>
    </ul>
  </li>
  <li><a href="#Decorating-methods">Decorating methods</a>
    <ul>
      <li><a href="#spliced_seq">spliced_seq</a></li>
      <li><a href="#location">location</a></li>
      <li><a href="#primary_id">primary_id</a></li>
      <li><a href="#phase">phase</a></li>
    </ul>
  </li>
  <li><a href="#Bio::RangeI-methods">Bio::RangeI methods</a>
    <ul>
      <li><a href="#start-">start()</a></li>
      <li><a href="#end-">end()</a></li>
      <li><a href="#strand-">strand()</a></li>
      <li><a href="#overlaps-">overlaps()</a></li>
      <li><a href="#contains-">contains()</a></li>
      <li><a href="#equals-">equals()</a></li>
      <li><a href="#intersection-">intersection()</a></li>
      <li><a href="#union-">union()</a></li>
    </ul>
  </li>
</ul>

<h1 id="NAME">NAME</h1>

<p>Bio::SeqFeatureI - Abstract interface of a Sequence Feature</p>

<h1 id="SYNOPSIS">SYNOPSIS</h1>

<pre><code><code>    <span class="comment"># get a seqfeature somehow, eg, from a Sequence with Features attached</span>
    
    <span class="keyword">foreach</span> <span class="variable">$feat</span> <span class="operator">(</span> <span class="variable">$seq</span><span class="operator">-&gt;</span><span class="variable">get_SeqFeatures</span><span class="operator">()</span> <span class="operator">)</span> <span class="operator">{</span>
       <span class="keyword">print</span> <span class="string">"Feature from "</span><span class="operator">,</span> <span class="variable">$feat</span><span class="operator">-&gt;</span><span class="variable">start</span><span class="operator">,</span> <span class="string">"to "</span><span class="operator">,</span>
               <span class="variable">$feat</span><span class="operator">-&gt;</span><span class="variable">end</span><span class="operator">,</span> <span class="string">" Primary tag  "</span><span class="operator">,</span> <span class="variable">$feat</span><span class="operator">-&gt;</span><span class="variable">primary_tag</span><span class="operator">,</span>
                  <span class="string">", produced by "</span><span class="operator">,</span> <span class="variable">$feat</span><span class="operator">-&gt;</span><span class="variable">source_tag</span><span class="operator">(),</span> <span class="string">"\n"</span><span class="operator">;</span>
    
       <span class="keyword">if</span><span class="operator">(</span> <span class="variable">$feat</span><span class="operator">-&gt;</span><span class="variable">strand</span> <span class="operator">==</span> <span class="number">0</span> <span class="operator">)</span> <span class="operator">{</span>
                    <span class="keyword">print</span> <span class="string">"Feature applicable to either strand\n"</span><span class="operator">;</span>
       <span class="operator">}</span> <span class="keyword">else</span> <span class="operator">{</span>
          <span class="keyword">print</span> <span class="string">"Feature on strand "</span><span class="operator">,</span> <span class="variable">$feat</span><span class="operator">-&gt;</span><span class="variable">strand</span><span class="operator">,</span><span class="string">"\n"</span><span class="operator">;</span> <span class="comment"># -1,1</span>
       <span class="operator">}</span>
    
       <span class="keyword">print</span> <span class="string">"feature location is "</span><span class="operator">,</span><span class="variable">$feat</span><span class="operator">-&gt;</span><span class="variable">start</span><span class="operator">,</span> <span class="string">".."</span><span class="operator">,</span>
          <span class="variable">$feat</span><span class="operator">-&gt;</span><span class="variable">end</span><span class="operator">,</span> <span class="string">" on strand "</span><span class="operator">,</span> <span class="variable">$feat</span><span class="operator">-&gt;</span><span class="variable">strand</span><span class="operator">,</span> <span class="string">"\n"</span><span class="operator">;</span>
       <span class="keyword">print</span> <span class="string">"easy utility to print locations in GenBank/EMBL way "</span><span class="operator">,</span>
          <span class="variable">$feat</span><span class="operator">-&gt;</span><span class="variable">location</span><span class="operator">-&gt;</span><span class="variable">to_FTstring</span><span class="operator">(),</span> <span class="string">"\n"</span><span class="operator">;</span>
    
       <span class="keyword">foreach</span> <span class="variable">$tag</span> <span class="operator">(</span> <span class="variable">$feat</span><span class="operator">-&gt;</span><span class="variable">get_all_tags</span><span class="operator">()</span> <span class="operator">)</span> <span class="operator">{</span>
                    <span class="keyword">print</span> <span class="string">"Feature has tag "</span><span class="operator">,</span> <span class="variable">$tag</span><span class="operator">,</span> <span class="string">" with values, "</span><span class="operator">,</span>
                      <span class="keyword">join</span><span class="operator">(</span><span class="string">' '</span><span class="operator">,</span><span class="variable">$feat</span><span class="operator">-&gt;</span><span class="variable">get_tag_values</span><span class="operator">(</span><span class="variable">$tag</span><span class="operator">)),</span> <span class="string">"\n"</span><span class="operator">;</span>
       <span class="operator">}</span>
            <span class="keyword">print</span> <span class="string">"new feature\n"</span> <span class="keyword">if</span> <span class="variable">$feat</span><span class="operator">-&gt;</span><span class="variable">has_tag</span><span class="operator">(</span><span class="string">'new'</span><span class="operator">);</span>
            <span class="comment"># features can have sub features</span>
            <span class="keyword">my</span> <span class="variable">@subfeat</span> <span class="operator">=</span> <span class="variable">$feat</span><span class="operator">-&gt;</span><span class="variable">get_SeqFeatures</span><span class="operator">();</span>
         <span class="operator">}</span>
</code></code></pre>

<h1 id="DESCRIPTION">DESCRIPTION</h1>

<p>This interface is the functions one can expect for any Sequence Feature, whatever its implementation or whether it is a more complex type (eg, a Gene). This object does not actually provide any implementation, it just provides the definitions of what methods one can call. See Bio::SeqFeature::Generic for a good standard implementation of this object</p>

<h1 id="FEEDBACK">FEEDBACK</h1>

<p>User feedback is an integral part of the evolution of this and other Bioperl modules. Send your comments and suggestions preferably to one of the Bioperl mailing lists. Your participation is much appreciated.</p>

<pre><code><code>  bioperl-l@bioperl.org                  - General discussion
  http://bioperl.org/wiki/Mailing_lists  - About the mailing lists</code></code></pre>

<h2 id="Support">Support</h2>

<p>Please direct usage questions or support issues to the mailing list:</p>

<p><i>bioperl-l@bioperl.org</i></p>

<p>rather than to the module maintainer directly. Many experienced and reponsive experts will be able look at the problem and quickly address it. Please include a thorough description of the problem with code and data examples if at all possible.</p>

<h2 id="Reporting-Bugs">Reporting Bugs</h2>

<p>Report bugs to the Bioperl bug tracking system to help us keep track the bugs and their resolution. Bug reports can be submitted via the web:</p>

<pre><code><code>  https://redmine.open-bio.org/projects/bioperl/</code></code></pre>

<h1 id="APPENDIX">APPENDIX</h1>

<p>The rest of the documentation details each of the object methods. Internal methods are usually preceded with a _</p>

<h1 id="Bio::SeqFeatureI-specific-methods">Bio::SeqFeatureI specific methods</h1>

<p>New method interfaces.</p>

<h2 id="get_SeqFeatures">get_SeqFeatures</h2>

<pre><code><code> <span class="variable">Title</span>   <span class="operator">:</span> <span class="variable">get_SeqFeatures</span>
 <span class="variable">Usage</span>   <span class="operator">:</span> <span class="variable">@feats</span> <span class="operator">=</span> <span class="variable">$feat</span><span class="operator">-&gt;</span><span class="variable">get_SeqFeatures</span><span class="operator">();</span>
 <span class="variable">Function</span><span class="operator">:</span> <span class="variable">Returns</span> <span class="variable">an</span> <span class="variable">array</span> <span class="variable">of</span> <span class="keyword">sub</span><span class="variable"> Sequence Features
 Returns : An array
 Args    : none
 </span>
</code></code></pre>

<h2 id="display_name">display_name</h2>

<pre><code><code> Title   : display_name
 Usage   : $name = $feat-&gt;display_name()
 Function: Returns the human-readable name of the feature for displays.
 Returns : a string
 Args    : none</code></code></pre>

<h2 id="primary_tag">primary_tag</h2>

<pre><code><code> Title   : primary_tag
 Usage   : $tag = $feat-&gt;primary_tag()
 Function: Returns the primary tag for a feature,
           eg &#39;exon&#39;
 Returns : a string
 Args    : none</code></code></pre>

<h2 id="source_tag">source_tag</h2>

<pre><code><code> Title   : source_tag
 Usage   : $tag = $feat-&gt;source_tag()
 Function: Returns the source tag for a feature,
           eg, &#39;genscan&#39;
 Returns : a string
 Args    : none</code></code></pre>

<h2 id="has_tag">has_tag</h2>

<pre><code><code> Title   : has_tag
 Usage   : $tag_exists = $self-&gt;has_tag(&#39;some_tag&#39;)
 Function: 
 Returns : TRUE if the specified tag exists, and FALSE otherwise
 Args    :</code></code></pre>

<h2 id="get_tag_values">get_tag_values</h2>

<pre><code><code> Title   : get_tag_values
 Usage   : @values = $self-&gt;get_tag_values(&#39;some_tag&#39;)
 Function: 
 Returns : An array comprising the values of the specified tag.
 Args    : a string</code></code></pre>

<p>throws an exception if there is no such tag</p>

<h2 id="get_tagset_values">get_tagset_values</h2>

<pre><code><code> Title   : get_tagset_values
 Usage   : @values = $self-&gt;get_tagset_values(qw(label transcript_id product))
 Function: 
 Returns : An array comprising the values of the specified tags, in order of tags
 Args    : An array of strings</code></code></pre>

<p>does NOT throw an exception if none of the tags are not present</p>

<p>this method is useful for getting a human-readable label for a SeqFeatureI; not all tags can be assumed to be present, so a list of possible tags in preferential order is provided</p>

<h2 id="get_all_tags">get_all_tags</h2>

<pre><code><code> Title   : get_all_tags
 Usage   : @tags = $feat-&gt;get_all_tags()
 Function: gives all tags for this feature
 Returns : an array of strings
 Args    : none</code></code></pre>

<h2 id="attach_seq">attach_seq</h2>

<pre><code><code> Title   : attach_seq
 Usage   : $sf-&gt;attach_seq($seq)
 Function: Attaches a Bio::Seq object to this feature. This
           Bio::Seq object is for the *entire* sequence: ie
           from 1 to 10000

           Note that it is not guaranteed that if you obtain a feature from
           an object in bioperl, it will have a sequence attached. Also,
           implementors of this interface can choose to provide an empty
           implementation of this method. I.e., there is also no guarantee
           that if you do attach a sequence, seq() or entire_seq() will not
           return undef.

           The reason that this method is here on the interface is to enable
           you to call it on every SeqFeatureI compliant object, and
           that it will be implemented in a useful way and set to a useful
           value for the great majority of use cases. Implementors who choose
           to ignore the call are encouraged to specifically state this in
           their documentation.

 Example :
 Returns : TRUE on success
 Args    : a Bio::PrimarySeqI compliant object</code></code></pre>

<h2 id="seq">seq</h2>

<pre><code><code> Title   : seq
 Usage   : $tseq = $sf-&gt;seq()
 Function: returns the truncated sequence (if there is a sequence attached)
           for this feature
 Example :
 Returns : sub seq (a Bio::PrimarySeqI compliant object) on attached sequence
           bounded by start &amp; end, or undef if there is no sequence attached
 Args    : none</code></code></pre>

<h2 id="entire_seq">entire_seq</h2>

<pre><code><code> Title   : entire_seq
 Usage   : $whole_seq = $sf-&gt;entire_seq()
 Function: gives the entire sequence that this seqfeature is attached to
 Example :
 Returns : a Bio::PrimarySeqI compliant object, or undef if there is no
           sequence attached
 Args    : none</code></code></pre>

<h2 id="seq_id">seq_id</h2>

<pre><code><code> Title   : seq_id
 Usage   : $obj-&gt;seq_id($newval)
 Function: There are many cases when you make a feature that you
           do know the sequence name, but do not know its actual
           sequence. This is an attribute such that you can store
           the ID (e.g., display_id) of the sequence.

           This attribute should *not* be used in GFF dumping, as
           that should come from the collection in which the seq
           feature was found.
 Returns : value of seq_id
 Args    : newvalue (optional)</code></code></pre>

<h2 id="gff_string">gff_string</h2>

<pre><code><code> <span class="variable">Title</span>   <span class="operator">:</span> <span class="variable">gff_string</span>
 <span class="variable">Usage</span>   <span class="operator">:</span> <span class="variable">$str</span> <span class="operator">=</span> <span class="variable">$feat</span><span class="operator">-&gt;</span><span class="variable">gff_string</span><span class="operator">;</span>
           <span class="variable">$str</span> <span class="operator">=</span> <span class="variable">$feat</span><span class="operator">-&gt;</span><span class="variable">gff_string</span><span class="operator">(</span><span class="variable">$gff_formatter</span><span class="operator">);</span>
 <span class="variable">Function</span><span class="operator">:</span> <span class="variable">Provides</span> <span class="variable">the</span> <span class="variable">feature</span> <span class="variable">information</span> <span class="variable">in</span> <span class="variable">GFF</span> <span class="keyword">format</span><span class="operator">.</span>
 
           <span class="variable">The</span> <span class="variable">implementation</span> <span class="variable">provided</span> <span class="variable">here</span> <span class="variable">returns</span> <span class="variable">GFF2</span> <span class="variable">by</span> <span class="keyword">default</span><span class="operator">.</span> <span class="variable">If</span> <span class="variable">you</span>
           <span class="variable">want</span> <span class="variable">a</span> <span class="variable">different</span> <span class="variable">version</span><span class="operator">,</span> <span class="variable">supply</span> <span class="variable">an</span> <span class="variable">object</span> <span class="variable">implementing</span> <span class="variable">a</span> <span class="variable">method</span>
           <span class="variable">gff_string</span><span class="operator">()</span> <span class="variable">accepting</span> <span class="variable">a</span> <span class="variable">SeqFeatureI</span> <span class="variable">object</span> <span class="variable">as</span> <span class="variable">argument</span><span class="operator">.</span> <span class="variable">E</span><span class="operator">.</span><span class="variable">g</span><span class="operator">.,</span> <span class="variable">to</span>
           <span class="variable">obtain</span> <span class="variable">GFF1</span> <span class="keyword">format</span><span class="operator">,</span> <span class="keyword">do</span> <span class="variable">the</span> <span class="variable">following</span><span class="operator">:</span>
 
                <span class="keyword">my</span> <span class="variable">$gffio</span> <span class="operator">=</span> <span class="variable">Bio::Tools::GFF</span><span class="operator">-&gt;</span><span class="variable">new</span><span class="operator">(</span><span class="string">-gff_version</span> <span class="operator">=&gt;</span> <span class="number">1</span><span class="operator">);</span>
                <span class="variable">$gff1str</span> <span class="operator">=</span> <span class="variable">$feat</span><span class="operator">-&gt;</span><span class="variable">gff_string</span><span class="operator">(</span><span class="variable">$gff1io</span><span class="operator">);</span>
 
 <span class="variable">Returns</span> <span class="operator">:</span> <span class="variable">A</span> <span class="variable">string</span>
 <span class="variable">Args</span>    <span class="operator">:</span> <span class="variable">Optionally</span><span class="operator">,</span> <span class="variable">an</span> <span class="variable">object</span> <span class="variable">implementing</span> <span class="variable">gff_string</span><span class="operator">().</span>
</code></code></pre>

<h2 id="static_gff_formatter">_static_gff_formatter</h2>

<pre><code><code> Title   : _static_gff_formatter
 Usage   :
 Function:
 Example :
 Returns :
 Args    :</code></code></pre>

<h1 id="Decorating-methods">Decorating methods</h1>

<p>These methods have an implementation provided by Bio::SeqFeatureI, but can be validly overwritten by subclasses</p>

<h2 id="spliced_seq">spliced_seq</h2>

<pre><code><code>  Title   : spliced_seq

  Usage   : $seq = $feature-&gt;spliced_seq()
            $seq = $feature_with_remote_locations-&gt;spliced_seq($db_for_seqs)

  Function: Provides a sequence of the feature which is the most
            semantically &quot;relevant&quot; feature for this sequence. A default
            implementation is provided which for simple cases returns just
            the sequence, but for split cases, loops over the split location
            to return the sequence. In the case of split locations with
            remote locations, eg

            join(AB000123:5567-5589,80..1144)

            in the case when a database object is passed in, it will attempt
            to retrieve the sequence from the database object, and &quot;Do the right thing&quot;,
            however if no database object is provided, it will generate the correct
            number of N&#39;s (DNA) or X&#39;s (protein, though this is unlikely).

            This function is deliberately &quot;magical&quot; attempting to second guess
            what a user wants as &quot;the&quot; sequence for this feature.

            Implementing classes are free to override this method with their
            own magic if they have a better idea what the user wants.

  Args    : [optional]
            -db        A L&lt;Bio::DB::RandomAccessI&gt; compliant object if
                       one needs to retrieve remote seqs.
            -nosort    boolean if the locations should not be sorted
                       by start location.  This may occur, for instance,
                       in a circular sequence where a gene span starts
                       before the end of the sequence and ends after the
                       sequence start. Example : join(15685..16260,1..207)
                                           (default = if sequence is_circular(), 1, otherwise 0)
                        -phase     truncates the returned sequence based on the
                                           intron phase (0,1,2).

  Returns : A L&lt;Bio::PrimarySeqI&gt; object</code></code></pre>

<h2 id="location">location</h2>

<pre><code><code> Title   : location
 Usage   : my $location = $seqfeature-&gt;location()
 Function: returns a location object suitable for identifying location
           of feature on sequence or parent feature
 Returns : Bio::LocationI object
 Args    : none</code></code></pre>

<h2 id="primary_id">primary_id</h2>

<pre><code><code> Title   : primary_id
 Usage   : $obj-&gt;primary_id($newval)
 Function:
 Example :
 Returns : value of primary_id (a scalar)
 Args    : on set, new value (a scalar or undef, optional)</code></code></pre>

<p>Primary ID is a synonym for the tag &#39;ID&#39;</p>

<h2 id="phase">phase</h2>

<pre><code><code> Title   : phase
 Usage   : $obj-&gt;phase($newval)
 Function: get/set this feature&#39;s phase.
 Example :
 Returns : undef if no phase is set,
           otherwise 0, 1, or 2 (the only valid values for phase)
 Args    : on set, the new value</code></code></pre>

<p>Most features do not have or need a defined phase.</p>

<p>For features representing a CDS, the phase indicates where the feature begins with reference to the reading frame. The phase is one of the integers 0, 1, or 2, indicating the number of bases that should be removed from the beginning of this feature to reach the first base of the next codon. In other words, a phase of &quot;0&quot; indicates that the next codon begins at the first base of the region described by the current line, a phase of &quot;1&quot; indicates that the next codon begins at the second base of this region, and a phase of &quot;2&quot; indicates that the codon begins at the third base of this region. This is NOT to be confused with the frame, which is simply start modulo 3.</p>

<p>For forward strand features, phase is counted from the start field. For reverse strand features, phase is counted from the end field.</p>

<h1 id="Bio::RangeI-methods">Bio::RangeI methods</h1>

<p>These methods are inherited from RangeI and can be used directly from a SeqFeatureI interface. Remember that a SeqFeature is-a RangeI, and so wherever you see RangeI you can use a feature ($r in the below documentation).</p>

<h2 id="start-">start()</h2>

<pre><code><code> See L&lt;Bio::RangeI&gt;</code></code></pre>

<h2 id="end-">end()</h2>

<pre><code><code> See L&lt;Bio::RangeI&gt;</code></code></pre>

<h2 id="strand-">strand()</h2>

<pre><code><code> See L&lt;Bio::RangeI&gt;</code></code></pre>

<h2 id="overlaps-">overlaps()</h2>

<pre><code><code> See L&lt;Bio::RangeI&gt;</code></code></pre>

<h2 id="contains-">contains()</h2>

<pre><code><code> See L&lt;Bio::RangeI&gt;</code></code></pre>

<h2 id="equals-">equals()</h2>

<pre><code><code> See L&lt;Bio::RangeI&gt;</code></code></pre>

<h2 id="intersection-">intersection()</h2>

<pre><code><code> See L&lt;Bio::RangeI&gt;</code></code></pre>

<h2 id="union-">union()</h2>

<pre><code><code> See L&lt;Bio::RangeI&gt;</code></code></pre>


</body>

</html>


