<?xml version="1.0" ?>
<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.0 Transitional//EN">
<html>
<head>
<!-- saved from url=(0017)http://localhost/ -->
<script language="JavaScript" src="../../../displayToc.js"></script>
<script language="JavaScript" src="../../../tocParas.js"></script>
<script language="JavaScript" src="../../../tocTab.js"></script>
<link rel="stylesheet" type="text/css" href="../../../scineplex.css">
<title></title>
<link rel="stylesheet" href="../../../Active.css" type="text/css" />
<meta http-equiv="content-type" content="text/html; charset=utf-8" />
<link rev="made" href="mailto:support@ActiveState.com" />
</head>

<body>



<ul id="index">
  <li><a href="#NAME">NAME</a></li>
  <li><a href="#SYNOPSIS">SYNOPSIS</a></li>
  <li><a href="#DESCRIPTION">DESCRIPTION</a></li>
  <li><a href="#FEEDBACK">FEEDBACK</a>
    <ul>
      <li><a href="#Mailing-Lists">Mailing Lists</a></li>
      <li><a href="#Support">Support</a></li>
      <li><a href="#Reporting-Bugs">Reporting Bugs</a></li>
    </ul>
  </li>
  <li><a href="#AUTHOR">AUTHOR</a></li>
  <li><a href="#CONTRIBUTORS">CONTRIBUTORS</a></li>
  <li><a href="#SEE-ALSO">SEE ALSO</a></li>
  <li><a href="#APPENDIX">APPENDIX</a>
    <ul>
      <li><a href="#new">new</a></li>
    </ul>
  </li>
  <li><a href="#Modifier-methods">Modifier methods</a>
    <ul>
      <li><a href="#add_seq">add_seq</a></li>
      <li><a href="#remove_seq">remove_seq</a></li>
      <li><a href="#purge">purge</a></li>
      <li><a href="#sort_alphabetically">sort_alphabetically</a></li>
      <li><a href="#sort_by_list">sort_by_list</a></li>
      <li><a href="#set_new_reference">set_new_reference</a></li>
      <li><a href="#uniq_seq">uniq_seq</a></li>
    </ul>
  </li>
  <li><a href="#Sequence-selection-methods">Sequence selection methods</a>
    <ul>
      <li><a href="#each_seq">each_seq</a></li>
      <li><a href="#each_alphabetically">each_alphabetically</a></li>
      <li><a href="#each_seq_with_id">each_seq_with_id</a></li>
      <li><a href="#get_seq_by_pos">get_seq_by_pos</a></li>
      <li><a href="#get_seq_by_id">get_seq_by_id</a></li>
      <li><a href="#seq_with_features">seq_with_features</a></li>
    </ul>
  </li>
  <li><a href="#Create-new-alignments">Create new alignments</a>
    <ul>
      <li><a href="#select">select</a></li>
      <li><a href="#select_noncont">select_noncont</a></li>
      <li><a href="#slice">slice</a></li>
      <li><a href="#remove_columns">remove_columns</a></li>
      <li><a href="#remove_gaps">remove_gaps</a></li>
    </ul>
  </li>
  <li><a href="#Change-sequences-within-the-MSA">Change sequences within the MSA</a>
    <ul>
      <li><a href="#splice_by_seq_pos">splice_by_seq_pos</a></li>
      <li><a href="#map_chars">map_chars</a></li>
      <li><a href="#uppercase">uppercase</a></li>
      <li><a href="#cigar_line">cigar_line</a></li>
      <li><a href="#match_line">match_line</a></li>
      <li><a href="#gap_line">gap_line</a></li>
      <li><a href="#all_gap_line">all_gap_line</a></li>
      <li><a href="#gap_col_matrix">gap_col_matrix</a></li>
      <li><a href="#match">match</a></li>
      <li><a href="#unmatch">unmatch</a></li>
    </ul>
  </li>
  <li><a href="#MSA-attributes">MSA attributes</a>
    <ul>
      <li><a href="#id">id</a></li>
      <li><a href="#accession">accession</a></li>
      <li><a href="#description">description</a></li>
      <li><a href="#missing_char">missing_char</a></li>
      <li><a href="#match_char">match_char</a></li>
      <li><a href="#gap_char">gap_char</a></li>
      <li><a href="#symbol_chars">symbol_chars</a></li>
    </ul>
  </li>
  <li><a href="#Alignment-descriptors">Alignment descriptors</a>
    <ul>
      <li><a href="#score">score</a></li>
      <li><a href="#consensus_string">consensus_string</a></li>
      <li><a href="#consensus_iupac">consensus_iupac</a></li>
      <li><a href="#consensus_meta">consensus_meta</a></li>
      <li><a href="#is_flush">is_flush</a></li>
      <li><a href="#length">length</a></li>
      <li><a href="#maxdisplayname_length">maxdisplayname_length</a></li>
      <li><a href="#max_metaname_length">max_metaname_length</a></li>
      <li><a href="#num_residues">num_residues</a></li>
      <li><a href="#num_sequences">num_sequences</a></li>
      <li><a href="#average_percentage_identity">average_percentage_identity</a></li>
      <li><a href="#percentage_identity">percentage_identity</a></li>
      <li><a href="#overall_percentage_identity">overall_percentage_identity</a></li>
    </ul>
  </li>
  <li><a href="#Alignment-positions">Alignment positions</a>
    <ul>
      <li><a href="#column_from_residue_number">column_from_residue_number</a></li>
    </ul>
  </li>
  <li><a href="#Sequence-names">Sequence names</a>
    <ul>
      <li><a href="#displayname">displayname</a></li>
      <li><a href="#set_displayname_count">set_displayname_count</a></li>
      <li><a href="#set_displayname_flat">set_displayname_flat</a></li>
      <li><a href="#set_displayname_normal">set_displayname_normal</a></li>
      <li><a href="#source">source</a></li>
      <li><a href="#set_displayname_safe">set_displayname_safe</a></li>
      <li><a href="#restore_displayname">restore_displayname</a></li>
      <li><a href="#sort_by_start">sort_by_start</a></li>
      <li><a href="#bracket_string">bracket_string</a></li>
      <li><a href="#methods-implementing-Bio::FeatureHolderI">methods implementing Bio::FeatureHolderI</a></li>
      <li><a href="#get_SeqFeatures">get_SeqFeatures</a></li>
      <li><a href="#add_SeqFeature">add_SeqFeature</a></li>
      <li><a href="#remove_SeqFeatures">remove_SeqFeatures</a></li>
      <li><a href="#feature_count">feature_count</a></li>
      <li><a href="#get_all_SeqFeatures">get_all_SeqFeatures</a></li>
      <li><a href="#methods-for-Bio::AnnotatableI">methods for Bio::AnnotatableI</a></li>
      <li><a href="#annotation">annotation</a></li>
    </ul>
  </li>
  <li><a href="#Deprecated-methods">Deprecated methods</a>
    <ul>
      <li><a href="#no_residues">no_residues</a></li>
      <li><a href="#no_sequences">no_sequences</a></li>
      <li><a href="#mask_columns">mask_columns</a></li>
    </ul>
  </li>
</ul>

<h1 id="NAME">NAME</h1>

<p>Bio::SimpleAlign - Multiple alignments held as a set of sequences</p>

<h1 id="SYNOPSIS">SYNOPSIS</h1>

<pre><code><code>  <span class="comment"># Use Bio::AlignIO to read in the alignment</span>
  <span class="variable">$str</span> <span class="operator">=</span> <span class="variable">Bio::AlignIO</span><span class="operator">-&gt;</span><span class="variable">new</span><span class="operator">(</span><span class="string">-file</span> <span class="operator">=&gt;</span> <span class="string">'t/data/testaln.pfam'</span><span class="operator">);</span>
  <span class="variable">$aln</span> <span class="operator">=</span> <span class="variable">$str</span><span class="operator">-&gt;</span><span class="variable">next_aln</span><span class="operator">();</span>
  
  <span class="comment"># Describe</span>
  <span class="keyword">print</span> <span class="variable">$aln</span><span class="operator">-&gt;</span><span class="variable">length</span><span class="operator">;</span>
  <span class="keyword">print</span> <span class="variable">$aln</span><span class="operator">-&gt;</span><span class="variable">num_residues</span><span class="operator">;</span>
  <span class="keyword">print</span> <span class="variable">$aln</span><span class="operator">-&gt;</span><span class="variable">is_flush</span><span class="operator">;</span>
  <span class="keyword">print</span> <span class="variable">$aln</span><span class="operator">-&gt;</span><span class="variable">num_sequences</span><span class="operator">;</span>
  <span class="keyword">print</span> <span class="variable">$aln</span><span class="operator">-&gt;</span><span class="variable">score</span><span class="operator">;</span>
  <span class="keyword">print</span> <span class="variable">$aln</span><span class="operator">-&gt;</span><span class="variable">percentage_identity</span><span class="operator">;</span>
  <span class="keyword">print</span> <span class="variable">$aln</span><span class="operator">-&gt;</span><span class="variable">consensus_string</span><span class="operator">(</span><span class="number">50</span><span class="operator">);</span>
  
  <span class="comment"># Find the position in the alignment for a sequence location</span>
  <span class="variable">$pos</span> <span class="operator">=</span> <span class="variable">$aln</span><span class="operator">-&gt;</span><span class="variable">column_from_residue_number</span><span class="operator">(</span><span class="string">'1433_LYCES'</span><span class="operator">,</span> <span class="number">14</span><span class="operator">);</span> <span class="comment"># = 6;</span>
  
  <span class="comment"># Extract sequences and check values for the alignment column $pos</span>
  <span class="keyword">foreach</span> <span class="variable">$seq</span> <span class="operator">(</span><span class="variable">$aln</span><span class="operator">-&gt;</span><span class="variable">each_seq</span><span class="operator">)</span> <span class="operator">{</span>
      <span class="variable">$res</span> <span class="operator">=</span> <span class="variable">$seq</span><span class="operator">-&gt;</span><span class="variable">subseq</span><span class="operator">(</span><span class="variable">$pos</span><span class="operator">,</span> <span class="variable">$pos</span><span class="operator">);</span>
      <span class="variable">$count</span><span class="operator">{</span><span class="variable">$res</span><span class="operator">}</span><span class="operator">++;</span>
  <span class="operator">}</span>
  <span class="keyword">foreach</span> <span class="variable">$res</span> <span class="operator">(</span><span class="keyword">keys</span> <span class="variable">%count</span><span class="operator">)</span> <span class="operator">{</span>
      <span class="keyword">printf</span> <span class="string">"Res: %s  Count: %2d\n"</span><span class="operator">,</span> <span class="variable">$res</span><span class="operator">,</span> <span class="variable">$count</span><span class="operator">{</span><span class="variable">$res</span><span class="operator">}</span><span class="operator">;</span>
  <span class="operator">}</span>
  
  <span class="comment"># Manipulate</span>
  <span class="variable">$aln</span><span class="operator">-&gt;</span><span class="variable">remove_seq</span><span class="operator">(</span><span class="variable">$seq</span><span class="operator">);</span>
  <span class="variable">$mini_aln</span> <span class="operator">=</span> <span class="variable">$aln</span><span class="operator">-&gt;</span><span class="variable">slice</span><span class="operator">(</span><span class="number">20</span><span class="operator">,</span><span class="number">30</span><span class="operator">);</span>  <span class="comment"># get a block of columns</span>
  <span class="variable">$mini_aln</span> <span class="operator">=</span> <span class="variable">$aln</span><span class="operator">-&gt;</span><span class="variable">select_noncont</span><span class="operator">(</span><span class="number">1</span><span class="operator">,</span><span class="number">3</span><span class="operator">,</span><span class="number">5</span><span class="operator">,</span><span class="number">7</span><span class="operator">,</span><span class="number">11</span><span class="operator">);</span> <span class="comment"># select certain sequences</span>
  <span class="variable">$new_aln</span> <span class="operator">=</span> <span class="variable">$aln</span><span class="operator">-&gt;</span><span class="variable">remove_columns</span><span class="operator">(</span><span class="operator">[</span><span class="number">20</span><span class="operator">,</span><span class="number">30</span><span class="operator">]</span><span class="operator">);</span> <span class="comment"># remove by position</span>
  <span class="variable">$new_aln</span> <span class="operator">=</span> <span class="variable">$aln</span><span class="operator">-&gt;</span><span class="variable">remove_columns</span><span class="operator">(</span><span class="operator">[</span><span class="string">'mismatch'</span><span class="operator">]</span><span class="operator">);</span> <span class="comment"># remove by property</span>
  
  <span class="comment"># Analyze</span>
  <span class="variable">$str</span> <span class="operator">=</span> <span class="variable">$aln</span><span class="operator">-&gt;</span><span class="variable">consensus_string</span><span class="operator">(</span><span class="variable">$threshold_percent</span><span class="operator">);</span>
  <span class="variable">$str</span> <span class="operator">=</span> <span class="variable">$aln</span><span class="operator">-&gt;</span><span class="variable">match_line</span><span class="operator">();</span>
  <span class="variable">$str</span> <span class="operator">=</span> <span class="variable">$aln</span><span class="operator">-&gt;</span><span class="variable">cigar_line</span><span class="operator">();</span>
  <span class="variable">$id</span> <span class="operator">=</span> <span class="variable">$aln</span><span class="operator">-&gt;</span><span class="variable">percentage_identity</span><span class="operator">;</span>
  
  <span class="comment"># See the module documentation for details and more methods.</span>
</code></code></pre>

<h1 id="DESCRIPTION">DESCRIPTION</h1>

<p>SimpleAlign is an object that handles a multiple sequence alignment (MSA). It is very permissive of types (it does not insist on sequences being all same length, for example). Think of it as a set of sequences with a whole series of built-in manipulations and methods for reading and writing alignments.</p>

<p>SimpleAlign uses <a href="../../../../.cpan/build/BioPerl-1.6.901-ghKY8U/blib/lib/Bio/LocatableSeq.html">Bio::LocatableSeq</a>, a subclass of <a href="../../../../.cpan/build/BioPerl-1.6.901-ghKY8U/blib/lib/Bio/PrimarySeq.html">Bio::PrimarySeq</a>, to store its sequences. These are subsequences with a start and end positions in the parent reference sequence. Each sequence in the SimpleAlign object is a Bio::LocatableSeq.</p>

<p>SimpleAlign expects the combination of name, start, and end for a given sequence to be unique in the alignment, and this is the key for the internal hashes (name, start, end are abbreviated <code><code>nse</code></code> in the code). However, in some cases people do not want the name/start-end to be displayed: either multiple names in an alignment or names specific to the alignment (ROA1_HUMAN_1, ROA1_HUMAN_2 etc). These names are called <code><code>displayname</code></code>, and generally is what is used to print out the alignment. They default to name/start-end.</p>

<p>The SimpleAlign Module is derived from the Align module by Ewan Birney.</p>

<h1 id="FEEDBACK">FEEDBACK</h1>

<h2 id="Mailing-Lists">Mailing Lists</h2>

<p>User feedback is an integral part of the evolution of this and other Bioperl modules. Send your comments and suggestions preferably to one of the Bioperl mailing lists. Your participation is much appreciated.</p>

<pre><code><code>  bioperl-l@bioperl.org                  - General discussion
  http://bioperl.org/wiki/Mailing_lists  - About the mailing lists</code></code></pre>

<h2 id="Support">Support</h2>

<p>Please direct usage questions or support issues to the mailing list:</p>

<p><i>bioperl-l@bioperl.org</i></p>

<p>rather than to the module maintainer directly. Many experienced and reponsive experts will be able look at the problem and quickly address it. Please include a thorough description of the problem with code and data examples if at all possible.</p>

<h2 id="Reporting-Bugs">Reporting Bugs</h2>

<p>Report bugs to the Bioperl bug tracking system to help us keep track the bugs and their resolution. Bug reports can be submitted via the web:</p>

<pre><code><code>  https://redmine.open-bio.org/projects/bioperl/</code></code></pre>

<h1 id="AUTHOR">AUTHOR</h1>

<p>Ewan Birney, birney@ebi.ac.uk</p>

<h1 id="CONTRIBUTORS">CONTRIBUTORS</h1>

<p>Allen Day, allenday-at-ucla.edu, Richard Adams, Richard.Adams-at-ed.ac.uk, David J. Evans, David.Evans-at-vir.gla.ac.uk, Heikki Lehvaslaiho, heikki-at-bioperl-dot-org, Allen Smith, allens-at-cpan.org, Jason Stajich, jason-at-bioperl.org, Anthony Underwood, aunderwood-at-phls.org.uk, Xintao Wei &amp; Giri Narasimhan, giri-at-cs.fiu.edu Brian Osborne, bosborne at alum.mit.edu Weigang Qiu, Weigang at GENECTR-HUNTER-CUNY-EDU Hongyu Zhang, forward at hongyu.org Jay Hannah, jay at jays.net Alexandr Bezginov, albezg at gmail.com</p>

<h1 id="SEE-ALSO">SEE ALSO</h1>

<p><a href="../../../../.cpan/build/BioPerl-1.6.901-ghKY8U/blib/lib/Bio/LocatableSeq.html">Bio::LocatableSeq</a></p>

<h1 id="APPENDIX">APPENDIX</h1>

<p>The rest of the documentation details each of the object methods. Internal methods are usually preceded with a _</p>

<h2 id="new">new</h2>

<pre><code><code> <span class="variable">Title</span>     <span class="operator">:</span> <span class="variable">new</span>
 <span class="variable">Usage</span>     <span class="operator">:</span> <span class="keyword">my</span> <span class="variable">$aln</span> <span class="operator">=</span> <span class="variable">Bio::SimpleAlign</span><span class="operator">-&gt;</span><span class="variable">new</span><span class="operator">();</span>
 <span class="variable">Function</span>  <span class="operator">:</span> <span class="variable">Creates</span> <span class="variable">a</span> <span class="variable">new</span> <span class="variable">simple</span> <span class="variable">align</span> <span class="variable">object</span>
 <span class="variable">Returns</span>   <span class="operator">:</span> <span class="variable">Bio::SimpleAlign</span>
 <span class="variable">Args</span>      <span class="operator">:</span> <span class="string">-source</span>     <span class="operator">=&gt;</span> <span class="variable">string</span> <span class="variable">representing</span> <span class="variable">the</span> <span class="variable">source</span> <span class="variable">program</span>
                            <span class="variable">where</span> <span class="variable">this</span> <span class="variable">alignment</span> <span class="variable">came</span> <span class="variable">from</span>
             <span class="operator">-</span><span class="string">annotation</span> <span class="operator">=&gt;</span> <span class="variable">Bio::AnnotationCollectionI</span>
             <span class="operator">-</span><span class="string">seq_annotation</span> <span class="operator">=&gt;</span> <span class="variable">Bio::AnnotationCollectionI</span> <span class="keyword">for</span> <span class="variable">sequences</span> <span class="operator">(</span><span class="variable">requires</span> <span class="operator">-</span><span class="variable">annotation</span> <span class="variable">also</span> <span class="variable">be</span> <span class="variable">set</span><span class="operator">)</span>
             <span class="string">-seqs</span>       <span class="operator">=&gt;</span> <span class="variable">array</span> <span class="keyword">ref</span> <span class="variable">containing</span> <span class="variable">Bio::LocatableSeq</span> <span class="keyword">or</span> <span class="variable">Bio::Seq::Meta</span>
             <span class="operator">-</span><span class="string">consensus</span>  <span class="operator">=&gt;</span> <span class="variable">consensus</span> <span class="variable">string</span>
             <span class="operator">-</span><span class="string">consensus_meta</span>  <span class="operator">=&gt;</span> <span class="variable">Bio::Seq::Meta</span> <span class="variable">object</span> <span class="variable">containing</span> <span class="variable">consensus</span> <span class="variable">met</span> <span class="variable">information</span> <span class="operator">(</span><span class="variable">kludge</span><span class="operator">)</span>
</code></code></pre>

<h1 id="Modifier-methods">Modifier methods</h1>

<p>These methods modify the MSA by adding, removing or shuffling complete sequences.</p>

<h2 id="add_seq">add_seq</h2>

<pre><code><code> <span class="variable">Title</span>     <span class="operator">:</span> <span class="variable">add_seq</span>
 <span class="variable">Usage</span>     <span class="operator">:</span> <span class="variable">$myalign</span><span class="operator">-&gt;</span><span class="variable">add_seq</span><span class="operator">(</span><span class="variable">$newseq</span><span class="operator">);</span>
             <span class="variable">$myalign</span><span class="operator">-&gt;</span><span class="variable">add_seq</span><span class="operator">(</span><span class="string">-SEQ</span><span class="operator">=&gt;</span><span class="variable">$newseq</span><span class="operator">,</span> <span class="string">-ORDER</span><span class="operator">=&gt;</span><span class="number">5</span><span class="operator">);</span>
 <span class="variable">Function</span>  <span class="operator">:</span> <span class="variable">Adds</span> <span class="variable">another</span> <span class="variable">sequence</span> <span class="variable">to</span> <span class="variable">the</span> <span class="variable">alignment</span><span class="operator">.</span> <span class="variable">*Does</span> <span class="keyword">not</span><span class="operator">*</span> <span class="variable">align</span>
             <span class="variable">it</span> <span class="operator">-</span> <span class="variable">just</span> <span class="variable">adds</span> <span class="variable">it</span> <span class="variable">to</span> <span class="variable">the</span> <span class="variable">hashes</span><span class="operator">.</span>
             <span class="variable">If</span> <span class="operator">-</span><span class="variable">ORDER</span> <span class="variable">is</span> <span class="variable">specified</span><span class="operator">,</span> <span class="variable">the</span> <span class="variable">sequence</span> <span class="variable">is</span> <span class="variable">inserted</span> <span class="variable">at</span> <span class="variable">the</span>
             <span class="variable">the</span> <span class="variable">position</span> <span class="variable">spec'd</span> <span class="variable">by</span> <span class="operator">-</span><span class="variable">ORDER</span><span class="operator">,</span> <span class="keyword">and</span> <span class="variable">existing</span> <span class="variable">sequences</span>
             <span class="variable">are</span> <span class="variable">pushed</span> <span class="variable">down</span> <span class="variable">the</span> <span class="variable">storage</span> <span class="variable">array</span><span class="operator">.</span>
 <span class="variable">Returns</span>   <span class="operator">:</span> <span class="variable">nothing</span>
 <span class="variable">Args</span>      <span class="operator">:</span> <span class="variable">A</span> <span class="variable">Bio::LocatableSeq</span> <span class="variable">object</span>
             <span class="variable">Positive</span> <span class="variable">integer</span> <span class="keyword">for</span> <span class="variable">the</span> <span class="variable">sequence</span> <span class="variable">position</span> <span class="operator">(</span><span class="variable">optional</span><span class="operator">)</span>
</code></code></pre>

<p>See <a href="../../../../.cpan/build/BioPerl-1.6.901-ghKY8U/blib/lib/Bio/LocatableSeq.html">Bio::LocatableSeq</a> for more information</p>

<h2 id="remove_seq">remove_seq</h2>

<pre><code><code> <span class="variable">Title</span>     <span class="operator">:</span> <span class="variable">remove_seq</span>
 <span class="variable">Usage</span>     <span class="operator">:</span> <span class="variable">$aln</span><span class="operator">-&gt;</span><span class="variable">remove_seq</span><span class="operator">(</span><span class="variable">$seq</span><span class="operator">);</span>
 <span class="variable">Function</span>  <span class="operator">:</span> <span class="variable">Removes</span> <span class="variable">a</span> <span class="variable">single</span> <span class="variable">sequence</span> <span class="variable">from</span> <span class="variable">an</span> <span class="variable">alignment</span>
 <span class="variable">Returns</span>   <span class="operator">:</span>
 <span class="variable">Argument</span>  <span class="operator">:</span> <span class="variable">a</span> <span class="variable">Bio::LocatableSeq</span> <span class="variable">object</span>
</code></code></pre>

<h2 id="purge">purge</h2>

<pre><code><code> <span class="variable">Title</span>   <span class="operator">:</span> <span class="variable">purge</span>
 <span class="variable">Usage</span>   <span class="operator">:</span> <span class="variable">$aln</span><span class="operator">-&gt;</span><span class="variable">purge</span><span class="operator">(</span><span class="number">0</span><span class="operator">.</span><span class="number">7</span><span class="operator">);</span>
 <span class="variable">Function</span><span class="operator">:</span> <span class="variable">Removes</span> <span class="variable">sequences</span> <span class="variable">above</span> <span class="keyword">given</span> <span class="variable">sequence</span> <span class="variable">similarity</span>
           <span class="variable">This</span> <span class="variable">function</span> <span class="variable">will</span> <span class="variable">grind</span> <span class="variable">on</span> <span class="variable">large</span> <span class="variable">alignments</span><span class="operator">.</span> <span class="variable">Beware</span><span class="operator">!</span>
 <span class="variable">Example</span> <span class="operator">:</span>
 <span class="variable">Returns</span> <span class="operator">:</span> <span class="variable">An</span> <span class="variable">array</span> <span class="variable">of</span> <span class="variable">the</span> <span class="variable">removed</span> <span class="variable">sequences</span>
 <span class="variable">Args</span>    <span class="operator">:</span> <span class="variable">float</span><span class="operator">,</span> <span class="variable">threshold</span> <span class="keyword">for</span> <span class="variable">similarity</span>
</code></code></pre>

<h2 id="sort_alphabetically">sort_alphabetically</h2>

<pre><code><code> Title     : sort_alphabetically
 Usage     : $ali-&gt;sort_alphabetically
 Function  : Changes the order of the alignment to alphabetical on name
             followed by numerical by number.
 Returns   :
 Argument  :</code></code></pre>

<h2 id="sort_by_list">sort_by_list</h2>

<pre><code><code> Title     : sort_by_list
 Usage     : $aln_ordered=$aln-&gt;sort_by_list($list_file)
 Function  : Arbitrarily order sequences in an alignment
 Returns   : A new Bio::SimpleAlign object
 Argument  : a file listing sequence names in intended order (one name per line)</code></code></pre>

<h2 id="set_new_reference">set_new_reference</h2>

<pre><code><code> Title     : set_new_reference
 Usage     : $aln-&gt;set_new_reference(3 or &#39;B31&#39;):  Select the 3rd sequence, or
             the sequence whoes name is &quot;B31&quot; (full, exact, and case-sensitive),
             as the reference (1st) sequence
 Function  : Change/Set a new reference (i.e., the first) sequence
 Returns   : a new Bio::SimpleAlign object.
             Throws an exception if designated sequence not found
 Argument  : a positive integer of sequence order, or a sequence name
             in the original alignment</code></code></pre>

<h2 id="uniq_seq">uniq_seq</h2>

<pre><code><code> Title     : uniq_seq
 Usage     : $aln-&gt;uniq_seq():  Remove identical sequences in
             in the alignment.  Ambiguous base (&quot;N&quot;, &quot;n&quot;) and
             leading and ending gaps (&quot;-&quot;) are NOT counted as
             differences.
 Function  : Make a new alignment of unique sequence types (STs)
 Returns   : 1a. if called in a scalar context, 
                a new Bio::SimpleAlign object (all sequences renamed as &quot;ST&quot;)
             1b. if called in an array context, 
                a new Bio::SimpleAlign object, and a hashref whose keys
                are sequence types, and whose values are arrayrefs to 
                lists of sequence ids within the corresponding sequence type
             2. if $aln-&gt;verbose &gt; 0, ST of each sequence is sent to 
                STDERR (in a tabular format)
 Argument  : None</code></code></pre>

<h1 id="Sequence-selection-methods">Sequence selection methods</h1>

<p>Methods returning one or more sequences objects.</p>

<h2 id="each_seq">each_seq</h2>

<pre><code><code> Title     : each_seq
 Usage     : foreach $seq ( $align-&gt;each_seq() )
 Function  : Gets a Seq object from the alignment
 Returns   : Seq object
 Argument  :</code></code></pre>

<h2 id="each_alphabetically">each_alphabetically</h2>

<pre><code><code> Title     : each_alphabetically
 Usage     : foreach $seq ( $ali-&gt;each_alphabetically() )
 Function  : Returns a sequence object, but the objects are returned
             in alphabetically sorted order.
             Does not change the order of the alignment.
 Returns   : Seq object
 Argument  :</code></code></pre>

<h2 id="each_seq_with_id">each_seq_with_id</h2>

<pre><code><code> Title     : each_seq_with_id
 Usage     : foreach $seq ( $align-&gt;each_seq_with_id() )
 Function  : Gets a Seq objects from the alignment, the contents
             being those sequences with the given name (there may be
             more than one)
 Returns   : Seq object
 Argument  : a seq name</code></code></pre>

<h2 id="get_seq_by_pos">get_seq_by_pos</h2>

<pre><code><code> Title     : get_seq_by_pos
 Usage     : $seq = $aln-&gt;get_seq_by_pos(3) # third sequence from the alignment
 Function  : Gets a sequence based on its position in the alignment.
             Numbering starts from 1.  Sequence positions larger than
             num_sequences() will thow an error.
 Returns   : a Bio::LocatableSeq object
 Args      : positive integer for the sequence position</code></code></pre>

<h2 id="get_seq_by_id">get_seq_by_id</h2>

<pre><code><code> Title     : get_seq_by_id
 Usage     : $seq = $aln-&gt;get_seq_by_id($name) # seq named $name
 Function  : Gets a sequence based on its name.
             Sequences that do not exist will warn and return undef
 Returns   : a Bio::LocatableSeq object
 Args      : string for sequence name</code></code></pre>

<h2 id="seq_with_features">seq_with_features</h2>

<pre><code><code> <span class="variable">Title</span>   <span class="operator">:</span> <span class="variable">seq_with_features</span>
 <span class="variable">Usage</span>   <span class="operator">:</span> <span class="variable">$seq</span> <span class="operator">=</span> <span class="variable">$aln</span><span class="operator">-&gt;</span><span class="variable">seq_with_features</span><span class="operator">(</span><span class="string">-pos</span> <span class="operator">=&gt;</span> <span class="number">1</span><span class="operator">,</span>
                                          <span class="string">-consensus</span> <span class="operator">=&gt;</span> <span class="number">60</span>
                                          <span class="operator">-</span><span class="string">mask</span> <span class="operator">=&gt;</span>
           <span class="keyword">sub</span><span class="variable"> </span><span class="operator">{</span> <span class="keyword">my</span> <span class="variable">$consensus</span> <span class="operator">=</span> <span class="keyword">shift</span><span class="operator">;</span>
 
                 <span class="keyword">for</span> <span class="keyword">my</span> <span class="variable">$i</span> <span class="operator">(</span><span class="number">1</span><span class="operator">..</span><span class="number">5</span><span class="operator">){</span>
                    <span class="keyword">my</span> <span class="variable">$n</span> <span class="operator">=</span> <span class="string">'N'</span> <span class="operator">x</span> <span class="variable">$i</span><span class="operator">;</span>
                    <span class="keyword">my</span> <span class="variable">$q</span> <span class="operator">=</span> <span class="string">'\?'</span> <span class="operator">x</span> <span class="variable">$i</span><span class="operator">;</span>
                    <span class="keyword">while</span><span class="operator">(</span><span class="variable">$consensus</span> <span class="operator">=~</span> <span class="regex">/[^?]</span><span class="variable">$q</span><span class="regex">[^?]/</span><span class="operator">){</span>
                       <span class="variable">$consensus</span> <span class="operator">=~</span> <span class="regex">s/([^?])</span><span class="variable">$q</span><span class="regex">([^?])/$1$n$2/</span><span class="operator">;</span>
                    <span class="operator">}</span>
                  <span class="operator">}</span>
                 <span class="keyword">return</span> <span class="variable">$consensus</span><span class="operator">;</span>
               <span class="operator">}</span>
                                         <span class="operator">);</span>
 <span class="variable">Function</span><span class="operator">:</span> <span class="variable">produces</span> <span class="variable">a</span> <span class="variable">Bio::Seq</span> <span class="variable">object</span> <span class="variable">by</span> <span class="variable">first</span> <span class="variable">splicing</span> <span class="variable">gaps</span> <span class="variable">from</span> <span class="operator">-</span><span class="keyword">pos</span>
           <span class="operator">(</span><span class="variable">by</span> <span class="variable">means</span> <span class="variable">of</span> <span class="variable">a</span> <span class="variable">splice_by_seq_pos</span><span class="operator">()</span> <span class="variable">call</span><span class="operator">),</span> <span class="variable">then</span> <span class="variable">creating</span>
           <span class="variable">features</span> <span class="variable">using</span> <span class="variable">non</span><span class="operator">-</span><span class="regex">? chars (by means of a consensus_string()
           call with stringency -consensus).
 Returns : a Bio::Seq object
 Args    : -pos : required. sequence from which to build the Bio::Seq
             object
           -consensus : optional, defaults to consensus_string()'s
             default cutoff value
           -mask : optional, a coderef to apply to consensus_string()'s
             output before building features.  this may be useful for
             closing gaps of 1 bp by masking over them with N, for
             instance
 </span>
</code></code></pre>

<h1 id="Create-new-alignments">Create new alignments</h1>

<p>The result of these methods are horizontal or vertical subsets of the current MSA.</p>

<h2 id="select">select</h2>

<pre><code><code> Title     : select
 Usage     : $aln2 = $aln-&gt;select(1, 3) # three first sequences
 Function  : Creates a new alignment from a continuous subset of
             sequences.  Numbering starts from 1.  Sequence positions
             larger than num_sequences() will thow an error.
 Returns   : a Bio::SimpleAlign object
 Args      : positive integer for the first sequence
             positive integer for the last sequence to include (optional)</code></code></pre>

<h2 id="select_noncont">select_noncont</h2>

<pre><code><code> Title     : select_noncont
 Usage     : # 1st and 3rd sequences, sorted
             $aln2 = $aln-&gt;select_noncont(1, 3)

             # 1st and 3rd sequences, sorted (same as first)
             $aln2 = $aln-&gt;select_noncont(3, 1)

             # 1st and 3rd sequences, unsorted
             $aln2 = $aln-&gt;select_noncont(&#39;nosort&#39;,3, 1)

 Function  : Creates a new alignment from a subset of sequences.  Numbering
             starts from 1.  Sequence positions larger than num_sequences() will
             throw an error.  Sorts the order added to new alignment by default,
             to prevent sorting pass &#39;nosort&#39; as the first argument in the list.
 Returns   : a Bio::SimpleAlign object
 Args      : array of integers for the sequences.  If the string &#39;nosort&#39; is
             passed as the first argument, the sequences will not be sorted
             in the new alignment but will appear in the order listed.</code></code></pre>

<h2 id="slice">slice</h2>

<pre><code><code> Title     : slice
 Usage     : $aln2 = $aln-&gt;slice(20,30)
 Function  : Creates a slice from the alignment inclusive of start and
             end columns, and the first column in the alignment is denoted 1.
             Sequences with no residues in the slice are excluded from the
             new alignment and a warning is printed. Slice beyond the length of
             the sequence does not do padding.
 Returns   : A Bio::SimpleAlign object
 Args      : Positive integer for start column, positive integer for end column,
             optional boolean which if true will keep gap-only columns in the newly
             created slice. Example:

             $aln2 = $aln-&gt;slice(20,30,1)</code></code></pre>

<h2 id="remove_columns">remove_columns</h2>

<pre><code><code> Title     : remove_columns
 Usage     : $aln2 = $aln-&gt;remove_columns([&#39;mismatch&#39;,&#39;weak&#39;]) or
             $aln2 = $aln-&gt;remove_columns([0,0],[6,8])
 Function  : Creates an aligment with columns removed corresponding to
             the specified type or by specifying the columns by number.
 Returns   : Bio::SimpleAlign object
 Args      : Array ref of types (&#39;match&#39;|&#39;weak&#39;|&#39;strong&#39;|&#39;mismatch&#39;|&#39;gaps&#39;|
             &#39;all_gaps_columns&#39;) or array ref where the referenced array
             contains a pair of integers that specify a range.
             The first column is 0</code></code></pre>

<h2 id="remove_gaps">remove_gaps</h2>

<pre><code><code> Title     : remove_gaps
 Usage     : $aln2 = $aln-&gt;remove_gaps
 Function  : Creates an aligment with gaps removed
 Returns   : a Bio::SimpleAlign object
 Args      : a gap character(optional) if none specified taken
                from $self-&gt;gap_char,
             [optional] $all_gaps_columns flag (1 or 0, default is 0)
                        indicates that only all-gaps columns should be deleted</code></code></pre>

<p>Used from method <a>remove_columns</a> in most cases. Set gap character using <a>gap_char()</a>.</p>

<h1 id="Change-sequences-within-the-MSA">Change sequences within the MSA</h1>

<p>These methods affect characters in all sequences without changing the alignment.</p>

<h2 id="splice_by_seq_pos">splice_by_seq_pos</h2>

<pre><code><code> <span class="variable">Title</span>   <span class="operator">:</span> <span class="variable">splice_by_seq_pos</span>
 <span class="variable">Usage</span>   <span class="operator">:</span> <span class="variable">$status</span> <span class="operator">=</span> <span class="variable">splice_by_seq_pos</span><span class="operator">(</span><span class="number">1</span><span class="operator">);</span>
 <span class="variable">Function</span><span class="operator">:</span> <span class="variable">splices</span> <span class="variable">all</span> <span class="variable">aligned</span> <span class="variable">sequences</span> <span class="variable">where</span> <span class="variable">the</span> <span class="variable">specified</span> <span class="variable">sequence</span>
           <span class="variable">has</span> <span class="variable">gaps</span><span class="operator">.</span>
 <span class="variable">Example</span> <span class="operator">:</span>
 <span class="variable">Returns</span> <span class="operator">:</span> <span class="number">1</span> <span class="variable">on</span> <span class="variable">success</span>
 <span class="variable">Args</span>    <span class="operator">:</span> <span class="variable">position</span> <span class="variable">of</span> <span class="variable">sequence</span> <span class="variable">to</span> <span class="keyword">splice</span> <span class="variable">by</span>
</code></code></pre>

<h2 id="map_chars">map_chars</h2>

<pre><code><code> Title     : map_chars
 Usage     : $ali-&gt;map_chars(&#39;\.&#39;,&#39;-&#39;)
 Function  : Does a s/$arg1/$arg2/ on the sequences. Useful for gap
             characters

             Notice that the from (arg1) is interpretted as a regex,
             so be careful about quoting meta characters (eg
             $ali-&gt;map_chars(&#39;.&#39;,&#39;-&#39;) wont do what you want)
 Returns   :
 Argument  : &#39;from&#39; rexexp
             &#39;to&#39; string</code></code></pre>

<h2 id="uppercase">uppercase</h2>

<pre><code><code> Title     : uppercase()
 Usage     : $ali-&gt;uppercase()
 Function  : Sets all the sequences to uppercase
 Returns   :
 Argument  :</code></code></pre>

<h2 id="cigar_line">cigar_line</h2>

<pre><code><code> Title    : cigar_line()
 Usage    : %cigars = $align-&gt;cigar_line()
 Function : Generates a &quot;cigar&quot; (Compact Idiosyncratic Gapped Alignment
            Report) line for each sequence in the alignment. Examples are
            &quot;1,60&quot; or &quot;5,10:12,58&quot;, where the numbers refer to conserved
            positions within the alignment. The keys of the hash are the
            NSEs (name/start/end) assigned to each sequence.
 Args     : threshold (optional, defaults to 100)
 Returns  : Hash of strings (cigar lines)</code></code></pre>

<h2 id="match_line">match_line</h2>

<pre><code><code> Title    : match_line()
 Usage    : $line = $align-&gt;match_line()
 Function : Generates a match line - much like consensus string
            except that a line indicating the &#39;*&#39; for a match.
 Args     : (optional) Match line characters (&#39;*&#39; by default)
            (optional) Strong match char (&#39;:&#39; by default)
            (optional) Weak match char (&#39;.&#39; by default)
 Returns  : String</code></code></pre>

<h2 id="gap_line">gap_line</h2>

<pre><code><code> Title    : gap_line()
 Usage    : $line = $align-&gt;gap_line()
 Function : Generates a gap line - much like consensus string
            except that a line where &#39;-&#39; represents gap
 Args     : (optional) gap line characters (&#39;-&#39; by default)
 Returns  : string</code></code></pre>

<h2 id="all_gap_line">all_gap_line</h2>

<pre><code><code> Title    : all_gap_line()
 Usage    : $line = $align-&gt;all_gap_line()
 Function : Generates a gap line - much like consensus string
            except that a line where &#39;-&#39; represents all-gap column
 Args     : (optional) gap line characters (&#39;-&#39; by default)
 Returns  : string</code></code></pre>

<h2 id="gap_col_matrix">gap_col_matrix</h2>

<pre><code><code> Title    : gap_col_matrix()
 Usage    : my $cols = $align-&gt;gap_col_matrix()
 Function : Generates an array of hashes where
            each entry in the array is a hash reference
            with keys of all the sequence names and
            and value of 1 or 0 if the sequence has a gap at that column
 Args     : (optional) gap line characters ($aln-&gt;gap_char or &#39;-&#39; by default)</code></code></pre>

<h2 id="match">match</h2>

<pre><code><code> Title     : match()
 Usage     : $ali-&gt;match()
 Function  : Goes through all columns and changes residues that are
             identical to residue in first sequence to match &#39;.&#39;
             character. Sets match_char.

             USE WITH CARE: Most MSA formats do not support match
             characters in sequences, so this is mostly for output
             only. NEXUS format (Bio::AlignIO::nexus) can handle
             it.
 Returns   : 1
 Argument  : a match character, optional, defaults to &#39;.&#39;</code></code></pre>

<h2 id="unmatch">unmatch</h2>

<pre><code><code> Title     : unmatch()
 Usage     : $ali-&gt;unmatch()
 Function  : Undoes the effect of method match. Unsets match_char.
 Returns   : 1
 Argument  : a match character, optional, defaults to &#39;.&#39;</code></code></pre>

<p>See <a href="../../../../.cpan/build/BioPerl-1.6.901-ghKY8U/blib/lib/Bio/DB/GFF/Aggregator/match.html">match</a> and <a>match_char</a></p>

<h1 id="MSA-attributes">MSA attributes</h1>

<p>Methods for setting and reading the MSA attributes.</p>

<p>Note that the methods defining character semantics depend on the user to set them sensibly. They are needed only by certain input/output methods. Unset them by setting to an empty string (&#39;&#39;).</p>

<h2 id="id">id</h2>

<pre><code><code> Title     : id
 Usage     : $myalign-&gt;id(&quot;Ig&quot;)
 Function  : Gets/sets the id field of the alignment
 Returns   : An id string
 Argument  : An id string (optional)</code></code></pre>

<h2 id="accession">accession</h2>

<pre><code><code> Title     : accession
 Usage     : $myalign-&gt;accession(&quot;PF00244&quot;)
 Function  : Gets/sets the accession field of the alignment
 Returns   : An acc string
 Argument  : An acc string (optional)</code></code></pre>

<h2 id="description">description</h2>

<pre><code><code> Title     : description
 Usage     : $myalign-&gt;description(&quot;14-3-3 proteins&quot;)
 Function  : Gets/sets the description field of the alignment
 Returns   : An description string
 Argument  : An description string (optional)</code></code></pre>

<h2 id="missing_char">missing_char</h2>

<pre><code><code> Title     : missing_char
 Usage     : $myalign-&gt;missing_char(&quot;?&quot;)
 Function  : Gets/sets the missing_char attribute of the alignment
             It is generally recommended to set it to &#39;n&#39; or &#39;N&#39;
             for nucleotides and to &#39;X&#39; for protein.
 Returns   : An missing_char string,
 Argument  : An missing_char string (optional)</code></code></pre>

<h2 id="match_char">match_char</h2>

<pre><code><code> Title     : match_char
 Usage     : $myalign-&gt;match_char(&#39;.&#39;)
 Function  : Gets/sets the match_char attribute of the alignment
 Returns   : An match_char string,
 Argument  : An match_char string (optional)</code></code></pre>

<h2 id="gap_char">gap_char</h2>

<pre><code><code> Title     : gap_char
 Usage     : $myalign-&gt;gap_char(&#39;-&#39;)
 Function  : Gets/sets the gap_char attribute of the alignment
 Returns   : An gap_char string, defaults to &#39;-&#39;
 Argument  : An gap_char string (optional)</code></code></pre>

<h2 id="symbol_chars">symbol_chars</h2>

<pre><code><code> <span class="variable">Title</span>   <span class="operator">:</span> <span class="variable">symbol_chars</span>
 <span class="variable">Usage</span>   <span class="operator">:</span> <span class="keyword">my</span> <span class="variable">@symbolchars</span> <span class="operator">=</span> <span class="variable">$aln</span><span class="operator">-&gt;</span><span class="variable">symbol_chars</span><span class="operator">;</span>
 <span class="variable">Function</span><span class="operator">:</span> <span class="variable">Returns</span> <span class="variable">all</span> <span class="variable">the</span> <span class="variable">seen</span> <span class="variable">symbols</span> <span class="operator">(</span><span class="variable">other</span> <span class="variable">than</span> <span class="variable">gaps</span><span class="operator">)</span>
 <span class="variable">Returns</span> <span class="operator">:</span> <span class="variable">array</span> <span class="variable">of</span> <span class="variable">characters</span> <span class="variable">that</span> <span class="variable">are</span> <span class="variable">the</span> <span class="variable">seen</span> <span class="variable">symbols</span>
 <span class="variable">Args</span>    <span class="operator">:</span> <span class="variable">boolean</span> <span class="variable">to</span> <span class="variable">include</span> <span class="variable">the</span> <span class="variable">gap</span><span class="operator">/</span><span class="variable">missing</span><span class="operator">/</span><span class="variable">match</span> <span class="variable">characters</span>
</code></code></pre>

<h1 id="Alignment-descriptors">Alignment descriptors</h1>

<p>These read only methods describe the MSA in various ways.</p>

<h2 id="score">score</h2>

<pre><code><code> Title     : score
 Usage     : $str = $ali-&gt;score()
 Function  : get/set a score of the alignment
 Returns   : a score for the alignment
 Argument  : an optional score to set</code></code></pre>

<h2 id="consensus_string">consensus_string</h2>

<pre><code><code> Title     : consensus_string
 Usage     : $str = $ali-&gt;consensus_string($threshold_percent)
 Function  : Makes a strict consensus
 Returns   : Consensus string
 Argument  : Optional threshold ranging from 0 to 100.
             The consensus residue has to appear at least threshold %
             of the sequences at a given location, otherwise a &#39;?&#39;
             character will be placed at that location.
             (Default value = 0%)</code></code></pre>

<h2 id="consensus_iupac">consensus_iupac</h2>

<pre><code><code> Title     : consensus_iupac
 Usage     : $str = $ali-&gt;consensus_iupac()
 Function  : Makes a consensus using IUPAC ambiguity codes from DNA
             and RNA. The output is in upper case except when gaps in
             a column force output to be in lower case.

             Note that if your alignment sequences contain a lot of
             IUPAC ambiquity codes you often have to manually set
             alphabet.  Bio::PrimarySeq::_guess_type thinks they
             indicate a protein sequence.
 Returns   : consensus string
 Argument  : none
 Throws    : on protein sequences</code></code></pre>

<h2 id="consensus_meta">consensus_meta</h2>

<pre><code><code> Title     : consensus_meta
 Usage     : $seqmeta = $ali-&gt;consensus_meta()
 Function  : Returns a Bio::Seq::Meta object containing the consensus
             strings derived from meta data analysis.
 Returns   : Bio::Seq::Meta 
 Argument  : Bio::Seq::Meta 
 Throws    : non-MetaI object</code></code></pre>

<h2 id="is_flush">is_flush</h2>

<pre><code><code> Title     : is_flush
 Usage     : if ( $ali-&gt;is_flush() )
 Function  : Tells you whether the alignment
           : is flush, i.e. all of the same length
 Returns   : 1 or 0
 Argument  :</code></code></pre>

<h2 id="length">length</h2>

<pre><code><code> Title     : length()
 Usage     : $len = $ali-&gt;length()
 Function  : Returns the maximum length of the alignment.
             To be sure the alignment is a block, use is_flush
 Returns   : Integer
 Argument  :</code></code></pre>

<h2 id="maxdisplayname_length">maxdisplayname_length</h2>

<pre><code><code> Title     : maxdisplayname_length
 Usage     : $ali-&gt;maxdisplayname_length()
 Function  : Gets the maximum length of the displayname in the
             alignment. Used in writing out various MSA formats.
 Returns   : integer
 Argument  :</code></code></pre>

<h2 id="max_metaname_length">max_metaname_length</h2>

<pre><code><code> Title     : max_metaname_length
 Usage     : $ali-&gt;max_metaname_length()
 Function  : Gets the maximum length of the meta name tags in the
             alignment for the sequences and for the alignment.
             Used in writing out various MSA formats.
 Returns   : integer
 Argument  : None</code></code></pre>

<h2 id="num_residues">num_residues</h2>

<pre><code><code> Title     : num_residues
 Usage     : $no = $ali-&gt;num_residues
 Function  : number of residues in total in the alignment
 Returns   : integer
 Argument  :
 Note      : replaces no_residues() </code></code></pre>

<h2 id="num_sequences">num_sequences</h2>

<pre><code><code> Title     : num_sequences
 Usage     : $depth = $ali-&gt;num_sequences
 Function  : number of sequence in the sequence alignment
 Returns   : integer
 Argument  : none
 Note      : replaces no_sequences()</code></code></pre>

<h2 id="average_percentage_identity">average_percentage_identity</h2>

<pre><code><code> Title   : average_percentage_identity
 Usage   : $id = $align-&gt;average_percentage_identity
 Function: The function uses a fast method to calculate the average
           percentage identity of the alignment
 Returns : The average percentage identity of the alignment
 Args    : None
 Notes   : This method implemented by Kevin Howe calculates a figure that is
           designed to be similar to the average pairwise identity of the
           alignment (identical in the absence of gaps), without having to
           explicitly calculate pairwise identities proposed by Richard Durbin.
           Validated by Ewan Birney ad Alex Bateman.</code></code></pre>

<h2 id="percentage_identity">percentage_identity</h2>

<pre><code><code> Title   : percentage_identity
 Usage   : $id = $align-&gt;percentage_identity
 Function: The function calculates the average percentage identity
           (aliased to average_percentage_identity)
 Returns : The average percentage identity
 Args    : None</code></code></pre>

<h2 id="overall_percentage_identity">overall_percentage_identity</h2>

<pre><code><code> Title   : overall_percentage_identity
 Usage   : $id = $align-&gt;overall_percentage_identity
           $id = $align-&gt;overall_percentage_identity(&#39;short&#39;)
 Function: The function calculates the percentage identity of
           the conserved columns
 Returns : The percentage identity of the conserved columns
 Args    : length value to use, optional defaults to alignment length
                 possible values: &#39;align&#39;, &#39;short&#39;, &#39;long&#39;</code></code></pre>

<p>The argument values &#39;short&#39; and &#39;long&#39; refer to shortest and longest sequence in the alignment. Method modification code by Hongyu Zhang.</p>

<h1 id="Alignment-positions">Alignment positions</h1>

<p>Methods to map a sequence position into an alignment column and back. column_from_residue_number() does the former. The latter is really a property of the sequence object and can done using <a>Bio::LocatableSeq::location_from_column</a>:</p>

<pre><code><code>    <span class="comment"># select somehow a sequence from the alignment, e.g.</span>
    <span class="keyword">my</span> <span class="variable">$seq</span> <span class="operator">=</span> <span class="variable">$aln</span><span class="operator">-&gt;</span><span class="variable">get_seq_by_pos</span><span class="operator">(</span><span class="number">1</span><span class="operator">);</span>
    <span class="comment">#$loc is undef or Bio::LocationI object</span>
    <span class="keyword">my</span> <span class="variable">$loc</span> <span class="operator">=</span> <span class="variable">$seq</span><span class="operator">-&gt;</span><span class="variable">location_from_column</span><span class="operator">(</span><span class="number">5</span><span class="operator">);</span>
</code></code></pre>

<h2 id="column_from_residue_number">column_from_residue_number</h2>

<pre><code><code> Title   : column_from_residue_number
 Usage   : $col = $ali-&gt;column_from_residue_number( $seqname, $resnumber)
 Function: This function gives the position in the alignment
           (i.e. column number) of the given residue number in the
           sequence with the given name. For example, for the
           alignment

             Seq1/91-97 AC..DEF.GH.
             Seq2/24-30 ACGG.RTY...
                Seq3/43-51 AC.DDEF.GHI

           column_from_residue_number( &quot;Seq1&quot;, 94 ) returns 6.
           column_from_residue_number( &quot;Seq2&quot;, 25 ) returns 2.
           column_from_residue_number( &quot;Seq3&quot;, 50 ) returns 10.

           An exception is thrown if the residue number would lie
           outside the length of the aligment
           (e.g. column_from_residue_number( &quot;Seq2&quot;, 22 )

          Note: If the the parent sequence is represented by more than
                one alignment sequence and the residue number is present in
                them, this method finds only the first one.

 Returns : A column number for the position in the alignment of the
           given residue in the given sequence (1 = first column)
 Args    : A sequence id/name (not a name/start-end)
           A residue number in the whole sequence (not just that
           segment of it in the alignment)</code></code></pre>

<h1 id="Sequence-names">Sequence names</h1>

<p>Methods to manipulate the display name. The default name based on the sequence id and subsequence positions can be overridden in various ways.</p>

<h2 id="displayname">displayname</h2>

<pre><code><code> Title     : displayname
 Usage     : $myalign-&gt;displayname(&quot;Ig&quot;, &quot;IgA&quot;)
 Function  : Gets/sets the display name of a sequence in the alignment
 Returns   : A display name string
 Argument  : name of the sequence
             displayname of the sequence (optional)</code></code></pre>

<h2 id="set_displayname_count">set_displayname_count</h2>

<pre><code><code> Title     : set_displayname_count
 Usage     : $ali-&gt;set_displayname_count
 Function  : Sets the names to be name_# where # is the number of
             times this name has been used.
 Returns   : 1, on success
 Argument  :</code></code></pre>

<h2 id="set_displayname_flat">set_displayname_flat</h2>

<pre><code><code> Title     : set_displayname_flat
 Usage     : $ali-&gt;set_displayname_flat()
 Function  : Makes all the sequences be displayed as just their name,
             not name/start-end
 Returns   : 1
 Argument  :</code></code></pre>

<h2 id="set_displayname_normal">set_displayname_normal</h2>

<pre><code><code> Title     : set_displayname_normal
 Usage     : $ali-&gt;set_displayname_normal()
 Function  : Makes all the sequences be displayed as name/start-end
 Returns   : 1, on success
 Argument  :</code></code></pre>

<h2 id="source">source</h2>

<pre><code><code> Title   : source
 Usage   : $obj-&gt;source($newval)
 Function: sets the Alignment source program
 Example :
 Returns : value of source
 Args    : newvalue (optional)</code></code></pre>

<h2 id="set_displayname_safe">set_displayname_safe</h2>

<pre><code><code> <span class="variable">Title</span>     <span class="operator">:</span> <span class="variable">set_displayname_safe</span>
 <span class="variable">Usage</span>     <span class="operator">:</span> <span class="operator">(</span><span class="variable">$new_aln</span><span class="operator">,</span> <span class="variable">$ref_name</span><span class="operator">)=</span><span class="variable">$ali</span><span class="operator">-&gt;</span><span class="variable">set_displayname_safe</span><span class="operator">(</span><span class="number">4</span><span class="operator">)</span>
 <span class="variable">Function</span>  <span class="operator">:</span> <span class="variable">Assign</span> <span class="variable">machine</span><span class="operator">-</span><span class="variable">generated</span> <span class="variable">serial</span> <span class="variable">names</span> <span class="variable">to</span> <span class="variable">sequences</span> <span class="variable">in</span> <span class="variable">input</span> <span class="variable">order</span><span class="operator">.</span>
             <span class="variable">Designed</span> <span class="variable">to</span> <span class="variable">protect</span> <span class="variable">names</span> <span class="variable">during</span> <span class="variable">PHYLIP</span> <span class="variable">runs</span><span class="operator">.</span> <span class="variable">Assign</span> <span class="number">10</span><span class="operator">-</span><span class="variable">char</span> <span class="variable">string</span>
             <span class="variable">in</span> <span class="variable">the</span> <span class="variable">form</span> <span class="variable">of</span> <span class="string">"S000000001"</span> <span class="variable">to</span> <span class="string">"S999999999"</span><span class="operator">.</span> <span class="variable">Restore</span> <span class="variable">the</span> <span class="variable">original</span>
             <span class="variable">names</span> <span class="variable">using</span> <span class="string">"restore_displayname"</span><span class="operator">.</span>
 <span class="variable">Returns</span>   <span class="operator">:</span> <span class="number">1.</span> <span class="variable">a</span> <span class="variable">new</span> <span class="variable">$aln</span> <span class="variable">with</span> <span class="keyword">system</span> <span class="variable">names</span><span class="operator">;</span>
             <span class="number">2.</span> <span class="variable">a</span> <span class="variable">hash</span> <span class="keyword">ref</span> <span class="keyword">for</span> <span class="variable">restoring</span> <span class="variable">names</span>
 <span class="variable">Argument</span>  <span class="operator">:</span> <span class="variable">Number</span> <span class="keyword">for</span> <span class="variable">id</span> <span class="keyword">length</span> <span class="operator">(</span><span class="keyword">default</span> <span class="number">10</span><span class="operator">)</span>
</code></code></pre>

<h2 id="restore_displayname">restore_displayname</h2>

<pre><code><code> Title     : restore_displayname
 Usage     : $aln_name_restored=$ali-&gt;restore_displayname($hash_ref)
 Function  : Restore original sequence names (after running
             $ali-&gt;set_displayname_safe)
 Returns   : a new $aln with names restored.
 Argument  : a hash reference of names from &quot;set_displayname_safe&quot;.</code></code></pre>

<h2 id="sort_by_start">sort_by_start</h2>

<pre><code><code> Title     : sort_by_start
 Usage     : $ali-&gt;sort_by_start
 Function  : Changes the order of the alignment to the start position of each
             subalignment    
 Returns   :
 Argument  :</code></code></pre>

<h2 id="bracket_string">bracket_string</h2>

<pre><code><code> <span class="variable">Title</span>     <span class="operator">:</span> <span class="variable">bracket_string</span>
 <span class="variable">Usage</span>     <span class="operator">:</span> <span class="keyword">my</span> <span class="variable">@params</span> <span class="operator">=</span> <span class="operator">(</span><span class="string">-refseq</span>     <span class="operator">=&gt;</span> <span class="string">'testseq'</span><span class="operator">,</span>
                           <span class="string">-allele1</span>    <span class="operator">=&gt;</span> <span class="string">'allele1'</span><span class="operator">,</span>
                           <span class="string">-allele2</span>    <span class="operator">=&gt;</span> <span class="string">'allele2'</span><span class="operator">,</span>
                           <span class="string">-delimiters</span> <span class="operator">=&gt;</span> <span class="string">'{}'</span><span class="operator">,</span>
                           <span class="string">-separator</span>  <span class="operator">=&gt;</span> <span class="string">'/'</span><span class="operator">);</span>
             <span class="variable">$str</span> <span class="operator">=</span> <span class="variable">$aln</span><span class="operator">-&gt;</span><span class="variable">bracket_string</span><span class="operator">(</span><span class="variable">@params</span><span class="operator">)</span>
 
 <span class="variable">Function</span> <span class="operator">:</span>  <span class="variable">When</span> <span class="variable">supplied</span> <span class="variable">with</span> <span class="variable">a</span> <span class="variable">list</span> <span class="variable">of</span> <span class="variable">parameters</span> <span class="operator">(</span><span class="variable">see</span> <span class="variable">below</span><span class="operator">),</span> <span class="variable">returns</span> <span class="variable">a</span>
             <span class="variable">string</span> <span class="variable">in</span> <span class="variable">BIC</span> <span class="keyword">format</span><span class="operator">.</span> <span class="variable">This</span> <span class="variable">is</span> <span class="variable">used</span> <span class="keyword">for</span> <span class="variable">allelic</span> <span class="variable">comparisons</span><span class="operator">.</span>
             <span class="variable">Briefly</span><span class="operator">,</span> <span class="keyword">if</span> <span class="variable">either</span> <span class="variable">allele</span> <span class="variable">contains</span> <span class="variable">a</span> <span class="variable">base</span> <span class="variable">change</span> <span class="keyword">when</span> <span class="variable">compared</span> <span class="variable">to</span>
             <span class="variable">the</span> <span class="variable">refseq</span><span class="operator">,</span> <span class="variable">the</span> <span class="variable">base</span> <span class="keyword">or</span> <span class="variable">gap</span> <span class="keyword">for</span> <span class="keyword">each</span> <span class="variable">allele</span> <span class="variable">is</span> <span class="variable">represented</span> <span class="variable">in</span>
             <span class="variable">brackets</span> <span class="variable">in</span> <span class="variable">the</span> <span class="variable">order</span> <span class="variable">present</span> <span class="variable">in</span> <span class="variable">the</span> <span class="string">'alleles'</span> <span class="variable">parameter</span><span class="operator">.</span>
 
             <span class="variable">For</span> <span class="variable">the</span> <span class="variable">following</span> <span class="variable">data</span><span class="operator">:</span>
 
             <span class="operator">&gt;</span><span class="variable">testseq</span>
             <span class="variable">GGATCCATTGCTACT</span>
             <span class="operator">&gt;</span><span class="variable">allele1</span>
             <span class="variable">GGATCCATTCCTACT</span>
             <span class="operator">&gt;</span><span class="variable">allele2</span>
             <span class="variable">GGAT</span><span class="operator">--</span><span class="variable">ATTCCTCCT</span>
 
             <span class="variable">the</span> <span class="variable">returned</span> <span class="variable">string</span> <span class="variable">with</span> <span class="variable">parameters</span> <span class="string">'refseq =&gt; testseq'</span> <span class="keyword">and</span>
             <span class="string">'alleles =&gt; [qw(allele1 allele2)]'</span> <span class="variable">would</span> <span class="variable">be</span><span class="operator">:</span>
 
             <span class="variable">GGAT</span><span class="operator">[</span><span class="variable">C</span><span class="operator">/-</span><span class="operator">][</span><span class="variable">C</span><span class="operator">/-</span><span class="operator">]</span><span class="variable">ATT</span><span class="operator">[</span><span class="variable">C</span><span class="operator">/</span><span class="variable">C</span><span class="operator">]</span><span class="variable">CT</span><span class="operator">[</span><span class="variable">A</span><span class="operator">/</span><span class="variable">C</span><span class="operator">]</span><span class="variable">CT</span>
 <span class="variable">Returns</span>   <span class="operator">:</span> <span class="variable">BIC</span><span class="operator">-</span><span class="variable">formatted</span> <span class="variable">string</span>
 <span class="variable">Argument</span>  <span class="operator">:</span> <span class="variable">Required</span> <span class="variable">args</span>
                <span class="variable">refseq</span>    <span class="operator">:</span> <span class="variable">string</span> <span class="operator">(</span><span class="variable">ID</span><span class="operator">)</span> <span class="variable">of</span> <span class="variable">the</span> <span class="variable">reference</span> <span class="variable">sequence</span> <span class="variable">used</span>
                            <span class="variable">as</span> <span class="variable">basis</span> <span class="keyword">for</span> <span class="variable">comparison</span>
                <span class="variable">allele1</span>   <span class="operator">:</span> <span class="variable">string</span> <span class="operator">(</span><span class="variable">ID</span><span class="operator">)</span> <span class="variable">of</span> <span class="variable">the</span> <span class="variable">first</span> <span class="variable">allele</span>
                <span class="variable">allele2</span>   <span class="operator">:</span> <span class="variable">string</span> <span class="operator">(</span><span class="variable">ID</span><span class="operator">)</span> <span class="variable">of</span> <span class="variable">the</span> <span class="variable">second</span> <span class="variable">allele</span>
             <span class="variable">Optional</span> <span class="variable">args</span>
                <span class="variable">delimiters</span><span class="operator">:</span> <span class="variable">two</span> <span class="variable">symbol</span> <span class="variable">string</span> <span class="variable">of</span> <span class="variable">left</span> <span class="keyword">and</span> <span class="variable">right</span> <span class="variable">delimiters</span><span class="operator">.</span>
                            <span class="variable">Only</span> <span class="variable">the</span> <span class="variable">first</span> <span class="variable">two</span> <span class="variable">symbols</span> <span class="variable">are</span> <span class="variable">used</span>
                            <span class="keyword">default</span> <span class="operator">=</span> <span class="string">'[]'</span>
                <span class="variable">separator</span> <span class="operator">:</span> <span class="variable">string</span> <span class="variable">used</span> <span class="variable">as</span> <span class="variable">a</span> <span class="variable">separator</span><span class="operator">.</span>  <span class="variable">Only</span> <span class="variable">the</span> <span class="variable">first</span>
                            <span class="variable">symbol</span> <span class="variable">is</span> <span class="variable">used</span>
                            <span class="keyword">default</span> <span class="operator">=</span> <span class="string">'/'</span>
 <span class="variable">Throws</span>    <span class="operator">:</span> <span class="variable">On</span> <span class="keyword">no</span> <span class="variable">refseq</span><span class="operator">/</span><span class="variable">alleles</span><span class="operator">,</span> <span class="keyword">or</span> <span class="variable">invalid</span> <span class="variable">refseq</span><span class="operator">/</span><span class="variable">alleles</span><span class="operator">.</span>
</code></code></pre>

<h2 id="methods-implementing-Bio::FeatureHolderI">methods implementing Bio::FeatureHolderI</h2>

<p>FeatureHolderI implementation to support labeled character sets like one would get from NEXUS represented data.</p>

<h2 id="get_SeqFeatures">get_SeqFeatures</h2>

<pre><code><code> Usage   : @features = $aln-&gt;get_SeqFeatures
 Function: Get the feature objects held by this feature holder.
 Example :
 Returns : an array of Bio::SeqFeatureI implementing objects
 Args    : optional filter coderef, taking a Bio::SeqFeatureI 
         : as argument, returning TRUE if wanted, FALSE if 
         : unwanted</code></code></pre>

<h2 id="add_SeqFeature">add_SeqFeature</h2>

<pre><code><code> <span class="variable">Usage</span>   <span class="operator">:</span> <span class="variable">$aln</span><span class="operator">-&gt;</span><span class="variable">add_SeqFeature</span><span class="operator">(</span><span class="variable">$subfeat</span><span class="operator">);</span>
 <span class="variable">Function</span><span class="operator">:</span> <span class="variable">adds</span> <span class="variable">a</span> <span class="variable">SeqFeature</span> <span class="variable">into</span> <span class="variable">the</span> <span class="variable">SeqFeature</span> <span class="variable">array</span><span class="operator">.</span>
 <span class="variable">Example</span> <span class="operator">:</span>
 <span class="variable">Returns</span> <span class="operator">:</span> <span class="variable">true</span> <span class="variable">on</span> <span class="variable">success</span>
 <span class="variable">Args</span>    <span class="operator">:</span> <span class="variable">a</span> <span class="variable">Bio::SeqFeatureI</span> <span class="variable">object</span>
 <span class="variable">Note</span>    <span class="operator">:</span> <span class="variable">This</span> <span class="variable">implementation</span> <span class="variable">is</span> <span class="keyword">not</span> <span class="variable">compliant</span>
           <span class="variable">with</span> <span class="variable">Bio::FeatureHolderI</span>
</code></code></pre>

<h2 id="remove_SeqFeatures">remove_SeqFeatures</h2>

<pre><code><code> Usage   : $obj-&gt;remove_SeqFeatures
 Function: Removes all SeqFeatures.  If you want to remove only a subset,
           remove that subset from the returned array, and add back the rest.
 Returns : The array of Bio::SeqFeatureI features that was
           deleted from this alignment.
 Args    : none</code></code></pre>

<h2 id="feature_count">feature_count</h2>

<pre><code><code> Title   : feature_count
 Usage   : $obj-&gt;feature_count()
 Function: Return the number of SeqFeatures attached to the alignment
 Returns : integer representing the number of SeqFeatures
 Args    : None</code></code></pre>

<h2 id="get_all_SeqFeatures">get_all_SeqFeatures</h2>

<pre><code><code> Title   : get_all_SeqFeatures
 Usage   : 
 Function: Get all SeqFeatures.
 Example :
 Returns : an array of Bio::SeqFeatureI implementing objects
 Args    : none
 Note    : Falls through to Bio::FeatureHolderI implementation.</code></code></pre>

<h2 id="methods-for-Bio::AnnotatableI">methods for Bio::AnnotatableI</h2>

<p>AnnotatableI implementation to support sequence alignments which contain annotation (NEXUS, Stockholm).</p>

<h2 id="annotation">annotation</h2>

<pre><code><code> Title   : annotation
 Usage   : $ann = $aln-&gt;annotation or 
           $aln-&gt;annotation($ann)
 Function: Gets or sets the annotation
 Returns : Bio::AnnotationCollectionI object
 Args    : None or Bio::AnnotationCollectionI object</code></code></pre>

<p>See <a href="../../../../.cpan/build/BioPerl-1.6.901-ghKY8U/blib/lib/Bio/AnnotationCollectionI.html">Bio::AnnotationCollectionI</a> and <a href="../../../../.cpan/build/BioPerl-1.6.901-ghKY8U/blib/lib/Bio/Annotation/Collection.html">Bio::Annotation::Collection</a> for more information</p>

<h1 id="Deprecated-methods">Deprecated methods</h1>

<h2 id="no_residues">no_residues</h2>

<pre><code><code> Title     : no_residues
 Usage     : $no = $ali-&gt;no_residues
 Function  : number of residues in total in the alignment
 Returns   : integer
 Argument  :
 Note      : deprecated in favor of num_residues() </code></code></pre>

<h2 id="no_sequences">no_sequences</h2>

<pre><code><code> Title     : no_sequences
 Usage     : $depth = $ali-&gt;no_sequences
 Function  : number of sequence in the sequence alignment
 Returns   : integer
 Argument  :
 Note      : deprecated in favor of num_sequences()</code></code></pre>

<h2 id="mask_columns">mask_columns</h2>

<pre><code><code> Title     : mask_columns
 Usage     : $aln2 = $aln-&gt;mask_columns(20,30)
 Function  : Masks a slice of the alignment inclusive of start and
             end columns, and the first column in the alignment is denoted 1.
             Mask beyond the length of the sequence does not do padding.
 Returns   : A Bio::SimpleAlign object
 Args      : Positive integer for start column, positive integer for end column,
             optional string value use for the mask. Example:

             $aln2 = $aln-&gt;mask_columns(20,30,&#39;?&#39;)
 Note      : Masking must use a character that is not used for gaps or
             frameshifts.  These can be adjusted using the relevant global
             variables, but be aware these may be (uncontrollably) modified
             elsewhere within BioPerl (see bug 2715)</code></code></pre>


</body>

</html>


