<?xml version="1.0" ?>
<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.0 Transitional//EN">
<html>
<head>
<!-- saved from url=(0017)http://localhost/ -->
<script language="JavaScript" src="../../../../displayToc.js"></script>
<script language="JavaScript" src="../../../../tocParas.js"></script>
<script language="JavaScript" src="../../../../tocTab.js"></script>
<link rel="stylesheet" type="text/css" href="../../../../scineplex.css">
<title></title>
<link rel="stylesheet" href="../../../../Active.css" type="text/css" />
<meta http-equiv="content-type" content="text/html; charset=utf-8" />
<link rev="made" href="mailto:support@ActiveState.com" />
</head>

<body>



<ul id="index">
  <li><a href="#NAME">NAME</a></li>
  <li><a href="#SYNOPSIS">SYNOPSIS</a></li>
  <li><a href="#DESCRIPTION">DESCRIPTION</a>
    <ul>
      <li><a href="#The-new-Constructor">The new() Constructor</a></li>
      <li><a href="#Non-SeqFeatureI-methods">Non-SeqFeatureI methods</a></li>
      <li><a href="#display_name">display_name</a></li>
      <li><a href="#accession_number">accession_number</a></li>
      <li><a href="#alphabet">alphabet</a></li>
      <li><a href="#desc">desc</a></li>
      <li><a href="#location">location</a></li>
      <li><a href="#location_string">location_string</a></li>
      <li><a href="#clone">clone</a></li>
      <li><a href="#refseq">refseq</a></li>
    </ul>
  </li>
  <li><a href="#SEE-ALSO">SEE ALSO</a></li>
  <li><a href="#AUTHOR">AUTHOR</a></li>
</ul>

<h1 id="NAME">NAME</h1>

<p>Bio::SeqFeature::Lite - Lightweight Bio::SeqFeatureI class</p>

<h1 id="SYNOPSIS">SYNOPSIS</h1>

<pre><code><code> <span class="comment"># create a simple feature with no internal structure</span>
 <span class="variable">$f</span> <span class="operator">=</span> <span class="variable">Bio::SeqFeature::Lite</span><span class="operator">-&gt;</span><span class="variable">new</span><span class="operator">(</span><span class="string">-start</span> <span class="operator">=&gt;</span> <span class="number">1000</span><span class="operator">,</span>
                                  <span class="string">-stop</span>  <span class="operator">=&gt;</span> <span class="number">2000</span><span class="operator">,</span>
                                  <span class="string">-type</span>  <span class="operator">=&gt;</span> <span class="string">'transcript'</span><span class="operator">,</span>
                                  <span class="string">-name</span>  <span class="operator">=&gt;</span> <span class="string">'alpha-1 antitrypsin'</span><span class="operator">,</span>
                                  <span class="string">-desc</span>  <span class="operator">=&gt;</span> <span class="string">'an enzyme inhibitor'</span><span class="operator">,</span>
                                 <span class="operator">);</span>
 
 <span class="comment"># create a feature composed of multiple segments, all of type "similarity"</span>
 <span class="variable">$f</span> <span class="operator">=</span> <span class="variable">Bio::SeqFeature::Lite</span><span class="operator">-&gt;</span><span class="variable">new</span><span class="operator">(</span><span class="string">-segments</span> <span class="operator">=&gt;</span> <span class="operator">[[</span><span class="number">1000</span><span class="operator">,</span><span class="number">1100</span><span class="operator">]</span><span class="operator">,</span><span class="operator">[</span><span class="number">1500</span><span class="operator">,</span><span class="number">1550</span><span class="operator">]</span><span class="operator">,</span><span class="operator">[</span><span class="number">1800</span><span class="operator">,</span><span class="number">2000</span><span class="operator">]]</span><span class="operator">,</span>
                                  <span class="string">-name</span>     <span class="operator">=&gt;</span> <span class="string">'ABC-3'</span><span class="operator">,</span>
                                  <span class="string">-type</span>     <span class="operator">=&gt;</span> <span class="string">'gapped_alignment'</span><span class="operator">,</span>
                                  <span class="string">-subtype</span>  <span class="operator">=&gt;</span> <span class="string">'similarity'</span><span class="operator">);</span>
 
 <span class="comment"># build up a gene exon by exon</span>
 <span class="variable">$e1</span> <span class="operator">=</span> <span class="variable">Bio::SeqFeature::Lite</span><span class="operator">-&gt;</span><span class="variable">new</span><span class="operator">(</span><span class="string">-start</span><span class="operator">=&gt;</span><span class="number">1</span><span class="operator">,</span><span class="string">-stop</span><span class="operator">=&gt;</span><span class="number">100</span><span class="operator">,</span><span class="string">-type</span><span class="operator">=&gt;</span><span class="string">'exon'</span><span class="operator">);</span>
 <span class="variable">$e2</span> <span class="operator">=</span> <span class="variable">Bio::SeqFeature::Lite</span><span class="operator">-&gt;</span><span class="variable">new</span><span class="operator">(</span><span class="string">-start</span><span class="operator">=&gt;</span><span class="number">150</span><span class="operator">,</span><span class="string">-stop</span><span class="operator">=&gt;</span><span class="number">200</span><span class="operator">,</span><span class="string">-type</span><span class="operator">=&gt;</span><span class="string">'exon'</span><span class="operator">);</span>
 <span class="variable">$e3</span> <span class="operator">=</span> <span class="variable">Bio::SeqFeature::Lite</span><span class="operator">-&gt;</span><span class="variable">new</span><span class="operator">(</span><span class="string">-start</span><span class="operator">=&gt;</span><span class="number">300</span><span class="operator">,</span><span class="string">-stop</span><span class="operator">=&gt;</span><span class="number">500</span><span class="operator">,</span><span class="string">-type</span><span class="operator">=&gt;</span><span class="string">'exon'</span><span class="operator">);</span>
 <span class="variable">$f</span>  <span class="operator">=</span> <span class="variable">Bio::SeqFeature::Lite</span><span class="operator">-&gt;</span><span class="variable">new</span><span class="operator">(</span><span class="string">-segments</span><span class="operator">=&gt;</span><span class="operator">[</span><span class="variable">$e1</span><span class="operator">,</span><span class="variable">$e2</span><span class="operator">,</span><span class="variable">$e3</span><span class="operator">]</span><span class="operator">,</span><span class="string">-type</span><span class="operator">=&gt;</span><span class="string">'gene'</span><span class="operator">);</span>
</code></code></pre>

<h1 id="DESCRIPTION">DESCRIPTION</h1>

<p>This is a simple Bio::SeqFeatureI-compliant object that is compatible with Bio::Graphics::Panel. With it you can create lightweight feature objects for drawing.</p>

<p>All methods are as described in <a href="../../../../../.cpan/build/BioPerl-1.6.901-ghKY8U/blib/lib/Bio/SeqFeatureI.html">Bio::SeqFeatureI</a> with the following additions:</p>

<h2 id="The-new-Constructor">The new() Constructor</h2>

<pre><code><code> <span class="variable">$feature</span> <span class="operator">=</span> <span class="variable">Bio::SeqFeature::Lite</span><span class="operator">-&gt;</span><span class="variable">new</span><span class="operator">(</span><span class="variable">@args</span><span class="operator">);</span>
</code></code></pre>

<p>This method creates a new feature object. You can create a simple feature that contains no subfeatures, or a hierarchically nested object.</p>

<p>Arguments are as follows:</p>

<pre><code><code>  -seq_id      the reference sequence
  -start       the start position of the feature
  -end         the stop position of the feature
  -stop        an alias for end
  -name        the feature name (returned by seqname())
  -type        the feature type (returned by primary_tag())
  -primary_tag the same as -type
  -source      the source tag
  -score       the feature score (for GFF compatibility)
  -desc        a description of the feature
  -segments    a list of subfeatures (see below)
  -subtype     the type to use when creating subfeatures
  -strand      the strand of the feature (one of -1, 0 or +1)
  -phase       the phase of the feature (0..2)
  -seq         a dna or protein sequence string to attach to feature
  -id          an alias for -name
  -seqname     an alias for -name
  -display_id  an alias for -name
  -display_name an alias for -name  (do you get the idea the API has changed?)
  -primary_id  unique database ID
  -url         a URL to link to when rendered with Bio::Graphics
  -attributes  a hashref of tag value attributes, in which the key is the tag
               and the value is an array reference of values
  -factory     a reference to a feature factory, used for compatibility with
               more obscure parts of Bio::DB::GFF</code></code></pre>

<p>The subfeatures passed in -segments may be an array of Bio::SeqFeature::Lite objects, or an array of [$start,$stop] pairs. Each pair should be a two-element array reference. In the latter case, the feature type passed in -subtype will be used when creating the subfeatures.</p>

<p>If no feature type is passed, then it defaults to &quot;feature&quot;.</p>

<h2 id="Non-SeqFeatureI-methods">Non-SeqFeatureI methods</h2>

<p>A number of new methods are provided for compatibility with Ace::Sequence, which has a slightly different API from SeqFeatureI:</p>

<dl>

<dt id="url-">url()</dt>
<dd>

<p>Get/set the URL that the graphical rendering of this feature will link to.</p>

</dd>
<dt id="add_segment-segments-">add_segment(@segments)</dt>
<dd>

<p>Add one or more segments (a subfeature). Segments can either be Feature objects, or [start,stop] arrays, as in the -segments argument to new(). The feature endpoints are automatically adjusted.</p>

</dd>
<dt id="segments-">segments()</dt>
<dd>

<p>An alias for sub_SeqFeature().</p>

</dd>
<dt id="get_SeqFeatures-">get_SeqFeatures()</dt>
<dd>

<p>Alias for sub_SeqFeature()</p>

</dd>
<dt id="get_all_SeqFeatures-">get_all_SeqFeatures()</dt>
<dd>

<p>Alias for sub_SeqFeature()</p>

</dd>
<dt id="merged_segments-">merged_segments()</dt>
<dd>

<p>Another alias for sub_SeqFeature().</p>

</dd>
<dt id="stop-">stop()</dt>
<dd>

<p>An alias for end().</p>

</dd>
<dt id="name-">name()</dt>
<dd>

<p>An alias for seqname().</p>

</dd>
<dt id="exons-">exons()</dt>
<dd>

<p>An alias for sub_SeqFeature() (you don&#39;t want to know why!)</p>

</dd>
</dl>

<h2 id="display_name">display_name</h2>

<pre><code><code> <span class="variable">Title</span>   <span class="operator">:</span> <span class="variable">display_name</span>
 <span class="variable">Usage</span>   <span class="operator">:</span> <span class="variable">$id</span> <span class="operator">=</span> <span class="variable">$obj</span><span class="operator">-&gt;</span><span class="variable">display_name</span> <span class="keyword">or</span> <span class="variable">$obj</span><span class="operator">-&gt;</span><span class="variable">display_name</span><span class="operator">(</span><span class="variable">$newid</span><span class="operator">);</span>
 <span class="variable">Function</span><span class="operator">:</span> <span class="variable">Gets</span> <span class="keyword">or</span> <span class="variable">sets</span> <span class="variable">the</span> <span class="variable">display</span> <span class="variable">id</span><span class="operator">,</span> <span class="variable">also</span> <span class="variable">known</span> <span class="variable">as</span> <span class="variable">the</span> <span class="variable">common</span> <span class="variable">name</span> <span class="variable">of</span>
           <span class="variable">the</span> <span class="variable">Seq</span> <span class="variable">object</span><span class="operator">.</span>
 
           <span class="variable">The</span> <span class="variable">semantics</span> <span class="variable">of</span> <span class="variable">this</span> <span class="variable">is</span> <span class="variable">that</span> <span class="variable">it</span> <span class="variable">is</span> <span class="variable">the</span> <span class="variable">most</span> <span class="variable">likely</span> <span class="variable">string</span>
           <span class="variable">to</span> <span class="variable">be</span> <span class="variable">used</span> <span class="variable">as</span> <span class="variable">an</span> <span class="variable">identifier</span> <span class="variable">of</span> <span class="variable">the</span> <span class="variable">sequence</span><span class="operator">,</span> <span class="keyword">and</span> <span class="variable">likely</span> <span class="variable">to</span>
           <span class="variable">have</span> <span class="string">"human"</span> <span class="variable">readability</span><span class="operator">.</span>  <span class="variable">The</span> <span class="variable">id</span> <span class="variable">is</span> <span class="variable">equivalent</span> <span class="variable">to</span> <span class="variable">the</span> <span class="variable">LOCUS</span>
           <span class="variable">field</span> <span class="variable">of</span> <span class="variable">the</span> <span class="variable">GenBank</span><span class="operator">/</span><span class="variable">EMBL</span> <span class="variable">databanks</span> <span class="keyword">and</span> <span class="variable">the</span> <span class="variable">ID</span> <span class="variable">field</span> <span class="variable">of</span> <span class="variable">the</span>
           <span class="variable">Swissprot</span><span class="operator">/</span><span class="variable">sptrembl</span> <span class="variable">database</span><span class="operator">.</span> <span class="variable">In</span> <span class="variable">fasta</span> <span class="keyword">format</span><span class="operator">,</span> <span class="variable">the</span> <span class="operator">&gt;(\</span><span class="variable">S</span><span class="operator">+)</span> <span class="variable">is</span>
           <span class="variable">presumed</span> <span class="variable">to</span> <span class="variable">be</span> <span class="variable">the</span> <span class="variable">id</span><span class="operator">,</span> <span class="variable">though</span> <span class="variable">some</span> <span class="variable">people</span> <span class="variable">overload</span> <span class="variable">the</span> <span class="variable">id</span>
           <span class="variable">to</span> <span class="variable">embed</span> <span class="variable">other</span> <span class="variable">information</span><span class="operator">.</span> <span class="variable">Bioperl</span> <span class="variable">does</span> <span class="keyword">not</span> <span class="keyword">use</span> <span class="variable">any</span>
           <span class="variable">embedded</span> <span class="variable">information</span> <span class="variable">in</span> <span class="variable">the</span> <span class="variable">ID</span> <span class="variable">field</span><span class="operator">,</span> <span class="keyword">and</span> <span class="variable">people</span> <span class="variable">are</span>
           <span class="variable">encouraged</span> <span class="variable">to</span> <span class="keyword">use</span> <span class="variable">other</span> <span class="variable">mechanisms</span> <span class="operator">(</span><span class="variable">accession</span> <span class="variable">field</span> <span class="keyword">for</span>
           <span class="variable">example</span><span class="operator">,</span> <span class="keyword">or</span> <span class="variable">extending</span> <span class="variable">the</span> <span class="variable">sequence</span> <span class="variable">object</span><span class="operator">)</span> <span class="variable">to</span> <span class="variable">solve</span> <span class="variable">this</span><span class="operator">.</span>
 
           <span class="variable">Notice</span> <span class="variable">that</span> <span class="variable">$seq</span><span class="operator">-&gt;</span><span class="variable">id</span><span class="operator">()</span> <span class="variable">maps</span> <span class="variable">to</span> <span class="variable">this</span> <span class="variable">function</span><span class="operator">,</span> <span class="variable">mainly</span> <span class="keyword">for</span>
           <span class="variable">legacy</span><span class="operator">/</span><span class="variable">convenience</span> <span class="variable">issues</span><span class="operator">.</span>
 <span class="variable">Returns</span> <span class="operator">:</span> <span class="variable">A</span> <span class="variable">string</span>
 <span class="variable">Args</span>    <span class="operator">:</span> <span class="variable">None</span> <span class="keyword">or</span> <span class="variable">a</span> <span class="variable">new</span> <span class="variable">id</span>
</code></code></pre>

<h2 id="accession_number">accession_number</h2>

<pre><code><code> <span class="variable">Title</span>   <span class="operator">:</span> <span class="variable">accession_number</span>
 <span class="variable">Usage</span>   <span class="operator">:</span> <span class="variable">$unique_biological_key</span> <span class="operator">=</span> <span class="variable">$obj</span><span class="operator">-&gt;</span><span class="variable">accession_number</span><span class="operator">;</span>
 <span class="variable">Function</span><span class="operator">:</span> <span class="variable">Returns</span> <span class="variable">the</span> <span class="variable">unique</span> <span class="variable">biological</span> <span class="variable">id</span> <span class="keyword">for</span> <span class="variable">a</span> <span class="variable">sequence</span><span class="operator">,</span> <span class="variable">commonly</span>
           <span class="variable">called</span> <span class="variable">the</span> <span class="variable">accession_number</span><span class="operator">.</span> <span class="variable">For</span> <span class="variable">sequences</span> <span class="variable">from</span> <span class="variable">established</span>
           <span class="variable">databases</span><span class="operator">,</span> <span class="variable">the</span> <span class="variable">implementors</span> <span class="variable">should</span> <span class="variable">try</span> <span class="variable">to</span> <span class="keyword">use</span> <span class="variable">the</span> <span class="variable">correct</span>
           <span class="variable">accession</span> <span class="variable">number</span><span class="operator">.</span> <span class="variable">Notice</span> <span class="variable">that</span> <span class="variable">primary_id</span><span class="operator">()</span> <span class="variable">provides</span> <span class="variable">the</span>
           <span class="variable">unique</span> <span class="variable">id</span> <span class="keyword">for</span> <span class="variable">the</span> <span class="variable">implemetation</span><span class="operator">,</span> <span class="variable">allowing</span> <span class="variable">multiple</span> <span class="variable">objects</span>
           <span class="variable">to</span> <span class="variable">have</span> <span class="variable">the</span> <span class="variable">same</span> <span class="variable">accession</span> <span class="variable">number</span> <span class="variable">in</span> <span class="variable">a</span> <span class="variable">particular</span> <span class="variable">implementation</span><span class="operator">.</span>
 
           <span class="variable">For</span> <span class="variable">sequences</span> <span class="variable">with</span> <span class="keyword">no</span> <span class="variable">accession</span> <span class="variable">number</span><span class="operator">,</span> <span class="variable">this</span> <span class="variable">method</span> <span class="variable">should</span> <span class="keyword">return</span>
           <span class="string">"unknown"</span><span class="operator">.</span>
 <span class="variable">Returns</span> <span class="operator">:</span> <span class="variable">A</span> <span class="variable">string</span>
 <span class="variable">Args</span>    <span class="operator">:</span> <span class="variable">None</span>
</code></code></pre>

<h2 id="alphabet">alphabet</h2>

<pre><code><code> <span class="variable">Title</span>   <span class="operator">:</span> <span class="variable">alphabet</span>
 <span class="variable">Usage</span>   <span class="operator">:</span> <span class="keyword">if</span><span class="operator">(</span> <span class="variable">$obj</span><span class="operator">-&gt;</span><span class="variable">alphabet</span> <span class="keyword">eq</span> <span class="string">'dna'</span> <span class="operator">)</span> <span class="operator">{</span> <span class="regex">/Do Something/</span> <span class="operator">}</span>
 <span class="variable">Function</span><span class="operator">:</span> <span class="variable">Returns</span> <span class="variable">the</span> <span class="variable">type</span> <span class="variable">of</span> <span class="variable">sequence</span> <span class="variable">being</span> <span class="variable">one</span> <span class="variable">of</span>
           <span class="string">'dna'</span><span class="operator">,</span> <span class="string">'rna'</span> <span class="keyword">or</span> <span class="string">'protein'</span><span class="operator">.</span> <span class="variable">This</span> <span class="variable">is</span> <span class="variable">case</span> <span class="variable">sensitive</span><span class="operator">.</span>
 
           <span class="variable">This</span> <span class="variable">is</span> <span class="keyword">not</span> <span class="variable">called</span> <span class="operator">&lt;</span><span class="variable">type</span><span class="operator">&gt;</span> <span class="variable">because</span> <span class="variable">this</span> <span class="variable">would</span> <span class="variable">cause</span>
           <span class="variable">upgrade</span> <span class="variable">problems</span> <span class="variable">from</span> <span class="variable">the</span> <span class="number">0</span><span class="operator">.</span><span class="number">5</span> <span class="keyword">and</span> <span class="variable">earlier</span> <span class="variable">Seq</span> <span class="variable">objects</span><span class="operator">.</span>
 
 <span class="variable">Returns</span> <span class="operator">:</span> <span class="variable">a</span> <span class="variable">string</span> <span class="variable">either</span> <span class="string">'dna'</span><span class="operator">,</span><span class="string">'rna'</span><span class="operator">,</span><span class="string">'protein'</span><span class="operator">.</span> <span class="variable">NB</span> <span class="operator">-</span> <span class="variable">the</span> <span class="variable">object</span> <span class="variable">must</span>
           <span class="variable">make</span> <span class="variable">a</span> <span class="variable">call</span> <span class="variable">of</span> <span class="variable">the</span> <span class="variable">type</span> <span class="operator">-</span> <span class="keyword">if</span> <span class="variable">there</span> <span class="variable">is</span> <span class="keyword">no</span> <span class="variable">type</span> <span class="variable">specified</span> <span class="variable">it</span>
           <span class="variable">has</span> <span class="variable">to</span> <span class="variable">guess</span><span class="operator">.</span>
 <span class="variable">Args</span>    <span class="operator">:</span> <span class="variable">none</span>
 <span class="variable">Status</span>  <span class="operator">:</span> <span class="variable">Virtual</span>
</code></code></pre>

<h2 id="desc">desc</h2>

<pre><code><code> Title   : desc
 Usage   : $seqobj-&gt;desc($string) or $seqobj-&gt;desc()
 Function: Sets or gets the description of the sequence
 Example :
 Returns : The description
 Args    : The description or none</code></code></pre>

<h2 id="location">location</h2>

<pre><code><code> Title   : location
 Usage   : my $location = $seqfeature-&gt;location()
 Function: returns a location object suitable for identifying location
           of feature on sequence or parent feature
 Returns : Bio::LocationI object
 Args    : none</code></code></pre>

<h2 id="location_string">location_string</h2>

<pre><code><code> Title   : location_string
 Usage   : my $string = $seqfeature-&gt;location_string()
 Function: Returns a location string in a format recognized by gbrowse
 Returns : a string
 Args    : none</code></code></pre>

<p>This is a convenience function used by the generic genome browser. It returns the location of the feature and its subfeatures in the compact form &quot;start1..end1,start2..end2,...&quot;. Use $seqfeature-&gt;location()-&gt;toFTString() to obtain a standard GenBank/EMBL location representation.</p>

<h2 id="clone">clone</h2>

<pre><code><code> Title   : clone
 Usage   : my $feature = $seqfeature-&gt;clone
 Function: Create a deep copy of the feature
 Returns : A copy of the feature
 Args    : none</code></code></pre>

<h2 id="refseq">refseq</h2>

<pre><code><code> Title   : refseq
 Usage   : $ref = $s-&gt;refseq([$newseq] [,$newseqclass])
 Function: get/set reference sequence
 Returns : current reference sequence
 Args    : new reference sequence and class (optional)
 Status  : Public</code></code></pre>

<p>This method will get or set the reference sequence. Called with no arguments, it returns the current reference sequence. Called with any Bio::SeqFeatureI object that provides the seq_id(), start(), end() and strand() methods.</p>

<p>The method will generate an exception if you attempt to set the reference sequence to a sequence that has a different seq_id from the current feature.</p>

<h1 id="SEE-ALSO">SEE ALSO</h1>

<p><a>Bio::Graphics::Feature</a></p>

<h1 id="AUTHOR">AUTHOR</h1>

<p>Lincoln Stein &lt;lstein@cshl.edu&gt;.</p>

<p>Copyright (c) 2006 Cold Spring Harbor Laboratory</p>

<p>This library is free software; you can redistribute it and/or modify it under the same terms as Perl itself. See DISCLAIMER.txt for disclaimers of warranty.</p>


</body>

</html>


