<?xml version="1.0" ?>
<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.0 Transitional//EN">
<html>
<head>
<!-- saved from url=(0017)http://localhost/ -->
<script language="JavaScript" src="../../../../displayToc.js"></script>
<script language="JavaScript" src="../../../../tocParas.js"></script>
<script language="JavaScript" src="../../../../tocTab.js"></script>
<link rel="stylesheet" type="text/css" href="../../../../scineplex.css">
<title></title>
<link rel="stylesheet" href="../../../../Active.css" type="text/css" />
<meta http-equiv="content-type" content="text/html; charset=utf-8" />
<link rev="made" href="mailto:support@ActiveState.com" />
</head>

<body>



<ul id="index">
  <li><a href="#NAME">NAME</a></li>
  <li><a href="#SYNOPSIS">SYNOPSIS</a></li>
  <li><a href="#DESCRIPTION">DESCRIPTION</a>
    <ul>
      <li><a href="#Coordinate_systems">Coordinate_systems</a></li>
      <li><a href="#Feature_collection">Feature_collection</a></li>
    </ul>
  </li>
  <li><a href="#FEEDBACK">FEEDBACK</a>
    <ul>
      <li><a href="#Mailing-Lists">Mailing Lists</a></li>
      <li><a href="#Support">Support</a></li>
      <li><a href="#Reporting-Bugs">Reporting Bugs</a></li>
    </ul>
  </li>
  <li><a href="#AUTHOR---Robson-Francisco-de-Souza">AUTHOR - Robson Francisco de Souza</a></li>
  <li><a href="#APPENDIX">APPENDIX</a></li>
  <li><a href="#Object-creator">Object creator</a>
    <ul>
      <li><a href="#new">new</a></li>
    </ul>
  </li>
  <li><a href="#Assembly-related-methods">Assembly related methods</a>
    <ul>
      <li><a href="#source">source</a></li>
      <li><a href="#assembly">assembly</a></li>
      <li><a href="#strand">strand</a></li>
      <li><a href="#upstream_neighbor">upstream_neighbor</a></li>
      <li><a href="#downstream_neighbor">downstream_neighbor</a></li>
    </ul>
  </li>
  <li><a href="#Contig-feature-collection-methods">Contig feature collection methods</a>
    <ul>
      <li><a href="#add_features">add_features</a></li>
      <li><a href="#remove_features">remove_features</a></li>
      <li><a href="#get_features_collection">get_features_collection</a></li>
      <li><a href="#remove_features_collection">remove_features_collection</a></li>
    </ul>
  </li>
  <li><a href="#Coordinate-systems-related-methods">Coordinate system&#39;s related methods</a>
    <ul>
      <li><a href="#change_coord">change_coord</a></li>
      <li><a href="#get_seq_coord">get_seq_coord</a></li>
      <li><a href="#set_seq_coord">set_seq_coord</a></li>
    </ul>
  </li>
  <li><a href="#Bio::Assembly::Contig-consensus-methods">Bio::Assembly::Contig consensus methods</a>
    <ul>
      <li><a href="#set_consensus_sequence">set_consensus_sequence</a></li>
      <li><a href="#set_consensus_quality">set_consensus_quality</a></li>
      <li><a href="#get_consensus_length">get_consensus_length</a></li>
      <li><a href="#get_consensus_sequence">get_consensus_sequence</a></li>
      <li><a href="#get_consensus_quality">get_consensus_quality</a></li>
    </ul>
  </li>
  <li><a href="#Bio::Assembly::Contig-aligned-sequences-methods">Bio::Assembly::Contig aligned sequences methods</a>
    <ul>
      <li><a href="#set_seq_qual">set_seq_qual</a></li>
      <li><a href="#get_seq_ids">get_seq_ids</a></li>
      <li><a href="#get_seq_feat_by_tag">get_seq_feat_by_tag</a></li>
      <li><a href="#get_seq_by_name">get_seq_by_name</a></li>
      <li><a href="#get_qual_by_name">get_qual_by_name</a></li>
    </ul>
  </li>
  <li><a href="#Bio::Align::AlignI-compatible-methods">Bio::Align::AlignI compatible methods</a>
    <ul>
      <li><a href="#Modifier-methods">Modifier methods</a></li>
      <li><a href="#add_seq">add_seq</a></li>
      <li><a href="#remove_seq">remove_seq</a></li>
      <li><a href="#purge">purge</a></li>
      <li><a href="#sort_alphabetically">sort_alphabetically</a></li>
      <li><a href="#Sequence-selection-methods">Sequence selection methods</a></li>
      <li><a href="#each_seq">each_seq</a></li>
      <li><a href="#each_alphabetically">each_alphabetically</a></li>
      <li><a href="#each_seq_with_id">each_seq_with_id</a></li>
      <li><a href="#get_seq_by_pos">get_seq_by_pos</a></li>
      <li><a href="#Create-new-alignments">Create new alignments</a></li>
      <li><a href="#select">select</a></li>
      <li><a href="#select_noncont">select_noncont</a></li>
      <li><a href="#slice">slice</a></li>
      <li><a href="#Change-sequences-within-the-MSE">Change sequences within the MSE</a></li>
      <li><a href="#map_chars">map_chars</a></li>
      <li><a href="#uppercase">uppercase</a></li>
      <li><a href="#match_line">match_line</a></li>
      <li><a href="#match">match</a></li>
      <li><a href="#unmatch">unmatch</a></li>
      <li><a href="#MSE-attibutes">MSE attibutes</a></li>
      <li><a href="#id">id</a></li>
      <li><a href="#missing_char">missing_char</a></li>
      <li><a href="#match_char">match_char</a></li>
      <li><a href="#gap_char">gap_char</a></li>
      <li><a href="#symbol_chars">symbol_chars</a></li>
      <li><a href="#Alignment-descriptors">Alignment descriptors</a></li>
      <li><a href="#consensus_string">consensus_string</a></li>
      <li><a href="#consensus_iupac">consensus_iupac</a></li>
      <li><a href="#is_flush">is_flush</a></li>
      <li><a href="#length">length</a></li>
      <li><a href="#maxname_length">maxname_length</a></li>
      <li><a href="#num_residues">num_residues</a></li>
      <li><a href="#num_sequences">num_sequences</a></li>
      <li><a href="#percentage_identity">percentage_identity</a></li>
      <li><a href="#overall_percentage_identity">overall_percentage_identity</a></li>
      <li><a href="#average_percentage_identity">average_percentage_identity</a></li>
      <li><a href="#Alignment-positions">Alignment positions</a></li>
      <li><a href="#column_from_residue_number">column_from_residue_number</a></li>
      <li><a href="#Sequence-names">Sequence names</a></li>
      <li><a href="#displayname">displayname</a></li>
      <li><a href="#set_displayname_count">set_displayname_count</a></li>
      <li><a href="#set_displayname_flat">set_displayname_flat</a></li>
      <li><a href="#set_displayname_normal">set_displayname_normal</a></li>
    </ul>
  </li>
  <li><a href="#Internal-Methods">Internal Methods</a>
    <ul>
      <li><a href="#binary_search">_binary_search</a></li>
      <li><a href="#compare">_compare</a></li>
      <li><a href="#nof_gaps">_nof_gaps</a></li>
      <li><a href="#padded_unpadded">_padded_unpadded</a></li>
      <li><a href="#unpadded_padded">_unpadded_padded</a></li>
      <li><a href="#register_gaps">_register_gaps</a></li>
    </ul>
  </li>
  <li><a href="#Deprecated-methods">Deprecated methods</a>
    <ul>
      <li><a href="#no_residues">no_residues</a></li>
      <li><a href="#no_sequences">no_sequences</a></li>
    </ul>
  </li>
</ul>

<h1 id="NAME">NAME</h1>

<p>Bio::Assembly::Contig - Perl module to hold and manipulate sequence assembly contigs.</p>

<h1 id="SYNOPSIS">SYNOPSIS</h1>

<pre><code><code>    <span class="comment"># Module loading</span>
    <span class="keyword">use</span> <span class="variable">Bio::Assembly::IO</span><span class="operator">;</span>
    
    <span class="comment"># Assembly loading methods</span>
    <span class="variable">$aio</span> <span class="operator">=</span> <span class="variable">Bio::Assembly::IO</span><span class="operator">-&gt;</span><span class="variable">new</span><span class="operator">(</span><span class="string">-file</span><span class="operator">=&gt;</span><span class="string">"test.ace.1"</span><span class="operator">,</span>
                                  <span class="string">-format</span><span class="operator">=&gt;</span><span class="string">'phrap'</span><span class="operator">);</span>
    
    <span class="variable">$assembly</span> <span class="operator">=</span> <span class="variable">$aio</span><span class="operator">-&gt;</span><span class="variable">next_assembly</span><span class="operator">;</span>
    <span class="keyword">foreach</span> <span class="variable">$contig</span> <span class="operator">(</span><span class="variable">$assembly</span><span class="operator">-&gt;</span><span class="variable">all_contigs</span><span class="operator">)</span> <span class="operator">{</span>
      <span class="comment"># do something</span>
    <span class="operator">}</span>
    
    <span class="comment"># OR, if you want to build the contig yourself,</span>
    
    <span class="keyword">use</span> <span class="variable">Bio::Assembly::Contig</span><span class="operator">;</span>
    <span class="variable">$c</span> <span class="operator">=</span> <span class="variable">Bio::Assembly::Contig</span><span class="operator">-&gt;</span><span class="variable">new</span><span class="operator">(</span><span class="string">-id</span><span class="operator">=&gt;</span><span class="string">"1"</span><span class="operator">);</span>
    
    <span class="variable">$ls</span>  <span class="operator">=</span> <span class="variable">Bio::LocatableSeq</span><span class="operator">-&gt;</span><span class="variable">new</span><span class="operator">(</span><span class="string">-seq</span><span class="operator">=&gt;</span><span class="string">"ACCG-T"</span><span class="operator">,</span>
                                  <span class="string">-id</span><span class="operator">=&gt;</span><span class="string">"r1"</span><span class="operator">,</span>
                                  <span class="string">-alphabet</span><span class="operator">=&gt;</span><span class="string">'dna'</span><span class="operator">);</span>
    <span class="variable">$ls2</span> <span class="operator">=</span> <span class="variable">Bio::LocatableSeq</span><span class="operator">-&gt;</span><span class="variable">new</span><span class="operator">(</span><span class="string">-seq</span><span class="operator">=&gt;</span><span class="string">"ACA-CG-T"</span><span class="operator">,</span>
                                  <span class="string">-id</span><span class="operator">=&gt;</span><span class="string">"r2"</span><span class="operator">,</span>
                                  <span class="string">-alphabet</span><span class="operator">=&gt;</span><span class="string">'dna'</span><span class="operator">);</span>
    
    <span class="variable">$ls_coord</span> <span class="operator">=</span> <span class="variable">Bio::SeqFeature::Generic</span><span class="operator">-&gt;</span><span class="variable">new</span><span class="operator">(</span><span class="string">-start</span><span class="operator">=&gt;</span><span class="number">3</span><span class="operator">,</span>
                                              <span class="string">-end</span><span class="operator">=&gt;</span><span class="number">8</span><span class="operator">,</span>
                                              <span class="string">-strand</span><span class="operator">=&gt;</span><span class="number">1</span><span class="operator">);</span>
    <span class="variable">$ls2_coord</span> <span class="operator">=</span> <span class="variable">Bio::SeqFeature::Generic</span><span class="operator">-&gt;</span><span class="variable">new</span><span class="operator">(</span><span class="string">-start</span><span class="operator">=&gt;</span><span class="number">1</span><span class="operator">,</span>
                                               <span class="string">-end</span><span class="operator">=&gt;</span><span class="number">8</span><span class="operator">,</span>
                                               <span class="string">-strand</span><span class="operator">=&gt;</span><span class="number">1</span><span class="operator">);</span>
    <span class="variable">$c</span><span class="operator">-&gt;</span><span class="variable">add_seq</span><span class="operator">(</span><span class="variable">$ls</span><span class="operator">);</span>
    <span class="variable">$c</span><span class="operator">-&gt;</span><span class="variable">add_seq</span><span class="operator">(</span><span class="variable">$ls2</span><span class="operator">);</span>
    <span class="variable">$c</span><span class="operator">-&gt;</span><span class="variable">set_seq_coord</span><span class="operator">(</span><span class="variable">$ls_coord</span><span class="operator">,</span><span class="variable">$ls</span><span class="operator">);</span>
    <span class="variable">$c</span><span class="operator">-&gt;</span><span class="variable">set_seq_coord</span><span class="operator">(</span><span class="variable">$ls2_coord</span><span class="operator">,</span><span class="variable">$ls2</span><span class="operator">);</span>
    
    <span class="variable">$con</span> <span class="operator">=</span> <span class="variable">Bio::LocatableSeq</span><span class="operator">-&gt;</span><span class="variable">new</span><span class="operator">(</span><span class="string">-seq</span><span class="operator">=&gt;</span><span class="string">"ACACCG-T"</span><span class="operator">,</span>
                                  <span class="string">-alphabet</span><span class="operator">=&gt;</span><span class="string">'dna'</span><span class="operator">);</span>
    <span class="variable">$c</span><span class="operator">-&gt;</span><span class="variable">set_consensus_sequence</span><span class="operator">(</span><span class="variable">$con</span><span class="operator">);</span>
    
    <span class="variable">$l</span> <span class="operator">=</span> <span class="variable">$c</span><span class="operator">-&gt;</span><span class="variable">change_coord</span><span class="operator">(</span><span class="string">'unaligned r2'</span><span class="operator">,</span><span class="string">'ungapped consensus'</span><span class="operator">,</span><span class="number">6</span><span class="operator">);</span>
    <span class="keyword">print</span> <span class="string">"6 in unaligned r2 =&gt; </span><span class="variable">$l</span><span class="string"> in ungapped consensus\n"</span><span class="operator">;</span>
</code></code></pre>

<h1 id="DESCRIPTION">DESCRIPTION</h1>

<p>A contig is as a set of sequences, locally aligned to each other, so that every sequence has overlapping regions with at least one sequence in the contig, such that a continuous of overlapping sequences is formed, allowing the deduction of a consensus sequence which may be longer than any of the sequences from which it was deduced.</p>

<p>In this documentation we refer to the overlapping sequences used to build the contig as &quot;aligned sequences&quot; and to the sequence deduced from the overlap of aligned sequences as the &quot;consensus&quot;. Methods to deduce the consensus sequence from aligned sequences were not yet implemented in this module, but its posssible to add a consensus sequence deduced by other means, e.g, by the assembly program used to build the alignment.</p>

<p>All aligned sequences in a Bio::Assembly::Contig must be Bio::Assembly::Locatable objects and have a unique ID. The unique ID restriction is due to the nature of the module&#39;s internal data structures and is also a request of some assembly programs. If two sequences with the same ID are added to a contig, the first sequence added is replaced by the second one.</p>

<h2 id="Coordinate_systems">Coordinate_systems</h2>

<p>There are four base coordinate systems in Bio::Assembly::Contig. When you need to access contig elements or data that exists on a certain range or location, you may be specifying coordinates in relation to different sequences, which may be either the contig consensus or one of the aligned sequences that were used to do the assembly.</p>

<pre><code><code> =========================================================
          Name           | Referenced sequence
 ---------------------------------------------------------
   &quot;gapped consensus&quot;    | Contig (with gaps)
   &quot;ungapped consensus&quot;  | Contig (without gaps)
   &quot;aligned $seqID&quot;      | sequence $seqID (with gaps)
   &quot;unaligned $seqID&quot;    | sequence $seqID (without gaps)
 =========================================================</code></code></pre>

<p>&quot;gapped consensus&quot; refers to positions in the aligned consensus sequence, which is the consensus sequence including the gaps inserted to align it agains the aligned sequences that were used to assemble the contig. So, its limits are [ 1, (consensus length + number of gaps in consensus) ]</p>

<p>&quot;ungapped consensus&quot; is a coordinate system based on the consensus sequence, but excluding consensus gaps. This is just the coordinate system that you have when considering the consensus sequence alone, instead of aligned to other sequences.</p>

<p>&quot;aligned $seqID&quot; refers to locations in the sequence $seqID after alignment of $seqID against the consensus sequence (reverse complementing the original sequence, if needed). Coordinate 1 in &quot;aligned $seqID&quot; is equivalent to the start location (first base) of $seqID in the consensus sequence, just like if the aligned sequence $seqID was a feature of the consensus sequence.</p>

<p>&quot;unaligned $seqID&quot; is equivalent to a location in the isolated sequence, just like you would have when considering the sequence alone, out of an alignment. When changing coordinates from &quot;aligned $seq2&quot; to &quot;unaligned $seq2&quot;, if $seq2 was reverse complemented when included in the alignment, the output coordinates will be reversed to fit that fact, i.e. 1 will be changed to length($seq2), 2 will be length($seq)-1 and so on.</p>

<p>An important note: when you change gap coordinates from a gapped system (&quot;gapped consensus&quot; or &quot;aligned $seqID&quot;) to a system that does not include gaps (&quot;ungapped consensus&quot; or &quot;unaligned $seqID&quot;), the position returned will be the first location before all gaps neighboring the input location.</p>

<h2 id="Feature_collection">Feature_collection</h2>

<p>Bio::Assembly::Contig stores much information about a contig in a Bio::Assembly::SeqFeature::Collection object. Relevant information on the alignment is accessed by selecting features based on their primary tags (e.g. all features which have a primary tag of the form &#39;_aligned_coord:$seqID&#39;, where $seqID is an aligned sequence ID, are coordinates for sequences in the contig alignment) and, by using methods from Bio::Assembly::SeqFeature::Collection, it&#39;s possible to select features by overlap with other features.</p>

<p>We suggest that you use the primary tags of features as identifiers for feature classes. By convention, features with primary tags starting with a &#39;_&#39; are generated by modules that populate the contig data structure and return the contig object, maybe as part of an assembly object, e.g. drivers from the Bio::Assembly::IO set.</p>

<p>Features in the features collection may be associated with particular aligned sequences. To obtain this, you must attach the sequence to the feature, using attach() seq from Bio::Assembly::SeqFeatureI, before you add the feature to the feature collection. We also suggest to add the sequence id to the primary tag, so that is easy to select feature for a particular sequence.</p>

<p>There is only one feature class that some methods in Bio::Assembly::Contig expect to find in the feature collection: features with primary tags of the form &#39;_aligned_coord:$seqID&#39;, where $seqID is the aligned sequence id (like returned by $seq-&gt;id()). These features describe the position (in &quot;gapped consensus&quot; coordinates) of aligned sequences, and the method set_seq_coord() automatically changes a feature&#39;s primary tag to this form whenever the feature is added to the collection by this method. Only two methods in Bio::Assembly::Contig will not work unless there are features from this class: change_coord() and get_seq_coord().</p>

<p>Other feature classes will be automatically available only when Bio::Assembly::Contig objects are created by a specific module. Such feature classes are (or should be) documented in the documentation of the module which create them, to which the user should refer.</p>

<h1 id="FEEDBACK">FEEDBACK</h1>

<h2 id="Mailing-Lists">Mailing Lists</h2>

<p>User feedback is an integral part of the evolution of this and other Bioperl modules. Send your comments and suggestions preferably to the Bioperl mailing lists Your participation is much appreciated.</p>

<pre><code><code>  bioperl-l@bioperl.org                  - General discussion
  http://bioperl.org/wiki/Mailing_lists  - About the mailing lists</code></code></pre>

<h2 id="Support">Support</h2>

<p>Please direct usage questions or support issues to the mailing list:</p>

<p><i>bioperl-l@bioperl.org</i></p>

<p>rather than to the module maintainer directly. Many experienced and reponsive experts will be able look at the problem and quickly address it. Please include a thorough description of the problem with code and data examples if at all possible.</p>

<h2 id="Reporting-Bugs">Reporting Bugs</h2>

<p>Report bugs to the Bioperl bug tracking system to help us keep track the bugs and their resolution. Bug reports can be submitted via the web:</p>

<pre><code><code>  https://redmine.open-bio.org/projects/bioperl/</code></code></pre>

<h1 id="AUTHOR---Robson-Francisco-de-Souza">AUTHOR - Robson Francisco de Souza</h1>

<p>rfsouza@citri.iq.usp.br</p>

<h1 id="APPENDIX">APPENDIX</h1>

<p>The rest of the documentation details each of the object methods. Internal methods are usually preceded with a _</p>

<h1 id="Object-creator">Object creator</h1>

<h2 id="new">new</h2>

<pre><code><code> <span class="variable">Title</span>     <span class="operator">:</span> <span class="variable">new</span>
 <span class="variable">Usage</span>     <span class="operator">:</span> <span class="keyword">my</span> <span class="variable">$contig</span> <span class="operator">=</span> <span class="variable">Bio::Assembly::Contig</span><span class="operator">-&gt;</span><span class="variable">new</span><span class="operator">();</span>
 <span class="variable">Function</span>  <span class="operator">:</span> <span class="variable">Creates</span> <span class="variable">a</span> <span class="variable">new</span> <span class="variable">contig</span> <span class="variable">object</span>
 <span class="variable">Returns</span>   <span class="operator">:</span> <span class="variable">Bio::Assembly::Contig</span>
 <span class="variable">Args</span>      <span class="operator">:</span> <span class="string">-id</span>         <span class="operator">=&gt;</span> <span class="variable">unique</span> <span class="variable">contig</span> <span class="variable">ID</span>
             <span class="operator">-</span><span class="string">source</span>     <span class="operator">=&gt;</span> <span class="variable">string</span> <span class="keyword">for</span> <span class="variable">the</span> <span class="variable">sequence</span> <span class="variable">assembly</span> <span class="variable">program</span> <span class="variable">used</span>
             <span class="operator">-</span><span class="string">collection</span> <span class="operator">=&gt;</span> <span class="variable">Bio::SeqFeature::CollectionI</span> <span class="variable">instance</span>
</code></code></pre>

<h1 id="Assembly-related-methods">Assembly related methods</h1>

<p>These methods exist to enable adding information about possible relations among contigs, e.g. when you already have a scaffold for your assembly, describing the ordering of contigs in the final assembly, but no sequences covering the gaps between neighboring contigs.</p>

<h2 id="source">source</h2>

<pre><code><code> <span class="variable">Title</span>     <span class="operator">:</span> <span class="variable">source</span>
 <span class="variable">Usage</span>     <span class="operator">:</span> <span class="variable">$contig</span><span class="operator">-&gt;</span><span class="variable">source</span><span class="operator">(</span><span class="variable">$program</span><span class="operator">);</span>
 <span class="variable">Function</span>  <span class="operator">:</span> <span class="variable">Get</span><span class="operator">/</span><span class="variable">Set</span> <span class="variable">program</span> <span class="variable">used</span> <span class="variable">to</span> <span class="variable">build</span> <span class="variable">this</span> <span class="variable">contig</span>
 <span class="variable">Returns</span>   <span class="operator">:</span> <span class="variable">string</span>
 <span class="variable">Argument</span>  <span class="operator">:</span> <span class="operator">[</span><span class="variable">optional</span><span class="operator">]</span> <span class="variable">string</span>
</code></code></pre>

<h2 id="assembly">assembly</h2>

<pre><code><code> <span class="variable">Title</span>     <span class="operator">:</span> <span class="variable">assembly</span>
 <span class="variable">Usage</span>     <span class="operator">:</span> <span class="variable">$contig</span><span class="operator">-&gt;</span><span class="variable">assembly</span><span class="operator">(</span><span class="variable">$assembly</span><span class="operator">);</span>
 <span class="variable">Function</span>  <span class="operator">:</span> <span class="variable">Get</span><span class="operator">/</span><span class="variable">Set</span> <span class="variable">assembly</span> <span class="variable">object</span> <span class="keyword">for</span> <span class="variable">this</span> <span class="variable">contig</span>
 <span class="variable">Returns</span>   <span class="operator">:</span> <span class="variable">a</span> <span class="variable">Bio::Assembly::Scaffold</span> <span class="variable">object</span>
 <span class="variable">Argument</span>  <span class="operator">:</span> <span class="variable">a</span> <span class="variable">Bio::Assembly::Scaffold</span> <span class="variable">object</span>
</code></code></pre>

<h2 id="strand">strand</h2>

<pre><code><code> <span class="variable">Title</span>     <span class="operator">:</span> <span class="variable">strand</span>
 <span class="variable">Usage</span>     <span class="operator">:</span> <span class="variable">$contig</span><span class="operator">-&gt;</span><span class="variable">strand</span><span class="operator">(</span><span class="variable">$num</span><span class="operator">);</span>
 <span class="variable">Function</span>  <span class="operator">:</span> <span class="variable">Get</span><span class="operator">/</span><span class="variable">Set</span> <span class="variable">contig</span> <span class="variable">orientation</span> <span class="variable">in</span> <span class="variable">a</span> <span class="variable">scaffold</span><span class="operator">/</span><span class="variable">assembly</span><span class="operator">.</span>
             <span class="variable">Its</span> <span class="variable">equivalent</span> <span class="variable">to</span> <span class="variable">the</span> <span class="variable">strand</span> <span class="variable">property</span> <span class="variable">of</span> <span class="variable">sequence</span>
             <span class="variable">objects</span> <span class="keyword">and</span> <span class="variable">sets</span> <span class="variable">whether</span> <span class="variable">the</span> <span class="variable">contig</span> <span class="variable">consensus</span> <span class="variable">should</span>
             <span class="variable">be</span> <span class="variable">reversed</span> <span class="keyword">and</span> <span class="variable">complemented</span> <span class="variable">before</span> <span class="variable">being</span> <span class="variable">added</span> <span class="variable">to</span> <span class="variable">a</span>
             <span class="variable">scaffold</span> <span class="keyword">or</span> <span class="variable">assembly</span><span class="operator">.</span>
 <span class="variable">Returns</span>   <span class="operator">:</span> <span class="variable">integer</span>
 <span class="variable">Argument</span>  <span class="operator">:</span> <span class="number">1</span> <span class="keyword">if</span> <span class="variable">orientaion</span> <span class="variable">is</span> <span class="variable">forward</span><span class="operator">,</span> <span class="operator">-</span><span class="number">1</span> <span class="keyword">if</span> <span class="keyword">reverse</span> <span class="keyword">and</span>
             <span class="number">0</span> <span class="keyword">if</span> <span class="variable">none</span>
</code></code></pre>

<h2 id="upstream_neighbor">upstream_neighbor</h2>

<pre><code><code> <span class="variable">Title</span>     <span class="operator">:</span> <span class="variable">upstream_neighbor</span>
 <span class="variable">Usage</span>     <span class="operator">:</span> <span class="variable">$contig</span><span class="operator">-&gt;</span><span class="variable">upstream_neighbor</span><span class="operator">(</span><span class="variable">$contig</span><span class="operator">);</span>
 <span class="variable">Function</span>  <span class="operator">:</span> <span class="variable">Get</span><span class="operator">/</span><span class="variable">Set</span> <span class="variable">a</span> <span class="variable">contig</span> <span class="variable">neighbor</span> <span class="keyword">for</span> <span class="variable">the</span> <span class="variable">current</span> <span class="variable">contig</span> <span class="keyword">when</span>
             <span class="variable">building</span> <span class="variable">a</span> <span class="variable">scaffold</span><span class="operator">.</span> <span class="variable">The</span> <span class="variable">upstream</span> <span class="variable">neighbor</span> <span class="variable">is</span>
             <span class="variable">located</span> <span class="variable">before</span> <span class="variable">$contig</span> <span class="variable">first</span> <span class="variable">base</span>
 <span class="variable">Returns</span>   <span class="operator">:</span> <span class="variable">nothing</span>
 <span class="variable">Argument</span>  <span class="operator">:</span> <span class="variable">Bio::Assembly::Contig</span>
</code></code></pre>

<h2 id="downstream_neighbor">downstream_neighbor</h2>

<pre><code><code> <span class="variable">Title</span>     <span class="operator">:</span> <span class="variable">downstream_neighbor</span>
 <span class="variable">Usage</span>     <span class="operator">:</span> <span class="variable">$contig</span><span class="operator">-&gt;</span><span class="variable">downstream_neighbor</span><span class="operator">(</span><span class="variable">$num</span><span class="operator">);</span>
 <span class="variable">Function</span>  <span class="operator">:</span> <span class="variable">Get</span><span class="operator">/</span><span class="variable">Set</span> <span class="variable">a</span> <span class="variable">contig</span> <span class="variable">neighbor</span> <span class="keyword">for</span> <span class="variable">the</span> <span class="variable">current</span> <span class="variable">contig</span> <span class="keyword">when</span>
             <span class="variable">building</span> <span class="variable">a</span> <span class="variable">scaffold</span><span class="operator">.</span> <span class="variable">The</span> <span class="variable">downstream</span> <span class="variable">neighbor</span> <span class="variable">is</span>
             <span class="variable">located</span> <span class="variable">after</span> <span class="variable">$contig</span> <span class="keyword">last</span> <span class="variable">base</span>
 <span class="variable">Returns</span>   <span class="operator">:</span> <span class="variable">nothing</span>
 <span class="variable">Argument</span>  <span class="operator">:</span> <span class="variable">Bio::Assembly::Contig</span>
</code></code></pre>

<h1 id="Contig-feature-collection-methods">Contig feature collection methods</h1>

<h2 id="add_features">add_features</h2>

<pre><code><code> Title     : add_features
 Usage     : $contig-&gt;add_features($feat,$flag)
 Function  :

             Add an array of features to the contig feature
             collection. The consensus sequence may be attached to the
             added feature, if $flag is set to 1. If $flag is 0 and
             the feature attached to one of the contig aligned
             sequences, the feature is registered as an aligned
             sequence feature. If $flag is 0 and the feature is not
             attched to any sequence in the contig, the feature is
             simply added to the feature collection and no attachment
             or registration is made.

             Note: You must attach aligned sequences to their features
             prior to calling add_features, otherwise you won&#39;t be
             able to access the feature through get_seq_feat_by_tag()
             method.

 Returns   : number of features added.
 Argument  :
             $feat : A reference to an array of Bio::SeqFeatureI
             $flag : boolean - true if consensus sequence object
                     should be attached to this feature, false if
                     no consensus attachment should be made.
                     Default: false.</code></code></pre>

<h2 id="remove_features">remove_features</h2>

<pre><code><code> Title     : remove_features
 Usage     : $contig-&gt;remove_features(@feat)
 Function  : Remove an array of contig features
 Returns   : true if successful
 Argument  : An array of Bio::SeqFeature::Generic (Bio::SeqFeatureI)</code></code></pre>

<h2 id="get_features_collection">get_features_collection</h2>

<pre><code><code> Title     : get_features_collection
 Usage     : $contig-&gt;get_features_collection()
 Function  : Get the collection of all contig features and seqfeatures
 Returns   : Bio::DB::SeqFeature::Store (Bio::SeqFeature::CollectionI)
 Argument  : none</code></code></pre>

<h2 id="remove_features_collection">remove_features_collection</h2>

<pre><code><code> Title     : remove_features_collection
 Usage     : $contig-&gt;remove_features_collection()
 Function  : Remove the collection of all contig features. It is useful
             to save some memory (when contig features are not needed).
 Returns   : none
 Argument  : none</code></code></pre>

<h1 id="Coordinate-systems-related-methods">Coordinate system&#39;s related methods</h1>

<p>See <a>Coordinate_Systems</a> above.</p>

<h2 id="change_coord">change_coord</h2>

<pre><code><code> Title     : change_coord
 Usage     : $contig-&gt;change_coord($in,$out,$query)
 Function  :

             Change coordinate system for $query.  This method
             transforms locations between coordinate systems described
             in section &quot;Coordinate Systems&quot; of this document.

             Note: this method will throw an exception when changing
             coordinates between &quot;ungapped consensus&quot; and other
             systems if consensus sequence was not set. It will also
             throw exceptions when changing coordinates among aligned
             sequence, either with or without gaps, and other systems
             if sequence locations were not set with set_seq_coord().

 Returns   : integer
 Argument  :
             $in    : [string]  input coordinate system
             $out   : [string]  output coordinate system
             $query : [integer] a position in a sequence</code></code></pre>

<h2 id="get_seq_coord">get_seq_coord</h2>

<pre><code><code> <span class="variable">Title</span>     <span class="operator">:</span> <span class="variable">get_seq_coord</span>
 <span class="variable">Usage</span>     <span class="operator">:</span> <span class="variable">$contig</span><span class="operator">-&gt;</span><span class="variable">get_seq_coord</span><span class="operator">(</span><span class="variable">$seq</span><span class="operator">);</span>
 <span class="variable">Function</span>  <span class="operator">:</span> <span class="variable">Get</span> <span class="string">"gapped consensus"</span> <span class="variable">location</span> <span class="keyword">for</span> <span class="variable">aligned</span> <span class="variable">sequence</span>
 <span class="variable">Returns</span>   <span class="operator">:</span> <span class="variable">Bio::SeqFeature::Generic</span> <span class="keyword">for</span> <span class="variable">coordinates</span> <span class="keyword">or</span> <span class="keyword">undef</span><span class="operator">.</span>
             <span class="variable">A</span> <span class="variable">warning</span> <span class="variable">is</span> <span class="variable">printed</span> <span class="keyword">if</span> <span class="variable">sequence</span> <span class="variable">coordinates</span> <span class="variable">were</span> <span class="keyword">not</span> <span class="variable">set</span><span class="operator">.</span>
 <span class="variable">Argument</span>  <span class="operator">:</span> <span class="variable">Bio::LocatableSeq</span> <span class="variable">object</span>
</code></code></pre>

<h2 id="set_seq_coord">set_seq_coord</h2>

<pre><code><code> <span class="variable">Title</span>     <span class="operator">:</span> <span class="variable">set_seq_coord</span>
 <span class="variable">Usage</span>     <span class="operator">:</span> <span class="variable">$contig</span><span class="operator">-&gt;</span><span class="variable">set_seq_coord</span><span class="operator">(</span><span class="variable">$feat</span><span class="operator">,</span><span class="variable">$seq</span><span class="operator">);</span>
 <span class="variable">Function</span>  <span class="operator">:</span>
 
             <span class="variable">Set</span> <span class="string">"gapped consensus"</span> <span class="variable">location</span> <span class="keyword">for</span> <span class="variable">an</span> <span class="variable">aligned</span>
             <span class="variable">sequence</span><span class="operator">.</span> <span class="variable">If</span> <span class="variable">the</span> <span class="variable">sequence</span> <span class="variable">was</span> <span class="variable">previously</span> <span class="variable">added</span> <span class="variable">using</span>
             <span class="variable">add_seq</span><span class="operator">,</span> <span class="variable">its</span> <span class="variable">coordinates</span> <span class="variable">are</span> <span class="variable">changed</span><span class="operator">/</span><span class="variable">set</span><span class="operator">.</span>  <span class="variable">Otherwise</span><span class="operator">,</span>
             <span class="variable">add_seq</span> <span class="variable">is</span> <span class="variable">called</span> <span class="keyword">and</span> <span class="variable">the</span> <span class="variable">sequence</span> <span class="variable">is</span> <span class="variable">added</span> <span class="variable">to</span> <span class="variable">the</span>
             <span class="variable">contig</span><span class="operator">.</span>
 
 <span class="variable">Returns</span>   <span class="operator">:</span> <span class="variable">Bio::SeqFeature::Generic</span> <span class="keyword">for</span> <span class="variable">old</span> <span class="variable">coordinates</span> <span class="keyword">or</span> <span class="keyword">undef</span><span class="operator">.</span>
 <span class="variable">Argument</span>  <span class="operator">:</span>
             <span class="variable">$feat</span>  <span class="operator">:</span> <span class="variable">a</span> <span class="variable">Bio::SeqFeature::Generic</span> <span class="variable">object</span>
                      <span class="variable">representing</span> <span class="variable">a</span> <span class="variable">location</span> <span class="keyword">for</span> <span class="variable">the</span>
                      <span class="variable">aligned</span> <span class="variable">sequence</span><span class="operator">,</span> <span class="variable">in</span> <span class="string">"gapped
                      consensus"</span> <span class="variable">coordinates</span><span class="operator">.</span>
 
             <span class="variable">Note</span><span class="operator">:</span> <span class="variable">the</span> <span class="variable">original</span> <span class="variable">feature</span> <span class="variable">primary</span> <span class="variable">tag</span> <span class="variable">will</span>
                   <span class="variable">be</span> <span class="variable">lost</span><span class="operator">.</span>
 
             <span class="variable">$seq</span>   <span class="operator">:</span> <span class="variable">a</span> <span class="variable">Bio::LocatableSeq</span> <span class="variable">object</span>
</code></code></pre>

<h1 id="Bio::Assembly::Contig-consensus-methods">Bio::Assembly::Contig consensus methods</h1>

<h2 id="set_consensus_sequence">set_consensus_sequence</h2>

<pre><code><code> Title     : set_consensus_sequence
 Usage     : $contig-&gt;set_consensus_sequence($seq)
 Function  : Set the consensus sequence object for this contig
 Returns   : consensus length
 Argument  : Bio::LocatableSeq</code></code></pre>

<h2 id="set_consensus_quality">set_consensus_quality</h2>

<pre><code><code> Title     : set_consensus_quality
 Usage     : $contig-&gt;set_consensus_quality($qual)
 Function  : Set the quality object for consensus sequence
 Returns   : nothing
 Argument  : Bio::Seq::QualI object</code></code></pre>

<h2 id="get_consensus_length">get_consensus_length</h2>

<pre><code><code> Title     : get_consensus_length
 Usage     : $contig-&gt;get_consensus_length()
 Function  : Get consensus sequence length
 Returns   : integer
 Argument  : none</code></code></pre>

<h2 id="get_consensus_sequence">get_consensus_sequence</h2>

<pre><code><code> Title     : get_consensus_sequence
 Usage     : $contig-&gt;get_consensus_sequence()
 Function  : Get a reference to the consensus sequence object
             for this contig
 Returns   : Bio::SeqI object
 Argument  : none</code></code></pre>

<h2 id="get_consensus_quality">get_consensus_quality</h2>

<pre><code><code> Title     : get_consensus_quality
 Usage     : $contig-&gt;get_consensus_quality()
 Function  : Get a reference to the consensus quality object
             for this contig.
 Returns   : A Bio::Seq::QualI object
 Argument  : none</code></code></pre>

<h1 id="Bio::Assembly::Contig-aligned-sequences-methods">Bio::Assembly::Contig aligned sequences methods</h1>

<h2 id="set_seq_qual">set_seq_qual</h2>

<pre><code><code> <span class="variable">Title</span>     <span class="operator">:</span> <span class="variable">set_seq_qual</span>
 <span class="variable">Usage</span>     <span class="operator">:</span> <span class="variable">$contig</span><span class="operator">-&gt;</span><span class="variable">set_seq_qual</span><span class="operator">(</span><span class="variable">$seq</span><span class="operator">,</span><span class="variable">$qual</span><span class="operator">);</span>
 <span class="variable">Function</span>  <span class="operator">:</span> <span class="variable">Adds</span> <span class="variable">quality</span> <span class="variable">to</span> <span class="variable">an</span> <span class="variable">aligned</span> <span class="variable">sequence</span><span class="operator">.</span>
 <span class="variable">Returns</span>   <span class="operator">:</span> <span class="variable">nothing</span>
 <span class="variable">Argument</span>  <span class="operator">:</span> <span class="variable">a</span> <span class="variable">Bio::LocatableSeq</span> <span class="variable">object</span> <span class="keyword">and</span>
             <span class="variable">a</span> <span class="variable">Bio::Seq::QualI</span> <span class="variable">object</span>
</code></code></pre>

<p>See <a href="../../../../../.cpan/build/BioPerl-1.6.901-ghKY8U/blib/lib/Bio/LocatableSeq.html">Bio::LocatableSeq</a> for more information.</p>

<h2 id="get_seq_ids">get_seq_ids</h2>

<pre><code><code> <span class="variable">Title</span>     <span class="operator">:</span> <span class="variable">get_seq_ids</span>
 <span class="variable">Usage</span>     <span class="operator">:</span> <span class="variable">$contig</span><span class="operator">-&gt;</span><span class="variable">get_seq_ids</span><span class="operator">(</span> <span class="string">-start</span> <span class="operator">=&gt;</span> <span class="variable">$start</span><span class="operator">,</span>
                                   <span class="string">-end</span>   <span class="operator">=&gt;</span> <span class="variable">$end</span><span class="operator">,</span>
                                   <span class="string">-type</span>  <span class="operator">=&gt;</span> <span class="string">"gapped A0QR67B08.b"</span> <span class="operator">);</span>
 <span class="variable">Function</span>  <span class="operator">:</span> <span class="variable">Get</span> <span class="variable">list</span> <span class="variable">of</span> <span class="variable">sequence</span> <span class="variable">IDs</span> <span class="variable">overlapping</span> <span class="variable">interval</span> <span class="operator">[</span><span class="variable">$start</span><span class="operator">,</span> <span class="variable">$end</span><span class="operator">]</span>
             <span class="variable">The</span> <span class="keyword">default</span> <span class="variable">interval</span> <span class="variable">is</span> <span class="operator">[</span><span class="number">1</span><span class="operator">,</span><span class="variable">$contig</span><span class="operator">-&gt;</span><span class="variable">length</span><span class="operator">]</span>
             <span class="variable">Default</span> <span class="variable">coordinate</span> <span class="keyword">system</span> <span class="variable">is</span> <span class="string">"gapped contig"</span>
 <span class="variable">Returns</span>   <span class="operator">:</span> <span class="variable">An</span> <span class="variable">array</span>
 <span class="variable">Argument</span>  <span class="operator">:</span> <span class="variable">A</span> <span class="variable">hash</span> <span class="variable">with</span> <span class="variable">optional</span> <span class="variable">elements</span><span class="operator">:</span>
             <span class="operator">-</span><span class="variable">start</span> <span class="operator">:</span> <span class="variable">consensus</span> <span class="variable">subsequence</span> <span class="variable">start</span>
             <span class="operator">-</span><span class="variable">end</span>   <span class="operator">:</span> <span class="variable">consensus</span> <span class="variable">subsequence</span> <span class="variable">end</span>
             <span class="operator">-</span><span class="variable">type</span>  <span class="operator">:</span> <span class="variable">the</span> <span class="variable">coordinate</span> <span class="keyword">system</span> <span class="variable">type</span> <span class="keyword">for</span> <span class="variable">$start</span> <span class="keyword">and</span> <span class="variable">$end</span> <span class="variable">arguments</span>
                      <span class="variable">Coordinate</span> <span class="keyword">system</span> <span class="variable">available</span> <span class="variable">are</span><span class="operator">:</span>
                      <span class="string">"gapped consensus"</span>   <span class="operator">:</span> <span class="variable">consensus</span> <span class="variable">coordinates</span> <span class="variable">with</span> <span class="variable">gaps</span>
                      <span class="string">"ungapped consensus"</span> <span class="operator">:</span> <span class="variable">consensus</span> <span class="variable">coordinates</span> <span class="variable">without</span> <span class="variable">gaps</span>
                      <span class="string">"aligned </span><span class="variable">$ReadID</span><span class="string">"</span>    <span class="operator">:</span> <span class="keyword">read</span> <span class="variable">$ReadID</span> <span class="variable">coordinates</span> <span class="variable">with</span> <span class="variable">gaps</span>
                      <span class="string">"unaligned </span><span class="variable">$ReadID</span><span class="string">"</span>  <span class="operator">:</span> <span class="keyword">read</span> <span class="variable">$ReadID</span> <span class="variable">coordinates</span> <span class="variable">without</span> <span class="variable">gaps</span>
</code></code></pre>

<h2 id="get_seq_feat_by_tag">get_seq_feat_by_tag</h2>

<pre><code><code> Title     : get_seq_feat_by_tag
 Usage     : $seq = $contig-&gt;get_seq_feat_by_tag($seq,&quot;_aligned_coord:$seqID&quot;)
 Function  : Get a sequence feature based on its primary_tag.
 Returns   : a Bio::SeqFeature object
 Argument  : a Bio::LocatableSeq and a string (feature primary tag)</code></code></pre>

<h2 id="get_seq_by_name">get_seq_by_name</h2>

<pre><code><code> Title     : get_seq_by_name
 Usage     : $seq = $contig-&gt;get_seq_by_name(&#39;Seq1&#39;)
 Function  : Gets a sequence based on its id.
 Returns   : a Bio::LocatableSeq object
             undef if name is not found
 Argument  : string</code></code></pre>

<h2 id="get_qual_by_name">get_qual_by_name</h2>

<pre><code><code> Title     : get_qual_by_name
 Usage     : $seq = $contig-&gt;get_qual_by_name(&#39;Seq1&#39;)
 Function  :

             Gets Bio::Seq::QualI object for a sequence
             through its id ( as given by $qual-&gt;id() ).

 Returns   : a Bio::Seq::QualI object.
             undef if name is not found
 Argument  : string</code></code></pre>

<h1 id="Bio::Align::AlignI-compatible-methods">Bio::Align::AlignI compatible methods</h1>

<h2 id="Modifier-methods">Modifier methods</h2>

<p>These methods modify the MSE by adding, removing or shuffling complete sequences.</p>

<h2 id="add_seq">add_seq</h2>

<pre><code><code> <span class="variable">Title</span>     <span class="operator">:</span> <span class="variable">add_seq</span>
 <span class="variable">Usage</span>     <span class="operator">:</span> <span class="variable">$contig</span><span class="operator">-&gt;</span><span class="variable">add_seq</span><span class="operator">(</span><span class="variable">$newseq</span><span class="operator">);</span>
 <span class="variable">Function</span>  <span class="operator">:</span>
 
             <span class="variable">Adds</span> <span class="variable">a</span> <span class="variable">sequence</span> <span class="variable">to</span> <span class="variable">the</span> <span class="variable">contig</span><span class="operator">.</span> <span class="variable">*Does</span><span class="operator">*</span>
             <span class="variable">*not</span><span class="operator">*</span> <span class="variable">align</span> <span class="variable">it</span> <span class="operator">-</span> <span class="variable">just</span> <span class="variable">adds</span> <span class="variable">it</span> <span class="variable">to</span> <span class="variable">the</span>
             <span class="variable">hashes</span><span class="operator">.</span>
 
 <span class="variable">Returns</span>   <span class="operator">:</span> <span class="variable">nothing</span>
 <span class="variable">Argument</span>  <span class="operator">:</span> <span class="variable">a</span> <span class="variable">Bio::LocatableSeq</span> <span class="variable">object</span>
</code></code></pre>

<p>See <a href="../../../../../.cpan/build/BioPerl-1.6.901-ghKY8U/blib/lib/Bio/LocatableSeq.html">Bio::LocatableSeq</a> for more information.</p>

<h2 id="remove_seq">remove_seq</h2>

<pre><code><code> <span class="variable">Title</span>     <span class="operator">:</span> <span class="variable">remove_seq</span>
 <span class="variable">Usage</span>     <span class="operator">:</span> <span class="variable">$contig</span><span class="operator">-&gt;</span><span class="variable">remove_seq</span><span class="operator">(</span><span class="variable">$seq</span><span class="operator">);</span>
 <span class="variable">Function</span>  <span class="operator">:</span> <span class="variable">Removes</span> <span class="variable">a</span> <span class="variable">single</span> <span class="variable">sequence</span> <span class="variable">from</span> <span class="variable">a</span> <span class="variable">contig</span>
 <span class="variable">Returns</span>   <span class="operator">:</span> <span class="number">1</span> <span class="variable">on</span> <span class="variable">success</span><span class="operator">,</span> <span class="number">0</span> <span class="variable">otherwise</span>
 <span class="variable">Argument</span>  <span class="operator">:</span> <span class="variable">a</span> <span class="variable">Bio::LocatableSeq</span> <span class="variable">object</span>
</code></code></pre>

<h2 id="purge">purge</h2>

<pre><code><code> <span class="variable">Title</span>   <span class="operator">:</span> <span class="variable">purge</span>
 <span class="variable">Usage</span>   <span class="operator">:</span> <span class="variable">$contig</span><span class="operator">-&gt;</span><span class="variable">purge</span><span class="operator">(</span><span class="number">0</span><span class="operator">.</span><span class="number">7</span><span class="operator">);</span>
 <span class="variable">Function</span><span class="operator">:</span>
 
           <span class="variable">Removes</span> <span class="variable">sequences</span> <span class="variable">above</span> <span class="variable">whatever</span> <span class="variable">%id</span><span class="operator">.</span>
 
           <span class="variable">This</span> <span class="variable">function</span> <span class="variable">will</span> <span class="variable">grind</span> <span class="variable">on</span> <span class="variable">large</span> <span class="variable">alignments</span><span class="operator">.</span> <span class="variable">Beware</span><span class="operator">!</span>
           <span class="operator">(</span><span class="variable">perhaps</span> <span class="keyword">not</span> <span class="variable">ideally</span> <span class="variable">implemented</span><span class="operator">)</span>
 
 <span class="variable">Example</span> <span class="operator">:</span>
 <span class="variable">Returns</span> <span class="operator">:</span> <span class="variable">An</span> <span class="variable">array</span> <span class="variable">of</span> <span class="variable">the</span> <span class="variable">removed</span> <span class="variable">sequences</span>
 <span class="variable">Argument</span><span class="operator">:</span>
</code></code></pre>

<h2 id="sort_alphabetically">sort_alphabetically</h2>

<pre><code><code> Title     : sort_alphabetically
 Usage     : $contig-&gt;sort_alphabetically
 Function  :

             Changes the order of the alignemnt to alphabetical on name
             followed by numerical by number.

 Returns   :
 Argument  :</code></code></pre>

<h2 id="Sequence-selection-methods">Sequence selection methods</h2>

<p>Methods returning one or more sequences objects.</p>

<h2 id="each_seq">each_seq</h2>

<pre><code><code> Title     : each_seq
 Usage     : foreach $seq ( $contig-&gt;each_seq() )
 Function  : Gets an array of Seq objects from the alignment
 Returns   : an array
 Argument  :</code></code></pre>

<h2 id="each_alphabetically">each_alphabetically</h2>

<pre><code><code> Title     : each_alphabetically
 Usage     : foreach $seq ( $contig-&gt;each_alphabetically() )
 Function  :

             Returns an array of sequence object sorted alphabetically
             by name and then by start point.
             Does not change the order of the alignment

 Returns   :
 Argument  :</code></code></pre>

<h2 id="each_seq_with_id">each_seq_with_id</h2>

<pre><code><code> Title     : each_seq_with_id
 Usage     : foreach $seq ( $contig-&gt;each_seq_with_id() )
 Function  :

             Gets an array of Seq objects from the
             alignment, the contents being those sequences
             with the given name (there may be more than one)

 Returns   : an array
 Argument  : a seq name</code></code></pre>

<h2 id="get_seq_by_pos">get_seq_by_pos</h2>

<pre><code><code> Title     : get_seq_by_pos
 Usage     : $seq = $contig-&gt;get_seq_by_pos(3)
 Function  :

             Gets a sequence based on its position in the alignment.
             Numbering starts from 1.  Sequence positions larger than
             num_sequences() will thow an error.

 Returns   : a Bio::LocatableSeq object
 Argument  : positive integer for the sequence osition</code></code></pre>

<h2 id="Create-new-alignments">Create new alignments</h2>

<p>The result of these methods are horizontal or vertical subsets of the current MSE.</p>

<h2 id="select">select</h2>

<pre><code><code> Title     : select
 Usage     : $contig2 = $contig-&gt;select(1, 3) # three first sequences
 Function  :

             Creates a new alignment from a continuous subset of
             sequences.  Numbering starts from 1.  Sequence positions
             larger than num_sequences() will thow an error.

 Returns   : a Bio::Assembly::Contig object
 Argument  : positive integer for the first sequence
             positive integer for the last sequence to include (optional)</code></code></pre>

<h2 id="select_noncont">select_noncont</h2>

<pre><code><code> Title     : select_noncont
 Usage     : $contig2 = $contig-&gt;select_noncont(1, 3) # first and 3rd sequences
 Function  :

             Creates a new alignment from a subset of
             sequences.  Numbering starts from 1.  Sequence positions
             larger than num_sequences() will throw an error.

 Returns   : a Bio::Assembly::Contig object
 Args      : array of integers for the sequences</code></code></pre>

<h2 id="slice">slice</h2>

<pre><code><code> Title     : slice
 Usage     : $contig2 = $contig-&gt;slice(20, 30)
 Function  :

             Creates a slice from the alignment inclusive of start and
             end columns.  Sequences with no residues in the slice are
             excluded from the new alignment and a warning is printed.
             Slice beyond the length of the sequence does not do
             padding.

 Returns   : a Bio::Assembly::Contig object
 Argument  : positive integer for start column
             positive integer for end column</code></code></pre>

<h2 id="Change-sequences-within-the-MSE">Change sequences within the MSE</h2>

<p>These methods affect characters in all sequences without changeing the alignment.</p>

<h2 id="map_chars">map_chars</h2>

<pre><code><code> Title     : map_chars
 Usage     : $contig-&gt;map_chars(&#39;\.&#39;,&#39;-&#39;)
 Function  :

             Does a s/$arg1/$arg2/ on the sequences. Useful for gap
             characters

             Notice that the from (arg1) is interpretted as a regex,
             so be careful about quoting meta characters (eg
             $contig-&gt;map_chars(&#39;.&#39;,&#39;-&#39;) wont do what you want)

 Returns   :
 Argument  : &#39;from&#39; rexexp
             &#39;to&#39; string</code></code></pre>

<h2 id="uppercase">uppercase</h2>

<pre><code><code> Title     : uppercase()
 Usage     : $contig-&gt;uppercase()
 Function  : Sets all the sequences to uppercase
 Returns   :
 Argument  :</code></code></pre>

<h2 id="match_line">match_line</h2>

<pre><code><code> Title    : match_line()
 Usage    : $contig-&gt;match_line()
 Function : Generates a match line - much like consensus string
            except that a line indicating the &#39;*&#39; for a match.
 Argument : (optional) Match line characters (&#39;*&#39; by default)
            (optional) Strong match char (&#39;:&#39; by default)
            (optional) Weak match char (&#39;.&#39; by default)</code></code></pre>

<h2 id="match">match</h2>

<pre><code><code> Title     : match()
 Usage     : $contig-&gt;match()
 Function  :

             Goes through all columns and changes residues that are
             identical to residue in first sequence to match &#39;.&#39;
             character. Sets match_char.

             USE WITH CARE: Most MSE formats do not support match
             characters in sequences, so this is mostly for output
             only. NEXUS format (Bio::AlignIO::nexus) can handle
             it.

 Returns   : 1
 Argument  : a match character, optional, defaults to &#39;.&#39;</code></code></pre>

<h2 id="unmatch">unmatch</h2>

<pre><code><code> Title     : unmatch()
 Usage     : $contig-&gt;unmatch()
 Function  :

             Undoes the effect of method match. Unsets match_char.

 Returns   : 1
 Argument  : a match character, optional, defaults to &#39;.&#39;</code></code></pre>

<h2 id="MSE-attibutes">MSE attibutes</h2>

<p>Methods for setting and reading the MSE attributes.</p>

<p>Note that the methods defining character semantics depend on the user to set them sensibly. They are needed only by certain input/output methods. Unset them by setting to an empty string (&#39;&#39;).</p>

<h2 id="id">id</h2>

<pre><code><code> Title     : id
 Usage     : $contig-&gt;id(&quot;Ig&quot;)
 Function  : Gets/sets the id field of the alignment
 Returns   : An id string
 Argument  : An id string (optional)</code></code></pre>

<h2 id="missing_char">missing_char</h2>

<pre><code><code> Title     : missing_char
 Usage     : $contig-&gt;missing_char(&quot;?&quot;)
 Function  : Gets/sets the missing_char attribute of the alignment
             It is generally recommended to set it to &#39;n&#39; or &#39;N&#39;
             for nucleotides and to &#39;X&#39; for protein.
 Returns   : An missing_char string,
 Argument  : An missing_char string (optional)</code></code></pre>

<h2 id="match_char">match_char</h2>

<pre><code><code> Title     : match_char
 Usage     : $contig-&gt;match_char(&#39;.&#39;)
 Function  : Gets/sets the match_char attribute of the alignment
 Returns   : An match_char string,
 Argument  : An match_char string (optional)</code></code></pre>

<h2 id="gap_char">gap_char</h2>

<pre><code><code> Title     : gap_char
 Usage     : $contig-&gt;gap_char(&#39;-&#39;)
 Function  : Gets/sets the gap_char attribute of the alignment
 Returns   : An gap_char string, defaults to &#39;-&#39;
 Argument  : An gap_char string (optional)</code></code></pre>

<h2 id="symbol_chars">symbol_chars</h2>

<pre><code><code> <span class="variable">Title</span>   <span class="operator">:</span> <span class="variable">symbol_chars</span>
 <span class="variable">Usage</span>   <span class="operator">:</span> <span class="keyword">my</span> <span class="variable">@symbolchars</span> <span class="operator">=</span> <span class="variable">$contig</span><span class="operator">-&gt;</span><span class="variable">symbol_chars</span><span class="operator">;</span>
 <span class="variable">Function</span><span class="operator">:</span> <span class="variable">Returns</span> <span class="variable">all</span> <span class="variable">the</span> <span class="variable">seen</span> <span class="variable">symbols</span> <span class="operator">(</span><span class="variable">other</span> <span class="variable">than</span> <span class="variable">gaps</span><span class="operator">)</span>
 <span class="variable">Returns</span> <span class="operator">:</span> <span class="variable">array</span> <span class="variable">of</span> <span class="variable">characters</span> <span class="variable">that</span> <span class="variable">are</span> <span class="variable">the</span> <span class="variable">seen</span> <span class="variable">symbols</span>
 <span class="variable">Argument</span><span class="operator">:</span> <span class="variable">boolean</span> <span class="variable">to</span> <span class="variable">include</span> <span class="variable">the</span> <span class="variable">gap</span><span class="operator">/</span><span class="variable">missing</span><span class="operator">/</span><span class="variable">match</span> <span class="variable">characters</span>
</code></code></pre>

<h2 id="Alignment-descriptors">Alignment descriptors</h2>

<p>These read only methods describe the MSE in various ways.</p>

<h2 id="consensus_string">consensus_string</h2>

<pre><code><code> Title     : consensus_string
 Usage     : $str = $contig-&gt;consensus_string($threshold_percent)
 Function  : Makes a strict consensus
 Returns   :
 Argument  : Optional threshold ranging from 0 to 100.
             The consensus residue has to appear at least threshold %
             of the sequences at a given location, otherwise a &#39;?&#39;
             character will be placed at that location.
             (Default value = 0%)</code></code></pre>

<h2 id="consensus_iupac">consensus_iupac</h2>

<pre><code><code> Title     : consensus_iupac
 Usage     : $str = $contig-&gt;consensus_iupac()
 Function  :

             Makes a consensus using IUPAC ambiguity codes from DNA
             and RNA. The output is in upper case except when gaps in
             a column force output to be in lower case.

             Note that if your alignment sequences contain a lot of
             IUPAC ambiquity codes you often have to manually set
             alphabet.  Bio::PrimarySeq::_guess_type thinks they
             indicate a protein sequence.

 Returns   : consensus string
 Argument  : none
 Throws    : on protein sequences</code></code></pre>

<h2 id="is_flush">is_flush</h2>

<pre><code><code> Title     : is_flush
 Usage     : if( $contig-&gt;is_flush() )
           :
           :
 Function  : Tells you whether the alignment
           : is flush, ie all of the same length
           :
           :
 Returns   : 1 or 0
 Argument  :</code></code></pre>

<h2 id="length">length</h2>

<pre><code><code> Title     : length()
 Usage     : $len = $contig-&gt;length()
 Function  : Returns the maximum length of the alignment.
             To be sure the alignment is a block, use is_flush
 Returns   :
 Argument  :</code></code></pre>

<h2 id="maxname_length">maxname_length</h2>

<pre><code><code> Title     : maxname_length
 Usage     : $contig-&gt;maxname_length()
 Function  :

             Gets the maximum length of the displayname in the
             alignment. Used in writing out various MSE formats.

 Returns   : integer
 Argument  :</code></code></pre>

<h2 id="num_residues">num_residues</h2>

<pre><code><code> Title     : num_residues
 Usage     : $no = $contig-&gt;num_residues
 Function  : number of residues in total in the alignment
 Returns   : integer
 Argument  :
 Note      : replaces no_residues</code></code></pre>

<h2 id="num_sequences">num_sequences</h2>

<pre><code><code> Title     : num_sequences
 Usage     : $depth = $contig-&gt;num_sequences
 Function  : number of sequence in the sequence alignment
 Returns   : integer
 Argument  : None
 Note      : replaces no_sequences</code></code></pre>

<h2 id="percentage_identity">percentage_identity</h2>

<pre><code><code> Title   : percentage_identity
 Usage   : $id = $contig-&gt;percentage_identity
 Function: The function calculates the percentage identity of the alignment
 Returns : The percentage identity of the alignment (as defined by the
                             implementation)
 Argument: None</code></code></pre>

<h2 id="overall_percentage_identity">overall_percentage_identity</h2>

<pre><code><code> Title   : percentage_identity
 Usage   : $id = $contig-&gt;percentage_identity
 Function: The function calculates the percentage identity of
           the conserved columns
 Returns : The percentage identity of the conserved columns
 Args    : None</code></code></pre>

<h2 id="average_percentage_identity">average_percentage_identity</h2>

<pre><code><code> Title   : average_percentage_identity
 Usage   : $id = $contig-&gt;average_percentage_identity
 Function: The function uses a fast method to calculate the average
           percentage identity of the alignment
 Returns : The average percentage identity of the alignment
 Args    : None</code></code></pre>

<h2 id="Alignment-positions">Alignment positions</h2>

<p>Methods to map a sequence position into an alignment column and back. column_from_residue_number() does the former. The latter is really a property of the sequence object and can done using <a>Bio::LocatableSeq::location_from_column</a>:</p>

<pre><code><code>    <span class="comment"># select somehow a sequence from the alignment, e.g.</span>
    <span class="keyword">my</span> <span class="variable">$seq</span> <span class="operator">=</span> <span class="variable">$contig</span><span class="operator">-&gt;</span><span class="variable">get_seq_by_pos</span><span class="operator">(</span><span class="number">1</span><span class="operator">);</span>
    <span class="comment">#$loc is undef or Bio::LocationI object</span>
    <span class="keyword">my</span> <span class="variable">$loc</span> <span class="operator">=</span> <span class="variable">$seq</span><span class="operator">-&gt;</span><span class="variable">location_from_column</span><span class="operator">(</span><span class="number">5</span><span class="operator">);</span>
</code></code></pre>

<h2 id="column_from_residue_number">column_from_residue_number</h2>

<pre><code><code> Title   : column_from_residue_number
 Usage   : $col = $contig-&gt;column_from_residue_number( $seqname, $resnumber)
 Function:

           This function gives the position in the alignment
           (i.e. column number) of the given residue number in the
           sequence with the given name. For example, for the
           alignment

           Seq1/91-97 AC..DEF.GH
           Seq2/24-30 ACGG.RTY..
           Seq3/43-51 AC.DDEFGHI

           column_from_residue_number( &quot;Seq1&quot;, 94 ) returns 5.
           column_from_residue_number( &quot;Seq2&quot;, 25 ) returns 2.
           column_from_residue_number( &quot;Seq3&quot;, 50 ) returns 9.

           An exception is thrown if the residue number would lie
           outside the length of the aligment
           (e.g. column_from_residue_number( &quot;Seq2&quot;, 22 )

      Note: If the the parent sequence is represented by more than
      one alignment sequence and the residue number is present in
      them, this method finds only the first one.

 Returns : A column number for the position in the alignment of the
           given residue in the given sequence (1 = first column)
 Args    : A sequence id/name (not a name/start-end)
           A residue number in the whole sequence (not just that
           segment of it in the alignment)</code></code></pre>

<h2 id="Sequence-names">Sequence names</h2>

<p>Methods to manipulate the display name. The default name based on the sequence id and subsequence positions can be overridden in various ways.</p>

<h2 id="displayname">displayname</h2>

<pre><code><code> Title     : displayname
 Usage     : $contig-&gt;displayname(&quot;Ig&quot;, &quot;IgA&quot;)
 Function  : Gets/sets the display name of a sequence in the alignment
           :
 Returns   : A display name string
 Argument  : name of the sequence
             displayname of the sequence (optional)</code></code></pre>

<h2 id="set_displayname_count">set_displayname_count</h2>

<pre><code><code> Title     : set_displayname_count
 Usage     : $contig-&gt;set_displayname_count
 Function  :

             Sets the names to be name_# where # is the number of
             times this name has been used.

 Returns   : None
 Argument  : None</code></code></pre>

<h2 id="set_displayname_flat">set_displayname_flat</h2>

<pre><code><code> Title     : set_displayname_flat
 Usage     : $contig-&gt;set_displayname_flat()
 Function  : Makes all the sequences be displayed as just their name,
             not name/start-end
 Returns   : 1
 Argument  : None</code></code></pre>

<h2 id="set_displayname_normal">set_displayname_normal</h2>

<pre><code><code> Title     : set_displayname_normal
 Usage     : $contig-&gt;set_displayname_normal()
 Function  : Makes all the sequences be displayed as name/start-end
 Returns   : None
 Argument  : None</code></code></pre>

<h1 id="Internal-Methods">Internal Methods</h1>

<h2 id="binary_search">_binary_search</h2>

<pre><code><code> Title     : _binary_search
 Usage     : _binary_search($list,$query)
 Function  :

             Find a number in a sorted list of numbers.  Return values
             may be on or two integers. One positive integer or zero
             (&gt;=0) is the index of the element that stores the queried
             value.  Two positive integers (or zero and another
             number) are the indexes of elements among which the
             queried value should be placed. Negative single values
             mean:

             -1: $query is smaller than smallest element in list
             -2: $query is greater than greatest element in list

 Returns   : array of integers
 Argument  :
             $list  : array reference
             $query : integer</code></code></pre>

<h2 id="compare">_compare</h2>

<pre><code><code>    Title   : _compare
    Usage   : _compare($arg1,$arg2)
    Function: Perform numeric or string comparisons
    Returns : integer (0, 1 or -1)
    Args    : values to be compared</code></code></pre>

<h2 id="nof_gaps">_nof_gaps</h2>

<pre><code><code>    Title   : _nof_gaps
    Usage   : _nof_gaps($array_ref, $query)
    Function: number of gaps found before position $query
    Returns : integer
    Args    :
              $array_ref : gap registry reference
              $query     : [integer] a position in a sequence</code></code></pre>

<h2 id="padded_unpadded">_padded_unpadded</h2>

<pre><code><code>    Title   : _padded_unpadded
    Usage   : _padded_unpadded($array_ref, $query)
    Function:

              Returns a coordinate corresponding to
              position $query after gaps were
              removed from a sequence.

    Returns : integer
    Args    :
              $array_ref : reference to this gap registry
              $query     : [integer] coordionate to change</code></code></pre>

<h2 id="unpadded_padded">_unpadded_padded</h2>

<pre><code><code>    Title   : _unpadded_padded
    Usage   : _unpadded_padded($array_ref, $query)
    Function:

              Returns the value corresponding to
              ungapped position $query when gaps are
              counted as valid sites in a sequence

    Returns :
    Args    : $array_ref = a reference to this sequence&#39;s gap registry
              $query = [integer] location to change</code></code></pre>

<h2 id="register_gaps">_register_gaps</h2>

<pre><code><code>    Title   : _register_gaps
    Usage   : $self-&gt;_register_gaps($seq, $array_ref)
    Function: stores gap locations for a sequence
    Returns : number of gaps found
    Args    :
              $seq       : sequence string
              $array_ref : a reference to an array,
                           where gap locations will
                           be stored</code></code></pre>

<h1 id="Deprecated-methods">Deprecated methods</h1>

<h2 id="no_residues">no_residues</h2>

<pre><code><code> Title     : no_residues
 Usage     : $no = $ali-&gt;no_residues
 Function  : number of residues in total in the alignment
 Returns   : integer
 Argument  :
 Note      : deprecated in favor of num_residues() </code></code></pre>

<h2 id="no_sequences">no_sequences</h2>

<pre><code><code> Title     : no_sequences
 Usage     : $depth = $ali-&gt;no_sequences
 Function  : number of sequence in the sequence alignment
 Returns   : integer
 Argument  :
 Note      : deprecated in favor of num_sequences()</code></code></pre>


</body>

</html>


