<?xml version="1.0" ?>
<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.0 Transitional//EN">
<html>
<head>
<!-- saved from url=(0017)http://localhost/ -->
<script language="JavaScript" src="../../../../displayToc.js"></script>
<script language="JavaScript" src="../../../../tocParas.js"></script>
<script language="JavaScript" src="../../../../tocTab.js"></script>
<link rel="stylesheet" type="text/css" href="../../../../scineplex.css">
<title></title>
<link rel="stylesheet" href="../../../../Active.css" type="text/css" />
<meta http-equiv="content-type" content="text/html; charset=utf-8" />
<link rev="made" href="mailto:support@ActiveState.com" />
</head>

<body>



<ul id="index">
  <li><a href="#NAME">NAME</a></li>
  <li><a href="#SYNOPSIS">SYNOPSIS</a></li>
  <li><a href="#DESCRIPTION">DESCRIPTION</a></li>
  <li><a href="#FEEDBACK">FEEDBACK</a>
    <ul>
      <li><a href="#Mailing-Lists">Mailing Lists</a></li>
      <li><a href="#Support">Support</a></li>
      <li><a href="#Reporting-Bugs">Reporting Bugs</a></li>
    </ul>
  </li>
  <li><a href="#AUTHOR---Robson-Francisco-de-Souza">AUTHOR - Robson Francisco de Souza</a></li>
  <li><a href="#APPENDIX">APPENDIX</a></li>
  <li><a href="#Accessing-general-assembly-data">Accessing general assembly data</a>
    <ul>
      <li><a href="#get_nof_contigs">get_nof_contigs</a></li>
      <li><a href="#get_nof_singlets">get_nof_singlets</a></li>
      <li><a href="#get_contig_ids">get_contig_ids</a></li>
      <li><a href="#get_singlet_ids">get_singlet_ids</a></li>
      <li><a href="#get_contig_by_id">get_contig_by_id</a></li>
      <li><a href="#get_singlet_by_id">get_singlet_by_id</a></li>
    </ul>
  </li>
  <li><a href="#Modifier-methods">Modifier methods</a>
    <ul>
      <li><a href="#add_contig">add_contig</a></li>
      <li><a href="#add_singlet">add_singlet</a></li>
      <li><a href="#remove_contigs">remove_contigs</a></li>
      <li><a href="#remove_singlets">remove_singlets</a></li>
    </ul>
  </li>
  <li><a href="#Contig-and-singlet-selection-methos">Contig and singlet selection methos</a>
    <ul>
      <li><a href="#select_contigs">select_contigs</a></li>
      <li><a href="#select_singlets">select_singlets</a></li>
      <li><a href="#all_contigs">all_contigs</a></li>
      <li><a href="#all_singlets">all_singlets</a></li>
    </ul>
  </li>
</ul>

<h1 id="NAME">NAME</h1>

<p>Bio::Assembly::ScaffoldI - Abstract Inteface of Sequence Assemblies</p>

<h1 id="SYNOPSIS">SYNOPSIS</h1>

<pre><code><code>    <span class="comment"># get a Bio::Assembly::ScaffoldI object somehow</span>
    
    <span class="keyword">foreach</span> <span class="keyword">my</span> <span class="variable">$contig</span> <span class="operator">(</span><span class="variable">$assembly</span><span class="operator">-&gt;</span><span class="variable">all_contigs</span><span class="operator">)</span> <span class="operator">{</span>
       <span class="comment"># do something (see Bio::Assembly::Contig)</span>
    <span class="operator">}</span>
</code></code></pre>

<h1 id="DESCRIPTION">DESCRIPTION</h1>

<p>This interface defines the basic set of methods an object should have to manipulate assembly data.</p>

<h1 id="FEEDBACK">FEEDBACK</h1>

<h2 id="Mailing-Lists">Mailing Lists</h2>

<p>User feedback is an integral part of the evolution of this and other Bioperl modules. Send your comments and suggestions preferably to the Bioperl mailing lists Your participation is much appreciated.</p>

<pre><code><code>  bioperl-l@bioperl.org                  - General discussion
  http://bioperl.org/wiki/Mailing_lists  - About the mailing lists</code></code></pre>

<h2 id="Support">Support</h2>

<p>Please direct usage questions or support issues to the mailing list:</p>

<p><i>bioperl-l@bioperl.org</i></p>

<p>rather than to the module maintainer directly. Many experienced and reponsive experts will be able look at the problem and quickly address it. Please include a thorough description of the problem with code and data examples if at all possible.</p>

<h2 id="Reporting-Bugs">Reporting Bugs</h2>

<p>Report bugs to the Bioperl bug tracking system to help us keep track the bugs and their resolution. Bug reports can be submitted via the web:</p>

<pre><code><code>  https://redmine.open-bio.org/projects/bioperl/</code></code></pre>

<h1 id="AUTHOR---Robson-Francisco-de-Souza">AUTHOR - Robson Francisco de Souza</h1>

<p>Email: rfsouza@citri.iq.usp.br</p>

<h1 id="APPENDIX">APPENDIX</h1>

<p>The rest of the documentation details each of the object methods. Internal methods are usually preceded with a _</p>

<h1 id="Accessing-general-assembly-data">Accessing general assembly data</h1>

<h2 id="get_nof_contigs">get_nof_contigs</h2>

<pre><code><code>        Title   : get_nof_contigs
        Usage   : $assembly-&gt;get_nof_contigs()
        Function: Get the number of contigs included in the assembly
        Returns : integer
        Args    : none</code></code></pre>

<h2 id="get_nof_singlets">get_nof_singlets</h2>

<pre><code><code>        Title   : get_nof_singlets
        Usage   : $assembly-&gt;get_nof_singlets()
        Function: Get the number of singlets included in the assembly
        Returns : integer
        Args    : none</code></code></pre>

<h2 id="get_contig_ids">get_contig_ids</h2>

<pre><code><code>        Title   : get_contig_ids
        Usage   : $assembly-&gt;get_contig_ids()
        Function: Access list of contig IDs from assembly
        Returns : an array if there are any contigs in the assembly.
                  undef otherwise
        Args    : an array of contig IDs</code></code></pre>

<h2 id="get_singlet_ids">get_singlet_ids</h2>

<pre><code><code>        Title   : get_singlet_ids
        Usage   : $assembly-&gt;get_singlet_ids()
        Function: Access list of singlet IDs from assembly
        Returns : an array if there are any singlets in the assembly.
                  undef otherwise
        Args    : an array of singlet IDs</code></code></pre>

<h2 id="get_contig_by_id">get_contig_by_id</h2>

<pre><code><code>    Title   : get_contig_by_id
    Usage   : $assembly-&gt;get_contig_by_id($id)
    Function: Get a reference for a contig from the assembly
    Returns : a Bio::Assembly::Contig object or undef
    Args    : [string] contig unique identifier (ID)</code></code></pre>

<h2 id="get_singlet_by_id">get_singlet_by_id</h2>

<pre><code><code>    Title   : get_singlet_by_id
    Usage   : $assembly-&gt;get_singlet_by_id()
    Function: Get a reference for a singlet from the assembly
    Returns : Bio::Assembly::Singlet object or undef
    Args    : [string] a singlet ID</code></code></pre>

<h1 id="Modifier-methods">Modifier methods</h1>

<p>Implementation of these methods is optional in the sense that read-only implementations may not have these. If an object implements one of them, it should however implement all.</p>

<h2 id="add_contig">add_contig</h2>

<pre><code><code>        Title   : add_contig
        Usage   : $assembly-&gt;add_contig($contig)
        Function: Add another contig to the Bio::Assembly::ScaffoldI object
        Returns : 1 on success, 0 otherwise
        Args    : a Bio::Assembly:Contig object

    See Bio::Assembly::Contig for more information</code></code></pre>

<h2 id="add_singlet">add_singlet</h2>

<pre><code><code>        Title   : add_singlet
        Usage   : $assembly-&gt;add_singlet($seq)
        Function: Add another singlet to the Bio::Assembly::ScaffoldI object
        Returns : 1 on success, 0 otherwise
        Args    : a Bio::Assembly::Singlet object</code></code></pre>

<h2 id="remove_contigs">remove_contigs</h2>

<pre><code><code>        Title   : remove_contigs
        Usage   : $assembly-&gt;remove_contigs(1..4)
        Function: Remove contig from assembly object
        Returns : a Bio::Assembly::Contig object
        Args    : a list of contig IDs

    See function get_contig_ids() above</code></code></pre>

<h2 id="remove_singlets">remove_singlets</h2>

<pre><code><code>        Title   : remove_singlets
        Usage   : $assembly-&gt;remove_singlets(1..4)
        Function: Remove singlets from assembly object
        Returns : an array of Bio::Assembly::Singlet objects
        Args    : an array of singlet IDs 

    See function get_singlet_ids() above</code></code></pre>

<h1 id="Contig-and-singlet-selection-methos">Contig and singlet selection methos</h1>

<h2 id="select_contigs">select_contigs</h2>

<pre><code><code>        Title   : select_contig
        Usage   : $assembly-&gt;select_contig
        Function: Selects an array of contigs from the assembly
        Returns : an array of Bio::Assembly::Contig objects
        Args    : an array of contig ids

    See function get_contig_ids() above</code></code></pre>

<h2 id="select_singlets">select_singlets</h2>

<pre><code><code>        Title   : select_singlets
        Usage   : $assembly-&gt;select_singlets(@list)
        Function: Selects an array of singlets from the assembly
        Returns : an array of Bio::Assembly::Singlet objects
        Args    : an array of singlet ids

    See function get_singlet_ids() above</code></code></pre>

<h2 id="all_contigs">all_contigs</h2>

<pre><code><code>        Title   : all_contigs
        Usage   : my @contigs = $assembly-&gt;all_contigs
        Function: Returns a list of all contigs in this assembly.
                  Contigs are both clusters and alignments of one
                  or more reads, with an associated consensus
                  sequence.
        Returns : array of Bio::Assembly::Contig
        Args    : none</code></code></pre>

<h2 id="all_singlets">all_singlets</h2>

<pre><code><code>    Title   : all_singlets
    Usage   : my @singlets = $assembly-&gt;all_singlets
    Function: Returns a list of all singlets in this assembly.
              Singlets are isolated reads, without non-vector
              matches to any other read in the assembly.
    Returns : array of Bio::Assembly::Singlet objects
    Args    : none</code></code></pre>


</body>

</html>


