<?xml version="1.0" ?>
<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.0 Transitional//EN">
<html>
<head>
<!-- saved from url=(0017)http://localhost/ -->
<script language="JavaScript" src="../../../../displayToc.js"></script>
<script language="JavaScript" src="../../../../tocParas.js"></script>
<script language="JavaScript" src="../../../../tocTab.js"></script>
<link rel="stylesheet" type="text/css" href="../../../../scineplex.css">
<title></title>
<link rel="stylesheet" href="../../../../Active.css" type="text/css" />
<meta http-equiv="content-type" content="text/html; charset=utf-8" />
<link rev="made" href="mailto:support@ActiveState.com" />
</head>

<body>



<ul id="index">
  <li><a href="#NAME">NAME</a></li>
  <li><a href="#SYNOPSIS">SYNOPSIS</a></li>
  <li><a href="#DESCRIPTION">DESCRIPTION</a></li>
  <li><a href="#FEEDBACK">FEEDBACK</a>
    <ul>
      <li><a href="#Mailing-Lists">Mailing Lists</a></li>
      <li><a href="#Support">Support</a></li>
      <li><a href="#Reporting-Bugs">Reporting Bugs</a></li>
    </ul>
  </li>
  <li><a href="#AUTHOR---Jason-Stajich">AUTHOR - Jason Stajich</a></li>
  <li><a href="#CONTRIBUTORS">CONTRIBUTORS</a></li>
  <li><a href="#APPENDIX">APPENDIX</a></li>
  <li><a href="#Modifier-methods">Modifier methods</a>
    <ul>
      <li><a href="#add_seq">add_seq</a></li>
      <li><a href="#remove_seq">remove_seq</a></li>
      <li><a href="#purge">purge</a></li>
      <li><a href="#sort_alphabetically">sort_alphabetically</a></li>
    </ul>
  </li>
  <li><a href="#Sequence-selection-methods">Sequence selection methods</a>
    <ul>
      <li><a href="#each_seq">each_seq</a></li>
      <li><a href="#each_alphabetically">each_alphabetically</a></li>
      <li><a href="#each_seq_with_id">each_seq_with_id</a></li>
      <li><a href="#get_seq_by_pos">get_seq_by_pos</a></li>
    </ul>
  </li>
  <li><a href="#Create-new-alignments">Create new alignments</a>
    <ul>
      <li><a href="#select">select</a></li>
      <li><a href="#select_noncont">select_noncont</a></li>
      <li><a href="#slice">slice</a></li>
    </ul>
  </li>
  <li><a href="#Change-sequences-within-the-MSE">Change sequences within the MSE</a>
    <ul>
      <li><a href="#map_chars">map_chars</a></li>
      <li><a href="#uppercase">uppercase</a></li>
      <li><a href="#match_line">match_line</a></li>
      <li><a href="#match">match</a></li>
      <li><a href="#unmatch">unmatch</a></li>
    </ul>
  </li>
  <li><a href="#MSE-attibutes">MSE attibutes</a>
    <ul>
      <li><a href="#id">id</a></li>
      <li><a href="#missing_char">missing_char</a></li>
      <li><a href="#match_char">match_char</a></li>
      <li><a href="#gap_char">gap_char</a></li>
      <li><a href="#symbol_chars">symbol_chars</a></li>
    </ul>
  </li>
  <li><a href="#Alignment-descriptors">Alignment descriptors</a>
    <ul>
      <li><a href="#consensus_string">consensus_string</a></li>
      <li><a href="#consensus_iupac">consensus_iupac</a></li>
      <li><a href="#is_flush">is_flush</a></li>
      <li><a href="#length">length</a></li>
      <li><a href="#maxname_length">maxname_length</a></li>
      <li><a href="#num_residues">num_residues</a></li>
      <li><a href="#num_sequences">num_sequences</a></li>
      <li><a href="#percentage_identity">percentage_identity</a></li>
      <li><a href="#overall_percentage_identity">overall_percentage_identity</a></li>
      <li><a href="#average_percentage_identity">average_percentage_identity</a></li>
    </ul>
  </li>
  <li><a href="#Alignment-positions">Alignment positions</a>
    <ul>
      <li><a href="#column_from_residue_number">column_from_residue_number</a></li>
    </ul>
  </li>
  <li><a href="#Sequence-names">Sequence names</a>
    <ul>
      <li><a href="#displayname">displayname</a></li>
      <li><a href="#set_displayname_count">set_displayname_count</a></li>
      <li><a href="#set_displayname_flat">set_displayname_flat</a></li>
      <li><a href="#set_displayname_normal">set_displayname_normal</a></li>
    </ul>
  </li>
  <li><a href="#Deprecated-methods">Deprecated methods</a>
    <ul>
      <li><a href="#no_residues">no_residues</a></li>
      <li><a href="#no_sequences">no_sequences</a></li>
    </ul>
  </li>
</ul>

<h1 id="NAME">NAME</h1>

<p>Bio::Align::AlignI - An interface for describing sequence alignments.</p>

<h1 id="SYNOPSIS">SYNOPSIS</h1>

<pre><code><code>  <span class="comment"># get a Bio::Align::AlignI somehow - typically using Bio::AlignIO system</span>
  <span class="comment"># some descriptors</span>
  <span class="keyword">print</span> <span class="variable">$aln</span><span class="operator">-&gt;</span><span class="variable">length</span><span class="operator">,</span> <span class="string">"\n"</span><span class="operator">;</span>
  <span class="keyword">print</span> <span class="variable">$aln</span><span class="operator">-&gt;</span><span class="variable">num_residues</span><span class="operator">,</span> <span class="string">"\n"</span><span class="operator">;</span>
  <span class="keyword">print</span> <span class="variable">$aln</span><span class="operator">-&gt;</span><span class="variable">is_flush</span><span class="operator">,</span> <span class="string">"\n"</span><span class="operator">;</span>
  <span class="keyword">print</span> <span class="variable">$aln</span><span class="operator">-&gt;</span><span class="variable">num_sequences</span><span class="operator">,</span> <span class="string">"\n"</span><span class="operator">;</span>
  <span class="keyword">print</span> <span class="variable">$aln</span><span class="operator">-&gt;</span><span class="variable">percentage_identity</span><span class="operator">,</span> <span class="string">"\n"</span><span class="operator">;</span>
  <span class="keyword">print</span> <span class="variable">$aln</span><span class="operator">-&gt;</span><span class="variable">consensus_string</span><span class="operator">(</span><span class="number">50</span><span class="operator">),</span> <span class="string">"\n"</span><span class="operator">;</span>
  
  <span class="comment"># find the position in the alignment for a sequence location</span>
  <span class="variable">$pos</span> <span class="operator">=</span> <span class="variable">$aln</span><span class="operator">-&gt;</span><span class="variable">column_from_residue_number</span><span class="operator">(</span><span class="string">'1433_LYCES'</span><span class="operator">,</span> <span class="number">14</span><span class="operator">);</span> <span class="comment"># = 6;</span>
  
  <span class="comment"># extract sequences and check values for the alignment column $pos</span>
  <span class="keyword">foreach</span> <span class="variable">$seq</span> <span class="operator">(</span><span class="variable">$aln</span><span class="operator">-&gt;</span><span class="variable">each_seq</span><span class="operator">)</span> <span class="operator">{</span>
      <span class="variable">$res</span> <span class="operator">=</span> <span class="variable">$seq</span><span class="operator">-&gt;</span><span class="variable">subseq</span><span class="operator">(</span><span class="variable">$pos</span><span class="operator">,</span> <span class="variable">$pos</span><span class="operator">);</span>
      <span class="variable">$count</span><span class="operator">{</span><span class="variable">$res</span><span class="operator">}</span><span class="operator">++;</span>
  <span class="operator">}</span>
  <span class="keyword">foreach</span> <span class="variable">$res</span> <span class="operator">(</span><span class="keyword">keys</span> <span class="variable">%count</span><span class="operator">)</span> <span class="operator">{</span>
      <span class="keyword">printf</span> <span class="string">"Res: %s  Count: %2d\n"</span><span class="operator">,</span> <span class="variable">$res</span><span class="operator">,</span> <span class="variable">$count</span><span class="operator">{</span><span class="variable">$res</span><span class="operator">}</span><span class="operator">;</span>
  <span class="operator">}</span>
</code></code></pre>

<h1 id="DESCRIPTION">DESCRIPTION</h1>

<p>This interface describes the basis for alignment objects.</p>

<h1 id="FEEDBACK">FEEDBACK</h1>

<h2 id="Mailing-Lists">Mailing Lists</h2>

<p>User feedback is an integral part of the evolution of this and other Bioperl modules. Send your comments and suggestions preferably to the Bioperl mailing list. Your participation is much appreciated.</p>

<pre><code><code>  bioperl-l@bioperl.org                  - General discussion
  http://bioperl.org/wiki/Mailing_lists  - About the mailing lists</code></code></pre>

<h2 id="Support">Support</h2>

<p>Please direct usage questions or support issues to the mailing list:</p>

<p><i>bioperl-l@bioperl.org</i></p>

<p>rather than to the module maintainer directly. Many experienced and reponsive experts will be able look at the problem and quickly address it. Please include a thorough description of the problem with code and data examples if at all possible.</p>

<h2 id="Reporting-Bugs">Reporting Bugs</h2>

<p>Report bugs to the Bioperl bug tracking system to help us keep track of the bugs and their resolution. Bug reports can be submitted via the web:</p>

<pre><code><code>  https://redmine.open-bio.org/projects/bioperl/</code></code></pre>

<h1 id="AUTHOR---Jason-Stajich">AUTHOR - Jason Stajich</h1>

<p>Email jason@bioperl.org</p>

<h1 id="CONTRIBUTORS">CONTRIBUTORS</h1>

<p>Ewan Birney, birney@ebi.ac.uk Heikki Lehvaslaiho, heikki-at-bioperl-dot-org</p>

<h1 id="APPENDIX">APPENDIX</h1>

<p>The rest of the documentation details each of the object methods. Internal methods are usually preceded with a _</p>

<h1 id="Modifier-methods">Modifier methods</h1>

<p>These methods modify the MSE by adding, removing or shuffling complete sequences.</p>

<h2 id="add_seq">add_seq</h2>

<pre><code><code> <span class="variable">Title</span>     <span class="operator">:</span> <span class="variable">add_seq</span>
 <span class="variable">Usage</span>     <span class="operator">:</span> <span class="variable">$myalign</span><span class="operator">-&gt;</span><span class="variable">add_seq</span><span class="operator">(</span><span class="variable">$newseq</span><span class="operator">);</span>
 <span class="variable">Function</span>  <span class="operator">:</span> <span class="variable">Adds</span> <span class="variable">another</span> <span class="variable">sequence</span> <span class="variable">to</span> <span class="variable">the</span> <span class="variable">alignment</span><span class="operator">.</span> <span class="variable">*Does</span> <span class="keyword">not</span><span class="operator">*</span> <span class="variable">align</span>
             <span class="variable">it</span> <span class="operator">-</span> <span class="variable">just</span> <span class="variable">adds</span> <span class="variable">it</span> <span class="variable">to</span> <span class="variable">the</span> <span class="variable">hashes</span><span class="operator">.</span>
 <span class="variable">Returns</span>   <span class="operator">:</span> <span class="variable">None</span>
 <span class="variable">Argument</span>  <span class="operator">:</span> <span class="variable">a</span> <span class="variable">Bio::LocatableSeq</span> <span class="variable">object</span>
             <span class="variable">order</span> <span class="operator">(</span><span class="variable">optional</span><span class="operator">)</span>
</code></code></pre>

<p>See <a href="../../../../../.cpan/build/BioPerl-1.6.901-ghKY8U/blib/lib/Bio/LocatableSeq.html">Bio::LocatableSeq</a> for more information.</p>

<h2 id="remove_seq">remove_seq</h2>

<pre><code><code> <span class="variable">Title</span>     <span class="operator">:</span> <span class="variable">remove_seq</span>
 <span class="variable">Usage</span>     <span class="operator">:</span> <span class="variable">$aln</span><span class="operator">-&gt;</span><span class="variable">remove_seq</span><span class="operator">(</span><span class="variable">$seq</span><span class="operator">);</span>
 <span class="variable">Function</span>  <span class="operator">:</span> <span class="variable">Removes</span> <span class="variable">a</span> <span class="variable">single</span> <span class="variable">sequence</span> <span class="variable">from</span> <span class="variable">an</span> <span class="variable">alignment</span>
 <span class="variable">Returns</span>   <span class="operator">:</span>
 <span class="variable">Argument</span>  <span class="operator">:</span> <span class="variable">a</span> <span class="variable">Bio::LocatableSeq</span> <span class="variable">object</span>
</code></code></pre>

<h2 id="purge">purge</h2>

<pre><code><code> <span class="variable">Title</span>   <span class="operator">:</span> <span class="variable">purge</span>
 <span class="variable">Usage</span>   <span class="operator">:</span> <span class="variable">$aln</span><span class="operator">-&gt;</span><span class="variable">purge</span><span class="operator">(</span><span class="number">0</span><span class="operator">.</span><span class="number">7</span><span class="operator">);</span>
 <span class="variable">Function</span><span class="operator">:</span>
 
           <span class="variable">Removes</span> <span class="variable">sequences</span> <span class="variable">above</span> <span class="variable">whatever</span> <span class="variable">%id</span><span class="operator">.</span>
 
           <span class="variable">This</span> <span class="variable">function</span> <span class="variable">will</span> <span class="variable">grind</span> <span class="variable">on</span> <span class="variable">large</span> <span class="variable">alignments</span><span class="operator">.</span> <span class="variable">Beware</span><span class="operator">!</span>
           <span class="operator">(</span><span class="variable">perhaps</span> <span class="keyword">not</span> <span class="variable">ideally</span> <span class="variable">implemented</span><span class="operator">)</span>
 
 <span class="variable">Example</span> <span class="operator">:</span>
 <span class="variable">Returns</span> <span class="operator">:</span> <span class="variable">An</span> <span class="variable">array</span> <span class="variable">of</span> <span class="variable">the</span> <span class="variable">removed</span> <span class="variable">sequences</span>
 <span class="variable">Argument</span><span class="operator">:</span>
</code></code></pre>

<h2 id="sort_alphabetically">sort_alphabetically</h2>

<pre><code><code> Title     : sort_alphabetically
 Usage     : $ali-&gt;sort_alphabetically
 Function  : 

             Changes the order of the alignment to alphabetical on name 
             followed by numerical by number.

 Returns   : an array
 Argument  : </code></code></pre>

<h1 id="Sequence-selection-methods">Sequence selection methods</h1>

<p>Methods returning one or more sequences objects.</p>

<h2 id="each_seq">each_seq</h2>

<pre><code><code> Title     : each_seq
 Usage     : foreach $seq ( $align-&gt;each_seq() ) 
 Function  : Gets an array of Seq objects from the alignment
 Returns   : an array
 Argument  : </code></code></pre>

<h2 id="each_alphabetically">each_alphabetically</h2>

<pre><code><code> Title     : each_alphabetically
 Usage     : foreach $seq ( $ali-&gt;each_alphabetically() )
 Function  :

             Returns an array of sequence object sorted alphabetically 
             by name and then by start point.
             Does not change the order of the alignment

 Returns   : 
 Argument  : </code></code></pre>

<h2 id="each_seq_with_id">each_seq_with_id</h2>

<pre><code><code> Title     : each_seq_with_id
 Usage     : foreach $seq ( $align-&gt;each_seq_with_id() ) 
 Function  : 

             Gets an array of Seq objects from the
             alignment, the contents being those sequences
             with the given name (there may be more than one)

 Returns   : an array
 Argument  : a seq name</code></code></pre>

<h2 id="get_seq_by_pos">get_seq_by_pos</h2>

<pre><code><code> Title     : get_seq_by_pos
 Usage     : $seq = $aln-&gt;get_seq_by_pos(3) # third sequence from the alignment
 Function  : 

             Gets a sequence based on its position in the alignment.
             Numbering starts from 1.  Sequence positions larger than
             num_sequences() will throw an error.

 Returns   : a Bio::LocatableSeq object
 Argument  : positive integer for the sequence position</code></code></pre>

<h1 id="Create-new-alignments">Create new alignments</h1>

<p>The result of these methods are horizontal or vertical subsets of the current MSE.</p>

<h2 id="select">select</h2>

<pre><code><code> Title     : select
 Usage     : $aln2 = $aln-&gt;select(1, 3) # three first sequences
 Function  : 

             Creates a new alignment from a continuous subset of
             sequences.  Numbering starts from 1.  Sequence positions
             larger than num_sequences() will throw an error.

 Returns   : a Bio::SimpleAlign object
 Argument  : positive integer for the first sequence
             positive integer for the last sequence to include (optional)</code></code></pre>

<h2 id="select_noncont">select_noncont</h2>

<pre><code><code> Title     : select_noncont
 Usage     : $aln2 = $aln-&gt;select_noncont(1, 3) # first and 3rd sequences
 Function  : 

             Creates a new alignment from a subset of
             sequences.  Numbering starts from 1.  Sequence positions
             larger than num_sequences() will throw an error.

 Returns   : a Bio::SimpleAlign object
 Args      : array of integers for the sequences</code></code></pre>

<h2 id="slice">slice</h2>

<pre><code><code> Title     : slice
 Usage     : $aln2 = $aln-&gt;slice(20, 30)
 Function  : 

             Creates a slice from the alignment inclusive of start and
             end columns.  Sequences with no residues in the slice are
             excluded from the new alignment and a warning is printed.
             Slice beyond the length of the sequence does not do
             padding.

 Returns   : a Bio::SimpleAlign object
 Argument  : positive integer for start column 
             positive integer for end column </code></code></pre>

<h1 id="Change-sequences-within-the-MSE">Change sequences within the MSE</h1>

<p>These methods affect characters in all sequences without changing the alignment.</p>

<h2 id="map_chars">map_chars</h2>

<pre><code><code> Title     : map_chars
 Usage     : $ali-&gt;map_chars(&#39;\.&#39;,&#39;-&#39;)
 Function  : 

             Does a s/$arg1/$arg2/ on the sequences. Useful for gap
             characters

             Notice that the from (arg1) is interpreted as a regex,
             so be careful about quoting meta characters (eg
             $ali-&gt;map_chars(&#39;.&#39;,&#39;-&#39;) wont do what you want)

 Returns   : None
 Argument  : &#39;from&#39; rexexp
             &#39;to&#39; string</code></code></pre>

<h2 id="uppercase">uppercase</h2>

<pre><code><code> Title     : uppercase()
 Usage     : $ali-&gt;uppercase()
 Function  : Sets all the sequences to uppercase
 Returns   : 
 Argument  : </code></code></pre>

<h2 id="match_line">match_line</h2>

<pre><code><code> Title    : match_line()
 Usage    : $align-&gt;match_line()
 Function : Generates a match line - much like consensus string
            except that a line indicating the &#39;*&#39; for a match.
 Argument : (optional) Match line characters (&#39;*&#39; by default)
            (optional) Strong match char (&#39;:&#39; by default)
            (optional) Weak match char (&#39;.&#39; by default)</code></code></pre>

<h2 id="match">match</h2>

<pre><code><code> Title     : match()
 Usage     : $ali-&gt;match()
 Function  : 

             Goes through all columns and changes residues that are
             identical to residue in first sequence to match &#39;.&#39;
             character. Sets match_char.

             USE WITH CARE: Most MSE formats do not support match
             characters in sequences, so this is mostly for output
             only. NEXUS format (Bio::AlignIO::nexus) can handle
             it.

 Returns   : 1
 Argument  : a match character, optional, defaults to &#39;.&#39;</code></code></pre>

<h2 id="unmatch">unmatch</h2>

<pre><code><code> Title     : unmatch()
 Usage     : $ali-&gt;unmatch()
 Function  : 

             Undoes the effect of method match. Unsets match_char.

 Returns   : 1
 Argument  : a match character, optional, defaults to &#39;.&#39;</code></code></pre>

<h1 id="MSE-attibutes">MSE attibutes</h1>

<p>Methods for setting and reading the MSE attributes.</p>

<p>Note that the methods defining character semantics depend on the user to set them sensibly. They are needed only by certain input/output methods. Unset them by setting to an empty string (&#39;&#39;).</p>

<h2 id="id">id</h2>

<pre><code><code> Title     : id
 Usage     : $myalign-&gt;id(&quot;Ig&quot;)
 Function  : Gets/sets the id field of the alignment
 Returns   : An id string
 Argument  : An id string (optional)</code></code></pre>

<h2 id="missing_char">missing_char</h2>

<pre><code><code> Title     : missing_char
 Usage     : $myalign-&gt;missing_char(&quot;?&quot;)
 Function  : Gets/sets the missing_char attribute of the alignment
             It is generally recommended to set it to &#39;n&#39; or &#39;N&#39; 
             for nucleotides and to &#39;X&#39; for protein. 
 Returns   : An missing_char string,
 Argument  : An missing_char string (optional)</code></code></pre>

<h2 id="match_char">match_char</h2>

<pre><code><code> Title     : match_char
 Usage     : $myalign-&gt;match_char(&#39;.&#39;)
 Function  : Gets/sets the match_char attribute of the alignment
 Returns   : An match_char string,
 Argument  : An match_char string (optional)</code></code></pre>

<h2 id="gap_char">gap_char</h2>

<pre><code><code> Title     : gap_char
 Usage     : $myalign-&gt;gap_char(&#39;-&#39;)
 Function  : Gets/sets the gap_char attribute of the alignment
 Returns   : An gap_char string, defaults to &#39;-&#39;
 Argument  : An gap_char string (optional)</code></code></pre>

<h2 id="symbol_chars">symbol_chars</h2>

<pre><code><code> <span class="variable">Title</span>   <span class="operator">:</span> <span class="variable">symbol_chars</span>
 <span class="variable">Usage</span>   <span class="operator">:</span> <span class="keyword">my</span> <span class="variable">@symbolchars</span> <span class="operator">=</span> <span class="variable">$aln</span><span class="operator">-&gt;</span><span class="variable">symbol_chars</span><span class="operator">;</span>
 <span class="variable">Function</span><span class="operator">:</span> <span class="variable">Returns</span> <span class="variable">all</span> <span class="variable">the</span> <span class="variable">seen</span> <span class="variable">symbols</span> <span class="operator">(</span><span class="variable">other</span> <span class="variable">than</span> <span class="variable">gaps</span><span class="operator">)</span>
 <span class="variable">Returns</span> <span class="operator">:</span> <span class="variable">array</span> <span class="variable">of</span> <span class="variable">characters</span> <span class="variable">that</span> <span class="variable">are</span> <span class="variable">the</span> <span class="variable">seen</span> <span class="variable">symbols</span>
 <span class="variable">Argument</span><span class="operator">:</span> <span class="variable">boolean</span> <span class="variable">to</span> <span class="variable">include</span> <span class="variable">the</span> <span class="variable">gap</span><span class="operator">/</span><span class="variable">missing</span><span class="operator">/</span><span class="variable">match</span> <span class="variable">characters</span>
</code></code></pre>

<h1 id="Alignment-descriptors">Alignment descriptors</h1>

<p>These read only methods describe the MSE in various ways.</p>

<h2 id="consensus_string">consensus_string</h2>

<pre><code><code> Title     : consensus_string
 Usage     : $str = $ali-&gt;consensus_string($threshold_percent)
 Function  : Makes a strict consensus 
 Returns   : consensus string
 Argument  : Optional threshold ranging from 0 to 100.
             The consensus residue has to appear at least threshold %
             of the sequences at a given location, otherwise a &#39;?&#39;
             character will be placed at that location.
             (Default value = 0%)</code></code></pre>

<h2 id="consensus_iupac">consensus_iupac</h2>

<pre><code><code> Title     : consensus_iupac
 Usage     : $str = $ali-&gt;consensus_iupac()
 Function  : 

             Makes a consensus using IUPAC ambiguity codes from DNA
             and RNA. The output is in upper case except when gaps in
             a column force output to be in lower case.

             Note that if your alignment sequences contain a lot of
             IUPAC ambiquity codes you often have to manually set
             alphabet.  Bio::PrimarySeq::_guess_type thinks they
             indicate a protein sequence.

 Returns   : consensus string
 Argument  : none
 Throws    : on protein sequences</code></code></pre>

<h2 id="is_flush">is_flush</h2>

<pre><code><code> Title     : is_flush
 Usage     : if( $ali-&gt;is_flush() )
           : 
           :
 Function  : Tells you whether the alignment 
           : is flush, ie all of the same length
           : 
           :
 Returns   : 1 or 0
 Argument  : </code></code></pre>

<h2 id="length">length</h2>

<pre><code><code> Title     : length()
 Usage     : $len = $ali-&gt;length() 
 Function  : Returns the maximum length of the alignment.
             To be sure the alignment is a block, use is_flush
 Returns   : integer
 Argument  : </code></code></pre>

<h2 id="maxname_length">maxname_length</h2>

<pre><code><code> Title     : maxname_length
 Usage     : $ali-&gt;maxname_length()
 Function  : 

             Gets the maximum length of the displayname in the
             alignment. Used in writing out various MSE formats.

 Returns   : integer
 Argument  : </code></code></pre>

<h2 id="num_residues">num_residues</h2>

<pre><code><code> Title     : num_residues
 Usage     : $no = $ali-&gt;num_residues
 Function  : number of residues in total in the alignment
 Returns   : integer
 Argument  :
 Note      : replaces no_residues</code></code></pre>

<h2 id="num_sequences">num_sequences</h2>

<pre><code><code> Title     : num_sequences
 Usage     : $depth = $ali-&gt;num_sequences
 Function  : number of sequence in the sequence alignment
 Returns   : integer
 Argument  : None
 Note      : replaces no_sequences</code></code></pre>

<h2 id="percentage_identity">percentage_identity</h2>

<pre><code><code> Title   : percentage_identity
 Usage   : $id = $align-&gt;percentage_identity
 Function: The function calculates the percentage identity of the alignment
 Returns : The percentage identity of the alignment (as defined by the 
           implementation)
 Argument: None</code></code></pre>

<h2 id="overall_percentage_identity">overall_percentage_identity</h2>

<pre><code><code> Title   : overall_percentage_identity
 Usage   : $id = $align-&gt;overall_percentage_identity
 Function: The function calculates the percentage identity of 
           the conserved columns
 Returns : The percentage identity of the conserved columns
 Args    : None</code></code></pre>

<h2 id="average_percentage_identity">average_percentage_identity</h2>

<pre><code><code> Title   : average_percentage_identity
 Usage   : $id = $align-&gt;average_percentage_identity
 Function: The function uses a fast method to calculate the average 
           percentage identity of the alignment
 Returns : The average percentage identity of the alignment
 Args    : None</code></code></pre>

<h1 id="Alignment-positions">Alignment positions</h1>

<p>Methods to map a sequence position into an alignment column and back. column_from_residue_number() does the former. The latter is really a property of the sequence object and can done using <a>Bio::LocatableSeq::location_from_column</a>:</p>

<pre><code><code>    <span class="comment"># select somehow a sequence from the alignment, e.g.</span>
    <span class="keyword">my</span> <span class="variable">$seq</span> <span class="operator">=</span> <span class="variable">$aln</span><span class="operator">-&gt;</span><span class="variable">get_seq_by_pos</span><span class="operator">(</span><span class="number">1</span><span class="operator">);</span>
    <span class="comment">#$loc is undef or Bio::LocationI object</span>
    <span class="keyword">my</span> <span class="variable">$loc</span> <span class="operator">=</span> <span class="variable">$seq</span><span class="operator">-&gt;</span><span class="variable">location_from_column</span><span class="operator">(</span><span class="number">5</span><span class="operator">);</span>
</code></code></pre>

<h2 id="column_from_residue_number">column_from_residue_number</h2>

<pre><code><code> Title   : column_from_residue_number
 Usage   : $col = $ali-&gt;column_from_residue_number( $seqname, $resnumber)
 Function:

           This function gives the position in the alignment
           (i.e. column number) of the given residue number in the
           sequence with the given name. For example, for the
           alignment

             Seq1/91-97 AC..DEF.GH
             Seq2/24-30 ACGG.RTY..
             Seq3/43-51 AC.DDEFGHI

           column_from_residue_number( &quot;Seq1&quot;, 94 ) returns 6.
           column_from_residue_number( &quot;Seq2&quot;, 25 ) returns 2.
           column_from_residue_number( &quot;Seq3&quot;, 50 ) returns 9.

           An exception is thrown if the residue number would lie
           outside the length of the alignment
           (e.g. column_from_residue_number( &quot;Seq2&quot;, 22 )

          Note: If the parent sequence is represented by more than one
          alignment sequence and the residue number is present in
          them, this method finds only the first one.

 Returns : A column number for the position in the alignment of the
           given residue in the given sequence (1 = first column)
 Args    : A sequence id/name (not a name/start-end)
           A residue number in the whole sequence (not just that
           segment of it in the alignment)</code></code></pre>

<h1 id="Sequence-names">Sequence names</h1>

<p>Methods to manipulate the display name. The default name based on the sequence id and subsequence positions can be overridden in various ways.</p>

<h2 id="displayname">displayname</h2>

<pre><code><code> Title     : displayname
 Usage     : $myalign-&gt;displayname(&quot;Ig&quot;, &quot;IgA&quot;)
 Function  : Gets/sets the display name of a sequence in the alignment
           :
 Returns   : A display name string
 Argument  : name of the sequence
             displayname of the sequence (optional)</code></code></pre>

<h2 id="set_displayname_count">set_displayname_count</h2>

<pre><code><code> Title     : set_displayname_count
 Usage     : $ali-&gt;set_displayname_count
 Function  : 

             Sets the names to be name_# where # is the number of
             times this name has been used.

 Returns   : None 
 Argument  : None</code></code></pre>

<h2 id="set_displayname_flat">set_displayname_flat</h2>

<pre><code><code> Title     : set_displayname_flat
 Usage     : $ali-&gt;set_displayname_flat()
 Function  : Makes all the sequences be displayed as just their name,
             not name/start-end
 Returns   : 1
 Argument  : None</code></code></pre>

<h2 id="set_displayname_normal">set_displayname_normal</h2>

<pre><code><code> Title     : set_displayname_normal
 Usage     : $ali-&gt;set_displayname_normal() 
 Function  : Makes all the sequences be displayed as name/start-end
 Returns   : None
 Argument  : None</code></code></pre>

<h1 id="Deprecated-methods">Deprecated methods</h1>

<h2 id="no_residues">no_residues</h2>

<pre><code><code> Title     : no_residues
 Usage     : $no = $ali-&gt;no_residues
 Function  : number of residues in total in the alignment
 Returns   : integer
 Argument  : 
 Note      : deprecated in favor of num_residues()</code></code></pre>

<h2 id="no_sequences">no_sequences</h2>

<pre><code><code> Title     : no_sequences
 Usage     : $depth = $ali-&gt;no_sequences
 Function  : number of sequence in the sequence alignment
 Returns   : integer
 Argument  : None
 Note      : deprecated in favor of num_sequences() </code></code></pre>


</body>

</html>


