<?xml version="1.0" ?>
<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.0 Transitional//EN">
<html>
<head>
<!-- saved from url=(0017)http://localhost/ -->
<script language="JavaScript" src="../../../../displayToc.js"></script>
<script language="JavaScript" src="../../../../tocParas.js"></script>
<script language="JavaScript" src="../../../../tocTab.js"></script>
<link rel="stylesheet" type="text/css" href="../../../../scineplex.css">
<title></title>
<link rel="stylesheet" href="../../../../Active.css" type="text/css" />
<meta http-equiv="content-type" content="text/html; charset=utf-8" />
<link rev="made" href="mailto:support@ActiveState.com" />
</head>

<body>



<ul id="index">
  <li><a href="#NAME">NAME</a></li>
  <li><a href="#SYNOPSIS">SYNOPSIS</a></li>
  <li><a href="#DESCRIPTION">DESCRIPTION</a></li>
  <li><a href="#FEEDBACK">FEEDBACK</a>
    <ul>
      <li><a href="#Mailing-Lists">Mailing Lists</a></li>
      <li><a href="#Support">Support</a></li>
      <li><a href="#Reporting-Bugs">Reporting Bugs</a></li>
    </ul>
  </li>
  <li><a href="#AUTHOR---Simon-K.-Chan">AUTHOR - Simon K. Chan</a></li>
  <li><a href="#APPENDIX">APPENDIX</a>
    <ul>
      <li><a href="#process">_process</a></li>
      <li><a href="#process_sciobj">_process_sciobj</a></li>
      <li><a href="#process_contig">_process_contig</a></li>
      <li><a href="#process_fragment_order">_process_fragment_order</a></li>
      <li><a href="#process_fragment_orientation">_process_fragment_orientation</a></li>
      <li><a href="#process_bio_sequence">_process_bio_sequence</a></li>
      <li><a href="#process_xrefs">_process_xrefs</a></li>
      <li><a href="#process_xref">_process_xref</a></li>
      <li><a href="#process_sequence_map">_process_sequence_map</a></li>
      <li><a href="#process_annotations">_process_annotations</a></li>
      <li><a href="#process_seq_feature">_process_seq_feature</a></li>
      <li><a href="#process_qualifier">_process_qualifier</a></li>
      <li><a href="#process_classification">_process_classification</a></li>
      <li><a href="#tag_processing_helper">_tag_processing_helper</a></li>
      <li><a href="#one_tag">_one_tag</a></li>
      <li><a href="#question_mark_tag">_question_mark_tag</a></li>
      <li><a href="#star_tag">_star_tag</a></li>
      <li><a href="#plus_tag">_plus_tag</a></li>
      <li><a href="#helper_store_attribute_list">_helper_store_attribute_list</a></li>
      <li><a href="#store_seqs">_store_seqs</a></li>
      <li><a href="#next_seq">next_seq</a></li>
      <li><a href="#next_primary_seq">next_primary_seq</a></li>
      <li><a href="#write_seq">write_seq</a></li>
      <li><a href="#write_each_record">_write_each_record</a></li>
      <li><a href="#write_seqfeature">_write_seqfeature</a></li>
      <li><a href="#filehandle">_filehandle</a></li>
      <li><a href="#throw">throw</a></li>
    </ul>
  </li>
</ul>

<h1 id="NAME">NAME</h1>

<p>Bio::SeqIO::agave - AGAVE sequence output stream.</p>

<h1 id="SYNOPSIS">SYNOPSIS</h1>

<p>It is probably best not to use this object directly, but rather go through the SeqIO handler system. Go:</p>

<pre><code><code>  <span class="variable">$in</span>  <span class="operator">=</span> <span class="variable">Bio::SeqIO</span><span class="operator">-&gt;</span><span class="variable">new</span><span class="operator">(</span><span class="string">'-file'</span>   <span class="operator">=&gt;</span> <span class="string">"</span><span class="variable">$file_in</span><span class="string">"</span><span class="operator">,</span>
                         <span class="string">'-format'</span> <span class="operator">=&gt;</span> <span class="string">'EMBL'</span><span class="operator">);</span>
  
  <span class="variable">$out</span> <span class="operator">=</span> <span class="variable">Bio::SeqIO</span><span class="operator">-&gt;</span><span class="variable">new</span><span class="operator">(</span><span class="string">'-file'</span>   <span class="operator">=&gt;</span> <span class="string">"&gt;</span><span class="variable">$file_out</span><span class="string">"</span><span class="operator">,</span>
                         <span class="string">'-format'</span> <span class="operator">=&gt;</span> <span class="string">'AGAVE'</span><span class="operator">);</span>
  
  <span class="keyword">while</span> <span class="operator">(</span><span class="keyword">my</span> <span class="variable">$seq</span> <span class="operator">=</span> <span class="variable">$in</span><span class="operator">-&gt;</span><span class="variable">next_seq</span><span class="operator">){</span>
        <span class="variable">$out</span><span class="operator">-&gt;</span><span class="variable">write_seq</span><span class="operator">(</span><span class="variable">$seq</span><span class="operator">);</span>
  <span class="operator">}</span>
</code></code></pre>

<h1 id="DESCRIPTION">DESCRIPTION</h1>

<p>This object can transform Bio::Seq objects to agave xml file and vice-versa. I (Simon) coded up this module because I needed a parser to extract data from AGAVE xml to be utitlized by the GenQuire genome annotation system (See http://www.bioinformatics.org/Genquire).</p>

<p>***NOTE*** At the moment, not all of the tags are implemented. In general, I followed the output format for the XEMBL project http://www.ebi.ac.uk/xembl/</p>

<h1 id="FEEDBACK">FEEDBACK</h1>

<h2 id="Mailing-Lists">Mailing Lists</h2>

<p>User feedback is an integral part of the evolution of this and other Bioperl modules. Send your comments and suggestions preferably to one of the Bioperl mailing lists. Your participation is much appreciated.</p>

<pre><code><code>  bioperl-l@bioperl.org                  - General discussion
  http://bioperl.org/wiki/Mailing_lists  - About the mailing lists</code></code></pre>

<h2 id="Support">Support</h2>

<p>Please direct usage questions or support issues to the mailing list:</p>

<p><i>bioperl-l@bioperl.org</i></p>

<p>rather than to the module maintainer directly. Many experienced and reponsive experts will be able look at the problem and quickly address it. Please include a thorough description of the problem with code and data examples if at all possible.</p>

<h2 id="Reporting-Bugs">Reporting Bugs</h2>

<p>Report bugs to the Bioperl bug tracking system to help us keep track the bugs and their resolution. Bug reports can be submitted via the web:</p>

<pre><code><code>  https://redmine.open-bio.org/projects/bioperl/</code></code></pre>

<h1 id="AUTHOR---Simon-K.-Chan">AUTHOR - Simon K. Chan</h1>

<p>Email:</p>

<h1 id="APPENDIX">APPENDIX</h1>

<p>The rest of the documentation details each of the object methods. Internal methods are usually preceded with a _</p>

<h2 id="process">_process</h2>

<pre><code><code>  Title    : _process
  Usage    : $self-&gt;_process
  Function : Parses the agave xml file.
  Args     : None.
  Returns  : Nothing.
  Note     : Method(s) that call(s) this method : _initialize
             Method(s) that this method calls   : _process_sciobj
             FIRST/START sub.</code></code></pre>

<h2 id="process_sciobj">_process_sciobj</h2>

<pre><code><code>  Title    : _process_sciobj
  Usage    : $self-&gt;_process_sciobj
  Function : Parses the data between the &lt;sciobj&gt;&lt;/sciobj&gt; tags.
  Args     : The string that holds the attributes for &lt;sciobj&gt;.
  Returns  : Data structure holding the values parsed between
             the &lt;sciobj&gt;&lt;/sciobj&gt; tags.
  Note     : Method(s) that call(s) this method : _process
             Method(s) that this method calls   :
             _helper_store_attribute_list , _process_contig</code></code></pre>

<h2 id="process_contig">_process_contig</h2>

<pre><code><code>  Title    : _process_contig
  Usage    : $self-&gt;_process_contig
  Function : Parses the data between the &lt;contig&gt;&lt;/contig&gt; tags.
  Args     : 2 scalars:
             - reference to a scalar holding the line to be parsed.
             - scalar holding the attributes for the &lt;contig&gt; tag
               to be parsed.
  Returns  : Data structure holding the values parsed between
             the &lt;contig&gt;&lt;/contig&gt; tags.
  Note     : Method(s) that call(s) this method : _process_sciobj
             Method(s) that this method calls   :
             _helper_store_attribute_list, _one_tag , _process_fragment_order</code></code></pre>

<h2 id="process_fragment_order">_process_fragment_order</h2>

<pre><code><code>  Title    : _process_fragment_order
  Usage    : $self-&gt;_process_fragment_order
  Function : Parses the data between the &lt;fragment_order&gt;&lt;/fragment_order&gt; tags.
  Args     : 2 scalars:
             - reference to a scalar holding the value of the line to be parsed.
             - reference to a data structure to store the &lt;fragment_order&gt; data.
  Returns  : Nothing.
  Note     : Method(s) that call(s) this method : _process_contig
             Method(s) that this method calls   :
             _helper_store_attribute_list , _process_fragment_orientation</code></code></pre>

<h2 id="process_fragment_orientation">_process_fragment_orientation</h2>

<pre><code><code>  Title    : _process_fragment_orientation
  Usage    : $self-&gt;_process_fragment_orientation
  Function : Parses the data between the &lt;fragment_orientation&gt; and
             &lt;/fragment_orientation&gt; tags.
  Args     : 2 scalars:
             - reference to a scalar holding the value of the line to be parsed.
             - reference to a data structure to store the &lt;fragment_orientation&gt; data.
  Returns  : Nothing.
  Note     : Method(s) that call(s) this method : _process_fragment_order</code></code></pre>

<p>Method(s) that this method calls : _helper_store_attribute_list , _process_bio_sequence</p>

<h2 id="process_bio_sequence">_process_bio_sequence</h2>

<pre><code><code>  Title    : _process_bio_sequence
  Usage    : $self-&gt;_process_bio_sequence
  Function : Parses the data between the &lt;bio_sequence&gt;&lt;/bio_sequence&gt; tags.
  Args     : 2 scalars:
             - reference to a scalar holding the value of the line to be parsed.
             - scalar holding the value of the attributes for &lt;bio_sequence&gt;
  Returns  : data structure holding the values between &lt;bio_sequence&gt;&lt;/bio_sequence&gt;
  Note     : Method(s) that call(s) this method : _process_fragment_orientation</code></code></pre>

<p>Method(s) that this method calls : _helper_store_attribute_list , _one_tag , _question_mark_tag , _star_tag , _process_alt_ids , _process_xrefs , _process_sequence_map</p>

<h2 id="process_xrefs">_process_xrefs</h2>

<pre><code><code>  Title    : _process_xrefs
  Usage    : $self-&gt;_process_xrefs
  Function : Parse the data between the &lt;xrefs&gt;&lt;/xrefs&gt; tags.
  Args     : reference to a scalar holding the value of the line to be parsed.
  Return   : Nothing.
  Note     : Method(s) that call(s) this method: _process_bio_sequence
             Method(s) that this method calls: _one_tag , _process_xref</code></code></pre>

<h2 id="process_xref">_process_xref</h2>

<pre><code><code>  Title    : _process_xref
  Usage    : $self-&gt;_process_xref
  Function : Parses the data between the &lt;xref&gt;&lt;/xref&gt; tags.
  Args     : 2 scalars:
             - reference to a scalar holding the value of the line to be parsed.
             - reference to a data structure to store the &lt;xref&gt; data.
  Returns  : Nothing.
  Note     : Method(s) that call(s) this method : _process_xrefs (note the &#39;s&#39; in &#39;xrefs&#39;)
             Method(s) that this method calls   : _helper_store_attribute_list , _star_tag</code></code></pre>

<h2 id="process_sequence_map">_process_sequence_map</h2>

<pre><code><code>  Title    : _process_sequence_map
  Usage    : $self-&gt;_process_sequence_map
  Function : Parses the data between the &lt;sequence_map&gt;&lt;/sequence_map&gt; tags.
  Args     : Reference to scalar holding the line to be parsed.
  Returns  : Data structure that holds the values that were parsed.
  Note     : Method(s) that call(s) this method : _process_bio_sequence
             Method(s) that this method calls   : _helper_store_attribute_list ,
                _question_mark_tag , _process_annotations</code></code></pre>

<h2 id="process_annotations">_process_annotations</h2>

<pre><code><code>  Title    : _process_annotations
  Usage    : $self-&gt;_process_annotations
  Function : Parse the data between the &lt;annotations&gt;&lt;/annotations&gt; tags.
  Args     : Reference to scalar holding the line to be parsed.
  Returns  : Data structure that holds the values that were parsed.
  Note     : Method(s) that call(s) this method : _process_sequence_map
             Method(s) that this method calls   : _process_seq_feature</code></code></pre>

<h2 id="process_seq_feature">_process_seq_feature</h2>

<pre><code><code>  Title    : _process_seq_feature
  Usage    : $self-&gt;_process_seq_feature
  Function : Parses the data between the &lt;seq_feature&gt;&lt;/seq_feature&gt; tag.
  Args     : 2 scalars:
             - Reference to scalar holding the line to be parsed.
             - Scalar holding the attributes for &lt;seq_feature&gt;.
  Returns  : Data structure holding the values parsed.
  Note     : Method(s) that call(s) this method: _process_annotations</code></code></pre>

<p>Method(s) that this method calls: _helper_store_attribute_list , _process_classification , _question_mark_tag , _one_tag , _process_evidence , _process_qualifier , _process_seq_feature , _process_related_annot</p>

<h2 id="process_qualifier">_process_qualifier</h2>

<pre><code><code>  Title    : _process_qualifier
  Usage    : $self-&gt;_process_qualifier
  Function : Parse the data between the &lt;qualifier&gt;&lt;/qualifier&gt; tags.
  Args     : 2 scalars:
             - reference to a scalar holding the value of the line to be parsed.
             - reference to a data structure to store the &lt;qualifer&gt; data.
  Returns  : Nothing.
  Note     : Method(s) that call(s) this method : _process_seq_feature
             Method(s) that this method calls   : _star_tag</code></code></pre>

<h2 id="process_classification">_process_classification</h2>

<pre><code><code>  Title   : _process_classification
  Usage   : $self-&gt;_process_classification
  Function: Parse the data between the &lt;classification&gt;&lt;/classification&gt; tags.
  Args    :   2 scalars:
            - reference to a scalar holding the value of the line to be parsed.
            - reference to a data structure to store the &lt;qualifer&gt; data.
  Returns : Nothing.
  Note    : Method(s) that call(s) this method: _process_seq_feature

  Method(s) that this method calls: _helper_store_attribute_list ,
  _question_mark_tag , _star_tag, _process_evidence</code></code></pre>

<h2 id="tag_processing_helper">_tag_processing_helper</h2>

<pre><code><code>  Title    : _tag_processing_helper
  Usage    : $self-&gt;_tag_processing_helper
  Function : Stores the tag value within the data structure.
             Also calls _helper_store_attribute_list to store the 
             attributes and their values in the data structure.
  Args     : 5 scalars:
             - Scalar holding the value of the attributes
             - Reference to a data structure to store the data for &lt;$tag_name&gt;
             - Scalar holding the tag name.
             - Scalar holding the value of the tag.
             - Scalar holding the value of either &#39;star&#39;, &#39;plus&#39;, 
               or &#39;question mark&#39; which specifies what type of method
               called this method.
  Returns  : Nothing.
  Note     : Method(s) that call(s) this method:
             Method(s) that this method calls: _helper_store_attribute_list</code></code></pre>

<h2 id="one_tag">_one_tag</h2>

<pre><code><code>  Title    : _one_tag
  Usage    : $self-&gt;_one_tag
  Function : A method to store data from tags that occurs just once.
  Args     : 2 scalars:
             - reference to a scalar holding the value of the line to be parsed.
             - reference to a data structure to store the data for &lt;$tag_name&gt;
  Returns  : Nothing.
  Note     : Method(s) that call(s) this method : many
             Method(s) that this method calls   : _tag_processing_helper</code></code></pre>

<h2 id="question_mark_tag">_question_mark_tag</h2>

<pre><code><code>  Title    : _question_mark_tag
  Usage    : $self-&gt;_question_mark_tag
  Function : Parses values from tags that occurs zero or one time. ie: tag_name?
  Args     : 3 scalars:
             - reference to a scalar holding the value of the line to be parsed.
             - reference to a data structure to store the data for &lt;$tag_name&gt;
             - scalar holding the name of the tag.
  Returns  : Nothing.
  Note     : Method(s) that call(s) this method : many.
             Method(s) that this method calls   : _tag_processing_helper</code></code></pre>

<h2 id="star_tag">_star_tag</h2>

<pre><code><code>  Title    : _star_tag
  Usage    : $self-&gt;_star_tag
  Function : Parses values from tags that occur zero or more times. ie: tag_name*
  Args     : 3 scalars:
             - reference to a scalar holding the value of the line to be parsed.
             - reference to a data structure to store the data for &lt;$tag_name&gt;
             - scalar holding the name of the tag.
  Returns  : Nothing.
  Note     : Method(s) that call(s) this method : many.
             Method(s) that this method calls   : _tag_processing_helper</code></code></pre>

<h2 id="plus_tag">_plus_tag</h2>

<pre><code><code>  Title    : _plus_tag
  Usage    : $self-&gt;_plus_tag
  Function : Handles &#39;plus&#39; tags (tags that occur one or more times).  tag_name+
  Args     : 3 scalars:
             - reference to a scalar holding the value of the line to be parsed.
             - reference to a data structure to store the data for &lt;$tag_name&gt;
             - scalar holding the name of the tag.
  Returns  : Nothing.
  Note     : Method(s) that call(s) this method : many.
             Method(s) that this method calls   : _star_tag</code></code></pre>

<h2 id="helper_store_attribute_list">_helper_store_attribute_list</h2>

<pre><code><code>  Title    : _helper_store_attribute_list
  Usage    : $self-&gt;_helper_store_attribute_list
  Function : A helper method used to store the attributes from
             the tags into the data structure.
  Args     : 2 scalars:
             - scalar holding the attribute values to be parsed.
             - reference to a data structure to store the data between the 2 tags.
  Returns  : Nothing.
  Note     : Method(s) that call(s) this method : Many.
             Method(s) that this method call(s) : None.</code></code></pre>

<h2 id="store_seqs">_store_seqs</h2>

<pre><code><code>  Title    : _store_seqs
  Usage    : $self-&gt;_store_seqs
  Function : This method is called once in the life time of the script.
             It stores the data parsed from the agave xml file into
             the Bio::Seq object.
  Args     : None.
  Returns  : Nothing.
  Note     : Method(s) that call(s) this method : next_seq
             Method(s) that this method calls   : None.</code></code></pre>

<h2 id="next_seq">next_seq</h2>

<pre><code><code>        Title    : next_seq
        Usage    : $seq = $stream-&gt;next_seq()
        Function : Returns the next sequence in the stream.
        Args     : None.
        Returns  : Bio::Seq object</code></code></pre>

<p>Method is called from the script. Method(s) that this method calls: _store_seqs (only once throughout the life time of script execution).</p>

<h2 id="next_primary_seq">next_primary_seq</h2>

<pre><code><code>  Title   : next_primary_seq
  Usage   : $seq = $stream-&gt;next_primary_seq()
  Function: returns the next primary sequence (ie no seq_features) in the stream
  Returns : Bio::PrimarySeq object
  Args    : NONE</code></code></pre>

<h2 id="write_seq">write_seq</h2>

<pre><code><code>  Title   : write_seq
  Usage   : Not Yet Implemented! $stream-&gt;write_seq(@seq)
  Function: writes the $seq object into the stream
  Returns : 1 for success and 0 for error
  Args    : Bio::Seq object</code></code></pre>

<h2 id="write_each_record">_write_each_record</h2>

<pre><code><code>  Title   : _write_each_record
  Usage   : $agave-&gt;_write_each_record( $seqI )
  Function: change data into agave format
  Returns : NONE
  Args    : Bio::SeqI object</code></code></pre>

<h2 id="write_seqfeature">_write_seqfeature</h2>

<pre><code><code>  Usage   : $agave-&gt;_write_each_record( $seqfeature, $write )
  Function: change seeqfeature data into agave format
  Returns : NONE
  Args    : Bio::SeqFeature object and XML::writer object</code></code></pre>

<h2 id="filehandle">_filehandle</h2>

<pre><code><code>  Title   : _filehandle
  Usage   : $obj-&gt;_filehandle($newval)
  Function:
  Example :
  Returns : value of _filehandle
  Args    : newvalue (optional)</code></code></pre>

<h2 id="throw">throw</h2>

<pre><code><code>  <span class="variable">Title</span>    <span class="operator">:</span> <span class="variable">throw</span>
  <span class="variable">Usage</span>    <span class="operator">:</span> <span class="variable">$self</span><span class="operator">-&gt;</span><span class="variable">throw</span><span class="operator">;</span>
  <span class="variable">Function</span> <span class="operator">:</span> <span class="variable">Throw's</span> <span class="variable">error</span> <span class="variable">message</span><span class="operator">.</span>  <span class="variable">Calls</span> <span class="variable">SeqIO's</span> <span class="variable">throw</span> <span class="variable">method</span><span class="operator">.</span>
  <span class="variable">Args</span>     <span class="operator">:</span> <span class="variable">Array</span> <span class="variable">of</span> <span class="variable">string</span><span class="operator">(</span><span class="regex">s), holding error message(s).
  Returns  : Nothing.
  Note     : Method(s)</span> <span class="variable">that</span> <span class="variable">call</span><span class="operator">(</span><span class="regex">s) this method: many.
             Method(s) that this method calls: Bio::SeqIO's throw method.
  </span>
</code></code></pre>


</body>

</html>


