<?xml version="1.0" ?>
<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.0 Transitional//EN">
<html>
<head>
<!-- saved from url=(0017)http://localhost/ -->
<script language="JavaScript" src="../../../../displayToc.js"></script>
<script language="JavaScript" src="../../../../tocParas.js"></script>
<script language="JavaScript" src="../../../../tocTab.js"></script>
<link rel="stylesheet" type="text/css" href="../../../../scineplex.css">
<title></title>
<link rel="stylesheet" href="../../../../Active.css" type="text/css" />
<meta http-equiv="content-type" content="text/html; charset=utf-8" />
<link rev="made" href="mailto:support@ActiveState.com" />
</head>

<body>



<ul id="index">
  <li><a href="#NAME">NAME</a></li>
  <li><a href="#SYNOPSIS">SYNOPSIS</a></li>
  <li><a href="#DESCRIPTION">DESCRIPTION</a></li>
  <li><a href="#FEEDBACK">FEEDBACK</a>
    <ul>
      <li><a href="#Mailing-Lists">Mailing Lists</a></li>
      <li><a href="#Reporting-Bugs">Reporting Bugs</a></li>
    </ul>
  </li>
  <li><a href="#AUTHOR---Warren-Kretzschmar">AUTHOR - Warren Kretzschmar</a></li>
  <li><a href="#COPYRIGHT">COPYRIGHT</a>
    <ul>
      <li><a href="#Public-Domain-Notice">Public Domain Notice</a></li>
    </ul>
  </li>
  <li><a href="#METHODS">METHODS</a>
    <ul>
      <li><a href="#INTERNAL-METHODS">INTERNAL METHODS</a>
        <ul>
          <li><a href="#initialize">_initialize</a></li>
          <li><a href="#read_start">_read_start</a></li>
        </ul>
      </li>
      <li><a href="#Methods-to-retrieve-mbsout-data">Methods to retrieve mbsout data</a>
        <ul>
          <li><a href="#get_segsites">get_segsites</a></li>
          <li><a href="#get_current_run_segsites">get_current_run_segsites</a></li>
          <li><a href="#get_pop_mut_param_per_site">get_pop_mut_param_per_site</a></li>
          <li><a href="#get_pop_recomb_param_per_site">get_pop_recomb_param_per_site</a></li>
          <li><a href="#get_nsites">get_nsites</a></li>
          <li><a href="#get_selpos">get_selpos</a></li>
          <li><a href="#get_nreps">get_nreps</a></li>
          <li><a href="#get_nfiles">get_nfiles</a></li>
          <li><a href="#get_traj_filename">get_traj_filename</a></li>
          <li><a href="#get_runs">get_runs</a></li>
          <li><a href="#get_Positions">get_Positions</a></li>
          <li><a href="#get_tot_run_haps">get_tot_run_haps</a></li>
          <li><a href="#get_mbs_info_line">get_mbs_info_line</a></li>
          <li><a href="#tot_haps">tot_haps</a></li>
          <li><a href="#next_run_num">next_run_num</a></li>
          <li><a href="#get_last_haps_run_num">get_last_haps_run_num</a></li>
          <li><a href="#get_last_read_hap_num">get_last_read_hap_num</a></li>
          <li><a href="#outgroup">outgroup</a></li>
          <li><a href="#get_next_seq">get_next_seq</a></li>
          <li><a href="#get_next_hap">get_next_hap</a></li>
          <li><a href="#get_next_run">get_next_run</a></li>
        </ul>
      </li>
      <li><a href="#METHODS-TO-RETRIEVE-CONSTANTS">METHODS TO RETRIEVE CONSTANTS</a>
        <ul>
          <li><a href="#base_conversion_table">base_conversion_table</a></li>
        </ul>
      </li>
    </ul>
  </li>
</ul>

<h1 id="NAME">NAME</h1>

<p>Bio::SeqIO::mbsout - input stream for output by Teshima et al.&#39;s mbs.</p>

<h1 id="SYNOPSIS">SYNOPSIS</h1>

<p>Do not use this module directly. Use it via the Bio::SeqIO class.</p>

<h1 id="DESCRIPTION">DESCRIPTION</h1>

<p>mbs (Teshima KM, Innan H (2009) mbs: modifying Hudson&#39;s ms software to generate samples of DNA sequences with a biallelic site under selection. BMC Bioinformatics 10: 166 ) can be found at http://www.biomedcentral.com/1471-2105/10/166/additional/.</p>

<p>Currently this object can be used to read output from mbs into seq objects. However, because bioperl has no support for haplotypes created using an infinite sites model (where &#39;1&#39; identifies a derived allele and &#39;0&#39; identifies an ancestral allele), the sequences returned by mbsout are coded using A, T, C and G. To decode the bases, use the sequence conversion table (a hash) returned by get_base_conversion_table(). In the table, 4 and 5 are used when the ancestry is unclear. This should not ever happen when creating files with mbs, but it will be used when creating mbsOUT files from a collection of seq objects ( To be added later ). Alternatively, use get_next_hap() to get a string with 1&#39;s and 0&#39;s instead of a seq object.</p>

<h1 id="FEEDBACK">FEEDBACK</h1>

<h2 id="Mailing-Lists">Mailing Lists</h2>

<p>User feedback is an integral part of the evolution of this and other Bioperl modules. Send your comments and suggestions preferably to the Bioperl mailing list. Your participation is much appreciated.</p>

<pre><code><code>  bioperl-l@bioperl.org                  - General discussion
  http://bioperl.org/wiki/Mailing_lists  - About the mailing lists</code></code></pre>

<h2 id="Reporting-Bugs">Reporting Bugs</h2>

<p>Report bugs to the Bioperl bug tracking system to help us keep track of the bugs and their resolution. Bug reports can be submitted via the web:</p>

<pre><code><code>  https://redmine.open-bio.org/projects/bioperl/  </code></code></pre>

<h1 id="AUTHOR---Warren-Kretzschmar">AUTHOR - Warren Kretzschmar</h1>

<p>This module was written by Warren Kretzschmar</p>

<p>email: wkretzsch@gmail.com</p>

<p>This module grew out of a parser written by Aida Andres.</p>

<h1 id="COPYRIGHT">COPYRIGHT</h1>

<h2 id="Public-Domain-Notice">Public Domain Notice</h2>

<p>This software/database is ``United States Government Work&#39;&#39; under the terms of the United States Copyright Act. It was written as part of the authors&#39; official duties for the United States Government and thus cannot be copyrighted. This software/database is freely available to the public for use without a copyright notice. Restrictions cannot be placed on its present or future use.</p>

<p>Although all reasonable efforts have been taken to ensure the accuracy and reliability of the software and data, the National Human Genome Research Institute (NHGRI) and the U.S. Government does not and cannot warrant the performance or results that may be obtained by using this software or data. NHGRI and the U.S. Government disclaims all warranties as to performance, merchantability or fitness for any particular purpose.</p>

<h1 id="METHODS">METHODS</h1>

<h2 id="INTERNAL-METHODS">INTERNAL METHODS</h2>

<h3 id="initialize">_initialize</h3>

<p>Title : _initialize Usage : $stream = Bio::SeqIO::mbsout-&gt;new($infile) Function: extracts basic information about the file. Returns : Bio::SeqIO object Args : no_og Details : include &#39;no_og&#39; flag = 0 if the last population of an mbsout file contains only one haplotype and you want the last haplotype to be treated as the outgroup.</p>

<h3 id="read_start">_read_start</h3>

<p>Title : _read_start Usage : $stream-&gt;_read_start() Function: reads from the filehandle $stream-&gt;{_filehandle} all information up to the first haplotype (sequence). Returns : void Args : none</p>

<h2 id="Methods-to-retrieve-mbsout-data">Methods to retrieve mbsout data</h2>

<h3 id="get_segsites">get_segsites</h3>

<p>Title : get_segsites Usage : $segsites = $stream-&gt;get_segsites() Function: returns the number segsites in the mbsout file (according to the mbsout header line). Returns : scalar Args : NONE</p>

<h3 id="get_current_run_segsites">get_current_run_segsites</h3>

<p>Title : get_current_run_segsites Usage : $segsites = $stream-&gt;get_current_run_segsites() Function: returns the number of segsites in the run of the last read haplotype (sequence). Returns : scalar Args : NONE</p>

<h3 id="get_pop_mut_param_per_site">get_pop_mut_param_per_site</h3>

<p>Title : get_pop_mut_param_per_site Usage : $pop_mut_param_per_site = $stream-&gt;get_pop_mut_param_per_site() Function: returns 4*N0*mu or the &quot;population mutation parameter per site&quot; Returns : scalar Args : NONE</p>

<h3 id="get_pop_recomb_param_per_site">get_pop_recomb_param_per_site</h3>

<p>Title : get_pop_recomb_param_per_site Usage : $pop_recomb_param_per_site = $stream-&gt;get_pop_recomb_param_per_site() Function: returns 4*N0*r or the &quot;population recombination parameter per site&quot; Returns : scalar Args : NONE</p>

<h3 id="get_nsites">get_nsites</h3>

<p>Title : get_nsites Usage : $nsites = $stream-&gt;get_nsites() Function: returns the number of sites simulated by mbs. Returns : scalar Args : NONE</p>

<h3 id="get_selpos">get_selpos</h3>

<p>Title : get_selpos Usage : $selpos = $stream-&gt;get_selpos() Function: returns the location on the chromosome where the allele is located that was selected for by mbs. Returns : scalar Args : NONE</p>

<h3 id="get_nreps">get_nreps</h3>

<p>Title : get_nreps Usage : $nreps = $stream-&gt;get_nreps() Function: returns the number replications done by mbs on each trajectory file to create the mbsout file. Returns : scalar Args : NONE</p>

<h3 id="get_nfiles">get_nfiles</h3>

<p>Title : get_nfiles Usage : $nfiles = $stream-&gt;get_nfiles() Function: returns the number of trajectory files used by mbs to create the mbsout file Returns : scalar Args : NONE</p>

<h3 id="get_traj_filename">get_traj_filename</h3>

<p>Title : get_traj_filename Usage : $traj_filename = $stream-&gt;get_traj_filename() Function: returns the prefix of the trajectory files used by mbs to create the mbsout file Returns : scalar Args : NONE</p>

<h3 id="get_runs">get_runs</h3>

<p>Title : get_runs Usage : $runs = $stream-&gt;get_runs() Function: returns the number of runs in the mbsout file Returns : scalar Args : NONE</p>

<h3 id="get_Positions">get_Positions</h3>

<p>Title : get_Positions Usage : @positions = $stream-&gt;get_Positions() Function: returns an array of the names of each segsite of the run of the last read hap. Returns : array Args : NONE</p>

<h3 id="get_tot_run_haps">get_tot_run_haps</h3>

<p>Title : get_tot_run_haps Usage : $number_of_haps_per_run = $stream-&gt;get_tot_run_haps() Function: returns the number of haplotypes (sequences) in each run of the mbsout file. Returns : scalar &gt;= 0 Args : NONE</p>

<h3 id="get_mbs_info_line">get_mbs_info_line</h3>

<p>Title : get_mbs_info_line Usage : $mbs_info_line = $stream-&gt;get_mbs_info_line() Function: returns the header line of the mbsout file. Returns : scalar Args : NONE</p>

<h3 id="tot_haps">tot_haps</h3>

<p>Title : tot_haps Usage : $number_of_haplotypes_in_file = $stream-&gt;tot_haps() Function: returns the number of haplotypes (sequences) in the mbsout file. Information gathered from mbsout header line. Returns : scalar Args : NONE</p>

<h3 id="next_run_num">next_run_num</h3>

<p>Title : next_run_num Usage : $next_run_number = $stream-&gt;next_run_num() Function: returns the number of the mbs run that the next haplotype (sequence) will be taken from (starting at 1). Returns undef if the complete file has been read. Returns : scalar &gt; 0 or undef Args : NONE</p>

<h3 id="get_last_haps_run_num">get_last_haps_run_num</h3>

<p>Title : get_last_haps_run_num Usage : $last_haps_run_number = $stream-&gt;get_last_haps_run_num() Function: returns the number of the ms run that the last haplotype (sequence) was taken from (starting at 1). Returns undef if no hap has been read yet. Returns : scalar &gt; 0 or undef Args : NONE</p>

<h3 id="get_last_read_hap_num">get_last_read_hap_num</h3>

<p>Title : get_last_read_hap_num Usage : $last_read_hap_num = $stream-&gt;get_last_read_hap_num() Function: returns the number (starting with 1) of the last haplotype read from the mbs file Returns : scalar &gt;= 0 Args : NONE Details : 0 means that no haplotype has been read yet.</p>

<h3 id="outgroup">outgroup</h3>

<p>Title : outgroup Usage : $outgroup = $stream-&gt;outgroup() Function: returns &#39;1&#39; if the mbsout object has an outgroup. Returns &#39;0&#39; otherwise. Returns : 1 or 0, currently always 0 Args : NONE Details : This method will return &#39;1&#39; only if the last population in the mbsout file contains only one haplotype. If the last population is not an outgroup then create the mbsout object using &#39;no_outgroup&#39; as input parameter for new() (see mbsout-&gt;new()).</p>

<pre><code><code>          Currently there exists no way of introducing an outgroup into an mbs 
          file, so this function will always return &#39;0&#39;.</code></code></pre>

<h3 id="get_next_seq">get_next_seq</h3>

<p>Title : get_next_seq Usage : $seq = $stream-&gt;get_next_seq() Function: reads and returns the next sequence (haplotype) in the stream Returns : Bio::Seq object Args : NONE Note : This function is included only to conform to convention. It only calls next_hap() and passes on that method&#39;s return value. Use next_hap() instead for better performance.</p>

<h3 id="get_next_hap">get_next_hap</h3>

<p>Title : get_next_hap Usage : $seq = $stream-&gt;get_next_hap() Function: reads and returns the next sequence (haplotype) in the stream. Returns void if all sequences in stream have been read. Returns : Bio::Seq object Args : NONE Note : Use this instead of get_next_seq().</p>

<h3 id="get_next_run">get_next_run</h3>

<p>Title : get_next_run Usage : @seqs = $stream-&gt;get_next_run() Function: reads and returns all the remaining sequences (haplotypes) in the mbs run of the next sequence. Returns : array of Bio::Seq objects Args : NONE</p>

<h2 id="METHODS-TO-RETRIEVE-CONSTANTS">METHODS TO RETRIEVE CONSTANTS</h2>

<h3 id="base_conversion_table">base_conversion_table</h3>

<p>Title : get_base_conversion_table Usage : $table_hash_ref = $stream-&gt;get_base_conversion_table() Function: returns a reference to a hash. The keys of the hash are the letters &#39;A&#39;,&#39;T&#39;,&#39;G&#39;,&#39;C&#39;. The values associated with each key are the value that each letter in the sequence of a seq object returned by a Bio::SeqIO::mbsout stream should be translated to. Returns : reference to a hash Args : NONE Synopsys:</p>

<pre><code><code>        <span class="comment"># retrieve the Bio::Seq object's sequence</span>
        <span class="keyword">my</span> <span class="variable">$haplotype</span> <span class="operator">=</span> <span class="variable">$seq</span><span class="operator">-&gt;</span><span class="variable">seq</span><span class="operator">;</span>
        <span class="keyword">my</span> <span class="variable">$rh_base_conversion_table</span> <span class="operator">=</span> <span class="variable">$stream</span><span class="operator">-&gt;</span><span class="variable">get_base_conversion_table</span><span class="operator">();</span>
        
        <span class="comment"># need to convert all letters to their corresponding numbers.</span>
        <span class="keyword">foreach</span> <span class="keyword">my</span> <span class="variable">$base</span> <span class="operator">(</span><span class="keyword">keys</span> <span class="variable">%</span><span class="operator">{</span><span class="variable">$rh_base_conversion_table</span><span class="operator">}</span><span class="operator">){</span>
                <span class="variable">$haplotype</span> <span class="operator">=~</span> <span class="regex">s/(</span><span class="variable">$base</span><span class="regex">)/$rh_base_conversion_table-&gt;{$base}/g</span><span class="operator">;</span>
        <span class="operator">}</span>
        
        <span class="comment"># $haplotype is now an ms style haplotype. (e.g. '100101101455')</span>
</code></code></pre>


</body>

</html>


