<?xml version="1.0" ?>
<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.0 Transitional//EN">
<html>
<head>
<!-- saved from url=(0017)http://localhost/ -->
<script language="JavaScript" src="../../../../displayToc.js"></script>
<script language="JavaScript" src="../../../../tocParas.js"></script>
<script language="JavaScript" src="../../../../tocTab.js"></script>
<link rel="stylesheet" type="text/css" href="../../../../scineplex.css">
<title></title>
<link rel="stylesheet" href="../../../../Active.css" type="text/css" />
<meta http-equiv="content-type" content="text/html; charset=utf-8" />
<link rev="made" href="mailto:support@ActiveState.com" />
</head>

<body>



<ul id="index">
  <li><a href="#NAME">NAME</a></li>
  <li><a href="#SYNOPSIS">SYNOPSIS</a></li>
  <li><a href="#DESCRIPTION">DESCRIPTION</a></li>
  <li><a href="#FEEDBACK">FEEDBACK</a>
    <ul>
      <li><a href="#Mailing-Lists">Mailing Lists</a></li>
      <li><a href="#Reporting-Bugs">Reporting Bugs</a></li>
    </ul>
  </li>
  <li><a href="#AUTHOR---Warren-Kretzschmar">AUTHOR - Warren Kretzschmar</a></li>
  <li><a href="#COPYRIGHT">COPYRIGHT</a>
    <ul>
      <li><a href="#Public-Domain-Notice">Public Domain Notice</a></li>
    </ul>
  </li>
  <li><a href="#METHODS">METHODS</a>
    <ul>
      <li><a href="#Methods-for-Internal-Use">Methods for Internal Use</a>
        <ul>
          <li><a href="#initialize">_initialize</a></li>
          <li><a href="#read_start">_read_start</a></li>
        </ul>
      </li>
      <li><a href="#Methods-to-Access-Data">Methods to Access Data</a>
        <ul>
          <li><a href="#get_segsites">get_segsites</a></li>
          <li><a href="#get_current_run_segsites">get_current_run_segsites</a></li>
          <li><a href="#get_runs">get_runs</a></li>
          <li><a href="#get_Seeds">get_Seeds</a></li>
          <li><a href="#get_Positions">get_Positions</a></li>
          <li><a href="#get_tot_run_haps">get_tot_run_haps</a></li>
          <li><a href="#get_ms_info_line">get_ms_info_line</a></li>
          <li><a href="#tot_haps">tot_haps</a></li>
          <li><a href="#get_Pops">get_Pops</a></li>
          <li><a href="#get_next_run_num">get_next_run_num</a></li>
          <li><a href="#get_last_haps_run_num">get_last_haps_run_num</a></li>
          <li><a href="#get_last_read_hap_num">get_last_read_hap_num</a></li>
          <li><a href="#outgroup">outgroup</a></li>
          <li><a href="#get_next_haps_pop_num">get_next_haps_pop_num</a></li>
          <li><a href="#get_next_seq">get_next_seq</a></li>
          <li><a href="#next_seq">next_seq</a></li>
          <li><a href="#get_next_hap">get_next_hap</a></li>
          <li><a href="#get_next_pop">get_next_pop</a></li>
          <li><a href="#next_run">next_run</a></li>
        </ul>
      </li>
      <li><a href="#Methods-to-Retrieve-Constants">Methods to Retrieve Constants</a>
        <ul>
          <li><a href="#base_conversion_table">base_conversion_table</a></li>
        </ul>
      </li>
    </ul>
  </li>
</ul>

<h1 id="NAME">NAME</h1>

<p>Bio::SeqIO::msout - input stream for output by Hudson&#39;s ms</p>

<h1 id="SYNOPSIS">SYNOPSIS</h1>

<p>Do not use this module directly. Use it via the Bio::SeqIO class.</p>

<h1 id="DESCRIPTION">DESCRIPTION</h1>

<p>ms ( Hudson, R. R. (2002) Generating samples under a Wright-Fisher neutral model. Bioinformatics 18:337-8 ) can be found at http://home.uchicago.edu/~rhudson1/source/mksamples.html.</p>

<p>Currently this object can be used to read output from ms into seq objects. However, because bioperl has no support for haplotypes created using an infinite sites model (where &#39;1&#39; identifies a derived allele and &#39;0&#39; identifies an ancestral allele), the sequences returned by msout are coded using A, T, C and G. To decode the bases, use the sequence conversion table (a hash) returned by get_base_conversion_table(). In the table, 4 and 5 are used when the ancestry is unclear. This should not ever happen when creating files with ms, but it will be used when creating msOUT files from a collection of seq objects ( To be added later ). Alternatively, use get_next_hap() to get a string with 1&#39;s and 0&#39;s instead of a seq object.</p>

<h1 id="FEEDBACK">FEEDBACK</h1>

<h2 id="Mailing-Lists">Mailing Lists</h2>

<p>User feedback is an integral part of the evolution of this and other Bioperl modules. Send your comments and suggestions preferably to the Bioperl mailing list. Your participation is much appreciated.</p>

<pre><code><code>  bioperl-l@bioperl.org                  - General discussion
  http://bioperl.org/wiki/Mailing_lists  - About the mailing lists</code></code></pre>

<h2 id="Reporting-Bugs">Reporting Bugs</h2>

<p>Report bugs to the Bioperl bug tracking system to help us keep track of the bugs and their resolution. Bug reports can be submitted via the web:</p>

<pre><code><code>  https://redmine.open-bio.org/projects/bioperl/  </code></code></pre>

<h1 id="AUTHOR---Warren-Kretzschmar">AUTHOR - Warren Kretzschmar</h1>

<p>This module was written by Warren Kretzschmar</p>

<p>email: wkretzsch@gmail.com</p>

<p>This module grew out of a parser written by Aida Andres.</p>

<h1 id="COPYRIGHT">COPYRIGHT</h1>

<h2 id="Public-Domain-Notice">Public Domain Notice</h2>

<p>This software/database is ``United States Government Work&#39;&#39; under the terms of the United States Copyright Act. It was written as part of the authors&#39; official duties for the United States Government and thus cannot be copyrighted. This software/database is freely available to the public for use without a copyright notice. Restrictions cannot be placed on its present or future use.</p>

<p>Although all reasonable efforts have been taken to ensure the accuracy and reliability of the software and data, the National Human Genome Research Institute (NHGRI) and the U.S. Government does not and cannot warrant the performance or results that may be obtained by using this software or data. NHGRI and the U.S. Government disclaims all warranties as to performance, merchantability or fitness for any particular purpose.</p>

<h1 id="METHODS">METHODS</h1>

<h2 id="Methods-for-Internal-Use">Methods for Internal Use</h2>

<h3 id="initialize">_initialize</h3>

<p>Title : _initialize Usage : $stream = Bio::SeqIO::msOUT-&gt;new($infile) Function: extracts basic information about the file. Returns : Bio::SeqIO object Args : no_og, gunzip, gzip Details : include &#39;no_og&#39; flag if the last population of an msout file contains only one haplotype and you don&#39;t want the last haplotype to be treated as the outgroup ( suggested when reading data created by ms ).</p>

<h3 id="read_start">_read_start</h3>

<p>Title : _read_start Usage : $stream-&gt;_read_start() Function: reads from the filehandle $stream-&gt;{_filehandle} all information up to the first haplotype (sequence). Closes the filehandle if all lines have been read. Returns : void Args : none</p>

<h2 id="Methods-to-Access-Data">Methods to Access Data</h2>

<h3 id="get_segsites">get_segsites</h3>

<p>Title : get_segsites Usage : $segsites = $stream-&gt;get_segsites() Function: returns the number of segsites in the msOUT file (according to the msOUT header line&#39;s -s option), or the current run&#39;s segsites if -s was not specified in the command line (in this case the number of segsites varies from run to run). Returns : scalar Args : NONE</p>

<h3 id="get_current_run_segsites">get_current_run_segsites</h3>

<p>Title : get_current_run_segsites Usage : $segsites = $stream-&gt;get_current_run_segsites() Function: returns the number of segsites in the run of the last read haplotype (sequence). Returns : scalar Args : NONE</p>

<h3 id="get_runs">get_runs</h3>

<p>Title : get_runs Usage : $runs = $stream-&gt;get_runs() Function: returns the number of runs in the msOUT file (according to the msinfo line) Returns : scalar Args : NONE</p>

<h3 id="get_Seeds">get_Seeds</h3>

<p>Title : get_Seeds Usage : @seeds = $stream-&gt;get_Seeds() Function: returns an array of the seeds used in the creation of the msOUT file. Returns : array Args : NONE Details : In older versions, ms used three seeds. Newer versions of ms seem to use only one (longer) seed. This function will return all the seeds found.</p>

<h3 id="get_Positions">get_Positions</h3>

<p>Title : get_Positions Usage : @positions = $stream-&gt;get_Positions() Function: returns an array of the names of each segsite of the run of the last read hap. Returns : array Args : NONE Details : The Positions may or may not vary from run to run depending on the options used with ms.</p>

<h3 id="get_tot_run_haps">get_tot_run_haps</h3>

<p>Title : get_tot_run_haps Usage : $number_of_haps_per_run = $stream-&gt;get_tot_run_haps() Function: returns the number of haplotypes (sequences) in each run of the msOUT file ( according to the msinfo line ). Returns : scalar &gt;= 0 Args : NONE Details : This number should not vary from run to run.</p>

<h3 id="get_ms_info_line">get_ms_info_line</h3>

<p>Title : get_ms_info_line Usage : $ms_info_line = $stream-&gt;get_ms_info_line() Function: returns the header line of the msOUT file. Returns : scalar Args : NONE</p>

<h3 id="tot_haps">tot_haps</h3>

<p>Title : tot_haps Usage : $number_of_haplotypes_in_file = $stream-&gt;tot_haps() Function: returns the number of haplotypes (sequences) in the msOUT file. Information gathered from msOUT header line. Returns : scalar Args : NONE</p>

<h3 id="get_Pops">get_Pops</h3>

<p>Title : get_Pops Usage : @pops = $stream-&gt;pops() Function: returns an array of population sizes (order taken from the -I flag in the msOUT header line). This array will include the last hap even if it looks like an outgroup. Returns : array of scalars &gt; 0 Args : NONE</p>

<h3 id="get_next_run_num">get_next_run_num</h3>

<p>Title : get_next_run_num Usage : $next_run_number = $stream-&gt;next_run_num() Function: returns the number of the ms run that the next haplotype (sequence) will be taken from (starting at 1). Returns undef if the complete file has been read. Returns : scalar &gt; 0 or undef Args : NONE</p>

<h3 id="get_last_haps_run_num">get_last_haps_run_num</h3>

<p>Title : get_last_haps_run_num Usage : $last_haps_run_number = $stream-&gt;get_last_haps_run_num() Function: returns the number of the ms run that the last haplotype (sequence) was taken from (starting at 1). Returns undef if no hap has been read yet. Returns : scalar &gt; 0 or undef Args : NONE</p>

<h3 id="get_last_read_hap_num">get_last_read_hap_num</h3>

<p>Title : get_last_read_hap_num Usage : $last_read_hap_num = $stream-&gt;get_last_read_hap_num() Function: returns the number (starting with 1) of the last haplotype read from the ms file Returns : scalar &gt;= 0 Args : NONE Details : 0 means that no haplotype has been read yet. Is reset to 0 every run.</p>

<h3 id="outgroup">outgroup</h3>

<p>Title : outgroup Usage : $outgroup = $stream-&gt;outgroup() Function: returns &#39;1&#39; if the msOUT stream has an outgroup. Returns &#39;0&#39; otherwise. Returns : &#39;1&#39; or &#39;0&#39; Args : NONE Details : This method will return &#39;1&#39; only if the last population in the msOUT file contains only one haplotype. If the last population is not an outgroup then create the msOUT object using &#39;no_og&#39; as input flag. Also, return 0, if the run has only one population.</p>

<h3 id="get_next_haps_pop_num">get_next_haps_pop_num</h3>

<p>Title : get_next_haps_pop_num Usage : ($next_haps_pop_num, $num_haps_left_in_pop) = $stream-&gt;get_next_haps_pop_num() Function: First return value is the population number (starting with 1) the next hap will come from. The second return value is the number of haps left to read in the population from which the next hap will come. Returns : (scalar &gt; 0, scalar &gt; 0) Args : NONE</p>

<h3 id="get_next_seq">get_next_seq</h3>

<p>Title : get_next_seq Usage : $seq = $stream-&gt;get_next_seq() Function: reads and returns the next sequence (haplotype) in the stream Returns : Bio::Seq object or void if end of file Args : NONE Note : This function is included only to conform to convention. The returned Bio::Seq object holds a halpotype in coded form. Use the hash returned by get_base_conversion_table() to convert &#39;A&#39;, &#39;T&#39;, &#39;C&#39;, &#39;G&#39; back into 1,2,4 and 5. Use get_next_hap() to retrieve the halptoype as a string of 1,2,4 and 5s instead.</p>

<h3 id="next_seq">next_seq</h3>

<p>Title : next_seq Usage : $seq = $stream-&gt;next_seq() Function: Alias to get_next_seq() Returns : Bio::Seq object or void if end of file Args : NONE Note : This function is only included for convention. It calls get_next_seq(). See get_next_seq() for details.</p>

<h3 id="get_next_hap">get_next_hap</h3>

<p>Title : get_next_hap Usage : $hap = $stream-&gt;next_hap() Function: reads and returns the next sequence (haplotype) in the stream. Returns undef if all sequences in stream have been read. Returns : Haplotype string (e.g. &#39;110110000101101045454000101&#39; Args : NONE Note : Use get_next_seq() if you want the halpotype returned as a Bio::Seq object.</p>

<h3 id="get_next_pop">get_next_pop</h3>

<p>Title : get_next_pop Usage : @seqs = $stream-&gt;next_pop() Function: reads and returns all the remaining sequences (haplotypes) in the population of the next sequence. Returns an empty list if no more haps remain to be read in the stream Returns : array of Bio::Seq objects Args : NONE</p>

<h3 id="next_run">next_run</h3>

<p>Title : next_run Usage : @seqs = $stream-&gt;next_run() Function: reads and returns all the remaining sequences (haplotypes) in the ms run of the next sequence. Returns an empty list if all haps have been read from the stream. Returns : array of Bio::Seq objects Args : NONE</p>

<h2 id="Methods-to-Retrieve-Constants">Methods to Retrieve Constants</h2>

<h3 id="base_conversion_table">base_conversion_table</h3>

<p>Title : get_base_conversion_table Usage : $table_hash_ref = $stream-&gt;get_base_conversion_table() Function: returns a reference to a hash. The keys of the hash are the letters &#39; A&#39;,&#39;T&#39;,&#39;G&#39;,&#39;C&#39;. The values associated with each key are the value that each letter in the sequence of a seq object returned by a Bio::SeqIO::msout stream should be translated to. Returns : reference to a hash Args : NONE Synopsys:</p>

<pre><code><code>        <span class="comment"># retrieve the Bio::Seq object's sequence</span>
        <span class="keyword">my</span> <span class="variable">$haplotype</span> <span class="operator">=</span> <span class="variable">$seq</span><span class="operator">-&gt;</span><span class="variable">seq</span><span class="operator">;</span>
        
        <span class="comment"># need to convert all letters to their corresponding numbers.</span>
        <span class="keyword">foreach</span> <span class="keyword">my</span> <span class="variable">$base</span> <span class="operator">(</span><span class="keyword">keys</span> <span class="variable">%</span><span class="operator">{</span><span class="variable">$rh_base_conversion_table</span><span class="operator">}</span><span class="operator">){</span>
                <span class="variable">$haplotype</span> <span class="operator">=~</span> <span class="regex">s/(</span><span class="variable">$base</span><span class="regex">)/$rh_base_conversion_table-&gt;{$base}/g</span><span class="operator">;</span>
        <span class="operator">}</span>
        
        <span class="comment"># $haplotype is now an ms style haplotype. (e.g. '100101101455')</span>
</code></code></pre>


</body>

</html>


