<?xml version="1.0" ?>
<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.0 Transitional//EN">
<html>
<head>
<!-- saved from url=(0017)http://localhost/ -->
<script language="JavaScript" src="../../../../displayToc.js"></script>
<script language="JavaScript" src="../../../../tocParas.js"></script>
<script language="JavaScript" src="../../../../tocTab.js"></script>
<link rel="stylesheet" type="text/css" href="../../../../scineplex.css">
<title></title>
<link rel="stylesheet" href="../../../../Active.css" type="text/css" />
<meta http-equiv="content-type" content="text/html; charset=utf-8" />
<link rev="made" href="mailto:support@ActiveState.com" />
</head>

<body>



<ul id="index">
  <li><a href="#NAME">NAME</a></li>
  <li><a href="#SYNOPSIS">SYNOPSIS</a></li>
  <li><a href="#DESCRIPTION">DESCRIPTION</a>
    <ul>
      <li><a href="#GN-Gene-name-line-management-details">GN (Gene name) line management details</a></li>
      <li><a href="#Optional-functions">Optional functions</a></li>
    </ul>
  </li>
  <li><a href="#FEEDBACK">FEEDBACK</a>
    <ul>
      <li><a href="#Mailing-Lists">Mailing Lists</a></li>
      <li><a href="#Support">Support</a></li>
      <li><a href="#Reporting-Bugs">Reporting Bugs</a></li>
    </ul>
  </li>
  <li><a href="#AUTHOR---Elia-Stupka">AUTHOR - Elia Stupka</a></li>
  <li><a href="#APPENDIX">APPENDIX</a>
    <ul>
      <li><a href="#next_seq">next_seq</a></li>
      <li><a href="#write_seq">write_seq</a></li>
      <li><a href="#generateCRCTable">_generateCRCTable</a></li>
      <li><a href="#crc32">_crc32</a></li>
      <li><a href="#crc64">_crc64</a></li>
      <li><a href="#print_swissprot_FTHelper">_print_swissprot_FTHelper</a></li>
      <li><a href="#read_swissprot_References">_read_swissprot_References</a></li>
      <li><a href="#read_swissprot_Species">_read_swissprot_Species</a></li>
      <li><a href="#filehandle">_filehandle</a></li>
      <li><a href="#read_FTHelper_swissprot">_read_FTHelper_swissprot</a></li>
      <li><a href="#write_line_swissprot">_write_line_swissprot</a></li>
      <li><a href="#write_line_swissprot_regex">_write_line_swissprot_regex</a></li>
      <li><a href="#post_sort">_post_sort</a></li>
      <li><a href="#show_dna">_show_dna</a></li>
      <li><a href="#id_generation_func">_id_generation_func</a></li>
      <li><a href="#ac_generation_func">_ac_generation_func</a></li>
      <li><a href="#sv_generation_func">_sv_generation_func</a></li>
      <li><a href="#kw_generation_func">_kw_generation_func</a></li>
    </ul>
  </li>
</ul>

<h1 id="NAME">NAME</h1>

<p>Bio::SeqIO::swiss - Swissprot sequence input/output stream</p>

<h1 id="SYNOPSIS">SYNOPSIS</h1>

<p>It is probably best not to use this object directly, but rather go through the SeqIO handler system:</p>

<pre><code><code>    <span class="keyword">use</span> <span class="variable">Bio::SeqIO</span><span class="operator">;</span>
    
    <span class="variable">$stream</span> <span class="operator">=</span> <span class="variable">Bio::SeqIO</span><span class="operator">-&gt;</span><span class="variable">new</span><span class="operator">(</span><span class="string">-file</span> <span class="operator">=&gt;</span> <span class="variable">$filename</span><span class="operator">,</span>
                              <span class="string">-format</span> <span class="operator">=&gt;</span> <span class="string">'swiss'</span><span class="operator">);</span>
    
    <span class="keyword">while</span> <span class="operator">(</span> <span class="keyword">my</span> <span class="variable">$seq</span> <span class="operator">=</span> <span class="variable">$stream</span><span class="operator">-&gt;</span><span class="variable">next_seq</span><span class="operator">()</span> <span class="operator">)</span> <span class="operator">{</span>
       <span class="comment"># do something with $seq</span>
    <span class="operator">}</span>
</code></code></pre>

<h1 id="DESCRIPTION">DESCRIPTION</h1>

<p>This object can transform Bio::Seq objects to and from Swiss-Pprot flat file databases.</p>

<p>There is a lot of flexibility here about how to dump things which needs to be documented.</p>

<h2 id="GN-Gene-name-line-management-details">GN (Gene name) line management details</h2>

<p>A Uniprot/Swiss-Prot entry holds information on one protein sequence. If that sequence is identical across genes and species, they are all merged into one entry. This creates complex needs for several annotation fields in swiss-prot format.</p>

<p>The latest syntax for GN line is described in the user manual:</p>

<pre><code><code>  http://www.expasy.ch/sprot/userman.html#GN_line</code></code></pre>

<p>Each of the possibly multiple genes in an entry can have Name, Synonyms (only if there is a name), OrderedLocusNames (names from genomic sequences) and ORFNames (temporary or cosmid names). &quot;Name&quot; here really means &quot;symbol&quot;. This complexity is now dealt with the following way:</p>

<p>A new Bio::AnnotationI class was created in order to store the data in tag-value pairs. This class (Bio::Annotation::TagTree) is stored in the Bio::Annotation::Collection object and is accessed like all other annotations. The tag name is &#39;gene_name&#39;.</p>

<p>There is a single Bio::Annotation::TagTree per sequence record, which corresponds to the original class that stored this data (Bio::Annotation::StructuredValue). Depending on how we progress this may change to represent each group of gene names.</p>

<p>For now, to access the gene name tree annotation, one uses the below method:</p>

<pre><code><code>   <span class="keyword">my</span> <span class="operator">(</span><span class="variable">$gene</span><span class="operator">)</span> <span class="operator">=</span> <span class="variable">$seq</span><span class="operator">-&gt;</span><span class="variable">annotation</span><span class="operator">-&gt;</span><span class="variable">get_Annotations</span><span class="operator">(</span><span class="string">'gene_name'</span><span class="operator">);</span>
</code></code></pre>

<p>If you are only interested in displaying the values, value() returns a string with similar formatting.</p>

<p>There are several ways to get directly at the information you want if you know the element (tag) for the data. For gene names all data is stored with the element-tag pairs:</p>

<pre><code><code>  <span class="string">"element1=tag1, tag2, tag3; element2=tag4, tag5;"</span>
</code></code></pre>

<p>This normally means the element will be &#39;Name&#39;, &#39;Synonyms&#39;, etc. and the gene names the values. Using findval(), you can do the following:</p>

<pre><code><code>  <span class="comment"># grab a flattened list of all gene names</span>
  <span class="keyword">my</span> <span class="variable">@names</span> <span class="operator">=</span> <span class="variable">$ann</span><span class="operator">-&gt;</span><span class="variable">findval</span><span class="operator">(</span><span class="string">'Name'</span><span class="operator">);</span>
  
  <span class="comment"># or iterated through the nodes and grab the name for each group</span>
  <span class="keyword">for</span> <span class="keyword">my</span> <span class="variable">$node</span> <span class="operator">(</span><span class="variable">$ann</span><span class="operator">-&gt;</span><span class="variable">findnode</span><span class="operator">(</span><span class="string">'gene_name'</span><span class="operator">))</span> <span class="operator">{</span>
     <span class="keyword">my</span> <span class="variable">@names</span> <span class="operator">=</span> <span class="variable">$node</span><span class="operator">-&gt;</span><span class="variable">findval</span><span class="operator">(</span><span class="string">'Name'</span><span class="operator">);</span>
  <span class="operator">}</span>
</code></code></pre>

<p>The current method for parsing gene name data (and reconstructing gene name output) is very generic. This is somewhat preemptive if, for instance, UniProt decides to update and add another element name to the current ones using the same formatting layout. Under those circumstances, one can iterate through the tag tree in a safe way and retrieve all node data like so.</p>

<pre><code><code>  <span class="comment"># retrieve the gene name nodes (groups like names, synonyms, etc).</span>
  <span class="keyword">for</span> <span class="keyword">my</span> <span class="variable">$ann</span> <span class="operator">(</span><span class="variable">$seq</span><span class="operator">-&gt;</span><span class="variable">annotation</span><span class="operator">-&gt;</span><span class="variable">get_Annotations</span><span class="operator">(</span><span class="string">'gene_name'</span><span class="operator">))</span> <span class="operator">{</span>
  
      <span class="comment"># each gene name group</span>
      <span class="keyword">for</span> <span class="keyword">my</span> <span class="variable">$node</span> <span class="operator">(</span><span class="variable">$ann</span><span class="operator">-&gt;</span><span class="variable">findnode</span><span class="operator">(</span><span class="string">'gene_name'</span><span class="operator">))</span> <span class="operator">{</span>
          <span class="keyword">print</span> <span class="string">"Gene name:\n"</span><span class="operator">;</span>
  
          <span class="comment"># each gene name node (tag =&gt; value pair)</span>
          <span class="keyword">for</span> <span class="keyword">my</span> <span class="variable">$n</span> <span class="operator">(</span><span class="variable">$node</span><span class="operator">-&gt;</span><span class="variable">children</span><span class="operator">)</span> <span class="operator">{</span>
              <span class="keyword">print</span> <span class="string">"\t"</span><span class="operator">.</span><span class="variable">$n</span><span class="operator">-&gt;</span><span class="variable">element</span><span class="operator">.</span><span class="string">": "</span><span class="operator">.</span><span class="variable">$n</span><span class="operator">-&gt;</span><span class="variable">children</span><span class="operator">.</span><span class="string">"\n"</span><span class="operator">;</span>
          <span class="operator">}</span>
      <span class="operator">}</span>
  <span class="operator">}</span>
</code></code></pre>

<p>For more uses see Bio::Annotation::TagTree.</p>

<p>Since Uniprot/Swiss-Prot format have been around for quite some time, the parser is also able to read in the older GN line syntax where genes are separated by AND and various symbols by OR. The first symbol is taken to be the &#39;Name&#39; and the remaining ones are stored as &#39;Synonyms&#39;.</p>

<p>Also, for UniProt output we support using other Bio::AnnotationI, but in this case we only use the stringified version of the annotation. This is to allow for backwards compatibility with code that previously used Bio::Annotation::SimpleValue or other Bio::AnnotationI classes.</p>

<h2 id="Optional-functions">Optional functions</h2>

<dl>

<dt id="show_dna-">_show_dna()</dt>
<dd>

<p>(output only) shows the dna or not</p>

</dd>
<dt id="post_sort-">_post_sort()</dt>
<dd>

<p>(output only) provides a sorting func which is applied to the FTHelpers before printing</p>

</dd>
<dt id="id_generation_func-">_id_generation_func()</dt>
<dd>

<p>This is function which is called as</p>

<pre><code><code>   <span class="keyword">print</span> <span class="string">"ID   "</span><span class="operator">,</span> <span class="variable">$func</span><span class="operator">(</span><span class="variable">$seq</span><span class="operator">),</span> <span class="string">"\n"</span><span class="operator">;</span>
</code></code></pre>

<p>To generate the ID line. If it is not there, it generates a sensible ID line using a number of tools.</p>

<p>If you want to output annotations in Swissprot format they need to be stored in a Bio::Annotation::Collection object which is accessible through the Bio::SeqI interface method <a>annotation()</a>.</p>

<p>The following are the names of the keys which are polled from a <a href="../../../../../.cpan/build/BioPerl-1.6.901-ghKY8U/blib/lib/Bio/Annotation/Collection.html">Bio::Annotation::Collection</a> object.</p>

<pre><code><code> reference   - Should contain Bio::Annotation::Reference objects
 comment     - Should contain Bio::Annotation::Comment objects
 dblink      - Should contain Bio::Annotation::DBLink objects
 gene_name   - Should contain Bio::Annotation::SimpleValue object</code></code></pre>

</dd>
</dl>

<h1 id="FEEDBACK">FEEDBACK</h1>

<h2 id="Mailing-Lists">Mailing Lists</h2>

<p>User feedback is an integral part of the evolution of this and other Bioperl modules. Send your comments and suggestions, preferably to one of the Bioperl mailing lists. Your participation is much appreciated.</p>

<pre><code><code>  bioperl-l@bioperl.org                  - General discussion
  http://bioperl.org/wiki/Mailing_lists  - About the mailing lists</code></code></pre>

<h2 id="Support">Support</h2>

<p>Please direct usage questions or support issues to the mailing list:</p>

<p><i>bioperl-l@bioperl.org</i></p>

<p>rather than to the module maintainer directly. Many experienced and reponsive experts will be able look at the problem and quickly address it. Please include a thorough description of the problem with code and data examples if at all possible.</p>

<h2 id="Reporting-Bugs">Reporting Bugs</h2>

<p>Report bugs to the Bioperl bug tracking system to help us keep track the bugs and their resolution. Bug reports can be submitted via the web:</p>

<pre><code><code>  https://redmine.open-bio.org/projects/bioperl/</code></code></pre>

<h1 id="AUTHOR---Elia-Stupka">AUTHOR - Elia Stupka</h1>

<p>Email elia@tll.org.sg</p>

<h1 id="APPENDIX">APPENDIX</h1>

<p>The rest of the documentation details each of the object methods. Internal methods are usually preceded with a _</p>

<h2 id="next_seq">next_seq</h2>

<pre><code><code> Title   : next_seq
 Usage   : $seq = $stream-&gt;next_seq()
 Function: returns the next sequence in the stream
 Returns : Bio::Seq object
 Args    :</code></code></pre>

<h2 id="write_seq">write_seq</h2>

<pre><code><code> Title   : write_seq
 Usage   : $stream-&gt;write_seq($seq)
 Function: writes the $seq object (must be seq) to the stream
 Returns : 1 for success and 0 for error
 Args    : array of 1 to n Bio::SeqI objects</code></code></pre>

<h2 id="generateCRCTable">_generateCRCTable</h2>

<pre><code><code> Title   : _generateCRCTable
 Usage   :
 Function:
 Example :
 Returns :
 Args    :</code></code></pre>

<h2 id="crc32">_crc32</h2>

<pre><code><code> Title   : _crc32
 Usage   :
 Function:
 Example :
 Returns :
 Args    :</code></code></pre>

<h2 id="crc64">_crc64</h2>

<pre><code><code> Title   : _crc64
 Usage   :
 Function:
 Example :
 Returns :
 Args    :</code></code></pre>

<h2 id="print_swissprot_FTHelper">_print_swissprot_FTHelper</h2>

<pre><code><code> Title   : _print_swissprot_FTHelper
 Usage   :
 Function:
 Example :
 Returns :
 Args    :</code></code></pre>

<h2 id="read_swissprot_References">_read_swissprot_References</h2>

<pre><code><code> Title   : _read_swissprot_References
 Usage   :
 Function: Reads references from swissprot format. Internal function really
 Example :
 Returns :
 Args    :</code></code></pre>

<h2 id="read_swissprot_Species">_read_swissprot_Species</h2>

<pre><code><code> <span class="variable">Title</span>   <span class="operator">:</span> <span class="variable">_read_swissprot_Species</span>
 <span class="variable">Usage</span>   <span class="operator">:</span>
 <span class="variable">Function</span><span class="operator">:</span> <span class="variable">Reads</span> <span class="variable">the</span> <span class="variable">swissprot</span> <span class="variable">Organism</span> <span class="variable">species</span> <span class="keyword">and</span> <span class="variable">classification</span>
           <span class="variable">lines</span><span class="operator">.</span>
             <span class="variable">Able</span> <span class="variable">to</span> <span class="variable">deal</span> <span class="variable">with</span> <span class="variable">unconventional</span> <span class="variable">species</span> <span class="variable">names</span><span class="operator">.</span>
 <span class="variable">Example</span> <span class="operator">:</span> <span class="variable">OS</span> <span class="variable">Unknown</span> <span class="variable">prokaryotic</span> <span class="variable">organism</span>
             <span class="variable">$genus</span> <span class="operator">=</span> <span class="keyword">undef</span> <span class="operator">;</span> <span class="variable">$species</span> <span class="operator">=</span> <span class="variable">Unknown</span> <span class="variable">prokaryotic</span> <span class="variable">organism</span>
 <span class="variable">Returns</span> <span class="operator">:</span> <span class="variable">A</span> <span class="variable">Bio::Species</span> <span class="variable">object</span>
 <span class="variable">Args</span>    <span class="operator">:</span>
</code></code></pre>

<h2 id="filehandle">_filehandle</h2>

<pre><code><code> Title   : _filehandle
 Usage   : $obj-&gt;_filehandle($newval)
 Function:
 Example :
 Returns : value of _filehandle
 Args    : newvalue (optional)</code></code></pre>

<h2 id="read_FTHelper_swissprot">_read_FTHelper_swissprot</h2>

<pre><code><code> Title   : _read_FTHelper_swissprot
 Usage   : _read_FTHelper_swissprot(\$buffer)
 Function: reads the next FT key line
 Example :
 Returns : Bio::SeqIO::FTHelper object
 Args    :</code></code></pre>

<h2 id="write_line_swissprot">_write_line_swissprot</h2>

<pre><code><code> Title   : _write_line_swissprot
 Usage   :
 Function: internal function
 Example :
 Returns :
 Args    :</code></code></pre>

<h2 id="write_line_swissprot_regex">_write_line_swissprot_regex</h2>

<pre><code><code> Title   : _write_line_swissprot_regex
 Usage   :
 Function: internal function for writing lines of specified
           length, with different first and the next line
           left hand headers and split at specific points in the
           text
 Example :
 Returns : nothing
 Args    : file handle, first header, second header, text-line, regex for line breaks, total line length</code></code></pre>

<h2 id="post_sort">_post_sort</h2>

<pre><code><code> Title   : _post_sort
 Usage   : $obj-&gt;_post_sort($newval)
 Function:
 Returns : value of _post_sort
 Args    : newvalue (optional)</code></code></pre>

<h2 id="show_dna">_show_dna</h2>

<pre><code><code> Title   : _show_dna
 Usage   : $obj-&gt;_show_dna($newval)
 Function:
 Returns : value of _show_dna
 Args    : newvalue (optional)</code></code></pre>

<h2 id="id_generation_func">_id_generation_func</h2>

<pre><code><code> Title   : _id_generation_func
 Usage   : $obj-&gt;_id_generation_func($newval)
 Function:
 Returns : value of _id_generation_func
 Args    : newvalue (optional)</code></code></pre>

<h2 id="ac_generation_func">_ac_generation_func</h2>

<pre><code><code> Title   : _ac_generation_func
 Usage   : $obj-&gt;_ac_generation_func($newval)
 Function:
 Returns : value of _ac_generation_func
 Args    : newvalue (optional)</code></code></pre>

<h2 id="sv_generation_func">_sv_generation_func</h2>

<pre><code><code> Title   : _sv_generation_func
 Usage   : $obj-&gt;_sv_generation_func($newval)
 Function:
 Returns : value of _sv_generation_func
 Args    : newvalue (optional)</code></code></pre>

<h2 id="kw_generation_func">_kw_generation_func</h2>

<pre><code><code> Title   : _kw_generation_func
 Usage   : $obj-&gt;_kw_generation_func($newval)
 Function:
 Returns : value of _kw_generation_func
 Args    : newvalue (optional)</code></code></pre>


</body>

</html>


