<?xml version="1.0" ?>
<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.0 Transitional//EN">
<html>
<head>
<!-- saved from url=(0017)http://localhost/ -->
<script language="JavaScript" src="../../../../displayToc.js"></script>
<script language="JavaScript" src="../../../../tocParas.js"></script>
<script language="JavaScript" src="../../../../tocTab.js"></script>
<link rel="stylesheet" type="text/css" href="../../../../scineplex.css">
<title></title>
<link rel="stylesheet" href="../../../../Active.css" type="text/css" />
<meta http-equiv="content-type" content="text/html; charset=utf-8" />
<link rev="made" href="mailto:support@ActiveState.com" />
</head>

<body>



<ul id="index">
  <li><a href="#NAME">NAME</a></li>
  <li><a href="#SYNOPSIS">SYNOPSIS</a></li>
  <li><a href="#DESCRIPTION">DESCRIPTION</a></li>
  <li><a href="#FEEDBACK">FEEDBACK</a>
    <ul>
      <li><a href="#Mailing-Lists">Mailing Lists</a></li>
      <li><a href="#Support">Support</a></li>
      <li><a href="#Reporting-Bugs">Reporting Bugs</a></li>
    </ul>
  </li>
  <li><a href="#SEE-ALSO">SEE ALSO</a></li>
  <li><a href="#AUTHOR">AUTHOR</a></li>
  <li><a href="#APPENDIX">APPENDIX</a>
    <ul>
      <li><a href="#next_seq">next_seq</a></li>
      <li><a href="#write_seq">write_seq</a></li>
      <li><a href="#get_seqs">_get_seqs</a></li>
      <li><a href="#get_species">_get_species</a></li>
      <li><a href="#create_species">_create_species</a></li>
      <li><a href="#assign_identifier">_assign_identifier</a></li>
      <li><a href="#convert_seqtype">_convert_seqtype</a></li>
      <li><a href="#get_idstring">_get_idstring</a></li>
      <li><a href="#get_writer">_get_writer</a></li>
      <li><a href="#close_writer">close_writer</a></li>
    </ul>
  </li>
</ul>

<h1 id="NAME">NAME</h1>

<p>Bio::SeqIO::tinyseq - reading/writing sequences in NCBI TinySeq format</p>

<h1 id="SYNOPSIS">SYNOPSIS</h1>

<p>Do not use this module directly; use the SeqIO handler system:</p>

<pre><code><code>  <span class="variable">$stream</span> <span class="operator">=</span> <span class="variable">Bio::SeqIO</span><span class="operator">-&gt;</span><span class="variable">new</span><span class="operator">(</span> <span class="string">-file</span> <span class="operator">=&gt;</span> <span class="variable">$filename</span><span class="operator">,</span> <span class="string">-format</span> <span class="operator">=&gt;</span> <span class="string">'tinyseq'</span> <span class="operator">);</span>
  
  <span class="keyword">while</span> <span class="operator">(</span> <span class="keyword">my</span> <span class="variable">$seq</span> <span class="operator">=</span> <span class="variable">$stream</span><span class="operator">-&gt;</span><span class="variable">next_seq</span> <span class="operator">)</span> <span class="operator">{</span>
    <span class="operator">....</span>
  <span class="operator">}</span>
</code></code></pre>

<h1 id="DESCRIPTION">DESCRIPTION</h1>

<p>This object reads and writes Bio::Seq objects to and from TinySeq XML format. A TinySeq is a lightweight XML file of sequence information, analgous to FASTA format.</p>

<p>See <a href="http://www.ncbi.nlm.nih.gov/dtd/NCBI_TSeq.mod.dtd">http://www.ncbi.nlm.nih.gov/dtd/NCBI_TSeq.mod.dtd</a> for the DTD.</p>

<h1 id="FEEDBACK">FEEDBACK</h1>

<h2 id="Mailing-Lists">Mailing Lists</h2>

<p>User feedback is an integral part of the evolution of this and other Bioperl modules. Send your comments and suggestions preferably to the Bioperl mailing list. Your participation is much appreciated.</p>

<pre><code><code>  bioperl-l@bioperl.org                  - General discussion
  http://bioperl.org/wiki/Mailing_lists  - About the mailing lists</code></code></pre>

<h2 id="Support">Support</h2>

<p>Please direct usage questions or support issues to the mailing list:</p>

<p><i>bioperl-l@bioperl.org</i></p>

<p>rather than to the module maintainer directly. Many experienced and reponsive experts will be able look at the problem and quickly address it. Please include a thorough description of the problem with code and data examples if at all possible.</p>

<h2 id="Reporting-Bugs">Reporting Bugs</h2>

<p>Report bugs to the Bioperl bug tracking system to help us keep track of the bugs and their resolution. Bug reports can be submitted via the web:</p>

<pre><code><code>  https://redmine.open-bio.org/projects/bioperl/</code></code></pre>

<h1 id="SEE-ALSO">SEE ALSO</h1>

<p><a href="../../../../../.cpan/build/BioPerl-1.6.901-ghKY8U/blib/lib/Bio/SeqIO.html">Bio::SeqIO</a>, <a href="../../../../../.cpan/build/BioPerl-1.6.901-ghKY8U/blib/lib/Bio/Seq.html">Bio::Seq</a>.</p>

<h1 id="AUTHOR">AUTHOR</h1>

<p>Donald Jackson, &lt;donald.jackson@bms.com&gt;</p>

<p>Parts of this module and the test script were patterned after Sheldon McKay&#39;s <a href="../../../../../.cpan/build/BioPerl-1.6.901-ghKY8U/blib/lib/Bio/SeqIO/game.html">Bio::SeqIO::game</a>. If it breaks, however, it&#39;s my fault not his ;).</p>

<h1 id="APPENDIX">APPENDIX</h1>

<p>The rest of the documentation details each of the object methods. Internal methods are usually preceded with a _</p>

<h2 id="next_seq">next_seq</h2>

<pre><code><code>  Title         : next_seq
  Usage         : $seq = $stream-&gt;next_seq()
  Function      : returns the next sequence in the stream
  Returns       : Bio::Seq object
  Args          : NONE</code></code></pre>

<h2 id="write_seq">write_seq</h2>

<pre><code><code>  <span class="variable">Title</span>         <span class="operator">:</span> <span class="variable">write_seq</span>
  <span class="variable">Usage</span>         <span class="operator">:</span> <span class="variable">$seq</span> <span class="operator">=</span> <span class="variable">$stream</span><span class="operator">-&gt;</span><span class="variable">write_seq</span><span class="operator">(</span><span class="variable">@sequence_objects</span><span class="operator">);</span> <span class="keyword">undef</span> <span class="variable">$stream</span>
  <span class="variable">Function</span>      <span class="operator">:</span> <span class="variable">outputs</span> <span class="variable">one</span> <span class="keyword">or</span> <span class="variable">more</span> <span class="variable">sequence</span> <span class="variable">objects</span> <span class="variable">as</span> <span class="variable">TinySeq</span> <span class="variable">XML</span>
  <span class="variable">Returns</span>       <span class="operator">:</span> <span class="number">1</span> <span class="variable">on</span> <span class="variable">success</span>
  <span class="variable">Args</span>          <span class="operator">:</span> <span class="variable">one</span> <span class="keyword">or</span> <span class="variable">more</span> <span class="variable">sequence</span> <span class="variable">objects</span> <span class="variable">as</span> <span class="variable">TinySeq</span> <span class="variable">XML</span>
</code></code></pre>

<p>Because the TSeq dtd includes closing tags after all sets are written, the output will not be complete until the program terminates or the object is forced out of scope (see close_writer()). May not perfectly reproduce TSeq_sid element for all sequences</p>

<h2 id="get_seqs">_get_seqs</h2>

<pre><code><code>  Title         : _get_seqs
  Usage         : Internal function - use next_seq() instead
  Function      : parses the XML and creates Bio::Seq objects
  Returns       : 1 on success
  Args          : NONE</code></code></pre>

<p>Currently stores all sequence objects into memory. I will work on do more of a stream-based approach</p>

<h2 id="get_species">_get_species</h2>

<pre><code><code>  Title         : _get_species
  Usage         : Internal function
  Function      : gets a Bio::Species object from cache or creates as needed
  Returns       : a Bio::Species object on success, undef on failure
  Args          : a classification string (eg &#39;Homo sapiens&#39;) and
                  a NCBI taxon id (optional)</code></code></pre>

<p>Objects are cached for parsing multiple sequence files.</p>

<h2 id="create_species">_create_species</h2>

<pre><code><code>  Title         : _create_species
  Usage         : Internal function
  Function      : creates a Bio::Species object
  Returns       : a Bio::Species object on success, undef on failure
  Args          : a classification string (eg &#39;Homo sapiens&#39;) and
                  a NCBI taxon id (optional)</code></code></pre>

<h2 id="assign_identifier">_assign_identifier</h2>

<pre><code><code>  Title         : _assign_identifier
  Usage         : Internal function
  Function      : looks for sequence accession
  Returns       : 1 on success
  Args          : NONE</code></code></pre>

<p>NCBI puts refseq accessions in TSeq_sid, others in TSeq_accver.</p>

<h2 id="convert_seqtype">_convert_seqtype</h2>

<pre><code><code>  Title         : _convert_seqtype
  Usage         : Internal function
  Function      : maps Bio::Seq::alphabet() values [dna/rna/protein] onto
                  TSeq_seqtype values [protein/nucleotide]</code></code></pre>

<h2 id="get_idstring">_get_idstring</h2>

<pre><code><code>  Title         : _get_idstring
  Usage         : Internal function
  Function      : parse accession and version info from TSeq_accver
                  or TSeq_sid</code></code></pre>

<h2 id="get_writer">_get_writer</h2>

<pre><code><code>  Title         : _get_writer
  Usage         : Internal function
  Function      : instantiate XML::Writer object if needed,
                  output initial XML</code></code></pre>

<h2 id="close_writer">close_writer</h2>

<pre><code><code>  Title         : close_writer
  Usage         : $self-&gt;close_writer()
  Function      : terminate XML output
  Args          : NONE
  Returns       : 1 on success</code></code></pre>

<p>Called automatically by DESTROY when object goes out of scope</p>


</body>

</html>


