<?xml version="1.0" ?>
<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.0 Transitional//EN">
<html>
<head>
<!-- saved from url=(0017)http://localhost/ -->
<script language="JavaScript" src="../../../../../../displayToc.js"></script>
<script language="JavaScript" src="../../../../../../tocParas.js"></script>
<script language="JavaScript" src="../../../../../../tocTab.js"></script>
<link rel="stylesheet" type="text/css" href="../../../../../../scineplex.css">
<title></title>
<link rel="stylesheet" href="../../../../../../Active.css" type="text/css" />
<meta http-equiv="content-type" content="text/html; charset=utf-8" />
<link rev="made" href="mailto:support@ActiveState.com" />
</head>

<body>



<ul id="index">
  <li><a href="#NAME">NAME</a></li>
  <li><a href="#SYNOPSIS">SYNOPSIS</a></li>
  <li><a href="#DESCRIPTION">DESCRIPTION</a>
    <ul>
      <li><a href="#Helping-the-module-find-your-executable">Helping the module find your executable</a></li>
      <li><a href="#How-it-works">How it works</a></li>
    </ul>
  </li>
  <li><a href="#PARAMETER-FOR-ALIGNMENT-COMPUTATION">PARAMETER FOR ALIGNMENT COMPUTATION</a>
    <ul>
      <li><a href="#KTUPLE">KTUPLE</a></li>
      <li><a href="#TOPDIAGS">TOPDIAGS</a></li>
      <li><a href="#WINDOW">WINDOW</a></li>
      <li><a href="#PAIRGAP">PAIRGAP</a></li>
      <li><a href="#FIXEDGAP">FIXEDGAP</a></li>
      <li><a href="#FLOATGAP">FLOATGAP</a></li>
      <li><a href="#MATRIX">MATRIX</a></li>
      <li><a href="#TYPE">TYPE</a></li>
      <li><a href="#OUTPUT">OUTPUT</a></li>
      <li><a href="#OUTFILE">OUTFILE</a></li>
      <li><a href="#TRANSMIT">TRANSMIT</a></li>
    </ul>
  </li>
  <li><a href="#FEEDBACK">FEEDBACK</a>
    <ul>
      <li><a href="#Mailing-Lists">Mailing Lists</a></li>
      <li><a href="#Support">Support</a></li>
      <li><a href="#Reporting-Bugs">Reporting Bugs</a></li>
    </ul>
  </li>
  <li><a href="#AUTHOR---Peter-Schattner">AUTHOR - Peter Schattner</a></li>
  <li><a href="#CONTRIBUTORS">CONTRIBUTORS</a></li>
  <li><a href="#APPENDIX">APPENDIX</a>
    <ul>
      <li><a href="#program_name">program_name</a></li>
      <li><a href="#program_dir">program_dir</a></li>
      <li><a href="#version">version</a></li>
      <li><a href="#run">run</a></li>
      <li><a href="#align">align</a></li>
      <li><a href="#profile_align">profile_align</a></li>
      <li><a href="#add_sequences">add_sequences</a></li>
      <li><a href="#tree">tree</a></li>
      <li><a href="#footprint">footprint</a></li>
      <li><a href="#run1">_run</a></li>
      <li><a href="#setinput-">_setinput()</a></li>
      <li><a href="#setparams-">_setparams()</a></li>
    </ul>
  </li>
</ul>

<h1 id="NAME">NAME</h1>

<p>Bio::Tools::Run::Alignment::Clustalw - Object for the calculation of a multiple sequence alignment from a set of unaligned sequences or alignments using the Clustalw program</p>

<h1 id="SYNOPSIS">SYNOPSIS</h1>

<pre><code><code>  <span class="comment">#  Build a clustalw alignment factory</span>
  <span class="variable">@params</span> <span class="operator">=</span> <span class="operator">(</span><span class="string">'ktuple'</span> <span class="operator">=&gt;</span> <span class="number">2</span><span class="operator">,</span> <span class="string">'matrix'</span> <span class="operator">=&gt;</span> <span class="string">'BLOSUM'</span><span class="operator">);</span>
  <span class="variable">$factory</span> <span class="operator">=</span> <span class="variable">Bio::Tools::Run::Alignment::Clustalw</span><span class="operator">-&gt;</span><span class="variable">new</span><span class="operator">(</span><span class="variable">@params</span><span class="operator">);</span>
  
  <span class="comment">#  Pass the factory a list of sequences to be aligned.        </span>
  <span class="variable">$inputfilename</span> <span class="operator">=</span> <span class="string">'t/data/cysprot.fa'</span><span class="operator">;</span>
  <span class="variable">$aln</span> <span class="operator">=</span> <span class="variable">$factory</span><span class="operator">-&gt;</span><span class="variable">align</span><span class="operator">(</span><span class="variable">$inputfilename</span><span class="operator">);</span> <span class="comment"># $aln is a SimpleAlign object.</span>
  <span class="comment"># or</span>
  <span class="variable">$seq_array_ref</span> <span class="operator">=</span> <span class="operator">\</span><span class="variable">@seq_array</span><span class="operator">;</span>
  <span class="comment"># where @seq_array is an array of Bio::Seq objects</span>
  <span class="variable">$aln</span> <span class="operator">=</span> <span class="variable">$factory</span><span class="operator">-&gt;</span><span class="variable">align</span><span class="operator">(</span><span class="variable">$seq_array_ref</span><span class="operator">);</span>
  
  <span class="comment"># Or one can pass the factory a pair of (sub)alignments</span>
  <span class="comment">#to be aligned against each other, e.g.:</span>
  <span class="variable">$aln</span> <span class="operator">=</span> <span class="variable">$factory</span><span class="operator">-&gt;</span><span class="variable">profile_align</span><span class="operator">(</span><span class="variable">$aln1</span><span class="operator">,</span><span class="variable">$aln2</span><span class="operator">);</span>
  <span class="comment"># where $aln1 and $aln2 are Bio::SimpleAlign objects.</span>
  
  <span class="comment"># Or one can pass the factory an alignment and one or more unaligned</span>
  <span class="comment"># sequences to be added to the alignment. For example:        </span>
  <span class="variable">$aln</span> <span class="operator">=</span> <span class="variable">$factory</span><span class="operator">-&gt;</span><span class="variable">profile_align</span><span class="operator">(</span><span class="variable">$aln1</span><span class="operator">,</span><span class="variable">$seq</span><span class="operator">);</span> <span class="comment"># $seq is a Bio::Seq object.</span>
  
  <span class="comment"># Get a tree of the sequences</span>
  <span class="variable">$tree</span> <span class="operator">=</span> <span class="variable">$factory</span><span class="operator">-&gt;</span><span class="variable">tree</span><span class="operator">(\</span><span class="variable">@seq_array</span><span class="operator">);</span>
  
  <span class="comment"># Get both an alignment and a tree</span>
  <span class="operator">(</span><span class="variable">$aln</span><span class="operator">,</span> <span class="variable">$tree</span><span class="operator">)</span> <span class="operator">=</span> <span class="variable">$factory</span><span class="operator">-&gt;</span><span class="variable">run</span><span class="operator">(\</span><span class="variable">@seq_array</span><span class="operator">);</span>
  
  <span class="comment"># Do a footprinting analysis on the supplied sequences, getting back the</span>
  <span class="comment"># most conserved sub-alignments</span>
  <span class="keyword">my</span> <span class="variable">@results</span> <span class="operator">=</span> <span class="variable">$factory</span><span class="operator">-&gt;</span><span class="variable">footprint</span><span class="operator">(\</span><span class="variable">@seq_array</span><span class="operator">);</span>
  <span class="keyword">foreach</span> <span class="keyword">my</span> <span class="variable">$result</span> <span class="operator">(</span><span class="variable">@results</span><span class="operator">)</span> <span class="operator">{</span>
    <span class="keyword">print</span> <span class="variable">$result</span><span class="operator">-&gt;</span><span class="variable">consensus_string</span><span class="operator">,</span> <span class="string">"\n"</span><span class="operator">;</span>
  <span class="operator">}</span>
  
  <span class="comment"># There are various additional options and input formats available.</span>
  <span class="comment"># See the DESCRIPTION section that follows for additional details.</span>
</code></code></pre>

<h1 id="DESCRIPTION">DESCRIPTION</h1>

<p>Note: this DESCRIPTION only documents the Bioperl interface to Clustalw. Clustalw, itself, is a large &amp; complex program - for more information regarding clustalw, please see the clustalw documentation which accompanies the clustalw distribution. Clustalw is available from (among others) ftp://ftp.ebi.ac.uk/pub/software/. Clustalw.pm has only been tested using version 1.8 of clustalw. Compatibility with earlier versions of the clustalw program is currently unknown. Before running Clustalw successfully it will be necessary: to install clustalw on your system, and to ensure that users have execute privileges for the clustalw program.</p>

<h2 id="Helping-the-module-find-your-executable">Helping the module find your executable</h2>

<p>You will need to enable Clustalw to find the clustalw program. This can be done in (at least) three ways:</p>

<pre><code><code> <span class="number">1.</span> <span class="variable">Make</span> <span class="variable">sure</span> <span class="variable">the</span> <span class="variable">clustalw</span> <span class="variable">executable</span> <span class="variable">is</span> <span class="variable">in</span> <span class="variable">your</span> <span class="variable">path</span> <span class="variable">so</span> <span class="variable">that</span>
    <span class="variable">which</span> <span class="variable">clustalw</span>
    <span class="variable">returns</span> <span class="variable">a</span> <span class="variable">clustalw</span> <span class="variable">executable</span> <span class="variable">on</span> <span class="variable">your</span> <span class="keyword">system</span><span class="operator">.</span>
 
 <span class="number">2.</span> <span class="variable">Define</span> <span class="variable">an</span> <span class="variable">environmental</span> <span class="variable">variable</span> <span class="variable">CLUSTALDIR</span> <span class="variable">which</span> <span class="variable">is</span> <span class="variable">a</span> 
    <span class="variable">directory</span> <span class="variable">which</span> <span class="variable">contains</span> <span class="variable">the</span> <span class="string">'clustalw'</span> <span class="variable">application</span><span class="operator">:</span>
    <span class="variable">In</span> <span class="variable">bash</span><span class="operator">:</span>
 
    <span class="variable">export</span> <span class="variable">CLUSTALDIR</span><span class="operator">=</span><span class="regex">/home/username</span><span class="operator">/</span><span class="variable">clustalw1</span><span class="number">.8</span>
 
    <span class="variable">In</span> <span class="variable">csh</span><span class="operator">/</span><span class="variable">tcsh</span><span class="operator">:</span>
 
    <span class="variable">setenv</span> <span class="variable">CLUSTALDIR</span> <span class="operator">/</span><span class="variable">home</span><span class="operator">/</span><span class="variable">username</span><span class="operator">/</span><span class="variable">clustalw1</span><span class="number">.8</span>
 
 <span class="number">3.</span> <span class="variable">Include</span> <span class="variable">a</span> <span class="variable">definition</span> <span class="variable">of</span> <span class="variable">an</span> <span class="variable">environmental</span> <span class="variable">variable</span> <span class="variable">CLUSTALDIR</span> <span class="variable">in</span>
    <span class="variable">every</span> <span class="variable">script</span> <span class="variable">that</span> <span class="variable">will</span> <span class="keyword">use</span> <span class="variable">this</span> <span class="variable">Clustalw</span> <span class="variable">wrapper</span> <span class="variable">module</span><span class="operator">,</span> <span class="variable">e</span><span class="operator">.</span><span class="variable">g</span><span class="operator">.:</span>
 
    <span class="keyword">BEGIN</span> <span class="operator">{</span> <span class="variable">$ENV</span><span class="operator">{</span><span class="string">CLUSTALDIR</span><span class="operator">}</span> <span class="operator">=</span> <span class="string">'/home/username/clustalw1.8/'</span> <span class="operator">}</span>
    <span class="keyword">use</span> <span class="variable">Bio::Tools::Run::Alignment::Clustalw</span><span class="operator">;</span>
</code></code></pre>

<p>If you are running an application on a webserver make sure the webserver environment has the proper PATH set or use the options 2 or 3 to set the variables.</p>

<h2 id="How-it-works">How it works</h2>

<p>Bio::Tools::Run::Alignment::Clustalw is an object for performing a multiple sequence alignment from a set of unaligned sequences and/or sub-alignments by means of the clustalw program.</p>

<p>Initially, a clustalw &quot;factory object&quot; is created. Optionally, the factory may be passed most of the parameters or switches of the clustalw program, e.g.:</p>

<pre><code><code>        <span class="variable">@params</span> <span class="operator">=</span> <span class="operator">(</span><span class="string">'ktuple'</span> <span class="operator">=&gt;</span> <span class="number">2</span><span class="operator">,</span> <span class="string">'matrix'</span> <span class="operator">=&gt;</span> <span class="string">'BLOSUM'</span><span class="operator">);</span>
        <span class="variable">$factory</span> <span class="operator">=</span> <span class="variable">Bio::Tools::Run::Alignment::Clustalw</span><span class="operator">-&gt;</span><span class="variable">new</span><span class="operator">(</span><span class="variable">@params</span><span class="operator">);</span>
</code></code></pre>

<p>Any parameters not explicitly set will remain as the defaults of the clustalw program. Additional parameters and switches (not available in clustalw) may also be set. Currently, the only such parameter is &quot;quiet&quot;, which when set to a non-zero value, suppresses clustalw terminal output. Not all clustalw parameters are supported at this stage.</p>

<p>By default, Clustalw output is returned solely in a the form of a Bio::SimpleAlign object which can then be printed and/or saved in multiple formats using the AlignIO.pm module. Optionally the raw clustalw output file can be saved if the calling script specifies an output file (with the clustalw parameter OUTFILE). Currently only the GCG-MSF output file formats is supported.</p>

<p>Not all parameters and features have been implemented yet in Perl format.</p>

<p>Alignment parameters can be changed and/or examined at any time after the factory has been created. The program checks that any parameter/switch being set/read is valid. However, currently no additional checks are included to check that parameters are of the proper type (eg string or numeric) or that their values are within the proper range. As an example, to change the value of the clustalw parameter ktuple to 3 and subsequently to check its value one would write:</p>

<pre><code><code>        <span class="variable">$ktuple</span> <span class="operator">=</span> <span class="number">3</span><span class="operator">;</span>
        <span class="variable">$factory</span><span class="operator">-&gt;</span><span class="variable">ktuple</span><span class="operator">(</span><span class="variable">$ktuple</span><span class="operator">);</span>
        <span class="variable">$get_ktuple</span> <span class="operator">=</span> <span class="variable">$factory</span><span class="operator">-&gt;</span><span class="variable">ktuple</span><span class="operator">();</span>
</code></code></pre>

<p>Once the factory has been created and the appropriate parameters set, one can call the method align() to align a set of unaligned sequences, or call profile_align() to add one or more sequences or a second alignment to an initial alignment.</p>

<p>Input to align() may consist of a set of unaligned sequences in the form of the name of file containing the sequences. For example,</p>

<pre><code><code>  <span class="variable">$inputfilename</span> <span class="operator">=</span> <span class="string">'t/data/cysprot.fa'</span><span class="operator">;</span> 
  <span class="variable">$aln</span> <span class="operator">=</span> <span class="variable">$factory</span><span class="operator">-</span><span class="variable">E</span><span class="operator">&lt;</span><span class="keyword">gt</span><span class="operator">&gt;</span><span class="variable">align</span><span class="operator">(</span><span class="variable">$inputfilename</span><span class="operator">);</span>
</code></code></pre>

<p>Alternately one can create an array of Bio::Seq objects somehow</p>

<pre><code><code>        <span class="variable">$str</span> <span class="operator">=</span> <span class="variable">Bio::SeqIO</span><span class="operator">-&gt;</span><span class="variable">new</span><span class="operator">(</span><span class="string">-file</span><span class="operator">=&gt;</span> <span class="string">'t/data/cysprot.fa'</span><span class="operator">,</span> <span class="string">-format</span> <span class="operator">=&gt;</span> <span class="string">'Fasta'</span><span class="operator">);</span>
        <span class="variable">@seq_array</span> <span class="operator">=();</span>
        <span class="keyword">while</span> <span class="operator">(</span> <span class="keyword">my</span> <span class="variable">$seq</span> <span class="operator">=</span> <span class="variable">$str</span><span class="operator">-&gt;</span><span class="variable">next_seq</span><span class="operator">()</span> <span class="operator">)</span> <span class="operator">{</span><span class="keyword">push</span> <span class="operator">(</span><span class="variable">@seq_array</span><span class="operator">,</span> <span class="variable">$seq</span><span class="operator">)</span> <span class="operator">;}</span>
</code></code></pre>

<p>and pass the factory a reference to that array</p>

<pre><code><code>        <span class="variable">$seq_array_ref</span> <span class="operator">=</span> <span class="operator">\</span><span class="variable">@seq_array</span><span class="operator">;</span>
        <span class="variable">$aln</span> <span class="operator">=</span> <span class="variable">$factory</span><span class="operator">-&gt;</span><span class="variable">align</span><span class="operator">(</span><span class="variable">$seq_array_ref</span><span class="operator">);</span>
</code></code></pre>

<p>In either case, align() returns a reference to a SimpleAlign object which can then used (see <a href="../../../../../../lib/perl5/Bio/SimpleAlign.html">Bio::SimpleAlign</a>).</p>

<p>Once an initial alignment exists, one can pass the factory additional sequence(s) to be added (ie aligned) to the original alignment. The alignment can be passed as either an alignment file or a Bio:SimpleAlign object. The unaligned sequence(s) can be passed as a filename or as an array of BioPerl sequence objects or as a single BioPerl Seq object. For example (to add a single sequence to an alignment),</p>

<pre><code><code>        <span class="variable">$str</span> <span class="operator">=</span> <span class="variable">Bio::AlignIO</span><span class="operator">-&gt;</span><span class="variable">new</span><span class="operator">(</span><span class="string">-file</span><span class="operator">=&gt;</span> <span class="string">'t/data/cysprot1a.msf'</span><span class="operator">);</span>
        <span class="variable">$aln</span> <span class="operator">=</span> <span class="variable">$str</span><span class="operator">-&gt;</span><span class="variable">next_aln</span><span class="operator">();</span>
        <span class="variable">$str1</span> <span class="operator">=</span> <span class="variable">Bio::SeqIO</span><span class="operator">-&gt;</span><span class="variable">new</span><span class="operator">(</span><span class="string">-file</span><span class="operator">=&gt;</span> <span class="string">'t/data/cysprot1b.fa'</span><span class="operator">);</span>
        <span class="variable">$seq</span> <span class="operator">=</span> <span class="variable">$str1</span><span class="operator">-&gt;</span><span class="variable">next_seq</span><span class="operator">();</span>
        <span class="variable">$aln</span> <span class="operator">=</span> <span class="variable">$factory</span><span class="operator">-&gt;</span><span class="variable">profile_align</span><span class="operator">(</span><span class="variable">$aln</span><span class="operator">,</span><span class="variable">$seq</span><span class="operator">);</span>
</code></code></pre>

<p>In either case, profile_align() returns a reference to a SimpleAlign object containing a new SimpleAlign object of the alignment with the additional sequence(s) added in.</p>

<p>Finally one can pass the factory a pair of (sub)alignments to be aligned against each other. The alignments can be passed in the form of either a pair of alignment files or a pair of Bio:SimpleAlign objects. For example,</p>

<pre><code><code>        <span class="variable">$profile1</span> <span class="operator">=</span> <span class="string">'t/data/cysprot1a.msf'</span><span class="operator">;</span>
        <span class="variable">$profile2</span> <span class="operator">=</span> <span class="string">'t/data/cysprot1b.msf'</span><span class="operator">;</span>
        <span class="variable">$aln</span> <span class="operator">=</span> <span class="variable">$factory</span><span class="operator">-&gt;</span><span class="variable">profile_align</span><span class="operator">(</span><span class="variable">$profile1</span><span class="operator">,</span><span class="variable">$profile2</span><span class="operator">);</span>
</code></code></pre>

<p>or</p>

<pre><code><code>        <span class="variable">$str1</span> <span class="operator">=</span> <span class="variable">Bio::AlignIO</span><span class="operator">-&gt;</span><span class="variable">new</span><span class="operator">(</span><span class="string">-file</span><span class="operator">=&gt;</span> <span class="string">'t/data/cysprot1a.msf'</span><span class="operator">);</span>
        <span class="variable">$aln1</span> <span class="operator">=</span> <span class="variable">$str1</span><span class="operator">-&gt;</span><span class="variable">next_aln</span><span class="operator">();</span>
        <span class="variable">$str2</span> <span class="operator">=</span> <span class="variable">Bio::AlignIO</span><span class="operator">-&gt;</span><span class="variable">new</span><span class="operator">(</span><span class="string">-file</span><span class="operator">=&gt;</span> <span class="string">'t/data/cysprot1b.msf'</span><span class="operator">);</span>
        <span class="variable">$aln2</span> <span class="operator">=</span> <span class="variable">$str2</span><span class="operator">-&gt;</span><span class="variable">next_aln</span><span class="operator">();</span>
        <span class="variable">$aln</span> <span class="operator">=</span> <span class="variable">$factory</span><span class="operator">-&gt;</span><span class="variable">profile_align</span><span class="operator">(</span><span class="variable">$aln1</span><span class="operator">,</span><span class="variable">$aln2</span><span class="operator">);</span>
</code></code></pre>

<p>In either case, profile_align() returns a reference to a SimpleAlign object containing an (super)alignment of the two input alignments.</p>

<p>For more examples of syntax and use of Clustalw, the user is encouraged to look at the script Clustalw.t in the t/ directory.</p>

<p>Note: Clustalw is still under development. Various features of the clustalw program have not yet been implemented. If you would like that a specific clustalw feature be added to this perl contact bioperl-l@bioperl.org.</p>

<p>These can be specified as paramters when instantiating a new Clustalw object, or through get/set methods of the same name (lowercase).</p>

<h1 id="PARAMETER-FOR-ALIGNMENT-COMPUTATION">PARAMETER FOR ALIGNMENT COMPUTATION</h1>

<h2 id="KTUPLE">KTUPLE</h2>

<pre><code><code> <span class="variable">Title</span>       <span class="operator">:</span> <span class="variable">KTUPLE</span>
 <span class="variable">Description</span> <span class="operator">:</span> <span class="operator">(</span><span class="variable">optional</span><span class="operator">)</span> <span class="variable">set</span> <span class="variable">the</span> <span class="variable">word</span> <span class="variable">size</span> <span class="variable">to</span> <span class="variable">be</span> <span class="variable">used</span> <span class="variable">in</span> <span class="variable">the</span> <span class="variable">alignment</span>
               <span class="variable">This</span> <span class="variable">is</span> <span class="variable">the</span> <span class="variable">size</span> <span class="variable">of</span> <span class="variable">exactly</span> <span class="variable">matching</span> <span class="variable">fragment</span> <span class="variable">that</span> <span class="variable">is</span> <span class="variable">used</span><span class="operator">.</span>
               <span class="variable">INCREASE</span> <span class="keyword">for</span> <span class="variable">speed</span> <span class="operator">(</span><span class="variable">max</span><span class="operator">=</span> <span class="number">2</span> <span class="keyword">for</span> <span class="variable">proteins</span><span class="operator">;</span> <span class="number">4</span> <span class="keyword">for</span> <span class="variable">DNA</span><span class="operator">),</span>
               <span class="variable">DECREASE</span> <span class="keyword">for</span> <span class="variable">sensitivity</span><span class="operator">.</span>
               <span class="variable">For</span> <span class="variable">longer</span> <span class="variable">sequences</span> <span class="operator">(</span><span class="variable">e</span><span class="operator">.</span><span class="variable">g</span><span class="operator">.</span> <span class="operator">&gt;</span><span class="number">1000</span> <span class="variable">residues</span><span class="operator">)</span> <span class="variable">you</span> <span class="variable">may</span>
               <span class="variable">need</span> <span class="variable">to</span> <span class="variable">increase</span> <span class="variable">the</span> <span class="keyword">default</span>
</code></code></pre>

<h2 id="TOPDIAGS">TOPDIAGS</h2>

<pre><code><code> <span class="variable">Title</span>       <span class="operator">:</span> <span class="variable">TOPDIAGS</span>
 <span class="variable">Description</span> <span class="operator">:</span> <span class="operator">(</span><span class="variable">optional</span><span class="operator">)</span> <span class="variable">number</span> <span class="variable">of</span> <span class="variable">best</span> <span class="variable">diagonals</span> <span class="variable">to</span> <span class="keyword">use</span>
               <span class="variable">The</span> <span class="variable">number</span> <span class="variable">of</span> <span class="variable">k</span><span class="operator">-</span><span class="variable">tuple</span> <span class="variable">matches</span> <span class="variable">on</span> <span class="keyword">each</span> <span class="variable">diagonal</span>
               <span class="operator">(</span><span class="variable">in</span> <span class="variable">an</span> <span class="variable">imaginary</span> <span class="variable">dot</span><span class="operator">-</span><span class="variable">matrix</span> <span class="variable">plot</span><span class="operator">)</span> <span class="variable">is</span> <span class="variable">calculated</span><span class="operator">.</span>
               <span class="variable">Only</span> <span class="variable">the</span> <span class="variable">best</span> <span class="variable">ones</span> <span class="operator">(</span><span class="variable">with</span> <span class="variable">most</span> <span class="variable">matches</span><span class="operator">)</span> <span class="variable">are</span> <span class="variable">used</span> <span class="variable">in</span>
               <span class="variable">the</span> <span class="variable">alignment</span><span class="operator">.</span>  <span class="variable">This</span> <span class="variable">parameter</span> <span class="variable">specifies</span> <span class="variable">how</span> <span class="variable">many</span><span class="operator">.</span>
               <span class="variable">Decrease</span> <span class="keyword">for</span> <span class="variable">speed</span><span class="operator">;</span> <span class="variable">increase</span> <span class="keyword">for</span> <span class="variable">sensitivity</span><span class="operator">.</span>
</code></code></pre>

<h2 id="WINDOW">WINDOW</h2>

<pre><code><code> <span class="variable">Title</span>       <span class="operator">:</span> <span class="variable">WINDOW</span>
 <span class="variable">Description</span> <span class="operator">:</span> <span class="operator">(</span><span class="variable">optional</span><span class="operator">)</span> <span class="variable">window</span> <span class="variable">size</span>
               <span class="variable">This</span> <span class="variable">is</span> <span class="variable">the</span> <span class="variable">number</span> <span class="variable">of</span> <span class="variable">diagonals</span> <span class="variable">around</span> <span class="keyword">each</span> <span class="variable">of</span> <span class="variable">the</span> <span class="string">'best'</span>
               <span class="variable">diagonals</span> <span class="variable">that</span> <span class="variable">will</span> <span class="variable">be</span> <span class="variable">used</span><span class="operator">.</span>  <span class="variable">Decrease</span> <span class="keyword">for</span> <span class="variable">speed</span><span class="operator">;</span>
               <span class="variable">increase</span> <span class="keyword">for</span> <span class="variable">sensitivity</span><span class="operator">.</span>
</code></code></pre>

<h2 id="PAIRGAP">PAIRGAP</h2>

<pre><code><code> Title       : PAIRGAP
 Description : (optional) gap penalty for pairwise alignments
               This is a penalty for each gap in the fast alignments.
               It has little affect on the speed or sensitivity except
               for extreme values.</code></code></pre>

<h2 id="FIXEDGAP">FIXEDGAP</h2>

<pre><code><code> Title       : FIXEDGAP
 Description : (optional) fixed length gap penalty</code></code></pre>

<h2 id="FLOATGAP">FLOATGAP</h2>

<pre><code><code> Title       : FLOATGAP
 Description : (optional) variable length gap penalty</code></code></pre>

<h2 id="MATRIX">MATRIX</h2>

<pre><code><code> Title       : MATRIX
 Default     : PAM100 for DNA - PAM250 for protein alignment
 Description : (optional) substitution matrix used in the multiple
               alignments. Depends on the version of clustalw as to
               what default matrix will be used

               PROTEIN WEIGHT MATRIX leads to a new menu where you are
               offered a choice of weight matrices. The default for
               proteins in version 1.8 is the PAM series derived by
               Gonnet and colleagues. Note, a series is used! The
               actual matrix that is used depends on how similar the
               sequences to be aligned at this alignment step
               are. Different matrices work differently at each
               evolutionary distance.

               DNA WEIGHT MATRIX leads to a new menu where a single
               matrix (not a series) can be selected. The default is
               the matrix used by BESTFIT for comparison of nucleic
               acid sequences.</code></code></pre>

<h2 id="TYPE">TYPE</h2>

<pre><code><code> Title       : TYPE
 Description : (optional) sequence type: protein or DNA. This allows
                    you to explicitly overide the programs attempt at
                    guessing the type of the sequence.  It is only useful
                    if you are using sequences with a VERY strange
                    composition.</code></code></pre>

<h2 id="OUTPUT">OUTPUT</h2>

<pre><code><code> Title       : OUTPUT
 Description : (optional) clustalw supports GCG or PHYLIP or PIR or
                Clustal format.  See the Bio::AlignIO modules for
                which formats are supported by bioperl.</code></code></pre>

<h2 id="OUTFILE">OUTFILE</h2>

<pre><code><code> Title       : OUTFILE
 Description : (optional) Name of clustalw output file. If not set
                    module will erase output file.  In any case alignment will
                    be returned in the form of SimpleAlign objects</code></code></pre>

<h2 id="TRANSMIT">TRANSMIT</h2>

<pre><code><code> Title       : TRANSMIT
 Description : (optional) transitions not weighted.  The default is to
                    weight transitions as more favourable than other
                    mismatches in DNA alignments.  This switch makes all
                    nucleotide mismatches equally weighted.</code></code></pre>

<h1 id="FEEDBACK">FEEDBACK</h1>

<h2 id="Mailing-Lists">Mailing Lists</h2>

<p>User feedback is an integral part of the evolution of this and other Bioperl modules. Send your comments and suggestions preferably to one of the Bioperl mailing lists. Your participation is much appreciated.</p>

<pre><code><code>  bioperl-l@bioperl.org                  - General discussion
  http://bioperl.org/wiki/Mailing_lists  - About the mailing lists</code></code></pre>

<h2 id="Support">Support</h2>

<p>Please direct usage questions or support issues to the mailing list:</p>

<p><i>bioperl-l@bioperl.org</i></p>

<p>rather than to the module maintainer directly. Many experienced and reponsive experts will be able look at the problem and quickly address it. Please include a thorough description of the problem with code and data examples if at all possible.</p>

<h2 id="Reporting-Bugs">Reporting Bugs</h2>

<p>Report bugs to the Bioperl bug tracking system to help us keep track the bugs and their resolution. Bug reports can be submitted via the web:</p>

<pre><code><code>  http://redmine.open-bio.org/projects/bioperl/</code></code></pre>

<h1 id="AUTHOR---Peter-Schattner">AUTHOR - Peter Schattner</h1>

<p>Email schattner@alum.mit.edu</p>

<h1 id="CONTRIBUTORS">CONTRIBUTORS</h1>

<p>Jason Stajich jason-AT-bioperl_DOT_org Sendu Bala bix@sendu.me.uk</p>

<h1 id="APPENDIX">APPENDIX</h1>

<p>The rest of the documentation details each of the object methods. Internal methods are usually preceded with a _</p>

<h2 id="program_name">program_name</h2>

<pre><code><code> Title   : program_name
 Usage   : $factory&gt;program_name()
 Function: holds the program name
 Returns:  string
 Args    : None</code></code></pre>

<h2 id="program_dir">program_dir</h2>

<pre><code><code> Title   : program_dir
 Usage   : $factory-&gt;program_dir(@params)
 Function: returns the program directory, obtained from ENV variable.
 Returns:  string
 Args    :</code></code></pre>

<h2 id="version">version</h2>

<pre><code><code> Title   : version
 Usage   : exit if $prog-&gt;version() &lt; 1.8
 Function: Determine the version number of the program
 Example :
 Returns : float or undef
 Args    : none</code></code></pre>

<h2 id="run">run</h2>

<pre><code><code> <span class="variable">Title</span>   <span class="operator">:</span> <span class="variable">run</span>
 <span class="variable">Usage</span>   <span class="operator">:</span> <span class="operator">(</span><span class="variable">$aln</span><span class="operator">,</span> <span class="variable">$tree</span><span class="operator">)</span> <span class="operator">=</span> <span class="variable">$factory</span><span class="operator">-&gt;</span><span class="variable">run</span><span class="operator">(</span><span class="variable">$inputfilename</span><span class="operator">);</span>
           <span class="operator">(</span><span class="variable">$aln</span><span class="operator">,</span> <span class="variable">$tree</span><span class="operator">)</span> <span class="operator">=</span> <span class="variable">$factory</span><span class="operator">-&gt;</span><span class="variable">run</span><span class="operator">(</span><span class="variable">$seq_array_ref</span><span class="operator">);</span>
 <span class="variable">Function</span><span class="operator">:</span> <span class="variable">Perform</span> <span class="variable">a</span> <span class="variable">multiple</span> <span class="variable">sequence</span> <span class="variable">alignment</span><span class="operator">,</span> <span class="variable">generating</span> <span class="variable">a</span> <span class="variable">tree</span> <span class="variable">at</span> <span class="variable">the</span> <span class="variable">same</span>
           <span class="keyword">time</span><span class="operator">.</span> <span class="operator">(</span><span class="variable">Like</span> <span class="variable">align</span><span class="operator">()</span> <span class="keyword">and</span> <span class="variable">tree</span><span class="operator">()</span> <span class="variable">combined</span><span class="operator">.)</span>
 <span class="variable">Returns</span> <span class="operator">:</span> <span class="variable">A</span> <span class="variable">SimpleAlign</span> <span class="variable">object</span> <span class="variable">containing</span> <span class="variable">the</span> <span class="variable">sequence</span> <span class="variable">alignment</span> <span class="keyword">and</span> <span class="variable">a</span>
           <span class="variable">Bio::Tree::Tree</span> <span class="variable">object</span> <span class="variable">with</span> <span class="variable">the</span> <span class="variable">tree</span> <span class="variable">relating</span> <span class="variable">the</span> <span class="variable">sequences</span><span class="operator">.</span>
 <span class="variable">Args</span>    <span class="operator">:</span> <span class="variable">Name</span> <span class="variable">of</span> <span class="variable">a</span> <span class="variable">file</span> <span class="variable">containing</span> <span class="variable">a</span> <span class="variable">set</span> <span class="variable">of</span> <span class="variable">unaligned</span> <span class="variable">fasta</span> <span class="variable">sequences</span>
           <span class="keyword">or</span> <span class="keyword">else</span> <span class="variable">an</span> <span class="variable">array</span> <span class="variable">of</span> <span class="variable">references</span> <span class="variable">to</span> <span class="variable">Bio::Seq</span> <span class="variable">objects</span><span class="operator">.</span>
</code></code></pre>

<h2 id="align">align</h2>

<pre><code><code> <span class="variable">Title</span>   <span class="operator">:</span> <span class="variable">align</span>
 <span class="variable">Usage</span>   <span class="operator">:</span> <span class="variable">$inputfilename</span> <span class="operator">=</span> <span class="string">'t/data/cysprot.fa'</span><span class="operator">;</span>
           <span class="variable">$aln</span> <span class="operator">=</span> <span class="variable">$factory</span><span class="operator">-&gt;</span><span class="variable">align</span><span class="operator">(</span><span class="variable">$inputfilename</span><span class="operator">);</span>
           <span class="keyword">or</span>
           <span class="variable">$seq_array_ref</span> <span class="operator">=</span> <span class="operator">\</span><span class="variable">@seq_array</span><span class="operator">;</span> <span class="comment"># @seq_array is array of Seq objs</span>
           <span class="variable">$aln</span> <span class="operator">=</span> <span class="variable">$factory</span><span class="operator">-&gt;</span><span class="variable">align</span><span class="operator">(</span><span class="variable">$seq_array_ref</span><span class="operator">);</span>
 <span class="variable">Function</span><span class="operator">:</span> <span class="variable">Perform</span> <span class="variable">a</span> <span class="variable">multiple</span> <span class="variable">sequence</span> <span class="variable">alignment</span>
 <span class="variable">Returns</span> <span class="operator">:</span> <span class="variable">Reference</span> <span class="variable">to</span> <span class="variable">a</span> <span class="variable">SimpleAlign</span> <span class="variable">object</span> <span class="variable">containing</span> <span class="variable">the</span>
           <span class="variable">sequence</span> <span class="variable">alignment</span><span class="operator">.</span>
 <span class="variable">Args</span>    <span class="operator">:</span> <span class="variable">Name</span> <span class="variable">of</span> <span class="variable">a</span> <span class="variable">file</span> <span class="variable">containing</span> <span class="variable">a</span> <span class="variable">set</span> <span class="variable">of</span> <span class="variable">unaligned</span> <span class="variable">fasta</span> <span class="variable">sequences</span>
           <span class="keyword">or</span> <span class="keyword">else</span> <span class="variable">an</span> <span class="variable">array</span> <span class="variable">of</span> <span class="variable">references</span> <span class="variable">to</span> <span class="variable">Bio::Seq</span> <span class="variable">objects</span><span class="operator">.</span>
 
 <span class="variable">Throws</span> <span class="variable">an</span> <span class="variable">exception</span> <span class="keyword">if</span> <span class="variable">argument</span> <span class="variable">is</span> <span class="keyword">not</span> <span class="variable">either</span> <span class="variable">a</span> <span class="variable">string</span> <span class="operator">(</span><span class="variable">eg</span> <span class="variable">a</span>
 <span class="variable">filename</span><span class="operator">)</span> <span class="keyword">or</span> <span class="variable">a</span> <span class="variable">reference</span> <span class="variable">to</span> <span class="variable">an</span> <span class="variable">array</span> <span class="variable">of</span> <span class="variable">Bio::Seq</span> <span class="variable">objects</span><span class="operator">.</span>  <span class="variable">If</span>
 <span class="variable">argument</span> <span class="variable">is</span> <span class="variable">string</span><span class="operator">,</span> <span class="variable">throws</span> <span class="variable">exception</span> <span class="keyword">if</span> <span class="variable">file</span> <span class="variable">corresponding</span> <span class="variable">to</span> <span class="variable">string</span>
 <span class="variable">name</span> <span class="variable">can</span> <span class="keyword">not</span> <span class="variable">be</span> <span class="variable">found</span><span class="operator">.</span> <span class="variable">If</span> <span class="variable">argument</span> <span class="variable">is</span> <span class="variable">Bio::Seq</span> <span class="variable">array</span><span class="operator">,</span> <span class="variable">throws</span>
 <span class="variable">exception</span> <span class="keyword">if</span> <span class="variable">less</span> <span class="variable">than</span> <span class="variable">two</span> <span class="variable">sequence</span> <span class="variable">objects</span> <span class="variable">are</span> <span class="variable">in</span> <span class="variable">array</span><span class="operator">.</span>
</code></code></pre>

<h2 id="profile_align">profile_align</h2>

<pre><code><code> <span class="variable">Title</span>   <span class="operator">:</span> <span class="variable">profile_align</span>
 <span class="variable">Usage</span>   <span class="operator">:</span> <span class="variable">$aln</span> <span class="operator">=</span> <span class="variable">$factory</span><span class="operator">-&gt;</span><span class="variable">profile_align</span><span class="operator">(</span><span class="variable">@simple_aligns</span><span class="operator">);</span>
           <span class="keyword">or</span>
           <span class="variable">$aln</span> <span class="operator">=</span> <span class="variable">$factory</span><span class="operator">-&gt;</span><span class="variable">profile_align</span><span class="operator">(</span><span class="variable">@subalignment_filenames</span><span class="operator">);</span>
 <span class="variable">Function</span><span class="operator">:</span> <span class="variable">Perform</span> <span class="variable">an</span> <span class="variable">alignment</span> <span class="variable">of</span> <span class="number">2</span> <span class="operator">(</span><span class="variable">sub</span><span class="operator">)</span><span class="variable">alignments</span>
 <span class="variable">Returns</span> <span class="operator">:</span> <span class="variable">Reference</span> <span class="variable">to</span> <span class="variable">a</span> <span class="variable">SimpleAlign</span> <span class="variable">object</span> <span class="variable">containing</span> <span class="variable">the</span> <span class="operator">(</span><span class="variable">super</span><span class="operator">)</span><span class="variable">alignment</span><span class="operator">.</span>
 <span class="variable">Args</span>    <span class="operator">:</span> <span class="variable">Names</span> <span class="variable">of</span> <span class="number">2</span> <span class="variable">files</span> <span class="variable">containing</span> <span class="variable">the</span> <span class="variable">subalignments</span>
           <span class="keyword">or</span> <span class="variable">references</span> <span class="variable">to</span> <span class="number">2</span> <span class="variable">Bio::SimpleAlign</span> <span class="variable">objects</span><span class="operator">.</span>
</code></code></pre>

<p>Throws an exception if arguments are not either strings (eg filenames) or references to SimpleAlign objects.</p>

<h2 id="add_sequences">add_sequences</h2>

<pre><code><code> Title   : add_sequences
 Usage   :
 Function: Align and add sequences into an alignment
 Example :
 Returns : Reference to a SimpleAlign object containing the (super)alignment.
 Args    : Names of 2 files, the first one containing an alignment and the second one containing sequences to be added
         or references to 2 Bio::SimpleAlign objects.</code></code></pre>

<p>Throws an exception if arguments are not either strings (eg filenames) or references to SimpleAlign objects.</p>

<h2 id="tree">tree</h2>

<pre><code><code> <span class="variable">Title</span>   <span class="operator">:</span> <span class="variable">tree</span>
 <span class="variable">Usage</span>   <span class="operator">:</span> <span class="variable">@params</span> <span class="operator">=</span> <span class="operator">(</span><span class="string">'bootstrap'</span> <span class="operator">=&gt;</span> <span class="number">1000</span><span class="operator">,</span> 
                            <span class="string">'tossgaps'</span>  <span class="operator">=&gt;</span> <span class="number">1</span><span class="operator">,</span> 
                            <span class="string">'kimura'</span>    <span class="operator">=&gt;</span> <span class="number">1</span><span class="operator">,</span> 
                            <span class="string">'seed'</span>      <span class="operator">=&gt;</span> <span class="number">121</span><span class="operator">,</span> 
                            <span class="string">'bootlabels'</span><span class="operator">=&gt;</span> <span class="string">'nodes'</span><span class="operator">,</span> 
                            <span class="string">'quiet'</span>     <span class="operator">=&gt;</span> <span class="number">1</span><span class="operator">);</span>
           <span class="variable">$factory</span> <span class="operator">=</span> <span class="variable">Bio::Tools::Run::Alignment::Clustalw</span><span class="operator">-&gt;</span><span class="variable">new</span><span class="operator">(</span><span class="variable">@params</span><span class="operator">);</span>
           <span class="variable">$tree_obj</span> <span class="operator">=</span> <span class="variable">$factory</span><span class="operator">-&gt;</span><span class="variable">tree</span><span class="operator">(</span><span class="variable">$aln_obj</span><span class="operator">);</span>
           <span class="keyword">or</span>
           <span class="variable">$tree_obj</span> <span class="operator">=</span> <span class="variable">$factory</span><span class="operator">-&gt;</span><span class="variable">tree</span><span class="operator">(</span><span class="variable">$treefilename</span><span class="operator">);</span>
 <span class="variable">Function</span><span class="operator">:</span> <span class="variable">Retrieve</span> <span class="variable">a</span> <span class="variable">tree</span> <span class="variable">corresponding</span> <span class="variable">to</span> <span class="variable">the</span> <span class="variable">input</span>
 <span class="variable">Returns</span> <span class="operator">:</span> <span class="variable">Bio::TreeIO</span> <span class="variable">object</span>
 <span class="variable">Args</span>    <span class="operator">:</span> <span class="variable">Bio::SimpleAlign</span> <span class="keyword">or</span> <span class="variable">filename</span> <span class="variable">of</span> <span class="variable">a</span> <span class="variable">tree</span>
</code></code></pre>

<h2 id="footprint">footprint</h2>

<pre><code><code> <span class="variable">Title</span>   <span class="operator">:</span> <span class="variable">footprint</span>
 <span class="variable">Usage</span>   <span class="operator">:</span> <span class="variable">@alns</span> <span class="operator">=</span> <span class="variable">$factory</span><span class="operator">-&gt;</span><span class="variable">footprint</span><span class="operator">(</span><span class="variable">$treefilename</span><span class="operator">,</span> <span class="variable">$window_size</span><span class="operator">,</span> <span class="variable">$diff</span><span class="operator">);</span>
           <span class="variable">@alns</span> <span class="operator">=</span> <span class="variable">$factory</span><span class="operator">-&gt;</span><span class="variable">footprint</span><span class="operator">(</span><span class="variable">$seqs_array_ref</span><span class="operator">);</span>
 <span class="variable">Function</span><span class="operator">:</span> <span class="variable">Aligns</span> <span class="variable">all</span> <span class="variable">the</span> <span class="variable">supplied</span> <span class="variable">sequences</span> <span class="keyword">and</span> <span class="variable">slices</span> <span class="variable">out</span> <span class="variable">of</span> <span class="variable">the</span> <span class="variable">alignment</span>
           <span class="variable">those</span> <span class="variable">regions</span> <span class="variable">along</span> <span class="variable">a</span> <span class="variable">sliding</span> <span class="variable">window</span> <span class="variable">who's</span> <span class="variable">tree</span> <span class="keyword">length</span> <span class="variable">differs</span>
           <span class="variable">significantly</span> <span class="variable">from</span> <span class="variable">the</span> <span class="variable">total</span> <span class="variable">average</span> <span class="variable">tree</span> <span class="keyword">length</span><span class="operator">.</span>
 <span class="variable">Returns</span> <span class="operator">:</span> <span class="variable">list</span> <span class="variable">of</span> <span class="variable">Bio::SimpleAlign</span> <span class="variable">objects</span>
 <span class="variable">Args</span>    <span class="operator">:</span> <span class="variable">first</span> <span class="variable">argument</span> <span class="variable">as</span> <span class="variable">per</span> <span class="variable">run</span><span class="operator">(),</span> <span class="variable">optional</span> <span class="variable">second</span> <span class="variable">argument</span> <span class="variable">to</span> <span class="variable">specify</span>
           <span class="variable">the</span> <span class="variable">size</span> <span class="variable">of</span> <span class="variable">the</span> <span class="variable">sliding</span> <span class="variable">window</span> <span class="operator">(</span><span class="keyword">default</span> <span class="number">5</span> <span class="variable">bp</span><span class="operator">)</span> <span class="keyword">and</span> <span class="variable">optional</span> <span class="variable">third</span>
           <span class="variable">argument</span> <span class="variable">to</span> <span class="variable">specify</span> <span class="variable">the</span> <span class="operator">% </span><span class="variable">difference</span> <span class="variable">from</span> <span class="variable">the</span> <span class="variable">total</span> <span class="variable">tree</span> <span class="keyword">length</span>
           <span class="variable">needed</span> <span class="keyword">for</span> <span class="variable">a</span> <span class="variable">window</span> <span class="variable">to</span> <span class="variable">be</span> <span class="variable">considered</span> <span class="variable">a</span> <span class="variable">footprint</span> <span class="operator">(</span><span class="keyword">default</span> <span class="number">33</span><span class="variable">%)</span><span class="operator">.</span>
</code></code></pre>

<h2 id="run1">_run</h2>

<pre><code><code> <span class="variable">Title</span>   <span class="operator">:</span> <span class="variable">_run</span>
 <span class="variable">Usage</span>   <span class="operator">:</span> <span class="variable">Internal</span> <span class="variable">function</span><span class="operator">,</span> <span class="keyword">not</span> <span class="variable">to</span> <span class="variable">be</span> <span class="variable">called</span> <span class="variable">directly</span> 
 <span class="variable">Function</span><span class="operator">:</span> <span class="variable">makes</span> <span class="variable">actual</span> <span class="keyword">system</span> <span class="variable">call</span> <span class="variable">to</span> <span class="variable">clustalw</span> <span class="variable">program</span>
 <span class="variable">Returns</span> <span class="operator">:</span> <span class="variable">nothing</span><span class="operator">;</span> <span class="variable">clustalw</span> <span class="variable">output</span> <span class="variable">is</span> <span class="variable">written</span> <span class="variable">to</span> <span class="variable">a</span>
           <span class="variable">temporary</span> <span class="variable">file</span> 
 <span class="variable">Args</span>    <span class="operator">:</span> <span class="variable">Name</span> <span class="variable">of</span> <span class="variable">a</span> <span class="variable">file</span> <span class="variable">containing</span> <span class="variable">a</span> <span class="variable">set</span> <span class="variable">of</span> <span class="variable">unaligned</span> <span class="variable">fasta</span> <span class="variable">sequences</span>
           <span class="keyword">and</span> <span class="variable">hash</span> <span class="variable">of</span> <span class="variable">parameters</span> <span class="variable">to</span> <span class="variable">be</span> <span class="variable">passed</span> <span class="variable">to</span> <span class="variable">clustalw</span>
</code></code></pre>

<h2 id="setinput-">_setinput()</h2>

<pre><code><code> Title   : _setinput
 Usage   : Internal function, not to be called directly 
 Function: Create input file for clustalw program
 Returns : name of file containing clustalw data input
 Args    : Seq or Align object reference or input file name</code></code></pre>

<h2 id="setparams-">_setparams()</h2>

<pre><code><code> Title   : _setparams
 Usage   : Internal function, not to be called directly 
 Function: Create parameter inputs for clustalw program
 Returns : parameter string to be passed to clustalw
           during align or profile_align
 Args    : name of calling object</code></code></pre>


</body>

</html>


