<?xml version="1.0" ?>
<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.0 Transitional//EN">
<html>
<head>
<!-- saved from url=(0017)http://localhost/ -->
<script language="JavaScript" src="../../../../../../displayToc.js"></script>
<script language="JavaScript" src="../../../../../../tocParas.js"></script>
<script language="JavaScript" src="../../../../../../tocTab.js"></script>
<link rel="stylesheet" type="text/css" href="../../../../../../scineplex.css">
<title></title>
<link rel="stylesheet" href="../../../../../../Active.css" type="text/css" />
<meta http-equiv="content-type" content="text/html; charset=utf-8" />
<link rev="made" href="mailto:support@ActiveState.com" />
</head>

<body>



<ul id="index">
  <li><a href="#NAME">NAME</a></li>
  <li><a href="#SYNOPSIS">SYNOPSIS</a></li>
  <li><a href="#DESCRIPTION">DESCRIPTION</a></li>
  <li><a href="#PROGRAM-VERSIONS">PROGRAM VERSIONS</a></li>
  <li><a href="#FEEDBACK">FEEDBACK</a>
    <ul>
      <li><a href="#Mailing-Lists">Mailing Lists</a></li>
      <li><a href="#Support">Support</a></li>
      <li><a href="#Reporting-Bugs">Reporting Bugs</a></li>
    </ul>
  </li>
  <li><a href="#AUTHOR---Shawn-Hoon">AUTHOR - Shawn Hoon</a></li>
  <li><a href="#APPENDIX">APPENDIX</a>
    <ul>
      <li><a href="#program_name">program_name</a></li>
      <li><a href="#program_dir">program_dir</a></li>
      <li><a href="#new">new</a></li>
      <li><a href="#version">version</a></li>
      <li><a href="#run">run</a></li>
      <li><a href="#input">_input</a></li>
      <li><a href="#run1">_run</a></li>
      <li><a href="#writeInput">_writeInput</a></li>
      <li><a href="#setparams">_setparams</a></li>
    </ul>
  </li>
</ul>

<h1 id="NAME">NAME</h1>

<p>Bio::Tools::Run::Alignment::Exonerate</p>

<h1 id="SYNOPSIS">SYNOPSIS</h1>

<pre><code><code>  <span class="keyword">use</span> <span class="variable">Bio::Tools::Run::Alignment::Exonerate</span><span class="operator">;</span>
  <span class="keyword">use</span> <span class="variable">Bio::SeqIO</span><span class="operator">;</span>
  
  <span class="keyword">my</span> <span class="variable">$qio</span> <span class="operator">=</span> <span class="variable">Bio::SeqIO</span><span class="operator">-&gt;</span><span class="variable">new</span><span class="operator">(</span><span class="string">-file</span><span class="operator">=&gt;</span><span class="variable">$ARGV</span><span class="operator">[</span><span class="number">0</span><span class="operator">]</span><span class="operator">,</span><span class="string">-format</span><span class="operator">=&gt;</span><span class="string">'fasta'</span><span class="operator">);</span>
  <span class="keyword">my</span> <span class="variable">$query</span> <span class="operator">=</span> <span class="variable">$qio</span><span class="operator">-&gt;</span><span class="variable">next_seq</span><span class="operator">();</span>
  <span class="keyword">my</span> <span class="variable">$tio</span> <span class="operator">=</span> <span class="variable">Bio::SeqIO</span><span class="operator">-&gt;</span><span class="variable">new</span><span class="operator">(</span><span class="string">-file</span><span class="operator">=&gt;</span><span class="variable">$ARGV</span><span class="operator">[</span><span class="number">1</span><span class="operator">]</span><span class="operator">,</span><span class="string">-format</span><span class="operator">=&gt;</span><span class="string">'fasta'</span><span class="operator">);</span>
  <span class="keyword">my</span> <span class="variable">$target</span> <span class="operator">=</span> <span class="variable">$sio</span><span class="operator">-&gt;</span><span class="variable">next_seq</span><span class="operator">();</span>
  
  <span class="comment">#exonerate parameters can all be passed via arguments parameter.</span>
  <span class="comment">#parameters passed are not checked for validity</span>
  
  <span class="keyword">my</span> <span class="variable">$run</span> <span class="operator">=</span> <span class="variable">Bio::Tools::Run::Alignment::Exonerate</span><span class="operator">-&gt;</span>
      <span class="variable">new</span><span class="operator">(</span><span class="string">arguments</span><span class="operator">=&gt;</span><span class="string">'--model est2genome --bestn 10'</span><span class="operator">);</span>
  <span class="keyword">my</span> <span class="variable">$searchio_obj</span> <span class="operator">=</span> <span class="variable">$run</span><span class="operator">-&gt;</span><span class="variable">run</span><span class="operator">(</span><span class="variable">$query</span><span class="operator">,</span><span class="variable">$target</span><span class="operator">);</span>
  
  <span class="keyword">while</span><span class="operator">(</span><span class="keyword">my</span> <span class="variable">$result</span> <span class="operator">=</span> <span class="variable">$searchio</span><span class="operator">-&gt;</span><span class="variable">next_result</span><span class="operator">){</span>
    <span class="keyword">while</span><span class="operator">(</span> <span class="keyword">my</span> <span class="variable">$hit</span> <span class="operator">=</span> <span class="variable">$result</span><span class="operator">-&gt;</span><span class="variable">next_hit</span> <span class="operator">)</span> <span class="operator">{</span>
      <span class="keyword">while</span><span class="operator">(</span> <span class="keyword">my</span> <span class="variable">$hsp</span> <span class="operator">=</span> <span class="variable">$hit</span><span class="operator">-&gt;</span><span class="variable">next_hsp</span> <span class="operator">)</span> <span class="operator">{</span>
        <span class="keyword">print</span> <span class="variable">$hsp</span><span class="operator">-&gt;</span><span class="variable">start</span><span class="operator">.</span><span class="string">"\t"</span><span class="operator">.</span><span class="variable">$hsp</span><span class="operator">-&gt;</span><span class="variable">end</span><span class="operator">.</span><span class="string">"\n"</span><span class="operator">;</span>
      <span class="operator">}</span>
    <span class="operator">}</span>
  <span class="operator">}</span>
</code></code></pre>

<h1 id="DESCRIPTION">DESCRIPTION</h1>

<p>Wrapper for Exonerate alignment program. You can get exonerate at http://www.ebi.ac.uk/~guy/exonerate/. This wrapper is written without parameter checking. All parameters are passed via the arugment parameter that is passed in the constructor. See SYNOPSIS. For exonerate parameters, run exonerate --help for more details.</p>

<h1 id="PROGRAM-VERSIONS">PROGRAM VERSIONS</h1>

<p>The tests have been shown to pass with exonorate versions 2.0 - 2.2.</p>

<h1 id="FEEDBACK">FEEDBACK</h1>

<h2 id="Mailing-Lists">Mailing Lists</h2>

<p>User feedback is an integral part of the evolution of this and other Bioperl modules. Send your comments and suggestions preferably to one of the Bioperl mailing lists. Your participation is much appreciated.</p>

<pre><code><code>  bioperl-l@bioperl.org                  - General discussion
  http://bioperl.org/wiki/Mailing_lists  - About the mailing lists</code></code></pre>

<h2 id="Support">Support</h2>

<p>Please direct usage questions or support issues to the mailing list:</p>

<p><i>bioperl-l@bioperl.org</i></p>

<p>rather than to the module maintainer directly. Many experienced and reponsive experts will be able look at the problem and quickly address it. Please include a thorough description of the problem with code and data examples if at all possible.</p>

<h2 id="Reporting-Bugs">Reporting Bugs</h2>

<p>Report bugs to the Bioperl bug tracking system to help us keep track the bugs and their resolution. Bug reports can be submitted via the web:</p>

<pre><code><code>  http://redmine.open-bio.org/projects/bioperl/</code></code></pre>

<h1 id="AUTHOR---Shawn-Hoon">AUTHOR - Shawn Hoon</h1>

<pre><code><code>  Email shawnh-at-stanford.edu</code></code></pre>

<h1 id="APPENDIX">APPENDIX</h1>

<p>The rest of the documentation details each of the object methods. Internal methods are usually preceded with a _</p>

<h2 id="program_name">program_name</h2>

<pre><code><code> Title   : program_name
 Usage   : $factory&gt;program_name()
 Function: holds the program name
 Returns:  string
 Args    : None</code></code></pre>

<h2 id="program_dir">program_dir</h2>

<pre><code><code> Title   : program_dir
 Usage   : $factory-&gt;program_dir(@params)
 Function: returns the program directory, obtained from ENV variable.
 Returns:  string
 Args    :</code></code></pre>

<h2 id="new">new</h2>

<pre><code><code> <span class="variable">Title</span>   <span class="operator">:</span> <span class="variable">new</span>
 <span class="variable">Usage</span>   <span class="operator">:</span> <span class="keyword">my</span> <span class="variable">$factory</span><span class="operator">=</span> <span class="variable">Bio::Tools::Run::Phrap</span><span class="operator">-&gt;</span><span class="variable">new</span><span class="operator">();</span>
 <span class="variable">Function</span><span class="operator">:</span> <span class="variable">creates</span> <span class="variable">a</span> <span class="variable">new</span> <span class="variable">Phrap</span> <span class="variable">factory</span>
 <span class="variable">Returns</span><span class="operator">:</span>  <span class="variable">Bio::Tools::Run::Phrap</span>
 <span class="variable">Args</span>    <span class="operator">:</span>
</code></code></pre>

<h2 id="version">version</h2>

<pre><code><code> Title   : version
 Usage   : exit if $prog-&gt;version() &lt; 1.8
 Function: Determine the version number of the program
 Example :
 Returns : float or undef
 Args    : none</code></code></pre>

<h2 id="run">run</h2>

<pre><code><code> Title   :   run()
 Usage   :   my $feats = $factory-&gt;run($seq)
 Function:   Runs Phrap 
 Returns :   An array of Bio::SeqFeature::Generic objects
 Args    :   A Bio::PrimarySeqI</code></code></pre>

<h2 id="input">_input</h2>

<pre><code><code> Title   :   _input
 Usage   :   $factory-&gt;_input($seqFile)
 Function:   get/set for input file
 Returns :
 Args    :</code></code></pre>

<h2 id="run1">_run</h2>

<pre><code><code> Title   :   _run
 Usage   :   $factory-&gt;_run()
 Function:   Makes a system call and runs Phrap
 Returns :   An array of Bio::SeqFeature::Generic objects
 Args    :</code></code></pre>

<h2 id="writeInput">_writeInput</h2>

<pre><code><code> Title   :   _writeInput
 Usage   :   $factory-&gt;_writeInput($query,$target)
 Function:   Creates a file from the given seq object
 Returns :   A string(filename)
 Args    :   Bio::PrimarySeqI</code></code></pre>

<h2 id="setparams">_setparams</h2>

<pre><code><code> Title   :  _setparams
 Usage   :  Internal function, not to be called directly
 Function:  creates a string of params to be used in the command string
 Example :
 Returns :  string of params
 Args    :  </code></code></pre>


</body>

</html>


