<?xml version="1.0" ?>
<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.0 Transitional//EN">
<html>
<head>
<!-- saved from url=(0017)http://localhost/ -->
<script language="JavaScript" src="../../../../../../displayToc.js"></script>
<script language="JavaScript" src="../../../../../../tocParas.js"></script>
<script language="JavaScript" src="../../../../../../tocTab.js"></script>
<link rel="stylesheet" type="text/css" href="../../../../../../scineplex.css">
<title></title>
<link rel="stylesheet" href="../../../../../../Active.css" type="text/css" />
<meta http-equiv="content-type" content="text/html; charset=utf-8" />
<link rev="made" href="mailto:support@ActiveState.com" />
</head>

<body>



<ul id="index">
  <li><a href="#NAME">NAME</a></li>
  <li><a href="#SYNOPSIS">SYNOPSIS</a></li>
  <li><a href="#DESCRIPTION">DESCRIPTION</a>
    <ul>
      <li><a href="#Helping-the-module-find-your-executable">Helping the module find your executable</a></li>
    </ul>
  </li>
  <li><a href="#PARAMETERS-FOR-ALIGNMENT-COMPUTATION">PARAMETERS FOR ALIGNMENT COMPUTATION</a>
    <ul>
      <li><a href="#IN">IN</a></li>
      <li><a href="#TYPE">TYPE</a></li>
      <li><a href="#PARAMETERS">PARAMETERS</a></li>
      <li><a href="#EXTEND">EXTEND</a></li>
      <li><a href="#DP_NORMALISE">DP_NORMALISE</a></li>
      <li><a href="#DP_MODE">DP_MODE</a></li>
      <li><a href="#KTUPLE">KTUPLE</a></li>
      <li><a href="#NDIAGS">NDIAGS</a></li>
      <li><a href="#DIAG_MODE">DIAG_MODE</a></li>
      <li><a href="#SIM_MATRIX">SIM_MATRIX</a></li>
      <li><a href="#MATRIX">MATRIX</a></li>
      <li><a href="#GAPOPEN">GAPOPEN</a></li>
      <li><a href="#GAPEXT">GAPEXT</a></li>
      <li><a href="#COSMETIC_PENALTY">COSMETIC_PENALTY</a></li>
      <li><a href="#TG_MODE">TG_MODE</a></li>
      <li><a href="#WEIGHT">WEIGHT</a></li>
      <li><a href="#SEQ_TO_ALIGN">SEQ_TO_ALIGN</a></li>
    </ul>
  </li>
  <li><a href="#PARAMETERS-FOR-TREE-COMPUTATION-AND-OUTPUT">PARAMETERS FOR TREE COMPUTATION AND OUTPUT</a>
    <ul>
      <li><a href="#NEWTREE">NEWTREE</a></li>
      <li><a href="#USETREE">USETREE</a></li>
      <li><a href="#TREE_MODE">TREE_MODE</a></li>
      <li><a href="#QUICKTREE">QUICKTREE</a></li>
    </ul>
  </li>
  <li><a href="#PARAMETERS-FOR-ALIGNMENT-OUTPUT">PARAMETERS FOR ALIGNMENT OUTPUT</a>
    <ul>
      <li><a href="#OUTFILE">OUTFILE</a></li>
      <li><a href="#OUTPUT">OUTPUT</a></li>
      <li><a href="#CASE">CASE</a></li>
      <li><a href="#CPU">CPU</a></li>
      <li><a href="#OUT_LIB">OUT_LIB</a></li>
      <li><a href="#OUTORDER">OUTORDER</a></li>
      <li><a href="#SEQNOS">SEQNOS</a></li>
    </ul>
  </li>
  <li><a href="#PARAMETERS-FOR-GENERIC-OUTPUT">PARAMETERS FOR GENERIC OUTPUT</a>
    <ul>
      <li><a href="#RUN_NAME">RUN_NAME</a></li>
      <li><a href="#ALIGN">ALIGN</a></li>
      <li><a href="#QUIET">QUIET</a></li>
      <li><a href="#CONVERT">CONVERT</a></li>
    </ul>
  </li>
  <li><a href="#FEEDBACK">FEEDBACK</a>
    <ul>
      <li><a href="#Mailing-Lists">Mailing Lists</a></li>
      <li><a href="#Support">Support</a></li>
      <li><a href="#Reporting-Bugs">Reporting Bugs</a></li>
    </ul>
  </li>
  <li><a href="#AUTHOR---Jason-Stajich-Peter-Schattner">AUTHOR - Jason Stajich, Peter Schattner</a></li>
  <li><a href="#APPENDIX">APPENDIX</a>
    <ul>
      <li><a href="#program_name">program_name</a></li>
      <li><a href="#program_dir">program_dir</a></li>
      <li><a href="#error_string">error_string</a></li>
      <li><a href="#version">version</a></li>
      <li><a href="#run">run</a></li>
      <li><a href="#align">align</a></li>
      <li><a href="#profile_align">profile_align</a></li>
      <li><a href="#run1">_run</a></li>
      <li><a href="#setinput">_setinput</a></li>
      <li><a href="#setparams">_setparams</a></li>
      <li><a href="#aformat">aformat</a></li>
      <li><a href="#methods">methods</a></li>
    </ul>
  </li>
  <li><a href="#Bio::Tools::Run::BaseWrapper-methods">Bio::Tools::Run::BaseWrapper methods</a>
    <ul>
      <li><a href="#no_param_checks">no_param_checks</a></li>
      <li><a href="#save_tempfiles">save_tempfiles</a></li>
      <li><a href="#outfile_name">outfile_name</a></li>
      <li><a href="#tempdir">tempdir</a></li>
      <li><a href="#cleanup">cleanup</a></li>
      <li><a href="#io">io</a></li>
    </ul>
  </li>
</ul>

<h1 id="NAME">NAME</h1>

<p>Bio::Tools::Run::Alignment::TCoffee - Object for the calculation of a multiple sequence alignment from a set of unaligned sequences or alignments using the TCoffee program</p>

<h1 id="SYNOPSIS">SYNOPSIS</h1>

<pre><code><code>  <span class="comment"># Build a tcoffee alignment factory</span>
  <span class="variable">@params</span> <span class="operator">=</span> <span class="operator">(</span><span class="string">'ktuple'</span> <span class="operator">=&gt;</span> <span class="number">2</span><span class="operator">,</span> <span class="string">'matrix'</span> <span class="operator">=&gt;</span> <span class="string">'BLOSUM'</span><span class="operator">);</span>
  <span class="variable">$factory</span> <span class="operator">=</span> <span class="variable">Bio::Tools::Run::Alignment::TCoffee</span><span class="operator">-&gt;</span><span class="variable">new</span><span class="operator">(</span><span class="variable">@params</span><span class="operator">);</span>
  
  <span class="comment"># Pass the factory a list of sequences to be aligned.</span>
  <span class="variable">$inputfilename</span> <span class="operator">=</span> <span class="string">'t/cysprot.fa'</span><span class="operator">;</span>
  <span class="comment"># $aln is a SimpleAlign object.</span>
  <span class="variable">$aln</span> <span class="operator">=</span> <span class="variable">$factory</span><span class="operator">-&gt;</span><span class="variable">align</span><span class="operator">(</span><span class="variable">$inputfilename</span><span class="operator">);</span>
  
  <span class="comment"># or where @seq_array is an array of Bio::Seq objects</span>
  <span class="variable">$seq_array_ref</span> <span class="operator">=</span> <span class="operator">\</span><span class="variable">@seq_array</span><span class="operator">;</span>
  <span class="variable">$aln</span> <span class="operator">=</span> <span class="variable">$factory</span><span class="operator">-&gt;</span><span class="variable">align</span><span class="operator">(</span><span class="variable">$seq_array_ref</span><span class="operator">);</span>
  
  <span class="comment"># Or one can pass the factory a pair of (sub)alignments</span>
  <span class="comment">#to be aligned against each other, e.g.:</span>
  
  <span class="comment"># where $aln1 and $aln2 are Bio::SimpleAlign objects.</span>
  <span class="variable">$aln</span> <span class="operator">=</span> <span class="variable">$factory</span><span class="operator">-&gt;</span><span class="variable">profile_align</span><span class="operator">(</span><span class="variable">$aln1</span><span class="operator">,</span><span class="variable">$aln2</span><span class="operator">);</span>
  
  <span class="comment"># Or one can pass the factory an alignment and one or more</span>
  <span class="comment"># unaligned sequences to be added to the alignment. For example:</span>
  
  <span class="comment"># $seq is a Bio::Seq object.</span>
  <span class="variable">$aln</span> <span class="operator">=</span> <span class="variable">$factory</span><span class="operator">-&gt;</span><span class="variable">profile_align</span><span class="operator">(</span><span class="variable">$aln1</span><span class="operator">,</span><span class="variable">$seq</span><span class="operator">);</span>
  
  <span class="comment">#There are various additional options and input formats available.</span>
  <span class="comment">#See the DESCRIPTION section that follows for additional details.</span>
</code></code></pre>

<h1 id="DESCRIPTION">DESCRIPTION</h1>

<p>Note: this DESCRIPTION only documents the (Bio)perl interface to TCoffee.</p>

<h2 id="Helping-the-module-find-your-executable">Helping the module find your executable</h2>

<p>You will need to enable TCoffee to find the t_coffee program. This can be done in (at least) three ways:</p>

<pre><code><code> <span class="number">1.</span> <span class="variable">Make</span> <span class="variable">sure</span> <span class="variable">the</span> <span class="variable">t_coffee</span> <span class="variable">executable</span> <span class="variable">is</span> <span class="variable">in</span> <span class="variable">your</span> <span class="variable">path</span> <span class="variable">so</span> <span class="variable">that</span>
    <span class="variable">which</span> <span class="variable">t_coffee</span> <span class="variable">returns</span> <span class="variable">a</span> <span class="variable">t_coffee</span> <span class="variable">executable</span> <span class="variable">on</span> <span class="variable">your</span> <span class="keyword">system</span><span class="operator">.</span>
 
 <span class="number">2.</span> <span class="variable">Define</span> <span class="variable">an</span> <span class="variable">environmental</span> <span class="variable">variable</span> <span class="variable">TCOFFEEDIR</span> <span class="variable">which</span> <span class="variable">is</span> <span class="variable">a</span> <span class="variable">dir</span> 
    <span class="variable">which</span> <span class="variable">contains</span> <span class="variable">the</span> <span class="string">'t_coffee'</span> <span class="variable">app</span><span class="operator">:</span>
    <span class="variable">In</span> <span class="variable">bash</span> 
    <span class="variable">export</span> <span class="variable">TCOFFEEDIR</span><span class="operator">=</span><span class="regex">/home/username</span><span class="operator">/</span><span class="variable">progs</span><span class="operator">/</span><span class="variable">T</span><span class="operator">-</span><span class="variable">COFFEE_distribution_Version_1</span><span class="number">.37</span><span class="operator">/</span><span class="variable">bin</span>
    <span class="variable">In</span> <span class="variable">csh</span><span class="operator">/</span><span class="variable">tcsh</span>
    <span class="variable">setenv</span> <span class="variable">TCOFFEEDIR</span> <span class="operator">/</span><span class="variable">home</span><span class="operator">/</span><span class="variable">username</span><span class="operator">/</span><span class="variable">progs</span><span class="operator">/</span><span class="variable">T</span><span class="operator">-</span><span class="variable">COFFEE_distribution_Version_1</span><span class="number">.37</span><span class="operator">/</span><span class="variable">bin</span>
 
 <span class="number">3.</span> <span class="variable">Include</span> <span class="variable">a</span> <span class="variable">definition</span> <span class="variable">of</span> <span class="variable">an</span> <span class="variable">environmental</span> <span class="variable">variable</span> <span class="variable">TCOFFEEDIR</span> <span class="variable">in</span>
    <span class="variable">every</span> <span class="variable">script</span> <span class="variable">that</span> <span class="variable">will</span> <span class="keyword">use</span> <span class="variable">this</span> <span class="variable">TCoffee</span> <span class="variable">wrapper</span> <span class="variable">module</span><span class="operator">.</span>
    <span class="keyword">BEGIN</span> <span class="operator">{</span> <span class="variable">$ENV</span><span class="operator">{</span><span class="string">TCOFFEDIR</span><span class="operator">}</span> <span class="operator">=</span> <span class="string">'/home/username/progs/T-COFFEE_distribution_Version_1.37/bin'</span> <span class="operator">}</span>
    <span class="keyword">use</span> <span class="variable">Bio::Tools::Run::Alignment::TCoffee</span><span class="operator">;</span>
</code></code></pre>

<p>If you are running an application on a webserver make sure the webserver environment has the proper PATH set or use the options 2 or 3 to set the variables.</p>

<h1 id="PARAMETERS-FOR-ALIGNMENT-COMPUTATION">PARAMETERS FOR ALIGNMENT COMPUTATION</h1>

<p>There are a number of possible parameters one can pass in TCoffee. One should really read the online manual for the best explanation of all the features. See http://igs-server.cnrs-mrs.fr/~cnotred/Documentation/t_coffee/t_coffee_doc.html</p>

<p>These can be specified as parameters when instantiating a new TCoffee object, or through get/set methods of the same name (lowercase).</p>

<h2 id="IN">IN</h2>

<pre><code><code> Title       : IN
 Description : (optional) input filename, this is specified when
               align so should not use this directly unless one
               understand TCoffee program very well.</code></code></pre>

<h2 id="TYPE">TYPE</h2>

<pre><code><code> Title       : TYPE
 Args        : [string] DNA, PROTEIN
 Description : (optional) set the sequence type, guessed automatically
               so should not use this directly</code></code></pre>

<h2 id="PARAMETERS">PARAMETERS</h2>

<pre><code><code> Title       : PARAMETERS
 Description : (optional) Indicates a file containing extra parameters</code></code></pre>

<h2 id="EXTEND">EXTEND</h2>

<pre><code><code> Title       : EXTEND
 Args        : 0, 1, or positive value
 Default     : 1
 Description : Flag indicating that library extension should be
               carried out when performing multiple alignments, if set
               to 0 then extension is not made, if set to 1 extension
               is made on all pairs in the library.  If extension is
               set to another positive value, the extension is only
               carried out on pairs having a weigth value superior to
               the specified limit.</code></code></pre>

<h2 id="DP_NORMALISE">DP_NORMALISE</h2>

<pre><code><code> Title       : DP_NORMALISE
 Args        : 0 or positive value
 Default     : 1000
 Description : When using a value different from 0, this flag sets the
               score of the highest scoring pair to 1000.</code></code></pre>

<h2 id="DP_MODE">DP_MODE</h2>

<pre><code><code> Title       : DP_MODE
 Args        : [string] gotoh_pair_wise, myers_miller_pair_wise,
               fasta_pair_wise cfasta_pair_wise
 Default     : cfast_fair_wise
 Description : Indicates the type of dynamic programming used by
               the program

    gotoh_pair_wise : implementation of the gotoh algorithm
    (quadratic in memory and time)

    myers_miller_pair_wise : implementation of the Myers and Miller
    dynamic programming algorithm ( quadratic in time and linear in
    space). This algorithm is recommended for very long sequences. It
    is about 2 time slower than gotoh. It only accepts tg_mode=1.

    fasta_pair_wise: implementation of the fasta algorithm. The
    sequence is hashed, looking for ktuples words. Dynamic programming
    is only carried out on the ndiag best scoring diagonals. This is
    much faster but less accurate than the two previous.

    cfasta_pair_wise : c stands for checked. It is the same
    algorithm. The dynamic programming is made on the ndiag best
    diagonals, and then on the 2*ndiags, and so on until the scores
    converge. Complexity will depend on the level of divergence of the
    sequences, but will usually be L*log(L), with an accuracy
    comparable to the two first mode ( this was checked on BaliBase).</code></code></pre>

<h2 id="KTUPLE">KTUPLE</h2>

<pre><code><code> Title       : KTUPLE
 Args        : numeric value
 Default     : 1 or 2 (1 for protein, 2 for DNA )

 Description : Indicates the ktuple size for cfasta_pair_wise dp_mode
               and fasta_pair_wise. It is set to 1 for proteins, and 2
               for DNA. The alphabet used for protein is not the 20
               letter code, but a mildly degenerated version, where
               some residues are grouped under one letter, based on
               physicochemical properties:
               rk, de, qh, vilm, fy (the other residues are
               not degenerated).</code></code></pre>

<h2 id="NDIAGS">NDIAGS</h2>

<pre><code><code> Title       : NDIAGS
 Args        : numeric value
 Default     : 0
 Description : Indicates the number of diagonals used by the
               fasta_pair_wise algorithm. When set to 0,
               n_diag=Log (length of the smallest sequence)</code></code></pre>

<h2 id="DIAG_MODE">DIAG_MODE</h2>

<pre><code><code> Title       : DIAG_MODE
 Args        : numeric value
 Default     : 0


 Description : Indicates the manner in which diagonals are scored
              during the fasta hashing.

              0 indicates that the score of a diagonal is equal to the
              sum of the scores of the exact matches it contains.


              1 indicates that this score is set equal to the score of
              the best uninterrupted segment

              1 can be useful when dealing with fragments of sequences.</code></code></pre>

<h2 id="SIM_MATRIX">SIM_MATRIX</h2>

<pre><code><code> Title       : SIM_MATRIX
 Args        : string
 Default     : vasiliky
 Description : Indicates the manner in which the amino acid is being
               degenerated when hashing. All the substitution matrix
               are acceptable. Categories will be defined as sub-group
               of residues all having a positive substitution score
               (they can overlap).

               If you wish to keep the non degenerated amino acid
               alphabet, use &#39;idmat&#39;</code></code></pre>

<h2 id="MATRIX">MATRIX</h2>

<pre><code><code> Title       : MATRIX
 Args        :
 Default     :
 Description : This flag is provided for compatibility with
               ClustalW. Setting matrix = &#39;blosum&#39; is equivalent to
               -in=Xblosum62mt , -matrix=pam is equivalent to
               in=Xpam250mt . Apart from this, the rules are similar
               to those applying when declaring a matrix with the
               -in=X fl</code></code></pre>

<h2 id="GAPOPEN">GAPOPEN</h2>

<pre><code><code> Title       : GAPOPEN
 Args        : numeric
 Default     : 0
 Description : Indicates the penalty applied for opening a gap. The
               penalty must be negative. If you provide a positive
               value, it will automatically be turned into a negative
               number. We recommend a value of 10 with pam matrices,
               and a value of 0 when a library is used.</code></code></pre>

<h2 id="GAPEXT">GAPEXT</h2>

<pre><code><code> Title       : GAPEXT
 Args        : numeric
 Default     : 0
 Description : Indicates the penalty applied for extending a gap.</code></code></pre>

<h2 id="COSMETIC_PENALTY">COSMETIC_PENALTY</h2>

<pre><code><code> Title       : COSMETIC_PENALTY
 Args        : numeric
 Default     : 100
 Description : Indicates the penalty applied for opening a gap. This
               penalty is set to a very low value. It will only have
               an influence on the portions of the alignment that are
               unalignable. It will not make them more correct, but
               only more pleasing to the eye ( i.e. Avoid stretches of
               lonely residues).

               The cosmetic penalty is automatically turned off if a
               substitution matrix is used rather than a library.</code></code></pre>

<h2 id="TG_MODE">TG_MODE</h2>

<pre><code><code> Title       : TG_MODE
 Args        : 0,1,2
 Default     : 1
 Description : (Terminal Gaps)
               0: indicates that terminal gaps must be panelized with
                  a gapopen and a gapext penalty.
               1: indicates that terminal gaps must be penalized only
                  with a gapext penalty
               2: indicates that terminal gaps must not be penalized.</code></code></pre>

<h2 id="WEIGHT">WEIGHT</h2>

<pre><code><code> Title       : WEIGHT
 Args        : sim or sim_&lt;matrix_name or matrix_file&gt; or integer value
 Default     : sim


 Description : Weight defines the way alignments are weighted when
               turned into a library.

               sim indicates that the weight equals the average
                   identity within the match residues.

               sim_matrix_name indicates the average identity with two
                   residues regarded as identical when their
                   substitution value is positive. The valid matrices
                   names are in matrices.h (pam250mt) . Matrices not
                   found in this header are considered to be
                   filenames. See the format section for matrices. For
                   instance, -weight=sim_pam250mt indicates that the
                   grouping used for similarity will be the set of
                   classes with positive substitutions. Other groups
                   include

                       sim_clustalw_col ( categories of clustalw
                       marked with :)

                       sim_clustalw_dot ( categories of clustalw
                       marked with .)


               Value indicates that all the pairs found in the
               alignments must be given the same weight equal to
               value. This is useful when the alignment one wishes to
               turn into a library must be given a pre-specified score
               (for instance if they come from a structure
               super-imposition program). Value is an integer:

                       -weight=1000

  Note       : Weight only affects methods that return an alignment to
               T-Coffee, such as ClustalW. On the contrary, the
               version of Lalign we use here returns a library where
               weights have already been applied and are therefore
               insensitive to the -weight flag.</code></code></pre>

<h2 id="SEQ_TO_ALIGN">SEQ_TO_ALIGN</h2>

<pre><code><code> Title       : SEQ_TO_ALIGN
 Args        : filename
 Default     : no file - align all the sequences

 Description : You may not wish to align all the sequences brought in
               by the -in flag. Supplying the seq_to_align flag allows
               for this, the file is simply a list of names in Fasta
               format.

               However, note that library extension will be carried out
               on all the sequences.</code></code></pre>

<h1 id="PARAMETERS-FOR-TREE-COMPUTATION-AND-OUTPUT">PARAMETERS FOR TREE COMPUTATION AND OUTPUT</h1>

<h2 id="NEWTREE">NEWTREE</h2>

<pre><code><code> Title       : NEWTREE
 Args        : treefile
 Default     : no file
 Description : Indicates the name of the new tree to compute. The
               default will be &lt;sequence_name&gt;.dnd, or &lt;run_name.dnd&gt;.
               Format is Phylip/Newick tree format</code></code></pre>

<h2 id="USETREE">USETREE</h2>

<pre><code><code> Title       : USETREE
 Args        : treefile
 Default     : no file specified
 Description : This flag indicates that rather than computing a new
               dendrogram, t_coffee can use a pre-computed one. The
               tree files are in phylips format and compatible with
               ClustalW. In most cases, using a pre-computed tree will
               halve the computation time required by t_coffee. It is
               also possible to use trees output by ClustalW or
               Phylips. Format is Phylips tree format</code></code></pre>

<h2 id="TREE_MODE">TREE_MODE</h2>

<pre><code><code> Title       : TREE_MODE
 Args        : slow, fast, very_fast
 Default     : very_fast
 Description : This flag indicates the method used for computing the
               dendrogram.
               slow : the chosen dp_mode using the extended library,
               fast : The fasta dp_mode using the extended library.
               very_fast: The fasta dp_mode using pam250mt.</code></code></pre>

<h2 id="QUICKTREE">QUICKTREE</h2>

<pre><code><code> Title       : QUICKTREE
 Args        :
 Default     :
 Description : This flag is kept for compatibility with ClustalW.
               It indicates that:  -tree_mode=very_fast</code></code></pre>

<h1 id="PARAMETERS-FOR-ALIGNMENT-OUTPUT">PARAMETERS FOR ALIGNMENT OUTPUT</h1>

<h2 id="OUTFILE">OUTFILE</h2>

<pre><code><code> Title       : OUTFILE
 Args        : out_aln file, default, no
 Default     : default ( yourseqfile.aln)
 Description : indicates name of output alignment file</code></code></pre>

<h2 id="OUTPUT">OUTPUT</h2>

<pre><code><code> Title       : OUTPUT
 Args        : format1, format2
 Default     : clustalw
 Description : Indicated format for outputting outputfile
               Supported formats are:

               clustalw_aln, clustalw: ClustalW format.
               gcg, msf_aln : Msf alignment.
               pir_aln : pir alignment.
               fasta_aln : fasta alignment.
               phylip : Phylip format.
               pir_seq : pir sequences (no gap).
               fasta_seq : fasta sequences (no gap).
    As well as:
                score_html : causes the output to be a reliability
                             plot in HTML
                score_pdf : idem in PDF.
                score_ps : idem in postscript.

    More than one format can be indicated:
                -output=clustalw,gcg, score_html</code></code></pre>

<h2 id="CASE">CASE</h2>

<pre><code><code> Title       : CASE
 Args        : upper, lower
 Default     : upper
 Description : triggers choice of the case for output</code></code></pre>

<h2 id="CPU">CPU</h2>

<pre><code><code> Title       : CPU
 Args        : value
 Default     : 0
 Description : Indicates the cpu time (micro seconds) that must be
               added to the t_coffee computation time.</code></code></pre>

<h2 id="OUT_LIB">OUT_LIB</h2>

<pre><code><code> Title       : OUT_LIB
 Args        : name of library, default, no
 Default     : default
 Description : Sets the name of the library output. Default implies
               &lt;run_name&gt;.tc_lib</code></code></pre>

<h2 id="OUTORDER">OUTORDER</h2>

<pre><code><code> Title       : OUTORDER
 Args        : input or aligned
 Default     : input
 Description : Sets the name of the library output. Default implies
               &lt;run_name&gt;.tc_lib</code></code></pre>

<h2 id="SEQNOS">SEQNOS</h2>

<pre><code><code> Title       : SEQNOS
 Args        : on or off
 Default     : off
 Description : Causes the output alignment to contain residue numbers
               at the end of each line:</code></code></pre>

<h1 id="PARAMETERS-FOR-GENERIC-OUTPUT">PARAMETERS FOR GENERIC OUTPUT</h1>

<h2 id="RUN_NAME">RUN_NAME</h2>

<pre><code><code> Title       : RUN_NAME
 Args        : your run name
 Default     :
 Description : This flag causes the prefix &lt;your sequences&gt; to be
               replaced by &lt;your run name&gt; when renaming the default
               files.</code></code></pre>

<h2 id="ALIGN">ALIGN</h2>

<pre><code><code> Title       : ALIGN
 Args        :
 Default     :
 Description : Indicates that the program must produce the
               alignment. This flag is here for compatibility with
               ClustalW</code></code></pre>

<h2 id="QUIET">QUIET</h2>

<pre><code><code> Title       : QUIET
 Args        : stderr, stdout, or filename, or nothing
 Default     : stderr
 Description : Redirects the standard output to either a file.
              -quiet on its own redirect the output to /dev/null.</code></code></pre>

<h2 id="CONVERT">CONVERT</h2>

<pre><code><code> Title       : CONVERT
 Args        :
 Default     :
 Description : Indicates that the program must not compute the
               alignment but simply convert all the sequences,
               alignments and libraries into the format indicated with
               -output. This flag can also be used if you simply want
               to compute a library ( i.e. You have an alignment and
               you want to turn it into a library).</code></code></pre>

<h1 id="FEEDBACK">FEEDBACK</h1>

<h2 id="Mailing-Lists">Mailing Lists</h2>

<p>User feedback is an integral part of the evolution of this and other Bioperl modules. Send your comments and suggestions preferably to one of the Bioperl mailing lists. Your participation is much appreciated.</p>

<pre><code><code>  bioperl-l@bioperl.org                  - General discussion
  http://bioperl.org/wiki/Mailing_lists  - About the mailing lists</code></code></pre>

<h2 id="Support">Support</h2>

<p>Please direct usage questions or support issues to the mailing list:</p>

<p><i>bioperl-l@bioperl.org</i></p>

<p>rather than to the module maintainer directly. Many experienced and reponsive experts will be able look at the problem and quickly address it. Please include a thorough description of the problem with code and data examples if at all possible.</p>

<h2 id="Reporting-Bugs">Reporting Bugs</h2>

<p>Report bugs to the Bioperl bug tracking system to help us keep track the bugs and their resolution. Bug reports can be submitted via the web:</p>

<pre><code><code> http://redmine.open-bio.org/projects/bioperl/</code></code></pre>

<h1 id="AUTHOR---Jason-Stajich-Peter-Schattner">AUTHOR - Jason Stajich, Peter Schattner</h1>

<p>Email jason-at-bioperl-dot-org, schattner@alum.mit.edu</p>

<h1 id="APPENDIX">APPENDIX</h1>

<p>The rest of the documentation details each of the object methods. Internal methods are usually preceded with a _</p>

<h2 id="program_name">program_name</h2>

<pre><code><code> Title   : program_name
 Usage   : $factory-&gt;program_name()
 Function: holds the program name
 Returns:  string
 Args    : None</code></code></pre>

<h2 id="program_dir">program_dir</h2>

<pre><code><code> Title   : program_dir
 Usage   : $factory-&gt;program_dir(@params)
 Function: returns the program directory, obtained from ENV variable.
 Returns:  string
 Args    :</code></code></pre>

<h2 id="error_string">error_string</h2>

<pre><code><code> Title   : error_string
 Usage   : $obj-&gt;error_string($newval)
 Function: Where the output from the last analysus run is stored.
 Returns : value of error_string
 Args    : newvalue (optional)</code></code></pre>

<h2 id="version">version</h2>

<pre><code><code> Title   : version
 Usage   : exit if $prog-&gt;version() &lt; 1.8
 Function: Determine the version number of the program
 Example :
 Returns : float or undef
 Args    : none</code></code></pre>

<h2 id="run">run</h2>

<pre><code><code> <span class="variable">Title</span>   <span class="operator">:</span> <span class="variable">run</span>
 <span class="variable">Usage</span>   <span class="operator">:</span> <span class="keyword">my</span> <span class="variable">$output</span> <span class="operator">=</span> <span class="variable">$application</span><span class="operator">-&gt;</span><span class="variable">run</span><span class="operator">(</span><span class="string">-seq</span>     <span class="operator">=&gt;</span> <span class="variable">$seq</span><span class="operator">,</span>
                                          <span class="string">-profile</span> <span class="operator">=&gt;</span> <span class="variable">$profile</span><span class="operator">,</span>
                                          <span class="string">-type</span>    <span class="operator">=&gt;</span> <span class="string">'profile-aln'</span><span class="operator">);</span>
 <span class="variable">Function</span><span class="operator">:</span> <span class="variable">Generic</span> <span class="variable">run</span> <span class="variable">of</span> <span class="variable">an</span> <span class="variable">application</span>
 <span class="variable">Returns</span> <span class="operator">:</span> <span class="variable">Bio::SimpleAlign</span> <span class="variable">object</span>
 <span class="variable">Args</span>    <span class="operator">:</span> <span class="variable">key</span><span class="operator">-</span><span class="variable">value</span> <span class="variable">parameters</span> <span class="variable">allowed</span> <span class="keyword">for</span> <span class="variable">TCoffee</span> <span class="variable">runs</span> <span class="variable">AND</span>
           <span class="operator">-</span><span class="string">type</span>     <span class="operator">=&gt;</span> <span class="variable">profile</span><span class="operator">-</span><span class="variable">aln</span> <span class="keyword">or</span> <span class="variable">alignment</span> <span class="keyword">for</span> <span class="variable">profile</span> <span class="variable">alignments</span> <span class="keyword">or</span>
                        <span class="variable">just</span> <span class="variable">multiple</span> <span class="variable">sequence</span> <span class="variable">alignment</span>
           <span class="operator">-</span><span class="string">seq</span>      <span class="operator">=&gt;</span> <span class="variable">either</span> <span class="variable">Bio::PrimarySeqI</span> <span class="variable">object</span> <span class="variable">OR</span>
                        <span class="variable">array</span> <span class="keyword">ref</span> <span class="variable">of</span> <span class="variable">Bio::PrimarySeqI</span> <span class="variable">objects</span> <span class="variable">OR</span>
                        <span class="variable">filename</span> <span class="variable">of</span> <span class="variable">sequences</span> <span class="variable">to</span> <span class="variable">run</span> <span class="variable">with</span>
           <span class="operator">-</span><span class="string">profile</span>  <span class="operator">=&gt;</span> <span class="variable">profile</span> <span class="variable">to</span> <span class="variable">align</span> <span class="variable">to</span><span class="operator">,</span> <span class="keyword">if</span> <span class="variable">this</span> <span class="variable">is</span> <span class="variable">an</span> <span class="variable">array</span> <span class="keyword">ref</span>
                        <span class="variable">will</span> <span class="variable">specify</span> <span class="variable">the</span> <span class="variable">first</span> <span class="variable">two</span> <span class="variable">entries</span> <span class="variable">as</span> <span class="variable">the</span> <span class="variable">two</span>
                        <span class="variable">profiles</span> <span class="variable">to</span> <span class="variable">align</span> <span class="variable">to</span> <span class="keyword">each</span> <span class="variable">other</span>
</code></code></pre>

<h2 id="align">align</h2>

<pre><code><code> <span class="variable">Title</span>   <span class="operator">:</span> <span class="variable">align</span>
 <span class="variable">Usage</span>   <span class="operator">:</span>
        <span class="variable">$inputfilename</span> <span class="operator">=</span> <span class="string">'t/data/cysprot.fa'</span><span class="operator">;</span>
        <span class="variable">$aln</span> <span class="operator">=</span> <span class="variable">$factory</span><span class="operator">-&gt;</span><span class="variable">align</span><span class="operator">(</span><span class="variable">$inputfilename</span><span class="operator">);</span>
 <span class="keyword">or</span>
        <span class="variable">$seq_array_ref</span> <span class="operator">=</span> <span class="operator">\</span><span class="variable">@seq_array</span><span class="operator">;</span> 
        <span class="comment"># @seq_array is array of Seq objs</span>
        <span class="variable">$aln</span> <span class="operator">=</span> <span class="variable">$factory</span><span class="operator">-&gt;</span><span class="variable">align</span><span class="operator">(</span><span class="variable">$seq_array_ref</span><span class="operator">);</span>
 <span class="variable">Function</span><span class="operator">:</span> <span class="variable">Perform</span> <span class="variable">a</span> <span class="variable">multiple</span> <span class="variable">sequence</span> <span class="variable">alignment</span>
 <span class="variable">Returns</span> <span class="operator">:</span> <span class="variable">Reference</span> <span class="variable">to</span> <span class="variable">a</span> <span class="variable">SimpleAlign</span> <span class="variable">object</span> <span class="variable">containing</span> <span class="variable">the</span>
           <span class="variable">sequence</span> <span class="variable">alignment</span><span class="operator">.</span>
 <span class="variable">Args</span>    <span class="operator">:</span> <span class="variable">Name</span> <span class="variable">of</span> <span class="variable">a</span> <span class="variable">file</span> <span class="variable">containing</span> <span class="variable">a</span> <span class="variable">set</span> <span class="variable">of</span> <span class="variable">unaligned</span> <span class="variable">fasta</span> <span class="variable">sequences</span>
           <span class="keyword">or</span> <span class="keyword">else</span> <span class="variable">an</span> <span class="variable">array</span> <span class="variable">of</span> <span class="variable">references</span> <span class="variable">to</span> <span class="variable">Bio::Seq</span> <span class="variable">objects</span><span class="operator">.</span>
 
 <span class="variable">Throws</span> <span class="variable">an</span> <span class="variable">exception</span> <span class="keyword">if</span> <span class="variable">argument</span> <span class="variable">is</span> <span class="keyword">not</span> <span class="variable">either</span> <span class="variable">a</span> <span class="variable">string</span> <span class="operator">(</span><span class="variable">eg</span> <span class="variable">a</span>
 <span class="variable">filename</span><span class="operator">)</span> <span class="keyword">or</span> <span class="variable">a</span> <span class="variable">reference</span> <span class="variable">to</span> <span class="variable">an</span> <span class="variable">array</span> <span class="variable">of</span> <span class="variable">Bio::Seq</span> <span class="variable">objects</span><span class="operator">.</span>  <span class="variable">If</span>
 <span class="variable">argument</span> <span class="variable">is</span> <span class="variable">string</span><span class="operator">,</span> <span class="variable">throws</span> <span class="variable">exception</span> <span class="keyword">if</span> <span class="variable">file</span> <span class="variable">corresponding</span> <span class="variable">to</span> <span class="variable">string</span>
 <span class="variable">name</span> <span class="variable">can</span> <span class="keyword">not</span> <span class="variable">be</span> <span class="variable">found</span><span class="operator">.</span> <span class="variable">If</span> <span class="variable">argument</span> <span class="variable">is</span> <span class="variable">Bio::Seq</span> <span class="variable">array</span><span class="operator">,</span> <span class="variable">throws</span>
 <span class="variable">exception</span> <span class="keyword">if</span> <span class="variable">less</span> <span class="variable">than</span> <span class="variable">two</span> <span class="variable">sequence</span> <span class="variable">objects</span> <span class="variable">are</span> <span class="variable">in</span> <span class="variable">array</span><span class="operator">.</span>
</code></code></pre>

<h2 id="profile_align">profile_align</h2>

<pre><code><code> Title   : profile_align
 Usage   :
 Function: Perform an alignment of 2 (sub)alignments
 Example :
 Returns : Reference to a SimpleAlign object containing the (super)alignment.
 Args    : Names of 2 files containing the subalignments
           or references to 2 Bio::SimpleAlign objects.
 Note    : Needs to be updated to run with newer TCoffee code, which
           allows more than two profile alignments.</code></code></pre>

<p>Throws an exception if arguments are not either strings (eg filenames) or references to SimpleAlign objects.</p>

<h2 id="run1">_run</h2>

<pre><code><code> <span class="variable">Title</span>   <span class="operator">:</span>  <span class="variable">_run</span>
 <span class="variable">Usage</span>   <span class="operator">:</span>  <span class="variable">Internal</span> <span class="variable">function</span><span class="operator">,</span> <span class="keyword">not</span> <span class="variable">to</span> <span class="variable">be</span> <span class="variable">called</span> <span class="variable">directly</span>        
 <span class="variable">Function</span><span class="operator">:</span>  <span class="variable">makes</span> <span class="variable">actual</span> <span class="keyword">system</span> <span class="variable">call</span> <span class="variable">to</span> <span class="variable">tcoffee</span> <span class="variable">program</span>
 <span class="variable">Example</span> <span class="operator">:</span>
 <span class="variable">Returns</span> <span class="operator">:</span> <span class="variable">nothing</span><span class="operator">;</span> <span class="variable">tcoffee</span> <span class="variable">output</span> <span class="variable">is</span> <span class="variable">written</span> <span class="variable">to</span> <span class="variable">a</span>
           <span class="variable">temporary</span> <span class="variable">file</span> <span class="variable">OR</span> <span class="variable">specified</span> <span class="variable">output</span> <span class="variable">file</span>
 <span class="variable">Args</span>    <span class="operator">:</span> <span class="variable">Name</span> <span class="variable">of</span> <span class="variable">a</span> <span class="variable">file</span> <span class="variable">containing</span> <span class="variable">a</span> <span class="variable">set</span> <span class="variable">of</span> <span class="variable">unaligned</span> <span class="variable">fasta</span> <span class="variable">sequences</span>
           <span class="keyword">and</span> <span class="variable">hash</span> <span class="variable">of</span> <span class="variable">parameters</span> <span class="variable">to</span> <span class="variable">be</span> <span class="variable">passed</span> <span class="variable">to</span> <span class="variable">tcoffee</span>
</code></code></pre>

<h2 id="setinput">_setinput</h2>

<pre><code><code> Title   :  _setinput
 Usage   :  Internal function, not to be called directly        
 Function:  Create input file for tcoffee program
 Example :
 Returns : name of file containing tcoffee data input AND
           type of file (if known, S for sequence, L for sequence library,
           A for sequence alignment)
 Args    : Seq or Align object reference or input file name</code></code></pre>

<h2 id="setparams">_setparams</h2>

<pre><code><code> Title   :  _setparams
 Usage   :  Internal function, not to be called directly        
 Function:  Create parameter inputs for tcoffee program
 Example :
 Returns : parameter string to be passed to tcoffee
           during align or profile_align
 Args    : name of calling object</code></code></pre>

<h2 id="aformat">aformat</h2>

<pre><code><code> <span class="variable">Title</span>   <span class="operator">:</span> <span class="variable">aformat</span>
 <span class="variable">Usage</span>   <span class="operator">:</span> <span class="keyword">my</span> <span class="variable">$alignmentformat</span> <span class="operator">=</span> <span class="variable">$self</span><span class="operator">-&gt;</span><span class="variable">aformat</span><span class="operator">();</span>
 <span class="variable">Function</span><span class="operator">:</span> <span class="variable">Get</span><span class="operator">/</span><span class="variable">Set</span> <span class="variable">alignment</span> <span class="keyword">format</span>
 <span class="variable">Returns</span> <span class="operator">:</span> <span class="variable">string</span>
 <span class="variable">Args</span>    <span class="operator">:</span> <span class="variable">string</span>
</code></code></pre>

<h2 id="methods">methods</h2>

<pre><code><code> Title   : methods
 Usage   : my @methods = $self-&gt;methods()
 Function: Get/Set Alignment methods - NOT VALIDATED
 Returns : array of strings
 Args    : arrayref of strings</code></code></pre>

<h1 id="Bio::Tools::Run::BaseWrapper-methods">Bio::Tools::Run::BaseWrapper methods</h1>

<h2 id="no_param_checks">no_param_checks</h2>

<pre><code><code> Title   : no_param_checks
 Usage   : $obj-&gt;no_param_checks($newval)
 Function: Boolean flag as to whether or not we should
           trust the sanity checks for parameter values  
 Returns : value of no_param_checks
 Args    : newvalue (optional)</code></code></pre>

<h2 id="save_tempfiles">save_tempfiles</h2>

<pre><code><code> Title   : save_tempfiles
 Usage   : $obj-&gt;save_tempfiles($newval)
 Function: 
 Returns : value of save_tempfiles
 Args    : newvalue (optional)</code></code></pre>

<h2 id="outfile_name">outfile_name</h2>

<pre><code><code> <span class="variable">Title</span>   <span class="operator">:</span> <span class="variable">outfile_name</span>
 <span class="variable">Usage</span>   <span class="operator">:</span> <span class="keyword">my</span> <span class="variable">$outfile</span> <span class="operator">=</span> <span class="variable">$tcoffee</span><span class="operator">-&gt;</span><span class="variable">outfile_name</span><span class="operator">();</span>
 <span class="variable">Function</span><span class="operator">:</span> <span class="variable">Get</span><span class="operator">/</span><span class="variable">Set</span> <span class="variable">the</span> <span class="variable">name</span> <span class="variable">of</span> <span class="variable">the</span> <span class="variable">output</span> <span class="variable">file</span> <span class="keyword">for</span> <span class="variable">this</span> <span class="variable">run</span>
           <span class="operator">(</span><span class="keyword">if</span> <span class="variable">you</span> <span class="variable">wanted</span> <span class="variable">to</span> <span class="keyword">do</span> <span class="variable">something</span> <span class="variable">special</span><span class="operator">)</span>
 <span class="variable">Returns</span> <span class="operator">:</span> <span class="variable">string</span>
 <span class="variable">Args</span>    <span class="operator">:</span> <span class="operator">[</span><span class="variable">optional</span><span class="operator">]</span> <span class="variable">string</span> <span class="variable">to</span> <span class="variable">set</span> <span class="variable">value</span> <span class="variable">to</span>
</code></code></pre>

<h2 id="tempdir">tempdir</h2>

<pre><code><code> <span class="variable">Title</span>   <span class="operator">:</span> <span class="variable">tempdir</span>
 <span class="variable">Usage</span>   <span class="operator">:</span> <span class="keyword">my</span> <span class="variable">$tmpdir</span> <span class="operator">=</span> <span class="variable">$self</span><span class="operator">-&gt;</span><span class="variable">tempdir</span><span class="operator">();</span>
 <span class="variable">Function</span><span class="operator">:</span> <span class="variable">Retrieve</span> <span class="variable">a</span> <span class="variable">temporary</span> <span class="variable">directory</span> <span class="variable">name</span> <span class="operator">(</span><span class="variable">which</span> <span class="variable">is</span> <span class="variable">created</span><span class="operator">)</span>
 <span class="variable">Returns</span> <span class="operator">:</span> <span class="variable">string</span> <span class="variable">which</span> <span class="variable">is</span> <span class="variable">the</span> <span class="variable">name</span> <span class="variable">of</span> <span class="variable">the</span> <span class="variable">temporary</span> <span class="variable">directory</span>
 <span class="variable">Args</span>    <span class="operator">:</span> <span class="variable">none</span>
</code></code></pre>

<h2 id="cleanup">cleanup</h2>

<pre><code><code> <span class="variable">Title</span>   <span class="operator">:</span> <span class="variable">cleanup</span>
 <span class="variable">Usage</span>   <span class="operator">:</span> <span class="variable">$tcoffee</span><span class="operator">-&gt;</span><span class="variable">cleanup</span><span class="operator">();</span>
 <span class="variable">Function</span><span class="operator">:</span> <span class="variable">Will</span> <span class="variable">cleanup</span> <span class="variable">the</span> <span class="variable">tempdir</span> <span class="variable">directory</span>
 <span class="variable">Returns</span> <span class="operator">:</span> <span class="variable">none</span>
 <span class="variable">Args</span>    <span class="operator">:</span> <span class="variable">none</span>
</code></code></pre>

<h2 id="io">io</h2>

<pre><code><code> Title   : io
 Usage   : $obj-&gt;io($newval)
 Function:  Gets a L&lt;Bio::Root::IO&gt; object
 Returns : L&lt;Bio::Root::IO&gt;
 Args    : none</code></code></pre>


</body>

</html>


