<?xml version="1.0" ?>
<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.0 Transitional//EN">
<html>
<head>
<!-- saved from url=(0017)http://localhost/ -->
<script language="JavaScript" src="../../../../../displayToc.js"></script>
<script language="JavaScript" src="../../../../../tocParas.js"></script>
<script language="JavaScript" src="../../../../../tocTab.js"></script>
<link rel="stylesheet" type="text/css" href="../../../../../scineplex.css">
<title></title>
<link rel="stylesheet" href="../../../../../Active.css" type="text/css" />
<meta http-equiv="content-type" content="text/html; charset=utf-8" />
<link rev="made" href="mailto:support@ActiveState.com" />
</head>

<body>



<ul id="index">
  <li><a href="#NAME">NAME</a></li>
  <li><a href="#SYNOPSIS">SYNOPSIS</a></li>
  <li><a href="#DESCRIPTION">DESCRIPTION</a></li>
  <li><a href="#TODO">TODO</a></li>
  <li><a href="#FEEDBACK">FEEDBACK</a>
    <ul>
      <li><a href="#Mailing-Lists">Mailing Lists</a></li>
      <li><a href="#Support">Support</a></li>
      <li><a href="#Reporting-Bugs">Reporting Bugs</a></li>
    </ul>
  </li>
  <li><a href="#AUTHOR---Chris-Mungall">AUTHOR - Chris Mungall</a></li>
  <li><a href="#APPENDIX">APPENDIX</a>
    <ul>
      <li><a href="#new">new</a></li>
      <li><a href="#name_feature">name_feature</a></li>
      <li><a href="#name_contained_features">name_contained_features</a></li>
      <li><a href="#generate_feature_name">generate_feature_name</a></li>
    </ul>
  </li>
</ul>

<h1 id="NAME">NAME</h1>

<p>Bio::SeqFeature::Tools::FeatureNamer - generates unique persistent names for features</p>

<h1 id="SYNOPSIS">SYNOPSIS</h1>

<pre><code><code>  <span class="keyword">use</span> <span class="variable">Bio::SeqIO</span><span class="operator">;</span>
  <span class="keyword">use</span> <span class="variable">Bio::SeqFeature::Tools::FeatureNamer</span><span class="operator">;</span>
  
  <span class="comment"># first fetch a genbank SeqI object</span>
  <span class="variable">$seqio</span> <span class="operator">=</span>
    <span class="variable">Bio::SeqIO</span><span class="operator">-&gt;</span><span class="variable">new</span><span class="operator">(</span><span class="string">-file</span><span class="operator">=&gt;</span><span class="string">'AE003644.gbk'</span><span class="operator">,</span>
                    <span class="string">-format</span><span class="operator">=&gt;</span><span class="string">'GenBank'</span><span class="operator">);</span>
  <span class="variable">$seq</span> <span class="operator">=</span> <span class="variable">$seqio</span><span class="operator">-&gt;</span><span class="variable">next_seq</span><span class="operator">();</span>
  
  <span class="variable">$namer</span> <span class="operator">=</span> <span class="variable">Bio::SeqFeature::Tools::FeatureNamer</span><span class="operator">-&gt;</span><span class="variable">new</span><span class="operator">;</span>
  <span class="keyword">my</span> <span class="variable">@features</span> <span class="operator">=</span> <span class="variable">$seq</span><span class="operator">-&gt;</span><span class="variable">get_SeqFeatures</span><span class="operator">;</span>
  <span class="keyword">foreach</span> <span class="keyword">my</span> <span class="variable">$feature</span> <span class="operator">(</span><span class="variable">@features</span><span class="operator">)</span> <span class="operator">{</span>
    <span class="variable">$namer</span><span class="operator">-&gt;</span><span class="variable">name_feature</span><span class="operator">(</span><span class="variable">$feature</span><span class="operator">)</span> <span class="keyword">unless</span> <span class="variable">$feature</span><span class="operator">-&gt;</span><span class="variable">display_name</span><span class="operator">;</span>
  <span class="operator">}</span>  
</code></code></pre>

<h1 id="DESCRIPTION">DESCRIPTION</h1>

<p>This is a helper class for providing names for SeqFeatures</p>

<p>The <a href="../../../../../../.cpan/build/BioPerl-1.6.901-ghKY8U/blib/lib/Bio/SeqFeatureI.html">Bio::SeqFeatureI</a> class provides a display_name method. Typically the display_name is not set when parsing formats such as genbank - instead properties such as <b>label</b>, <b>product</b> or <b>gene</b> are set in a somewhat inconsistent manner.</p>

<p>In addition, when generating subfeatures (for example, exons that are subfeatures of a transcript feature), it is often desirable to name these subfeatures before either exporting to another format or reporting to the user.</p>

<p>This module is intended to help given uniform display_names to features and their subfeatures.</p>

<h1 id="TODO">TODO</h1>

<p>Currently the naming policy is hardcoded. It may be desirable to allow plugging in variations on naming policies; this could be done either by subclassing, anonymous subroutines (closures) or parameterization. Contact the author if you feel you have need for a different naming policy</p>

<h1 id="FEEDBACK">FEEDBACK</h1>

<h2 id="Mailing-Lists">Mailing Lists</h2>

<p>User feedback is an integral part of the evolution of this and other Bioperl modules. Send your comments and suggestions preferably to the Bioperl mailing lists Your participation is much appreciated.</p>

<pre><code><code>  bioperl-l@bioperl.org                  - General discussion
  http://bioperl.org/wiki/Mailing_lists  - About the mailing lists</code></code></pre>

<h2 id="Support">Support</h2>

<p>Please direct usage questions or support issues to the mailing list:</p>

<p><i>bioperl-l@bioperl.org</i></p>

<p>rather than to the module maintainer directly. Many experienced and reponsive experts will be able look at the problem and quickly address it. Please include a thorough description of the problem with code and data examples if at all possible.</p>

<h2 id="Reporting-Bugs">Reporting Bugs</h2>

<p>report bugs to the Bioperl bug tracking system to help us keep track the bugs and their resolution. Bug reports can be submitted via the web:</p>

<pre><code><code>  https://redmine.open-bio.org/projects/bioperl/</code></code></pre>

<h1 id="AUTHOR---Chris-Mungall">AUTHOR - Chris Mungall</h1>

<p>Email: cjm AT fruitfly DOT org</p>

<h1 id="APPENDIX">APPENDIX</h1>

<p>The rest of the documentation details each of the object methods. Internal methods are usually preceded with a _</p>

<h2 id="new">new</h2>

<pre><code><code> <span class="variable">Title</span>   <span class="operator">:</span> <span class="variable">new</span>
 <span class="variable">Usage</span>   <span class="operator">:</span> <span class="variable">$unflattener</span> <span class="operator">=</span> <span class="variable">Bio::SeqFeature::Tools::FeatureNamer</span><span class="operator">-&gt;</span><span class="variable">new</span><span class="operator">();</span>
 <span class="variable">Function</span><span class="operator">:</span> <span class="variable">constructor</span>
 <span class="variable">Example</span> <span class="operator">:</span> 
 <span class="variable">Returns</span> <span class="operator">:</span> <span class="variable">a</span> <span class="variable">new</span> <span class="variable">Bio::SeqFeature::Tools::FeatureNamer</span>
 <span class="variable">Args</span>    <span class="operator">:</span> <span class="variable">see</span> <span class="variable">below</span>
</code></code></pre>

<h2 id="name_feature">name_feature</h2>

<pre><code><code> <span class="variable">Title</span>   <span class="operator">:</span> <span class="variable">name_feature</span>
 <span class="variable">Usage</span>   <span class="operator">:</span> <span class="variable">$namer</span><span class="operator">-&gt;</span><span class="variable">name_feature</span><span class="operator">(</span><span class="variable">$sf</span><span class="operator">);</span>
 <span class="variable">Function</span><span class="operator">:</span> <span class="variable">sets</span> <span class="variable">display_name</span>
 <span class="variable">Example</span> <span class="operator">:</span>
 <span class="variable">Returns</span> <span class="operator">:</span> 
 <span class="variable">Args</span>    <span class="operator">:</span> <span class="variable">L</span><span class="operator">&lt;</span><span class="variable">Bio::SeqFeatureI</span><span class="operator">&gt;</span>
</code></code></pre>

<p>This method calls generate_feature_name() and uses the returned value to set the display_name of the feature</p>

<h2 id="name_contained_features">name_contained_features</h2>

<pre><code><code> <span class="variable">Title</span>   <span class="operator">:</span> <span class="variable">name_contained_features</span>
 <span class="variable">Usage</span>   <span class="operator">:</span> <span class="variable">$namer</span><span class="operator">-&gt;</span><span class="variable">name_contained_features</span><span class="operator">(</span><span class="variable">$sf</span><span class="operator">);</span>
 <span class="variable">Function</span><span class="operator">:</span> <span class="variable">sets</span> <span class="variable">display_name</span> <span class="keyword">for</span> <span class="variable">all</span> <span class="variable">features</span> <span class="variable">contained</span> <span class="variable">by</span> <span class="variable">sf</span>
 <span class="variable">Example</span> <span class="operator">:</span>
 <span class="variable">Returns</span> <span class="operator">:</span> 
 <span class="variable">Args</span>    <span class="operator">:</span> <span class="variable">L</span><span class="operator">&lt;</span><span class="variable">Bio::SeqFeatureI</span><span class="operator">&gt;</span>
</code></code></pre>

<p>iterates through all subfeatures of a certain feature (using get_all_SeqFeatures) and names each subfeatures, based on the generated name for the holder feature</p>

<p>A subfeature is named by concatenating the generated name of the container feature with the type and a number.</p>

<p>For example, if the containing feature is a gene with display name <b>dpp</b>, subfeatures will be named dpp-mRNA-1 dpp-mRNA2 dpp-exon1 dpp-exon2 etc</p>

<h2 id="generate_feature_name">generate_feature_name</h2>

<pre><code><code> <span class="variable">Title</span>   <span class="operator">:</span> <span class="variable">generate_feature_name</span>
 <span class="variable">Usage</span>   <span class="operator">:</span> <span class="variable">$name</span> <span class="operator">=</span> <span class="variable">$namer</span><span class="operator">-&gt;</span><span class="variable">generate_feature_name</span><span class="operator">(</span><span class="variable">$sf</span><span class="operator">);</span>
 <span class="variable">Function</span><span class="operator">:</span> <span class="variable">derives</span> <span class="variable">a</span> <span class="variable">sensible</span> <span class="variable">human</span> <span class="variable">readable</span> <span class="variable">name</span> <span class="keyword">for</span> <span class="variable">a</span> <span class="variable">$sf</span>
 <span class="variable">Example</span> <span class="operator">:</span>
 <span class="variable">Returns</span> <span class="operator">:</span> <span class="variable">str</span>
 <span class="variable">Args</span>    <span class="operator">:</span> <span class="variable">L</span><span class="operator">&lt;</span><span class="variable">Bio::SeqFeatureI</span><span class="operator">&gt;</span>
</code></code></pre>

<p>returns a generated name (but does not actually set display_name).</p>

<p>If display_name is already set, the method will return this</p>

<p>Otherwise, the name will depend on the property:</p>

<dl>

<dt id="label">label</dt>
<dd>

</dd>
<dt id="product">product</dt>
<dd>

</dd>
<dt id="gene">gene</dt>
<dd>

</dd>
<dt id="locus_tag">locus_tag</dt>
<dd>

</dd>
</dl>

<p>(in order of priority)</p>


</body>

</html>


